{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad84acda",
   "metadata": {},
   "source": [
    "This tutorial will demonstrate how to pre-process single-cell raw UMI counts to generate expression matrices that can be used as input to Tensor-cell2cell. \n",
    "\n",
    "At some point in the pipeline, we must account for batch. Batch-correction is important since Tensor-cell2cell considers multiple balf_samples to extract context-dependent patterns, and we want to make sure we are capturing true biological signals rather than sample-specific differences due to technical variability. \n",
    "\n",
    "Ideally, we can use single-cell RNAseq batch correction methods. There are a few potential problems with this approach:\n",
    "\n",
    "1) Batch correction methods often return a matrix in a reduced space and thus does not have the original gene features included, which is needed for LR scoring.\n",
    "\n",
    "2) Some cell-cell communication tools expect data in other formats, such as log(1+CPM)\n",
    "\n",
    "3) Batch correction methods that do return gene counts often return negative counts which can result in negative LR scores. Negative values in the tensor can bias non-negative TCD, the main algorithm used in Tensor-cell2cell.  \n",
    "\n",
    "In this tutorial, and its companion 01B for R users, we will show pre-processing from raw counts to batch corrected counts. Problem 1 can simply be dealt with by only using batch correction methods that return the original gene features. Problem 2-3 will be discussed further in Tutorials XXX. Essentially, Problems 2-3 can both be dealth with by instead directly introducing a technical covariate to account for batch directly to the decomposition. Problem 3 can also be dealt with either by masking negative values or using a TCD approach that does not have a non-negative constraint. \n",
    "\n",
    "There are a number of workflows to achieve this; here we demonstrate a typical workflow using the popular single-cell analysis software scanpy to generate an AnnData object which can be used downstream."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dbff2e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "seed = 888"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b704b19",
   "metadata": {},
   "source": [
    "Load the BALF COVID dataset, which is described here: https://doi.org/10.1038/s41591-020-0901-9.\n",
    "\n",
    "This dataset contains 12 samples, each associated with \"Healthy Control\", \"Moderate\", or \"Severe\" COVID contexts. \n",
    "\n",
    "You can download the 12 scRNAseq .h5 files under the samples section here: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926\n",
    "You can also download the metadata file: TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5d757698",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data_path = '/data2/eric/Tensor-Revisions/data/COVID-19/'\n",
    "from cell2cell.datasets.load_data import load_balf\n",
    "\n",
    "balf_samples = c2c.data.load_balf(expression_data_path)\n",
    "\n",
    "# expression_data_path = '/data2/eric/Tensor-Revisions/data/COVID-19/'\n",
    "# balf_samples = load_balf(expression_data_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "761e41d9",
   "metadata": {},
   "source": [
    "balf_samples\n",
    "is a dictionary with keys as the sample name and values as a scanpy object with relevant metadata and raw UMI counts."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8eb15f8",
   "metadata": {},
   "source": [
    "We begin with a basic preprocessing of each sample based on QC metrics. See https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html#Preprocessing for details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3a511c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample, adata_ in balf_samples.items():\n",
    "    adata = adata_.copy()\n",
    "    sc.pp.filter_cells(adata, min_genes=30)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "    # filter cells based on QC metrics\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    adata = adata[adata.obs.n_genes_by_counts < 6000, :]\n",
    "    adata = adata[adata.obs.pct_counts_mt < 10, :]\n",
    "    \n",
    "    balf_samples[sample] = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32c99290",
   "metadata": {},
   "source": [
    "Next, we normalize the raw UMI counts. We recommend log(1+CPM) normalization, as this maintains non-negative counts and is the input for many communication scoring functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "daa05c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample, adata_ in balf_samples.items():\n",
    "    adata = adata_.copy()\n",
    "    # CPM normalize\n",
    "    sc.pp.normalize_total(adata, target_sum=1e6)\n",
    "\n",
    "    # logarithmize \n",
    "    sc.pp.log1p(adata)\n",
    "    \n",
    "    balf_samples[sample] = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e3ccc31",
   "metadata": {},
   "source": [
    "Finally, we apply a batch correction. The goal here is to account for sample-to-sample technical variability. In this case, we show Combat since it is built in with scanpy. \n",
    "\n",
    "Note, the final input matrices to Tensor-cell2cell must be non-negative. We will demonstrate workarounds to negative counts in the tensor building tutorial. \n",
    "\n",
    "See 10.1186/s13619-020-00041-9 for a benchmarking of Scanpy's batch correction methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3cd6a78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_var = 'Sample_ID' # the batch variable in the metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a84d4ba",
   "metadata": {},
   "source": [
    "Batch correction using combat:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "22729c4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge the balf_samples\n",
    "balf_corrected = sc.concat(balf_samples.values())\n",
    "balf_corrected.obs_names_make_unique()\n",
    "\n",
    "# store log(1+CPM) values in \"raw\" attribute\n",
    "balf_corrected.raw = balf_corrected \n",
    "\n",
    "# do the batch correction\n",
    "sc.pp.combat(balf_corrected, key = batch_var) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "454e3110",
   "metadata": {},
   "source": [
    "The next two cells, unused, show examples of other methods for batch correction . See https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/scanpy/scanpy_03_integration.html for more tutorials on batch correction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78bd0c4c",
   "metadata": {},
   "source": [
    "Batch correction with scanorama:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7cb66089",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import scanorama\n",
    "\n",
    "# # merge all the balf_samples into a single object\n",
    "# balf_log = sc.concat(balf_samples.values())\n",
    "# balf_log.obs_names_make_unique()\n",
    "\n",
    "# # correct with scanorama\n",
    "# balf_corrected = scanorama.correct_scanpy(adatas=list(balf_samples.values()), return_dimred=False)\n",
    "\n",
    "# # aggregate into one object\n",
    "# balf_corrected = sc.concat(balf_corrected) \n",
    "# balf_corrected.obs_names_make_unique()\n",
    "\n",
    "# # store log(1+CPM) values in \"raw\" attribute\n",
    "# balf_corrected.raw = balf_log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b2d77e9",
   "metadata": {},
   "source": [
    "Batch correction using a simple linear regression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "06f7be4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # merge the balf_samples\n",
    "# balf_corrected = sc.concat(balf_samples.values())\n",
    "# balf_corrected.obs_names_make_unique()\n",
    "\n",
    "# # store log(1+CPM) values in \"raw\" attribute\n",
    "# balf_corrected.raw = balf_corrected\n",
    "\n",
    "# # do the batch correction\n",
    "# sc.pp.regress_out(balf_corrected, keys = batch_var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29a53fb8",
   "metadata": {},
   "source": [
    "Calculate a PCA manifold on the batch-corrected counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3a44fc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the top 2000 highly variable genes\n",
    "sc.pp.highly_variable_genes(balf_corrected, n_top_genes = 2000)\n",
    "\n",
    "# get PCA to 100 PCs\n",
    "sc.tl.pca(balf_corrected, use_highly_variable = True, svd_solver='arpack', random_state = seed, \n",
    "         n_comps = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd88998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: make this corrected object, the raw data, and metadata available to download somewhere\n",
    "out_path = '/data3/hratch/c2c_general/'\n",
    "balf_corrected.write_h5ad(out_path + 'batch_corrected_balf_covid.h5ad') # 6.7Gb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f03121ab",
   "metadata": {},
   "source": [
    "The final \"balf_corrected\" AnnData object has the following attributes:\n",
    "1) X: batch-correct counts matrix (preferably non-negative) <br>\n",
    "2) obs: cell metadata that includes the cell group (cluster or type), Sample ID, and Context <br>\n",
    "3) raw: log(1+CPM) normalized AnnData object <br>\n",
    "4) obsm['X_pca']: the cell manifold \n",
    "\n",
    "Regardless of the preprocessing pipeline used, these four pieces of information will be necessary for some parts of the Tensor-cell2cell analyses. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "393eb5de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGGAATCGC-1</th>\n",
       "      <th>AAACCTGTCCAGAAGG-1</th>\n",
       "      <th>AAACCTGTCCAGTAGT-1</th>\n",
       "      <th>AAACCTGTCTGGGCCA-1</th>\n",
       "      <th>AAACGGGCACGAGGTA-1</th>\n",
       "      <th>AAACGGGGTACATCCA-1</th>\n",
       "      <th>AAACGGGGTCTCCCTA-1</th>\n",
       "      <th>AAACGGGTCTAGAGTC-1</th>\n",
       "      <th>AAAGATGTCGTGGGAA-1</th>\n",
       "      <th>AAAGCAAAGGGATACC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGGTTAGCACGCCT-1</th>\n",
       "      <th>TTTGGTTAGTGGTAAT-1</th>\n",
       "      <th>TTTGGTTAGTTGTAGA-1-1</th>\n",
       "      <th>TTTGGTTCATACTACG-1</th>\n",
       "      <th>TTTGTCAAGATTACCC-1</th>\n",
       "      <th>TTTGTCAAGTGGTAAT-1</th>\n",
       "      <th>TTTGTCACAGAAGCAC-1</th>\n",
       "      <th>TTTGTCATCAACCAAC-1</th>\n",
       "      <th>TTTGTCATCCAAACAC-1</th>\n",
       "      <th>TTTGTCATCGCGTTTC-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>-0.026211</td>\n",
       "      <td>3.884047</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>4.615011</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>-0.026211</td>\n",
       "      <td>4.423844</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "      <td>-0.031741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>-0.276392</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>2.241069</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "      <td>-0.434179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>-0.045777</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "      <td>-0.019993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHN1</th>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>-0.061546</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>-0.444072</td>\n",
       "      <td>4.559903</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>-0.444072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>0.562998</td>\n",
       "      <td>6.038211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 63102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAACCTGAGGAATCGC-1  AAACCTGTCCAGAAGG-1  AAACCTGTCCAGTAGT-1  \\\n",
       "LINC00115           -0.026211            3.884047           -0.026211   \n",
       "NOC2L               -0.276392           -0.276392           -0.276392   \n",
       "KLHL17              -0.045777           -0.045777           -0.045777   \n",
       "PLEKHN1             -0.061546           -0.061546           -0.061546   \n",
       "HES4                -0.444072            4.559903           -0.444072   \n",
       "\n",
       "           AAACCTGTCTGGGCCA-1  AAACGGGCACGAGGTA-1  AAACGGGGTACATCCA-1  \\\n",
       "LINC00115           -0.026211            4.615011           -0.026211   \n",
       "NOC2L               -0.276392           -0.276392           -0.276392   \n",
       "KLHL17              -0.045777           -0.045777           -0.045777   \n",
       "PLEKHN1             -0.061546           -0.061546           -0.061546   \n",
       "HES4                -0.444072           -0.444072           -0.444072   \n",
       "\n",
       "           AAACGGGGTCTCCCTA-1  AAACGGGTCTAGAGTC-1  AAAGATGTCGTGGGAA-1  \\\n",
       "LINC00115           -0.026211           -0.026211           -0.026211   \n",
       "NOC2L               -0.276392           -0.276392           -0.276392   \n",
       "KLHL17              -0.045777           -0.045777           -0.045777   \n",
       "PLEKHN1             -0.061546           -0.061546           -0.061546   \n",
       "HES4                -0.444072           -0.444072           -0.444072   \n",
       "\n",
       "           AAAGCAAAGGGATACC-1  ...  TTTGGTTAGCACGCCT-1  TTTGGTTAGTGGTAAT-1  \\\n",
       "LINC00115            4.423844  ...           -0.031741           -0.031741   \n",
       "NOC2L               -0.276392  ...           -0.434179           -0.434179   \n",
       "KLHL17              -0.045777  ...           -0.019993           -0.019993   \n",
       "PLEKHN1             -0.061546  ...            0.002228            0.002228   \n",
       "HES4                -0.444072  ...            0.562998            0.562998   \n",
       "\n",
       "           TTTGGTTAGTTGTAGA-1-1  TTTGGTTCATACTACG-1  TTTGTCAAGATTACCC-1  \\\n",
       "LINC00115             -0.031741           -0.031741           -0.031741   \n",
       "NOC2L                  2.241069           -0.434179           -0.434179   \n",
       "KLHL17                -0.019993           -0.019993           -0.019993   \n",
       "PLEKHN1                0.002228            0.002228            0.002228   \n",
       "HES4                   0.562998            0.562998            0.562998   \n",
       "\n",
       "           TTTGTCAAGTGGTAAT-1  TTTGTCACAGAAGCAC-1  TTTGTCATCAACCAAC-1  \\\n",
       "LINC00115           -0.031741           -0.031741           -0.031741   \n",
       "NOC2L               -0.434179           -0.434179           -0.434179   \n",
       "KLHL17              -0.019993           -0.019993           -0.019993   \n",
       "PLEKHN1              0.002228            0.002228            0.002228   \n",
       "HES4                 0.562998            0.562998            0.562998   \n",
       "\n",
       "           TTTGTCATCCAAACAC-1  TTTGTCATCGCGTTTC-1  \n",
       "LINC00115           -0.031741           -0.031741  \n",
       "NOC2L               -0.434179           -0.434179  \n",
       "KLHL17              -0.019993           -0.019993  \n",
       "PLEKHN1              0.002228            0.002228  \n",
       "HES4                 0.562998            6.038211  \n",
       "\n",
       "[5 rows x 63102 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# corrected counts matrix\n",
    "balf_corrected.to_df().T.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e79d443a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Context</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGAATCGC-1</th>\n",
       "      <td>C148</td>\n",
       "      <td>Severe_Covid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>606</td>\n",
       "      <td>606</td>\n",
       "      <td>1342.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>4.470939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGAAGG-1</th>\n",
       "      <td>C148</td>\n",
       "      <td>Severe_Covid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>2035</td>\n",
       "      <td>2034</td>\n",
       "      <td>7297.0</td>\n",
       "      <td>334.0</td>\n",
       "      <td>4.577224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGTAGT-1</th>\n",
       "      <td>C148</td>\n",
       "      <td>Severe_Covid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>1660</td>\n",
       "      <td>1658</td>\n",
       "      <td>4959.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>6.533575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCTGGGCCA-1</th>\n",
       "      <td>C148</td>\n",
       "      <td>Severe_Covid</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>4965</td>\n",
       "      <td>4964</td>\n",
       "      <td>31956.0</td>\n",
       "      <td>1374.0</td>\n",
       "      <td>4.299662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGCACGAGGTA-1</th>\n",
       "      <td>C148</td>\n",
       "      <td>Severe_Covid</td>\n",
       "      <td>T</td>\n",
       "      <td>1290</td>\n",
       "      <td>1288</td>\n",
       "      <td>2892.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>4.114799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Sample_ID       Context    cell_type  n_genes  \\\n",
       "AAACCTGAGGAATCGC-1      C148  Severe_Covid  Macrophages      606   \n",
       "AAACCTGTCCAGAAGG-1      C148  Severe_Covid  Macrophages     2035   \n",
       "AAACCTGTCCAGTAGT-1      C148  Severe_Covid  Macrophages     1660   \n",
       "AAACCTGTCTGGGCCA-1      C148  Severe_Covid  Macrophages     4965   \n",
       "AAACGGGCACGAGGTA-1      C148  Severe_Covid            T     1290   \n",
       "\n",
       "                    n_genes_by_counts  total_counts  total_counts_mt  \\\n",
       "AAACCTGAGGAATCGC-1                606        1342.0             60.0   \n",
       "AAACCTGTCCAGAAGG-1               2034        7297.0            334.0   \n",
       "AAACCTGTCCAGTAGT-1               1658        4959.0            324.0   \n",
       "AAACCTGTCTGGGCCA-1               4964       31956.0           1374.0   \n",
       "AAACGGGCACGAGGTA-1               1288        2892.0            119.0   \n",
       "\n",
       "                    pct_counts_mt  \n",
       "AAACCTGAGGAATCGC-1       4.470939  \n",
       "AAACCTGTCCAGAAGG-1       4.577224  \n",
       "AAACCTGTCCAGTAGT-1       6.533575  \n",
       "AAACCTGTCTGGGCCA-1       4.299662  \n",
       "AAACGGGCACGAGGTA-1       4.114799  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cell metadata\n",
    "balf_corrected.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "399e7f0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGGAATCGC-1</th>\n",
       "      <th>AAACCTGTCCAGAAGG-1</th>\n",
       "      <th>AAACCTGTCCAGTAGT-1</th>\n",
       "      <th>AAACCTGTCTGGGCCA-1</th>\n",
       "      <th>AAACGGGCACGAGGTA-1</th>\n",
       "      <th>AAACGGGGTACATCCA-1</th>\n",
       "      <th>AAACGGGGTCTCCCTA-1</th>\n",
       "      <th>AAACGGGTCTAGAGTC-1</th>\n",
       "      <th>AAAGATGTCGTGGGAA-1</th>\n",
       "      <th>AAAGCAAAGGGATACC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGGTTAGCACGCCT-1</th>\n",
       "      <th>TTTGGTTAGTGGTAAT-1</th>\n",
       "      <th>TTTGGTTAGTTGTAGA-1-1</th>\n",
       "      <th>TTTGGTTCATACTACG-1</th>\n",
       "      <th>TTTGTCAAGATTACCC-1</th>\n",
       "      <th>TTTGTCAAGTGGTAAT-1</th>\n",
       "      <th>TTTGTCACAGAAGCAC-1</th>\n",
       "      <th>TTTGTCATCAACCAAC-1</th>\n",
       "      <th>TTTGTCATCCAAACAC-1</th>\n",
       "      <th>TTTGTCATCGCGTTTC-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.927562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.848695</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.607794</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.458022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHN1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.021334</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.693111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 63102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAACCTGAGGAATCGC-1  AAACCTGTCCAGAAGG-1  AAACCTGTCCAGTAGT-1  \\\n",
       "LINC00115                 0.0            4.927562                 0.0   \n",
       "NOC2L                     0.0            0.000000                 0.0   \n",
       "KLHL17                    0.0            0.000000                 0.0   \n",
       "PLEKHN1                   0.0            0.000000                 0.0   \n",
       "HES4                      0.0            6.021334                 0.0   \n",
       "\n",
       "           AAACCTGTCTGGGCCA-1  AAACGGGCACGAGGTA-1  AAACGGGGTACATCCA-1  \\\n",
       "LINC00115                 0.0            5.848695                 0.0   \n",
       "NOC2L                     0.0            0.000000                 0.0   \n",
       "KLHL17                    0.0            0.000000                 0.0   \n",
       "PLEKHN1                   0.0            0.000000                 0.0   \n",
       "HES4                      0.0            0.000000                 0.0   \n",
       "\n",
       "           AAACGGGGTCTCCCTA-1  AAACGGGTCTAGAGTC-1  AAAGATGTCGTGGGAA-1  \\\n",
       "LINC00115                 0.0                 0.0                 0.0   \n",
       "NOC2L                     0.0                 0.0                 0.0   \n",
       "KLHL17                    0.0                 0.0                 0.0   \n",
       "PLEKHN1                   0.0                 0.0                 0.0   \n",
       "HES4                      0.0                 0.0                 0.0   \n",
       "\n",
       "           AAAGCAAAGGGATACC-1  ...  TTTGGTTAGCACGCCT-1  TTTGGTTAGTGGTAAT-1  \\\n",
       "LINC00115            5.607794  ...                 0.0                 0.0   \n",
       "NOC2L                0.000000  ...                 0.0                 0.0   \n",
       "KLHL17               0.000000  ...                 0.0                 0.0   \n",
       "PLEKHN1              0.000000  ...                 0.0                 0.0   \n",
       "HES4                 0.000000  ...                 0.0                 0.0   \n",
       "\n",
       "           TTTGGTTAGTTGTAGA-1-1  TTTGGTTCATACTACG-1  TTTGTCAAGATTACCC-1  \\\n",
       "LINC00115              0.000000                 0.0                 0.0   \n",
       "NOC2L                  3.458022                 0.0                 0.0   \n",
       "KLHL17                 0.000000                 0.0                 0.0   \n",
       "PLEKHN1                0.000000                 0.0                 0.0   \n",
       "HES4                   0.000000                 0.0                 0.0   \n",
       "\n",
       "           TTTGTCAAGTGGTAAT-1  TTTGTCACAGAAGCAC-1  TTTGTCATCAACCAAC-1  \\\n",
       "LINC00115                 0.0                 0.0                 0.0   \n",
       "NOC2L                     0.0                 0.0                 0.0   \n",
       "KLHL17                    0.0                 0.0                 0.0   \n",
       "PLEKHN1                   0.0                 0.0                 0.0   \n",
       "HES4                      0.0                 0.0                 0.0   \n",
       "\n",
       "           TTTGTCATCCAAACAC-1  TTTGTCATCGCGTTTC-1  \n",
       "LINC00115                 0.0            0.000000  \n",
       "NOC2L                     0.0            0.000000  \n",
       "KLHL17                    0.0            0.000000  \n",
       "PLEKHN1                   0.0            0.000000  \n",
       "HES4                      0.0            3.693111  \n",
       "\n",
       "[5 rows x 63102 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# log(1+CPM) counts matrix\n",
    "balf_corrected.raw.to_adata().to_df().T.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4f27239e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>...</th>\n",
       "      <th>PC91</th>\n",
       "      <th>PC92</th>\n",
       "      <th>PC93</th>\n",
       "      <th>PC94</th>\n",
       "      <th>PC95</th>\n",
       "      <th>PC96</th>\n",
       "      <th>PC97</th>\n",
       "      <th>PC98</th>\n",
       "      <th>PC99</th>\n",
       "      <th>PC100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGAATCGC-1</th>\n",
       "      <td>-12.279510</td>\n",
       "      <td>-11.569539</td>\n",
       "      <td>2.333932</td>\n",
       "      <td>-0.792567</td>\n",
       "      <td>-2.818772</td>\n",
       "      <td>2.152842</td>\n",
       "      <td>1.414543</td>\n",
       "      <td>-2.819450</td>\n",
       "      <td>-0.843046</td>\n",
       "      <td>-0.359107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.636127</td>\n",
       "      <td>-0.555673</td>\n",
       "      <td>-1.174195</td>\n",
       "      <td>0.382731</td>\n",
       "      <td>-1.154741</td>\n",
       "      <td>1.880821</td>\n",
       "      <td>-2.175864</td>\n",
       "      <td>0.964448</td>\n",
       "      <td>2.638466</td>\n",
       "      <td>2.882710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGAAGG-1</th>\n",
       "      <td>4.993004</td>\n",
       "      <td>-9.634439</td>\n",
       "      <td>-4.017168</td>\n",
       "      <td>6.867777</td>\n",
       "      <td>-3.659773</td>\n",
       "      <td>2.181219</td>\n",
       "      <td>2.381791</td>\n",
       "      <td>5.002475</td>\n",
       "      <td>4.327793</td>\n",
       "      <td>-2.199009</td>\n",
       "      <td>...</td>\n",
       "      <td>1.056841</td>\n",
       "      <td>-0.549429</td>\n",
       "      <td>-1.792231</td>\n",
       "      <td>-1.502628</td>\n",
       "      <td>3.608827</td>\n",
       "      <td>0.246818</td>\n",
       "      <td>2.478500</td>\n",
       "      <td>-0.412751</td>\n",
       "      <td>2.093467</td>\n",
       "      <td>-2.841379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGTAGT-1</th>\n",
       "      <td>-4.327682</td>\n",
       "      <td>-9.474691</td>\n",
       "      <td>-0.373740</td>\n",
       "      <td>-1.574611</td>\n",
       "      <td>1.026707</td>\n",
       "      <td>0.264881</td>\n",
       "      <td>-0.642445</td>\n",
       "      <td>2.645939</td>\n",
       "      <td>0.768080</td>\n",
       "      <td>1.401344</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.219072</td>\n",
       "      <td>-1.509040</td>\n",
       "      <td>0.535264</td>\n",
       "      <td>-2.544001</td>\n",
       "      <td>-2.009928</td>\n",
       "      <td>0.517890</td>\n",
       "      <td>-0.087034</td>\n",
       "      <td>0.123314</td>\n",
       "      <td>0.314210</td>\n",
       "      <td>-3.137026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCTGGGCCA-1</th>\n",
       "      <td>21.996510</td>\n",
       "      <td>2.566066</td>\n",
       "      <td>-3.944390</td>\n",
       "      <td>-13.916259</td>\n",
       "      <td>7.445151</td>\n",
       "      <td>3.117134</td>\n",
       "      <td>1.296663</td>\n",
       "      <td>2.814822</td>\n",
       "      <td>-1.560574</td>\n",
       "      <td>-0.914551</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.616594</td>\n",
       "      <td>-0.872157</td>\n",
       "      <td>-2.061459</td>\n",
       "      <td>-0.744164</td>\n",
       "      <td>-1.183877</td>\n",
       "      <td>-4.804202</td>\n",
       "      <td>-1.866552</td>\n",
       "      <td>2.740477</td>\n",
       "      <td>1.292481</td>\n",
       "      <td>-1.160910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGCACGAGGTA-1</th>\n",
       "      <td>-18.817038</td>\n",
       "      <td>4.994419</td>\n",
       "      <td>-9.384811</td>\n",
       "      <td>-6.844615</td>\n",
       "      <td>-3.056098</td>\n",
       "      <td>2.774960</td>\n",
       "      <td>-5.017649</td>\n",
       "      <td>0.400225</td>\n",
       "      <td>0.813033</td>\n",
       "      <td>0.663741</td>\n",
       "      <td>...</td>\n",
       "      <td>1.176409</td>\n",
       "      <td>0.047743</td>\n",
       "      <td>-0.064789</td>\n",
       "      <td>2.626716</td>\n",
       "      <td>0.151935</td>\n",
       "      <td>0.872773</td>\n",
       "      <td>-1.538529</td>\n",
       "      <td>-0.425076</td>\n",
       "      <td>1.315260</td>\n",
       "      <td>-0.842731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          PC1        PC2       PC3        PC4       PC5  \\\n",
       "AAACCTGAGGAATCGC-1 -12.279510 -11.569539  2.333932  -0.792567 -2.818772   \n",
       "AAACCTGTCCAGAAGG-1   4.993004  -9.634439 -4.017168   6.867777 -3.659773   \n",
       "AAACCTGTCCAGTAGT-1  -4.327682  -9.474691 -0.373740  -1.574611  1.026707   \n",
       "AAACCTGTCTGGGCCA-1  21.996510   2.566066 -3.944390 -13.916259  7.445151   \n",
       "AAACGGGCACGAGGTA-1 -18.817038   4.994419 -9.384811  -6.844615 -3.056098   \n",
       "\n",
       "                         PC6       PC7       PC8       PC9      PC10  ...  \\\n",
       "AAACCTGAGGAATCGC-1  2.152842  1.414543 -2.819450 -0.843046 -0.359107  ...   \n",
       "AAACCTGTCCAGAAGG-1  2.181219  2.381791  5.002475  4.327793 -2.199009  ...   \n",
       "AAACCTGTCCAGTAGT-1  0.264881 -0.642445  2.645939  0.768080  1.401344  ...   \n",
       "AAACCTGTCTGGGCCA-1  3.117134  1.296663  2.814822 -1.560574 -0.914551  ...   \n",
       "AAACGGGCACGAGGTA-1  2.774960 -5.017649  0.400225  0.813033  0.663741  ...   \n",
       "\n",
       "                        PC91      PC92      PC93      PC94      PC95  \\\n",
       "AAACCTGAGGAATCGC-1  0.636127 -0.555673 -1.174195  0.382731 -1.154741   \n",
       "AAACCTGTCCAGAAGG-1  1.056841 -0.549429 -1.792231 -1.502628  3.608827   \n",
       "AAACCTGTCCAGTAGT-1 -1.219072 -1.509040  0.535264 -2.544001 -2.009928   \n",
       "AAACCTGTCTGGGCCA-1 -1.616594 -0.872157 -2.061459 -0.744164 -1.183877   \n",
       "AAACGGGCACGAGGTA-1  1.176409  0.047743 -0.064789  2.626716  0.151935   \n",
       "\n",
       "                        PC96      PC97      PC98      PC99     PC100  \n",
       "AAACCTGAGGAATCGC-1  1.880821 -2.175864  0.964448  2.638466  2.882710  \n",
       "AAACCTGTCCAGAAGG-1  0.246818  2.478500 -0.412751  2.093467 -2.841379  \n",
       "AAACCTGTCCAGTAGT-1  0.517890 -0.087034  0.123314  0.314210 -3.137026  \n",
       "AAACCTGTCTGGGCCA-1 -4.804202 -1.866552  2.740477  1.292481 -1.160910  \n",
       "AAACGGGCACGAGGTA-1  0.872773 -1.538529 -0.425076  1.315260 -0.842731  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cell manifold\n",
    "pd.DataFrame(balf_corrected.obsm['X_pca'], \n",
    "            columns = ['PC' + str(i) for i in range(1, 101)], \n",
    "                      index = balf_corrected.obs.index.tolist()).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a2d139",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from typing import Dict\n",
    "# def split_adata(adata, sample_col = 'Sample_ID'):\n",
    "#     \"\"\"Split an AnnData object with corrected counts into its respective balf_samples.\n",
    "\n",
    "#     Parameters\n",
    "#     ----------\n",
    "#     adata : AnnData\n",
    "#         merged AnnData object across balf_samples (see sc.concat)\n",
    "#     sample_col : str, optional\n",
    "#         the metadata (adata.obs) column specifying the balf_samples, by default 'Sample_ID'\n",
    "\n",
    "#     Returns\n",
    "#     -------\n",
    "#     balf_samples : Dict[str, AnnData]\n",
    "#         the set of AnnData objects corresponding to each sample\n",
    "#     \"\"\"\n",
    "    \n",
    "#     balf_samples = {sample: adata[adata.obs[adata.obs[sample_col] == sample].index] for sample in adata.obs[sample_col].unique()}\n",
    "#     return balf_samples\n",
    "\n",
    "\n",
    "# balf_corrected_split = split_adata(adata=balf_corrected)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:FDA_collab] *",
   "language": "python",
   "name": "conda-env-FDA_collab-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
