<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>API Documentation - cell2cell</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "API Documentation";
        var mkdocs_page_input_path = "documentation.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> cell2cell
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">API Documentation</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell">cell2cell</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.analysis">analysis</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.analysis.cell2cell_pipelines">cell2cell_pipelines</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.cell2cell_pipelines.BulkInteractions">BulkInteractions</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions">SingleCellInteractions</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.cell2cell_pipelines.initialize_interaction_space">initialize_interaction_space()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream">tensor_downstream</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream.compute_gini_coefficients">compute_gini_coefficients()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream.flatten_factor_ccc_networks">flatten_factor_ccc_networks()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream.get_factor_specific_ccc_networks">get_factor_specific_ccc_networks()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream.get_joint_loadings">get_joint_loadings()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_downstream.get_lr_by_cell_pairs">get_lr_by_cell_pairs()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.analysis.tensor_pipelines">tensor_pipelines</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.analysis.tensor_pipelines.run_tensor_cell2cell_pipeline">run_tensor_cell2cell_pipeline()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.clustering">clustering</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.clustering.cluster_interactions">cluster_interactions</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.clustering.cluster_interactions.compute_distance">compute_distance()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.clustering.cluster_interactions.compute_linkage">compute_linkage()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.clustering.cluster_interactions.get_clusters_from_linkage">get_clusters_from_linkage()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.core">core</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.core.cci_scores">cci_scores</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.compute_braycurtis_like_cci_score">compute_braycurtis_like_cci_score()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.compute_count_score">compute_count_score()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.compute_icellnet_score">compute_icellnet_score()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.compute_jaccard_like_cci_score">compute_jaccard_like_cci_score()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.matmul_bray_curtis_like">matmul_bray_curtis_like()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.matmul_cosine">matmul_cosine()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.matmul_count_active">matmul_count_active()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cci_scores.matmul_jaccard_like">matmul_jaccard_like()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.core.cell">cell</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cell.Cell">Cell</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.cell.get_cells_from_rnaseq">get_cells_from_rnaseq()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.core.communication_scores">communication_scores</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.aggregate_ccc_matrices">aggregate_ccc_matrices()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.compute_ccc_matrix">compute_ccc_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.get_binary_scores">get_binary_scores()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.get_continuous_scores">get_continuous_scores()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.score_expression_mean">score_expression_mean()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.communication_scores.score_expression_product">score_expression_product()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.core.interaction_space">interaction_space</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.interaction_space.InteractionSpace">InteractionSpace</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.interaction_space.generate_interaction_elements">generate_interaction_elements()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.core.interaction_space.generate_pairs">generate_pairs()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.datasets">datasets</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.datasets.anndata">anndata</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.anndata.balf_covid">balf_covid()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.datasets.gsea_data">gsea_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.gsea_data.gsea_msig">gsea_msig()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.datasets.heuristic_data">heuristic_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.heuristic_data.HeuristicGOTerms">HeuristicGOTerms</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.datasets.random_data">random_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.random_data.generate_random_cci_scores">generate_random_cci_scores()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.random_data.generate_random_metadata">generate_random_metadata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.random_data.generate_random_ppi">generate_random_ppi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.random_data.generate_random_rnaseq">generate_random_rnaseq()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.datasets.toy_data">toy_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.toy_data.generate_toy_distance">generate_toy_distance()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.toy_data.generate_toy_metadata">generate_toy_metadata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.toy_data.generate_toy_ppi">generate_toy_ppi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.datasets.toy_data.generate_toy_rnaseq">generate_toy_rnaseq()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.external">external</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.external.goenrich">goenrich</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.goenrich.gene2go">gene2go()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.goenrich.goa">goa()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.goenrich.ontology">ontology()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.goenrich.sgd">sgd()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.external.gseapy">gseapy</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.gseapy.generate_lr_geneset">generate_lr_geneset()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.gseapy.load_gmt">load_gmt()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.gseapy.run_gsea">run_gsea()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.external.pcoa">pcoa</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa.pcoa">pcoa()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa.pcoa_biplot">pcoa_biplot()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.external.pcoa_utils">pcoa_utils</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.center_distance_matrix">center_distance_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.corr">corr()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.e_matrix">e_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.f_matrix">f_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.mean_and_std">mean_and_std()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.scale">scale()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.pcoa_utils.svd_rank">svd_rank()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.external.umap">umap</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.external.umap.run_umap">run_umap()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.io">io</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.io.directories">directories</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.directories.create_directory">create_directory()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.directories.get_files_from_directory">get_files_from_directory()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.io.read_data">read_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_cutoffs">load_cutoffs()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_go_annotations">load_go_annotations()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_go_terms">load_go_terms()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_metadata">load_metadata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_ppi">load_ppi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_rnaseq">load_rnaseq()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_table">load_table()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_tables_from_directory">load_tables_from_directory()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_tensor">load_tensor()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_tensor_factors">load_tensor_factors()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.read_data.load_variable_with_pickle">load_variable_with_pickle()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.io.save_data">save_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.io.save_data.export_variable_with_pickle">export_variable_with_pickle()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.plotting">plotting</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.aesthetics">aesthetics</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.aesthetics.generate_legend">generate_legend()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.aesthetics.get_colors_from_labels">get_colors_from_labels()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.aesthetics.map_colors_to_metadata">map_colors_to_metadata()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.ccc_plot">ccc_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.ccc_plot.clustermap_ccc">clustermap_ccc()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.cci_plot">cci_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.cci_plot.clustermap_cci">clustermap_cci()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.circular_plot">circular_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.circos_plot">circos_plot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.determine_small_radius">determine_small_radius()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.generate_circos_legend">generate_circos_legend()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.get_arc_angles">get_arc_angles()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.get_cartesian">get_cartesian()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.get_node_colors">get_node_colors()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.get_readable_ccc_matrix">get_readable_ccc_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.circular_plot.sort_nodes">sort_nodes()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.factor_plot">factor_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.factor_plot.ccc_networks_plot">ccc_networks_plot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.factor_plot.context_boxplot">context_boxplot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.factor_plot.loading_clustermap">loading_clustermap()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.pcoa_plot">pcoa_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.pcoa_plot.pcoa_3dplot">pcoa_3dplot()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.pval_plot">pval_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.pval_plot.dot_plot">dot_plot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.pval_plot.generate_dot_plot">generate_dot_plot()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.tensor_plot">tensor_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.generate_plot_df">generate_plot_df()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.plot_elbow">plot_elbow()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.plot_multiple_run_elbow">plot_multiple_run_elbow()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.reorder_dimension_elements">reorder_dimension_elements()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.tensor_factors_plot">tensor_factors_plot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.tensor_plot.tensor_factors_plot_from_loadings">tensor_factors_plot_from_loadings()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.plotting.umap_plot">umap_plot</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.plotting.umap_plot.umap_biplot">umap_biplot()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.preprocessing">preprocessing</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.cutoffs">cutoffs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.cutoffs.get_constant_cutoff">get_constant_cutoff()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.cutoffs.get_cutoffs">get_cutoffs()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.cutoffs.get_global_percentile_cutoffs">get_global_percentile_cutoffs()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.cutoffs.get_local_percentile_cutoffs">get_local_percentile_cutoffs()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.find_elements">find_elements</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.find_elements.find_duplicates">find_duplicates()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.find_elements.get_element_abundances">get_element_abundances()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.find_elements.get_elements_over_fraction">get_elements_over_fraction()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.gene_ontology">gene_ontology</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.gene_ontology.find_all_children_of_go_term">find_all_children_of_go_term()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.gene_ontology.find_go_terms_from_keyword">find_go_terms_from_keyword()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.gene_ontology.get_genes_from_go_hierarchy">get_genes_from_go_hierarchy()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.gene_ontology.get_genes_from_go_terms">get_genes_from_go_terms()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data">integrate_data</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data.get_modified_rnaseq">get_modified_rnaseq()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data.get_ppi_dict_from_go_terms">get_ppi_dict_from_go_terms()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data.get_ppi_dict_from_proteins">get_ppi_dict_from_proteins()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data.get_thresholded_rnaseq">get_thresholded_rnaseq()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.integrate_data.get_weighted_ppi">get_weighted_ppi()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes">manipulate_dataframes</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.check_presence_in_dataframe">check_presence_in_dataframe()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.check_symmetry">check_symmetry()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.convert_to_distance_matrix">convert_to_distance_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.shuffle_cols_in_df">shuffle_cols_in_df()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.shuffle_dataframe">shuffle_dataframe()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.shuffle_rows_in_df">shuffle_rows_in_df()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.manipulate_dataframes.subsample_dataframe">subsample_dataframe()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.ppi">ppi</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.bidirectional_ppi_for_cci">bidirectional_ppi_for_cci()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.filter_complex_ppi_by_proteins">filter_complex_ppi_by_proteins()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.filter_ppi_by_proteins">filter_ppi_by_proteins()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.filter_ppi_network">filter_ppi_network()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.get_all_to_all_ppi">get_all_to_all_ppi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.get_filtered_ppi_network">get_filtered_ppi_network()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.get_genes_from_complexes">get_genes_from_complexes()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.get_one_group_to_other_ppi">get_one_group_to_other_ppi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.preprocess_ppi_data">preprocess_ppi_data()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.remove_ppi_bidirectionality">remove_ppi_bidirectionality()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.ppi.simplify_ppi">simplify_ppi()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq">rnaseq</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.add_complexes_to_expression">add_complexes_to_expression()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.aggregate_single_cells">aggregate_single_cells()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.divide_expression_by_max">divide_expression_by_max()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.divide_expression_by_mean">divide_expression_by_mean()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.drop_empty_genes">drop_empty_genes()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.log10_transformation">log10_transformation()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.rnaseq.scale_expression_by_sum">scale_expression_by_sum()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.preprocessing.signal">signal</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.preprocessing.signal.smooth_curve">smooth_curve()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.spatial">spatial</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.spatial.distances">distances</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.distances.celltype_pair_distance">celltype_pair_distance()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.distances.pairwise_celltype_distances">pairwise_celltype_distances()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.spatial.filtering">filtering</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.filtering.dist_filter_liana">dist_filter_liana()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.filtering.dist_filter_tensor">dist_filter_tensor()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.spatial.neighborhoods">neighborhoods</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.neighborhoods.add_sliding_window_info_to_adata">add_sliding_window_info_to_adata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.neighborhoods.calculate_window_size">calculate_window_size()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.neighborhoods.create_sliding_windows">create_sliding_windows()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.spatial.neighborhoods.create_spatial_grid">create_spatial_grid()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.stats">stats</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.stats.enrichment">enrichment</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.enrichment.fisher_representation">fisher_representation()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.enrichment.hypergeom_representation">hypergeom_representation()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.stats.gini">gini</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.gini.gini_coefficient">gini_coefficient()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.stats.multitest">multitest</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.multitest.compute_fdrcorrection_asymmetric_matrix">compute_fdrcorrection_asymmetric_matrix()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.multitest.compute_fdrcorrection_symmetric_matrix">compute_fdrcorrection_symmetric_matrix()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.stats.permutation">permutation</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.permutation.compute_pvalue_from_dist">compute_pvalue_from_dist()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.permutation.pvalue_from_dist">pvalue_from_dist()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.permutation.random_switching_ppi_labels">random_switching_ppi_labels()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.stats.permutation.run_label_permutation">run_label_permutation()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.tensor">tensor</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.external_scores">external_scores</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.external_scores.dataframes_to_tensor">dataframes_to_tensor()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.factor_manipulation">factor_manipulation</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.factor_manipulation.normalize_factors">normalize_factors()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.factor_manipulation.shuffle_factors">shuffle_factors()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.factorization">factorization</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.factorization.normalized_error">normalized_error()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.metrics">metrics</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.metrics.correlation_index">correlation_index()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.metrics.pairwise_correlation_index">pairwise_correlation_index()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.subset">subset</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.subset.find_element_indexes">find_element_indexes()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.subset.subset_metadata">subset_metadata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.subset.subset_tensor">subset_tensor()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.tensor">tensor</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.BaseTensor">BaseTensor</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.InteractionTensor">InteractionTensor</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.PreBuiltTensor">PreBuiltTensor</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.aggregate_ccc_tensor">aggregate_ccc_tensor()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.build_context_ccc_tensor">build_context_ccc_tensor()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.generate_ccc_tensor">generate_ccc_tensor()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.generate_tensor_metadata">generate_tensor_metadata()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor.interactions_to_tensor">interactions_to_tensor()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.tensor.tensor_manipulation">tensor_manipulation</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.tensor.tensor_manipulation.concatenate_interaction_tensors">concatenate_interaction_tensors()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell2cell.utils">utils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.utils.networks">networks</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.utils.networks.export_network_to_cytoscape">export_network_to_cytoscape()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.utils.networks.export_network_to_gephi">export_network_to_gephi()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.utils.networks.generate_network_from_adjacency">generate_network_from_adjacency()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell2cell.utils.parallel_computing">parallel_computing</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.utils.parallel_computing.agents_number">agents_number()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#cell2cell.utils.parallel_computing.parallel_spatial_ccis">parallel_spatial_ccis()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">cell2cell Tutorials</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/Toy-Example-BulkPipeline/">Cell-cell communication from bulk dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/Toy-Example-SingleCellPipeline/">Cell-cell communication from single-cell dataset</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tensor-cell2cell Tutorials</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/ASD/01-Tensor-Factorization-ASD/">Obtaining patterns of cell-cell communication with Tensor-cell2cell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/ASD/02-Factor-Specific-ASD/">Downstream analysis 1: Factor-specific analyses</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/ASD/03-GSEA-ASD/">Downstream analysis 2: Gene Set Enrichment Analysis</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/Tensor-cell2cell-Spatial/">Inspecting CCC patterns from spatial transcriptomics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tutorials/GPU-Example/">Running Tensor-cell2cell on your own GPU or on Google Colab's GPU</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">cell2cell</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li><li>API Documentation</li>
    <li class="wy-breadcrumbs-aside">
        <a href="https://github.com/earmingol/cell2cell/edit/master/docs/documentation.md"
          class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>

          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation-for-cell2cell">Documentation for <em>cell2cell</em></h1>
<p>This documentation is for our <em>cell2cell</em> suite, which includes the <a href="https://doi.org/10.1371/journal.pcbi.1010715">regular cell2cell</a>
and <a href="https://doi.org/10.1038/s41467-022-31369-2">Tensor-cell2cell</a> tools. The former is for inferring cell-cell interactions
and communication in one sample or context, while the latter is for deconvolving complex patterns
of cell-cell communication across multiple samples or contexts simultaneously into interpretable factors
representing patterns of communication.</p>
<p>Here, multiple classes and functions are implemented to facilitate the analyses, including a variety of
visualizations to simplify the interpretation of results:</p>
<ul>
<li><strong>cell2cell.analysis</strong> : Includes simplified pipelines for running the analyses, and functions for downstream analyses of Tensor-cell2cell</li>
<li><strong>cell2cell.clustering</strong> : Includes multiple scipy-based functions for performing clustering methods.</li>
<li><strong>cell2cell.core</strong> : Includes the core functions for inferring cell-cell interactions and communication. It includes scoring methods, cell classes, and interaction spaces.</li>
<li><strong>cell2cell.datasets</strong> : Includes toy datasets and annotations for testing functions in basic scenarios.</li>
<li><strong>cell2cell.external</strong> : Includes built-in approaches borrowed from other tools to avoid incompatibilities (e.g. UMAP, tensorly, and PCoA).</li>
<li><strong>cell2cell.io</strong> : Includes functions for opening and saving diverse types of files.</li>
<li><strong>cell2cell.plotting</strong> : Includes all the visualization options that <em>cell2cell</em> offers.</li>
<li><strong>cell2cell.preprocessing</strong> : Includes functions for manipulating data and variables (e.g. data preprocessing, integration, permutation, among others).</li>
<li><strong>cell2cell.spatial</strong> : Includes filtering of cell-cell interactions results given intercellular distance, as well as defining neighborhoods by grids or moving windows.</li>
<li><strong>cell2cell.stats</strong> : Includes statistical analyses such as enrichment analysis, multiple test correction methods, permutation approaches, and Gini coefficient.</li>
<li><strong>cell2cell.tensor</strong> : Includes all functions pertinent to the analysis of <em>Tensor-cell2cell</em></li>
<li><strong>cell2cell.utils</strong> : Includes general utilities for analyzing networks and performing parallel computing.</li>
</ul>
<p>Below, all the inputs, parameters (including their different options), and outputs are detailed. Source code of the functions is also included.</p>


  <div class="doc doc-object doc-module">

<a id="cell2cell"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">











  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.analysis">
        <code>analysis</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines">
        <code>cell2cell_pipelines</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions">
        <code>
BulkInteractions        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Interaction class with all necessary methods to run the cell2cell pipeline
on a bulk RNA-seq dataset. Cells here could be represented by tissues, samples
or any bulk organization of cells.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment. Columns are samples
    and rows are genes.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>metadata : pandas.Dataframe, default=None
    Metadata associated with the samples in the RNA-seq dataset.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>communication_score : str, default='expression_thresholding'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39; : Computes the joint presence of a ligand from a
                             sender cell and of a receptor on a receiver cell
                             from binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
</code></pre></div>

<p>cci_score : str, default='bray_curtis'
    Scoring function to aggregate the communication scores between a pair of
    cells. It computes an overall potential of cell-cell interactions.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score.
- &#39;jaccard&#39; : Jaccard-like score.
- &#39;count&#39; : Number of LR pairs that the pair of cells use.
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>cci_type : str, default='undirected'
    Specifies whether computing the cci_score in a directed or undirected
    way. For a pair of cells A and B, directed means that the ligands are
    considered only from cell A and receptors only from cell B or viceversa.
    While undirected simultaneously considers signaling from cell A to
    cell B and from cell B to cell A.</p>
<p>sample_col : str, default='sampleID'
    Column-name for the samples in the metadata.</p>
<p>group_col : str, default='tissue'
    Column-name for the grouping information associated with the samples
    in the metadata.</p>
<p>expression_threshold : float, default=10
    Threshold value to binarize gene expression when using
    communication_score='expression_thresholding'. Units have to be the
    same as the rnaseq_data matrix (e.g., TPMs, counts, etc.).</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions--attributes">Attributes</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment. Columns are samples
    and rows are genes.</p>
<p>metadata : pandas.DataFrame
    Metadata associated with the samples in the RNA-seq dataset.</p>
<p>index_col : str
    Column-name for the samples in the metadata.</p>
<p>group_col : str
    Column-name for the grouping information associated with the samples
    in the metadata.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used for
    inferring the cell-cell interactions and communication.</p>
<p>complex_sep : str
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>ref_ppi : pandas.DataFrame
    Reference list of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication. It could be the
    same as 'ppi_data' if ppi_data is not bidirectional (that is, contains
    ProtA-ProtB interaction as well as ProtB-ProtA interaction). ref_ppi must
    be undirected (contains only ProtA-ProtB and not ProtB-ProtA interaction).</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>analysis_setup : dict
    Contains main setup for running the cell-cell interactions and communication
    analyses.
    Three main setups are needed (passed as keys):</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">communication_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">communication</span><span class="w"> </span><span class="n">score</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">detect</span><span class="w"></span>
<span class="w">    </span><span class="n">active</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">pair</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">cell</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_thresholding</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_product</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_mean</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_gmean</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">scoring</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">communication</span><span class="w"></span>
<span class="w">    </span><span class="n">scores</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">bray_curtis</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">jaccard</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">count</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">icellnet</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_type</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interaction</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">two</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"> </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"></span>
<span class="w">    </span><span class="n">undirected</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">directed</span><span class="p">,</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">other</span><span class="w"></span>
<span class="w">    </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">separately</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="n">respect</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">second</span><span class="w"></span>
<span class="w">    </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptor</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">first</span><span class="w"> </span><span class="n">one</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">So</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">undirected</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">directed</span><span class="p">&#39;</span><span class="w"></span>
</code></pre></div>

<p>cutoff_setup : dict
    Contains two keys: 'type' and 'parameter'. The first key represent the
    way to use a cutoff or threshold, while parameter is the value used
    to binarize the expression values.
    The key 'type' can be:</p>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"></span>
<span class="w">        </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">        </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"></span>
<span class="w">        </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"></span>
<span class="w">        </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">        </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"></span>
<span class="w">        </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"></span>
<span class="w">        </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">        </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">        </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">        </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all the required elements to perform the
    cell-cell interaction and communication analysis between every pair of cells.
    After performing the analyses, the results are stored in this object.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">BulkInteractions</span><span class="p">:</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Interaction class with all necessary methods to run the cell2cell pipeline</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    on a bulk RNA-seq dataset. Cells here could be represented by tissues, samples</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    or any bulk organization of cells.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment. Columns are samples</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        and rows are genes.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    metadata : pandas.Dataframe, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Metadata associated with the samples in the RNA-seq dataset.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    communication_score : str, default=&#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;expression_thresholding&#39; : Computes the joint presence of a ligand from a</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                                     sender cell and of a receptor on a receiver cell</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                                     from binarizing their gene expression levels.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    cci_score : str, default=&#39;bray_curtis&#39;</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Scoring function to aggregate the communication scores between a pair of</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        cells. It computes an overall potential of cell-cell interactions.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells use.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    cci_type : str, default=&#39;undirected&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Specifies whether computing the cci_score in a directed or undirected</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        way. For a pair of cells A and B, directed means that the ligands are</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        considered only from cell A and receptors only from cell B or viceversa.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        While undirected simultaneously considers signaling from cell A to</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        cell B and from cell B to cell A.</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    sample_col : str, default=&#39;sampleID&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        Column-name for the samples in the metadata.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    group_col : str, default=&#39;tissue&#39;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Column-name for the grouping information associated with the samples</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        in the metadata.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    expression_threshold : float, default=10</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Threshold value to binarize gene expression when using</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        communication_score=&#39;expression_thresholding&#39;. Units have to be the</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        same as the rnaseq_data matrix (e.g., TPMs, counts, etc.).</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment. Columns are samples</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        and rows are genes.</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">    metadata : pandas.DataFrame</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        Metadata associated with the samples in the RNA-seq dataset.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">    index_col : str</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">        Column-name for the samples in the metadata.</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">    group_col : str</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">        Column-name for the grouping information associated with the samples</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">        in the metadata.</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used for</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a><span class="sd">    complex_sep : str</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a><span class="sd">    complex_agg_method : str</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a><span class="sd">    ref_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a><span class="sd">        Reference list of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a><span class="sd">        for inferring the cell-cell interactions and communication. It could be the</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a><span class="sd">        same as &#39;ppi_data&#39; if ppi_data is not bidirectional (that is, contains</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a><span class="sd">        ProtA-ProtB interaction as well as ProtB-ProtA interaction). ref_ppi must</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a><span class="sd">        be undirected (contains only ProtA-ProtB and not ProtB-ProtA interaction).</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a><span class="sd">    analysis_setup : dict</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a><span class="sd">        Contains main setup for running the cell-cell interactions and communication</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a><span class="sd">        analyses.</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a><span class="sd">        Three main setups are needed (passed as keys):</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a><span class="sd">        - &#39;communication_score&#39; : is the type of communication score used to detect</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a><span class="sd">            active ligand-receptor pairs between each pair of cell.</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a><span class="sd">            - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a><span class="sd">            - &#39;expression_product&#39;</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a><span class="sd">            - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a><span class="sd">            - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a><span class="sd">        - &#39;cci_score&#39; : is the scoring function to aggregate the communication</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a><span class="sd">            scores.</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a><span class="sd">            - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a><span class="sd">            - &#39;jaccard&#39;</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a><span class="sd">            - &#39;count&#39;</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a><span class="sd">            - &#39;icellnet&#39;</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a><span class="sd">        - &#39;cci_type&#39; : is the type of interaction between two cells. If it is</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a><span class="sd">            undirected, all ligands and receptors are considered from both cells.</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a><span class="sd">            If it is directed, ligands from one cell and receptors from the other</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a><span class="sd">            are considered separately with respect to ligands from the second</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a><span class="sd">            cell and receptor from the first one.</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a><span class="sd">            So, it can be:</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a><span class="sd">            - &#39;undirected&#39;</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a><span class="sd">            - &#39;directed&#39;</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a><span class="sd">    cutoff_setup : dict</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a><span class="sd">        Contains two keys: &#39;type&#39; and &#39;parameter&#39;. The first key represent the</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a><span class="sd">        way to use a cutoff or threshold, while parameter is the value used</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a><span class="sd">        to binarize the expression values.</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a><span class="sd">        The key &#39;type&#39; can be:</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a><span class="sd">            - &#39;local_percentile&#39; : computes the value of a given percentile, for each</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a><span class="sd">                gene independently. In this case, the parameter corresponds to the</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a><span class="sd">                percentile to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a><span class="sd">            - &#39;global_percentile&#39; : computes the value of a given percentile from all</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a><span class="sd">                genes and samples simultaneously. In this case, the parameter</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a><span class="sd">                corresponds to the percentile to compute, as a float value between</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a><span class="sd">                0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a><span class="sd">            - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is the</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a><span class="sd">                path of that file. It must contain the same genes as index and same</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a><span class="sd">                samples as columns.</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a><span class="sd">            - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a><span class="sd">                for each gene in each sample. This allows to use specific cutoffs for</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a><span class="sd">                each sample. The columns here must be the same as the ones in the</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a><span class="sd">                rnaseq_data.</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a><span class="sd">            - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a><span class="sd">                for each gene in only one column. These cutoffs will be applied to</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a><span class="sd">                all samples.</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a><span class="sd">            - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a><span class="sd">                expression is greater than the value input in the parameter.</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a><span class="sd">        Interaction space that contains all the required elements to perform the</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a><span class="sd">        cell-cell interaction and communication analysis between every pair of cells.</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a><span class="sd">        After performing the analyses, the results are stored in this object.</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                 <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="s1">&#39;undirected&#39;</span><span class="p">,</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>                 <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;sampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;tissue&#39;</span><span class="p">,</span> <span class="n">expression_threshold</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>                 <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>        <span class="c1"># Placeholders</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">index_col</span> <span class="o">=</span> <span class="n">sample_col</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">group_col</span> <span class="o">=</span> <span class="n">group_col</span>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">complex_sep</span> <span class="o">=</span> <span class="n">complex_sep</span>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">complex_agg_method</span> <span class="o">=</span> <span class="n">complex_agg_method</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span> <span class="o">=</span> <span class="n">interaction_columns</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a>        <span class="c1"># Analysis setup</span>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;communication_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">communication_score</span>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_score</span>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a>        <span class="c1"># Initialize PPI</span>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-221" name="__codelineno-0-221" href="#__codelineno-0-221"></a>        <span class="n">ppi_data_</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-222" name="__codelineno-0-222" href="#__codelineno-0-222"></a>                                               <span class="n">proteins</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
<a id="__codelineno-0-223" name="__codelineno-0-223" href="#__codelineno-0-223"></a>                                               <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-224" name="__codelineno-0-224" href="#__codelineno-0-224"></a>                                               <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-225" name="__codelineno-0-225" href="#__codelineno-0-225"></a>                                               <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-226" name="__codelineno-0-226" href="#__codelineno-0-226"></a>
<a id="__codelineno-0-227" name="__codelineno-0-227" href="#__codelineno-0-227"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data_</span><span class="p">,</span>
<a id="__codelineno-0-228" name="__codelineno-0-228" href="#__codelineno-0-228"></a>                                                        <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-229" name="__codelineno-0-229" href="#__codelineno-0-229"></a>                                                        <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-230" name="__codelineno-0-230" href="#__codelineno-0-230"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-231" name="__codelineno-0-231" href="#__codelineno-0-231"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-232" name="__codelineno-0-232" href="#__codelineno-0-232"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-233" name="__codelineno-0-233" href="#__codelineno-0-233"></a>                                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-234" name="__codelineno-0-234" href="#__codelineno-0-234"></a>                                                          <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-235" name="__codelineno-0-235" href="#__codelineno-0-235"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-236" name="__codelineno-0-236" href="#__codelineno-0-236"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-237" name="__codelineno-0-237" href="#__codelineno-0-237"></a>
<a id="__codelineno-0-238" name="__codelineno-0-238" href="#__codelineno-0-238"></a>        <span class="c1"># Thresholding</span>
<a id="__codelineno-0-239" name="__codelineno-0-239" href="#__codelineno-0-239"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;constant_value&#39;</span>
<a id="__codelineno-0-240" name="__codelineno-0-240" href="#__codelineno-0-240"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">[</span><span class="s1">&#39;parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">expression_threshold</span>
<a id="__codelineno-0-241" name="__codelineno-0-241" href="#__codelineno-0-241"></a>
<a id="__codelineno-0-242" name="__codelineno-0-242" href="#__codelineno-0-242"></a>        <span class="c1"># Interaction Space</span>
<a id="__codelineno-0-243" name="__codelineno-0-243" href="#__codelineno-0-243"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span> <span class="o">=</span> <span class="n">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-244" name="__codelineno-0-244" href="#__codelineno-0-244"></a>                                                              <span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-245" name="__codelineno-0-245" href="#__codelineno-0-245"></a>                                                              <span class="n">cutoff_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-246" name="__codelineno-0-246" href="#__codelineno-0-246"></a>                                                              <span class="n">analysis_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">,</span>
<a id="__codelineno-0-247" name="__codelineno-0-247" href="#__codelineno-0-247"></a>                                                              <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-248" name="__codelineno-0-248" href="#__codelineno-0-248"></a>                                                              <span class="n">complex_agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-249" name="__codelineno-0-249" href="#__codelineno-0-249"></a>                                                              <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-250" name="__codelineno-0-250" href="#__codelineno-0-250"></a>                                                              <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-251" name="__codelineno-0-251" href="#__codelineno-0-251"></a>
<a id="__codelineno-0-252" name="__codelineno-0-252" href="#__codelineno-0-252"></a>    <span class="k">def</span> <span class="nf">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-253" name="__codelineno-0-253" href="#__codelineno-0-253"></a>        <span class="sd">&#39;&#39;&#39;Computes overall CCI scores for each pair of cells.</span>
<a id="__codelineno-0-254" name="__codelineno-0-254" href="#__codelineno-0-254"></a>
<a id="__codelineno-0-255" name="__codelineno-0-255" href="#__codelineno-0-255"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-256" name="__codelineno-0-256" href="#__codelineno-0-256"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-257" name="__codelineno-0-257" href="#__codelineno-0-257"></a><span class="sd">        cci_score : str, default=None</span>
<a id="__codelineno-0-258" name="__codelineno-0-258" href="#__codelineno-0-258"></a><span class="sd">            Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-259" name="__codelineno-0-259" href="#__codelineno-0-259"></a><span class="sd">            a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-260" name="__codelineno-0-260" href="#__codelineno-0-260"></a><span class="sd">            interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-261" name="__codelineno-0-261" href="#__codelineno-0-261"></a><span class="sd">            attribute analysis_setup of this object.</span>
<a id="__codelineno-0-262" name="__codelineno-0-262" href="#__codelineno-0-262"></a><span class="sd">            Options:</span>
<a id="__codelineno-0-263" name="__codelineno-0-263" href="#__codelineno-0-263"></a>
<a id="__codelineno-0-264" name="__codelineno-0-264" href="#__codelineno-0-264"></a><span class="sd">            - &#39;bray_curtis&#39; : Bray-Curtis-like score.</span>
<a id="__codelineno-0-265" name="__codelineno-0-265" href="#__codelineno-0-265"></a><span class="sd">            - &#39;jaccard&#39; : Jaccard-like score.</span>
<a id="__codelineno-0-266" name="__codelineno-0-266" href="#__codelineno-0-266"></a><span class="sd">            - &#39;count&#39; : Number of LR pairs that the pair of cells use.</span>
<a id="__codelineno-0-267" name="__codelineno-0-267" href="#__codelineno-0-267"></a><span class="sd">            - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-268" name="__codelineno-0-268" href="#__codelineno-0-268"></a>
<a id="__codelineno-0-269" name="__codelineno-0-269" href="#__codelineno-0-269"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-270" name="__codelineno-0-270" href="#__codelineno-0-270"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-271" name="__codelineno-0-271" href="#__codelineno-0-271"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-272" name="__codelineno-0-272" href="#__codelineno-0-272"></a>
<a id="__codelineno-0-273" name="__codelineno-0-273" href="#__codelineno-0-273"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-274" name="__codelineno-0-274" href="#__codelineno-0-274"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-275" name="__codelineno-0-275" href="#__codelineno-0-275"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-276" name="__codelineno-0-276" href="#__codelineno-0-276"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="n">cci_score</span><span class="o">=</span><span class="n">cci_score</span><span class="p">,</span>
<a id="__codelineno-0-277" name="__codelineno-0-277" href="#__codelineno-0-277"></a>                                                           <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-278" name="__codelineno-0-278" href="#__codelineno-0-278"></a>                                                           <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-279" name="__codelineno-0-279" href="#__codelineno-0-279"></a>
<a id="__codelineno-0-280" name="__codelineno-0-280" href="#__codelineno-0-280"></a>    <span class="k">def</span> <span class="nf">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-281" name="__codelineno-0-281" href="#__codelineno-0-281"></a>                                              <span class="n">interaction_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-282" name="__codelineno-0-282" href="#__codelineno-0-282"></a>        <span class="sd">&#39;&#39;&#39;Computes the communication scores for each LR pairs in</span>
<a id="__codelineno-0-283" name="__codelineno-0-283" href="#__codelineno-0-283"></a><span class="sd">        a given pair of sender-receiver cell</span>
<a id="__codelineno-0-284" name="__codelineno-0-284" href="#__codelineno-0-284"></a>
<a id="__codelineno-0-285" name="__codelineno-0-285" href="#__codelineno-0-285"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-286" name="__codelineno-0-286" href="#__codelineno-0-286"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-287" name="__codelineno-0-287" href="#__codelineno-0-287"></a><span class="sd">        communication_score : str, default=None</span>
<a id="__codelineno-0-288" name="__codelineno-0-288" href="#__codelineno-0-288"></a><span class="sd">            Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-289" name="__codelineno-0-289" href="#__codelineno-0-289"></a><span class="sd">            a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-290" name="__codelineno-0-290" href="#__codelineno-0-290"></a><span class="sd">            If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-291" name="__codelineno-0-291" href="#__codelineno-0-291"></a><span class="sd">            will be used.</span>
<a id="__codelineno-0-292" name="__codelineno-0-292" href="#__codelineno-0-292"></a><span class="sd">            Available communication_scores are:</span>
<a id="__codelineno-0-293" name="__codelineno-0-293" href="#__codelineno-0-293"></a>
<a id="__codelineno-0-294" name="__codelineno-0-294" href="#__codelineno-0-294"></a><span class="sd">            - &#39;expresion_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-295" name="__codelineno-0-295" href="#__codelineno-0-295"></a><span class="sd">                                         ligand from a sender cell and of</span>
<a id="__codelineno-0-296" name="__codelineno-0-296" href="#__codelineno-0-296"></a><span class="sd">                                         a receptor on a receiver cell from</span>
<a id="__codelineno-0-297" name="__codelineno-0-297" href="#__codelineno-0-297"></a><span class="sd">                                         binarizing their gene expression levels.</span>
<a id="__codelineno-0-298" name="__codelineno-0-298" href="#__codelineno-0-298"></a><span class="sd">            - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-299" name="__codelineno-0-299" href="#__codelineno-0-299"></a><span class="sd">                                  of a ligand from a sender cell and the</span>
<a id="__codelineno-0-300" name="__codelineno-0-300" href="#__codelineno-0-300"></a><span class="sd">                                  expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-301" name="__codelineno-0-301" href="#__codelineno-0-301"></a><span class="sd">            - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-302" name="__codelineno-0-302" href="#__codelineno-0-302"></a><span class="sd">                                    of a ligand from a sender cell and the</span>
<a id="__codelineno-0-303" name="__codelineno-0-303" href="#__codelineno-0-303"></a><span class="sd">                                    expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-304" name="__codelineno-0-304" href="#__codelineno-0-304"></a><span class="sd">            - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-305" name="__codelineno-0-305" href="#__codelineno-0-305"></a><span class="sd">                                  of a ligand from a sender cell and the</span>
<a id="__codelineno-0-306" name="__codelineno-0-306" href="#__codelineno-0-306"></a><span class="sd">                                  expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-307" name="__codelineno-0-307" href="#__codelineno-0-307"></a>
<a id="__codelineno-0-308" name="__codelineno-0-308" href="#__codelineno-0-308"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-309" name="__codelineno-0-309" href="#__codelineno-0-309"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-310" name="__codelineno-0-310" href="#__codelineno-0-310"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-311" name="__codelineno-0-311" href="#__codelineno-0-311"></a>
<a id="__codelineno-0-312" name="__codelineno-0-312" href="#__codelineno-0-312"></a><span class="sd">        ref_ppi_data : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-313" name="__codelineno-0-313" href="#__codelineno-0-313"></a><span class="sd">            Reference list of protein-protein interactions (or</span>
<a id="__codelineno-0-314" name="__codelineno-0-314" href="#__codelineno-0-314"></a><span class="sd">            ligand-receptor pairs) used for inferring the cell-cell</span>
<a id="__codelineno-0-315" name="__codelineno-0-315" href="#__codelineno-0-315"></a><span class="sd">            interactions and communication. It could be the same as</span>
<a id="__codelineno-0-316" name="__codelineno-0-316" href="#__codelineno-0-316"></a><span class="sd">            &#39;ppi_data&#39; if ppi_data is not bidirectional (that is,</span>
<a id="__codelineno-0-317" name="__codelineno-0-317" href="#__codelineno-0-317"></a><span class="sd">            contains ProtA-ProtB interaction as well as ProtB-ProtA</span>
<a id="__codelineno-0-318" name="__codelineno-0-318" href="#__codelineno-0-318"></a><span class="sd">            interaction). ref_ppi must be undirected (contains only</span>
<a id="__codelineno-0-319" name="__codelineno-0-319" href="#__codelineno-0-319"></a><span class="sd">            ProtA-ProtB and not ProtB-ProtA interaction). If None</span>
<a id="__codelineno-0-320" name="__codelineno-0-320" href="#__codelineno-0-320"></a><span class="sd">            the one stored in the attribute ref_ppi will be used.</span>
<a id="__codelineno-0-321" name="__codelineno-0-321" href="#__codelineno-0-321"></a>
<a id="__codelineno-0-322" name="__codelineno-0-322" href="#__codelineno-0-322"></a><span class="sd">        interaction_columns : tuple, default=None</span>
<a id="__codelineno-0-323" name="__codelineno-0-323" href="#__codelineno-0-323"></a><span class="sd">            Contains the names of the columns where to find the</span>
<a id="__codelineno-0-324" name="__codelineno-0-324" href="#__codelineno-0-324"></a><span class="sd">            partners in a dataframe of protein-protein interactions.</span>
<a id="__codelineno-0-325" name="__codelineno-0-325" href="#__codelineno-0-325"></a><span class="sd">            If the list is for ligand-receptor pairs, the first column</span>
<a id="__codelineno-0-326" name="__codelineno-0-326" href="#__codelineno-0-326"></a><span class="sd">            is for the ligands and the second for the receptors. If</span>
<a id="__codelineno-0-327" name="__codelineno-0-327" href="#__codelineno-0-327"></a><span class="sd">            None, the one stored in the attribute interaction_columns</span>
<a id="__codelineno-0-328" name="__codelineno-0-328" href="#__codelineno-0-328"></a><span class="sd">            will be used.</span>
<a id="__codelineno-0-329" name="__codelineno-0-329" href="#__codelineno-0-329"></a>
<a id="__codelineno-0-330" name="__codelineno-0-330" href="#__codelineno-0-330"></a><span class="sd">        cells : list=None</span>
<a id="__codelineno-0-331" name="__codelineno-0-331" href="#__codelineno-0-331"></a><span class="sd">            List of cells to consider.</span>
<a id="__codelineno-0-332" name="__codelineno-0-332" href="#__codelineno-0-332"></a>
<a id="__codelineno-0-333" name="__codelineno-0-333" href="#__codelineno-0-333"></a><span class="sd">        cci_type : str, default=None</span>
<a id="__codelineno-0-334" name="__codelineno-0-334" href="#__codelineno-0-334"></a><span class="sd">            Type of interaction between two cells. Used to specify</span>
<a id="__codelineno-0-335" name="__codelineno-0-335" href="#__codelineno-0-335"></a><span class="sd">            if we want to consider a LR pair in both directions.</span>
<a id="__codelineno-0-336" name="__codelineno-0-336" href="#__codelineno-0-336"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-337" name="__codelineno-0-337" href="#__codelineno-0-337"></a>
<a id="__codelineno-0-338" name="__codelineno-0-338" href="#__codelineno-0-338"></a><span class="sd">            - &#39;undirected&#39;</span>
<a id="__codelineno-0-339" name="__codelineno-0-339" href="#__codelineno-0-339"></a><span class="sd">            - &#39;directed&#39;</span>
<a id="__codelineno-0-340" name="__codelineno-0-340" href="#__codelineno-0-340"></a>
<a id="__codelineno-0-341" name="__codelineno-0-341" href="#__codelineno-0-341"></a><span class="sd">            If None, &#39;directed&#39; will be used.</span>
<a id="__codelineno-0-342" name="__codelineno-0-342" href="#__codelineno-0-342"></a>
<a id="__codelineno-0-343" name="__codelineno-0-343" href="#__codelineno-0-343"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-344" name="__codelineno-0-344" href="#__codelineno-0-344"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-345" name="__codelineno-0-345" href="#__codelineno-0-345"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-346" name="__codelineno-0-346" href="#__codelineno-0-346"></a>        <span class="k">if</span> <span class="n">interaction_columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-347" name="__codelineno-0-347" href="#__codelineno-0-347"></a>            <span class="n">interaction_columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span> <span class="c1"># Used only for ref_ppi_data</span>
<a id="__codelineno-0-348" name="__codelineno-0-348" href="#__codelineno-0-348"></a>
<a id="__codelineno-0-349" name="__codelineno-0-349" href="#__codelineno-0-349"></a>        <span class="k">if</span> <span class="n">ref_ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-350" name="__codelineno-0-350" href="#__codelineno-0-350"></a>            <span class="n">ref_ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span>
<a id="__codelineno-0-351" name="__codelineno-0-351" href="#__codelineno-0-351"></a>
<a id="__codelineno-0-352" name="__codelineno-0-352" href="#__codelineno-0-352"></a>        <span class="k">if</span> <span class="n">cci_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-353" name="__codelineno-0-353" href="#__codelineno-0-353"></a>            <span class="n">cci_type</span> <span class="o">=</span> <span class="s1">&#39;directed&#39;</span>
<a id="__codelineno-0-354" name="__codelineno-0-354" href="#__codelineno-0-354"></a>
<a id="__codelineno-0-355" name="__codelineno-0-355" href="#__codelineno-0-355"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-356" name="__codelineno-0-356" href="#__codelineno-0-356"></a>
<a id="__codelineno-0-357" name="__codelineno-0-357" href="#__codelineno-0-357"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-358" name="__codelineno-0-358" href="#__codelineno-0-358"></a>                                                                     <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-359" name="__codelineno-0-359" href="#__codelineno-0-359"></a>                                                                     <span class="n">ref_ppi_data</span><span class="o">=</span><span class="n">ref_ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-360" name="__codelineno-0-360" href="#__codelineno-0-360"></a>                                                                     <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-361" name="__codelineno-0-361" href="#__codelineno-0-361"></a>                                                                     <span class="n">cells</span><span class="o">=</span><span class="n">cells</span><span class="p">,</span>
<a id="__codelineno-0-362" name="__codelineno-0-362" href="#__codelineno-0-362"></a>                                                                     <span class="n">cci_type</span><span class="o">=</span><span class="n">cci_type</span><span class="p">,</span>
<a id="__codelineno-0-363" name="__codelineno-0-363" href="#__codelineno-0-363"></a>                                                                     <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-364" name="__codelineno-0-364" href="#__codelineno-0-364"></a>
<a id="__codelineno-0-365" name="__codelineno-0-365" href="#__codelineno-0-365"></a>    <span class="nd">@property</span>
<a id="__codelineno-0-366" name="__codelineno-0-366" href="#__codelineno-0-366"></a>    <span class="k">def</span> <span class="nf">interaction_elements</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-367" name="__codelineno-0-367" href="#__codelineno-0-367"></a>        <span class="sd">&#39;&#39;&#39;Returns the interaction elements within an interaction space.&#39;&#39;&#39;</span>
<a id="__codelineno-0-368" name="__codelineno-0-368" href="#__codelineno-0-368"></a>        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_elements&#39;</span><span class="p">):</span>
<a id="__codelineno-0-369" name="__codelineno-0-369" href="#__codelineno-0-369"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span>
<a id="__codelineno-0-370" name="__codelineno-0-370" href="#__codelineno-0-370"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-371" name="__codelineno-0-371" href="#__codelineno-0-371"></a>            <span class="k">return</span> <span class="kc">None</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">






  <div class="doc doc-object doc-attribute">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions.interaction_elements">
<code class="highlight language-python"><span class="n">interaction_elements</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-property"><code>property</code></small>
      <small class="doc doc-property doc-property-readonly"><code>readonly</code></small>
  </span>

</h5>

    <div class="doc doc-contents ">

      <p>Returns the interaction elements within an interaction space.</p>
    </div>

  </div>







  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions.compute_pairwise_cci_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes overall CCI scores for each pair of cells.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions.compute_pairwise_cci_scores--parameters">Parameters</h6>
<p>cci_score : str, default=None
    Scoring function to aggregate the communication scores between
    a pair of cells. It computes an overall potential of cell-cell
    interactions. If None, it will use the one stored in the
    attribute analysis_setup of this object.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score.
- &#39;jaccard&#39; : Jaccard-like score.
- &#39;count&#39; : Number of LR pairs that the pair of cells use.
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes overall CCI scores for each pair of cells.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cci_score : str, default=None</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        attribute analysis_setup of this object.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells use.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="n">cci_score</span><span class="o">=</span><span class="n">cci_score</span><span class="p">,</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>                                                       <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>                                                       <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions.compute_pairwise_communication_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes the communication scores for each LR pairs in
a given pair of sender-receiver cell</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.BulkInteractions.compute_pairwise_communication_scores--parameters">Parameters</h6>
<p>communication_score : str, default=None
    Type of communication score to infer the potential use of
    a given ligand-receptor pair by a pair of cells/tissues/samples.
    If None, the score stored in the attribute analysis_setup
    will be used.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expresion_thresholding&#39; : Computes the joint presence of a
                             ligand from a sender cell and of
                             a receptor on a receiver cell from
                             binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression
                        of a ligand from a sender cell and the
                        expression of a receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>ref_ppi_data : pandas.DataFrame, default=None
    Reference list of protein-protein interactions (or
    ligand-receptor pairs) used for inferring the cell-cell
    interactions and communication. It could be the same as
    'ppi_data' if ppi_data is not bidirectional (that is,
    contains ProtA-ProtB interaction as well as ProtB-ProtA
    interaction). ref_ppi must be undirected (contains only
    ProtA-ProtB and not ProtB-ProtA interaction). If None
    the one stored in the attribute ref_ppi will be used.</p>
<p>interaction_columns : tuple, default=None
    Contains the names of the columns where to find the
    partners in a dataframe of protein-protein interactions.
    If the list is for ligand-receptor pairs, the first column
    is for the ligands and the second for the receptors. If
    None, the one stored in the attribute interaction_columns
    will be used.</p>
<p>cells : list=None
    List of cells to consider.</p>
<p>cci_type : str, default=None
    Type of interaction between two cells. Used to specify
    if we want to consider a LR pair in both directions.
    It can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">undirected</span><span class="s1">&#39;</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">directed</span><span class="s1">&#39;</span>

<span class="k">If</span> <span class="nv">None</span>, <span class="s1">&#39;</span><span class="s">directed</span><span class="s1">&#39;</span> <span class="nv">will</span> <span class="nv">be</span> <span class="nv">used</span>.
</code></pre></div>

<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Computes the communication scores for each LR pairs in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a given pair of sender-receiver cell</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    communication_score : str, default=None</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;expresion_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">                                     ligand from a sender cell and of</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">                                     a receptor on a receiver cell from</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">                                     binarizing their gene expression levels.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                                of a ligand from a sender cell and the</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                                expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    ref_ppi_data : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Reference list of protein-protein interactions (or</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        ligand-receptor pairs) used for inferring the cell-cell</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        interactions and communication. It could be the same as</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        &#39;ppi_data&#39; if ppi_data is not bidirectional (that is,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        contains ProtA-ProtB interaction as well as ProtB-ProtA</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        interaction). ref_ppi must be undirected (contains only</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        ProtA-ProtB and not ProtB-ProtA interaction). If None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        the one stored in the attribute ref_ppi will be used.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    interaction_columns : tuple, default=None</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Contains the names of the columns where to find the</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        partners in a dataframe of protein-protein interactions.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        If the list is for ligand-receptor pairs, the first column</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        is for the ligands and the second for the receptors. If</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        None, the one stored in the attribute interaction_columns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    cells : list=None</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        List of cells to consider.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    cci_type : str, default=None</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Type of interaction between two cells. Used to specify</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        if we want to consider a LR pair in both directions.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        - &#39;undirected&#39;</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        - &#39;directed&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        If None, &#39;directed&#39; will be used.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">interaction_columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">interaction_columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span> <span class="c1"># Used only for ref_ppi_data</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">ref_ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="n">ref_ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="k">if</span> <span class="n">cci_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="n">cci_type</span> <span class="o">=</span> <span class="s1">&#39;directed&#39;</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                                 <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                                                                 <span class="n">ref_ppi_data</span><span class="o">=</span><span class="n">ref_ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>                                                                 <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                                                                 <span class="n">cells</span><span class="o">=</span><span class="n">cells</span><span class="p">,</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                                                                 <span class="n">cci_type</span><span class="o">=</span><span class="n">cci_type</span><span class="p">,</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>                                                                 <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions">
        <code>
SingleCellInteractions        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Interaction class with all necessary methods to run the cell2cell pipeline
    on a single-cell RNA-seq dataset.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame or scanpy.AnnData
    Gene expression data for a single-cell RNA-seq experiment. If it is a
    dataframe columns are single cells and rows are genes, while if it is
    a AnnData object, columns are genes and rows are single cells.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>metadata : pandas.Dataframe
    Metadata containing the cell types for each single cells in the
    RNA-seq dataset.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>communication_score : str, default='expression_thresholding'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39; : Computes the joint presence of a ligand from a
                             sender cell and of a receptor on a receiver cell
                             from binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
</code></pre></div>

<p>cci_score : str, default='bray_curtis'
    Scoring function to aggregate the communication scores between a pair of
    cells. It computes an overall potential of cell-cell interactions.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score.
- &#39;jaccard&#39; : Jaccard-like score.
- &#39;count&#39; : Number of LR pairs that the pair of cells use.
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>cci_type : str, default='undirected'
    Specifies whether computing the cci_score in a directed or undirected
    way. For a pair of cells A and B, directed means that the ligands are
    considered only from cell A and receptors only from cell B or viceversa.
    While undirected simultaneously considers signaling from cell A to
    cell B and from cell B to cell A.</p>
<p>expression_threshold : float, default=0.2
    Threshold value to binarize gene expression when using
    communication_score='expression_thresholding'. Units have to be the
    same as the aggregated gene expression matrix (e.g., counts, fraction
    of cells with non-zero counts, etc.).</p>
<p>aggregation_method : str, default='nn_cell_fraction'
    Specifies the method to use to aggregate gene expression of single
    cells into their respective cell types. Used to perform the CCI
    analysis since it is on the cell types rather than single cells.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">nn_cell_fraction</span><span class="s1">&#39;</span> : <span class="nv">Among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span>, <span class="nv">it</span>
    <span class="nv">calculates</span> <span class="nv">the</span> <span class="nv">fraction</span> <span class="nv">of</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">with</span> <span class="nv">non</span><span class="o">-</span><span class="nv">zero</span> <span class="nv">count</span> <span class="nv">values</span>
    <span class="nv">of</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">average</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">gene</span> <span class="nv">expression</span> <span class="nv">among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span>
    <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
</code></pre></div>

<p>barcode_col : str, default='barcodes'
    Column-name for the single cells in the metadata.</p>
<p>celltype_col : str, default='celltypes'
    Column-name in the metadata for the grouping single cells into cell types
    by the selected aggregation method.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions--attributes">Attributes</h6>
<p>rnaseq_data : pandas.DataFrame or scanpy.AnnData
    Gene expression data for a single-cell RNA-seq experiment. If it is a
    dataframe columns are single cells and rows are genes, while if it is
    a AnnData object, columns are genes and rows are single cells.</p>
<p>metadata : pandas.DataFrame
    Metadata containing the cell types for each single cells in the
    RNA-seq dataset.</p>
<p>index_col : str
    Column-name for the single cells in the metadata.</p>
<p>group_col : str
    Column-name in the metadata for the grouping single cells into cell types
    by the selected aggregation method.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used for
    inferring the cell-cell interactions and communication.</p>
<p>complex_sep : str
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>ref_ppi : pandas.DataFrame
    Reference list of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication. It could be the
    same as 'ppi_data' if ppi_data is not bidirectional (that is, contains
    ProtA-ProtB interaction as well as ProtB-ProtA interaction). ref_ppi must
    be undirected (contains only ProtA-ProtB and not ProtB-ProtA interaction).</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>analysis_setup : dict
    Contains main setup for running the cell-cell interactions and communication
    analyses.
    Three main setups are needed (passed as keys):</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">communication_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">communication</span><span class="w"> </span><span class="n">score</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">detect</span><span class="w"></span>
<span class="w">    </span><span class="n">active</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">pair</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">cell</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_thresholding</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_product</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_mean</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_gmean</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">scoring</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">communication</span><span class="w"></span>
<span class="w">    </span><span class="n">scores</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">bray_curtis</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">jaccard</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">count</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">icellnet</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_type</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interaction</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">two</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"> </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"></span>
<span class="w">    </span><span class="n">undirected</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">directed</span><span class="p">,</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">other</span><span class="w"></span>
<span class="w">    </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">separately</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="n">respect</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">second</span><span class="w"></span>
<span class="w">    </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptor</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">first</span><span class="w"> </span><span class="n">one</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">So</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">undirected</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">directed</span><span class="p">&#39;</span><span class="w"></span>
</code></pre></div>

<p>cutoff_setup : dict
    Contains two keys: 'type' and 'parameter'. The first key represent the
    way to use a cutoff or threshold, while parameter is the value used
    to binarize the expression values.
    The key 'type' can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"></span>
<span class="w">    </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"></span>
<span class="w">    </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"></span>
<span class="w">    </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">    </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"></span>
<span class="w">    </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"></span>
<span class="w">    </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">    </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">    </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all the required elements to perform the
    cell-cell interaction and communication analysis between every pair of cells.
    After performing the analyses, the results are stored in this object.</p>
<p>aggregation_method : str
    Specifies the method to use to aggregate gene expression of single
    cells into their respective cell types. Used to perform the CCI
    analysis since it is on the cell types rather than single cells.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">nn_cell_fraction</span><span class="s1">&#39;</span> : <span class="nv">Among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span>, <span class="nv">it</span>
    <span class="nv">calculates</span> <span class="nv">the</span> <span class="nv">fraction</span> <span class="nv">of</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">with</span> <span class="nv">non</span><span class="o">-</span><span class="nv">zero</span> <span class="nv">count</span> <span class="nv">values</span>
    <span class="nv">of</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">average</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">gene</span> <span class="nv">expression</span> <span class="nv">among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span>
    <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
</code></pre></div>

<p>ccc_permutation_pvalues : pandas.DataFrame
    Contains the P-values of the permutation analysis on the
    communication scores.</p>
<p>cci_permutation_pvalues : pandas.DataFrame
    Contains the P-values of the permutation analysis on the
    CCI scores.</p>
<p>__adata : boolean
    Auxiliary variable used for storing whether rnaseq_data
    is an AnnData object.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">SingleCellInteractions</span><span class="p">:</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Interaction class with all necessary methods to run the cell2cell pipeline</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">        on a single-cell RNA-seq dataset.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame or scanpy.AnnData</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for a single-cell RNA-seq experiment. If it is a</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        dataframe columns are single cells and rows are genes, while if it is</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        a AnnData object, columns are genes and rows are single cells.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    metadata : pandas.Dataframe</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Metadata containing the cell types for each single cells in the</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        RNA-seq dataset.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    communication_score : str, default=&#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        - &#39;expression_thresholding&#39; : Computes the joint presence of a ligand from a</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                                     sender cell and of a receptor on a receiver cell</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">                                     from binarizing their gene expression levels.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    cci_score : str, default=&#39;bray_curtis&#39;</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Scoring function to aggregate the communication scores between a pair of</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        cells. It computes an overall potential of cell-cell interactions.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells use.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    cci_type : str, default=&#39;undirected&#39;</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Specifies whether computing the cci_score in a directed or undirected</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        way. For a pair of cells A and B, directed means that the ligands are</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        considered only from cell A and receptors only from cell B or viceversa.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        While undirected simultaneously considers signaling from cell A to</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        cell B and from cell B to cell A.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    expression_threshold : float, default=0.2</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        Threshold value to binarize gene expression when using</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        communication_score=&#39;expression_thresholding&#39;. Units have to be the</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        same as the aggregated gene expression matrix (e.g., counts, fraction</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        of cells with non-zero counts, etc.).</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    aggregation_method : str, default=&#39;nn_cell_fraction&#39;</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Specifies the method to use to aggregate gene expression of single</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        cells into their respective cell types. Used to perform the CCI</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        analysis since it is on the cell types rather than single cells.</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">        - &#39;nn_cell_fraction&#39; : Among the single cells composing a cell type, it</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">            calculates the fraction of single cells with non-zero count values</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">            of a given gene.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">        - &#39;average&#39; : Computes the average gene expression among the single cells</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">            composing a cell type for a given gene.</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">    barcode_col : str, default=&#39;barcodes&#39;</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">        Column-name for the single cells in the metadata.</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">    celltype_col : str, default=&#39;celltypes&#39;</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        Column-name in the metadata for the grouping single cells into cell types</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        by the selected aggregation method.</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">    rnaseq_data : pandas.DataFrame or scanpy.AnnData</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">        Gene expression data for a single-cell RNA-seq experiment. If it is a</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        dataframe columns are single cells and rows are genes, while if it is</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">        a AnnData object, columns are genes and rows are single cells.</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">    metadata : pandas.DataFrame</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a><span class="sd">        Metadata containing the cell types for each single cells in the</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a><span class="sd">        RNA-seq dataset.</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a><span class="sd">    index_col : str</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a><span class="sd">        Column-name for the single cells in the metadata.</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a><span class="sd">    group_col : str</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a><span class="sd">        Column-name in the metadata for the grouping single cells into cell types</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a><span class="sd">        by the selected aggregation method.</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used for</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a><span class="sd">        inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a><span class="sd">    complex_sep : str</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a><span class="sd">    complex_agg_method : str</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a><span class="sd">    ref_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a><span class="sd">        Reference list of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a><span class="sd">        for inferring the cell-cell interactions and communication. It could be the</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a><span class="sd">        same as &#39;ppi_data&#39; if ppi_data is not bidirectional (that is, contains</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a><span class="sd">        ProtA-ProtB interaction as well as ProtB-ProtA interaction). ref_ppi must</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a><span class="sd">        be undirected (contains only ProtA-ProtB and not ProtB-ProtA interaction).</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a><span class="sd">    analysis_setup : dict</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a><span class="sd">        Contains main setup for running the cell-cell interactions and communication</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a><span class="sd">        analyses.</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a><span class="sd">        Three main setups are needed (passed as keys):</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a><span class="sd">        - &#39;communication_score&#39; : is the type of communication score used to detect</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a><span class="sd">            active ligand-receptor pairs between each pair of cell.</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a><span class="sd">            - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a><span class="sd">            - &#39;expression_product&#39;</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a><span class="sd">            - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a><span class="sd">            - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a><span class="sd">        - &#39;cci_score&#39; : is the scoring function to aggregate the communication</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a><span class="sd">            scores.</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a><span class="sd">            - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a><span class="sd">            - &#39;jaccard&#39;</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a><span class="sd">            - &#39;count&#39;</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a><span class="sd">            - &#39;icellnet&#39;</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a><span class="sd">        - &#39;cci_type&#39; : is the type of interaction between two cells. If it is</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a><span class="sd">            undirected, all ligands and receptors are considered from both cells.</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a><span class="sd">            If it is directed, ligands from one cell and receptors from the other</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a><span class="sd">            are considered separately with respect to ligands from the second</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a><span class="sd">            cell and receptor from the first one.</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a><span class="sd">            So, it can be:</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a><span class="sd">            - &#39;undirected&#39;</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a><span class="sd">            - &#39;directed&#39;</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a><span class="sd">    cutoff_setup : dict</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a><span class="sd">        Contains two keys: &#39;type&#39; and &#39;parameter&#39;. The first key represent the</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a><span class="sd">        way to use a cutoff or threshold, while parameter is the value used</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a><span class="sd">        to binarize the expression values.</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a><span class="sd">        The key &#39;type&#39; can be:</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a><span class="sd">        - &#39;local_percentile&#39; : computes the value of a given percentile, for each</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a><span class="sd">            gene independently. In this case, the parameter corresponds to the</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a><span class="sd">            percentile to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a><span class="sd">        - &#39;global_percentile&#39; : computes the value of a given percentile from all</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a><span class="sd">            genes and samples simultaneously. In this case, the parameter</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a><span class="sd">            corresponds to the percentile to compute, as a float value between</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a><span class="sd">            0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a><span class="sd">        - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is the</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a><span class="sd">            path of that file. It must contain the same genes as index and same</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a><span class="sd">            samples as columns.</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a><span class="sd">        - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a><span class="sd">            for each gene in each sample. This allows to use specific cutoffs for</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a><span class="sd">            each sample. The columns here must be the same as the ones in the</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a><span class="sd">            rnaseq_data.</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a><span class="sd">        - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a><span class="sd">            for each gene in only one column. These cutoffs will be applied to</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a><span class="sd">            all samples.</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a><span class="sd">        - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a><span class="sd">            expression is greater than the value input in the parameter.</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a><span class="sd">        Interaction space that contains all the required elements to perform the</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a><span class="sd">        cell-cell interaction and communication analysis between every pair of cells.</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a><span class="sd">        After performing the analyses, the results are stored in this object.</span>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a><span class="sd">    aggregation_method : str</span>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a><span class="sd">        Specifies the method to use to aggregate gene expression of single</span>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a><span class="sd">        cells into their respective cell types. Used to perform the CCI</span>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a><span class="sd">        analysis since it is on the cell types rather than single cells.</span>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a><span class="sd">        - &#39;nn_cell_fraction&#39; : Among the single cells composing a cell type, it</span>
<a id="__codelineno-0-221" name="__codelineno-0-221" href="#__codelineno-0-221"></a><span class="sd">            calculates the fraction of single cells with non-zero count values</span>
<a id="__codelineno-0-222" name="__codelineno-0-222" href="#__codelineno-0-222"></a><span class="sd">            of a given gene.</span>
<a id="__codelineno-0-223" name="__codelineno-0-223" href="#__codelineno-0-223"></a><span class="sd">        - &#39;average&#39; : Computes the average gene expression among the single cells</span>
<a id="__codelineno-0-224" name="__codelineno-0-224" href="#__codelineno-0-224"></a><span class="sd">            composing a cell type for a given gene.</span>
<a id="__codelineno-0-225" name="__codelineno-0-225" href="#__codelineno-0-225"></a>
<a id="__codelineno-0-226" name="__codelineno-0-226" href="#__codelineno-0-226"></a><span class="sd">    ccc_permutation_pvalues : pandas.DataFrame</span>
<a id="__codelineno-0-227" name="__codelineno-0-227" href="#__codelineno-0-227"></a><span class="sd">        Contains the P-values of the permutation analysis on the</span>
<a id="__codelineno-0-228" name="__codelineno-0-228" href="#__codelineno-0-228"></a><span class="sd">        communication scores.</span>
<a id="__codelineno-0-229" name="__codelineno-0-229" href="#__codelineno-0-229"></a>
<a id="__codelineno-0-230" name="__codelineno-0-230" href="#__codelineno-0-230"></a><span class="sd">    cci_permutation_pvalues : pandas.DataFrame</span>
<a id="__codelineno-0-231" name="__codelineno-0-231" href="#__codelineno-0-231"></a><span class="sd">        Contains the P-values of the permutation analysis on the</span>
<a id="__codelineno-0-232" name="__codelineno-0-232" href="#__codelineno-0-232"></a><span class="sd">        CCI scores.</span>
<a id="__codelineno-0-233" name="__codelineno-0-233" href="#__codelineno-0-233"></a>
<a id="__codelineno-0-234" name="__codelineno-0-234" href="#__codelineno-0-234"></a><span class="sd">    __adata : boolean</span>
<a id="__codelineno-0-235" name="__codelineno-0-235" href="#__codelineno-0-235"></a><span class="sd">        Auxiliary variable used for storing whether rnaseq_data</span>
<a id="__codelineno-0-236" name="__codelineno-0-236" href="#__codelineno-0-236"></a><span class="sd">        is an AnnData object.</span>
<a id="__codelineno-0-237" name="__codelineno-0-237" href="#__codelineno-0-237"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-238" name="__codelineno-0-238" href="#__codelineno-0-238"></a>    <span class="n">compute_pairwise_cci_scores</span> <span class="o">=</span> <span class="n">BulkInteractions</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span>
<a id="__codelineno-0-239" name="__codelineno-0-239" href="#__codelineno-0-239"></a>    <span class="n">compute_pairwise_communication_scores</span> <span class="o">=</span>  <span class="n">BulkInteractions</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span>
<a id="__codelineno-0-240" name="__codelineno-0-240" href="#__codelineno-0-240"></a>    <span class="n">interaction_elements</span> <span class="o">=</span> <span class="n">BulkInteractions</span><span class="o">.</span><span class="n">interaction_elements</span>
<a id="__codelineno-0-241" name="__codelineno-0-241" href="#__codelineno-0-241"></a>
<a id="__codelineno-0-242" name="__codelineno-0-242" href="#__codelineno-0-242"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-243" name="__codelineno-0-243" href="#__codelineno-0-243"></a>                 <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="s1">&#39;undirected&#39;</span><span class="p">,</span>
<a id="__codelineno-0-244" name="__codelineno-0-244" href="#__codelineno-0-244"></a>                 <span class="n">expression_threshold</span><span class="o">=</span><span class="mf">0.20</span><span class="p">,</span> <span class="n">aggregation_method</span><span class="o">=</span><span class="s1">&#39;nn_cell_fraction&#39;</span><span class="p">,</span> <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcodes&#39;</span><span class="p">,</span>
<a id="__codelineno-0-245" name="__codelineno-0-245" href="#__codelineno-0-245"></a>                 <span class="n">celltype_col</span><span class="o">=</span><span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-246" name="__codelineno-0-246" href="#__codelineno-0-246"></a>        <span class="c1"># Placeholders</span>
<a id="__codelineno-0-247" name="__codelineno-0-247" href="#__codelineno-0-247"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span>
<a id="__codelineno-0-248" name="__codelineno-0-248" href="#__codelineno-0-248"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span>
<a id="__codelineno-0-249" name="__codelineno-0-249" href="#__codelineno-0-249"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">index_col</span> <span class="o">=</span> <span class="n">barcode_col</span>
<a id="__codelineno-0-250" name="__codelineno-0-250" href="#__codelineno-0-250"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">group_col</span> <span class="o">=</span> <span class="n">celltype_col</span>
<a id="__codelineno-0-251" name="__codelineno-0-251" href="#__codelineno-0-251"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">aggregation_method</span> <span class="o">=</span> <span class="n">aggregation_method</span>
<a id="__codelineno-0-252" name="__codelineno-0-252" href="#__codelineno-0-252"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-253" name="__codelineno-0-253" href="#__codelineno-0-253"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-254" name="__codelineno-0-254" href="#__codelineno-0-254"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">complex_sep</span> <span class="o">=</span> <span class="n">complex_sep</span>
<a id="__codelineno-0-255" name="__codelineno-0-255" href="#__codelineno-0-255"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">complex_agg_method</span> <span class="o">=</span> <span class="n">complex_agg_method</span>
<a id="__codelineno-0-256" name="__codelineno-0-256" href="#__codelineno-0-256"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span> <span class="o">=</span> <span class="n">interaction_columns</span>
<a id="__codelineno-0-257" name="__codelineno-0-257" href="#__codelineno-0-257"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ccc_permutation_pvalues</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-258" name="__codelineno-0-258" href="#__codelineno-0-258"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cci_permutation_pvalues</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-259" name="__codelineno-0-259" href="#__codelineno-0-259"></a>
<a id="__codelineno-0-260" name="__codelineno-0-260" href="#__codelineno-0-260"></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">scanpy</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
<a id="__codelineno-0-261" name="__codelineno-0-261" href="#__codelineno-0-261"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">__adata</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-262" name="__codelineno-0-262" href="#__codelineno-0-262"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-263" name="__codelineno-0-263" href="#__codelineno-0-263"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-264" name="__codelineno-0-264" href="#__codelineno-0-264"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">__adata</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-265" name="__codelineno-0-265" href="#__codelineno-0-265"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-266" name="__codelineno-0-266" href="#__codelineno-0-266"></a>
<a id="__codelineno-0-267" name="__codelineno-0-267" href="#__codelineno-0-267"></a>        <span class="c1"># Analysis</span>
<a id="__codelineno-0-268" name="__codelineno-0-268" href="#__codelineno-0-268"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;communication_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">communication_score</span>
<a id="__codelineno-0-269" name="__codelineno-0-269" href="#__codelineno-0-269"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_score</span>
<a id="__codelineno-0-270" name="__codelineno-0-270" href="#__codelineno-0-270"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-271" name="__codelineno-0-271" href="#__codelineno-0-271"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-272" name="__codelineno-0-272" href="#__codelineno-0-272"></a>
<a id="__codelineno-0-273" name="__codelineno-0-273" href="#__codelineno-0-273"></a>        <span class="c1"># Initialize PPI</span>
<a id="__codelineno-0-274" name="__codelineno-0-274" href="#__codelineno-0-274"></a>        <span class="n">ppi_data_</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-275" name="__codelineno-0-275" href="#__codelineno-0-275"></a>                                               <span class="n">proteins</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
<a id="__codelineno-0-276" name="__codelineno-0-276" href="#__codelineno-0-276"></a>                                               <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-277" name="__codelineno-0-277" href="#__codelineno-0-277"></a>                                               <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-278" name="__codelineno-0-278" href="#__codelineno-0-278"></a>                                               <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-279" name="__codelineno-0-279" href="#__codelineno-0-279"></a>
<a id="__codelineno-0-280" name="__codelineno-0-280" href="#__codelineno-0-280"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data_</span><span class="p">,</span>
<a id="__codelineno-0-281" name="__codelineno-0-281" href="#__codelineno-0-281"></a>                                                        <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-282" name="__codelineno-0-282" href="#__codelineno-0-282"></a>                                                        <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-283" name="__codelineno-0-283" href="#__codelineno-0-283"></a>
<a id="__codelineno-0-284" name="__codelineno-0-284" href="#__codelineno-0-284"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-285" name="__codelineno-0-285" href="#__codelineno-0-285"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span>
<a id="__codelineno-0-286" name="__codelineno-0-286" href="#__codelineno-0-286"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-287" name="__codelineno-0-287" href="#__codelineno-0-287"></a>                                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-288" name="__codelineno-0-288" href="#__codelineno-0-288"></a>                                                          <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-289" name="__codelineno-0-289" href="#__codelineno-0-289"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-290" name="__codelineno-0-290" href="#__codelineno-0-290"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-291" name="__codelineno-0-291" href="#__codelineno-0-291"></a>
<a id="__codelineno-0-292" name="__codelineno-0-292" href="#__codelineno-0-292"></a>        <span class="c1"># Thresholding</span>
<a id="__codelineno-0-293" name="__codelineno-0-293" href="#__codelineno-0-293"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;constant_value&#39;</span>
<a id="__codelineno-0-294" name="__codelineno-0-294" href="#__codelineno-0-294"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">[</span><span class="s1">&#39;parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">expression_threshold</span>
<a id="__codelineno-0-295" name="__codelineno-0-295" href="#__codelineno-0-295"></a>
<a id="__codelineno-0-296" name="__codelineno-0-296" href="#__codelineno-0-296"></a>
<a id="__codelineno-0-297" name="__codelineno-0-297" href="#__codelineno-0-297"></a>        <span class="c1"># Aggregate single-cell RNA-Seq data</span>
<a id="__codelineno-0-298" name="__codelineno-0-298" href="#__codelineno-0-298"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">aggregated_expression</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">aggregate_single_cells</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-299" name="__codelineno-0-299" href="#__codelineno-0-299"></a>                                                                   <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
<a id="__codelineno-0-300" name="__codelineno-0-300" href="#__codelineno-0-300"></a>                                                                   <span class="n">barcode_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index_col</span><span class="p">,</span>
<a id="__codelineno-0-301" name="__codelineno-0-301" href="#__codelineno-0-301"></a>                                                                   <span class="n">celltype_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-302" name="__codelineno-0-302" href="#__codelineno-0-302"></a>                                                                   <span class="n">method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">aggregation_method</span><span class="p">,</span>
<a id="__codelineno-0-303" name="__codelineno-0-303" href="#__codelineno-0-303"></a>                                                                   <span class="n">transposed</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">__adata</span><span class="p">)</span>
<a id="__codelineno-0-304" name="__codelineno-0-304" href="#__codelineno-0-304"></a>
<a id="__codelineno-0-305" name="__codelineno-0-305" href="#__codelineno-0-305"></a>        <span class="c1"># Interaction Space</span>
<a id="__codelineno-0-306" name="__codelineno-0-306" href="#__codelineno-0-306"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span> <span class="o">=</span> <span class="n">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">aggregated_expression</span><span class="p">,</span>
<a id="__codelineno-0-307" name="__codelineno-0-307" href="#__codelineno-0-307"></a>                                                              <span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-308" name="__codelineno-0-308" href="#__codelineno-0-308"></a>                                                              <span class="n">cutoff_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-309" name="__codelineno-0-309" href="#__codelineno-0-309"></a>                                                              <span class="n">analysis_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">,</span>
<a id="__codelineno-0-310" name="__codelineno-0-310" href="#__codelineno-0-310"></a>                                                              <span class="n">complex_sep</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-311" name="__codelineno-0-311" href="#__codelineno-0-311"></a>                                                              <span class="n">complex_agg_method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-312" name="__codelineno-0-312" href="#__codelineno-0-312"></a>                                                              <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-313" name="__codelineno-0-313" href="#__codelineno-0-313"></a>                                                              <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-314" name="__codelineno-0-314" href="#__codelineno-0-314"></a>
<a id="__codelineno-0-315" name="__codelineno-0-315" href="#__codelineno-0-315"></a>    <span class="k">def</span> <span class="nf">permute_cell_labels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">evaluation</span><span class="o">=</span><span class="s1">&#39;communication&#39;</span><span class="p">,</span> <span class="n">fdr_correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-316" name="__codelineno-0-316" href="#__codelineno-0-316"></a>                            <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-317" name="__codelineno-0-317" href="#__codelineno-0-317"></a>        <span class="sd">&#39;&#39;&#39;Performs permutation analysis of cell-type labels. Detects</span>
<a id="__codelineno-0-318" name="__codelineno-0-318" href="#__codelineno-0-318"></a><span class="sd">        significant CCI or communication scores.</span>
<a id="__codelineno-0-319" name="__codelineno-0-319" href="#__codelineno-0-319"></a>
<a id="__codelineno-0-320" name="__codelineno-0-320" href="#__codelineno-0-320"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-321" name="__codelineno-0-321" href="#__codelineno-0-321"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-322" name="__codelineno-0-322" href="#__codelineno-0-322"></a><span class="sd">        permutations : int, default=100</span>
<a id="__codelineno-0-323" name="__codelineno-0-323" href="#__codelineno-0-323"></a><span class="sd">            Number of permutations where in each of them a random</span>
<a id="__codelineno-0-324" name="__codelineno-0-324" href="#__codelineno-0-324"></a><span class="sd">            shuffle of cell-type labels is performed, followed of</span>
<a id="__codelineno-0-325" name="__codelineno-0-325" href="#__codelineno-0-325"></a><span class="sd">            computing CCI or communication scores to create a null</span>
<a id="__codelineno-0-326" name="__codelineno-0-326" href="#__codelineno-0-326"></a><span class="sd">            distribution.</span>
<a id="__codelineno-0-327" name="__codelineno-0-327" href="#__codelineno-0-327"></a>
<a id="__codelineno-0-328" name="__codelineno-0-328" href="#__codelineno-0-328"></a><span class="sd">        evaluation : str, default=&#39;communication&#39;</span>
<a id="__codelineno-0-329" name="__codelineno-0-329" href="#__codelineno-0-329"></a><span class="sd">            Whether calculating P-values for CCI or communication scores.</span>
<a id="__codelineno-0-330" name="__codelineno-0-330" href="#__codelineno-0-330"></a>
<a id="__codelineno-0-331" name="__codelineno-0-331" href="#__codelineno-0-331"></a><span class="sd">            - &#39;interactions&#39; : For CCI scores.</span>
<a id="__codelineno-0-332" name="__codelineno-0-332" href="#__codelineno-0-332"></a><span class="sd">            - &#39;communication&#39; : For communication scores.</span>
<a id="__codelineno-0-333" name="__codelineno-0-333" href="#__codelineno-0-333"></a>
<a id="__codelineno-0-334" name="__codelineno-0-334" href="#__codelineno-0-334"></a><span class="sd">        fdr_correction : boolean, default=True</span>
<a id="__codelineno-0-335" name="__codelineno-0-335" href="#__codelineno-0-335"></a><span class="sd">            Whether performing a multiple test correction after</span>
<a id="__codelineno-0-336" name="__codelineno-0-336" href="#__codelineno-0-336"></a><span class="sd">            computing P-values. In this case corresponds to an</span>
<a id="__codelineno-0-337" name="__codelineno-0-337" href="#__codelineno-0-337"></a><span class="sd">            FDR or Benjamini-Hochberg correction, using an alpha</span>
<a id="__codelineno-0-338" name="__codelineno-0-338" href="#__codelineno-0-338"></a><span class="sd">            equal to 0.05.</span>
<a id="__codelineno-0-339" name="__codelineno-0-339" href="#__codelineno-0-339"></a>
<a id="__codelineno-0-340" name="__codelineno-0-340" href="#__codelineno-0-340"></a><span class="sd">        random_state : int, default=None</span>
<a id="__codelineno-0-341" name="__codelineno-0-341" href="#__codelineno-0-341"></a><span class="sd">            Seed for randomization.</span>
<a id="__codelineno-0-342" name="__codelineno-0-342" href="#__codelineno-0-342"></a>
<a id="__codelineno-0-343" name="__codelineno-0-343" href="#__codelineno-0-343"></a><span class="sd">        verbose : boolean, default=False</span>
<a id="__codelineno-0-344" name="__codelineno-0-344" href="#__codelineno-0-344"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-345" name="__codelineno-0-345" href="#__codelineno-0-345"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-346" name="__codelineno-0-346" href="#__codelineno-0-346"></a>        <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-347" name="__codelineno-0-347" href="#__codelineno-0-347"></a>            <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-348" name="__codelineno-0-348" href="#__codelineno-0-348"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_communication_scores() before permutation analysis.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-349" name="__codelineno-0-349" href="#__codelineno-0-349"></a>            <span class="n">score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-350" name="__codelineno-0-350" href="#__codelineno-0-350"></a>        <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-351" name="__codelineno-0-351" href="#__codelineno-0-351"></a>            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-352" name="__codelineno-0-352" href="#__codelineno-0-352"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_interactions() before permutation analysis.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-353" name="__codelineno-0-353" href="#__codelineno-0-353"></a>            <span class="n">score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-354" name="__codelineno-0-354" href="#__codelineno-0-354"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-355" name="__codelineno-0-355" href="#__codelineno-0-355"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not a valid evaluation&#39;</span><span class="p">)</span>
<a id="__codelineno-0-356" name="__codelineno-0-356" href="#__codelineno-0-356"></a>
<a id="__codelineno-0-357" name="__codelineno-0-357" href="#__codelineno-0-357"></a>        <span class="n">randomized_scores</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-358" name="__codelineno-0-358" href="#__codelineno-0-358"></a>
<a id="__codelineno-0-359" name="__codelineno-0-359" href="#__codelineno-0-359"></a>        <span class="n">analysis_setup</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-360" name="__codelineno-0-360" href="#__codelineno-0-360"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span>
<a id="__codelineno-0-361" name="__codelineno-0-361" href="#__codelineno-0-361"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-362" name="__codelineno-0-362" href="#__codelineno-0-362"></a>            <span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span>
<a id="__codelineno-0-363" name="__codelineno-0-363" href="#__codelineno-0-363"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;directed&#39;</span><span class="p">:</span>
<a id="__codelineno-0-364" name="__codelineno-0-364" href="#__codelineno-0-364"></a>                <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-365" name="__codelineno-0-365" href="#__codelineno-0-365"></a>                                                         <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-366" name="__codelineno-0-366" href="#__codelineno-0-366"></a>                                                         <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-367" name="__codelineno-0-367" href="#__codelineno-0-367"></a>            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-368" name="__codelineno-0-368" href="#__codelineno-0-368"></a>                <span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span>
<a id="__codelineno-0-369" name="__codelineno-0-369" href="#__codelineno-0-369"></a>
<a id="__codelineno-0-370" name="__codelineno-0-370" href="#__codelineno-0-370"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">permutations</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="ow">not</span> <span class="n">verbose</span><span class="p">):</span>
<a id="__codelineno-0-371" name="__codelineno-0-371" href="#__codelineno-0-371"></a>            <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-372" name="__codelineno-0-372" href="#__codelineno-0-372"></a>                <span class="n">seed</span> <span class="o">=</span> <span class="n">random_state</span> <span class="o">+</span> <span class="n">i</span>
<a id="__codelineno-0-373" name="__codelineno-0-373" href="#__codelineno-0-373"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-374" name="__codelineno-0-374" href="#__codelineno-0-374"></a>                <span class="n">seed</span> <span class="o">=</span> <span class="n">random_state</span>
<a id="__codelineno-0-375" name="__codelineno-0-375" href="#__codelineno-0-375"></a>
<a id="__codelineno-0-376" name="__codelineno-0-376" href="#__codelineno-0-376"></a>            <span class="n">randomized_meta</span> <span class="o">=</span> <span class="n">manipulate_dataframes</span><span class="o">.</span><span class="n">shuffle_cols_in_df</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
<a id="__codelineno-0-377" name="__codelineno-0-377" href="#__codelineno-0-377"></a>                                                                       <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-378" name="__codelineno-0-378" href="#__codelineno-0-378"></a>                                                                       <span class="n">random_state</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
<a id="__codelineno-0-379" name="__codelineno-0-379" href="#__codelineno-0-379"></a>
<a id="__codelineno-0-380" name="__codelineno-0-380" href="#__codelineno-0-380"></a>            <span class="n">aggregated_expression</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">aggregate_single_cells</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-381" name="__codelineno-0-381" href="#__codelineno-0-381"></a>                                                                  <span class="n">metadata</span><span class="o">=</span><span class="n">randomized_meta</span><span class="p">,</span>
<a id="__codelineno-0-382" name="__codelineno-0-382" href="#__codelineno-0-382"></a>                                                                  <span class="n">barcode_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index_col</span><span class="p">,</span>
<a id="__codelineno-0-383" name="__codelineno-0-383" href="#__codelineno-0-383"></a>                                                                  <span class="n">celltype_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-384" name="__codelineno-0-384" href="#__codelineno-0-384"></a>                                                                  <span class="n">method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">aggregation_method</span><span class="p">,</span>
<a id="__codelineno-0-385" name="__codelineno-0-385" href="#__codelineno-0-385"></a>                                                                  <span class="n">transposed</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">__adata</span><span class="p">)</span>
<a id="__codelineno-0-386" name="__codelineno-0-386" href="#__codelineno-0-386"></a>
<a id="__codelineno-0-387" name="__codelineno-0-387" href="#__codelineno-0-387"></a>            <span class="n">interaction_space</span> <span class="o">=</span> <span class="n">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">aggregated_expression</span><span class="p">,</span>
<a id="__codelineno-0-388" name="__codelineno-0-388" href="#__codelineno-0-388"></a>                                                             <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-389" name="__codelineno-0-389" href="#__codelineno-0-389"></a>                                                             <span class="n">cutoff_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-390" name="__codelineno-0-390" href="#__codelineno-0-390"></a>                                                             <span class="n">analysis_setup</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">,</span>
<a id="__codelineno-0-391" name="__codelineno-0-391" href="#__codelineno-0-391"></a>                                                             <span class="n">complex_sep</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-392" name="__codelineno-0-392" href="#__codelineno-0-392"></a>                                                             <span class="n">complex_agg_method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-393" name="__codelineno-0-393" href="#__codelineno-0-393"></a>                                                             <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-394" name="__codelineno-0-394" href="#__codelineno-0-394"></a>                                                             <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-395" name="__codelineno-0-395" href="#__codelineno-0-395"></a>
<a id="__codelineno-0-396" name="__codelineno-0-396" href="#__codelineno-0-396"></a>            <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-397" name="__codelineno-0-397" href="#__codelineno-0-397"></a>                <span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-398" name="__codelineno-0-398" href="#__codelineno-0-398"></a>                <span class="n">randomized_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<a id="__codelineno-0-399" name="__codelineno-0-399" href="#__codelineno-0-399"></a>            <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-400" name="__codelineno-0-400" href="#__codelineno-0-400"></a>                <span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-401" name="__codelineno-0-401" href="#__codelineno-0-401"></a>                <span class="n">randomized_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<a id="__codelineno-0-402" name="__codelineno-0-402" href="#__codelineno-0-402"></a>
<a id="__codelineno-0-403" name="__codelineno-0-403" href="#__codelineno-0-403"></a>        <span class="n">randomized_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">randomized_scores</span><span class="p">)</span>
<a id="__codelineno-0-404" name="__codelineno-0-404" href="#__codelineno-0-404"></a>        <span class="n">base_scores</span> <span class="o">=</span> <span class="n">score</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<a id="__codelineno-0-405" name="__codelineno-0-405" href="#__codelineno-0-405"></a>        <span class="n">pvals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">base_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-406" name="__codelineno-0-406" href="#__codelineno-0-406"></a>        <span class="n">n_pvals</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">base_scores</span><span class="p">)</span>
<a id="__codelineno-0-407" name="__codelineno-0-407" href="#__codelineno-0-407"></a>        <span class="n">randomized_scores</span> <span class="o">=</span> <span class="n">randomized_scores</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_pvals</span><span class="p">))</span>
<a id="__codelineno-0-408" name="__codelineno-0-408" href="#__codelineno-0-408"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_pvals</span><span class="p">):</span>
<a id="__codelineno-0-409" name="__codelineno-0-409" href="#__codelineno-0-409"></a>            <span class="n">dist</span> <span class="o">=</span> <span class="n">randomized_scores</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-410" name="__codelineno-0-410" href="#__codelineno-0-410"></a>            <span class="n">dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dist</span><span class="p">,</span> <span class="n">base_scores</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<a id="__codelineno-0-411" name="__codelineno-0-411" href="#__codelineno-0-411"></a>            <span class="n">pvals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">permutation</span><span class="o">.</span><span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="o">=</span><span class="n">base_scores</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-412" name="__codelineno-0-412" href="#__codelineno-0-412"></a>                                                            <span class="n">dist</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span>
<a id="__codelineno-0-413" name="__codelineno-0-413" href="#__codelineno-0-413"></a>                                                            <span class="n">consider_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-414" name="__codelineno-0-414" href="#__codelineno-0-414"></a>                                                            <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;different&#39;</span>
<a id="__codelineno-0-415" name="__codelineno-0-415" href="#__codelineno-0-415"></a>                                                            <span class="p">)</span>
<a id="__codelineno-0-416" name="__codelineno-0-416" href="#__codelineno-0-416"></a>        <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pvals</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">score</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">score</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">score</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-417" name="__codelineno-0-417" href="#__codelineno-0-417"></a>
<a id="__codelineno-0-418" name="__codelineno-0-418" href="#__codelineno-0-418"></a>        <span class="k">if</span> <span class="n">fdr_correction</span><span class="p">:</span>
<a id="__codelineno-0-419" name="__codelineno-0-419" href="#__codelineno-0-419"></a>            <span class="n">symmetric</span> <span class="o">=</span> <span class="n">manipulate_dataframes</span><span class="o">.</span><span class="n">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">pval_df</span><span class="p">)</span>
<a id="__codelineno-0-420" name="__codelineno-0-420" href="#__codelineno-0-420"></a>            <span class="k">if</span> <span class="n">symmetric</span><span class="p">:</span>
<a id="__codelineno-0-421" name="__codelineno-0-421" href="#__codelineno-0-421"></a>                <span class="n">pval_df</span> <span class="o">=</span> <span class="n">multitest</span><span class="o">.</span><span class="n">compute_fdrcorrection_symmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">pval_df</span><span class="p">,</span>
<a id="__codelineno-0-422" name="__codelineno-0-422" href="#__codelineno-0-422"></a>                                                                           <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<a id="__codelineno-0-423" name="__codelineno-0-423" href="#__codelineno-0-423"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-424" name="__codelineno-0-424" href="#__codelineno-0-424"></a>                <span class="n">pval_df</span> <span class="o">=</span> <span class="n">multitest</span><span class="o">.</span><span class="n">compute_fdrcorrection_asymmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">pval_df</span><span class="p">,</span>
<a id="__codelineno-0-425" name="__codelineno-0-425" href="#__codelineno-0-425"></a>                                                                            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<a id="__codelineno-0-426" name="__codelineno-0-426" href="#__codelineno-0-426"></a>
<a id="__codelineno-0-427" name="__codelineno-0-427" href="#__codelineno-0-427"></a>        <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-428" name="__codelineno-0-428" href="#__codelineno-0-428"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ccc_permutation_pvalues</span> <span class="o">=</span> <span class="n">pval_df</span>
<a id="__codelineno-0-429" name="__codelineno-0-429" href="#__codelineno-0-429"></a>        <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-430" name="__codelineno-0-430" href="#__codelineno-0-430"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">cci_permutation_pvalues</span> <span class="o">=</span> <span class="n">pval_df</span>
<a id="__codelineno-0-431" name="__codelineno-0-431" href="#__codelineno-0-431"></a>        <span class="k">return</span> <span class="n">pval_df</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">






  <div class="doc doc-object doc-attribute">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.interaction_elements">
<code class="highlight language-python"><span class="n">interaction_elements</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-property"><code>property</code></small>
      <small class="doc doc-property doc-property-readonly"><code>readonly</code></small>
  </span>

</h5>

    <div class="doc doc-contents ">

      <p>Returns the interaction elements within an interaction space.</p>
    </div>

  </div>







  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.compute_pairwise_cci_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes overall CCI scores for each pair of cells.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.compute_pairwise_cci_scores--parameters">Parameters</h6>
<p>cci_score : str, default=None
    Scoring function to aggregate the communication scores between
    a pair of cells. It computes an overall potential of cell-cell
    interactions. If None, it will use the one stored in the
    attribute analysis_setup of this object.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score.
- &#39;jaccard&#39; : Jaccard-like score.
- &#39;count&#39; : Number of LR pairs that the pair of cells use.
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes overall CCI scores for each pair of cells.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cci_score : str, default=None</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        attribute analysis_setup of this object.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells use.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="n">cci_score</span><span class="o">=</span><span class="n">cci_score</span><span class="p">,</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>                                                       <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>                                                       <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.compute_pairwise_communication_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes the communication scores for each LR pairs in
a given pair of sender-receiver cell</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.compute_pairwise_communication_scores--parameters">Parameters</h6>
<p>communication_score : str, default=None
    Type of communication score to infer the potential use of
    a given ligand-receptor pair by a pair of cells/tissues/samples.
    If None, the score stored in the attribute analysis_setup
    will be used.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expresion_thresholding&#39; : Computes the joint presence of a
                             ligand from a sender cell and of
                             a receptor on a receiver cell from
                             binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression
                        of a ligand from a sender cell and the
                        expression of a receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>ref_ppi_data : pandas.DataFrame, default=None
    Reference list of protein-protein interactions (or
    ligand-receptor pairs) used for inferring the cell-cell
    interactions and communication. It could be the same as
    'ppi_data' if ppi_data is not bidirectional (that is,
    contains ProtA-ProtB interaction as well as ProtB-ProtA
    interaction). ref_ppi must be undirected (contains only
    ProtA-ProtB and not ProtB-ProtA interaction). If None
    the one stored in the attribute ref_ppi will be used.</p>
<p>interaction_columns : tuple, default=None
    Contains the names of the columns where to find the
    partners in a dataframe of protein-protein interactions.
    If the list is for ligand-receptor pairs, the first column
    is for the ligands and the second for the receptors. If
    None, the one stored in the attribute interaction_columns
    will be used.</p>
<p>cells : list=None
    List of cells to consider.</p>
<p>cci_type : str, default=None
    Type of interaction between two cells. Used to specify
    if we want to consider a LR pair in both directions.
    It can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">undirected</span><span class="s1">&#39;</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">directed</span><span class="s1">&#39;</span>

<span class="k">If</span> <span class="nv">None</span>, <span class="s1">&#39;</span><span class="s">directed</span><span class="s1">&#39;</span> <span class="nv">will</span> <span class="nv">be</span> <span class="nv">used</span>.
</code></pre></div>

<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Computes the communication scores for each LR pairs in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a given pair of sender-receiver cell</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    communication_score : str, default=None</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;expresion_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">                                     ligand from a sender cell and of</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">                                     a receptor on a receiver cell from</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">                                     binarizing their gene expression levels.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                                of a ligand from a sender cell and the</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                                expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    ref_ppi_data : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Reference list of protein-protein interactions (or</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        ligand-receptor pairs) used for inferring the cell-cell</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        interactions and communication. It could be the same as</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        &#39;ppi_data&#39; if ppi_data is not bidirectional (that is,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        contains ProtA-ProtB interaction as well as ProtB-ProtA</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        interaction). ref_ppi must be undirected (contains only</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        ProtA-ProtB and not ProtB-ProtA interaction). If None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        the one stored in the attribute ref_ppi will be used.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    interaction_columns : tuple, default=None</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Contains the names of the columns where to find the</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        partners in a dataframe of protein-protein interactions.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        If the list is for ligand-receptor pairs, the first column</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        is for the ligands and the second for the receptors. If</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        None, the one stored in the attribute interaction_columns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    cells : list=None</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        List of cells to consider.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    cci_type : str, default=None</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Type of interaction between two cells. Used to specify</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        if we want to consider a LR pair in both directions.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        - &#39;undirected&#39;</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        - &#39;directed&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        If None, &#39;directed&#39; will be used.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">interaction_columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">interaction_columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span> <span class="c1"># Used only for ref_ppi_data</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">ref_ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="n">ref_ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="k">if</span> <span class="n">cci_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="n">cci_type</span> <span class="o">=</span> <span class="s1">&#39;directed&#39;</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                                 <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                                                                 <span class="n">ref_ppi_data</span><span class="o">=</span><span class="n">ref_ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>                                                                 <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                                                                 <span class="n">cells</span><span class="o">=</span><span class="n">cells</span><span class="p">,</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                                                                 <span class="n">cci_type</span><span class="o">=</span><span class="n">cci_type</span><span class="p">,</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>                                                                 <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.permute_cell_labels">
<code class="highlight language-python"><span class="n">permute_cell_labels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">evaluation</span><span class="o">=</span><span class="s1">&#39;communication&#39;</span><span class="p">,</span> <span class="n">fdr_correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Performs permutation analysis of cell-type labels. Detects
significant CCI or communication scores.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions.permute_cell_labels--parameters">Parameters</h6>
<p>permutations : int, default=100
    Number of permutations where in each of them a random
    shuffle of cell-type labels is performed, followed of
    computing CCI or communication scores to create a null
    distribution.</p>
<p>evaluation : str, default='communication'
    Whether calculating P-values for CCI or communication scores.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">interactions</span><span class="s1">&#39;</span> : <span class="k">For</span> <span class="nv">CCI</span> <span class="nv">scores</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">communication</span><span class="s1">&#39;</span> : <span class="k">For</span> <span class="nv">communication</span> <span class="nv">scores</span>.
</code></pre></div>

<p>fdr_correction : boolean, default=True
    Whether performing a multiple test correction after
    computing P-values. In this case corresponds to an
    FDR or Benjamini-Hochberg correction, using an alpha
    equal to 0.05.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">permute_cell_labels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">evaluation</span><span class="o">=</span><span class="s1">&#39;communication&#39;</span><span class="p">,</span> <span class="n">fdr_correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                        <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Performs permutation analysis of cell-type labels. Detects</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    significant CCI or communication scores.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    permutations : int, default=100</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Number of permutations where in each of them a random</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        shuffle of cell-type labels is performed, followed of</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        computing CCI or communication scores to create a null</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        distribution.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    evaluation : str, default=&#39;communication&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Whether calculating P-values for CCI or communication scores.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;interactions&#39; : For CCI scores.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;communication&#39; : For communication scores.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    fdr_correction : boolean, default=True</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Whether performing a multiple test correction after</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        computing P-values. In this case corresponds to an</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        FDR or Benjamini-Hochberg correction, using an alpha</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        equal to 0.05.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_communication_scores() before permutation analysis.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_interactions() before permutation analysis.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not a valid evaluation&#39;</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">randomized_scores</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">analysis_setup</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;ccc_type&#39;</span><span class="p">]</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;directed&#39;</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>                                                     <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                                                     <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>            <span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_ppi</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">permutations</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="ow">not</span> <span class="n">verbose</span><span class="p">):</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>            <span class="n">seed</span> <span class="o">=</span> <span class="n">random_state</span> <span class="o">+</span> <span class="n">i</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>            <span class="n">seed</span> <span class="o">=</span> <span class="n">random_state</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">randomized_meta</span> <span class="o">=</span> <span class="n">manipulate_dataframes</span><span class="o">.</span><span class="n">shuffle_cols_in_df</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                                                                   <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                                                                   <span class="n">random_state</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="n">aggregated_expression</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">aggregate_single_cells</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                                                              <span class="n">metadata</span><span class="o">=</span><span class="n">randomized_meta</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                                                              <span class="n">barcode_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index_col</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                                              <span class="n">celltype_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                                              <span class="n">method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">aggregation_method</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                                              <span class="n">transposed</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">__adata</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">interaction_space</span> <span class="o">=</span> <span class="n">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">aggregated_expression</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                         <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                         <span class="n">cutoff_setup</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                                         <span class="n">analysis_setup</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">,</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                                                         <span class="n">complex_sep</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                                                         <span class="n">complex_agg_method</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                         <span class="n">interaction_columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                                                         <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>            <span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>            <span class="n">randomized_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>            <span class="n">interaction_space</span><span class="o">.</span><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>            <span class="n">randomized_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">randomized_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">randomized_scores</span><span class="p">)</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="n">base_scores</span> <span class="o">=</span> <span class="n">score</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">base_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">n_pvals</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">base_scores</span><span class="p">)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="n">randomized_scores</span> <span class="o">=</span> <span class="n">randomized_scores</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_pvals</span><span class="p">))</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_pvals</span><span class="p">):</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>        <span class="n">dist</span> <span class="o">=</span> <span class="n">randomized_scores</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="n">dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dist</span><span class="p">,</span> <span class="n">base_scores</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="n">pvals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">permutation</span><span class="o">.</span><span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="o">=</span><span class="n">base_scores</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>                                                        <span class="n">dist</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                                                        <span class="n">consider_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                                                        <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;different&#39;</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                                                        <span class="p">)</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pvals</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">score</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">score</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">score</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="k">if</span> <span class="n">fdr_correction</span><span class="p">:</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>        <span class="n">symmetric</span> <span class="o">=</span> <span class="n">manipulate_dataframes</span><span class="o">.</span><span class="n">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">pval_df</span><span class="p">)</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>        <span class="k">if</span> <span class="n">symmetric</span><span class="p">:</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>            <span class="n">pval_df</span> <span class="o">=</span> <span class="n">multitest</span><span class="o">.</span><span class="n">compute_fdrcorrection_symmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">pval_df</span><span class="p">,</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>                                                                       <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>            <span class="n">pval_df</span> <span class="o">=</span> <span class="n">multitest</span><span class="o">.</span><span class="n">compute_fdrcorrection_asymmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">pval_df</span><span class="p">,</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>                                                                        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ccc_permutation_pvalues</span> <span class="o">=</span> <span class="n">pval_df</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cci_permutation_pvalues</span> <span class="o">=</span> <span class="n">pval_df</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>    <span class="k">return</span> <span class="n">pval_df</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>





  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.cell2cell_pipelines.initialize_interaction_space">
<code class="highlight language-python"><span class="n">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">cutoff_setup</span><span class="p">,</span> <span class="n">analysis_setup</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Initializes a InteractionSpace object to perform the analyses</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.initialize_interaction_space--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are samples
    and rows are genes.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>cutoff_setup : dict
    Contains two keys: 'type' and 'parameter'. The first key represent the
    way to use a cutoff or threshold, while parameter is the value used
    to binarize the expression values.
    The key 'type' can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"></span>
<span class="w">    </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"></span>
<span class="w">    </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"></span>
<span class="w">    </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">    </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"></span>
<span class="w">    </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"></span>
<span class="w">    </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">    </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">    </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>analysis_setup : dict
    Contains main setup for running the cell-cell interactions and communication
    analyses.
    Three main setups are needed (passed as keys):</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">communication_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">communication</span><span class="w"> </span><span class="n">score</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">detect</span><span class="w"></span>
<span class="w">    </span><span class="n">active</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">pair</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">cell</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_thresholding</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_product</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_mean</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_gmean</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">scoring</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">communication</span><span class="w"></span>
<span class="w">    </span><span class="n">scores</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">bray_curtis</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">jaccard</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">count</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">icellnet</span><span class="p">&#39;</span><span class="w"></span>

<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_type</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interaction</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">two</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"> </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"></span>
<span class="w">    </span><span class="n">undirected</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">directed</span><span class="p">,</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">other</span><span class="w"></span>
<span class="w">    </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">separately</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="n">respect</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">second</span><span class="w"></span>
<span class="w">    </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptor</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">first</span><span class="w"> </span><span class="n">one</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">So</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">undirected</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">directed</span><span class="p">&#39;</span><span class="w"></span>
</code></pre></div>

<p>excluded_cells : list, default=None
    List of cells in the rnaseq_data to be excluded. If None, all cells
    are considered.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.analysis.cell2cell_pipelines.initialize_interaction_space--returns">Returns</h6>
<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all the required elements to perform the
    cell-cell interaction and communication analysis between every pair of cells.
    After performing the analyses, the results are stored in this object.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/cell2cell_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">initialize_interaction_space</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">cutoff_setup</span><span class="p">,</span> <span class="n">analysis_setup</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                 <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                                 <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Initializes a InteractionSpace object to perform the analyses</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        experiment after aggregation into cell types. Columns are samples</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        and rows are genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    cutoff_setup : dict</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Contains two keys: &#39;type&#39; and &#39;parameter&#39;. The first key represent the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        way to use a cutoff or threshold, while parameter is the value used</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        to binarize the expression values.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        The key &#39;type&#39; can be:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;local_percentile&#39; : computes the value of a given percentile, for each</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">            gene independently. In this case, the parameter corresponds to the</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">            percentile to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        - &#39;global_percentile&#39; : computes the value of a given percentile from all</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">            genes and samples simultaneously. In this case, the parameter</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">            corresponds to the percentile to compute, as a float value between</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">            0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is the</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">            path of that file. It must contain the same genes as index and same</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">            samples as columns.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">            for each gene in each sample. This allows to use specific cutoffs for</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">            each sample. The columns here must be the same as the ones in the</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">            rnaseq_data.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">            for each gene in only one column. These cutoffs will be applied to</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">            all samples.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">            expression is greater than the value input in the parameter.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    analysis_setup : dict</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Contains main setup for running the cell-cell interactions and communication</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        analyses.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Three main setups are needed (passed as keys):</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        - &#39;communication_score&#39; : is the type of communication score used to detect</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">            active ligand-receptor pairs between each pair of cell.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">            - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">            - &#39;expression_product&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">            - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">            - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        - &#39;cci_score&#39; : is the scoring function to aggregate the communication</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">            scores.</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">            - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">            - &#39;jaccard&#39;</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">            - &#39;count&#39;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">            - &#39;icellnet&#39;</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        - &#39;cci_type&#39; : is the type of interaction between two cells. If it is</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">            undirected, all ligands and receptors are considered from both cells.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">            If it is directed, ligands from one cell and receptors from the other</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">            are considered separately with respect to ligands from the second</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">            cell and receptor from the first one.</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">            So, it can be:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">            - &#39;undirected&#39;</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">            - &#39;directed&#39;</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">    excluded_cells : list, default=None</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        List of cells in the rnaseq_data to be excluded. If None, all cells</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        are considered.</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    -------</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">        Interaction space that contains all the required elements to perform the</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">        cell-cell interaction and communication analysis between every pair of cells.</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        After performing the analyses, the results are stored in this object.</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>        <span class="n">excluded_cells</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="n">included_cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">((</span><span class="nb">set</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">))))</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="n">interaction_space</span> <span class="o">=</span> <span class="n">ispace</span><span class="o">.</span><span class="n">InteractionSpace</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="p">[</span><span class="n">included_cells</span><span class="p">],</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>                                                <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>                                                <span class="n">gene_cutoffs</span><span class="o">=</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>                                                <span class="n">communication_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;communication_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>                                                <span class="n">cci_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>                                                <span class="n">cci_type</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">],</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>                                                <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>                                                <span class="n">complex_agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                                                <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>                                                <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>    <span class="k">return</span> <span class="n">interaction_space</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream">
        <code>tensor_downstream</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream.compute_gini_coefficients">
<code class="highlight language-python"><span class="n">compute_gini_coefficients</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes Gini coefficient on the distribution of edge weights
in each factor-specific cell-cell communication network. Factors
obtained from the tensor decomposition with Tensor-cell2cell.</p>
<h6 id="cell2cell.analysis.tensor_downstream.compute_gini_coefficients--parameters">Parameters</h6>
<p>result : any Tensor class in cell2cell.tensor.tensor or a dict
    Either a Tensor type or a dictionary which resulted from the tensor
    decomposition. If it is a dict, it should be the one in, for example,
    InteractionTensor.factors</p>
<p>sender_label : str
    Label for the dimension of sender cells. Usually found in
    InteractionTensor.order_labels</p>
<p>receiver_label : str
    Label for the dimension of receiver cells. Usually found in
    InteractionTensor.order_labels</p>
<h6 id="cell2cell.analysis.tensor_downstream.compute_gini_coefficients--returns">Returns</h6>
<p>gini_df : pandas.DataFrame
    Dataframe containing the Gini coefficient of each factor from
    a tensor decomposition. Calculated on the factor-specific
    cell-cell communication networks.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_downstream.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_gini_coefficients</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes Gini coefficient on the distribution of edge weights</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in each factor-specific cell-cell communication network. Factors</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    obtained from the tensor decomposition with Tensor-cell2cell.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    result : any Tensor class in cell2cell.tensor.tensor or a dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Either a Tensor type or a dictionary which resulted from the tensor</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        decomposition. If it is a dict, it should be the one in, for example,</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        InteractionTensor.factors</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    sender_label : str</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Label for the dimension of sender cells. Usually found in</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    receiver_label : str</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Label for the dimension of receiver cells. Usually found in</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    gini_df : pandas.DataFrame</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Dataframe containing the Gini coefficient of each factor from</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        a tensor decomposition. Calculated on the factor-specific</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        cell-cell communication networks.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s1">&#39;factors&#39;</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">factors</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;A tensor factorization must be run on the tensor before calling this function.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">pass</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;result is not of a valid type. It must be an InteractionTensor or a dict.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">factors</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">sender_label</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">receiver_label</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">ginis</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">factor_net</span> <span class="o">=</span> <span class="n">get_joint_loadings</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="p">,</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                                        <span class="n">dim1</span><span class="o">=</span><span class="n">sender_label</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                        <span class="n">dim2</span><span class="o">=</span><span class="n">receiver_label</span><span class="p">,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                        <span class="n">factor</span><span class="o">=</span><span class="n">f</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                                        <span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">gini</span> <span class="o">=</span> <span class="n">gini_coefficient</span><span class="p">(</span><span class="n">factor_net</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">ginis</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">f</span><span class="p">,</span> <span class="n">gini</span><span class="p">))</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">gini_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">ginis</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Factor&#39;</span><span class="p">,</span> <span class="s1">&#39;Gini&#39;</span><span class="p">])</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">return</span> <span class="n">gini_df</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream.flatten_factor_ccc_networks">
<code class="highlight language-python"><span class="n">flatten_factor_ccc_networks</span><span class="p">(</span><span class="n">networks</span><span class="p">,</span> <span class="n">orderby</span><span class="o">=</span><span class="s1">&#39;senders&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Flattens all adjacency matrices in the factor-specific
cell-cell communication networks. It generates a matrix
where rows are factors and columns are cell-cell pairs.</p>
<h6 id="cell2cell.analysis.tensor_downstream.flatten_factor_ccc_networks--parameters">Parameters</h6>
<p>networks : dict
    A dictionary containing a pandas.DataFrame for each of the factors
    (factor names are the keys of the dict). These dataframes are the
    adjacency matrices of the CCC networks.</p>
<p>orderby : str
    Order of the flatten cell-cell pairs. Options are 'senders' and
    'receivers'. 'senders' means to flatten the matrices in a way that
    all cell-cell pairs with a same sender cell are put next to each others.
    'receivers' means the same, but by considering the receiver cell instead.</p>
<h6 id="cell2cell.analysis.tensor_downstream.flatten_factor_ccc_networks--returns">Returns</h6>
<p>flatten_networks : pandas.DataFrame
    A dataframe wherein rows contains a factor-specific network. Columns are
    the directed cell-cell pairs.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_downstream.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">flatten_factor_ccc_networks</span><span class="p">(</span><span class="n">networks</span><span class="p">,</span> <span class="n">orderby</span><span class="o">=</span><span class="s1">&#39;senders&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Flattens all adjacency matrices in the factor-specific</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell-cell communication networks. It generates a matrix</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    where rows are factors and columns are cell-cell pairs.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    networks : dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        A dictionary containing a pandas.DataFrame for each of the factors</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        (factor names are the keys of the dict). These dataframes are the</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        adjacency matrices of the CCC networks.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    orderby : str</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Order of the flatten cell-cell pairs. Options are &#39;senders&#39; and</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        &#39;receivers&#39;. &#39;senders&#39; means to flatten the matrices in a way that</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        all cell-cell pairs with a same sender cell are put next to each others.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        &#39;receivers&#39; means the same, but by considering the receiver cell instead.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    flatten_networks : pandas.DataFrame</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        A dataframe wherein rows contains a factor-specific network. Columns are</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        the directed cell-cell pairs.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">senders</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">networks</span><span class="o">.</span><span class="n">values</span><span class="p">()]))</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">receivers</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">networks</span><span class="o">.</span><span class="n">values</span><span class="p">()]))</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="n">orderby</span> <span class="o">==</span> <span class="s1">&#39;senders&#39;</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">cell_pairs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="o">+</span> <span class="s1">&#39; --&gt; &#39;</span> <span class="o">+</span> <span class="n">r</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">senders</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">receivers</span><span class="p">]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">flatten_order</span> <span class="o">=</span> <span class="s1">&#39;C&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">elif</span> <span class="n">orderby</span> <span class="o">==</span> <span class="s1">&#39;receivers&#39;</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">cell_pairs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="o">+</span> <span class="s1">&#39; --&gt; &#39;</span> <span class="o">+</span> <span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">receivers</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">senders</span><span class="p">]</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="n">flatten_order</span> <span class="o">=</span> <span class="s1">&#39;F&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`orderby` must be either &#39;senders&#39; or &#39;receivers&#39;.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span><span class="n">v</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">flatten_order</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">networks</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">flatten_networks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>                                    <span class="n">index</span><span class="o">=</span><span class="n">cell_pairs</span><span class="p">,</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>                                    <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">networks</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>                                    <span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">return</span> <span class="n">flatten_networks</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream.get_factor_specific_ccc_networks">
<code class="highlight language-python"><span class="n">get_factor_specific_ccc_networks</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates adjacency matrices for each of the factors
obtained from a tensor decomposition. These matrices represent a
cell-cell communication directed network.</p>
<h6 id="cell2cell.analysis.tensor_downstream.get_factor_specific_ccc_networks--parameters">Parameters</h6>
<p>result : any Tensor class in cell2cell.tensor.tensor or a dict
    Either a Tensor type or a dictionary which resulted from the tensor
    decomposition. If it is a dict, it should be the one in, for example,
    InteractionTensor.factors</p>
<p>sender_label : str
    Label for the dimension of sender cells. Usually found in
    InteractionTensor.order_labels</p>
<p>receiver_label : str
    Label for the dimension of receiver cells. Usually found in
    InteractionTensor.order_labels</p>
<h6 id="cell2cell.analysis.tensor_downstream.get_factor_specific_ccc_networks--returns">Returns</h6>
<p>networks : dict
    A dictionary containing a pandas.DataFrame for each of the factors
    (factor names are the keys of the dict). These dataframes are the
    adjacency matrices of the CCC networks.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_downstream.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_factor_specific_ccc_networks</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Generates adjacency matrices for each of the factors</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    obtained from a tensor decomposition. These matrices represent a</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    cell-cell communication directed network.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    result : any Tensor class in cell2cell.tensor.tensor or a dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Either a Tensor type or a dictionary which resulted from the tensor</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        decomposition. If it is a dict, it should be the one in, for example,</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        InteractionTensor.factors</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    sender_label : str</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Label for the dimension of sender cells. Usually found in</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    receiver_label : str</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Label for the dimension of receiver cells. Usually found in</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    networks : dict</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        A dictionary containing a pandas.DataFrame for each of the factors</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        (factor names are the keys of the dict). These dataframes are the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        adjacency matrices of the CCC networks.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s1">&#39;factors&#39;</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">factors</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;A tensor factorization must be run on the tensor before calling this function.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">pass</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;result is not of a valid type. It must be an InteractionTensor or a dict.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">factors</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">sender_label</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">receiver_label</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">networks</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">networks</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_joint_loadings</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="p">,</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                                         <span class="n">dim1</span><span class="o">=</span><span class="n">sender_label</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                         <span class="n">dim2</span><span class="o">=</span><span class="n">receiver_label</span><span class="p">,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                         <span class="n">factor</span><span class="o">=</span><span class="n">f</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                                         <span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">return</span> <span class="n">networks</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream.get_joint_loadings">
<code class="highlight language-python"><span class="n">get_joint_loadings</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">dim1</span><span class="p">,</span> <span class="n">dim2</span><span class="p">,</span> <span class="n">factor</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Creates the joint loading distribution between two tensor dimensions for a
given factor output from decomposition.</p>
<h6 id="cell2cell.analysis.tensor_downstream.get_joint_loadings--parameters">Parameters</h6>
<p>result : any Tensor class in cell2cell.tensor.tensor or a dict
    Either a Tensor type or a dictionary which resulted from the tensor
    decomposition. If it is a dict, it should be the one in, for example,
    InteractionTensor.factors</p>
<p>dim1 : str
    One of the tensor dimensions (options are in the keys of the dict,
    or interaction.factors.keys())</p>
<p>dim2 : str
    A second tensor dimension (options are in the keys of the dict,
    or interaction.factors.keys())</p>
<div class="admonition factor">
<p class="admonition-title">str</p>
<p>One of the factors output from the decomposition (e.g. 'Factor 1').</p>
</div>
<h6 id="cell2cell.analysis.tensor_downstream.get_joint_loadings--returns">Returns</h6>
<p>joint_dist : pandas.DataFrame
    Joint distribution of factor loadings for the specified dimensions.
    Rows correspond to elements in dim1 and columns to elements in dim2.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_downstream.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_joint_loadings</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">dim1</span><span class="p">,</span> <span class="n">dim2</span><span class="p">,</span> <span class="n">factor</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Creates the joint loading distribution between two tensor dimensions for a</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    given factor output from decomposition.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    result : any Tensor class in cell2cell.tensor.tensor or a dict</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Either a Tensor type or a dictionary which resulted from the tensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        decomposition. If it is a dict, it should be the one in, for example,</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        InteractionTensor.factors</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    dim1 : str</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        One of the tensor dimensions (options are in the keys of the dict,</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        or interaction.factors.keys())</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    dim2 : str</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        A second tensor dimension (options are in the keys of the dict,</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        or interaction.factors.keys())</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    factor: str</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        One of the factors output from the decomposition (e.g. &#39;Factor 1&#39;).</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    joint_dist : pandas.DataFrame</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Joint distribution of factor loadings for the specified dimensions.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Rows correspond to elements in dim1 and columns to elements in dim2.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s1">&#39;factors&#39;</span><span class="p">):</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">factors</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;A tensor factorization must be run on the tensor before calling this function.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">pass</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;result is not of a valid type. It must be an InteractionTensor or a dict.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">assert</span> <span class="n">dim1</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="s1">&#39;The specified dimension &#39;</span> <span class="o">+</span> <span class="n">dim1</span> <span class="o">+</span> <span class="s1">&#39; is not present in the `result` input&#39;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">assert</span> <span class="n">dim2</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="s1">&#39;The specified dimension &#39;</span> <span class="o">+</span> <span class="n">dim2</span> <span class="o">+</span> <span class="s1">&#39; is not present in the `result` input&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">vec1</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">dim1</span><span class="p">][</span><span class="n">factor</span><span class="p">]</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">vec2</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">dim2</span><span class="p">][</span><span class="n">factor</span><span class="p">]</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="c1"># Calculate the outer product</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">joint_dist</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">vec1</span><span class="p">,</span> <span class="n">vec2</span><span class="p">),</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                              <span class="n">index</span><span class="o">=</span><span class="n">vec1</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>                              <span class="n">columns</span><span class="o">=</span><span class="n">vec2</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">joint_dist</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">dim1</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">joint_dist</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">dim2</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">return</span> <span class="n">joint_dist</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_downstream.get_lr_by_cell_pairs">
<code class="highlight language-python"><span class="n">get_lr_by_cell_pairs</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">lr_label</span><span class="p">,</span> <span class="n">sender_label</span><span class="p">,</span> <span class="n">receiver_label</span><span class="p">,</span> <span class="n">order_cells_by</span><span class="o">=</span><span class="s1">&#39;receivers&#39;</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lr_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Returns a dataframe containing the product loadings of a specific combination
of ligand-receptor pair and sender-receiver pair.</p>
<h6 id="cell2cell.analysis.tensor_downstream.get_lr_by_cell_pairs--parameters">Parameters</h6>
<p>result : any Tensor class in cell2cell.tensor.tensor or a dict
    Either a Tensor type or a dictionary which resulted from the tensor
    decomposition. If it is a dict, it should be the one in, for example,
    InteractionTensor.factors</p>
<p>lr_label : str
    Label for the dimension of the ligand-receptor pairs. Usually found in
    InteractionTensor.order_labels</p>
<p>sender_label : str
    Label for the dimension of sender cells. Usually found in
    InteractionTensor.order_labels</p>
<p>receiver_label : str
    Label for the dimension of receiver cells. Usually found in
    InteractionTensor.order_labels</p>
<p>order_cells_by : str, default='receivers'
    Order of the returned dataframe. Options are 'senders' and
    'receivers'. 'senders' means to order the dataframe in a way that
    all cell-cell pairs with a same sender cell are put next to each others.
    'receivers' means the same, but by considering the receiver cell instead.</p>
<p>factor : str, default=None
    Name of the factor to be used to compute the product loadings.
    If None, all factors will be included to compute them.</p>
<p>cci_threshold : float, default=None
    Threshold to be applied on the product loadings of the sender-cell pairs.
    If specified, only cell-cell pairs with a product loading above the
    threshold at least in one of the factors included will be included
    in the returned dataframe.</p>
<p>lr_threshold : float, default=None
    Threshold to be applied on the ligand-receptor loadings.
    If specified, only LR pairs with a loading above the
    threshold at least in one of the factors included will be included
    in the returned dataframe.</p>
<h6 id="cell2cell.analysis.tensor_downstream.get_lr_by_cell_pairs--returns">Returns</h6>
<p>cci_lr : pandas.DataFrame
    Dataframe containing the product loadings of a specific combination 
    of ligand-receptor pair and sender-receiver pair. If the factor is specified,
    the returned dataframe will contain the product loadings of that factor.
    If the factor is not specified, the returned dataframe will contain the
    product loadings across all factors.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_downstream.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_lr_by_cell_pairs</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">lr_label</span><span class="p">,</span> <span class="n">sender_label</span><span class="p">,</span> <span class="n">receiver_label</span><span class="p">,</span> <span class="n">order_cells_by</span><span class="o">=</span><span class="s1">&#39;receivers&#39;</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                         <span class="n">cci_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lr_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Returns a dataframe containing the product loadings of a specific combination</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    of ligand-receptor pair and sender-receiver pair.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    result : any Tensor class in cell2cell.tensor.tensor or a dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Either a Tensor type or a dictionary which resulted from the tensor</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        decomposition. If it is a dict, it should be the one in, for example,</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        InteractionTensor.factors</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    lr_label : str</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Label for the dimension of the ligand-receptor pairs. Usually found in</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    sender_label : str</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Label for the dimension of sender cells. Usually found in</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    receiver_label : str</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Label for the dimension of receiver cells. Usually found in</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        InteractionTensor.order_labels</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    order_cells_by : str, default=&#39;receivers&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Order of the returned dataframe. Options are &#39;senders&#39; and</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        &#39;receivers&#39;. &#39;senders&#39; means to order the dataframe in a way that</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        all cell-cell pairs with a same sender cell are put next to each others.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        &#39;receivers&#39; means the same, but by considering the receiver cell instead.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    factor : str, default=None</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Name of the factor to be used to compute the product loadings.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        If None, all factors will be included to compute them.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    cci_threshold : float, default=None</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Threshold to be applied on the product loadings of the sender-cell pairs.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        If specified, only cell-cell pairs with a product loading above the</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        threshold at least in one of the factors included will be included</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        in the returned dataframe.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    lr_threshold : float, default=None</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Threshold to be applied on the ligand-receptor loadings.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        If specified, only LR pairs with a loading above the</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        threshold at least in one of the factors included will be included</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        in the returned dataframe.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    -------</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    cci_lr : pandas.DataFrame</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Dataframe containing the product loadings of a specific combination </span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        of ligand-receptor pair and sender-receiver pair. If the factor is specified,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        the returned dataframe will contain the product loadings of that factor.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        If the factor is not specified, the returned dataframe will contain the</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        product loadings across all factors.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s1">&#39;factors&#39;</span><span class="p">):</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">factors</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;A tensor factorization must be run on the tensor before calling this function.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">pass</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;result is not of a valid type. It must be an InteractionTensor or a dict.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="k">assert</span> <span class="n">lr_label</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="s1">&#39;The specified dimension &#39;</span> <span class="o">+</span> <span class="n">lr_label</span> <span class="o">+</span> <span class="s1">&#39; is not present in the `result` input&#39;</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">assert</span> <span class="n">sender_label</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="s1">&#39;The specified dimension &#39;</span> <span class="o">+</span> <span class="n">sender_label</span> <span class="o">+</span> <span class="s1">&#39; is not present in the `result` input&#39;</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="k">assert</span> <span class="n">receiver_label</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="s1">&#39;The specified dimension &#39;</span> <span class="o">+</span> <span class="n">receiver_label</span> <span class="o">+</span> <span class="s1">&#39; is not present in the `result` input&#39;</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="c1"># Sort factors</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">sorted_factors</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">lr_label</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="c1"># Get CCI network per factor</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="n">networks</span> <span class="o">=</span> <span class="n">get_factor_specific_ccc_networks</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                <span class="n">sender_label</span><span class="o">=</span><span class="n">sender_label</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                                <span class="n">receiver_label</span><span class="o">=</span><span class="n">receiver_label</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="c1"># Flatten networks</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="n">network_by_factors</span> <span class="o">=</span> <span class="n">flatten_factor_ccc_networks</span><span class="p">(</span><span class="n">networks</span><span class="o">=</span><span class="n">networks</span><span class="p">,</span> <span class="n">orderby</span><span class="o">=</span><span class="n">order_cells_by</span><span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="c1"># Get final dataframe</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="n">df1</span> <span class="o">=</span> <span class="n">network_by_factors</span><span class="p">[</span><span class="n">sorted_factors</span><span class="p">]</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="n">df2</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">lr_label</span><span class="p">][</span><span class="n">sorted_factors</span><span class="p">]</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>    <span class="k">if</span> <span class="n">factor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="p">[</span><span class="n">factor</span><span class="p">]</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[</span><span class="n">factor</span><span class="p">]</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="k">if</span> <span class="n">cci_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>            <span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="p">[(</span><span class="n">df1</span> <span class="o">&gt;</span> <span class="n">cci_threshold</span><span class="p">)]</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>        <span class="k">if</span> <span class="n">lr_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>            <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[(</span><span class="n">df2</span> <span class="o">&gt;</span> <span class="n">lr_threshold</span><span class="p">)]</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">df1</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="k">if</span> <span class="n">cci_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>            <span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="p">[(</span><span class="n">df1</span><span class="o">.</span><span class="n">T</span> <span class="o">&gt;</span> <span class="n">cci_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()]</span>  <span class="c1"># Top sender-receiver pairs</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="k">if</span> <span class="n">lr_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>            <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[(</span><span class="n">df2</span><span class="o">.</span><span class="n">T</span> <span class="o">&gt;</span> <span class="n">lr_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()]</span>  <span class="c1"># Top LR Pairs</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="n">cci_lr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                          <span class="n">columns</span><span class="o">=</span><span class="n">df1</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>                          <span class="n">index</span><span class="o">=</span><span class="n">df2</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                          <span class="p">)</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">cci_lr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Sender-Receiver Pair&#39;</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="n">cci_lr</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Ligand-Receptor Pair&#39;</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="k">return</span> <span class="n">cci_lr</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.analysis.tensor_pipelines">
        <code>tensor_pipelines</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.analysis.tensor_pipelines.run_tensor_cell2cell_pipeline">
<code class="highlight language-python"><span class="n">run_tensor_cell2cell_pipeline</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">tensor_metadata</span><span class="p">,</span> <span class="n">copy_tensor</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tf_optimization</span><span class="o">=</span><span class="s1">&#39;regular&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">elbow_metric</span><span class="o">=</span><span class="s1">&#39;error&#39;</span><span class="p">,</span> <span class="n">smooth_elbow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">upper_rank</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">tf_init</span><span class="o">=</span><span class="s1">&#39;random&#39;</span><span class="p">,</span> <span class="n">tf_svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span> <span class="n">fig_fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">output_folder</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_fig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fig_format</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Runs basic pipeline of Tensor-cell2cell (excluding downstream analyses).</p>
<h6 id="cell2cell.analysis.tensor_pipelines.run_tensor_cell2cell_pipeline--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor.</p>
<p>tensor_metadata : list
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element.</p>
<p>copy_tensor : boolean, default=False
    Whether generating a copy of the original tensor to avoid modifying it.</p>
<p>rank : int, default=None
    Rank of the Tensor Factorization (number of factors to deconvolve the original
    tensor). If None, it will automatically inferred from an elbow analysis.</p>
<p>tf_optimization : str, default='regular'
    It defines whether performing an optimization with higher number of iterations,
    independent factorization runs, and higher resolution (lower tolerance),
    or with lower number of iterations, factorization runs, and resolution.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">regular</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">uses</span><span class="w"> </span><span class="mh">100</span><span class="w"> </span><span class="n">max</span><span class="w"> </span><span class="n">iterations</span><span class="p">,</span><span class="w"> </span><span class="mh">1</span><span class="w"> </span><span class="n">factorization</span><span class="w"> </span><span class="n">run</span><span class="p">,</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="mf">10e-7</span><span class="w"> </span><span class="n">tolerance</span><span class="p">.</span><span class="w"></span>
<span class="w">              </span><span class="n">Faster</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">run</span><span class="p">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">robust</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">uses</span><span class="w"> </span><span class="mh">500</span><span class="w"> </span><span class="n">max</span><span class="w"> </span><span class="n">iterations</span><span class="p">,</span><span class="w"> </span><span class="mh">100</span><span class="w"> </span><span class="n">factorization</span><span class="w"> </span><span class="n">runs</span><span class="p">,</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="mf">10e-8</span><span class="w"> </span><span class="n">tolerance</span><span class="p">.</span><span class="w"></span>
<span class="w">             </span><span class="n">Slower</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">run</span><span class="p">.</span><span class="w"></span>
</code></pre></div>

<p>random_state : boolean, default=None
    Seed for randomization.</p>
<p _cupy_="'cupy'," _jax_="'jax'," _mxnet_="'mxnet'," _numpy_="'numpy'," _pytorch_="'pytorch'," _tensorflow_="'tensorflow'">backend : str, default=None
    Backend that TensorLy will use to perform calculations
    on this tensor. When None, the default backend used is
    the currently active backend, usually is ('numpy'). Options are:</p>
<p None="None" _cpu_="'cpu'," _cuda:0_="'cuda:0',">device : str, default=None
    Device to use when backend allows multiple devices. Options are:</p>
<p>elbow_metric : str, default='error'
    Metric to perform the elbow analysis (y-axis).</p>
<div class="codehilite"><pre><span></span><code>    - &#39;error&#39; : Normalized error to compute the elbow.
    - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).
</code></pre></div>

<p>smooth_elbow : boolean, default=False
    Whether smoothing the elbow-analysis curve with a Savitzky-Golay filter.</p>
<p>upper_rank : int, default=25
    Upper bound of ranks to explore with the elbow analysis.</p>
<p _random_="random" _svd_="svd,">tf_init : str, default='random'
    Initialization method for computing the Tensor Factorization.</p>
<p>tf_svd : str, default='numpy_svd'
    Function to compute the SVD for initializing the Tensor Factorization,
    acceptable values in tensorly.SVD_FUNS</p>
<p>cmaps : list, default=None
    A list of colormaps used for coloring elements in each dimension. The length
    of this list is equal to the number of dimensions of the tensor. If None, all
    dimensions will be colores with the colormap 'gist_rainbow'.</p>
<p>sample_col : str, default='Element'
    Name of the column containing the element names in the metadata.</p>
<p>group_col : str, default='Category'
    Name of the column containing the metadata or grouping information for each
    element in the metadata.</p>
<p>fig_fontsize : int, default=14
    Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</p>
<p>output_folder : str, default=None
    Path to the folder where the figures generated will be saved.
    If None, figures will not be saved.</p>
<p>output_fig : boolean, default=True
    Whether generating the figures with matplotlib.</p>
<p>fig_format : str, default='pdf'
    Format to store figures when an <code>output_folder</code> is specified
    and <code>output_fig</code> is True. Otherwise, this is not necessary.</p>
<p>**kwargs : dict
        Extra arguments for the tensor factorization according to inputs in
        tensorly.</p>
<h6 id="cell2cell.analysis.tensor_pipelines.run_tensor_cell2cell_pipeline--returns">Returns</h6>
<p>interaction_tensor : cell2cell.tensor.tensor.BaseTensor
    Either the original input <code>interaction_tensor</code> or a copy of it.
    This also stores the results from running the Tensor-cell2cell
    pipeline in the corresponding attributes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/analysis/tensor_pipelines.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">run_tensor_cell2cell_pipeline</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">tensor_metadata</span><span class="p">,</span> <span class="n">copy_tensor</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                  <span class="n">tf_optimization</span><span class="o">=</span><span class="s1">&#39;regular&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                                  <span class="n">elbow_metric</span><span class="o">=</span><span class="s1">&#39;error&#39;</span><span class="p">,</span> <span class="n">smooth_elbow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">upper_rank</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">tf_init</span><span class="o">=</span><span class="s1">&#39;random&#39;</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                                  <span class="n">tf_svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>                                  <span class="n">fig_fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">output_folder</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_fig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fig_format</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Runs basic pipeline of Tensor-cell2cell (excluding downstream analyses).</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        cell2cell.tensor.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    tensor_metadata : list</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        element.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    copy_tensor : boolean, default=False</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether generating a copy of the original tensor to avoid modifying it.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    rank : int, default=None</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Rank of the Tensor Factorization (number of factors to deconvolve the original</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        tensor). If None, it will automatically inferred from an elbow analysis.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    tf_optimization : str, default=&#39;regular&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        It defines whether performing an optimization with higher number of iterations,</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        independent factorization runs, and higher resolution (lower tolerance),</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        or with lower number of iterations, factorization runs, and resolution.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        - &#39;regular&#39; : It uses 100 max iterations, 1 factorization run, and 10e-7 tolerance.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">                      Faster to run.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        - &#39;robust&#39; : It uses 500 max iterations, 100 factorization runs, and 10e-8 tolerance.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">                     Slower to run.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    random_state : boolean, default=None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    backend : str, default=None</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Backend that TensorLy will use to perform calculations</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        on this tensor. When None, the default backend used is</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        the currently active backend, usually is (&#39;numpy&#39;). Options are:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        {&#39;cupy&#39;, &#39;jax&#39;, &#39;mxnet&#39;, &#39;numpy&#39;, &#39;pytorch&#39;, &#39;tensorflow&#39;}</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        Device to use when backend allows multiple devices. Options are:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">         {&#39;cpu&#39;, &#39;cuda:0&#39;, None}</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    elbow_metric : str, default=&#39;error&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Metric to perform the elbow analysis (y-axis).</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">            - &#39;error&#39; : Normalized error to compute the elbow.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">            - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    smooth_elbow : boolean, default=False</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Whether smoothing the elbow-analysis curve with a Savitzky-Golay filter.</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">    upper_rank : int, default=25</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        Upper bound of ranks to explore with the elbow analysis.</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    tf_init : str, default=&#39;random&#39;</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        Initialization method for computing the Tensor Factorization.</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        {svd, random}</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    tf_svd : str, default=&#39;numpy_svd&#39;</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        Function to compute the SVD for initializing the Tensor Factorization,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        acceptable values in tensorly.SVD_FUNS</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    cmaps : list, default=None</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        A list of colormaps used for coloring elements in each dimension. The length</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        of this list is equal to the number of dimensions of the tensor. If None, all</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">        dimensions will be colores with the colormap &#39;gist_rainbow&#39;.</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">    sample_col : str, default=&#39;Element&#39;</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        Name of the column containing the element names in the metadata.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">    group_col : str, default=&#39;Category&#39;</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        Name of the column containing the metadata or grouping information for each</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        element in the metadata.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">    fig_fontsize : int, default=14</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">        Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">    output_folder : str, default=None</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">        Path to the folder where the figures generated will be saved.</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">        If None, figures will not be saved.</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">    output_fig : boolean, default=True</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        Whether generating the figures with matplotlib.</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">    fig_format : str, default=&#39;pdf&#39;</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        Format to store figures when an `output_folder` is specified</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">        and `output_fig` is True. Otherwise, this is not necessary.</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">            Extra arguments for the tensor factorization according to inputs in</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">            tensorly.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">    -------</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">    interaction_tensor : cell2cell.tensor.tensor.BaseTensor</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        Either the original input `interaction_tensor` or a copy of it.</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">        This also stores the results from running the Tensor-cell2cell</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a><span class="sd">        pipeline in the corresponding attributes.</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="k">if</span> <span class="n">copy_tensor</span><span class="p">:</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>        <span class="n">interaction_tensor</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="n">dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="c1">### OUTPUT FILENAMES ###</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">if</span> <span class="n">output_folder</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="n">elbow_filename</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="n">tf_filename</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>        <span class="n">loading_filename</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="n">elbow_filename</span> <span class="o">=</span> <span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/Elbow.</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fig_format</span><span class="p">)</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>        <span class="n">tf_filename</span> <span class="o">=</span> <span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/Tensor-Factorization.</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fig_format</span><span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="n">loading_filename</span> <span class="o">=</span> <span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/Loadings.xlsx&#39;</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>    <span class="c1">### PALETTE COLORS FOR ELEMENTS IN TENSOR DIMS ###</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">if</span> <span class="n">cmaps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="n">cmap_5d</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;tab10&#39;</span><span class="p">,</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="s1">&#39;Dark2_r&#39;</span><span class="p">,</span> <span class="s1">&#39;tab20&#39;</span><span class="p">,</span> <span class="s1">&#39;tab20&#39;</span><span class="p">]</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="n">cmap_4d</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;plasma&#39;</span><span class="p">,</span> <span class="s1">&#39;Dark2_r&#39;</span><span class="p">,</span> <span class="s1">&#39;tab20&#39;</span><span class="p">,</span> <span class="s1">&#39;tab20&#39;</span><span class="p">]</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="k">if</span> <span class="n">dim</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>            <span class="n">cmaps</span> <span class="o">=</span> <span class="n">cmap_5d</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="k">elif</span> <span class="n">dim</span> <span class="o">&lt;=</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>            <span class="n">cmaps</span> <span class="o">=</span> <span class="n">cmap_4d</span><span class="p">[</span><span class="o">-</span><span class="n">dim</span><span class="p">:]</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Tensor of dimension higher to 5 is not supported&#39;</span><span class="p">)</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">cmaps</span><span class="p">)</span> <span class="o">==</span> <span class="n">dim</span><span class="p">,</span> <span class="s2">&quot;`cmap` must be of the same len of dimensions in the tensor.&quot;</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>    <span class="c1">### FACTORIZATION PARAMETERS ###</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>    <span class="k">if</span> <span class="n">tf_optimization</span> <span class="o">==</span> <span class="s1">&#39;robust&#39;</span><span class="p">:</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>        <span class="n">elbow_runs</span> <span class="o">=</span> <span class="mi">20</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>        <span class="n">tf_runs</span> <span class="o">=</span> <span class="mi">100</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>        <span class="n">tol</span> <span class="o">=</span> <span class="mf">1e-8</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="n">n_iter_max</span> <span class="o">=</span> <span class="mi">500</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>    <span class="k">elif</span> <span class="n">tf_optimization</span> <span class="o">==</span> <span class="s1">&#39;regular&#39;</span><span class="p">:</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="n">elbow_runs</span> <span class="o">=</span> <span class="mi">10</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>        <span class="n">tf_runs</span> <span class="o">=</span> <span class="mi">1</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>        <span class="n">tol</span> <span class="o">=</span> <span class="mf">1e-7</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>        <span class="n">n_iter_max</span> <span class="o">=</span> <span class="mi">100</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`factorization_type` must be either &#39;robust&#39; or &#39;regular&#39;.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>    <span class="k">if</span> <span class="n">backend</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>        <span class="n">tl</span><span class="o">.</span><span class="n">set_backend</span><span class="p">(</span><span class="n">backend</span><span class="p">)</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>    <span class="k">if</span> <span class="n">device</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">to_device</span><span class="p">(</span><span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>    <span class="c1">### ANALYSIS ###</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>    <span class="c1"># Elbow</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>    <span class="k">if</span> <span class="n">rank</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running Elbow Analysis&#39;</span><span class="p">)</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="n">fig1</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">elbow_rank_selection</span><span class="p">(</span><span class="n">upper_rank</span><span class="o">=</span><span class="n">upper_rank</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>                                                              <span class="n">runs</span><span class="o">=</span><span class="n">elbow_runs</span><span class="p">,</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>                                                              <span class="n">init</span><span class="o">=</span><span class="n">tf_init</span><span class="p">,</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>                                                              <span class="n">svd</span><span class="o">=</span><span class="n">tf_svd</span><span class="p">,</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>                                                              <span class="n">automatic_elbow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>                                                              <span class="n">metric</span><span class="o">=</span><span class="n">elbow_metric</span><span class="p">,</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>                                                              <span class="n">output_fig</span><span class="o">=</span><span class="n">output_fig</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>                                                              <span class="n">smooth</span><span class="o">=</span><span class="n">smooth_elbow</span><span class="p">,</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                                                              <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>                                                              <span class="n">fontsize</span><span class="o">=</span><span class="n">fig_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                                                              <span class="n">filename</span><span class="o">=</span><span class="n">elbow_filename</span><span class="p">,</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>                                                              <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>                                                              <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>                                                              <span class="p">)</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="n">rank</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">rank</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>    <span class="c1"># Factorization</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running Tensor Factorization&#39;</span><span class="p">)</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>    <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">compute_tensor_factorization</span><span class="p">(</span><span class="n">rank</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                                                    <span class="n">init</span><span class="o">=</span><span class="n">tf_init</span><span class="p">,</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>                                                    <span class="n">svd</span><span class="o">=</span><span class="n">tf_svd</span><span class="p">,</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>                                                    <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>                                                    <span class="n">runs</span><span class="o">=</span><span class="n">tf_runs</span><span class="p">,</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                                                    <span class="n">normalize_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>                                                    <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                                                    <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>                                                    <span class="p">)</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>    <span class="c1">### EXPORT RESULTS ###</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>    <span class="k">if</span> <span class="n">output_folder</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Generating Outputs&#39;</span><span class="p">)</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>        <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">export_factor_loadings</span><span class="p">(</span><span class="n">loading_filename</span><span class="p">)</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>    <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>        <span class="n">fig2</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">tensor_factors_plot</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">=</span><span class="n">interaction_tensor</span><span class="p">,</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                                         <span class="n">metadata</span><span class="o">=</span><span class="n">tensor_metadata</span><span class="p">,</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>                                         <span class="n">sample_col</span><span class="o">=</span><span class="n">sample_col</span><span class="p">,</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>                                         <span class="n">group_col</span><span class="o">=</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>                                         <span class="n">meta_cmaps</span><span class="o">=</span><span class="n">cmaps</span><span class="p">,</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>                                         <span class="n">fontsize</span><span class="o">=</span><span class="n">fig_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>                                         <span class="n">filename</span><span class="o">=</span><span class="n">tf_filename</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>                                         <span class="p">)</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a>    <span class="k">return</span> <span class="n">interaction_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.clustering">
        <code>clustering</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.clustering.cluster_interactions">
        <code>cluster_interactions</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.clustering.cluster_interactions.compute_distance">
<code class="highlight language-python"><span class="n">compute_distance</span><span class="p">(</span><span class="n">data_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the pairwise distance between elements in a
matrix of shape m x n. Uses the function
scipy.spatial.distance.pdist</p>
<h6 id="cell2cell.clustering.cluster_interactions.compute_distance--parameters">Parameters</h6>
<p>data_matrix : pandas.DataFrame or ndarray
    A m x n matrix used to compute the distances</p>
<p>axis : int, default=0
    To decide on which elements to compute the distance.
    If axis=0, the distances will be between elements in
    the rows, while axis=1 will lead to distances between
    elements in the columns.</p>
<p>metric : str, default='euclidean'
    The distance metric to use. The distance function can be 'braycurtis',
    'canberra', 'chebyshev', 'cityblock', 'correlation', 'cosine', 'dice',
    'euclidean', 'hamming', 'jaccard', 'jensenshannon', 'kulsinski',
    'mahalanobis', 'matching', 'minkowski', 'rogerstanimoto', 'russellrao',
    'seuclidean', 'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule'.</p>
<h6 id="cell2cell.clustering.cluster_interactions.compute_distance--returns">Returns</h6>
<p>D : ndarray
    Returns a condensed distance matrix Y. For each i and j (where i &lt; j &lt; m),
    where m is the number of original observations. The metric
    dist(u=X[i], v=X[j]) is computed and stored in entry
    m * i + j - ((i + 2) * (i + 1)) // 2.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/clustering/cluster_interactions.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_distance</span><span class="p">(</span><span class="n">data_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the pairwise distance between elements in a</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    matrix of shape m x n. Uses the function</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    scipy.spatial.distance.pdist</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    data_matrix : pandas.DataFrame or ndarray</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        A m x n matrix used to compute the distances</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        To decide on which elements to compute the distance.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        If axis=0, the distances will be between elements in</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        the rows, while axis=1 will lead to distances between</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        elements in the columns.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    metric : str, default=&#39;euclidean&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        The distance metric to use. The distance function can be &#39;braycurtis&#39;,</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        &#39;canberra&#39;, &#39;chebyshev&#39;, &#39;cityblock&#39;, &#39;correlation&#39;, &#39;cosine&#39;, &#39;dice&#39;,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        &#39;euclidean&#39;, &#39;hamming&#39;, &#39;jaccard&#39;, &#39;jensenshannon&#39;, &#39;kulsinski&#39;,</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        &#39;mahalanobis&#39;, &#39;matching&#39;, &#39;minkowski&#39;, &#39;rogerstanimoto&#39;, &#39;russellrao&#39;,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        &#39;seuclidean&#39;, &#39;sokalmichener&#39;, &#39;sokalsneath&#39;, &#39;sqeuclidean&#39;, &#39;yule&#39;.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    D : ndarray</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Returns a condensed distance matrix Y. For each i and j (where i &lt; j &lt; m),</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        where m is the number of original observations. The metric</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        dist(u=X[i], v=X[j]) is computed and stored in entry</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        m * i + j - ((i + 2) * (i + 1)) // 2.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">data_matrix</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">data_matrix</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">data_matrix</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">D</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">pdist</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">elif</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">D</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">pdist</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not valid axis. Use 0 or 1.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">return</span> <span class="n">D</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.clustering.cluster_interactions.compute_linkage">
<code class="highlight language-python"><span class="n">compute_linkage</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Returns a linkage for a given distance matrix using a specific method.</p>
<h6 id="cell2cell.clustering.cluster_interactions.compute_linkage--parameters">Parameters</h6>
<p>distance_matrix : numpy.ndarray
    A square array containing the distance between a given row and a
    given column. Diagonal elements must be zero.</p>
<p>method : str, 'ward' by default
    Method to compute the linkage. It could be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;single&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;average&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;weighted&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;centroid&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;median&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;ward&#39;</span><span class="w"></span>
<span class="k">For</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">details</span><span class="p">,</span><span class="w"> </span><span class="k">go</span><span class="w"> </span><span class="k">to</span><span class="err">:</span><span class="w"></span>
<span class="nl">https</span><span class="p">:</span><span class="o">//</span><span class="n">docs</span><span class="p">.</span><span class="n">scipy</span><span class="p">.</span><span class="n">org</span><span class="o">/</span><span class="n">doc</span><span class="o">/</span><span class="n">scipy</span><span class="o">-</span><span class="mf">0.19.0</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">generated</span><span class="o">/</span><span class="n">scipy</span><span class="p">.</span><span class="n">cluster</span><span class="p">.</span><span class="n">hierarchy</span><span class="p">.</span><span class="n">linkage</span><span class="p">.</span><span class="n">html</span><span class="w"></span>
</code></pre></div>

<p>optimal_ordering : boolean, default=True
    Whether sorting the leaf of the dendrograms to have a minimal distance
    between successive leaves. For more information, see
    scipy.cluster.hierarchy.optimal_leaf_ordering</p>
<h6 id="cell2cell.clustering.cluster_interactions.compute_linkage--returns">Returns</h6>
<p>Z : numpy.ndarray
    The hierarchical clustering encoded as a linkage matrix.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/clustering/cluster_interactions.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_linkage</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Returns a linkage for a given distance matrix using a specific method.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    distance_matrix : numpy.ndarray</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A square array containing the distance between a given row and a</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        given column. Diagonal elements must be zero.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    method : str, &#39;ward&#39; by default</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Method to compute the linkage. It could be:</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        - &#39;single&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;complete&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;average&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;weighted&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;centroid&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;median&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        - &#39;ward&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        For more details, go to:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.cluster.hierarchy.linkage.html</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    optimal_ordering : boolean, default=True</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Whether sorting the leaf of the dendrograms to have a minimal distance</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        between successive leaves. For more information, see</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        scipy.cluster.hierarchy.optimal_leaf_ordering</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    -------</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    Z : numpy.ndarray</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        The hierarchical clustering encoded as a linkage matrix.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="o">~</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span> <span class="o">==</span> <span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;The matrix is not symmetric&#39;</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Compute linkage</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">D</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">Z</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">D</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_ordering</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">return</span> <span class="n">Z</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.clustering.cluster_interactions.get_clusters_from_linkage">
<code class="highlight language-python"><span class="n">get_clusters_from_linkage</span><span class="p">(</span><span class="n">linkage</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">criterion</span><span class="o">=</span><span class="s1">&#39;maxclust&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Gets clusters from a linkage given a threshold and a criterion.</p>
<h6 id="cell2cell.clustering.cluster_interactions.get_clusters_from_linkage--parameters">Parameters</h6>
<p>linkage : numpy.ndarray
    The hierarchical clustering encoded with the matrix returned by
    the linkage function (Z).</p>
<p>threshold : float
    The threshold to apply when forming flat clusters.</p>
<p>criterion : str, 'maxclust' by default
    The criterion to use in forming flat clusters. Depending on the
    criterion, the threshold has different meanings. More information on:
    https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.cluster.hierarchy.fcluster.html</p>
<p>labels : array-like, None by default
    List of labels of the elements contained in the linkage. The order
    must match the order they were provided when generating the linkage.</p>
<h6 id="cell2cell.clustering.cluster_interactions.get_clusters_from_linkage--returns">Returns</h6>
<p>clusters : dict
    A dictionary containing the clusters obtained. The keys correspond to
    the cluster numbers and the vaues to a list with element names given the
    labels, or the element index based on the linkage.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/clustering/cluster_interactions.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_clusters_from_linkage</span><span class="p">(</span><span class="n">linkage</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">criterion</span><span class="o">=</span><span class="s1">&#39;maxclust&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Gets clusters from a linkage given a threshold and a criterion.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    linkage : numpy.ndarray</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        The hierarchical clustering encoded with the matrix returned by</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the linkage function (Z).</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    threshold : float</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        The threshold to apply when forming flat clusters.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    criterion : str, &#39;maxclust&#39; by default</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        The criterion to use in forming flat clusters. Depending on the</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        criterion, the threshold has different meanings. More information on:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.cluster.hierarchy.fcluster.html</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    labels : array-like, None by default</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        List of labels of the elements contained in the linkage. The order</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        must match the order they were provided when generating the linkage.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    -------</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    clusters : dict</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        A dictionary containing the clusters obtained. The keys correspond to</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        the cluster numbers and the vaues to a list with element names given the</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        labels, or the element index based on the linkage.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">cluster_ids</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">fcluster</span><span class="p">(</span><span class="n">linkage</span><span class="p">,</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">criterion</span><span class="o">=</span><span class="n">criterion</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">clusters</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">cluster_ids</span><span class="p">):</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="n">clusters</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cluster_ids</span><span class="p">):</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>            <span class="n">clusters</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">labels</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="n">clusters</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">return</span> <span class="n">clusters</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.core">
        <code>core</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.core.cci_scores">
        <code>cci_scores</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.compute_braycurtis_like_cci_score">
<code class="highlight language-python"><span class="n">compute_braycurtis_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates a Bray-Curtis-like score for the interaction between
two cells based on their intercellular protein-protein
interactions such as ligand-receptor interactions.</p>
<h6 id="cell2cell.core.cci_scores.compute_braycurtis_like_cci_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the receiver.</p>
<h6 id="cell2cell.core.cci_scores.compute_braycurtis_like_cci_score--returns">Returns</h6>
<p>cci_score : float
    Overall score for the interaction between a pair of
    cell-types/tissues/samples. In this case is a
    Bray-Curtis-like score.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_braycurtis_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Calculates a Bray-Curtis-like score for the interaction between</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    two cells based on their intercellular protein-protein</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    interactions such as ligand-receptor interactions.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        First cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        this is the sender.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Second cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        this is the receiver.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    cci_score : float</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell-types/tissues/samples. In this case is a</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Bray-Curtis-like score.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># Bray Curtis similarity</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c1</span> <span class="o">*</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c2</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">denominator</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">cci_score</span> <span class="o">=</span> <span class="n">numerator</span> <span class="o">/</span> <span class="n">denominator</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">return</span> <span class="n">cci_score</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.compute_count_score">
<code class="highlight language-python"><span class="n">compute_count_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates the number of active protein-protein interactions
for the interaction between two cells, which could be the number
of active ligand-receptor interactions.</p>
<h6 id="cell2cell.core.cci_scores.compute_count_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the receiver.</p>
<h6 id="cell2cell.core.cci_scores.compute_count_score--returns">Returns</h6>
<p>cci_score : float
    Overall score for the interaction between a pair of
    cell-types/tissues/samples.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_count_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Calculates the number of active protein-protein interactions</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    for the interaction between two cells, which could be the number</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    of active ligand-receptor interactions.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        First cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        this is the sender.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Second cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        this is the receiver.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    cci_score : float</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">mult</span> <span class="o">=</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">cci_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">mult</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="c1"># Count all active pathways (different to zero)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">return</span> <span class="n">cci_score</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.compute_icellnet_score">
<code class="highlight language-python"><span class="n">compute_icellnet_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates the sum of communication scores
for the interaction between two cells. Based on ICELLNET.</p>
<h6 id="cell2cell.core.cci_scores.compute_icellnet_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the receiver.</p>
<h6 id="cell2cell.core.cci_scores.compute_icellnet_score--returns">Returns</h6>
<p>cci_score : float
    Overall score for the interaction between a pair of
    cell-types/tissues/samples.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_icellnet_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Calculates the sum of communication scores</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    for the interaction between two cells. Based on ICELLNET.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        First cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        this is the sender.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Second cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        this is the receiver.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    cci_score : float</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">mult</span> <span class="o">=</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">cci_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">mult</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">return</span> <span class="n">cci_score</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.compute_jaccard_like_cci_score">
<code class="highlight language-python"><span class="n">compute_jaccard_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates a Jaccard-like score for the interaction between
two cells based on their intercellular protein-protein
interactions such as ligand-receptor interactions.</p>
<h6 id="cell2cell.core.cci_scores.compute_jaccard_like_cci_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute interaction
    between a pair of them. In a directed interaction,
    this is the receiver.</p>
<h6 id="cell2cell.core.cci_scores.compute_jaccard_like_cci_score--returns">Returns</h6>
<p>cci_score : float
    Overall score for the interaction between a pair of
    cell-types/tissues/samples. In this case it is a
    Jaccard-like score.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_jaccard_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Calculates a Jaccard-like score for the interaction between</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    two cells based on their intercellular protein-protein</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    interactions such as ligand-receptor interactions.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        First cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        this is the sender.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Second cell-type/tissue/sample to compute interaction</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        between a pair of them. In a directed interaction,</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        this is the receiver.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    cci_score : float</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell-types/tissues/samples. In this case it is a</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Jaccard-like score.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># Extended Jaccard similarity</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c1</span> <span class="o">*</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">c2</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span><span class="p">)</span> <span class="o">-</span> <span class="n">numerator</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">denominator</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">cci_score</span> <span class="o">=</span> <span class="n">numerator</span> <span class="o">/</span> <span class="n">denominator</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">return</span> <span class="n">cci_score</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.matmul_bray_curtis_like">
<code class="highlight language-python"><span class="n">matmul_bray_curtis_like</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes Bray-Curtis-like scores using matrices of proteins by
cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_bray_curtis_like--parameters">Parameters</h6>
<p>A_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<p>B_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_bray_curtis_like--returns">Returns</h6>
<p>bray_curtis : numpy.array
    Matrix MxM, representing the CCI score for all pairs of
    cell-types/tissues/samples. In directed interactions,
    the vertical axis (axis 0) represents the senders, while
    the horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">matmul_bray_curtis_like</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes Bray-Curtis-like scores using matrices of proteins by</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    cell-types/tissues/samples.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    A_scores : array-like</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    B_scores : array-like</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    bray_curtis : numpy.array</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Matrix MxM, representing the CCI score for all pairs of</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples. In directed interactions,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        the vertical axis (axis 0) represents the senders, while</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        the horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="n">A_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">ppi_score</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">ppi_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span> <span class="n">B_scores</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">A_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">A_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">B_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">B_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">A_module</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">A_module</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span> <span class="o">+</span> <span class="n">B_module</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">bray_curtis</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="mf">2.0</span><span class="o">*</span><span class="n">numerator</span><span class="p">,</span> <span class="n">denominator</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">return</span> <span class="n">bray_curtis</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.matmul_cosine">
<code class="highlight language-python"><span class="n">matmul_cosine</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes cosine-similarity scores using matrices of proteins by
cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_cosine--parameters">Parameters</h6>
<p>A_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<p>B_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_cosine--returns">Returns</h6>
<p>cosine : numpy.array
    Matrix MxM, representing the CCI score for all pairs of
    cell-types/tissues/samples. In directed interactions,
    the vertical axis (axis 0) represents the senders, while
    the horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">matmul_cosine</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes cosine-similarity scores using matrices of proteins by</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    cell-types/tissues/samples.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    A_scores : array-like</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    B_scores : array-like</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    cosine : numpy.array</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Matrix MxM, representing the CCI score for all pairs of</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples. In directed interactions,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        the vertical axis (axis 0) represents the senders, while</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        the horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="n">A_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">ppi_score</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">ppi_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span> <span class="n">B_scores</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">A_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">A_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">**</span> <span class="mf">0.5</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">B_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">B_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">**</span> <span class="mf">0.5</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">A_module</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">A_module</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span> <span class="o">*</span> <span class="n">B_module</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">cosine</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">numerator</span><span class="p">,</span> <span class="n">denominator</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">return</span> <span class="n">cosine</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.matmul_count_active">
<code class="highlight language-python"><span class="n">matmul_count_active</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the count of active protein-protein interactions
used for intercellular communication using matrices of proteins by
cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_count_active--parameters">Parameters</h6>
<p>A_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<p>B_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_count_active--returns">Returns</h6>
<p>counts : numpy.array
    Matrix MxM, representing the CCI score for all pairs of
    cell-types/tissues/samples. In directed interactions,
    the vertical axis (axis 0) represents the senders, while
    the horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">matmul_count_active</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the count of active protein-protein interactions</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    used for intercellular communication using matrices of proteins by</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell-types/tissues/samples.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    A_scores : array-like</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    B_scores : array-like</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    counts : numpy.array</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Matrix MxM, representing the CCI score for all pairs of</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell-types/tissues/samples. In directed interactions,</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        the vertical axis (axis 0) represents the senders, while</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        the horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="n">A_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">ppi_score</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">ppi_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span> <span class="n">B_scores</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">return</span> <span class="n">counts</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cci_scores.matmul_jaccard_like">
<code class="highlight language-python"><span class="n">matmul_jaccard_like</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes Jaccard-like scores using matrices of proteins by
cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_jaccard_like--parameters">Parameters</h6>
<p>A_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<p>B_scores : array-like
    Matrix of size NxM, where N are the proteins in the first
    column of a list of PPIs and M are the
    cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.cci_scores.matmul_jaccard_like--returns">Returns</h6>
<p>jaccard : numpy.array
    Matrix MxM, representing the CCI score for all pairs of
    cell-types/tissues/samples. In directed interactions,
    the vertical axis (axis 0) represents the senders, while
    the horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cci_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">matmul_jaccard_like</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes Jaccard-like scores using matrices of proteins by</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    cell-types/tissues/samples.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    A_scores : array-like</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    B_scores : array-like</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Matrix of size NxM, where N are the proteins in the first</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        column of a list of PPIs and M are the</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    jaccard : numpy.array</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Matrix MxM, representing the CCI score for all pairs of</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples. In directed interactions,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        the vertical axis (axis 0) represents the senders, while</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        the horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="n">A_scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">ppi_score</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">ppi_score</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">ppi_score</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span> <span class="n">B_scores</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">A_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">A_scores</span><span class="p">,</span> <span class="n">A_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">B_module</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">B_scores</span><span class="p">,</span> <span class="n">B_scores</span><span class="p">),</span> <span class="n">ppi_score</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">A_module</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">A_module</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span> <span class="o">+</span> <span class="n">B_module</span> <span class="o">-</span> <span class="n">numerator</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">jaccard</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">numerator</span><span class="p">,</span> <span class="n">denominator</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">return</span> <span class="n">jaccard</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.core.cell">
        <code>cell</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.core.cell.Cell">
        <code>
Cell        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Specific cell-type/tissue/organ element in a RNAseq dataset.</p>
<h6 id="cell2cell.core.cell.Cell--parameters">Parameters</h6>
<p>sc_rnaseq_data : pandas.DataFrame
    A gene expression matrix. Contains only one column that
    corresponds to cell-type/tissue/sample, while the genes
    are rows and the specific. Column name will be the label
    of the instance.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.cell.Cell--attributes">Attributes</h6>
<p>id : int
    ID number of the instance generated.</p>
<p>type : str
    Name of the respective cell-type/tissue/sample.</p>
<p>rnaseq_data : pandas.DataFrame
    Copy of sc_rnaseq_data.</p>
<p>weighted_ppi : pandas.DataFrame
    Dataframe created from a list of protein-protein interactions,
    here the columns of the interacting proteins are replaced by
    a score or a preprocessed gene expression of the respective
    proteins.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cell.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">Cell</span><span class="p">:</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Specific cell-type/tissue/organ element in a RNAseq dataset.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    sc_rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A gene expression matrix. Contains only one column that</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        corresponds to cell-type/tissue/sample, while the genes</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        are rows and the specific. Column name will be the label</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        of the instance.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    id : int</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        ID number of the instance generated.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    type : str</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Name of the respective cell-type/tissue/sample.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Copy of sc_rnaseq_data.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    weighted_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Dataframe created from a list of protein-protein interactions,</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        here the columns of the interacting proteins are replaced by</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        a score or a preprocessed gene expression of the respective</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        proteins.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">_id_counter</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Number of active instances</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">_id</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># Unique ID</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sc_rnaseq_data</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">Cell</span><span class="o">.</span><span class="n">_id</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">Cell</span><span class="o">.</span><span class="n">_id_counter</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="n">Cell</span><span class="o">.</span><span class="n">_id</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sc_rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="c1"># RNAseq datasets</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">sc_rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="c1"># Binary ppi datasets</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">weighted_ppi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">])</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="c1"># Object created</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;New cell instance created for &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="k">def</span> <span class="fm">__del__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="n">Cell</span><span class="o">.</span><span class="n">_id_counter</span> <span class="o">-=</span> <span class="mi">1</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">)</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="fm">__repr__</span> <span class="o">=</span> <span class="fm">__str__</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">











  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.core.cell.Cell.__repr__">
<code class="highlight language-python"><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h5>

    <div class="doc doc-contents ">

      <p>Return str(self).</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cell.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">type</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.cell.get_cells_from_rnaseq">
<code class="highlight language-python"><span class="n">get_cells_from_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cell_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Creates new instances of Cell based on the RNAseq data of each
cell-type/tissue/sample in a gene expression matrix.</p>
<h6 id="cell2cell.core.cell.get_cells_from_rnaseq--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>cell_columns : array-like, default=None
    List of names of cell-types/tissues/samples in the dataset
    to be used. If None, all columns will be used.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.cell.get_cells_from_rnaseq--returns">Returns</h6>
<p>cells : dict
    Dictionary containing all Cell instances generated from a RNAseq dataset.
    The keys of this dictionary are the names of the corresponding Cell instances.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/cell.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_cells_from_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cell_columns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Creates new instances of Cell based on the RNAseq data of each</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell-type/tissue/sample in a gene expression matrix.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    cell_columns : array-like, default=None</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of names of cell-types/tissues/samples in the dataset</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        to be used. If None, all columns will be used.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    cells : dict</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Dictionary containing all Cell instances generated from a RNAseq dataset.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        The keys of this dictionary are the names of the corresponding Cell instances.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generating objects according to RNAseq datasets provided&quot;</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">cells</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="n">cell_columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">cell_columns</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cell_columns</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">cells</span><span class="p">[</span><span class="n">cell</span><span class="p">]</span> <span class="o">=</span> <span class="n">Cell</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">[[</span><span class="n">cell</span><span class="p">]],</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">cells</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.core.communication_scores">
        <code>communication_scores</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.aggregate_ccc_matrices">
<code class="highlight language-python"><span class="n">aggregate_ccc_matrices</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Aggregates matrices of communication scores. Each
matrix has the communication scores across all pairs
of cell-types/tissues/samples for a different
pair of interacting proteins.</p>
<h6 id="cell2cell.core.communication_scores.aggregate_ccc_matrices--parameters">Parameters</h6>
<p>ccc_matrices : list
    List of matrices of communication scores. Each matrix
    is for an specific pair of interacting proteins.</p>
<p>method : str, default='gmean'.
    Method to aggregate the matrices element-wise.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;gmean&#39; : Geometric mean in an element-wise way.
- &#39;sum&#39; : Sum in an element-wise way.
- &#39;mean&#39; : Mean in an element-wise way.
</code></pre></div>

<h6 id="cell2cell.core.communication_scores.aggregate_ccc_matrices--returns">Returns</h6>
<p>aggregated_ccc_matrix : numpy.array
    A matrix contiaining aggregated communication scores
    from multiple PPIs. It's shape is of MxM, where M are all
    cell-types/tissues/samples. In directed interactions, the
    vertical axis (axis 0) represents the senders, while the
    horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">aggregate_ccc_matrices</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Aggregates matrices of communication scores. Each</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    matrix has the communication scores across all pairs</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    of cell-types/tissues/samples for a different</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    pair of interacting proteins.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ccc_matrices : list</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of matrices of communication scores. Each matrix</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        is for an specific pair of interacting proteins.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    method : str, default=&#39;gmean&#39;.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Method to aggregate the matrices element-wise.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean in an element-wise way.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;sum&#39; : Sum in an element-wise way.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;mean&#39; : Mean in an element-wise way.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    -------</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    aggregated_ccc_matrix : numpy.array</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        A matrix contiaining aggregated communication scores</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        from multiple PPIs. It&#39;s shape is of MxM, where M are all</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        cell-types/tissues/samples. In directed interactions, the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        vertical axis (axis 0) represents the senders, while the</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;gmean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">aggregated_ccc_matrix</span> <span class="o">=</span> <span class="n">gmean</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;sum&#39;</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">aggregated_ccc_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">aggregated_ccc_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Not a valid method&quot;</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">return</span> <span class="n">aggregated_ccc_matrix</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.compute_ccc_matrix">
<code class="highlight language-python"><span class="n">compute_ccc_matrix</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="p">,</span> <span class="n">prot_b_exp</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes communication scores for an specific
protein-protein interaction using vectors of gene expression
levels for a given interacting protein produced by
different cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.communication_scores.compute_ccc_matrix--parameters">Parameters</h6>
<p>prot_a_exp : array-like
    Vector with gene expression levels for an interacting protein A
    in a given PPI. Coordinates are different cell-types/tissues/samples.</p>
<p>prot_b_exp : array-like
    Vector with gene expression levels for an interacting protein B
    in a given PPI. Coordinates are different cell-types/tissues/samples.</p>
<p>communication_score : str, default='expression_product'
    Scoring function for computing the communication score.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_product&#39; : Multiplication between the expression
    of the interacting proteins.
- &#39;expression_mean&#39; : Average between the expression
    of the interacting proteins.
- &#39;expression_gmean&#39; : Geometric mean between the expression
    of the interacting proteins.
</code></pre></div>

<h6 id="cell2cell.core.communication_scores.compute_ccc_matrix--returns">Returns</h6>
<p>communication_scores : numpy.array
    Matrix MxM, representing the CCC scores of an specific PPI
    across all pairs of cell-types/tissues/samples. M are all
    cell-types/tissues/samples. In directed interactions, the
    vertical axis (axis 0) represents the senders, while the
    horizontal axis (axis 1) represents the receivers.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_ccc_matrix</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="p">,</span> <span class="n">prot_b_exp</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes communication scores for an specific</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    protein-protein interaction using vectors of gene expression</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    levels for a given interacting protein produced by</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    different cell-types/tissues/samples.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    prot_a_exp : array-like</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Vector with gene expression levels for an interacting protein A</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        in a given PPI. Coordinates are different cell-types/tissues/samples.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    prot_b_exp : array-like</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Vector with gene expression levels for an interacting protein B</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        in a given PPI. Coordinates are different cell-types/tissues/samples.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    communication_score : str, default=&#39;expression_product&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Scoring function for computing the communication score.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;expression_product&#39; : Multiplication between the expression</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">            of the interacting proteins.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;expression_mean&#39; : Average between the expression</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">            of the interacting proteins.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;expression_gmean&#39; : Geometric mean between the expression</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">            of the interacting proteins.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    communication_scores : numpy.array</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Matrix MxM, representing the CCC scores of an specific PPI</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        across all pairs of cell-types/tissues/samples. M are all</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        cell-types/tissues/samples. In directed interactions, the</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        vertical axis (axis 0) represents the senders, while the</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        horizontal axis (axis 1) represents the receivers.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">if</span> <span class="n">communication_score</span> <span class="o">==</span> <span class="s1">&#39;expression_product&#39;</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="p">,</span> <span class="n">prot_b_exp</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">elif</span> <span class="n">communication_score</span> <span class="o">==</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">prot_b_exp</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="n">prot_b_exp</span><span class="p">))</span> <span class="o">/</span> <span class="mf">2.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">elif</span> <span class="n">communication_score</span> <span class="o">==</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">outer</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="p">,</span> <span class="n">prot_b_exp</span><span class="p">))</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Not a valid communication_score&quot;</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">return</span> <span class="n">communication_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.get_binary_scores">
<code class="highlight language-python"><span class="n">get_binary_scores</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes binary communication scores for all
protein-protein interactions between a pair of
cell-types/tissues/samples. This corresponds to
an AND function between binary values for each
interacting protein coming from each cell.</p>
<h6 id="cell2cell.core.communication_scores.get_binary_scores--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the receiver.</p>
<p>ppi_score : array-like, default=None
    An array with a weight for each PPI. The weight
    multiplies the communication scores.</p>
<h6 id="cell2cell.core.communication_scores.get_binary_scores--returns">Returns</h6>
<p>communication_scores : numpy.array
    An array with the communication scores for each intercellular
    PPI.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_binary_scores</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes binary communication scores for all</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    protein-protein interactions between a pair of</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell-types/tissues/samples. This corresponds to</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    an AND function between binary values for each</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    interacting protein coming from each cell.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    ppi_score : array-like, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        An array with a weight for each PPI. The weight</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        multiplies the communication scores.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    communication_scores : numpy.array</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        An array with the communication scores for each intercellular</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        PPI.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span> <span class="o">*</span> <span class="n">ppi_score</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">return</span> <span class="n">communication_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.get_continuous_scores">
<code class="highlight language-python"><span class="n">get_continuous_scores</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes continuous communication scores for all
protein-protein interactions between a pair of
cell-types/tissues/samples. This corresponds to
a specific scoring function between preprocessed continuous
expression values for each interacting protein coming from
each cell.</p>
<h6 id="cell2cell.core.communication_scores.get_continuous_scores--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the receiver.</p>
<p>ppi_score : array-like, default=None
    An array with a weight for each PPI. The weight
    multiplies the communication scores.</p>
<p>method : str, default='expression_product'
    Scoring function for computing the communication score.
    Options are:
        - 'expression_product' : Multiplication between the expression
            of the interacting proteins. One coming from cell1 and the
            other from cell2.
        - 'expression_mean' : Average between the expression
            of the interacting proteins. One coming from cell1 and the
            other from cell2.
        - 'expression_gmean' : Geometric mean between the expression
            of the interacting proteins. One coming from cell1 and the
            other from cell2.</p>
<h6 id="cell2cell.core.communication_scores.get_continuous_scores--returns">Returns</h6>
<p>communication_scores : numpy.array
    An array with the communication scores for each intercellular
    PPI.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_continuous_scores</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes continuous communication scores for all</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    protein-protein interactions between a pair of</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell-types/tissues/samples. This corresponds to</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    a specific scoring function between preprocessed continuous</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    expression values for each interacting protein coming from</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    each cell.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    ppi_score : array-like, default=None</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        An array with a weight for each PPI. The weight</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        multiplies the communication scores.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    method : str, default=&#39;expression_product&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Scoring function for computing the communication score.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">            - &#39;expression_product&#39; : Multiplication between the expression</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                of the interacting proteins. One coming from cell1 and the</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                other from cell2.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">            - &#39;expression_mean&#39; : Average between the expression</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                of the interacting proteins. One coming from cell1 and the</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                other from cell2.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">            - &#39;expression_gmean&#39; : Geometric mean between the expression</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">                of the interacting proteins. One coming from cell1 and the</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                other from cell2.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    -------</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    communication_scores : numpy.array</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        An array with the communication scores for each intercellular</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        PPI.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">c1</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">c2</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">weighted_ppi</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;expression_product&#39;</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">score_expression_product</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">score_expression_mean</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">score_expression_product</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">))</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> is not implemented yet&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">))</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">if</span> <span class="n">ppi_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">))</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="n">communication_scores</span> <span class="o">=</span> <span class="n">communication_scores</span> <span class="o">*</span> <span class="n">ppi_score</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">return</span> <span class="n">communication_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.score_expression_mean">
<code class="highlight language-python"><span class="n">score_expression_mean</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the expression product score</p>
<h6 id="cell2cell.core.communication_scores.score_expression_mean--parameters">Parameters</h6>
<p>c1 : array-like
    A 1D-array containing the preprocessed expression values
    for the interactors in the first column of a list of
    protein-protein interactions.</p>
<p>c2 : array-like
    A 1D-array containing the preprocessed expression values
    for the interactors in the second column of a list of
    protein-protein interactions.</p>
<h6 id="cell2cell.core.communication_scores.score_expression_mean--returns">Returns</h6>
<p>(c1 + c2)/2. : array-like
    Average of vectors.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">score_expression_mean</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the expression product score</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    c1 : array-like</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A 1D-array containing the preprocessed expression values</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        for the interactors in the first column of a list of</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        protein-protein interactions.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    c2 : array-like</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A 1D-array containing the preprocessed expression values</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        for the interactors in the second column of a list of</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        protein-protein interactions.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    (c1 + c2)/2. : array-like</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Average of vectors.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">return</span> <span class="p">(</span><span class="n">c1</span> <span class="o">+</span> <span class="n">c2</span><span class="p">)</span><span class="o">/</span><span class="mf">2.</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.communication_scores.score_expression_product">
<code class="highlight language-python"><span class="n">score_expression_product</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the expression product score</p>
<h6 id="cell2cell.core.communication_scores.score_expression_product--parameters">Parameters</h6>
<p>c1 : array-like
    A 1D-array containing the preprocessed expression values
    for the interactors in the first column of a list of
    protein-protein interactions.</p>
<p>c2 : array-like
    A 1D-array containing the preprocessed expression values
    for the interactors in the second column of a list of
    protein-protein interactions.</p>
<h6 id="cell2cell.core.communication_scores.score_expression_product--returns">Returns</h6>
<p>c1 * c2 : array-like
    Multiplication of vectors.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/communication_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">score_expression_product</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the expression product score</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    c1 : array-like</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A 1D-array containing the preprocessed expression values</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        for the interactors in the first column of a list of</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        protein-protein interactions.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    c2 : array-like</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A 1D-array containing the preprocessed expression values</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        for the interactors in the second column of a list of</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        protein-protein interactions.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    c1 * c2 : array-like</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Multiplication of vectors.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">return</span> <span class="n">c1</span> <span class="o">*</span> <span class="n">c2</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.core.interaction_space">
        <code>interaction_space</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.core.interaction_space.InteractionSpace">
        <code>
InteractionSpace        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Interaction space that contains all the required elements to perform the analysis between every pair of cells.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>gene_cutoffs : dict
    Contains two keys: 'type' and 'parameter'. The first key represent the
    way to use a cutoff or threshold, while parameter is the value used
    to binarize the expression values.
    The key 'type' can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"></span>
<span class="w">    </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"></span>
<span class="w">    </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"></span>
<span class="w">    </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">    </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"></span>
<span class="w">    </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"></span>
<span class="w">    </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">    </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">    </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>communication_score : str, default='expression_thresholding'
    Type of communication score used to detect active ligand-receptor
    pairs between each pair of cell. See
    cell2cell.core.communication_scores for more details.
    It can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39;
- &#39;expression_product&#39;
- &#39;expression_mean&#39;
- &#39;expression_gmean&#39;
</code></pre></div>

<p>cci_score : str, default='bray_curtis'
    Scoring function to aggregate the communication scores. See
    cell2cell.core.cci_scores for more details.
    It can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39;
- &#39;jaccard&#39;
- &#39;count&#39;
- &#39;icellnet&#39;
</code></pre></div>

<p>cci_type : str, default='undirected'
    Type of interaction between two cells. If it is undirected, all ligands
    and receptors are considered from both cells. If it is directed, ligands
    from one cell and receptors from the other are considered separately with
    respect to ligands from the second cell and receptor from the first one.
    So, it can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;undirected&#39;
- &#39;directed&#39;
</code></pre></div>

<p>cci_matrix_template : pandas.DataFrame, default=None
    A matrix of shape MxM where M are cell-types/tissues/samples. This
    is used as template for storing CCI scores. It may be useful
    for specifying which pairs of cells to consider.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace--attributes">Attributes</h6>
<p>communication_score : str
    Type of communication score used to detect active ligand-receptor
    pairs between each pair of cell. See
    cell2cell.core.communication_scores for more details.
    It can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39;
- &#39;expression_product&#39;
- &#39;expression_mean&#39;
- &#39;expression_gmean&#39;
</code></pre></div>

<p>cci_score : str
    Scoring function to aggregate the communication scores. See
    cell2cell.core.cci_scores for more details.
    It can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39;
- &#39;jaccard&#39;
- &#39;count&#39;
- &#39;icellnet&#39;
</code></pre></div>

<p>cci_type : str
    Type of interaction between two cells. If it is undirected, all ligands
    and receptors are considered from both cells. If it is directed, ligands
    from one cell and receptors from the other are considered separately with
    respect to ligands from the second cell and receptor from the first one.
    So, it can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;undirected&#39;
- &#39;directed&#39;
</code></pre></div>

<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>modified_rnaseq_data : pandas.DataFrame
    Preprocessed gene expression data for a bulk or single-cell RNA-seq experiment.
    Columns are are cell-types/tissues/samples and rows are genes. The preprocessing
    may correspond to scoring the gene expression as binary or continuous values
    depending on the scoring function for cell-cell interactions/communication.</p>
<p>interaction_elements : dict
    Dictionary containing all the pairs of cells considered (under
    the key of 'pairs'), Cell instances (under key 'cells')
    which include all cells/tissues/organs with their associated datasets
    (rna_seq, weighted_ppi, etc) and a Cell-Cell Interaction Matrix
    to store CCI scores(under key 'cci_matrix'). A communication matrix
    is also stored in this object when the communication scores are
    computed in the InteractionSpace class (under key
    'communication_matrix')</p>
<p>distance_matrix : pandas.DataFrame
    Contains distances for each pair of cells, computed from
    the CCI scores previously obtained (and stored in
    interaction_elements['cci_matrix'].</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">InteractionSpace</span><span class="p">():</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Interaction space that contains all the required elements to perform the analysis between every pair of cells.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    gene_cutoffs : dict</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Contains two keys: &#39;type&#39; and &#39;parameter&#39;. The first key represent the</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        way to use a cutoff or threshold, while parameter is the value used</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        to binarize the expression values.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        The key &#39;type&#39; can be:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;local_percentile&#39; : computes the value of a given percentile, for each</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">            gene independently. In this case, the parameter corresponds to the</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">            percentile to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;global_percentile&#39; : computes the value of a given percentile from all</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">            genes and samples simultaneously. In this case, the parameter</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">            corresponds to the percentile to compute, as a float value between</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">            0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is the</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">            path of that file. It must contain the same genes as index and same</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">            samples as columns.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">            for each gene in each sample. This allows to use specific cutoffs for</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">            each sample. The columns here must be the same as the ones in the</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">            rnaseq_data.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">            for each gene in only one column. These cutoffs will be applied to</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">            all samples.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">            expression is greater than the value input in the parameter.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    communication_score : str, default=&#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Type of communication score used to detect active ligand-receptor</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        pairs between each pair of cell. See</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        cell2cell.core.communication_scores for more details.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        - &#39;expression_product&#39;</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    cci_score : str, default=&#39;bray_curtis&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Scoring function to aggregate the communication scores. See</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        cell2cell.core.cci_scores for more details.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        - &#39;jaccard&#39;</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        - &#39;count&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        - &#39;icellnet&#39;</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    cci_type : str, default=&#39;undirected&#39;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Type of interaction between two cells. If it is undirected, all ligands</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        and receptors are considered from both cells. If it is directed, ligands</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        from one cell and receptors from the other are considered separately with</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        respect to ligands from the second cell and receptor from the first one.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        So, it can be:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        - &#39;undirected&#39;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        - &#39;directed&#39;</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    cci_matrix_template : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        A matrix of shape MxM where M are cell-types/tissues/samples. This</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        is used as template for storing CCI scores. It may be useful</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">        for specifying which pairs of cells to consider.</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    communication_score : str</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">        Type of communication score used to detect active ligand-receptor</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">        pairs between each pair of cell. See</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">        cell2cell.core.communication_scores for more details.</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a><span class="sd">        - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">        - &#39;expression_product&#39;</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a><span class="sd">        - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a><span class="sd">        - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a><span class="sd">    cci_score : str</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a><span class="sd">        Scoring function to aggregate the communication scores. See</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a><span class="sd">        cell2cell.core.cci_scores for more details.</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a><span class="sd">        - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a><span class="sd">        - &#39;jaccard&#39;</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a><span class="sd">        - &#39;count&#39;</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a><span class="sd">        - &#39;icellnet&#39;</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a><span class="sd">    cci_type : str</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a><span class="sd">        Type of interaction between two cells. If it is undirected, all ligands</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a><span class="sd">        and receptors are considered from both cells. If it is directed, ligands</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a><span class="sd">        from one cell and receptors from the other are considered separately with</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a><span class="sd">        respect to ligands from the second cell and receptor from the first one.</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a><span class="sd">        So, it can be:</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a><span class="sd">        - &#39;undirected&#39;</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a><span class="sd">        - &#39;directed&#39;</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a><span class="sd">    modified_rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a><span class="sd">        Preprocessed gene expression data for a bulk or single-cell RNA-seq experiment.</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a><span class="sd">        Columns are are cell-types/tissues/samples and rows are genes. The preprocessing</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a><span class="sd">        may correspond to scoring the gene expression as binary or continuous values</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a><span class="sd">        depending on the scoring function for cell-cell interactions/communication.</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a><span class="sd">    interaction_elements : dict</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a><span class="sd">        Dictionary containing all the pairs of cells considered (under</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a><span class="sd">        the key of &#39;pairs&#39;), Cell instances (under key &#39;cells&#39;)</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a><span class="sd">        which include all cells/tissues/organs with their associated datasets</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a><span class="sd">        (rna_seq, weighted_ppi, etc) and a Cell-Cell Interaction Matrix</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a><span class="sd">        to store CCI scores(under key &#39;cci_matrix&#39;). A communication matrix</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a><span class="sd">        is also stored in this object when the communication scores are</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a><span class="sd">        computed in the InteractionSpace class (under key</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a><span class="sd">        &#39;communication_matrix&#39;)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a><span class="sd">    distance_matrix : pandas.DataFrame</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a><span class="sd">        Contains distances for each pair of cells, computed from</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a><span class="sd">        the CCI scores previously obtained (and stored in</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a><span class="sd">        interaction_elements[&#39;cci_matrix&#39;].</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">gene_cutoffs</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>                 <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="s1">&#39;undirected&#39;</span><span class="p">,</span> <span class="n">cci_matrix_template</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>                 <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="o">=</span> <span class="n">communication_score</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cci_score</span> <span class="o">=</span> <span class="n">cci_score</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span> <span class="o">=</span> <span class="n">cci_type</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="o">==</span> <span class="s1">&#39;expression_thresholding&#39;</span><span class="p">:</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>            <span class="k">if</span> <span class="s1">&#39;type&#39;</span> <span class="ow">in</span> <span class="n">gene_cutoffs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>                <span class="n">cutoff_values</span> <span class="o">=</span> <span class="n">cutoffs</span><span class="o">.</span><span class="n">get_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>                                                    <span class="n">parameters</span><span class="o">=</span><span class="n">gene_cutoffs</span><span class="p">,</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                                                    <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If dataframe is not included in gene_cutoffs, please provide the type of method to obtain them.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>            <span class="n">cutoff_values</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>        <span class="n">prot_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="n">prot_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_a</span> <span class="o">!=</span> <span class="s1">&#39;A&#39;</span><span class="p">):</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;B&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_b</span> <span class="o">!=</span> <span class="s1">&#39;B&#39;</span><span class="p">):</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">)</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">prot_a</span> <span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">prot_b</span> <span class="p">:</span> <span class="s1">&#39;B&#39;</span><span class="p">})</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>        <span class="k">if</span> <span class="s1">&#39;score&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">modified_rnaseq</span> <span class="o">=</span> <span class="n">integrate_data</span><span class="o">.</span><span class="n">get_modified_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                                                                  <span class="n">cutoffs</span><span class="o">=</span><span class="n">cutoff_values</span><span class="p">,</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>                                                                  <span class="n">communication_score</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>                                                                  <span class="p">)</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span> <span class="o">=</span> <span class="n">generate_interaction_elements</span><span class="p">(</span><span class="n">modified_rnaseq</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">modified_rnaseq</span><span class="p">,</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>                                                                  <span class="n">ppi_data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>                                                                  <span class="n">cci_matrix_template</span><span class="o">=</span><span class="n">cci_matrix_template</span><span class="p">,</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>                                                                  <span class="n">cci_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span><span class="p">,</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>                                                                  <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>                                                                  <span class="n">complex_agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                                                                  <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;ppi_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>    <span class="k">def</span> <span class="nf">pair_cci_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>        <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a><span class="sd">        Computes a CCI score for a pair of cells.</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a><span class="sd">        cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a><span class="sd">            First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a><span class="sd">            score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a><span class="sd">        cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a><span class="sd">            Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a><span class="sd">            score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a><span class="sd">        cci_score : str, default=&#39;bray_curtis&#39;</span>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a><span class="sd">            Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a><span class="sd">            a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a><span class="sd">            interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-221" name="__codelineno-0-221" href="#__codelineno-0-221"></a><span class="sd">            attribute analysis_setup of this object.</span>
<a id="__codelineno-0-222" name="__codelineno-0-222" href="#__codelineno-0-222"></a><span class="sd">            Options:</span>
<a id="__codelineno-0-223" name="__codelineno-0-223" href="#__codelineno-0-223"></a>
<a id="__codelineno-0-224" name="__codelineno-0-224" href="#__codelineno-0-224"></a><span class="sd">            - &#39;bray_curtis&#39; : Bray-Curtis-like score</span>
<a id="__codelineno-0-225" name="__codelineno-0-225" href="#__codelineno-0-225"></a><span class="sd">            - &#39;jaccard&#39; : Jaccard-like score</span>
<a id="__codelineno-0-226" name="__codelineno-0-226" href="#__codelineno-0-226"></a><span class="sd">            - &#39;count&#39; : Number of LR pairs that the pair of cells uses</span>
<a id="__codelineno-0-227" name="__codelineno-0-227" href="#__codelineno-0-227"></a><span class="sd">            - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-228" name="__codelineno-0-228" href="#__codelineno-0-228"></a>
<a id="__codelineno-0-229" name="__codelineno-0-229" href="#__codelineno-0-229"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-230" name="__codelineno-0-230" href="#__codelineno-0-230"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-231" name="__codelineno-0-231" href="#__codelineno-0-231"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-232" name="__codelineno-0-232" href="#__codelineno-0-232"></a>
<a id="__codelineno-0-233" name="__codelineno-0-233" href="#__codelineno-0-233"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-234" name="__codelineno-0-234" href="#__codelineno-0-234"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-235" name="__codelineno-0-235" href="#__codelineno-0-235"></a>
<a id="__codelineno-0-236" name="__codelineno-0-236" href="#__codelineno-0-236"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-237" name="__codelineno-0-237" href="#__codelineno-0-237"></a><span class="sd">        -------</span>
<a id="__codelineno-0-238" name="__codelineno-0-238" href="#__codelineno-0-238"></a><span class="sd">        cci_score : float</span>
<a id="__codelineno-0-239" name="__codelineno-0-239" href="#__codelineno-0-239"></a><span class="sd">            Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-240" name="__codelineno-0-240" href="#__codelineno-0-240"></a><span class="sd">            cell-types/tissues/samples. In this case it is a</span>
<a id="__codelineno-0-241" name="__codelineno-0-241" href="#__codelineno-0-241"></a><span class="sd">            Jaccard-like score.</span>
<a id="__codelineno-0-242" name="__codelineno-0-242" href="#__codelineno-0-242"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-243" name="__codelineno-0-243" href="#__codelineno-0-243"></a>
<a id="__codelineno-0-244" name="__codelineno-0-244" href="#__codelineno-0-244"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-245" name="__codelineno-0-245" href="#__codelineno-0-245"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing interaction score between </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell1</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>
<a id="__codelineno-0-246" name="__codelineno-0-246" href="#__codelineno-0-246"></a>
<a id="__codelineno-0-247" name="__codelineno-0-247" href="#__codelineno-0-247"></a>        <span class="k">if</span> <span class="n">use_ppi_score</span><span class="p">:</span>
<a id="__codelineno-0-248" name="__codelineno-0-248" href="#__codelineno-0-248"></a>            <span class="n">ppi_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-249" name="__codelineno-0-249" href="#__codelineno-0-249"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-250" name="__codelineno-0-250" href="#__codelineno-0-250"></a>            <span class="n">ppi_score</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-251" name="__codelineno-0-251" href="#__codelineno-0-251"></a>        <span class="c1"># Calculate cell-cell interaction score</span>
<a id="__codelineno-0-252" name="__codelineno-0-252" href="#__codelineno-0-252"></a>        <span class="k">if</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;bray_curtis&#39;</span><span class="p">:</span>
<a id="__codelineno-0-253" name="__codelineno-0-253" href="#__codelineno-0-253"></a>            <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_braycurtis_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-254" name="__codelineno-0-254" href="#__codelineno-0-254"></a>        <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;jaccard&#39;</span><span class="p">:</span>
<a id="__codelineno-0-255" name="__codelineno-0-255" href="#__codelineno-0-255"></a>            <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_jaccard_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-256" name="__codelineno-0-256" href="#__codelineno-0-256"></a>        <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;count&#39;</span><span class="p">:</span>
<a id="__codelineno-0-257" name="__codelineno-0-257" href="#__codelineno-0-257"></a>            <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_count_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-258" name="__codelineno-0-258" href="#__codelineno-0-258"></a>        <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;icellnet&#39;</span><span class="p">:</span>
<a id="__codelineno-0-259" name="__codelineno-0-259" href="#__codelineno-0-259"></a>            <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_icellnet_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-260" name="__codelineno-0-260" href="#__codelineno-0-260"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-261" name="__codelineno-0-261" href="#__codelineno-0-261"></a>            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;CCI score </span><span class="si">{}</span><span class="s2"> to compute pairwise cell-interactions is not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cci_score</span><span class="p">))</span>
<a id="__codelineno-0-262" name="__codelineno-0-262" href="#__codelineno-0-262"></a>        <span class="k">return</span> <span class="n">cci_value</span>
<a id="__codelineno-0-263" name="__codelineno-0-263" href="#__codelineno-0-263"></a>
<a id="__codelineno-0-264" name="__codelineno-0-264" href="#__codelineno-0-264"></a>    <span class="k">def</span> <span class="nf">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-265" name="__codelineno-0-265" href="#__codelineno-0-265"></a>        <span class="sd">&#39;&#39;&#39;Computes overall CCI scores for each pair of cells.</span>
<a id="__codelineno-0-266" name="__codelineno-0-266" href="#__codelineno-0-266"></a>
<a id="__codelineno-0-267" name="__codelineno-0-267" href="#__codelineno-0-267"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-268" name="__codelineno-0-268" href="#__codelineno-0-268"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-269" name="__codelineno-0-269" href="#__codelineno-0-269"></a><span class="sd">        cci_score : str, default=None</span>
<a id="__codelineno-0-270" name="__codelineno-0-270" href="#__codelineno-0-270"></a><span class="sd">            Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-271" name="__codelineno-0-271" href="#__codelineno-0-271"></a><span class="sd">            a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-272" name="__codelineno-0-272" href="#__codelineno-0-272"></a><span class="sd">            interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-273" name="__codelineno-0-273" href="#__codelineno-0-273"></a><span class="sd">            attribute analysis_setup of this object.</span>
<a id="__codelineno-0-274" name="__codelineno-0-274" href="#__codelineno-0-274"></a><span class="sd">            Options:</span>
<a id="__codelineno-0-275" name="__codelineno-0-275" href="#__codelineno-0-275"></a>
<a id="__codelineno-0-276" name="__codelineno-0-276" href="#__codelineno-0-276"></a><span class="sd">            - &#39;bray_curtis&#39; : Bray-Curtis-like score</span>
<a id="__codelineno-0-277" name="__codelineno-0-277" href="#__codelineno-0-277"></a><span class="sd">            - &#39;jaccard&#39; : Jaccard-like score</span>
<a id="__codelineno-0-278" name="__codelineno-0-278" href="#__codelineno-0-278"></a><span class="sd">            - &#39;count&#39; : Number of LR pairs that the pair of cells uses</span>
<a id="__codelineno-0-279" name="__codelineno-0-279" href="#__codelineno-0-279"></a><span class="sd">            - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-280" name="__codelineno-0-280" href="#__codelineno-0-280"></a>
<a id="__codelineno-0-281" name="__codelineno-0-281" href="#__codelineno-0-281"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-282" name="__codelineno-0-282" href="#__codelineno-0-282"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-283" name="__codelineno-0-283" href="#__codelineno-0-283"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-284" name="__codelineno-0-284" href="#__codelineno-0-284"></a>
<a id="__codelineno-0-285" name="__codelineno-0-285" href="#__codelineno-0-285"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-286" name="__codelineno-0-286" href="#__codelineno-0-286"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-287" name="__codelineno-0-287" href="#__codelineno-0-287"></a>
<a id="__codelineno-0-288" name="__codelineno-0-288" href="#__codelineno-0-288"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-289" name="__codelineno-0-289" href="#__codelineno-0-289"></a><span class="sd">        -------</span>
<a id="__codelineno-0-290" name="__codelineno-0-290" href="#__codelineno-0-290"></a><span class="sd">        self.interaction_elements[&#39;cci_matrix&#39;] : pandas.DataFrame</span>
<a id="__codelineno-0-291" name="__codelineno-0-291" href="#__codelineno-0-291"></a><span class="sd">            Contains CCI scores for each pair of cells</span>
<a id="__codelineno-0-292" name="__codelineno-0-292" href="#__codelineno-0-292"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-293" name="__codelineno-0-293" href="#__codelineno-0-293"></a>        <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-294" name="__codelineno-0-294" href="#__codelineno-0-294"></a>            <span class="n">cci_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_score</span>
<a id="__codelineno-0-295" name="__codelineno-0-295" href="#__codelineno-0-295"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-296" name="__codelineno-0-296" href="#__codelineno-0-296"></a>            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cci_score</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
<a id="__codelineno-0-297" name="__codelineno-0-297" href="#__codelineno-0-297"></a>
<a id="__codelineno-0-298" name="__codelineno-0-298" href="#__codelineno-0-298"></a>        <span class="c1">### Compute pairwise physical interactions</span>
<a id="__codelineno-0-299" name="__codelineno-0-299" href="#__codelineno-0-299"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-300" name="__codelineno-0-300" href="#__codelineno-0-300"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing pairwise interactions&quot;</span><span class="p">)</span>
<a id="__codelineno-0-301" name="__codelineno-0-301" href="#__codelineno-0-301"></a>
<a id="__codelineno-0-302" name="__codelineno-0-302" href="#__codelineno-0-302"></a>        <span class="c1"># Compute pair by pair</span>
<a id="__codelineno-0-303" name="__codelineno-0-303" href="#__codelineno-0-303"></a>        <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-304" name="__codelineno-0-304" href="#__codelineno-0-304"></a>            <span class="n">cell1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<a id="__codelineno-0-305" name="__codelineno-0-305" href="#__codelineno-0-305"></a>            <span class="n">cell2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<a id="__codelineno-0-306" name="__codelineno-0-306" href="#__codelineno-0-306"></a>            <span class="n">cci_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pair_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-307" name="__codelineno-0-307" href="#__codelineno-0-307"></a>                                            <span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-308" name="__codelineno-0-308" href="#__codelineno-0-308"></a>                                            <span class="n">cci_score</span><span class="o">=</span><span class="n">cci_score</span><span class="p">,</span>
<a id="__codelineno-0-309" name="__codelineno-0-309" href="#__codelineno-0-309"></a>                                            <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-310" name="__codelineno-0-310" href="#__codelineno-0-310"></a>                                            <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-311" name="__codelineno-0-311" href="#__codelineno-0-311"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">cci_value</span>
<a id="__codelineno-0-312" name="__codelineno-0-312" href="#__codelineno-0-312"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-313" name="__codelineno-0-313" href="#__codelineno-0-313"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">cci_value</span>
<a id="__codelineno-0-314" name="__codelineno-0-314" href="#__codelineno-0-314"></a>
<a id="__codelineno-0-315" name="__codelineno-0-315" href="#__codelineno-0-315"></a>        <span class="c1"># Compute using matmul -&gt; Too slow and uses a lot of memory TODO: Try to optimize this</span>
<a id="__codelineno-0-316" name="__codelineno-0-316" href="#__codelineno-0-316"></a>        <span class="c1"># if cci_score == &#39;bray_curtis&#39;:</span>
<a id="__codelineno-0-317" name="__codelineno-0-317" href="#__codelineno-0-317"></a>        <span class="c1">#     cci_matrix = cci_scores.matmul_bray_curtis_like(self.interaction_elements[&#39;A_score&#39;],</span>
<a id="__codelineno-0-318" name="__codelineno-0-318" href="#__codelineno-0-318"></a>        <span class="c1">#                                                     self.interaction_elements[&#39;B_score&#39;])</span>
<a id="__codelineno-0-319" name="__codelineno-0-319" href="#__codelineno-0-319"></a>        <span class="c1"># self.interaction_elements[&#39;cci_matrix&#39;] = pd.DataFrame(cci_matrix,</span>
<a id="__codelineno-0-320" name="__codelineno-0-320" href="#__codelineno-0-320"></a>        <span class="c1">#                                                        index=self.interaction_elements[&#39;cell_names&#39;],</span>
<a id="__codelineno-0-321" name="__codelineno-0-321" href="#__codelineno-0-321"></a>        <span class="c1">#                                                        columns=self.interaction_elements[&#39;cell_names&#39;]</span>
<a id="__codelineno-0-322" name="__codelineno-0-322" href="#__codelineno-0-322"></a>        <span class="c1">#                                                        )</span>
<a id="__codelineno-0-323" name="__codelineno-0-323" href="#__codelineno-0-323"></a>
<a id="__codelineno-0-324" name="__codelineno-0-324" href="#__codelineno-0-324"></a>        <span class="c1"># Generate distance matrix</span>
<a id="__codelineno-0-325" name="__codelineno-0-325" href="#__codelineno-0-325"></a>        <span class="k">if</span> <span class="o">~</span><span class="p">(</span><span class="n">cci_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;icellnet&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-326" name="__codelineno-0-326" href="#__codelineno-0-326"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-327" name="__codelineno-0-327" href="#__codelineno-0-327"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-328" name="__codelineno-0-328" href="#__codelineno-0-328"></a>            <span class="c1">#self.distance_matrix = self.interaction_elements[&#39;cci_matrix&#39;].div(self.interaction_elements[&#39;cci_matrix&#39;].max().max()).apply(lambda x: 1 - x)</span>
<a id="__codelineno-0-329" name="__codelineno-0-329" href="#__codelineno-0-329"></a>            <span class="c1"># Regularized distance</span>
<a id="__codelineno-0-330" name="__codelineno-0-330" href="#__codelineno-0-330"></a>            <span class="n">mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">])</span>
<a id="__codelineno-0-331" name="__codelineno-0-331" href="#__codelineno-0-331"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">mean</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-332" name="__codelineno-0-332" href="#__codelineno-0-332"></a>        <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>  <span class="c1"># Make diagonal zero (delete autocrine-interactions)</span>
<a id="__codelineno-0-333" name="__codelineno-0-333" href="#__codelineno-0-333"></a>
<a id="__codelineno-0-334" name="__codelineno-0-334" href="#__codelineno-0-334"></a>    <span class="k">def</span> <span class="nf">pair_communication_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span>
<a id="__codelineno-0-335" name="__codelineno-0-335" href="#__codelineno-0-335"></a>                                 <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-336" name="__codelineno-0-336" href="#__codelineno-0-336"></a>        <span class="sd">&#39;&#39;&#39;Computes a communication score for each protein-protein interaction</span>
<a id="__codelineno-0-337" name="__codelineno-0-337" href="#__codelineno-0-337"></a><span class="sd">        between a pair of cells.</span>
<a id="__codelineno-0-338" name="__codelineno-0-338" href="#__codelineno-0-338"></a>
<a id="__codelineno-0-339" name="__codelineno-0-339" href="#__codelineno-0-339"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-340" name="__codelineno-0-340" href="#__codelineno-0-340"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-341" name="__codelineno-0-341" href="#__codelineno-0-341"></a><span class="sd">        cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-342" name="__codelineno-0-342" href="#__codelineno-0-342"></a><span class="sd">            First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-343" name="__codelineno-0-343" href="#__codelineno-0-343"></a><span class="sd">            score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-344" name="__codelineno-0-344" href="#__codelineno-0-344"></a>
<a id="__codelineno-0-345" name="__codelineno-0-345" href="#__codelineno-0-345"></a><span class="sd">        cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-346" name="__codelineno-0-346" href="#__codelineno-0-346"></a><span class="sd">            Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-347" name="__codelineno-0-347" href="#__codelineno-0-347"></a><span class="sd">            score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-348" name="__codelineno-0-348" href="#__codelineno-0-348"></a>
<a id="__codelineno-0-349" name="__codelineno-0-349" href="#__codelineno-0-349"></a><span class="sd">        communication_score : str, default=None</span>
<a id="__codelineno-0-350" name="__codelineno-0-350" href="#__codelineno-0-350"></a><span class="sd">            Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-351" name="__codelineno-0-351" href="#__codelineno-0-351"></a><span class="sd">            a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-352" name="__codelineno-0-352" href="#__codelineno-0-352"></a><span class="sd">            If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-353" name="__codelineno-0-353" href="#__codelineno-0-353"></a><span class="sd">            will be used.</span>
<a id="__codelineno-0-354" name="__codelineno-0-354" href="#__codelineno-0-354"></a><span class="sd">            Available communication_scores are:</span>
<a id="__codelineno-0-355" name="__codelineno-0-355" href="#__codelineno-0-355"></a>
<a id="__codelineno-0-356" name="__codelineno-0-356" href="#__codelineno-0-356"></a><span class="sd">            - &#39;expression_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-357" name="__codelineno-0-357" href="#__codelineno-0-357"></a><span class="sd">                                         ligand from a sender cell and of</span>
<a id="__codelineno-0-358" name="__codelineno-0-358" href="#__codelineno-0-358"></a><span class="sd">                                         a receptor on a receiver cell from</span>
<a id="__codelineno-0-359" name="__codelineno-0-359" href="#__codelineno-0-359"></a><span class="sd">                                         binarizing their gene expression levels.</span>
<a id="__codelineno-0-360" name="__codelineno-0-360" href="#__codelineno-0-360"></a><span class="sd">            - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-361" name="__codelineno-0-361" href="#__codelineno-0-361"></a><span class="sd">                                  of a ligand from a sender cell and the</span>
<a id="__codelineno-0-362" name="__codelineno-0-362" href="#__codelineno-0-362"></a><span class="sd">                                  expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-363" name="__codelineno-0-363" href="#__codelineno-0-363"></a><span class="sd">            - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-364" name="__codelineno-0-364" href="#__codelineno-0-364"></a><span class="sd">                                    of a ligand from a sender cell and the</span>
<a id="__codelineno-0-365" name="__codelineno-0-365" href="#__codelineno-0-365"></a><span class="sd">                                    expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-366" name="__codelineno-0-366" href="#__codelineno-0-366"></a><span class="sd">            - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-367" name="__codelineno-0-367" href="#__codelineno-0-367"></a><span class="sd">                                   of a ligand from a sender cell and the</span>
<a id="__codelineno-0-368" name="__codelineno-0-368" href="#__codelineno-0-368"></a><span class="sd">                                   expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-369" name="__codelineno-0-369" href="#__codelineno-0-369"></a>
<a id="__codelineno-0-370" name="__codelineno-0-370" href="#__codelineno-0-370"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-371" name="__codelineno-0-371" href="#__codelineno-0-371"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-372" name="__codelineno-0-372" href="#__codelineno-0-372"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-373" name="__codelineno-0-373" href="#__codelineno-0-373"></a>
<a id="__codelineno-0-374" name="__codelineno-0-374" href="#__codelineno-0-374"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-375" name="__codelineno-0-375" href="#__codelineno-0-375"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-376" name="__codelineno-0-376" href="#__codelineno-0-376"></a>
<a id="__codelineno-0-377" name="__codelineno-0-377" href="#__codelineno-0-377"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-378" name="__codelineno-0-378" href="#__codelineno-0-378"></a><span class="sd">        -------</span>
<a id="__codelineno-0-379" name="__codelineno-0-379" href="#__codelineno-0-379"></a><span class="sd">        communication_scores : numpy.array</span>
<a id="__codelineno-0-380" name="__codelineno-0-380" href="#__codelineno-0-380"></a><span class="sd">            An array with the communication scores for each intercellular</span>
<a id="__codelineno-0-381" name="__codelineno-0-381" href="#__codelineno-0-381"></a><span class="sd">            PPI.</span>
<a id="__codelineno-0-382" name="__codelineno-0-382" href="#__codelineno-0-382"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-383" name="__codelineno-0-383" href="#__codelineno-0-383"></a>        <span class="c1"># TODO: Implement communication scores</span>
<a id="__codelineno-0-384" name="__codelineno-0-384" href="#__codelineno-0-384"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-385" name="__codelineno-0-385" href="#__codelineno-0-385"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing communication score between </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell1</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>
<a id="__codelineno-0-386" name="__codelineno-0-386" href="#__codelineno-0-386"></a>
<a id="__codelineno-0-387" name="__codelineno-0-387" href="#__codelineno-0-387"></a>        <span class="c1"># Check that new score is the same type as score used to build interaction space (binary or continuous)</span>
<a id="__codelineno-0-388" name="__codelineno-0-388" href="#__codelineno-0-388"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">])</span> \
<a id="__codelineno-0-389" name="__codelineno-0-389" href="#__codelineno-0-389"></a>                <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-390" name="__codelineno-0-390" href="#__codelineno-0-390"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot use </span><span class="si">{}</span><span class="s1"> for this interaction space&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-391" name="__codelineno-0-391" href="#__codelineno-0-391"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">])</span> \
<a id="__codelineno-0-392" name="__codelineno-0-392" href="#__codelineno-0-392"></a>                <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-393" name="__codelineno-0-393" href="#__codelineno-0-393"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot use </span><span class="si">{}</span><span class="s1"> for this interaction space&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-394" name="__codelineno-0-394" href="#__codelineno-0-394"></a>
<a id="__codelineno-0-395" name="__codelineno-0-395" href="#__codelineno-0-395"></a>        <span class="k">if</span> <span class="n">use_ppi_score</span><span class="p">:</span>
<a id="__codelineno-0-396" name="__codelineno-0-396" href="#__codelineno-0-396"></a>            <span class="n">ppi_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-397" name="__codelineno-0-397" href="#__codelineno-0-397"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-398" name="__codelineno-0-398" href="#__codelineno-0-398"></a>            <span class="n">ppi_score</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-399" name="__codelineno-0-399" href="#__codelineno-0-399"></a>
<a id="__codelineno-0-400" name="__codelineno-0-400" href="#__codelineno-0-400"></a>        <span class="k">if</span> <span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-401" name="__codelineno-0-401" href="#__codelineno-0-401"></a>            <span class="n">communication_value</span> <span class="o">=</span> <span class="n">communication_scores</span><span class="o">.</span><span class="n">get_binary_scores</span><span class="p">(</span><span class="n">cell1</span><span class="o">=</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-402" name="__codelineno-0-402" href="#__codelineno-0-402"></a>                                                                         <span class="n">cell2</span><span class="o">=</span><span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-403" name="__codelineno-0-403" href="#__codelineno-0-403"></a>                                                                         <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-404" name="__codelineno-0-404" href="#__codelineno-0-404"></a>        <span class="k">elif</span> <span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-405" name="__codelineno-0-405" href="#__codelineno-0-405"></a>              <span class="n">communication_value</span> <span class="o">=</span> <span class="n">communication_scores</span><span class="o">.</span><span class="n">get_continuous_scores</span><span class="p">(</span><span class="n">cell1</span><span class="o">=</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-406" name="__codelineno-0-406" href="#__codelineno-0-406"></a>                                                                               <span class="n">cell2</span><span class="o">=</span><span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-407" name="__codelineno-0-407" href="#__codelineno-0-407"></a>                                                                               <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-408" name="__codelineno-0-408" href="#__codelineno-0-408"></a>                                                                               <span class="n">method</span><span class="o">=</span><span class="n">communication_score</span><span class="p">)</span>
<a id="__codelineno-0-409" name="__codelineno-0-409" href="#__codelineno-0-409"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-410" name="__codelineno-0-410" href="#__codelineno-0-410"></a>            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
<a id="__codelineno-0-411" name="__codelineno-0-411" href="#__codelineno-0-411"></a>                <span class="s2">&quot;Communication score </span><span class="si">{}</span><span class="s2"> to compute pairwise cell-communication is not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-412" name="__codelineno-0-412" href="#__codelineno-0-412"></a>        <span class="k">return</span> <span class="n">communication_value</span>
<a id="__codelineno-0-413" name="__codelineno-0-413" href="#__codelineno-0-413"></a>
<a id="__codelineno-0-414" name="__codelineno-0-414" href="#__codelineno-0-414"></a>    <span class="k">def</span> <span class="nf">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-415" name="__codelineno-0-415" href="#__codelineno-0-415"></a>                                              <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-416" name="__codelineno-0-416" href="#__codelineno-0-416"></a>        <span class="sd">&#39;&#39;&#39;Computes the communication scores for each LR pairs in</span>
<a id="__codelineno-0-417" name="__codelineno-0-417" href="#__codelineno-0-417"></a><span class="sd">        a given pair of sender-receiver cell</span>
<a id="__codelineno-0-418" name="__codelineno-0-418" href="#__codelineno-0-418"></a>
<a id="__codelineno-0-419" name="__codelineno-0-419" href="#__codelineno-0-419"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-420" name="__codelineno-0-420" href="#__codelineno-0-420"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-421" name="__codelineno-0-421" href="#__codelineno-0-421"></a><span class="sd">        communication_score : str, default=None</span>
<a id="__codelineno-0-422" name="__codelineno-0-422" href="#__codelineno-0-422"></a><span class="sd">            Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-423" name="__codelineno-0-423" href="#__codelineno-0-423"></a><span class="sd">            a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-424" name="__codelineno-0-424" href="#__codelineno-0-424"></a><span class="sd">            If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-425" name="__codelineno-0-425" href="#__codelineno-0-425"></a><span class="sd">            will be used.</span>
<a id="__codelineno-0-426" name="__codelineno-0-426" href="#__codelineno-0-426"></a><span class="sd">            Available communication_scores are:</span>
<a id="__codelineno-0-427" name="__codelineno-0-427" href="#__codelineno-0-427"></a>
<a id="__codelineno-0-428" name="__codelineno-0-428" href="#__codelineno-0-428"></a><span class="sd">            - &#39;expression_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-429" name="__codelineno-0-429" href="#__codelineno-0-429"></a><span class="sd">                                         ligand from a sender cell and of</span>
<a id="__codelineno-0-430" name="__codelineno-0-430" href="#__codelineno-0-430"></a><span class="sd">                                         a receptor on a receiver cell from</span>
<a id="__codelineno-0-431" name="__codelineno-0-431" href="#__codelineno-0-431"></a><span class="sd">                                         binarizing their gene expression levels.</span>
<a id="__codelineno-0-432" name="__codelineno-0-432" href="#__codelineno-0-432"></a><span class="sd">            - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-433" name="__codelineno-0-433" href="#__codelineno-0-433"></a><span class="sd">                                  of a ligand from a sender cell and the</span>
<a id="__codelineno-0-434" name="__codelineno-0-434" href="#__codelineno-0-434"></a><span class="sd">                                  expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-435" name="__codelineno-0-435" href="#__codelineno-0-435"></a><span class="sd">            - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-436" name="__codelineno-0-436" href="#__codelineno-0-436"></a><span class="sd">                                    of a ligand from a sender cell and the</span>
<a id="__codelineno-0-437" name="__codelineno-0-437" href="#__codelineno-0-437"></a><span class="sd">                                    expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-438" name="__codelineno-0-438" href="#__codelineno-0-438"></a><span class="sd">            - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-439" name="__codelineno-0-439" href="#__codelineno-0-439"></a><span class="sd">                                    of a ligand from a sender cell and the</span>
<a id="__codelineno-0-440" name="__codelineno-0-440" href="#__codelineno-0-440"></a><span class="sd">                                    expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-441" name="__codelineno-0-441" href="#__codelineno-0-441"></a>
<a id="__codelineno-0-442" name="__codelineno-0-442" href="#__codelineno-0-442"></a><span class="sd">        use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-443" name="__codelineno-0-443" href="#__codelineno-0-443"></a><span class="sd">            Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-444" name="__codelineno-0-444" href="#__codelineno-0-444"></a><span class="sd">            to compute the scores.</span>
<a id="__codelineno-0-445" name="__codelineno-0-445" href="#__codelineno-0-445"></a>
<a id="__codelineno-0-446" name="__codelineno-0-446" href="#__codelineno-0-446"></a><span class="sd">        ref_ppi_data : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-447" name="__codelineno-0-447" href="#__codelineno-0-447"></a><span class="sd">            Reference list of protein-protein interactions (or</span>
<a id="__codelineno-0-448" name="__codelineno-0-448" href="#__codelineno-0-448"></a><span class="sd">            ligand-receptor pairs) used for inferring the cell-cell</span>
<a id="__codelineno-0-449" name="__codelineno-0-449" href="#__codelineno-0-449"></a><span class="sd">            interactions and communication. It could be the same as</span>
<a id="__codelineno-0-450" name="__codelineno-0-450" href="#__codelineno-0-450"></a><span class="sd">            &#39;ppi_data&#39; if ppi_data is not bidirectional (that is,</span>
<a id="__codelineno-0-451" name="__codelineno-0-451" href="#__codelineno-0-451"></a><span class="sd">            contains ProtA-ProtB interaction as well as ProtB-ProtA</span>
<a id="__codelineno-0-452" name="__codelineno-0-452" href="#__codelineno-0-452"></a><span class="sd">            interaction). ref_ppi must be undirected (contains only</span>
<a id="__codelineno-0-453" name="__codelineno-0-453" href="#__codelineno-0-453"></a><span class="sd">            ProtA-ProtB and not ProtB-ProtA interaction). If None</span>
<a id="__codelineno-0-454" name="__codelineno-0-454" href="#__codelineno-0-454"></a><span class="sd">            the one stored in the attribute ref_ppi will be used.</span>
<a id="__codelineno-0-455" name="__codelineno-0-455" href="#__codelineno-0-455"></a>
<a id="__codelineno-0-456" name="__codelineno-0-456" href="#__codelineno-0-456"></a><span class="sd">        interaction_columns : tuple, default=None</span>
<a id="__codelineno-0-457" name="__codelineno-0-457" href="#__codelineno-0-457"></a><span class="sd">            Contains the names of the columns where to find the</span>
<a id="__codelineno-0-458" name="__codelineno-0-458" href="#__codelineno-0-458"></a><span class="sd">            partners in a dataframe of protein-protein interactions.</span>
<a id="__codelineno-0-459" name="__codelineno-0-459" href="#__codelineno-0-459"></a><span class="sd">            If the list is for ligand-receptor pairs, the first column</span>
<a id="__codelineno-0-460" name="__codelineno-0-460" href="#__codelineno-0-460"></a><span class="sd">            is for the ligands and the second for the receptors. If</span>
<a id="__codelineno-0-461" name="__codelineno-0-461" href="#__codelineno-0-461"></a><span class="sd">            None, the one stored in the attribute interaction_columns</span>
<a id="__codelineno-0-462" name="__codelineno-0-462" href="#__codelineno-0-462"></a><span class="sd">            will be used</span>
<a id="__codelineno-0-463" name="__codelineno-0-463" href="#__codelineno-0-463"></a>
<a id="__codelineno-0-464" name="__codelineno-0-464" href="#__codelineno-0-464"></a><span class="sd">        cells : list=None</span>
<a id="__codelineno-0-465" name="__codelineno-0-465" href="#__codelineno-0-465"></a><span class="sd">            List of cells to consider.</span>
<a id="__codelineno-0-466" name="__codelineno-0-466" href="#__codelineno-0-466"></a>
<a id="__codelineno-0-467" name="__codelineno-0-467" href="#__codelineno-0-467"></a><span class="sd">        cci_type : str, default=None</span>
<a id="__codelineno-0-468" name="__codelineno-0-468" href="#__codelineno-0-468"></a><span class="sd">            Type of interaction between two cells. Used to specify</span>
<a id="__codelineno-0-469" name="__codelineno-0-469" href="#__codelineno-0-469"></a><span class="sd">            if we want to consider a LR pair in both directions.</span>
<a id="__codelineno-0-470" name="__codelineno-0-470" href="#__codelineno-0-470"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-471" name="__codelineno-0-471" href="#__codelineno-0-471"></a><span class="sd">                - &#39;undirected&#39;</span>
<a id="__codelineno-0-472" name="__codelineno-0-472" href="#__codelineno-0-472"></a><span class="sd">                - &#39;directed</span>
<a id="__codelineno-0-473" name="__codelineno-0-473" href="#__codelineno-0-473"></a><span class="sd">            If None, the one stored in the attribute analysis_setup</span>
<a id="__codelineno-0-474" name="__codelineno-0-474" href="#__codelineno-0-474"></a><span class="sd">            will be used.</span>
<a id="__codelineno-0-475" name="__codelineno-0-475" href="#__codelineno-0-475"></a>
<a id="__codelineno-0-476" name="__codelineno-0-476" href="#__codelineno-0-476"></a><span class="sd">        verbose : boolean, default=True</span>
<a id="__codelineno-0-477" name="__codelineno-0-477" href="#__codelineno-0-477"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-478" name="__codelineno-0-478" href="#__codelineno-0-478"></a>
<a id="__codelineno-0-479" name="__codelineno-0-479" href="#__codelineno-0-479"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-480" name="__codelineno-0-480" href="#__codelineno-0-480"></a><span class="sd">        -------</span>
<a id="__codelineno-0-481" name="__codelineno-0-481" href="#__codelineno-0-481"></a><span class="sd">        self.interaction_elements[&#39;communication_matrix&#39;] : pandas.DataFrame</span>
<a id="__codelineno-0-482" name="__codelineno-0-482" href="#__codelineno-0-482"></a><span class="sd">            Contains communication scores for each LR pair in a</span>
<a id="__codelineno-0-483" name="__codelineno-0-483" href="#__codelineno-0-483"></a><span class="sd">            given pair of sender-receiver cells.</span>
<a id="__codelineno-0-484" name="__codelineno-0-484" href="#__codelineno-0-484"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-485" name="__codelineno-0-485" href="#__codelineno-0-485"></a>        <span class="k">if</span> <span class="n">communication_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-486" name="__codelineno-0-486" href="#__codelineno-0-486"></a>            <span class="n">communication_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span>
<a id="__codelineno-0-487" name="__codelineno-0-487" href="#__codelineno-0-487"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-488" name="__codelineno-0-488" href="#__codelineno-0-488"></a>            <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">communication_score</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
<a id="__codelineno-0-489" name="__codelineno-0-489" href="#__codelineno-0-489"></a>
<a id="__codelineno-0-490" name="__codelineno-0-490" href="#__codelineno-0-490"></a>        <span class="c1"># Cells to consider</span>
<a id="__codelineno-0-491" name="__codelineno-0-491" href="#__codelineno-0-491"></a>        <span class="k">if</span> <span class="n">cells</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-492" name="__codelineno-0-492" href="#__codelineno-0-492"></a>            <span class="n">cells</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cell_names&#39;</span><span class="p">]</span>
<a id="__codelineno-0-493" name="__codelineno-0-493" href="#__codelineno-0-493"></a>
<a id="__codelineno-0-494" name="__codelineno-0-494" href="#__codelineno-0-494"></a>        <span class="c1"># Labels:</span>
<a id="__codelineno-0-495" name="__codelineno-0-495" href="#__codelineno-0-495"></a>        <span class="k">if</span> <span class="n">cci_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-496" name="__codelineno-0-496" href="#__codelineno-0-496"></a>            <span class="n">cell_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]</span>
<a id="__codelineno-0-497" name="__codelineno-0-497" href="#__codelineno-0-497"></a>        <span class="k">elif</span> <span class="n">cci_type</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span><span class="p">:</span>
<a id="__codelineno-0-498" name="__codelineno-0-498" href="#__codelineno-0-498"></a>            <span class="n">cell_pairs</span> <span class="o">=</span> <span class="n">generate_pairs</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cci_type</span><span class="p">)</span>
<a id="__codelineno-0-499" name="__codelineno-0-499" href="#__codelineno-0-499"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-500" name="__codelineno-0-500" href="#__codelineno-0-500"></a>            <span class="c1">#cell_pairs = generate_pairs(cells, self.cci_type) # Think about other scenarios that may need this line</span>
<a id="__codelineno-0-501" name="__codelineno-0-501" href="#__codelineno-0-501"></a>            <span class="n">cell_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]</span>
<a id="__codelineno-0-502" name="__codelineno-0-502" href="#__codelineno-0-502"></a>        <span class="n">col_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">cell_pairs</span><span class="p">]</span>
<a id="__codelineno-0-503" name="__codelineno-0-503" href="#__codelineno-0-503"></a>
<a id="__codelineno-0-504" name="__codelineno-0-504" href="#__codelineno-0-504"></a>        <span class="c1"># Ref PPI data</span>
<a id="__codelineno-0-505" name="__codelineno-0-505" href="#__codelineno-0-505"></a>        <span class="k">if</span> <span class="n">ref_ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-506" name="__codelineno-0-506" href="#__codelineno-0-506"></a>            <span class="n">ref_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-507" name="__codelineno-0-507" href="#__codelineno-0-507"></a>            <span class="n">keep_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<a id="__codelineno-0-508" name="__codelineno-0-508" href="#__codelineno-0-508"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-509" name="__codelineno-0-509" href="#__codelineno-0-509"></a>            <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-510" name="__codelineno-0-510" href="#__codelineno-0-510"></a>            <span class="n">prot_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-511" name="__codelineno-0-511" href="#__codelineno-0-511"></a>            <span class="n">prot_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-512" name="__codelineno-0-512" href="#__codelineno-0-512"></a>            <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span> <span class="ow">in</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_a</span> <span class="o">!=</span> <span class="s1">&#39;A&#39;</span><span class="p">):</span>
<a id="__codelineno-0-513" name="__codelineno-0-513" href="#__codelineno-0-513"></a>                <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<a id="__codelineno-0-514" name="__codelineno-0-514" href="#__codelineno-0-514"></a>            <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;B&#39;</span> <span class="ow">in</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_b</span> <span class="o">!=</span> <span class="s1">&#39;B&#39;</span><span class="p">):</span>
<a id="__codelineno-0-515" name="__codelineno-0-515" href="#__codelineno-0-515"></a>                <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">)</span>
<a id="__codelineno-0-516" name="__codelineno-0-516" href="#__codelineno-0-516"></a>            <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">prot_a</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">prot_b</span><span class="p">:</span> <span class="s1">&#39;B&#39;</span><span class="p">})</span>
<a id="__codelineno-0-517" name="__codelineno-0-517" href="#__codelineno-0-517"></a>            <span class="n">ref_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ref_ppi</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-518" name="__codelineno-0-518" href="#__codelineno-0-518"></a>            <span class="n">keep_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">ref_ppi</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-519" name="__codelineno-0-519" href="#__codelineno-0-519"></a>
<a id="__codelineno-0-520" name="__codelineno-0-520" href="#__codelineno-0-520"></a>        <span class="c1"># DataFrame to Store values</span>
<a id="__codelineno-0-521" name="__codelineno-0-521" href="#__codelineno-0-521"></a>        <span class="n">communication_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">ref_index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col_labels</span><span class="p">)</span>
<a id="__codelineno-0-522" name="__codelineno-0-522" href="#__codelineno-0-522"></a>
<a id="__codelineno-0-523" name="__codelineno-0-523" href="#__codelineno-0-523"></a>        <span class="c1">### Compute pairwise physical interactions</span>
<a id="__codelineno-0-524" name="__codelineno-0-524" href="#__codelineno-0-524"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-525" name="__codelineno-0-525" href="#__codelineno-0-525"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing pairwise communication&quot;</span><span class="p">)</span>
<a id="__codelineno-0-526" name="__codelineno-0-526" href="#__codelineno-0-526"></a>
<a id="__codelineno-0-527" name="__codelineno-0-527" href="#__codelineno-0-527"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pair</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cell_pairs</span><span class="p">):</span>
<a id="__codelineno-0-528" name="__codelineno-0-528" href="#__codelineno-0-528"></a>            <span class="n">cell1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<a id="__codelineno-0-529" name="__codelineno-0-529" href="#__codelineno-0-529"></a>            <span class="n">cell2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<a id="__codelineno-0-530" name="__codelineno-0-530" href="#__codelineno-0-530"></a>
<a id="__codelineno-0-531" name="__codelineno-0-531" href="#__codelineno-0-531"></a>            <span class="n">comm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pair_communication_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-532" name="__codelineno-0-532" href="#__codelineno-0-532"></a>                                                       <span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-533" name="__codelineno-0-533" href="#__codelineno-0-533"></a>                                                       <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-534" name="__codelineno-0-534" href="#__codelineno-0-534"></a>                                                       <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-535" name="__codelineno-0-535" href="#__codelineno-0-535"></a>                                                       <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-536" name="__codelineno-0-536" href="#__codelineno-0-536"></a>            <span class="n">kept_values</span> <span class="o">=</span> <span class="n">comm_score</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">keep_index</span><span class="p">]</span>
<a id="__codelineno-0-537" name="__codelineno-0-537" href="#__codelineno-0-537"></a>            <span class="n">communication_matrix</span><span class="p">[</span><span class="n">col_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">kept_values</span>
<a id="__codelineno-0-538" name="__codelineno-0-538" href="#__codelineno-0-538"></a>
<a id="__codelineno-0-539" name="__codelineno-0-539" href="#__codelineno-0-539"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">communication_matrix</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">










  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_cci_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes overall CCI scores for each pair of cells.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_cci_scores--parameters">Parameters</h6>
<p>cci_score : str, default=None
    Scoring function to aggregate the communication scores between
    a pair of cells. It computes an overall potential of cell-cell
    interactions. If None, it will use the one stored in the
    attribute analysis_setup of this object.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score
- &#39;jaccard&#39; : Jaccard-like score
- &#39;count&#39; : Number of LR pairs that the pair of cells uses
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_cci_scores--returns">Returns</h6>
<p>self.interaction_elements['cci_matrix'] : pandas.DataFrame
    Contains CCI scores for each pair of cells</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_cci_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes overall CCI scores for each pair of cells.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cci_score : str, default=None</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        attribute analysis_setup of this object.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells uses</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    -------</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    self.interaction_elements[&#39;cci_matrix&#39;] : pandas.DataFrame</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Contains CCI scores for each pair of cells</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">cci_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_score</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cci_score</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="c1">### Compute pairwise physical interactions</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing pairwise interactions&quot;</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="c1"># Compute pair by pair</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">cell1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">cell2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="n">cci_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pair_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                        <span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                        <span class="n">cci_score</span><span class="o">=</span><span class="n">cci_score</span><span class="p">,</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                                        <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                                        <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">cci_value</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">cci_value</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="c1"># Compute using matmul -&gt; Too slow and uses a lot of memory TODO: Try to optimize this</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="c1"># if cci_score == &#39;bray_curtis&#39;:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="c1">#     cci_matrix = cci_scores.matmul_bray_curtis_like(self.interaction_elements[&#39;A_score&#39;],</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="c1">#                                                     self.interaction_elements[&#39;B_score&#39;])</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="c1"># self.interaction_elements[&#39;cci_matrix&#39;] = pd.DataFrame(cci_matrix,</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="c1">#                                                        index=self.interaction_elements[&#39;cell_names&#39;],</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="c1">#                                                        columns=self.interaction_elements[&#39;cell_names&#39;]</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="c1">#                                                        )</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="c1"># Generate distance matrix</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="k">if</span> <span class="o">~</span><span class="p">(</span><span class="n">cci_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;icellnet&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="c1">#self.distance_matrix = self.interaction_elements[&#39;cci_matrix&#39;].div(self.interaction_elements[&#39;cci_matrix&#39;].max().max()).apply(lambda x: 1 - x)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="c1"># Regularized distance</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="n">mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">])</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">mean</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>  <span class="c1"># Make diagonal zero (delete autocrine-interactions)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_communication_scores">
<code class="highlight language-python"><span class="n">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes the communication scores for each LR pairs in
a given pair of sender-receiver cell</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_communication_scores--parameters">Parameters</h6>
<p>communication_score : str, default=None
    Type of communication score to infer the potential use of
    a given ligand-receptor pair by a pair of cells/tissues/samples.
    If None, the score stored in the attribute analysis_setup
    will be used.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39; : Computes the joint presence of a
                             ligand from a sender cell and of
                             a receptor on a receiver cell from
                             binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression
                        of a ligand from a sender cell and the
                        expression of a receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                        of a ligand from a sender cell and the
                        expression of a receptor on a receiver cell.
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>ref_ppi_data : pandas.DataFrame, default=None
    Reference list of protein-protein interactions (or
    ligand-receptor pairs) used for inferring the cell-cell
    interactions and communication. It could be the same as
    'ppi_data' if ppi_data is not bidirectional (that is,
    contains ProtA-ProtB interaction as well as ProtB-ProtA
    interaction). ref_ppi must be undirected (contains only
    ProtA-ProtB and not ProtB-ProtA interaction). If None
    the one stored in the attribute ref_ppi will be used.</p>
<p>interaction_columns : tuple, default=None
    Contains the names of the columns where to find the
    partners in a dataframe of protein-protein interactions.
    If the list is for ligand-receptor pairs, the first column
    is for the ligands and the second for the receptors. If
    None, the one stored in the attribute interaction_columns
    will be used</p>
<p>cells : list=None
    List of cells to consider.</p>
<p>cci_type : str, default=None
    Type of interaction between two cells. Used to specify
    if we want to consider a LR pair in both directions.
    It can be:
        - 'undirected'
        - 'directed
    If None, the one stored in the attribute analysis_setup
    will be used.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.compute_pairwise_communication_scores--returns">Returns</h6>
<p>self.interaction_elements['communication_matrix'] : pandas.DataFrame
    Contains communication scores for each LR pair in a
    given pair of sender-receiver cells.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pairwise_communication_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ref_ppi_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Computes the communication scores for each LR pairs in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a given pair of sender-receiver cell</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    communication_score : str, default=None</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;expression_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">                                     ligand from a sender cell and of</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">                                     a receptor on a receiver cell from</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">                                     binarizing their gene expression levels.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                                of a ligand from a sender cell and the</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                                expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                of a ligand from a sender cell and the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                                expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    ref_ppi_data : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Reference list of protein-protein interactions (or</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        ligand-receptor pairs) used for inferring the cell-cell</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        interactions and communication. It could be the same as</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        &#39;ppi_data&#39; if ppi_data is not bidirectional (that is,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        contains ProtA-ProtB interaction as well as ProtB-ProtA</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        interaction). ref_ppi must be undirected (contains only</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        ProtA-ProtB and not ProtB-ProtA interaction). If None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        the one stored in the attribute ref_ppi will be used.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    interaction_columns : tuple, default=None</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Contains the names of the columns where to find the</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        partners in a dataframe of protein-protein interactions.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        If the list is for ligand-receptor pairs, the first column</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        is for the ligands and the second for the receptors. If</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        None, the one stored in the attribute interaction_columns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        will be used</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    cells : list=None</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        List of cells to consider.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    cci_type : str, default=None</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Type of interaction between two cells. Used to specify</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        if we want to consider a LR pair in both directions.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">            - &#39;undirected&#39;</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">            - &#39;directed</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        If None, the one stored in the attribute analysis_setup</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    -------</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    self.interaction_elements[&#39;communication_matrix&#39;] : pandas.DataFrame</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        Contains communication scores for each LR pair in a</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        given pair of sender-receiver cells.</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">if</span> <span class="n">communication_score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">communication_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">communication_score</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="c1"># Cells to consider</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="k">if</span> <span class="n">cells</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cell_names&#39;</span><span class="p">]</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="c1"># Labels:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="k">if</span> <span class="n">cci_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="n">cell_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="k">elif</span> <span class="n">cci_type</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cci_type</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">cell_pairs</span> <span class="o">=</span> <span class="n">generate_pairs</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cci_type</span><span class="p">)</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="c1">#cell_pairs = generate_pairs(cells, self.cci_type) # Think about other scenarios that may need this line</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="n">cell_pairs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">col_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">cell_pairs</span><span class="p">]</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="c1"># Ref PPI data</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="k">if</span> <span class="n">ref_ppi_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>        <span class="n">ref_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="n">keep_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="n">prot_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="n">prot_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span> <span class="ow">in</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_a</span> <span class="o">!=</span> <span class="s1">&#39;A&#39;</span><span class="p">):</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>            <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;B&#39;</span> <span class="ow">in</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">prot_b</span> <span class="o">!=</span> <span class="s1">&#39;B&#39;</span><span class="p">):</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>            <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">)</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>        <span class="n">ref_ppi</span> <span class="o">=</span> <span class="n">ref_ppi</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">prot_a</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">prot_b</span><span class="p">:</span> <span class="s1">&#39;B&#39;</span><span class="p">})</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>        <span class="n">ref_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ref_ppi</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>        <span class="n">keep_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">ref_ppi</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="c1"># DataFrame to Store values</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="n">communication_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">ref_index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col_labels</span><span class="p">)</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="c1">### Compute pairwise physical interactions</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing pairwise communication&quot;</span><span class="p">)</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pair</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cell_pairs</span><span class="p">):</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">cell1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="n">cell2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">][</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="n">comm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pair_communication_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>                                                   <span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>                                                   <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                                                   <span class="n">use_ppi_score</span><span class="o">=</span><span class="n">use_ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>                                                   <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="n">kept_values</span> <span class="o">=</span> <span class="n">comm_score</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">keep_index</span><span class="p">]</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>        <span class="n">communication_matrix</span><span class="p">[</span><span class="n">col_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">kept_values</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">communication_matrix</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.core.interaction_space.InteractionSpace.pair_cci_score">
<code class="highlight language-python"><span class="n">pair_cci_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes a CCI score for a pair of cells.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.pair_cci_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the receiver.</p>
<p>cci_score : str, default='bray_curtis'
    Scoring function to aggregate the communication scores between
    a pair of cells. It computes an overall potential of cell-cell
    interactions. If None, it will use the one stored in the
    attribute analysis_setup of this object.
    Options:</p>
<div class="codehilite"><pre><span></span><code>- &#39;bray_curtis&#39; : Bray-Curtis-like score
- &#39;jaccard&#39; : Jaccard-like score
- &#39;count&#39; : Number of LR pairs that the pair of cells uses
- &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.pair_cci_score--returns">Returns</h6>
<p>cci_score : float
    Overall score for the interaction between a pair of
    cell-types/tissues/samples. In this case it is a
    Jaccard-like score.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pair_cci_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">cci_score</span><span class="o">=</span><span class="s1">&#39;bray_curtis&#39;</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes a CCI score for a pair of cells.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    cci_score : str, default=&#39;bray_curtis&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Scoring function to aggregate the communication scores between</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        a pair of cells. It computes an overall potential of cell-cell</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        interactions. If None, it will use the one stored in the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        attribute analysis_setup of this object.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Options:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;bray_curtis&#39; : Bray-Curtis-like score</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;jaccard&#39; : Jaccard-like score</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        - &#39;count&#39; : Number of LR pairs that the pair of cells uses</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;icellnet&#39; : Sum of the L-R expression product of a pair of cells</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    -------</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    cci_score : float</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Overall score for the interaction between a pair of</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        cell-types/tissues/samples. In this case it is a</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Jaccard-like score.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing interaction score between </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell1</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">if</span> <span class="n">use_ppi_score</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="c1"># Calculate cell-cell interaction score</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">if</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;bray_curtis&#39;</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_braycurtis_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;jaccard&#39;</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_jaccard_like_cci_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;count&#39;</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_count_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">elif</span> <span class="n">cci_score</span> <span class="o">==</span> <span class="s1">&#39;icellnet&#39;</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="n">cci_value</span> <span class="o">=</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">compute_icellnet_score</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;CCI score </span><span class="si">{}</span><span class="s2"> to compute pairwise cell-interactions is not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cci_score</span><span class="p">))</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">return</span> <span class="n">cci_value</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.core.interaction_space.InteractionSpace.pair_communication_score">
<code class="highlight language-python"><span class="n">pair_communication_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes a communication score for each protein-protein interaction
between a pair of cells.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.pair_communication_score--parameters">Parameters</h6>
<p>cell1 : cell2cell.core.cell.Cell
    First cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the sender.</p>
<p>cell2 : cell2cell.core.cell.Cell
    Second cell-type/tissue/sample to compute the communication
    score. In a directed interaction, this is the receiver.</p>
<p>communication_score : str, default=None
    Type of communication score to infer the potential use of
    a given ligand-receptor pair by a pair of cells/tissues/samples.
    If None, the score stored in the attribute analysis_setup
    will be used.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39; : Computes the joint presence of a
                             ligand from a sender cell and of
                             a receptor on a receiver cell from
                             binarizing their gene expression levels.
- &#39;expression_mean&#39; : Computes the average between the expression
                      of a ligand from a sender cell and the
                      expression of a receptor on a receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression
                        of a ligand from a sender cell and the
                        expression of a receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                       of a ligand from a sender cell and the
                       expression of a receptor on a receiver cell.
</code></pre></div>

<p>use_ppi_score : boolean, default=False
    Whether using a weight of LR pairs specified in the ppi_data
    to compute the scores.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.InteractionSpace.pair_communication_score--returns">Returns</h6>
<p>communication_scores : numpy.array
    An array with the communication scores for each intercellular
    PPI.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pair_communication_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                             <span class="n">use_ppi_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Computes a communication score for each protein-protein interaction</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    between a pair of cells.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    cell1 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        First cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        score. In a directed interaction, this is the sender.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    cell2 : cell2cell.core.cell.Cell</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Second cell-type/tissue/sample to compute the communication</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        score. In a directed interaction, this is the receiver.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    communication_score : str, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Type of communication score to infer the potential use of</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        a given ligand-receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        If None, the score stored in the attribute analysis_setup</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        will be used.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;expression_thresholding&#39; : Computes the joint presence of a</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                                     ligand from a sender cell and of</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">                                     a receptor on a receiver cell from</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                     binarizing their gene expression levels.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                              of a ligand from a sender cell and the</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                              expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                                of a ligand from a sender cell and the</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                                expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                               of a ligand from a sender cell and the</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">                               expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    use_ppi_score : boolean, default=False</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Whether using a weight of LR pairs specified in the ppi_data</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        to compute the scores.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    -------</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    communication_scores : numpy.array</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        An array with the communication scores for each intercellular</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        PPI.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="c1"># TODO: Implement communication scores</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing communication score between </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell1</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="c1"># Check that new score is the same type as score used to build interaction space (binary or continuous)</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">])</span> \
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>            <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot use </span><span class="si">{}</span><span class="s1"> for this interaction space&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">])</span> \
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>            <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">]):</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot use </span><span class="si">{}</span><span class="s1"> for this interaction space&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="k">if</span> <span class="n">use_ppi_score</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="n">ppi_score</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">,</span> <span class="s1">&#39;differential_combinations&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">communication_value</span> <span class="o">=</span> <span class="n">communication_scores</span><span class="o">.</span><span class="n">get_binary_scores</span><span class="p">(</span><span class="n">cell1</span><span class="o">=</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                                                     <span class="n">cell2</span><span class="o">=</span><span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                                                     <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">elif</span> <span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_correlation&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>          <span class="n">communication_value</span> <span class="o">=</span> <span class="n">communication_scores</span><span class="o">.</span><span class="n">get_continuous_scores</span><span class="p">(</span><span class="n">cell1</span><span class="o">=</span><span class="n">cell1</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                                                           <span class="n">cell2</span><span class="o">=</span><span class="n">cell2</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                                           <span class="n">ppi_score</span><span class="o">=</span><span class="n">ppi_score</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                                           <span class="n">method</span><span class="o">=</span><span class="n">communication_score</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>            <span class="s2">&quot;Communication score </span><span class="si">{}</span><span class="s2"> to compute pairwise cell-communication is not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="k">return</span> <span class="n">communication_value</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>




  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.interaction_space.generate_interaction_elements">
<code class="highlight language-python"><span class="n">generate_interaction_elements</span><span class="p">(</span><span class="n">modified_rnaseq</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="s1">&#39;undirected&#39;</span><span class="p">,</span> <span class="n">cci_matrix_template</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Create all elements needed to perform the analyses of pairwise
cell-cell interactions/communication. Corresponds to the interaction
elements used by the class InteractionSpace.</p>
<h6 id="cell2cell.core.interaction_space.generate_interaction_elements--parameters">Parameters</h6>
<p>modified_rnaseq : pandas.DataFrame
    Preprocessed gene expression data for a bulk or single-cell RNA-seq experiment.
    Columns are are cell-types/tissues/samples and rows are genes. The preprocessing
    may correspond to scoring the gene expression as binary or continuous values
    depending on the scoring function for cell-cell interactions/communication.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used for
    inferring the cell-cell interactions and communication.</p>
<p>cci_type : str, default='undirected'
    Specifies whether computing the cci_score in a directed or undirected
    way. For a pair of cells A and B, directed means that the ligands are
    considered only from cell A and receptors only from cell B or viceversa.
    While undirected simultaneously considers signaling from cell A to
    cell B and from cell B to cell A.</p>
<p>cci_matrix_template : pandas.DataFrame, default=None
    A matrix of shape MxM where M are cell-types/tissues/samples. This
    is used as template for storing CCI scores. It may be useful
    for specifying which pairs of cells to consider.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.core.interaction_space.generate_interaction_elements--returns">Returns</h6>
<p>interaction_elements : dict
    Dictionary containing all the pairs of cells considered (under
    the key of 'pairs'), Cell instances (under key 'cells')
    which include all cells/tissues/organs with their associated datasets
    (rna_seq, weighted_ppi, etc) and a Cell-Cell Interaction Matrix
    to store CCI scores(under key 'cci_matrix'). A communication matrix
    is also stored in this object when the communication scores are
    computed in the InteractionSpace class (under key
    'communication_score')</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_interaction_elements</span><span class="p">(</span><span class="n">modified_rnaseq</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">cci_type</span><span class="o">=</span><span class="s1">&#39;undirected&#39;</span><span class="p">,</span> <span class="n">cci_matrix_template</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                  <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                                  <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Create all elements needed to perform the analyses of pairwise</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    cell-cell interactions/communication. Corresponds to the interaction</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    elements used by the class InteractionSpace.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    modified_rnaseq : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Preprocessed gene expression data for a bulk or single-cell RNA-seq experiment.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Columns are are cell-types/tissues/samples and rows are genes. The preprocessing</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        may correspond to scoring the gene expression as binary or continuous values</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        depending on the scoring function for cell-cell interactions/communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used for</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    cci_type : str, default=&#39;undirected&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Specifies whether computing the cci_score in a directed or undirected</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        way. For a pair of cells A and B, directed means that the ligands are</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        considered only from cell A and receptors only from cell B or viceversa.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        While undirected simultaneously considers signaling from cell A to</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        cell B and from cell B to cell A.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    cci_matrix_template : pandas.DataFrame, default=None</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        A matrix of shape MxM where M are cell-types/tissues/samples. This</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        is used as template for storing CCI scores. It may be useful</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        for specifying which pairs of cells to consider.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    -------</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    interaction_elements : dict</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Dictionary containing all the pairs of cells considered (under</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        the key of &#39;pairs&#39;), Cell instances (under key &#39;cells&#39;)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        which include all cells/tissues/organs with their associated datasets</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        (rna_seq, weighted_ppi, etc) and a Cell-Cell Interaction Matrix</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        to store CCI scores(under key &#39;cci_matrix&#39;). A communication matrix</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        is also stored in this object when the communication scores are</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        computed in the InteractionSpace class (under key</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        &#39;communication_score&#39;)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Creating Interaction Space&#39;</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="c1"># Include complex expression</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="n">col_a_genes</span><span class="p">,</span> <span class="n">complex_a</span><span class="p">,</span> <span class="n">col_b_genes</span><span class="p">,</span> <span class="n">complex_b</span><span class="p">,</span> <span class="n">complexes</span> <span class="o">=</span> <span class="n">get_genes_from_complexes</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                                                                             <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                                                             <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                                                             <span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>        <span class="n">modified_rnaseq</span> <span class="o">=</span> <span class="n">add_complexes_to_expression</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">modified_rnaseq</span><span class="p">,</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                                                      <span class="n">complexes</span><span class="o">=</span><span class="n">complexes</span><span class="p">,</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                                                      <span class="n">agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                      <span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="c1"># Cells</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="n">cell_instances</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">modified_rnaseq</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>  <span class="c1"># @Erick, check if position 0 of columns contain index header.</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="n">cell_number</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_instances</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>    <span class="c1"># Generate pairwise interactions</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="n">pairwise_interactions</span> <span class="o">=</span> <span class="n">generate_pairs</span><span class="p">(</span><span class="n">cell_instances</span><span class="p">,</span> <span class="n">cci_type</span><span class="p">)</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>    <span class="c1"># Interaction elements</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">interaction_elements</span> <span class="o">=</span> <span class="p">{}</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cell_names&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_instances</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;pairs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pairwise_interactions</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell</span><span class="o">.</span><span class="n">get_cells_from_rnaseq</span><span class="p">(</span><span class="n">modified_rnaseq</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="c1"># Cell-specific scores in PPIs</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>    <span class="c1"># For matmul functions</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="c1">#interaction_elements[&#39;A_score&#39;] = np.array([], dtype=np.int64)#.reshape(ppi_data.shape[0],0)</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="c1">#interaction_elements[&#39;B_score&#39;] = np.array([], dtype=np.int64)#.reshape(ppi_data.shape[0],0)</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="c1"># For &#39;for&#39; loop</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">for</span> <span class="n">cell_instance</span> <span class="ow">in</span> <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cells&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="n">cell_instance</span><span class="o">.</span><span class="n">weighted_ppi</span> <span class="o">=</span> <span class="n">integrate_data</span><span class="o">.</span><span class="n">get_weighted_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                                                                     <span class="n">modified_rnaseq_data</span><span class="o">=</span><span class="n">cell_instance</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">,</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>                                                                     <span class="n">column</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">,</span> <span class="c1"># value is in each cell</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>                                                                     <span class="p">)</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>        <span class="c1">#interaction_elements[&#39;A_score&#39;] = np.hstack([interaction_elements[&#39;A_score&#39;], cell_instance.weighted_ppi[&#39;A&#39;].values])</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>        <span class="c1">#interaction_elements[&#39;B_score&#39;] = np.hstack([interaction_elements[&#39;B_score&#39;], cell_instance.weighted_ppi[&#39;B&#39;].values])</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="c1"># Cell-cell interaction matrix</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="k">if</span> <span class="n">cci_matrix_template</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>        <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">cell_number</span><span class="p">,</span> <span class="n">cell_number</span><span class="p">)),</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>                                                          <span class="n">columns</span><span class="o">=</span><span class="n">cell_instances</span><span class="p">,</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>                                                          <span class="n">index</span><span class="o">=</span><span class="n">cell_instances</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cci_matrix_template</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">return</span> <span class="n">interaction_elements</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.core.interaction_space.generate_pairs">
<code class="highlight language-python"><span class="n">generate_pairs</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cci_type</span><span class="p">,</span> <span class="n">self_interaction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a list of pairs of interacting cell-types/tissues/samples.</p>
<h6 id="cell2cell.core.interaction_space.generate_pairs--parameters">Parameters</h6>
<p>cells : list
    A lyst of cell-type/tissue/sample names.</p>
<p>cci_type : str,
    Type of interactions.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;directed&#39; : Directed cell-cell interactions, so pair A-B is different
    to pair B-A and both are considered.
- &#39;undirected&#39; : Undirected cell-cell interactions, so pair A-B is equal
    to pair B-A and just one of them is considered.
</code></pre></div>

<p>self_interaction : boolean, default=True
    Whether considering autocrine interactions (pair A-A, B-B, etc).</p>
<p>remove_duplicates : boolean, default=True
    Whether removing duplicates when a list of cells is passed and names are
    duplicated. If False and a list [A, A, B] is passed, pairs could be
    [A-A, A-A, A-B, A-A, A-A, A-B, B-A, B-A, B-B] when self_interaction is True
    and cci_type is 'directed'. In the same scenario but when remove_duplicates
    is True, the resulting list would be [A-A, A-B, B-A, B-B].</p>
<h6 id="cell2cell.core.interaction_space.generate_pairs--returns">Returns</h6>
<p>pairs : list
    List with pairs of interacting cell-types/tissues/samples.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/core/interaction_space.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_pairs</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cci_type</span><span class="p">,</span> <span class="n">self_interaction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a list of pairs of interacting cell-types/tissues/samples.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cells : list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A lyst of cell-type/tissue/sample names.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    cci_type : str,</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Type of interactions.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;directed&#39; : Directed cell-cell interactions, so pair A-B is different</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">            to pair B-A and both are considered.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;undirected&#39; : Undirected cell-cell interactions, so pair A-B is equal</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">            to pair B-A and just one of them is considered.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    self_interaction : boolean, default=True</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether considering autocrine interactions (pair A-A, B-B, etc).</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    remove_duplicates : boolean, default=True</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Whether removing duplicates when a list of cells is passed and names are</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        duplicated. If False and a list [A, A, B] is passed, pairs could be</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        [A-A, A-A, A-B, A-A, A-A, A-B, B-A, B-A, B-B] when self_interaction is True</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        and cci_type is &#39;directed&#39;. In the same scenario but when remove_duplicates</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        is True, the resulting list would be [A-A, A-B, B-A, B-B].</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    pairs : list</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        List with pairs of interacting cell-types/tissues/samples.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">if</span> <span class="n">self_interaction</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">if</span> <span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;directed&#39;</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cells</span><span class="p">))</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="c1">#pairs = list(itertools.combinations(cells + cells, 2)) # Directed</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">elif</span> <span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span> <span class="o">+</span> <span class="p">[(</span><span class="n">c</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">]</span> <span class="c1"># Undirected</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;CCI type has to be directed or undirected&quot;</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="k">if</span> <span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;directed&#39;</span><span class="p">:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>            <span class="n">pairs_</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cells</span><span class="p">))</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pairs_</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                <span class="k">if</span> <span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">p</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                    <span class="k">continue</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>                    <span class="n">pairs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="k">elif</span> <span class="n">cci_type</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;CCI type has to be directed or undirected&quot;</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">if</span> <span class="n">remove_duplicates</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pairs</span><span class="p">))</span>  <span class="c1"># Remove duplicates</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">return</span> <span class="n">pairs</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.datasets">
        <code>datasets</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.datasets.anndata">
        <code>anndata</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.anndata.balf_covid">
<code class="highlight language-python"><span class="n">balf_covid</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s1">&#39;BALF-COVID19-Liao_et_al-NatMed-2020.h5ad&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>BALF samples from COVID-19 patients
The data consists in 63k immune and epithelial cells in lungs
from 3 control, 3 moderate COVID-19, and 6 severe COVID-19 patients.</p>
<p>This dataset was previously published in [1], and this objects contains
the raw counts for the annotated cell types available in:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926</p>
<p>References:
[1] Liao, M., Liu, Y., Yuan, J. et al.
    Single-cell landscape of bronchoalveolar immune cells in patients
    with COVID-19. Nat Med 26, 842844 (2020).
    https://doi.org/10.1038/s41591-020-0901-9</p>
<h6 id="cell2cell.datasets.anndata.balf_covid--parameters">Parameters</h6>
<div class="codehilite"><pre><span></span><code><span class="n">filename</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">str</span><span class="p">,</span><span class="w"> </span><span class="n">default</span><span class="o">=</span><span class="s1">&#39;BALF-COVID19-Liao_et_al-NatMed-2020.h5ad&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">Path</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">h5ad</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">was</span><span class="w"> </span><span class="n">manually</span><span class="w"> </span><span class="n">downloaded</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<h6 id="cell2cell.datasets.anndata.balf_covid--returns">Returns</h6>
<div class="codehilite"><pre><span></span><code>Annotated data matrix.
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/anndata.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">balf_covid</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s1">&#39;BALF-COVID19-Liao_et_al-NatMed-2020.h5ad&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;BALF samples from COVID-19 patients</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    The data consists in 63k immune and epithelial cells in lungs</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    from 3 control, 3 moderate COVID-19, and 6 severe COVID-19 patients.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    This dataset was previously published in [1], and this objects contains</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    the raw counts for the annotated cell types available in:</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    References:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    [1] Liao, M., Liu, Y., Yuan, J. et al.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Single-cell landscape of bronchoalveolar immune cells in patients</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        with COVID-19. Nat Med 26, 842844 (2020).</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        https://doi.org/10.1038/s41591-020-0901-9</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        filename : str, default=&#39;BALF-COVID19-Liao_et_al-NatMed-2020.h5ad&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">            Path to the h5ad file in case it was manually downloaded.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    -------</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Annotated data matrix.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://zenodo.org/record/7535867/files/BALF-COVID19-Liao_et_al-NatMed-2020.h5ad&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">adata</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">backup_url</span><span class="o">=</span><span class="n">url</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.datasets.gsea_data">
        <code>gsea_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.gsea_data.gsea_msig">
<code class="highlight language-python"><span class="n">gsea_msig</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="o">=</span><span class="s1">&#39;GOBP&#39;</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Load a MSigDB from a gmt file</p>
<h6 id="cell2cell.datasets.gsea_data.gsea_msig--parameters">Parameters</h6>
<p>organism : str, default='human'
    Organism for whom the DB will be loaded.
    Available options are {'human', 'mouse'}.</p>
<div class="admonition pathwaydb">
<p class="admonition-title">str, default='GOBP'</p>
<p>Molecular Signature Database to load.
Available options are {'GOBP', 'KEGG', 'Reactome'}</p>
</div>
<p>readable_name : boolean, default=False
    If True, the pathway names are transformed to a more readable format.
    That is, removing underscores and pathway DB name at the beginning.</p>
<h6 id="cell2cell.datasets.gsea_data.gsea_msig--returns">Returns</h6>
<p>pathway_per_gene : defaultdict
    Dictionary containing all genes in the DB as keys, and
    their values are lists with their pathway annotations.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/gsea_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">gsea_msig</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="o">=</span><span class="s1">&#39;GOBP&#39;</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Load a MSigDB from a gmt file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    organism : str, default=&#39;human&#39;</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Organism for whom the DB will be loaded.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Available options are {&#39;human&#39;, &#39;mouse&#39;}.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    pathwaydb: str, default=&#39;GOBP&#39;</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Molecular Signature Database to load.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Available options are {&#39;GOBP&#39;, &#39;KEGG&#39;, &#39;Reactome&#39;}</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    readable_name : boolean, default=False</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        If True, the pathway names are transformed to a more readable format.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        That is, removing underscores and pathway DB name at the beginning.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    pathway_per_gene : defaultdict</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Dictionary containing all genes in the DB as keys, and</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        their values are lists with their pathway annotations.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">_check_pathwaydb</span><span class="p">(</span><span class="n">organism</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">pathway_per_gene</span> <span class="o">=</span> <span class="n">load_gmt</span><span class="p">(</span><span class="n">readable_name</span><span class="o">=</span><span class="n">readable_name</span><span class="p">,</span> <span class="o">**</span><span class="n">PATHWAY_DATA</span><span class="p">[</span><span class="n">organism</span><span class="p">][</span><span class="n">pathwaydb</span><span class="p">])</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">return</span> <span class="n">pathway_per_gene</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.datasets.heuristic_data">
        <code>heuristic_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.datasets.heuristic_data.HeuristicGOTerms">
        <code>
HeuristicGOTerms        </code>



</h4>

    <div class="doc doc-contents ">

      <p>GO terms for contact and secreted proteins.</p>
<h6 id="cell2cell.datasets.heuristic_data.HeuristicGOTerms--attributes">Attributes</h6>
<p>contact_go_terms : list
    List of GO terms associated with proteins that
    participate in contact interactions (usually
    on the surface of cells).</p>
<p>mediator_go_terms : list
    List of GO terms associated with secreted
    proteins that mediate intercellular interactions
    or communication.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/heuristic_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">HeuristicGOTerms</span><span class="p">:</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;GO terms for contact and secreted proteins.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    contact_go_terms : list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        List of GO terms associated with proteins that</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        participate in contact interactions (usually</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        on the surface of cells).</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    mediator_go_terms : list</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        List of GO terms associated with secreted</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        proteins that mediate intercellular interactions</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        or communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">contact_go_terms</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;GO:0007155&#39;</span><span class="p">,</span>  <span class="c1"># Cell adhesion</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>                                 <span class="s1">&#39;GO:0022608&#39;</span><span class="p">,</span>  <span class="c1"># Multicellular organism adhesion</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>                                 <span class="s1">&#39;GO:0098740&#39;</span><span class="p">,</span>  <span class="c1"># Multiorganism cell adhesion</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>                                 <span class="s1">&#39;GO:0098743&#39;</span><span class="p">,</span>  <span class="c1"># Cell aggregation</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>                                 <span class="s1">&#39;GO:0030054&#39;</span><span class="p">,</span>  <span class="c1"># Cell-junction #</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>                                 <span class="s1">&#39;GO:0009986&#39;</span><span class="p">,</span>  <span class="c1"># Cell surface #</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>                                 <span class="s1">&#39;GO:0097610&#39;</span><span class="p">,</span>  <span class="c1"># Cell surface forrow</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>                                 <span class="s1">&#39;GO:0007160&#39;</span><span class="p">,</span>  <span class="c1"># Cell-matrix adhesion</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>                                 <span class="s1">&#39;GO:0043235&#39;</span><span class="p">,</span>  <span class="c1"># Receptor complex,</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>                                 <span class="s1">&#39;GO:0008305&#39;</span><span class="p">,</span>  <span class="c1"># Integrin complex,</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>                                 <span class="s1">&#39;GO:0043113&#39;</span><span class="p">,</span>  <span class="c1"># Receptor clustering</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>                                 <span class="s1">&#39;GO:0009897&#39;</span><span class="p">,</span>  <span class="c1"># External side of plasma membrane #</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>                                 <span class="s1">&#39;GO:0038023&#39;</span><span class="p">,</span>  <span class="c1"># Signaling receptor activity #</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>                                 <span class="p">]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">mediator_go_terms</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;GO:0005615&#39;</span><span class="p">,</span>  <span class="c1"># Extracellular space</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>                                  <span class="s1">&#39;GO:0005576&#39;</span><span class="p">,</span>  <span class="c1"># Extracellular region</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>                                  <span class="s1">&#39;GO:0031012&#39;</span><span class="p">,</span>  <span class="c1"># Extracellular matrix</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>                                  <span class="s1">&#39;GO:0005201&#39;</span><span class="p">,</span>  <span class="c1"># Extracellular matrix structural constituent</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>                                  <span class="s1">&#39;GO:1990430&#39;</span><span class="p">,</span>  <span class="c1"># Extracellular matrix protein binding</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>                                  <span class="s1">&#39;GO:0048018&#39;</span><span class="p">,</span>  <span class="c1"># Receptor ligand activity #</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>                                  <span class="p">]</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">












  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.datasets.random_data">
        <code>random_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.random_data.generate_random_cci_scores">
<code class="highlight language-python"><span class="n">generate_random_cci_scores</span><span class="p">(</span><span class="n">cell_number</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">symmetric</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a square cell-cell interaction
matrix with random scores.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_cci_scores--parameters">Parameters</h6>
<p>cell_number : int
    Number of cells.</p>
<p>labels : list, default=None
    List containing labels for each cells. Length of
    this list must match the cell_number.</p>
<p>symmetric : boolean, default=True
    Whether generating a symmetric CCI matrix.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_cci_scores--returns">Returns</h6>
<p>cci_matrix : pandas.DataFrame
    Matrix with rows and columns as cells. Values
    represent a random CCI score between 0 and 1.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/random_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_random_cci_scores</span><span class="p">(</span><span class="n">cell_number</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">symmetric</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a square cell-cell interaction</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    matrix with random scores.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    cell_number : int</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Number of cells.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    labels : list, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List containing labels for each cells. Length of</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        this list must match the cell_number.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    symmetric : boolean, default=True</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Whether generating a symmetric CCI matrix.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    cci_matrix : pandas.DataFrame</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Matrix with rows and columns as cells. Values</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        represent a random CCI score between 0 and 1.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">cell_number</span><span class="p">,</span> <span class="s2">&quot;Lenght of labels must match cell_number&quot;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cell-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cell_number</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">cci_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">((</span><span class="n">cell_number</span><span class="p">,</span> <span class="n">cell_number</span><span class="p">))</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">if</span> <span class="n">symmetric</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">cci_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">cci_scores</span> <span class="o">+</span> <span class="n">cci_scores</span><span class="o">.</span><span class="n">T</span><span class="p">)</span> <span class="o">/</span> <span class="mf">2.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">cci_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cci_scores</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">labels</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">return</span> <span class="n">cci_matrix</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.random_data.generate_random_metadata">
<code class="highlight language-python"><span class="n">generate_random_metadata</span><span class="p">(</span><span class="n">cell_labels</span><span class="p">,</span> <span class="n">group_number</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly assigns groups to cell labels.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_metadata--parameters">Parameters</h6>
<p>cell_labels : list
    A list of cell labels.</p>
<p>group_number : int
    Number of major groups of cells.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_metadata--returns">Returns</h6>
<p>metadata : pandas.DataFrame
    DataFrame containing the major groups that each cell
    received randomly (under column 'Group'). Cells are
    under the column 'Cell'.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/random_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_random_metadata</span><span class="p">(</span><span class="n">cell_labels</span><span class="p">,</span> <span class="n">group_number</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Randomly assigns groups to cell labels.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cell_labels : list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A list of cell labels.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    group_number : int</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Number of major groups of cells.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    metadata : pandas.DataFrame</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        DataFrame containing the major groups that each cell</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        received randomly (under column &#39;Group&#39;). Cells are</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        under the column &#39;Cell&#39;.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Cell&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_labels</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">groups</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">group_number</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;Cell&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">return</span> <span class="n">metadata</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.random_data.generate_random_ppi">
<code class="highlight language-python"><span class="n">generate_random_ppi</span><span class="p">(</span><span class="n">max_size</span><span class="p">,</span> <span class="n">interactors_A</span><span class="p">,</span> <span class="n">interactors_B</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a random list of protein-protein interactions.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_ppi--parameters">Parameters</h6>
<p>max_size : int
    Maximum size of interactions to obtain. Since the PPIs
    are obtained by independently resampling interactors A and B
    rather than creating all possible combinations (it may demand too much
    memory), some PPIs can be duplicated and when dropping them
    results into a smaller number of PPIs than the max_size.</p>
<p>interactors_A : list
    A list of protein names to include in the first column of
    the PPIs.</p>
<p>interactors_B : list, default=None
    A list of protein names to include in the second columns
    of the PPIs. If None, interactors_A will be used as
    interactors_B too.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_ppi--returns">Returns</h6>
<p>ppi_data : pandas.DataFrame
    DataFrame containing a list of protein-protein interactions.
    It has three columns: 'A', 'B', and 'score' for interactors
    A, B and weights of interactions, respectively.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/random_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_random_ppi</span><span class="p">(</span><span class="n">max_size</span><span class="p">,</span> <span class="n">interactors_A</span><span class="p">,</span> <span class="n">interactors_B</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a random list of protein-protein interactions.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    max_size : int</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Maximum size of interactions to obtain. Since the PPIs</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        are obtained by independently resampling interactors A and B</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        rather than creating all possible combinations (it may demand too much</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        memory), some PPIs can be duplicated and when dropping them</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        results into a smaller number of PPIs than the max_size.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    interactors_A : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A list of protein names to include in the first column of</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        the PPIs.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    interactors_B : list, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        A list of protein names to include in the second columns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        of the PPIs. If None, interactors_A will be used as</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        interactors_B too.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        DataFrame containing a list of protein-protein interactions.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        It has three columns: &#39;A&#39;, &#39;B&#39;, and &#39;score&#39; for interactors</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        A, B and weights of interactions, respectively.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">interactors_B</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="k">assert</span> <span class="n">max_size</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interactors_A</span><span class="p">)</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">interactors_B</span><span class="p">),</span> <span class="s2">&quot;The maximum size can&#39;t be greater than all combinations between partners A and B&quot;</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="k">assert</span> <span class="n">max_size</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interactors_A</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;The maximum size can&#39;t be greater than all combinations of partners A&quot;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Generating random PPI network.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">def</span> <span class="nf">small_block_ppi</span><span class="p">(</span><span class="n">size</span><span class="p">,</span> <span class="n">interactors_A</span><span class="p">,</span> <span class="n">interactors_B</span><span class="p">,</span> <span class="n">random_state</span><span class="p">):</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="n">random_state</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="k">if</span> <span class="n">interactors_B</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>            <span class="n">interactors_B</span> <span class="o">=</span> <span class="n">interactors_A</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">col_A</span> <span class="o">=</span> <span class="n">resample</span><span class="p">(</span><span class="n">interactors_A</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">col_B</span> <span class="o">=</span> <span class="n">resample</span><span class="p">(</span><span class="n">interactors_B</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">col_A</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">ppi_data</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">col_B</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">ppi_data</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="n">ppi_data</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="k">return</span> <span class="n">ppi_data</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">small_block_ppi</span><span class="p">(</span><span class="n">max_size</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="n">interactors_A</span><span class="p">,</span> <span class="n">interactors_B</span><span class="p">,</span> <span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="c1"># TODO: This part need to be fixed, it does not converge to the max_size -&gt; len((set(A)) * len(set(B) - set(A)))</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="c1"># while ppi_data.shape[0] &lt; size:</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="c1">#     if random_state is not None:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="c1">#         random_state += 2</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="c1">#     b = small_block_ppi(size, interactors_A, interactors_B, random_state)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="c1">#     print(b)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="c1">#     ppi_data = pd.concat([ppi_data, b])</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="c1">#     ppi_data = ppi.remove_ppi_bidirectionality(ppi_data, (&#39;A&#39;, &#39;B&#39;), verbose=verbose)</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="c1">#     ppi_data = ppi_data.drop_duplicates()</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="c1">#     ppi_data.dropna()</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="c1">#     ppi_data.reset_index(inplace=True, drop=True)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="c1">#     print(ppi_data.shape[0])</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="k">if</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_size</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">max_size</span><span class="p">)),</span> <span class="p">:]</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="n">ppi_data</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="k">return</span> <span class="n">ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.random_data.generate_random_rnaseq">
<code class="highlight language-python"><span class="n">generate_random_rnaseq</span><span class="p">(</span><span class="n">size</span><span class="p">,</span> <span class="n">row_names</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a RNA-seq dataset that is normally distributed gene-wise and size
normalized (each column sums up to a million).</p>
<h6 id="cell2cell.datasets.random_data.generate_random_rnaseq--parameters">Parameters</h6>
<p>size : int
    Number of cell-types/tissues/samples (columns).</p>
<p>row_names : array-like
    List containing the name of genes (rows).</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.datasets.random_data.generate_random_rnaseq--returns">Returns</h6>
<p>df : pandas.DataFrame
    Dataframe containing gene expression given the list
    of genes for each cell-type/tissue/sample.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/random_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_random_rnaseq</span><span class="p">(</span><span class="n">size</span><span class="p">,</span> <span class="n">row_names</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Generates a RNA-seq dataset that is normally distributed gene-wise and size</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    normalized (each column sums up to a million).</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    size : int</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Number of cell-types/tissues/samples (columns).</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    row_names : array-like</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        List containing the name of genes (rows).</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Dataframe containing gene expression given the list</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        of genes for each cell-type/tissue/sample.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Generating random RNA-seq dataset.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cell-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">row_names</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">))</span>    <span class="c1"># Normal distribution</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="nb">min</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">amin</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="nb">min</span> <span class="o">=</span> <span class="nb">min</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="nb">min</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span> <span class="o">+</span> <span class="nb">min</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">row_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Normalizing random RNA-seq dataset (into TPM)&#39;</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">scale_expression_by_sum</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_value</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">return</span> <span class="n">df</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.datasets.toy_data">
        <code>toy_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.toy_data.generate_toy_distance">
<code class="highlight language-python"><span class="n">generate_toy_distance</span><span class="p">()</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a square matrix with cell-cell distance.</p>
<h6 id="cell2cell.datasets.toy_data.generate_toy_distance--returns">Returns</h6>
<p>distance : pandas.DataFrame
    DataFrame with Euclidean-like distance between each
    pair of cells in the toy RNA-seq dataset.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/toy_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_toy_distance</span><span class="p">():</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a square matrix with cell-cell distance.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    -------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    distance : pandas.DataFrame</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        DataFrame with Euclidean-like distance between each</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        pair of cells in the toy RNA-seq dataset.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">,</span> <span class="mf">12.0</span><span class="p">,</span> <span class="mf">5.0</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">],</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>                       <span class="p">[</span><span class="mf">10.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">15.0</span><span class="p">,</span> <span class="mf">8.0</span><span class="p">,</span> <span class="mf">9.0</span><span class="p">],</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>                       <span class="p">[</span><span class="mf">12.0</span><span class="p">,</span> <span class="mf">15.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">],</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>                       <span class="p">[</span><span class="mf">5.0</span><span class="p">,</span> <span class="mf">8.0</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">6.5</span><span class="p">],</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>                       <span class="p">[</span><span class="mf">3.0</span><span class="p">,</span> <span class="mf">9.0</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">,</span> <span class="mf">6.5</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>                       <span class="p">])</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="n">distance</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>                            <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="s1">&#39;C2&#39;</span><span class="p">,</span> <span class="s1">&#39;C3&#39;</span><span class="p">,</span> <span class="s1">&#39;C4&#39;</span><span class="p">,</span> <span class="s1">&#39;C5&#39;</span><span class="p">],</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>                            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="s1">&#39;C2&#39;</span><span class="p">,</span> <span class="s1">&#39;C3&#39;</span><span class="p">,</span> <span class="s1">&#39;C4&#39;</span><span class="p">,</span> <span class="s1">&#39;C5&#39;</span><span class="p">]</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>                            <span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">return</span> <span class="n">distance</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.toy_data.generate_toy_metadata">
<code class="highlight language-python"><span class="n">generate_toy_metadata</span><span class="p">()</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates metadata for cells in the toy RNA-seq dataset.</p>
<h6 id="cell2cell.datasets.toy_data.generate_toy_metadata--returns">Returns</h6>
<p>metadata : pandas.DataFrame
    DataFrame with metadata for each cell. Metadata contains the
    major groups of those cells.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/toy_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_toy_metadata</span><span class="p">():</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates metadata for cells in the toy RNA-seq dataset.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    -------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    metadata : pandas.DataFrame</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        DataFrame with metadata for each cell. Metadata contains the</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        major groups of those cells.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([[</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="s1">&#39;G1&#39;</span><span class="p">],</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>                       <span class="p">[</span><span class="s1">&#39;C2&#39;</span><span class="p">,</span> <span class="s1">&#39;G2&#39;</span><span class="p">],</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>                       <span class="p">[</span><span class="s1">&#39;C3&#39;</span><span class="p">,</span> <span class="s1">&#39;G3&#39;</span><span class="p">],</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>                       <span class="p">[</span><span class="s1">&#39;C4&#39;</span><span class="p">,</span> <span class="s1">&#39;G3&#39;</span><span class="p">],</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>                       <span class="p">[</span><span class="s1">&#39;C5&#39;</span><span class="p">,</span> <span class="s1">&#39;G1&#39;</span><span class="p">]</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>                       <span class="p">])</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="s1">&#39;Groups&#39;</span><span class="p">])</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">return</span> <span class="n">metadata</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.toy_data.generate_toy_ppi">
<code class="highlight language-python"><span class="n">generate_toy_ppi</span><span class="p">(</span><span class="n">prot_complex</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a toy list of protein-protein interactions.</p>
<h6 id="cell2cell.datasets.toy_data.generate_toy_ppi--parameters">Parameters</h6>
<p>prot_complex : boolean, default=False
    Whether including PPIs where interactors could contain
    multimeric complexes.</p>
<h6 id="cell2cell.datasets.toy_data.generate_toy_ppi--returns">Returns</h6>
<p>ppi : pandas.DataFrame
    Dataframe containing PPIs. Columns are 'A' (first interacting
    partners), 'B' (second interacting partners) and 'score'
    for weighting each PPI.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/toy_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_toy_ppi</span><span class="p">(</span><span class="n">prot_complex</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a toy list of protein-protein interactions.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    prot_complex : boolean, default=False</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Whether including PPIs where interactors could contain</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        multimeric complexes.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    -------</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    ppi : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Dataframe containing PPIs. Columns are &#39;A&#39; (first interacting</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        partners), &#39;B&#39; (second interacting partners) and &#39;score&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        for weighting each PPI.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="k">if</span> <span class="n">prot_complex</span><span class="p">:</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([[</span><span class="s1">&#39;Protein-A&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-B&#39;</span><span class="p">],</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-C&#39;</span><span class="p">],</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-C&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-A&#39;</span><span class="p">],</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-B&#39;</span><span class="p">],</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-A&#39;</span><span class="p">],</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-E&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-F&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-C&amp;Protein-E&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-E&#39;</span><span class="p">],</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-A&amp;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>                           <span class="p">])</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([[</span><span class="s1">&#39;Protein-A&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-B&#39;</span><span class="p">],</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-C&#39;</span><span class="p">],</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-C&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-A&#39;</span><span class="p">],</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-B&#39;</span><span class="p">],</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-A&#39;</span><span class="p">],</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-E&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-F&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-C&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-E&#39;</span><span class="p">],</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>                           <span class="p">[</span><span class="s1">&#39;Protein-A&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>                           <span class="p">])</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">ppi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">])</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">ppi</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">return</span> <span class="n">ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.datasets.toy_data.generate_toy_rnaseq">
<code class="highlight language-python"><span class="n">generate_toy_rnaseq</span><span class="p">()</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a toy RNA-seq dataset</p>
<h6 id="cell2cell.datasets.toy_data.generate_toy_rnaseq--returns">Returns</h6>
<p>rnaseq : pandas.DataFrame
    DataFrame contianing the toy RNA-seq dataset. Columns
    are cells and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/datasets/toy_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_toy_rnaseq</span><span class="p">():</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a toy RNA-seq dataset</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    -------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    rnaseq : pandas.DataFrame</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        DataFrame contianing the toy RNA-seq dataset. Columns</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        are cells and rows are genes.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>                       <span class="p">[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">30</span><span class="p">],</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>                       <span class="p">[</span><span class="mi">18</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>                       <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">22</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>                       <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">27</span><span class="p">,</span> <span class="mi">15</span><span class="p">],</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>                       <span class="p">[</span><span class="mi">30</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>                       <span class="p">])</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">rnaseq</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>                          <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Protein-A&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-B&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-C&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-D&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-E&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein-F&#39;</span><span class="p">],</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>                          <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="s1">&#39;C2&#39;</span><span class="p">,</span> <span class="s1">&#39;C3&#39;</span><span class="p">,</span> <span class="s1">&#39;C4&#39;</span><span class="p">,</span> <span class="s1">&#39;C5&#39;</span><span class="p">]</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>                          <span class="p">)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">rnaseq</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;gene_id&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">return</span> <span class="n">rnaseq</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.external">
        <code>external</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.external.goenrich">
        <code>goenrich</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">











  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.goenrich.gene2go">
<code class="highlight language-python"><span class="n">gene2go</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">tax_id</span><span class="o">=</span><span class="mi">9606</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>read go-annotation file</p>
<p>:param filename: protein or gene identifier column
:param experimental: use only experimentally validated annotations
:param tax_id: filter according to taxon</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/goenrich.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">gene2go</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">tax_id</span><span class="o">=</span><span class="mi">9606</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot; read go-annotation file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    :param filename: protein or gene identifier column</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    :param experimental: use only experimentally validated annotations</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    :param tax_id: filter according to taxon</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>    <span class="n">defaults</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;comment&#39;</span><span class="p">:</span> <span class="s1">&#39;#&#39;</span><span class="p">,</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>                <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="n">GENE2GO_COLUMNS</span><span class="p">}</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="n">defaults</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds</span><span class="p">)</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>    <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">defaults</span><span class="p">)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="n">retain_mask</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">tax_id</span> <span class="o">==</span> <span class="n">tax_id</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>    <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="o">~</span><span class="n">retain_mask</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">if</span> <span class="n">experimental</span><span class="p">:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="n">retain_mask</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">Evidence</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">EXPERIMENTAL_EVIDENCE</span><span class="p">)</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="o">~</span><span class="n">retain_mask</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">return</span> <span class="n">result</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.goenrich.goa">
<code class="highlight language-python"><span class="n">goa</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>read go-annotation file</p>
<p>:param filename: protein or gene identifier column
:param experimental: use only experimentally validated annotations</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/goenrich.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">goa</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot; read go-annotation file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    :param filename: protein or gene identifier column</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    :param experimental: use only experimentally validated annotations</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>    <span class="n">defaults</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;comment&#39;</span><span class="p">:</span> <span class="s1">&#39;!&#39;</span><span class="p">,</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>                <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="n">GENE_ASSOCIATION_COLUMNS</span><span class="p">}</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">if</span> <span class="n">experimental</span> <span class="ow">and</span> <span class="s1">&#39;usecols&#39;</span> <span class="ow">in</span> <span class="n">kwds</span><span class="p">:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="n">kwds</span><span class="p">[</span><span class="s1">&#39;usecols&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span><span class="s1">&#39;evidence_code&#39;</span><span class="p">,)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="n">defaults</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds</span><span class="p">)</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>    <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">defaults</span><span class="p">)</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">if</span> <span class="n">experimental</span><span class="p">:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="n">retain_mask</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">evidence_code</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">EXPERIMENTAL_EVIDENCE</span><span class="p">)</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="o">~</span><span class="n">retain_mask</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">return</span> <span class="n">result</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.goenrich.ontology">
<code class="highlight language-python"><span class="n">ontology</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>read ontology from file
:param file: file path of file handle</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/goenrich.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">ontology</span><span class="p">(</span><span class="n">file</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot; read ontology from file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    :param file: file path of file handle</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="n">O</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">()</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>        <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>        <span class="n">we_opened_file</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="n">f</span> <span class="o">=</span> <span class="n">file</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>        <span class="n">we_opened_file</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>    <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>        <span class="n">tokens</span> <span class="o">=</span> <span class="n">_tokenize</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>        <span class="n">terms</span> <span class="o">=</span> <span class="n">_filter_terms</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="n">entries</span> <span class="o">=</span> <span class="n">_parse_terms</span><span class="p">(</span><span class="n">terms</span><span class="p">)</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">nodes</span><span class="p">,</span> <span class="n">edges</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">entries</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="n">O</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">nodes</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>        <span class="n">O</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">edges</span><span class="p">))</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="n">O</span><span class="o">.</span><span class="n">graph</span><span class="p">[</span><span class="s1">&#39;roots&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="p">:</span> <span class="n">n</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">O</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>                <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;namespace&#39;</span><span class="p">]}</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">finally</span><span class="p">:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="k">if</span> <span class="n">we_opened_file</span><span class="p">:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>            <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">for</span> <span class="n">root</span> <span class="ow">in</span> <span class="n">O</span><span class="o">.</span><span class="n">graph</span><span class="p">[</span><span class="s1">&#39;roots&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">depth</span> <span class="ow">in</span> <span class="n">nx</span><span class="o">.</span><span class="n">shortest_path_length</span><span class="p">(</span><span class="n">O</span><span class="p">,</span> <span class="n">root</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>            <span class="n">node</span> <span class="o">=</span> <span class="n">O</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">n</span><span class="p">]</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>            <span class="n">node</span><span class="p">[</span><span class="s1">&#39;depth&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">depth</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;depth&#39;</span><span class="p">,</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;inf&#39;</span><span class="p">)))</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">return</span> <span class="n">O</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.goenrich.sgd">
<code class="highlight language-python"><span class="n">sgd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>read yeast genome database go-annotation file
:param filename: protein or gene identifier column
:param experimental: use only experimentally validated annotations</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/goenrich.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">sgd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot; read yeast genome database go-annotation file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    :param filename: protein or gene identifier column</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    :param experimental: use only experimentally validated annotations</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="k">return</span> <span class="n">goa</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">experimental</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.external.gseapy">
        <code>gseapy</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">









  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.gseapy.generate_lr_geneset">
<code class="highlight language-python"><span class="n">generate_lr_geneset</span><span class="p">(</span><span class="n">lr_list</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lr_sep</span><span class="o">=</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="n">pathway_per_gene</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="o">=</span><span class="s1">&#39;GOBP&#39;</span><span class="p">,</span> <span class="n">min_pathways</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">max_pathways</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">output_folder</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generate a gene set from a list of LR pairs.</p>
<h6 id="cell2cell.external.gseapy.generate_lr_geneset--parameters">Parameters</h6>
<p>lr_list : list
    List of LR pairs.</p>
<p>complex_sep : str, default=None
    Separator of the members of a complex. If None, the ligand and receptor are assumed to be single genes.</p>
<p>lr_sep : str, default='^'
    Separator of the ligand and receptor in the LR pair.</p>
<p>pathway_per_gene : dict, default=None
    Dictionary with genes as keys and pathways as values.
    You can pass this if you are using different annotations than those
    available resources in <code>cell2cell.datasets.gsea_data.gsea_msig()</code>.</p>
<p>organism : str, default='human'
    Organism for whom the DB will be loaded.
    Available options are {'human', 'mouse'}.</p>
<div class="admonition pathwaydb">
<p class="admonition-title">str, default='GOBP'</p>
<p>Molecular Signature Database to load.
Available options are {'GOBP', 'KEGG', 'Reactome'}</p>
</div>
<p>min_pathways : int, default=15
    Minimum number of pathways that a LR pair can be annotated to.</p>
<p>max_pathways : int, default=10000
    Maximum number of pathways that a LR pair can be annotated to.</p>
<p>readable_name : boolean, default=False
    If True, the pathway names are transformed to a more readable format.</p>
<p>output_folder : str, default=None
    Path to store the GMT file. If None, it stores the gmt file in the
    current directory.</p>
<h6 id="cell2cell.external.gseapy.generate_lr_geneset--returns">Returns</h6>
<p>lr_set : dict
    Dictionary with pathways as keys and LR pairs as values.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/gseapy.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_lr_geneset</span><span class="p">(</span><span class="n">lr_list</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lr_sep</span><span class="o">=</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="n">pathway_per_gene</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="o">=</span><span class="s1">&#39;GOBP&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                        <span class="n">min_pathways</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">max_pathways</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">output_folder</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Generate a gene set from a list of LR pairs.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    lr_list : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of LR pairs.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Separator of the members of a complex. If None, the ligand and receptor are assumed to be single genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    lr_sep : str, default=&#39;^&#39;</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Separator of the ligand and receptor in the LR pair.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    pathway_per_gene : dict, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Dictionary with genes as keys and pathways as values.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        You can pass this if you are using different annotations than those</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        available resources in `cell2cell.datasets.gsea_data.gsea_msig()`.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    organism : str, default=&#39;human&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Organism for whom the DB will be loaded.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Available options are {&#39;human&#39;, &#39;mouse&#39;}.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    pathwaydb: str, default=&#39;GOBP&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Molecular Signature Database to load.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Available options are {&#39;GOBP&#39;, &#39;KEGG&#39;, &#39;Reactome&#39;}</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    min_pathways : int, default=15</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Minimum number of pathways that a LR pair can be annotated to.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    max_pathways : int, default=10000</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Maximum number of pathways that a LR pair can be annotated to.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    readable_name : boolean, default=False</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        If True, the pathway names are transformed to a more readable format.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    output_folder : str, default=None</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Path to store the GMT file. If None, it stores the gmt file in the</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        current directory.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    -------</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    lr_set : dict</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Dictionary with pathways as keys and LR pairs as values.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="c1"># Check if the LR gene set is already available</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">_check_pathwaydb</span><span class="p">(</span><span class="n">organism</span><span class="p">,</span> <span class="n">pathwaydb</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="c1"># Obtain annotations</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">gmt_info</span> <span class="o">=</span> <span class="n">PATHWAY_DATA</span><span class="p">[</span><span class="n">organism</span><span class="p">][</span><span class="n">pathwaydb</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">if</span> <span class="n">output_folder</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">gmt_info</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="n">gmt_info</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">])</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">if</span> <span class="n">pathway_per_gene</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">pathway_per_gene</span> <span class="o">=</span> <span class="n">load_gmt</span><span class="p">(</span><span class="n">readable_name</span><span class="o">=</span><span class="n">readable_name</span><span class="p">,</span> <span class="o">**</span><span class="n">gmt_info</span><span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="c1"># Dictionary to save the LR interaction (key) and the annotated pathways (values).</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">pathway_sets</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="c1"># Iterate through the interactions in the LR DB.</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">for</span> <span class="n">lr_label</span> <span class="ow">in</span> <span class="n">lr_list</span><span class="p">:</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">lr</span> <span class="o">=</span> <span class="n">lr_label</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">lr_sep</span><span class="p">)</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="c1"># Gene members of the ligand and the receptor in the LR pair</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>            <span class="n">ligands</span> <span class="o">=</span> <span class="p">[</span><span class="n">lr</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="n">receptors</span> <span class="o">=</span> <span class="p">[</span><span class="n">lr</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>            <span class="n">ligands</span> <span class="o">=</span> <span class="n">lr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">complex_sep</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>            <span class="n">receptors</span> <span class="o">=</span> <span class="n">lr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">complex_sep</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="c1"># Find pathways associated with all members of the ligand</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ligand</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ligands</span><span class="p">):</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                <span class="n">ligand_pathways</span> <span class="o">=</span> <span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">ligand</span><span class="p">]</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                <span class="n">ligand_pathways</span> <span class="o">=</span> <span class="n">ligand_pathways</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">ligand</span><span class="p">])</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="c1"># Find pathways associated with all members of the receptor</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">receptor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">receptors</span><span class="p">):</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                <span class="n">receptor_pathways</span> <span class="o">=</span> <span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">receptor</span><span class="p">]</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>                <span class="n">receptor_pathways</span> <span class="o">=</span> <span class="n">receptor_pathways</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">receptor</span><span class="p">])</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="c1"># Keep only pathways that are in both ligand and receptor.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="n">lr_pathways</span> <span class="o">=</span> <span class="n">ligand_pathways</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">receptor_pathways</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">lr_pathways</span><span class="p">:</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>            <span class="n">pathway_sets</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="n">pathway_sets</span><span class="p">[</span><span class="n">p</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">([</span><span class="n">lr_label</span><span class="p">])</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="n">lr_set</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">pathway_sets</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="k">if</span> <span class="n">min_pathways</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">max_pathways</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>            <span class="n">lr_set</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>    <span class="k">return</span> <span class="n">lr_set</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.gseapy.load_gmt">
<code class="highlight language-python"><span class="n">load_gmt</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">backup_url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Load a GMT file.</p>
<h6 id="cell2cell.external.gseapy.load_gmt--parameters">Parameters</h6>
<p>filename : str
    Path to the GMT file.</p>
<p>backup_url : str, default=None
    URL to download the GMT file from if not present locally.</p>
<p>readable_name : boolean, default=False
    If True, the pathway names are transformed to a more readable format.
    That is, removing underscores and pathway DB name at the beginning.</p>
<h6 id="cell2cell.external.gseapy.load_gmt--returns">Returns</h6>
<p>pathway_per_gene : dict
    Dictionary with genes as keys and pathways as values.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/gseapy.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_gmt</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">backup_url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">readable_name</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Load a GMT file.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Path to the GMT file.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    backup_url : str, default=None</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        URL to download the GMT file from if not present locally.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    readable_name : boolean, default=False</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        If True, the pathway names are transformed to a more readable format.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        That is, removing underscores and pathway DB name at the beginning.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    pathway_per_gene : dict</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Dictionary with genes as keys and pathways as values.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">if</span> <span class="n">path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="k">if</span> <span class="n">backup_url</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>            <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>                <span class="n">_download</span><span class="p">(</span><span class="n">backup_url</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>  <span class="c1"># invalid URL</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Invalid filename or URL&#39;</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>            <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Invalid filename&#39;</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">pathway_per_gene</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">with</span> <span class="n">f</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>            <span class="n">l</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="n">l</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">l</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>            <span class="n">l</span> <span class="o">=</span> <span class="p">[</span><span class="n">pw</span> <span class="k">for</span> <span class="n">pw</span> <span class="ow">in</span> <span class="n">l</span> <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;http&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pw</span><span class="p">)]</span>  <span class="c1"># Remove website info</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>            <span class="n">pathway_name</span> <span class="o">=</span> <span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>            <span class="k">if</span> <span class="n">readable_name</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                <span class="n">pathway_name</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pathway_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                <span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">pathway_per_gene</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">pathway_name</span><span class="p">]))</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">return</span> <span class="n">pathway_per_gene</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.gseapy.run_gsea">
<code class="highlight language-python"><span class="n">run_gsea</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">lr_set</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">999</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">significance_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Run GSEA using the LR gene set.</p>
<h6 id="cell2cell.external.gseapy.run_gsea--parameters">Parameters</h6>
<p>loadings : pandas.DataFrame
    Dataframe with the loadings of the LR pairs for each factor.</p>
<p>lr_set : dict
    Dictionary with pathways as keys and LR pairs as values.
    LR pairs must match the indexes in the loadings dataframe.</p>
<p>output_folder : str
    Path to the output folder.</p>
<p>weight : int, default=1
    Weight to use for score underlying the GSEA (parameter p).</p>
<p>min_size : int, default=15
    Minimum number of LR pairs that a pathway must contain.</p>
<p>permutations : int, default=999
    Number of permutations to use for the GSEA. The total permutations
    will be this number plus 1 (this extra case is the unpermuted one).</p>
<p>processes : int, default=6
    Number of processes to use for the GSEA.</p>
<p>random_state : int, default=6
    Random seed to use for the GSEA.</p>
<p>significance_threshold : float, default=0.05
    Significance threshold to use for the FDR correction.</p>
<h6 id="cell2cell.external.gseapy.run_gsea--returns">Returns</h6>
<p>pvals : pandas.DataFrame
    Dataframe containing the P-values for each pathway (rows)
    in each of the factors (columns).</p>
<p>score : pandas.DataFrame
    Dataframe containing the Normalized Enrichment Scores (NES)
    for each pathway (rows) in each of the factors (columns).</p>
<p>gsea_df : pandas.DataFrame
    Dataframe with the detailed GSEA results.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/gseapy.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">run_gsea</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">lr_set</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">999</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>             <span class="n">random_state</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">significance_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Run GSEA using the LR gene set.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    loadings : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Dataframe with the loadings of the LR pairs for each factor.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    lr_set : dict</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Dictionary with pathways as keys and LR pairs as values.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        LR pairs must match the indexes in the loadings dataframe.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    output_folder : str</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Path to the output folder.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    weight : int, default=1</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Weight to use for score underlying the GSEA (parameter p).</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    min_size : int, default=15</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Minimum number of LR pairs that a pathway must contain.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    permutations : int, default=999</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Number of permutations to use for the GSEA. The total permutations</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        will be this number plus 1 (this extra case is the unpermuted one).</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    processes : int, default=6</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Number of processes to use for the GSEA.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    random_state : int, default=6</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Random seed to use for the GSEA.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    significance_threshold : float, default=0.05</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Significance threshold to use for the FDR correction.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    -------</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    pvals : pandas.DataFrame</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Dataframe containing the P-values for each pathway (rows)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        in each of the factors (columns).</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    score : pandas.DataFrame</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Dataframe containing the Normalized Enrichment Scores (NES)</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        for each pathway (rows) in each of the factors (columns).</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    gsea_df : pandas.DataFrame</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Dataframe with the detailed GSEA results.</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">fdrcorrection</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="kn">from</span> <span class="nn">cell2cell.io.directories</span> <span class="kn">import</span> <span class="n">create_directory</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="n">create_directory</span><span class="p">(</span><span class="n">output_folder</span><span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">gseapy</span> <span class="o">=</span> <span class="n">_check_if_gseapy</span><span class="p">()</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">lr_set_</span> <span class="o">=</span> <span class="n">lr_set</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">loadings</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="c1"># Rank LR pairs of each factor by their respective loadings</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">test</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">factor</span><span class="p">]]</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="n">test</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="n">test</span> <span class="o">=</span> <span class="n">test</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="n">test</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="c1"># RUN GSEA</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">gseapy</span><span class="o">.</span><span class="n">prerank</span><span class="p">(</span><span class="n">rnk</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                       <span class="n">gene_sets</span><span class="o">=</span><span class="n">lr_set_</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                       <span class="n">min_size</span><span class="o">=</span><span class="n">min_size</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                       <span class="n">weighted_score_type</span><span class="o">=</span><span class="n">weight</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                       <span class="n">processes</span><span class="o">=</span><span class="n">processes</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                       <span class="n">permutation_num</span><span class="o">=</span><span class="n">permutations</span><span class="p">,</span>  <span class="c1"># reduce number to speed up testing</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                       <span class="n">outdir</span><span class="o">=</span><span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/GSEA/&#39;</span> <span class="o">+</span> <span class="n">factor</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="c1"># Adjust P-values</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="n">terms</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="n">factors</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="n">nes</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="n">p_report</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/GSEA/&#39;</span> <span class="o">+</span> <span class="n">factor</span> <span class="o">+</span> <span class="s1">&#39;/gseapy.gene_set.prerank.report.csv&#39;</span><span class="p">)</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="n">pval</span> <span class="o">=</span> <span class="n">p_report</span><span class="p">[</span><span class="s1">&#39;NOM p-val&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="n">pvals</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">pval</span><span class="p">)</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">terms</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">p_report</span><span class="o">.</span><span class="n">Term</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="n">factors</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">factor</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pval</span><span class="p">))</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="n">nes</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">p_report</span><span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>    <span class="n">gsea_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span><span class="n">factors</span><span class="p">,</span> <span class="n">terms</span><span class="p">,</span> <span class="n">nes</span><span class="p">,</span> <span class="n">pvals</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Factor&#39;</span><span class="p">,</span> <span class="s1">&#39;Term&#39;</span><span class="p">,</span> <span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;P-value&#39;</span><span class="p">])</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">gsea_df</span> <span class="o">=</span> <span class="n">gsea_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;nan&#39;</span><span class="p">]</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">])</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span> <span class="o">/</span> <span class="p">(</span><span class="n">permutations</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">])</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="c1"># Corrected P-value</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;Adj. P-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fdrcorrection</span><span class="p">(</span><span class="n">gsea_df</span><span class="p">[</span><span class="s1">&#39;P-value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                                            <span class="n">alpha</span><span class="o">=</span><span class="n">significance_threshold</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>    <span class="n">gsea_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">output_folder</span> <span class="o">+</span> <span class="s1">&#39;/GSEA/GSEA-Adj-Pvals.xlsx&#39;</span><span class="p">)</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">gsea_df</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s2">&quot;Term&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;Factor&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;Adj. P-value&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mf">1.</span><span class="p">)</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>    <span class="n">scores</span> <span class="o">=</span> <span class="n">gsea_df</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s2">&quot;Term&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;Factor&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s2">&quot;NES&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="c1"># Sort factors</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="n">sorted_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="k">if</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="n">pvals</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span> <span class="ow">in</span> <span class="n">scores</span><span class="o">.</span><span class="n">columns</span><span class="p">)]</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">pvals</span><span class="p">[</span><span class="n">sorted_columns</span><span class="p">]</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="n">scores</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">sorted_columns</span><span class="p">]</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="k">return</span> <span class="n">pvals</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">gsea_df</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.external.pcoa">
        <code>pcoa</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa.pcoa">
<code class="highlight language-python"><span class="n">pcoa</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;eigh&#39;</span><span class="p">,</span> <span class="n">number_of_dimensions</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Perform Principal Coordinate Analysis.
Principal Coordinate Analysis (PCoA) is a method similar
to Principal Components Analysis (PCA) with the difference that PCoA
operates on distance matrices, typically with non-euclidian and thus
ecologically meaningful distances like UniFrac in microbiome research.
In ecology, the euclidean distance preserved by Principal
Component Analysis (PCA) is often not a good choice because it
deals poorly with double zeros (Species have unimodal
distributions along environmental gradients, so if a species is
absent from two sites at the same site, it can't be known if an
environmental variable is too high in one of them and too low in
the other, or too low in both, etc. On the other hand, if an
species is present in two sites, that means that the sites are
similar.).
Note that the returned eigenvectors are not normalized to unit length.
Parameters</p>
<hr />
<p>distance_matrix : pandas.DataFrame
    A distance matrix.
method : str, optional
    Eigendecomposition method to use in performing PCoA.
    By default, uses SciPy's <code>eigh</code>, which computes exact
    eigenvectors and eigenvalues for all dimensions. The alternate
    method, <code>fsvd</code>, uses faster heuristic eigendecomposition but loses
    accuracy. The magnitude of accuracy lost is dependent on dataset.
number_of_dimensions : int, optional
    Dimensions to reduce the distance matrix to. This number determines
    how many eigenvectors and eigenvalues will be returned.
    By default, equal to the number of dimensions of the distance matrix,
    as default eigendecomposition using SciPy's <code>eigh</code> method computes
    all eigenvectors and eigenvalues. If using fast heuristic
    eigendecomposition through <code>fsvd</code>, a desired number of dimensions
    should be specified. Note that the default eigendecomposition
    method <code>eigh</code> does not natively support a specifying number of
    dimensions to reduce a matrix to, so if this parameter is specified,
    all eigenvectors and eigenvalues will be simply be computed with no
    speed gain, and only the number specified by <code>number_of_dimensions</code>
    will be returned. Specifying a value of <code>0</code>, the default, will
    set <code>number_of_dimensions</code> equal to the number of dimensions of the
    specified <code>distance_matrix</code>.
inplace : bool, optional
    If true, centers a distance matrix in-place in a manner that reduces
    memory consumption.
Returns</p>
<hr />
<p>OrdinationResults
    Object that stores the PCoA results, including eigenvalues, the
    proportion explained by each of them, and transformed sample
    coordinates.
See Also</p>
<hr />
<p>OrdinationResults
Notes</p>
<hr />
<p>.. note:: If the distance is not euclidean (for example if it is a
    semimetric and the triangle inequality doesn't hold),
    negative eigenvalues can appear. There are different ways
    to deal with that problem (see Legendre &amp; Legendre 1998, \S
    9.2.3), but none are currently implemented here.
    However, a warning is raised whenever negative eigenvalues
    appear, allowing the user to decide if they can be safely
    ignored.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pcoa</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;eigh&quot;</span><span class="p">,</span> <span class="n">number_of_dimensions</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>         <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Perform Principal Coordinate Analysis.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Principal Coordinate Analysis (PCoA) is a method similar</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    to Principal Components Analysis (PCA) with the difference that PCoA</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    operates on distance matrices, typically with non-euclidian and thus</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ecologically meaningful distances like UniFrac in microbiome research.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    In ecology, the euclidean distance preserved by Principal</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Component Analysis (PCA) is often not a good choice because it</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    deals poorly with double zeros (Species have unimodal</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    distributions along environmental gradients, so if a species is</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    absent from two sites at the same site, it can&#39;t be known if an</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    environmental variable is too high in one of them and too low in</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    the other, or too low in both, etc. On the other hand, if an</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    species is present in two sites, that means that the sites are</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    similar.).</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Note that the returned eigenvectors are not normalized to unit length.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    distance_matrix : pandas.DataFrame</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        A distance matrix.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    method : str, optional</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Eigendecomposition method to use in performing PCoA.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        By default, uses SciPy&#39;s `eigh`, which computes exact</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        eigenvectors and eigenvalues for all dimensions. The alternate</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        method, `fsvd`, uses faster heuristic eigendecomposition but loses</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        accuracy. The magnitude of accuracy lost is dependent on dataset.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    number_of_dimensions : int, optional</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Dimensions to reduce the distance matrix to. This number determines</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        how many eigenvectors and eigenvalues will be returned.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        By default, equal to the number of dimensions of the distance matrix,</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        as default eigendecomposition using SciPy&#39;s `eigh` method computes</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        all eigenvectors and eigenvalues. If using fast heuristic</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        eigendecomposition through `fsvd`, a desired number of dimensions</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        should be specified. Note that the default eigendecomposition</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        method `eigh` does not natively support a specifying number of</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        dimensions to reduce a matrix to, so if this parameter is specified,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        all eigenvectors and eigenvalues will be simply be computed with no</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        speed gain, and only the number specified by `number_of_dimensions`</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        will be returned. Specifying a value of `0`, the default, will</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        set `number_of_dimensions` equal to the number of dimensions of the</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        specified `distance_matrix`.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    inplace : bool, optional</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        If true, centers a distance matrix in-place in a manner that reduces</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        memory consumption.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    -------</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    OrdinationResults</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Object that stores the PCoA results, including eigenvalues, the</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        proportion explained by each of them, and transformed sample</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        coordinates.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    See Also</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    --------</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    OrdinationResults</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    Notes</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    -----</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">    .. note:: If the distance is not euclidean (for example if it is a</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        semimetric and the triangle inequality doesn&#39;t hold),</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        negative eigenvalues can appear. There are different ways</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        to deal with that problem (see Legendre &amp; Legendre 1998, \S</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        9.2.3), but none are currently implemented here.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        However, a warning is raised whenever negative eigenvalues</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        appear, allowing the user to decide if they can be safely</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">convert_to_distance_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="c1"># Center distance matrix, a requirement for PCoA here</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="n">matrix_data</span> <span class="o">=</span> <span class="n">center_distance_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="n">inplace</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="c1"># If no dimension specified, by default will compute all eigenvectors</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="c1"># and eigenvalues</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="k">if</span> <span class="n">number_of_dimensions</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;fsvd&quot;</span> <span class="ow">and</span> <span class="n">matrix_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>            <span class="n">warn</span><span class="p">(</span><span class="s2">&quot;FSVD: since no value for number_of_dimensions is specified, &quot;</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                 <span class="s2">&quot;PCoA for all dimensions will be computed, which may &quot;</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                 <span class="s2">&quot;result in long computation time if the original &quot;</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                 <span class="s2">&quot;distance matrix is large.&quot;</span><span class="p">,</span> <span class="ne">RuntimeWarning</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="c1"># distance_matrix is guaranteed to be square</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="n">number_of_dimensions</span> <span class="o">=</span> <span class="n">matrix_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="k">elif</span> <span class="n">number_of_dimensions</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Invalid operation: cannot reduce distance matrix &#39;</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>                         <span class="s1">&#39;to negative dimensions using PCoA. Did you intend &#39;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>                         <span class="s1">&#39;to specify the default value &quot;0&quot;, which sets &#39;</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>                         <span class="s1">&#39;the number_of_dimensions equal to the &#39;</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                         <span class="s1">&#39;dimensionality of the given distance matrix?&#39;</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="c1"># Perform eigendecomposition</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;eigh&quot;</span><span class="p">:</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="c1"># eigh does not natively support specifying number_of_dimensions, i.e.</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>        <span class="c1"># there are no speed gains unlike in FSVD. Later, we slice off unwanted</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>        <span class="c1"># dimensions to conform the result of eigh to the specified</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="c1"># number_of_dimensions.</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigh</span><span class="p">(</span><span class="n">matrix_data</span><span class="p">)</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="n">long_method_name</span> <span class="o">=</span> <span class="s2">&quot;Principal Coordinate Analysis&quot;</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;fsvd&quot;</span><span class="p">:</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">_fsvd</span><span class="p">(</span><span class="n">matrix_data</span><span class="p">,</span> <span class="n">number_of_dimensions</span><span class="p">)</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>        <span class="n">long_method_name</span> <span class="o">=</span> <span class="s2">&quot;Approximate Principal Coordinate Analysis &quot;</span> \
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                           <span class="s2">&quot;using FSVD&quot;</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>            <span class="s2">&quot;PCoA eigendecomposition method </span><span class="si">{}</span><span class="s2"> not supported.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">))</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="c1"># cogent makes eigenvalues positive by taking the</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="c1"># abs value, but that doesn&#39;t seem to be an approach accepted</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="c1"># by L&amp;L to deal with negative eigenvalues. We raise a warning</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="c1"># in that case. First, we make values close to 0 equal to 0.</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="n">negative_close_to_zero</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">eigvals</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="n">eigvals</span><span class="p">[</span><span class="n">negative_close_to_zero</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">eigvals</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>        <span class="n">warn</span><span class="p">(</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>            <span class="s2">&quot;The result contains negative eigenvalues.&quot;</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>            <span class="s2">&quot; Please compare their magnitude with the magnitude of some&quot;</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>            <span class="s2">&quot; of the largest positive eigenvalues. If the negative ones&quot;</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>            <span class="s2">&quot; are smaller, it&#39;s probably safe to ignore them, but if they&quot;</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>            <span class="s2">&quot; are large in magnitude, the results won&#39;t be useful. See the&quot;</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>            <span class="s2">&quot; Notes section for more details. The smallest eigenvalue is&quot;</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>            <span class="s2">&quot; </span><span class="si">{0}</span><span class="s2"> and the largest is </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">eigvals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                                                  <span class="n">eigvals</span><span class="o">.</span><span class="n">max</span><span class="p">()),</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>            <span class="ne">RuntimeWarning</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="p">)</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>    <span class="c1"># eigvals might not be ordered, so we first sort them, then analogously</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="c1"># sort the eigenvectors by the ordering of the eigenvalues too</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>    <span class="n">idxs_descending</span> <span class="o">=</span> <span class="n">eigvals</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>    <span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[</span><span class="n">idxs_descending</span><span class="p">]</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>    <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">idxs_descending</span><span class="p">]</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>    <span class="c1"># If we return only the coordinates that make sense (i.e., that have a</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>    <span class="c1"># corresponding positive eigenvalue), then Jackknifed Beta Diversity</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>    <span class="c1"># won&#39;t work as it expects all the OrdinationResults to have the same</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>    <span class="c1"># number of coordinates. In order to solve this issue, we return the</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>    <span class="c1"># coordinates that have a negative eigenvalue as 0</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>    <span class="n">num_positive</span> <span class="o">=</span> <span class="p">(</span><span class="n">eigvals</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">num_positive</span><span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">num_positive</span><span class="p">:]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>    <span class="n">eigvals</span><span class="p">[</span><span class="n">num_positive</span><span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">eigvals</span><span class="p">[</span><span class="n">num_positive</span><span class="p">:]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;fsvd&quot;</span><span class="p">:</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>        <span class="c1"># Since the dimension parameter, hereafter referred to as &#39;d&#39;,</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>        <span class="c1"># restricts the number of eigenvalues and eigenvectors that FSVD</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>        <span class="c1"># computes, we need to use an alternative method to compute the sum</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="c1"># of all eigenvalues, used to compute the array of proportions</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>        <span class="c1"># explained. Otherwise, the proportions calculated will only be</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="c1"># relative to d number of dimensions computed; whereas we want</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>        <span class="c1"># it to be relative to the entire dimensionality of the</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>        <span class="c1"># centered distance matrix.</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>        <span class="c1"># An alternative method of calculating th sum of eigenvalues is by</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="c1"># computing the trace of the centered distance matrix.</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="c1"># See proof outlined here: https://goo.gl/VAYiXx</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>        <span class="n">sum_eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">trace</span><span class="p">(</span><span class="n">matrix_data</span><span class="p">)</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>        <span class="c1"># Calculate proportions the usual way</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="n">sum_eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>    <span class="n">proportion_explained</span> <span class="o">=</span> <span class="n">eigvals</span> <span class="o">/</span> <span class="n">sum_eigenvalues</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>    <span class="c1"># In case eigh is used, eigh computes all eigenvectors and -values.</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>    <span class="c1"># So if number_of_dimensions was specified, we manually need to ensure</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>    <span class="c1"># only the requested number of dimensions</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>    <span class="c1"># (number of eigenvectors and eigenvalues, respectively) are returned.</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>    <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="p">:</span><span class="n">number_of_dimensions</span><span class="p">]</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>    <span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[:</span><span class="n">number_of_dimensions</span><span class="p">]</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>    <span class="n">proportion_explained</span> <span class="o">=</span> <span class="n">proportion_explained</span><span class="p">[:</span><span class="n">number_of_dimensions</span><span class="p">]</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>    <span class="c1"># Scale eigenvalues to have length = sqrt(eigenvalue). This</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="c1"># works because np.linalg.eigh returns normalized</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>    <span class="c1"># eigenvectors. Each row contains the coordinates of the</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>    <span class="c1"># objects in the space of principal coordinates. Note that at</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>    <span class="c1"># least one eigenvalue is zero because only n-1 axes are</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>    <span class="c1"># needed to represent n points in a euclidean space.</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>    <span class="n">coordinates</span> <span class="o">=</span> <span class="n">eigvecs</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>    <span class="n">axis_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;PC</span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">number_of_dimensions</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>    <span class="n">ordination_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;short_method_name&#39;</span> <span class="p">:</span> <span class="s2">&quot;PCoA&quot;</span><span class="p">,</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>                       <span class="s1">&#39;long_method_name&#39;</span> <span class="p">:</span> <span class="n">long_method_name</span><span class="p">,</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>                       <span class="s1">&#39;eigvals&#39;</span> <span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">eigvals</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">axis_labels</span><span class="p">),</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>                       <span class="s1">&#39;samples&#39;</span> <span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">coordinates</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">axis_labels</span><span class="p">),</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>                       <span class="s1">&#39;proportion_explained&#39;</span> <span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">proportion_explained</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">axis_labels</span><span class="p">)</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                       <span class="p">}</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>    <span class="k">return</span> <span class="n">ordination_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa.pcoa_biplot">
<code class="highlight language-python"><span class="n">pcoa_biplot</span><span class="p">(</span><span class="n">ordination</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Compute the projection of descriptors into a PCoA matrix
This implementation is as described in Chapter 9 of Legendre &amp; Legendre,
Numerical Ecology 3rd edition.
Parameters</p>
<hr />
<div class="admonition ordination">
<p class="admonition-title">OrdinationResults</p>
<p>The computed principal coordinates analysis of dimensions (n, c) where
the matrix <code>y</code> will be projected onto.</p>
</div>
<div class="admonition y">
<p class="admonition-title">DataFrame</p>
<p>Samples by features table of dimensions (n, m). These can be
environmental features or abundance counts. This table should be
normalized in cases of dimensionally heterogenous physical variables.</p>
</div>
<h6 id="cell2cell.external.pcoa.pcoa_biplot--returns">Returns</h6>
<p>OrdinationResults
    The modified input object that includes projected features onto the
    ordination space in the <code>features</code> attribute.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pcoa_biplot</span><span class="p">(</span><span class="n">ordination</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Compute the projection of descriptors into a PCoA matrix</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    This implementation is as described in Chapter 9 of Legendre &amp; Legendre,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Numerical Ecology 3rd edition.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ordination: OrdinationResults</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        The computed principal coordinates analysis of dimensions (n, c) where</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the matrix ``y`` will be projected onto.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    y: DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Samples by features table of dimensions (n, m). These can be</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        environmental features or abundance counts. This table should be</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        normalized in cases of dimensionally heterogenous physical variables.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    -------</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    OrdinationResults</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The modified input object that includes projected features onto the</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        ordination space in the ``features`` attribute.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="c1"># acknowledge that most saved ordinations lack a name, however if they have</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="c1"># a name, it should be PCoA</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;short_method_name&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span> <span class="ow">and</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>            <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;short_method_name&#39;</span><span class="p">]</span><span class="o">!=</span> <span class="s1">&#39;PCoA&#39;</span><span class="p">):</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;This biplot computation can only be performed in a &#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>                         <span class="s1">&#39;PCoA matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;The eigenvectors and the descriptors must describe &#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>                         <span class="s1">&#39;the same samples.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">eigvals</span> <span class="o">=</span> <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;eigvals&#39;</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">coordinates</span> <span class="o">=</span> <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">N</span> <span class="o">=</span> <span class="n">coordinates</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="c1"># align the descriptors and eigenvectors in a sample-wise fashion</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">y</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">coordinates</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="c1"># S_pc from equation 9.44</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="c1"># Represents the covariance matrix between the features matrix and the</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="c1"># column-centered eigenvectors of the pcoa.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">spc</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">*</span> <span class="n">y</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">scale</span><span class="p">(</span><span class="n">coordinates</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="c1"># U_proj from equation 9.55, is the matrix of descriptors to be projected.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="c1">#</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="c1"># Only get the power of non-zero values, otherwise this will raise a</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="c1"># divide by zero warning. There shouldn&#39;t be negative eigenvalues(?)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">Uproj</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">spc</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">eigvals</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">,</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>                                                      <span class="n">where</span><span class="o">=</span><span class="n">eigvals</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)))</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;features&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Uproj</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                                       <span class="n">index</span><span class="o">=</span><span class="n">y</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                                       <span class="n">columns</span><span class="o">=</span><span class="n">coordinates</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">return</span> <span class="n">ordination</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.external.pcoa_utils">
        <code>pcoa_utils</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.center_distance_matrix">
<code class="highlight language-python"><span class="n">center_distance_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Centers a distance matrix.
Note: If the used distance was euclidean, pairwise distances
needn't be computed from the data table Y because F_matrix =
Y.dot(Y.T) (if Y has been centered).
But since we're expecting distance_matrix to be non-euclidian,
we do the following computation as per
Numerical Ecology (Legendre &amp; Legendre 1998).
Parameters</p>
<hr />
<p>distance_matrix : 2D array_like
    Distance matrix.
inplace : bool, optional
    Whether or not to center the given distance matrix in-place, which
    is more efficient in terms of memory and computation.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">center_distance_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Centers a distance matrix.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Note: If the used distance was euclidean, pairwise distances</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    needn&#39;t be computed from the data table Y because F_matrix =</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Y.dot(Y.T) (if Y has been centered).</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    But since we&#39;re expecting distance_matrix to be non-euclidian,</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    we do the following computation as per</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Numerical Ecology (Legendre &amp; Legendre 1998).</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    distance_matrix : 2D array_like</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Distance matrix.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    inplace : bool, optional</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Whether or not to center the given distance matrix in-place, which</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        is more efficient in terms of memory and computation.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="k">return</span> <span class="n">_f_matrix_inplace</span><span class="p">(</span><span class="n">_e_matrix_inplace</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">))</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="k">return</span> <span class="n">f_matrix</span><span class="p">(</span><span class="n">e_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.corr">
<code class="highlight language-python"><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes correlation between columns of <code>x</code>, or <code>x</code> and <code>y</code>.
Correlation is covariance of (columnwise) standardized matrices,
so each matrix is first centered and scaled to have variance one,
and then their covariance is computed.
Parameters</p>
<hr />
<p>x : 2D array_like
    Matrix of shape (n, p). Correlation between its columns will
    be computed.
y : 2D array_like, optional
    Matrix of shape (n, q). If provided, the correlation is
    computed between the columns of <code>x</code> and the columns of
    <code>y</code>. Else, it's computed between the columns of <code>x</code>.
Returns</p>
<hr />
<p>correlation
    Matrix of computed correlations. Has shape (p, p) if <code>y</code> is
    not provided, else has shape (p, q).</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">corr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Computes correlation between columns of `x`, or `x` and `y`.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Correlation is covariance of (columnwise) standardized matrices,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    so each matrix is first centered and scaled to have variance one,</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    and then their covariance is computed.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    x : 2D array_like</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Matrix of shape (n, p). Correlation between its columns will</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        be computed.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    y : 2D array_like, optional</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Matrix of shape (n, q). If provided, the correlation is</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        computed between the columns of `x` and the columns of</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        `y`. Else, it&#39;s computed between the columns of `x`.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    -------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    correlation</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Matrix of computed correlations. Has shape (p, p) if `y` is</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        not provided, else has shape (p, q).</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">if</span> <span class="n">y</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>        <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="k">if</span> <span class="n">y</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Both matrices must have the same number of rows&quot;</span><span class="p">)</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">scale</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">x</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">y</span> <span class="o">=</span> <span class="n">x</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="c1"># Notice that scaling was performed with ddof=0 (dividing by n,</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="c1"># the default), so now we need to remove it by also using ddof=0</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="c1"># (dividing by n)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="o">/</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.e_matrix">
<code class="highlight language-python"><span class="n">e_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Compute E matrix from a distance matrix.
Squares and divides by -2 the input elementwise. Eq. 9.20 in
Legendre &amp; Legendre 1998.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">e_matrix</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Compute E matrix from a distance matrix.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Squares and divides by -2 the input elementwise. Eq. 9.20 in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Legendre &amp; Legendre 1998.&quot;&quot;&quot;</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="k">return</span> <span class="n">distance_matrix</span> <span class="o">*</span> <span class="n">distance_matrix</span> <span class="o">/</span> <span class="o">-</span><span class="mi">2</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.f_matrix">
<code class="highlight language-python"><span class="n">f_matrix</span><span class="p">(</span><span class="n">E_matrix</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Compute F matrix from E matrix.
Centring step: for each element, the mean of the corresponding
row and column are substracted, and the mean of the whole
matrix is added. Eq. 9.21 in Legendre &amp; Legendre 1998.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">f_matrix</span><span class="p">(</span><span class="n">E_matrix</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Compute F matrix from E matrix.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Centring step: for each element, the mean of the corresponding</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    row and column are substracted, and the mean of the whole</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    matrix is added. Eq. 9.21 in Legendre &amp; Legendre 1998.&quot;&quot;&quot;</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="n">row_means</span> <span class="o">=</span> <span class="n">E_matrix</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>    <span class="n">col_means</span> <span class="o">=</span> <span class="n">E_matrix</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>    <span class="n">matrix_mean</span> <span class="o">=</span> <span class="n">E_matrix</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>    <span class="k">return</span> <span class="n">E_matrix</span> <span class="o">-</span> <span class="n">row_means</span> <span class="o">-</span> <span class="n">col_means</span> <span class="o">+</span> <span class="n">matrix_mean</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.mean_and_std">
<code class="highlight language-python"><span class="n">mean_and_std</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">with_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">with_std</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Compute the weighted average and standard deviation along the
specified axis.
Parameters</p>
<hr />
<p>a : array_like
    Calculate average and standard deviation of these values.
axis : int, optional
    Axis along which the statistics are computed. The default is
    to compute them on the flattened array.
weights : array_like, optional
    An array of weights associated with the values in <code>a</code>. Each
    value in <code>a</code> contributes to the average according to its
    associated weight. The weights array can either be 1-D (in
    which case its length must be the size of <code>a</code> along the given
    axis) or of the same shape as <code>a</code>. If <code>weights=None</code>, then all
    data in <code>a</code> are assumed to have a weight equal to one.
with_mean : bool, optional, defaults to True
    Compute average if True.
with_std : bool, optional, defaults to True
    Compute standard deviation if True.
ddof : int, optional, defaults to 0
    It means delta degrees of freedom. Variance is calculated by
    dividing by <code>n - ddof</code> (where <code>n</code> is the number of
    elements). By default it computes the maximum likelyhood
    estimator.
Returns</p>
<hr />
<p>average, std
    Return the average and standard deviation along the specified
    axis. If any of them was not required, returns <code>None</code> instead</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">mean_and_std</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">with_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">with_std</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                 <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&quot;&quot;&quot;Compute the weighted average and standard deviation along the</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    specified axis.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    a : array_like</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Calculate average and standard deviation of these values.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    axis : int, optional</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Axis along which the statistics are computed. The default is</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        to compute them on the flattened array.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    weights : array_like, optional</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        An array of weights associated with the values in `a`. Each</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        value in `a` contributes to the average according to its</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        associated weight. The weights array can either be 1-D (in</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        which case its length must be the size of `a` along the given</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        axis) or of the same shape as `a`. If `weights=None`, then all</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        data in `a` are assumed to have a weight equal to one.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    with_mean : bool, optional, defaults to True</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Compute average if True.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    with_std : bool, optional, defaults to True</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Compute standard deviation if True.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    ddof : int, optional, defaults to 0</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        It means delta degrees of freedom. Variance is calculated by</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        dividing by `n - ddof` (where `n` is the number of</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        elements). By default it computes the maximum likelyhood</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        estimator.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    average, std</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Return the average and standard deviation along the specified</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        axis. If any of them was not required, returns `None` instead</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">with_mean</span> <span class="ow">or</span> <span class="n">with_std</span><span class="p">):</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Either the mean or standard deviation need to be&quot;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>                         <span class="s2">&quot; computed.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">weights</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">avg</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span> <span class="k">if</span> <span class="n">with_mean</span> <span class="k">else</span> <span class="kc">None</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="n">std</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="n">ddof</span><span class="p">)</span> <span class="k">if</span> <span class="n">with_std</span> <span class="k">else</span> <span class="kc">None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">avg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="k">if</span> <span class="n">with_std</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>            <span class="k">if</span> <span class="n">axis</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                <span class="n">variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">((</span><span class="n">a</span> <span class="o">-</span> <span class="n">avg</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                <span class="c1"># Make sure that the subtraction to compute variance works for</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>                <span class="c1"># multidimensional arrays</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>                <span class="n">a_rolled</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">rollaxis</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>                <span class="c1"># Numpy doesn&#39;t have a weighted std implementation, but this is</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>                <span class="c1"># stable and fast</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                <span class="n">variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">((</span><span class="n">a_rolled</span> <span class="o">-</span> <span class="n">avg</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                                      <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>            <span class="k">if</span> <span class="n">ddof</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Don&#39;t waste time if variance doesn&#39;t need scaling</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>                <span class="k">if</span> <span class="n">axis</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>                    <span class="n">variance</span> <span class="o">*=</span> <span class="n">a</span><span class="o">.</span><span class="n">size</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">size</span> <span class="o">-</span> <span class="n">ddof</span><span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>                    <span class="n">variance</span> <span class="o">*=</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="n">axis</span><span class="p">]</span> <span class="o">/</span> <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="n">axis</span><span class="p">]</span> <span class="o">-</span> <span class="n">ddof</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>            <span class="n">std</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">variance</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>            <span class="n">std</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">avg</span> <span class="o">=</span> <span class="n">avg</span> <span class="k">if</span> <span class="n">with_mean</span> <span class="k">else</span> <span class="kc">None</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="k">return</span> <span class="n">avg</span><span class="p">,</span> <span class="n">std</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.scale">
<code class="highlight language-python"><span class="n">scale</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">with_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">with_std</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Scale array by columns to have weighted average 0 and standard
deviation 1.
Parameters</p>
<hr />
<p>a : array_like
    2D array whose columns are standardized according to the
    weights.
weights : array_like, optional
    Array of weights associated with the columns of <code>a</code>. By
    default, the scaling is unweighted.
with_mean : bool, optional, defaults to True
    Center columns to have 0 weighted mean.
with_std : bool, optional, defaults to True
    Scale columns to have unit weighted std.
ddof : int, optional, defaults to 0
    If with_std is True, variance is calculated by dividing by <code>n
    - ddof</code> (where <code>n</code> is the number of elements). By default it
    computes the maximum likelyhood stimator.
copy : bool, optional, defaults to True
    Whether to perform the standardization in place, or return a
    new copy of <code>a</code>.
Returns</p>
<hr />
<p>2D ndarray
    Scaled array.
Notes</p>
<hr />
<p>Wherever std equals 0, it is replaced by 1 in order to avoid
division by zero.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">scale</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">with_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">with_std</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Scale array by columns to have weighted average 0 and standard</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    deviation 1.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    a : array_like</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        2D array whose columns are standardized according to the</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        weights.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    weights : array_like, optional</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Array of weights associated with the columns of `a`. By</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        default, the scaling is unweighted.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    with_mean : bool, optional, defaults to True</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Center columns to have 0 weighted mean.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    with_std : bool, optional, defaults to True</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Scale columns to have unit weighted std.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ddof : int, optional, defaults to 0</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        If with_std is True, variance is calculated by dividing by `n</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - ddof` (where `n` is the number of elements). By default it</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        computes the maximum likelyhood stimator.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    copy : bool, optional, defaults to True</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Whether to perform the standardization in place, or return a</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        new copy of `a`.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    -------</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    2D ndarray</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Scaled array.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    Notes</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    -----</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    Wherever std equals 0, it is replaced by 1 in order to avoid</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    division by zero.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">if</span> <span class="n">copy</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">a</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">avg</span><span class="p">,</span> <span class="n">std</span> <span class="o">=</span> <span class="n">mean_and_std</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span> <span class="n">with_mean</span><span class="o">=</span><span class="n">with_mean</span><span class="p">,</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>                            <span class="n">with_std</span><span class="o">=</span><span class="n">with_std</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="n">ddof</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">if</span> <span class="n">with_mean</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="n">a</span> <span class="o">-=</span> <span class="n">avg</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">if</span> <span class="n">with_std</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="n">std</span><span class="p">[</span><span class="n">std</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">a</span> <span class="o">/=</span> <span class="n">std</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">return</span> <span class="n">a</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.pcoa_utils.svd_rank">
<code class="highlight language-python"><span class="n">svd_rank</span><span class="p">(</span><span class="n">M_shape</span><span class="p">,</span> <span class="n">S</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Matrix rank of <code>M</code> given its singular values <code>S</code>.
See <code>np.linalg.matrix_rank</code> for a rationale on the tolerance
(we're not using that function because it doesn't let us reuse a
precomputed SVD).</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/pcoa_utils.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">svd_rank</span><span class="p">(</span><span class="n">M_shape</span><span class="p">,</span> <span class="n">S</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Matrix rank of `M` given its singular values `S`.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    See `np.linalg.matrix_rank` for a rationale on the tolerance</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    (we&#39;re not using that function because it doesn&#39;t let us reuse a</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    precomputed SVD).&quot;&quot;&quot;</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="k">if</span> <span class="n">tol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>        <span class="n">tol</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">*</span> <span class="nb">max</span><span class="p">(</span><span class="n">M_shape</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">finfo</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span><span class="o">.</span><span class="n">eps</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">S</span> <span class="o">&gt;</span> <span class="n">tol</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.external.umap">
        <code>umap</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.external.umap.run_umap">
<code class="highlight language-python"><span class="n">run_umap</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Runs UMAP on a expression matrix.
Parameters</p>
<hr />
<p>rnaseq_data : pandas.DataFrame
    A dataframe of gene expression values wherein the rows are the genes or
    embeddings of a dimensionality reduction method and columns the cells,
    tissues or samples.</p>
<p>axis : int, default=0
    An axis of the dataframe (0 across rows, 1 across columns).
    Across rows means that the UMAP is to compare genes, while
    across columns is to compare cells, tissues or samples.</p>
<p>metric : str, default='euclidean'
    The distance metric to use. The distance function can be 'braycurtis',
    'canberra', 'chebyshev', 'cityblock', 'correlation', 'cosine', 'dice',
    'euclidean', 'hamming', 'jaccard', 'jensenshannon', 'kulsinski',
    'mahalanobis', 'matching', 'minkowski', 'rogerstanimoto', 'russellrao',
    'seuclidean', 'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule'.</p>
<div class="admonition min_dist">
<p class="admonition-title">float, default=0.4</p>
<p>The effective minimum distance between embedded points. Smaller values
will result in a more clustered/clumped embedding where nearby points
on the manifold are drawn closer together, while larger values will
result on a more even dispersal of points. The value should be set
relative to the <code>spread</code> value, which determines the scale at which
embedded points will be spread out.</p>
</div>
<div class="admonition n_neighbors">
<p class="admonition-title">float, default=8</p>
<p>The size of local neighborhood (in terms of number of neighboring
sample points) used for manifold approximation. Larger values
result in more global views of the manifold, while smaller
values result in more local data being preserved. In general
values should be in the range 2 to 100.</p>
</div>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>**kwargs : dict
    Extra arguments for UMAP as defined in umap.UMAP.</p>
<h6 id="cell2cell.external.umap.run_umap--returns">Returns</h6>
<p>umap_df : pandas.DataFrame
    Dataframe containing the UMAP embeddings for the axis analyzed.
    Contains columns 'umap1 and 'umap2'.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/external/umap.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">run_umap</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Runs UMAP on a expression matrix.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">        A dataframe of gene expression values wherein the rows are the genes or</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        embeddings of a dimensionality reduction method and columns the cells,</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        tissues or samples.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        An axis of the dataframe (0 across rows, 1 across columns).</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Across rows means that the UMAP is to compare genes, while</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        across columns is to compare cells, tissues or samples.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    metric : str, default=&#39;euclidean&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        The distance metric to use. The distance function can be &#39;braycurtis&#39;,</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        &#39;canberra&#39;, &#39;chebyshev&#39;, &#39;cityblock&#39;, &#39;correlation&#39;, &#39;cosine&#39;, &#39;dice&#39;,</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        &#39;euclidean&#39;, &#39;hamming&#39;, &#39;jaccard&#39;, &#39;jensenshannon&#39;, &#39;kulsinski&#39;,</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        &#39;mahalanobis&#39;, &#39;matching&#39;, &#39;minkowski&#39;, &#39;rogerstanimoto&#39;, &#39;russellrao&#39;,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        &#39;seuclidean&#39;, &#39;sokalmichener&#39;, &#39;sokalsneath&#39;, &#39;sqeuclidean&#39;, &#39;yule&#39;.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    min_dist: float, default=0.4</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        The effective minimum distance between embedded points. Smaller values</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        will result in a more clustered/clumped embedding where nearby points</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        on the manifold are drawn closer together, while larger values will</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        result on a more even dispersal of points. The value should be set</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        relative to the ``spread`` value, which determines the scale at which</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        embedded points will be spread out.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    n_neighbors: float, default=8</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        The size of local neighborhood (in terms of number of neighboring</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        sample points) used for manifold approximation. Larger values</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        result in more global views of the manifold, while smaller</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        values result in more local data being preserved. In general</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        values should be in the range 2 to 100.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Extra arguments for UMAP as defined in umap.UMAP.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    -------</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    umap_df : pandas.DataFrame</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Dataframe containing the UMAP embeddings for the axis analyzed.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Contains columns &#39;umap1 and &#39;umap2&#39;.</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="c1"># Organize data</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">rnaseq_data</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">elif</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The parameter axis must be either 0 or 1.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="c1"># Compute distances</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">D</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">pdist</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">D_sq</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="c1"># Run UMAP</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="n">model</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">metric</span><span class="o">=</span><span class="s2">&quot;precomputed&quot;</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                      <span class="n">min_dist</span><span class="o">=</span><span class="n">min_dist</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                      <span class="n">n_neighbors</span><span class="o">=</span><span class="n">n_neighbors</span><span class="p">,</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                      <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                      <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                      <span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="n">trans_D</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">D_sq</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="c1"># Organize results</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">umap_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">trans_D</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;umap1&#39;</span><span class="p">,</span> <span class="s1">&#39;umap2&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="k">return</span> <span class="n">umap_df</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.io">
        <code>io</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.io.directories">
        <code>directories</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.directories.create_directory">
<code class="highlight language-python"><span class="n">create_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Creates a directory.</p>
<p>Uses a path to create a directory. It creates
all intermediate folders before creating the
leaf folder.</p>
<h6 id="cell2cell.io.directories.create_directory--parameters">Parameters</h6>
<p>pathname : str
    Full path of the folder to create.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/directories.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">create_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Creates a directory.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Uses a path to create a directory. It creates</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    all intermediate folders before creating the</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    leaf folder.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    pathname : str</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Full path of the folder to create.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">pathname</span><span class="p">):</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">pathname</span><span class="p">)</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> was created successfully.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pathname</span><span class="p">))</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> already exists.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pathname</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.directories.get_files_from_directory">
<code class="highlight language-python"><span class="n">get_files_from_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">,</span> <span class="n">dir_in_filepath</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains a list of filenames in a folder.</p>
<h6 id="cell2cell.io.directories.get_files_from_directory--parameters">Parameters</h6>
<p>pathname : str
    Full path of the folder to explore.</p>
<p>dir_in_filepath : boolean, default=False
    Whether adding <code>pathname</code> to the filenames</p>
<h6 id="cell2cell.io.directories.get_files_from_directory--returns">Returns</h6>
<p>filenames : list
    A list containing the names (strings) of the files
    in the folder.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/directories.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_files_from_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">,</span> <span class="n">dir_in_filepath</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Obtains a list of filenames in a folder.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    pathname : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Full path of the folder to explore.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    dir_in_filepath : boolean, default=False</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Whether adding `pathname` to the filenames</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    filenames : list</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        A list containing the names (strings) of the files</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        in the folder.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">fsencode</span><span class="p">(</span><span class="n">pathname</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">pathname</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">fsdecode</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">if</span> <span class="n">dir_in_filepath</span> <span class="k">else</span> <span class="n">os</span><span class="o">.</span><span class="n">fsdecode</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">)]</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">return</span> <span class="n">filenames</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.io.read_data">
        <code>read_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_cutoffs">
<code class="highlight language-python"><span class="n">load_cutoffs</span><span class="p">(</span><span class="n">cutoff_file</span><span class="p">,</span> <span class="n">gene_column</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">drop_nangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log_transformation</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads a table of cutoff of thresholding values for each gene.</p>
<h6 id="cell2cell.io.read_data.load_cutoffs--parameters">Parameters</h6>
<p>cutoff_file : str
    Absolute path to a file containing thresholding values for genes.
    Genes are rows and threshold values are in the only column beyond
    the one containing the gene names.</p>
<p>gene_column : str, default=None
    Column name where the gene labels are contained. If None, the
    first column will be assummed to contain gene names.</p>
<p>drop_nangenes : boolean, default=True
    Whether dropping empty genes across all columns.</p>
<p>log_transformation : boolean, default=False
    Whether applying a log10 transformation on the data.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for loading files the function
    cell2cell.io.read_data.load_table</p>
<h6 id="cell2cell.io.read_data.load_cutoffs--returns">Returns</h6>
<p>cutoff_data : pandas.DataFrame
    Dataframe with the cutoff values for each gene. Rows are genes
    and just one column is included, which corresponds to 'value',
    wherein the thresholding or cutoff values are contained.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_cutoffs</span><span class="p">(</span><span class="n">cutoff_file</span><span class="p">,</span> <span class="n">gene_column</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">drop_nangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log_transformation</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Loads a table of cutoff of thresholding values for each gene.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cutoff_file : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Absolute path to a file containing thresholding values for genes.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Genes are rows and threshold values are in the only column beyond</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the one containing the gene names.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    gene_column : str, default=None</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Column name where the gene labels are contained. If None, the</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        first column will be assummed to contain gene names.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    drop_nangenes : boolean, default=True</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Whether dropping empty genes across all columns.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    log_transformation : boolean, default=False</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether applying a log10 transformation on the data.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Extra arguments for loading files the function</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        cell2cell.io.read_data.load_table</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    cutoff_data : pandas.DataFrame</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Dataframe with the cutoff values for each gene. Rows are genes</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        and just one column is included, which corresponds to &#39;value&#39;,</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        wherein the thresholding or cutoff values are contained.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Opening Cutoff datasets from </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cutoff_file</span><span class="p">))</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">load_table</span><span class="p">(</span><span class="n">cutoff_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">gene_column</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">cutoff_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">gene_column</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">cutoff_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">cutoff_data</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Keep only numeric datasets</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">cutoff_data</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">number</span><span class="p">])</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">if</span> <span class="n">drop_nangenes</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">drop_empty_genes</span><span class="p">(</span><span class="n">cutoff_data</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">if</span> <span class="n">log_transformation</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">cutoff_data</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">log10_transformation</span><span class="p">(</span><span class="n">cutoff_data</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="n">cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cutoff_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;value&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="n">cutoff_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">cols</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">return</span> <span class="n">cutoff_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_go_annotations">
<code class="highlight language-python"><span class="n">load_go_annotations</span><span class="p">(</span><span class="n">goa_file</span><span class="p">,</span> <span class="n">experimental_evidence</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads GO annotations for each gene in a given organism.</p>
<h6 id="cell2cell.io.read_data.load_go_annotations--parameters">Parameters</h6>
<p>goa_file : str
    Absolute path to an ga file. It could be an URL as for example:
    goa_file = 'http://current.geneontology.org/annotations/wb.gaf.gz'</p>
<p>experimental_evidence : boolean, default=True
    Whether considering only annotations with experimental evidence
    (at least one article/evidence).</p>
<p>verbose : boolean, default=True
        Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.io.read_data.load_go_annotations--returns">Returns</h6>
<p>goa : pandas.DataFrame
    Dataframe containing information about GO term annotations of each
    gene for a given organism according to the ga file.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_go_annotations</span><span class="p">(</span><span class="n">goa_file</span><span class="p">,</span> <span class="n">experimental_evidence</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Loads GO annotations for each gene in a given organism.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    goa_file : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Absolute path to an ga file. It could be an URL as for example:</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        goa_file = &#39;http://current.geneontology.org/annotations/wb.gaf.gz&#39;</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    experimental_evidence : boolean, default=True</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Whether considering only annotations with experimental evidence</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        (at least one article/evidence).</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    goa : pandas.DataFrame</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Dataframe containing information about GO term annotations of each</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        gene for a given organism according to the ga file.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="kn">import</span> <span class="nn">cell2cell.external.goenrich</span> <span class="k">as</span> <span class="nn">goenrich</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Opening GO annotations from </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">goa_file</span><span class="p">))</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">goa</span> <span class="o">=</span> <span class="n">goenrich</span><span class="o">.</span><span class="n">goa</span><span class="p">(</span><span class="n">goa_file</span><span class="p">,</span> <span class="n">experimental_evidence</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">goa_cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">goa</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">goa</span> <span class="o">=</span> <span class="n">goa</span><span class="p">[</span><span class="n">goa_cols</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">goa_cols</span><span class="p">[</span><span class="mi">4</span><span class="p">]]]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;db&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Name&#39;</span><span class="p">,</span> <span class="s1">&#39;GO&#39;</span><span class="p">]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">goa</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">new_cols</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="nb">print</span><span class="p">(</span><span class="n">goa_file</span> <span class="o">+</span> <span class="s1">&#39; was correctly loaded&#39;</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">return</span> <span class="n">goa</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_go_terms">
<code class="highlight language-python"><span class="n">load_go_terms</span><span class="p">(</span><span class="n">go_terms_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads GO term information from a obo-basic file.</p>
<h6 id="cell2cell.io.read_data.load_go_terms--parameters">Parameters</h6>
<p>go_terms_file : str
    Absolute path to an obo file. It could be an URL as for example:
    go_terms_file = 'http://purl.obolibrary.org/obo/go/go-basic.obo'</p>
<p>verbose : boolean, default=True
        Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.io.read_data.load_go_terms--returns">Returns</h6>
<p>go_terms : networkx.Graph
    NetworkX Graph containing GO terms datasets from .obo file.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_go_terms</span><span class="p">(</span><span class="n">go_terms_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Loads GO term information from a obo-basic file.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    go_terms_file : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Absolute path to an obo file. It could be an URL as for example:</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        go_terms_file = &#39;http://purl.obolibrary.org/obo/go/go-basic.obo&#39;</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    -------</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    go_terms : networkx.Graph</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        NetworkX Graph containing GO terms datasets from .obo file.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="kn">import</span> <span class="nn">cell2cell.external.goenrich</span> <span class="k">as</span> <span class="nn">goenrich</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Opening GO terms from </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">go_terms_file</span><span class="p">))</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">go_terms</span> <span class="o">=</span> <span class="n">goenrich</span><span class="o">.</span><span class="n">ontology</span><span class="p">(</span><span class="n">go_terms_file</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="nb">print</span><span class="p">(</span><span class="n">go_terms_file</span> <span class="o">+</span> <span class="s1">&#39; was correctly loaded&#39;</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">go_terms</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_metadata">
<code class="highlight language-python"><span class="n">load_metadata</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">,</span> <span class="n">cell_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads a metadata table for a given list of cells.</p>
<h6 id="cell2cell.io.read_data.load_metadata--parameters">Parameters</h6>
<p>metadata_file : str
    Absolute path to a file containing a metadata table for
    cell-types/tissues/samples in a RNA-seq dataset.</p>
<p>cell_labels : list, default=None
    List of cell-types/tissues/samples to consider. Names must
    match the labels in the metadata table. These names must
    be contained in the values of the column indicated
    by index_col.</p>
<p>index_col : str, default=None
    Column to be consider the index of the metadata.
    If None, the index will be the numbers of the rows.</p>
<p>**kwargs : dict
    Extra arguments for loading files the function
    cell2cell.io.read_data.load_table</p>
<h6 id="cell2cell.io.read_data.load_metadata--returns">Returns</h6>
<p>meta : pandas.DataFrame
    Metadata for the cell-types/tissues/samples provided.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_metadata</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">,</span> <span class="n">cell_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Loads a metadata table for a given list of cells.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    metadata_file : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Absolute path to a file containing a metadata table for</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        cell-types/tissues/samples in a RNA-seq dataset.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    cell_labels : list, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List of cell-types/tissues/samples to consider. Names must</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        match the labels in the metadata table. These names must</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        be contained in the values of the column indicated</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        by index_col.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    index_col : str, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Column to be consider the index of the metadata.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        If None, the index will be the numbers of the rows.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Extra arguments for loading files the function</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell2cell.io.read_data.load_table</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    meta : pandas.DataFrame</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Metadata for the cell-types/tissues/samples provided.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">meta</span> <span class="o">=</span> <span class="n">load_table</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">index_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">index_col</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">meta</span><span class="o">.</span><span class="n">columns</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">indexing</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="n">indexing</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">cell_labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">meta</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">meta</span><span class="p">[</span><span class="n">index_col</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_labels</span><span class="p">)]</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">indexing</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">meta</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">index_col</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">return</span> <span class="n">meta</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_ppi">
<code class="highlight language-python"><span class="n">load_ppi</span><span class="p">(</span><span class="n">ppi_file</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rnaseq_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strna</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads a list of protein-protein interactions from a table and
returns it in a simplified format.</p>
<h6 id="cell2cell.io.read_data.load_ppi--parameters">Parameters</h6>
<p>ppi_file : str
    Absolute path to a file containing a list of protein-protein
    interactions.</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors. Example: ('partner_A', 'partner_B').</p>
<p>sort_values : str, default=None
    Column name of a column used for sorting the table. If it is not None,
    that the column, and the whole dataframe, we will be ordered in an
    ascending manner.</p>
<p>score : str, default=None
    Column name of a column containing weights to consider in the cell-cell
    interactions/communication analyses. If None, no weights are used and
    PPIs are assumed to have an equal contribution to CCI and CCC scores.</p>
<p>rnaseq_genes : list, default=None
    List of genes in a RNA-seq dataset to filter the list of PPIs. If None,
    the entire list will be used.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44". If None, it is assummed
    that the list does not contains complexes.</p>
<p>dropna : boolean, default=False
    Whether dropping PPIs with any missing information.</p>
<p>strna : str, default=''
    If dropna is False, missing values will be filled with strna.</p>
<p>upper_letter_comparison : boolean, default=False
    Whether making uppercase the gene names in the expression matrices and the
    protein names in the ppi_data to match their names and integrate their
    respective expression level. Useful when there are inconsistencies in the
    names between the expression matrix and the ligand-receptor annotations.</p>
<p>**kwargs : dict
    Extra arguments for loading files the function
    cell2cell.io.read_data.load_table</p>
<h6 id="cell2cell.io.read_data.load_ppi--returns">Returns</h6>
<p>simplified_ppi : pandas.DataFrame
    A simplified list of PPIs. In this case, interaction_columns are renamed
    into 'A' and 'B' for the first and second interacting proteins, respectively.
    A third column 'score' is included, containing weights of PPIs.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_ppi</span><span class="p">(</span><span class="n">ppi_file</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rnaseq_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>             <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strna</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Loads a list of protein-protein interactions from a table and</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    returns it in a simplified format.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ppi_file : str</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Absolute path to a file containing a list of protein-protein</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        interactions.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        for the receptors. Example: (&#39;partner_A&#39;, &#39;partner_B&#39;).</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    sort_values : str, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Column name of a column used for sorting the table. If it is not None,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        that the column, and the whole dataframe, we will be ordered in an</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        ascending manner.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    score : str, default=None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Column name of a column containing weights to consider in the cell-cell</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        interactions/communication analyses. If None, no weights are used and</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        PPIs are assumed to have an equal contribution to CCI and CCC scores.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    rnaseq_genes : list, default=None</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        List of genes in a RNA-seq dataset to filter the list of PPIs. If None,</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        the entire list will be used.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;. If None, it is assummed</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        that the list does not contains complexes.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    dropna : boolean, default=False</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Whether dropping PPIs with any missing information.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    strna : str, default=&#39;&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        If dropna is False, missing values will be filled with strna.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    upper_letter_comparison : boolean, default=False</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Whether making uppercase the gene names in the expression matrices and the</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        protein names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        respective expression level. Useful when there are inconsistencies in the</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        names between the expression matrix and the ligand-receptor annotations.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Extra arguments for loading files the function</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        cell2cell.io.read_data.load_table</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    -------</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    simplified_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        A simplified list of PPIs. In this case, interaction_columns are renamed</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        into &#39;A&#39; and &#39;B&#39; for the first and second interacting proteins, respectively.</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        A third column &#39;score&#39; is included, containing weights of PPIs.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Opening PPI datasets from </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ppi_file</span><span class="p">))</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">load_table</span><span class="p">(</span><span class="n">ppi_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>  <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">simplified_ppi</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">preprocess_ppi_data</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                                             <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                                             <span class="n">sort_values</span><span class="o">=</span><span class="n">sort_values</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                                             <span class="n">score</span><span class="o">=</span><span class="n">score</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                             <span class="n">rnaseq_genes</span><span class="o">=</span><span class="n">rnaseq_genes</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                             <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                             <span class="n">dropna</span><span class="o">=</span><span class="n">dropna</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                                             <span class="n">strna</span><span class="o">=</span><span class="n">strna</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                             <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                             <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">return</span> <span class="n">simplified_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_rnaseq">
<code class="highlight language-python"><span class="n">load_rnaseq</span><span class="p">(</span><span class="n">rnaseq_file</span><span class="p">,</span> <span class="n">gene_column</span><span class="p">,</span> <span class="n">drop_nangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log_transformation</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Loads a gene expression matrix for a RNA-seq experiment. Preprocessing
steps can be done on-the-fly.</p>
<h6 id="cell2cell.io.read_data.load_rnaseq--parameters">Parameters</h6>
<p>rnaseq_file : str
    Absolute path to a file containing a gene expression matrix. Genes
    are rows and cell-types/tissues/samples are columns.</p>
<p>gene_column : str
    Column name where the gene labels are contained.</p>
<p>drop_nangenes : boolean, default=True
    Whether dropping empty genes across all columns.</p>
<p>log_transformation : boolean, default=False
    Whether applying a log10 transformation on the data.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for loading files the function
    cell2cell.io.read_data.load_table</p>
<h6 id="cell2cell.io.read_data.load_rnaseq--returns">Returns</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_rnaseq</span><span class="p">(</span><span class="n">rnaseq_file</span><span class="p">,</span> <span class="n">gene_column</span><span class="p">,</span> <span class="n">drop_nangenes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log_transformation</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Loads a gene expression matrix for a RNA-seq experiment. Preprocessing</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    steps can be done on-the-fly.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_file : str</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Absolute path to a file containing a gene expression matrix. Genes</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        are rows and cell-types/tissues/samples are columns.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    gene_column : str</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Column name where the gene labels are contained.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    drop_nangenes : boolean, default=True</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Whether dropping empty genes across all columns.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    log_transformation : boolean, default=False</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether applying a log10 transformation on the data.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Extra arguments for loading files the function</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        cell2cell.io.read_data.load_table</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Opening RNAseq datasets from </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">rnaseq_file</span><span class="p">))</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">load_table</span><span class="p">(</span><span class="n">rnaseq_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">gene_column</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">gene_column</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="c1"># Keep only numeric datasets</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">number</span><span class="p">])</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">if</span> <span class="n">drop_nangenes</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">drop_empty_genes</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">if</span> <span class="n">log_transformation</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">log10_transformation</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">return</span> <span class="n">rnaseq_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_table">
<code class="highlight language-python"><span class="n">load_table</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Opens a file containing a table into a pandas dataframe.</p>
<h6 id="cell2cell.io.read_data.load_table--parameters">Parameters</h6>
<p>filename : str
    Absolute path to a file storing a table.</p>
<p>format : str, default='auto'
    Format of the file.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;auto&#39; : Automatically determines the format given
    the file extension. Files ending with .gz will be
    consider as tsv files.
- &#39;excel&#39; : An excel file, either .xls or .xlsx
- &#39;csv&#39; : Comma separated value format
- &#39;tsv&#39; : Tab separated value format
- &#39;txt&#39; : Text file
</code></pre></div>

<p>sep : str, default='        '
    Separation between columns. Examples are: '     ', ' ', ';', ',', etc.</p>
<p>sheet_name : str, int, list, or None, default=0
    Strings are used for sheet names. Integers are used in zero-indexed
    sheet positions. Lists of strings/integers are used to request
    multiple sheets. Specify None to get all sheets.
    Available cases:</p>
<div class="codehilite"><pre><span></span><code>- Defaults to 0: 1st sheet as a DataFrame
- 1: 2nd sheet as a DataFrame
- &quot;Sheet1&quot;: Load sheet with name Sheet1
- [0, 1, &quot;Sheet5&quot;]: Load first, second and sheet named
    Sheet5 as a dict of DataFrame
- None: All sheets.
</code></pre></div>

<p>compression : str, or None, default=infer
    For on-the-fly decompression of on-disk data. If infer, detects
    compression from the following extensions: .gz, .bz2, .zip, or .xz
    (otherwise no decompression). If using zip, the ZIP file must contain
    only one data file to be read in. Set to None for no decompression.
    Options: {infer, gzip, bz2, zip, xz, None}</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for loading files with the respective pandas function
    given the format of the file.</p>
<h6 id="cell2cell.io.read_data.load_table--returns">Returns</h6>
<p>table : pandas.DataFrame
    Dataframe containing the table stored in a file.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_table</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Opens a file containing a table into a pandas dataframe.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Absolute path to a file storing a table.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    format : str, default=&#39;auto&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Format of the file.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;auto&#39; : Automatically determines the format given</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">            the file extension. Files ending with .gz will be</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">            consider as tsv files.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;excel&#39; : An excel file, either .xls or .xlsx</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;csv&#39; : Comma separated value format</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;tsv&#39; : Tab separated value format</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;txt&#39; : Text file</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    sep : str, default=&#39;\t&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Separation between columns. Examples are: &#39;\t&#39;, &#39; &#39;, &#39;;&#39;, &#39;,&#39;, etc.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    sheet_name : str, int, list, or None, default=0</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Strings are used for sheet names. Integers are used in zero-indexed</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        sheet positions. Lists of strings/integers are used to request</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        multiple sheets. Specify None to get all sheets.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Available cases:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - Defaults to 0: 1st sheet as a DataFrame</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        - 1: 2nd sheet as a DataFrame</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &quot;Sheet1&quot;: Load sheet with name Sheet1</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - [0, 1, &quot;Sheet5&quot;]: Load first, second and sheet named</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">            Sheet5 as a dict of DataFrame</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        - None: All sheets.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    compression : str, or None, default=infer</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        For on-the-fly decompression of on-disk data. If infer, detects</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        compression from the following extensions: .gz, .bz2, .zip, or .xz</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        (otherwise no decompression). If using zip, the ZIP file must contain</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        only one data file to be read in. Set to None for no decompression.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Options: {infer, gzip, bz2, zip, xz, None}</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        Extra arguments for loading files with the respective pandas function</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        given the format of the file.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    -------</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    table : pandas.DataFrame</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Dataframe containing the table stored in a file.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">return</span> <span class="kc">None</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;.gz&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>            <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;csv&#39;</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>            <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="k">elif</span> <span class="p">(</span><span class="s1">&#39;.xlsx&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="s1">&#39;.xls&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>            <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;excel&#39;</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="k">elif</span> <span class="p">(</span><span class="s1">&#39;.csv&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;csv&#39;</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>            <span class="n">compression</span><span class="o">=</span><span class="kc">None</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>        <span class="k">elif</span> <span class="p">(</span><span class="s1">&#39;.tsv&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="s1">&#39;.txt&#39;</span> <span class="ow">in</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>            <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;csv&#39;</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>            <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>            <span class="n">compression</span><span class="o">=</span><span class="kc">None</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;excel&#39;</span><span class="p">:</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>        <span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="n">sheet_name</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">elif</span> <span class="p">(</span><span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;csv&#39;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;tsv&#39;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;txt&#39;</span><span class="p">):</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">sep</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Specify a correct format&quot;</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="k">return</span> <span class="kc">None</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="nb">print</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s1">&#39; was correctly loaded&#39;</span><span class="p">)</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>    <span class="k">return</span> <span class="n">table</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_tables_from_directory">
<code class="highlight language-python"><span class="n">load_tables_from_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">,</span> <span class="n">extension</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Opens all tables with the same extension in a folder.</p>
<h6 id="cell2cell.io.read_data.load_tables_from_directory--parameters">Parameters</h6>
<p>pathname : str
    Full path of the folder to explore.</p>
<p>extension : str
    Extension of the file.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;excel&#39; : An excel file, either .xls or .xlsx
- &#39;csv&#39; : Comma separated value format
- &#39;tsv&#39; : Tab separated value format
- &#39;txt&#39; : Text file
</code></pre></div>

<p>sep : str, default='        '
    Separation between columns. Examples are: '     ', ' ', ';', ',', etc.</p>
<p>sheet_name : str, int, list, or None, default=0
    Strings are used for sheet names. Integers are used in zero-indexed
    sheet positions. Lists of strings/integers are used to request
    multiple sheets. Specify None to get all sheets.
    Available cases:</p>
<div class="codehilite"><pre><span></span><code>- Defaults to 0: 1st sheet as a DataFrame
- 1: 2nd sheet as a DataFrame
- &quot;Sheet1&quot;: Load sheet with name Sheet1
- [0, 1, &quot;Sheet5&quot;]: Load first, second and sheet named
    Sheet5 as a dict of DataFrame
- None: All sheets.
</code></pre></div>

<p>compression : str, or None, default=infer
    For on-the-fly decompression of on-disk data. If infer, detects
    compression from the following extensions: .gz, .bz2, .zip, or .xz
    (otherwise no decompression). If using zip, the ZIP file must contain
    only one data file to be read in. Set to None for no decompression.
    Options: {gzip, bz2, zip, xz, None}</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for loading files with the respective pandas function
    given the format of the file.</p>
<h6 id="cell2cell.io.read_data.load_tables_from_directory--returns">Returns</h6>
<p>data : dict
    Dictionary containing the tables (pandas.DataFrame) loaded from the files.
    Keys are the filenames without the extension and values are the dataframes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_tables_from_directory</span><span class="p">(</span><span class="n">pathname</span><span class="p">,</span> <span class="n">extension</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Opens all tables with the same extension in a folder.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    pathname : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Full path of the folder to explore.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    extension : str</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Extension of the file.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        - &#39;excel&#39; : An excel file, either .xls or .xlsx</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        - &#39;csv&#39; : Comma separated value format</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;tsv&#39; : Tab separated value format</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;txt&#39; : Text file</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    sep : str, default=&#39;\t&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Separation between columns. Examples are: &#39;\t&#39;, &#39; &#39;, &#39;;&#39;, &#39;,&#39;, etc.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    sheet_name : str, int, list, or None, default=0</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Strings are used for sheet names. Integers are used in zero-indexed</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        sheet positions. Lists of strings/integers are used to request</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        multiple sheets. Specify None to get all sheets.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Available cases:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - Defaults to 0: 1st sheet as a DataFrame</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - 1: 2nd sheet as a DataFrame</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &quot;Sheet1&quot;: Load sheet with name Sheet1</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - [0, 1, &quot;Sheet5&quot;]: Load first, second and sheet named</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">            Sheet5 as a dict of DataFrame</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - None: All sheets.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    compression : str, or None, default=infer</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        For on-the-fly decompression of on-disk data. If infer, detects</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        compression from the following extensions: .gz, .bz2, .zip, or .xz</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        (otherwise no decompression). If using zip, the ZIP file must contain</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        only one data file to be read in. Set to None for no decompression.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Options: {gzip, bz2, zip, xz, None}</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Extra arguments for loading files with the respective pandas function</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        given the format of the file.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    -------</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    data : dict</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Dictionary containing the tables (pandas.DataFrame) loaded from the files.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        Keys are the filenames without the extension and values are the dataframes.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">assert</span> <span class="n">extension</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;excel&#39;</span><span class="p">,</span> <span class="s1">&#39;csv&#39;</span><span class="p">,</span> <span class="s1">&#39;tsv&#39;</span><span class="p">,</span> <span class="s1">&#39;txt&#39;</span><span class="p">],</span> <span class="s2">&quot;Enter a valid `extension`.&quot;</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">filenames</span> <span class="o">=</span> <span class="n">get_files_from_directory</span><span class="p">(</span><span class="n">pathname</span><span class="o">=</span><span class="n">pathname</span><span class="p">,</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>                                         <span class="n">dir_in_filepath</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">if</span> <span class="n">compression</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="n">comp</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="k">assert</span> <span class="n">compression</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="s1">&#39;bz2&#39;</span><span class="p">,</span> <span class="s1">&#39;zip&#39;</span><span class="p">,</span> <span class="s1">&#39;xz&#39;</span><span class="p">],</span> <span class="s2">&quot;Enter a valid `compression`.&quot;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="n">comp</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="n">compression</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="n">extension</span> <span class="o">+</span> <span class="n">comp</span><span class="p">):</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Loading </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>            <span class="n">sample</span> <span class="o">=</span> <span class="n">basename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="n">extension</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>            <span class="n">data</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">load_table</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                      <span class="nb">format</span><span class="o">=</span><span class="n">extension</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                                      <span class="n">sep</span><span class="o">=</span><span class="n">sep</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                      <span class="n">sheet_name</span><span class="o">=</span><span class="n">sheet_name</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                      <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                      <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="k">return</span> <span class="n">data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_tensor">
<code class="highlight language-python"><span class="n">load_tensor</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Imports a communication tensor that could be used
with Tensor-cell2cell.</p>
<h6 id="cell2cell.io.read_data.load_tensor--parameters">Parameters</h6>
<p>filename : str
    Absolute path to a file storing a communication tensor
    that was previously saved by using pickle.</p>
<p _cupy_="'cupy'," _jax_="'jax'," _mxnet_="'mxnet'," _numpy_="'numpy'," _pytorch_="'pytorch'," _tensorflow_="'tensorflow'">backend : str, default=None
    Backend that TensorLy will use to perform calculations
    on this tensor. When None, the default backend used is
    the currently active backend, usually is ('numpy'). Options are:</p>
<p None="None" _cpu_="'cpu'," _cuda:0_="'cuda:0',">device : str, default=None
    Device to use when backend allows using multiple devices. Options are:</p>
<h6 id="cell2cell.io.read_data.load_tensor--returns">Returns</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_tensor</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Imports a communication tensor that could be used</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    with Tensor-cell2cell.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Absolute path to a file storing a communication tensor</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        that was previously saved by using pickle.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    backend : str, default=None</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Backend that TensorLy will use to perform calculations</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        on this tensor. When None, the default backend used is</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        the currently active backend, usually is (&#39;numpy&#39;). Options are:</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        {&#39;cupy&#39;, &#39;jax&#39;, &#39;mxnet&#39;, &#39;numpy&#39;, &#39;pytorch&#39;, &#39;tensorflow&#39;}</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Device to use when backend allows using multiple devices. Options are:</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">         {&#39;cpu&#39;, &#39;cuda:0&#39;, None}</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    -------</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        cell2cell.tensor.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">interaction_tensor</span> <span class="o">=</span> <span class="n">load_variable_with_pickle</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="s1">&#39;tl&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">globals</span><span class="p">():</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="kn">import</span> <span class="nn">tensorly</span> <span class="k">as</span> <span class="nn">tl</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">if</span> <span class="n">backend</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">tl</span><span class="o">.</span><span class="n">set_backend</span><span class="p">(</span><span class="n">backend</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="n">device</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">if</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="k">if</span> <span class="n">tl</span><span class="o">.</span><span class="n">get_backend</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pytorch&#39;</span><span class="p">,</span> <span class="s1">&#39;tensorflow&#39;</span><span class="p">]:</span>  <span class="c1"># Potential TODO: Include other backends that support different devices</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="k">if</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>            <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>            <span class="k">if</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">return</span> <span class="n">interaction_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_tensor_factors">
<code class="highlight language-python"><span class="n">load_tensor_factors</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Imports factors previously exported from a tensor
decomposition done in a cell2cell.tensor.BaseTensor-like object.</p>
<h6 id="cell2cell.io.read_data.load_tensor_factors--parameters">Parameters</h6>
<p>filename : str
    Absolute path to a file storing an excel file containing
    the factors, their loadings, and element names for each
    of the dimensions of a previously decomposed tensor.</p>
<h6 id="cell2cell.io.read_data.load_tensor_factors--returns">Returns</h6>
<p>factors : collections.OrderedDict
    An ordered dictionary wherein keys are the names of each
    tensor dimension, and values are the loadings in a pandas.DataFrame.
    In this dataframe, rows are the elements of the respective dimension
    and columns are the factors from the tensor factorization. Values
    are the corresponding loadings.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_tensor_factors</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Imports factors previously exported from a tensor</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    decomposition done in a cell2cell.tensor.BaseTensor-like object.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Absolute path to a file storing an excel file containing</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the factors, their loadings, and element names for each</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        of the dimensions of a previously decomposed tensor.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    factors : collections.OrderedDict</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        An ordered dictionary wherein keys are the names of each</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        tensor dimension, and values are the loadings in a pandas.DataFrame.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        In this dataframe, rows are the elements of the respective dimension</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        and columns are the factors from the tensor factorization. Values</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        are the corresponding loadings.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">xls</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelFile</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">factors</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">for</span> <span class="n">sheet_name</span> <span class="ow">in</span> <span class="n">xls</span><span class="o">.</span><span class="n">sheet_names</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">factors</span><span class="p">[</span><span class="n">sheet_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">xls</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">sheet_name</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">return</span> <span class="n">factors</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.read_data.load_variable_with_pickle">
<code class="highlight language-python"><span class="n">load_variable_with_pickle</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Imports a large size variable stored in a file previously
exported with pickle.</p>
<h6 id="cell2cell.io.read_data.load_variable_with_pickle--parameters">Parameters</h6>
<p>filename : str
    Absolute path to a file storing a python variable that
    was previously created by using pickle.</p>
<h6 id="cell2cell.io.read_data.load_variable_with_pickle--returns">Returns</h6>
<p>variable : a python variable
    The variable of interest.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/read_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">load_variable_with_pickle</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Imports a large size variable stored in a file previously</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    exported with pickle.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Absolute path to a file storing a python variable that</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        was previously created by using pickle.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    variable : a python variable</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        The variable of interest.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="n">max_bytes</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="mi">31</span> <span class="o">-</span> <span class="mi">1</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">bytes_in</span> <span class="o">=</span> <span class="nb">bytearray</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">input_size</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">input_size</span><span class="p">,</span> <span class="n">max_bytes</span><span class="p">):</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>            <span class="n">bytes_in</span> <span class="o">+=</span> <span class="n">f_in</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">max_bytes</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">variable</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">bytes_in</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">return</span> <span class="n">variable</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.io.save_data">
        <code>save_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.io.save_data.export_variable_with_pickle">
<code class="highlight language-python"><span class="n">export_variable_with_pickle</span><span class="p">(</span><span class="n">variable</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Exports a large size variable in a python readable way
using pickle.</p>
<h6 id="cell2cell.io.save_data.export_variable_with_pickle--parameters">Parameters</h6>
<p>variable : a python variable
    Variable to export</p>
<p>filename : str
    Complete path to the file wherein the variable will be
    stored. For example:
    /home/user/variable.pkl</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/io/save_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">export_variable_with_pickle</span><span class="p">(</span><span class="n">variable</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Exports a large size variable in a python readable way</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    using pickle.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    variable : a python variable</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Variable to export</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Complete path to the file wherein the variable will be</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        stored. For example:</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        /home/user/variable.pkl</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="n">max_bytes</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="mi">31</span> <span class="o">-</span> <span class="mi">1</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">bytes_out</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">variable</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">bytes_out</span><span class="p">),</span> <span class="n">max_bytes</span><span class="p">):</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>            <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">bytes_out</span><span class="p">[</span><span class="n">idx</span><span class="p">:</span><span class="n">idx</span> <span class="o">+</span> <span class="n">max_bytes</span><span class="p">])</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="nb">print</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39; was correctly saved.&#39;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.plotting">
        <code>plotting</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.aesthetics">
        <code>aesthetics</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.aesthetics.generate_legend">
<code class="highlight language-python"><span class="n">generate_legend</span><span class="p">(</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.01</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Legend&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">sorted_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Adds a legend to a previous plot or displays an independent legend
given specific colors for labels.</p>
<h6 id="cell2cell.plotting.aesthetics.generate_legend--parameters">Parameters</h6>
<p>color_dict : dict
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. Keys are the labels and values are the RGBA
    tuples.</p>
<p>loc : str, default='center left'
    Alignment of the legend given the location specieid in bbox_to_anchor.</p>
<p>bbox_to_anchor : tuple, default=(1.01, 0.5)
    Location of the legend in a (X, Y) format. For example, if you want
    your axes legend located at the figure's top right-hand corner instead
    of the axes' corner, simply specify the corner's location and the
    coordinate system of that location, which in this case would be (1, 1).</p>
<p>ncol : int, default=1
    Number of columns to display the legend.</p>
<p>fancybox : boolean, default=True
    Whether round edges should be enabled around the FancyBboxPatch which
    makes up the legend's background.</p>
<p>shadow : boolean, default=True
    Whether to draw a shadow behind the legend.</p>
<p>title : str, default='Legend'
    Title of the legend box</p>
<p>fontsize : int, default=14
    Size of the text in the legends.</p>
<p>sorted_labels : boolean, default=True
    Whether alphabetically sorting the labels.</p>
<p>fig : matplotlib.figure.Figure, default=None
    Figure object to add a legend. If fig=None and ax=None, a new empty
    figure will be generated.</p>
<p>ax : matplotlib.axes.Axes, default=None
    Axes instance for a plot.</p>
<h6 id="cell2cell.plotting.aesthetics.generate_legend--returns">Returns</h6>
<p>legend1 : matplotlib.legend.Legend
    A legend object in a figure.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/aesthetics.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_legend</span><span class="p">(</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.01</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Legend&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">sorted_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Adds a legend to a previous plot or displays an independent legend</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    given specific colors for labels.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    color_dict : dict</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        of major groups of cells. Keys are the labels and values are the RGBA</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        tuples.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    loc : str, default=&#39;center left&#39;</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Alignment of the legend given the location specieid in bbox_to_anchor.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    bbox_to_anchor : tuple, default=(1.01, 0.5)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Location of the legend in a (X, Y) format. For example, if you want</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        your axes legend located at the figure&#39;s top right-hand corner instead</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        of the axes&#39; corner, simply specify the corner&#39;s location and the</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        coordinate system of that location, which in this case would be (1, 1).</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    ncol : int, default=1</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Number of columns to display the legend.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    fancybox : boolean, default=True</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Whether round edges should be enabled around the FancyBboxPatch which</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        makes up the legend&#39;s background.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    shadow : boolean, default=True</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether to draw a shadow behind the legend.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    title : str, default=&#39;Legend&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Title of the legend box</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Size of the text in the legends.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    sorted_labels : boolean, default=True</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Whether alphabetically sorting the labels.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    fig : matplotlib.figure.Figure, default=None</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Figure object to add a legend. If fig=None and ax=None, a new empty</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        figure will be generated.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    ax : matplotlib.axes.Axes, default=None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Axes instance for a plot.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    -------</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    legend1 : matplotlib.legend.Legend</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        A legend object in a figure.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="n">color_patches</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">if</span> <span class="n">sorted_labels</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">iteritems</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">color_dict</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="n">iteritems</span> <span class="o">=</span> <span class="n">color_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">:</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="n">color_patches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">patches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">v</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)))</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">legend1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">color_patches</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                             <span class="n">loc</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                             <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="n">bbox_to_anchor</span><span class="p">,</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                             <span class="n">ncol</span><span class="o">=</span><span class="n">ncol</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                             <span class="n">fancybox</span><span class="o">=</span><span class="n">fancybox</span><span class="p">,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                             <span class="n">shadow</span><span class="o">=</span><span class="n">shadow</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                             <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                             <span class="n">title_fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                             <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="n">legend1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">color_patches</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                            <span class="n">loc</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                            <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="n">bbox_to_anchor</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                            <span class="n">ncol</span><span class="o">=</span><span class="n">ncol</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                            <span class="n">fancybox</span><span class="o">=</span><span class="n">fancybox</span><span class="p">,</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                            <span class="n">shadow</span><span class="o">=</span><span class="n">shadow</span><span class="p">,</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                            <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                            <span class="n">title_fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                            <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="k">return</span> <span class="n">legend1</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.aesthetics.get_colors_from_labels">
<code class="highlight language-python"><span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates colors for each label in a list given a colormap</p>
<h6 id="cell2cell.plotting.aesthetics.get_colors_from_labels--parameters">Parameters</h6>
<p>labels : list
    A list of labels to assign a color.</p>
<p>cmap : str, default='gist_rainbow'
    A matplotlib color palette name.</p>
<p>factor : int, default=1
    Factor to amplify the separation of colors.</p>
<h6 id="cell2cell.plotting.aesthetics.get_colors_from_labels--returns">Returns</h6>
<p>colors : dict
    A dictionary where the keys are the labels and the values
    correspond to the assigned colors.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/aesthetics.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates colors for each label in a list given a colormap</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    labels : list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A list of labels to assign a color.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    cmap : str, default=&#39;gist_rainbow&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        A matplotlib color palette name.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    factor : int, default=1</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Factor to amplify the separation of colors.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    -------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    colors : dict</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        A dictionary where the keys are the labels and the values</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        correspond to the assigned colors.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">assert</span> <span class="n">factor</span> <span class="o">&gt;=</span> <span class="mi">1</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">colors</span> <span class="o">=</span> <span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="p">())</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">factor</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">factor</span><span class="p">)</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">cm_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">is_number</span> <span class="o">=</span> <span class="nb">all</span><span class="p">((</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">e</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">e</span><span class="p">,</span> <span class="nb">int</span><span class="p">))</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">is_number</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">NUM_COLORS</span> <span class="o">=</span> <span class="n">factor</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">label</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">colors</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="n">colors</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">cm_</span><span class="p">((</span><span class="mi">1</span> <span class="o">+</span> <span class="p">((</span><span class="n">factor</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="n">factor</span><span class="p">))</span> <span class="o">*</span> <span class="n">i</span> <span class="o">/</span> <span class="n">NUM_COLORS</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">max_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">min_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">norm</span> <span class="o">=</span> <span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=-</span><span class="n">min_</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">max_</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">m</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">colors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>            <span class="n">colors</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">to_rgba</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">return</span> <span class="n">colors</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.aesthetics.map_colors_to_metadata">
<code class="highlight language-python"><span class="n">map_colors_to_metadata</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">ref_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Assigns a color to elements in a dataframe containing metadata.</p>
<h6 id="cell2cell.plotting.aesthetics.map_colors_to_metadata--parameters">Parameters</h6>
<p>metadata : pandas.DataFrame
    A dataframe with metadata for specific elements.</p>
<p>ref_df : pandas.DataFrame
    A dataframe whose columns contains a subset of
    elements in the metadata.</p>
<p>colors : dict, default=None
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. If colors is specified, cmap will be
    ignored.</p>
<p>sample_col : str, default='#SampleID'
    Column in the metadata for elements to color.</p>
<p>group_col : str, default='Groups'
    Column in the metadata containing the major groups of the elements
    to color.</p>
<p>cmap : str, default='gist_rainbow'
    Name of the color palette for coloring the major groups of elements.</p>
<h6 id="cell2cell.plotting.aesthetics.map_colors_to_metadata--returns">Returns</h6>
<p>new_colors : pandas.DataFrame
    A pandas dataframe where the index is the list of elements in the
    sample_col and the column group_col contains the colors assigned
    to each element given their groups.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/aesthetics.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">map_colors_to_metadata</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">ref_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                           <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Assigns a color to elements in a dataframe containing metadata.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    metadata : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe with metadata for specific elements.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ref_df : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A dataframe whose columns contains a subset of</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        elements in the metadata.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    colors : dict, default=None</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        of major groups of cells. If colors is specified, cmap will be</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    sample_col : str, default=&#39;#SampleID&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Column in the metadata for elements to color.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    group_col : str, default=&#39;Groups&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Column in the metadata containing the major groups of the elements</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        to color.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    cmap : str, default=&#39;gist_rainbow&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Name of the color palette for coloring the major groups of elements.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    -------</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    new_colors : pandas.DataFrame</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        A pandas dataframe where the index is the list of elements in the</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        sample_col and the column group_col contains the colors assigned</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        to each element given their groups.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">if</span> <span class="n">ref_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">meta_</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">ref_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">meta_</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">labels</span> <span class="o">=</span> <span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">if</span> <span class="n">colors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">colors</span> <span class="o">=</span> <span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="n">upd_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">v</span><span class="p">,</span> <span class="p">(</span><span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">))</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">labels</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">colors</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="n">colors</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">upd_dict</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="n">new_colors</span> <span class="o">=</span> <span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">new_colors</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">meta_</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">new_colors</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">group_col</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="k">return</span> <span class="n">new_colors</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.ccc_plot">
        <code>ccc_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.ccc_plot.clustermap_ccc">
<code class="highlight language-python"><span class="n">clustermap_ccc</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cell_labels</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;SENDER-CELL&#39;</span><span class="p">,</span> <span class="s1">&#39;RECEIVER-CELL&#39;</span><span class="p">),</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;jaccard&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">only_used_lr</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;Presence&#39;</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">row_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">col_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a clustermap (heatmap + dendrograms from a hierarchical
clustering) based on CCC scores for each LR pair in every cell-cell pair.</p>
<h6 id="cell2cell.plotting.ccc_plot.clustermap_ccc--parameters">Parameters</h6>
<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all a distance matrix after running the
    the method compute_pairwise_communication_scores. Alternatively, this
    object can be a numpy-array or a pandas DataFrame. Also, a
    SingleCellInteractions or a BulkInteractions object after running
    the method compute_pairwise_communication_scores.</p>
<p>metadata : pandas.Dataframe, default=None
    Metadata associated with the cells, cell types or samples in the
    matrix containing CCC scores. If None, cells will not be colored
    by major groups.</p>
<p>sample_col : str, default='#SampleID'
    Column in the metadata for the cells, cell types or samples
    in the matrix containing CCC scores.</p>
<p>group_col : str, default='Groups'
    Column in the metadata containing the major groups of cells, cell types
    or samples in the matrix with CCC scores.</p>
<p>meta_cmap : str, default='gist_rainbow'
    Name of the color palette for coloring the major groups of cells.</p>
<p>colors : dict, default=None
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. If colors is specified, meta_cmap will be
    ignored.</p>
<p>cell_labels : tuple, default=('SENDER-CELL','RECEIVER-CELL')
    A tuple containing the labels for indicating the group colors of
    sender and receiver cells if metadata or colors are provided.</p>
<p>metric : str, default='jaccard'
    The distance metric to use. The distance function can be 'braycurtis',
    'canberra', 'chebyshev', 'cityblock', 'correlation', 'cosine', 'dice',
    'euclidean', 'hamming', 'jaccard', 'jensenshannon', 'kulsinski',
    'mahalanobis', 'matching', 'minkowski', 'rogerstanimoto', 'russellrao',
    'seuclidean', 'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule'.</p>
<p>method : str, default='ward'
    Clustering method for computing a linkage as in
    scipy.cluster.hierarchy.linkage</p>
<p>optimal_leaf : boolean, default=True
    Whether sorting the leaf of the dendrograms to have a minimal distance
    between successive leaves. For more information, see
    scipy.cluster.hierarchy.optimal_leaf_ordering</p>
<p>excluded_cells : list, default=None
    List containing cell names that are present in the interaction_space
    object but that will be excluded from this plot.</p>
<p>title : str, default=''
    Title of the clustermap.</p>
<p>only_used_lr : boolean, default=True
    Whether displaying or not only LR pairs that were used at least by
    one pair of cells. If True, those LR pairs that were not used will
    not be displayed.</p>
<p>cbar_title : str, default='CCI score'
    Title for the colorbar, depending on the score employed.</p>
<p>cbar_fontsize : int, default=12
    Font size for the colorbar title as well as labels for axes X and Y.</p>
<p>row_fontsize : int, default=8
    Font size for the rows in the clustermap (ligand-receptor pairs).</p>
<p>col_fontsize : int, default=8
    Font size for the columns in the clustermap (sender-receiver cell pairs).</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<p>**kwargs : dict
    Dictionary containing arguments for the seaborn.clustermap function.</p>
<h6 id="cell2cell.plotting.ccc_plot.clustermap_ccc--returns">Returns</h6>
<p>fig : seaborn.matrix.ClusterGrid
    A seaborn ClusterGrid instance.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/ccc_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">clustermap_ccc</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                   <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cell_labels</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;SENDER-CELL&#39;</span><span class="p">,</span><span class="s1">&#39;RECEIVER-CELL&#39;</span><span class="p">),</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                   <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;jaccard&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                   <span class="n">only_used_lr</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;Presence&#39;</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">row_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">col_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>                   <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="sd">&#39;&#39;&#39;Generates a clustermap (heatmap + dendrograms from a hierarchical</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    clustering) based on CCC scores for each LR pair in every cell-cell pair.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Interaction space that contains all a distance matrix after running the</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        the method compute_pairwise_communication_scores. Alternatively, this</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        object can be a numpy-array or a pandas DataFrame. Also, a</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        SingleCellInteractions or a BulkInteractions object after running</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        the method compute_pairwise_communication_scores.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    metadata : pandas.Dataframe, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Metadata associated with the cells, cell types or samples in the</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        matrix containing CCC scores. If None, cells will not be colored</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        by major groups.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    sample_col : str, default=&#39;#SampleID&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Column in the metadata for the cells, cell types or samples</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        in the matrix containing CCC scores.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    group_col : str, default=&#39;Groups&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Column in the metadata containing the major groups of cells, cell types</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        or samples in the matrix with CCC scores.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    meta_cmap : str, default=&#39;gist_rainbow&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Name of the color palette for coloring the major groups of cells.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    colors : dict, default=None</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        of major groups of cells. If colors is specified, meta_cmap will be</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    cell_labels : tuple, default=(&#39;SENDER-CELL&#39;,&#39;RECEIVER-CELL&#39;)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        A tuple containing the labels for indicating the group colors of</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        sender and receiver cells if metadata or colors are provided.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    metric : str, default=&#39;jaccard&#39;</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        The distance metric to use. The distance function can be &#39;braycurtis&#39;,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        &#39;canberra&#39;, &#39;chebyshev&#39;, &#39;cityblock&#39;, &#39;correlation&#39;, &#39;cosine&#39;, &#39;dice&#39;,</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        &#39;euclidean&#39;, &#39;hamming&#39;, &#39;jaccard&#39;, &#39;jensenshannon&#39;, &#39;kulsinski&#39;,</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        &#39;mahalanobis&#39;, &#39;matching&#39;, &#39;minkowski&#39;, &#39;rogerstanimoto&#39;, &#39;russellrao&#39;,</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        &#39;seuclidean&#39;, &#39;sokalmichener&#39;, &#39;sokalsneath&#39;, &#39;sqeuclidean&#39;, &#39;yule&#39;.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    method : str, default=&#39;ward&#39;</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Clustering method for computing a linkage as in</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        scipy.cluster.hierarchy.linkage</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    optimal_leaf : boolean, default=True</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Whether sorting the leaf of the dendrograms to have a minimal distance</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        between successive leaves. For more information, see</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        scipy.cluster.hierarchy.optimal_leaf_ordering</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    excluded_cells : list, default=None</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        List containing cell names that are present in the interaction_space</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        object but that will be excluded from this plot.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    title : str, default=&#39;&#39;</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Title of the clustermap.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    only_used_lr : boolean, default=True</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        Whether displaying or not only LR pairs that were used at least by</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        one pair of cells. If True, those LR pairs that were not used will</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        not be displayed.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">    cbar_title : str, default=&#39;CCI score&#39;</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        Title for the colorbar, depending on the score employed.</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">    cbar_fontsize : int, default=12</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        Font size for the colorbar title as well as labels for axes X and Y.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    row_fontsize : int, default=8</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Font size for the rows in the clustermap (ligand-receptor pairs).</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    col_fontsize : int, default=8</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        Font size for the columns in the clustermap (sender-receiver cell pairs).</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        Dictionary containing arguments for the seaborn.clustermap function.</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">    -------</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">    fig : seaborn.matrix.ClusterGrid</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        A seaborn ClusterGrid instance.</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_elements&#39;</span><span class="p">):</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as an InteractionSpace class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_communication_scores() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                             <span class="s1">&#39; object to generate a communication matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>            <span class="n">df_</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="k">elif</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a communication matrix&#39;</span><span class="p">)</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>        <span class="n">df_</span> <span class="o">=</span> <span class="n">interaction_space</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_space&#39;</span><span class="p">):</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a Interactions class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>        <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_communication_scores() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>                             <span class="s1">&#39; object to generate a communication matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>            <span class="n">df_</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_communication_scores() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>                         <span class="s1">&#39; object to generate a communication matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="n">included_cells</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="k">for</span> <span class="n">cells</span> <span class="ow">in</span> <span class="n">df_</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>            <span class="n">include</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>            <span class="k">for</span> <span class="n">excluded_cell</span> <span class="ow">in</span> <span class="n">excluded_cells</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                <span class="k">if</span> <span class="n">excluded_cell</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">:</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>                    <span class="n">include</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>            <span class="k">if</span> <span class="n">include</span><span class="p">:</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                <span class="n">included_cells</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>        <span class="n">included_cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="p">[</span><span class="n">included_cells</span><span class="p">]</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>    <span class="k">if</span> <span class="n">only_used_lr</span><span class="p">:</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>        <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="p">[(</span><span class="n">df_</span><span class="o">.</span><span class="n">T</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()]</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>    <span class="c1"># Clustering</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>    <span class="n">dm_rows</span> <span class="o">=</span> <span class="n">compute_distance</span><span class="p">(</span><span class="n">df_</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="n">row_linkage</span> <span class="o">=</span> <span class="n">compute_linkage</span><span class="p">(</span><span class="n">dm_rows</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_leaf</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>    <span class="n">dm_cols</span> <span class="o">=</span> <span class="n">compute_distance</span><span class="p">(</span><span class="n">df_</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>    <span class="n">col_linkage</span> <span class="o">=</span> <span class="n">compute_linkage</span><span class="p">(</span><span class="n">dm_cols</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_leaf</span><span class="p">)</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>    <span class="c1"># Colors</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="n">metadata2</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>        <span class="n">meta_</span> <span class="o">=</span> <span class="n">metadata2</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>            <span class="n">meta_</span> <span class="o">=</span> <span class="n">meta_</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">meta_</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)]</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>        <span class="n">labels</span> <span class="o">=</span> <span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>        <span class="k">if</span> <span class="n">colors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>            <span class="n">colors</span> <span class="o">=</span> <span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">meta_cmap</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>            <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">elem</span> <span class="ow">in</span> <span class="n">colors</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">elem</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>        <span class="n">col_colors_L</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">included_cells</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">colors</span><span class="p">[</span><span class="n">metadata2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">metadata2</span><span class="p">[</span><span class="n">sample_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>                                                                                           <span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="n">col_colors_L</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">included_cells</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>        <span class="n">col_colors_L</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">cell_labels</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="n">col_colors_R</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">included_cells</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">colors</span><span class="p">[</span><span class="n">metadata2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">metadata2</span><span class="p">[</span><span class="n">sample_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>                                                                                           <span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>        <span class="n">col_colors_R</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">included_cells</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="n">col_colors_R</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">cell_labels</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="c1"># Clustermap</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df_</span><span class="p">,</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>                             <span class="n">cmap</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">dark_palette</span><span class="p">(</span><span class="s1">&#39;red&#39;</span><span class="p">),</span> <span class="c1">#plt.get_cmap(&#39;YlGnBu_r&#39;),</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>                             <span class="n">col_linkage</span><span class="o">=</span><span class="n">col_linkage</span><span class="p">,</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>                             <span class="n">row_linkage</span><span class="o">=</span><span class="n">row_linkage</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>                             <span class="n">col_colors</span><span class="o">=</span><span class="p">[</span><span class="n">col_colors_L</span><span class="p">,</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                                         <span class="n">col_colors_R</span><span class="p">],</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>                             <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                             <span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_col_colors</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">2.</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_col_colors</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cell_labels</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">row_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_col_colors</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">tick_right</span><span class="p">()</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">fig</span><span class="o">.</span><span class="n">ax_col_colors</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">visible</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df_</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>                             <span class="n">cmap</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">dark_palette</span><span class="p">(</span><span class="s1">&#39;red&#39;</span><span class="p">),</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                             <span class="n">col_linkage</span><span class="o">=</span><span class="n">col_linkage</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>                             <span class="n">row_linkage</span><span class="o">=</span><span class="n">row_linkage</span><span class="p">,</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                             <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>                             <span class="p">)</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>    <span class="c1"># Title</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">title</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">24</span><span class="p">)</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>    <span class="c1"># Color bar label</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>    <span class="n">cbar</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">colorbar</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>    <span class="n">cbar</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">cbar_title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">cbar_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>    <span class="n">cbar</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_label_position</span><span class="p">(</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>    <span class="c1"># Change tick labels</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>    <span class="n">xlabels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> \
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>               <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">()]</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>    <span class="n">ylabels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">))</span> \
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>               <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">()]</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a>    <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">,</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a>                                   <span class="n">fontsize</span><span class="o">=</span><span class="n">col_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a>                                   <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a>                                   <span class="n">rotation_mode</span><span class="o">=</span><span class="s1">&#39;anchor&#39;</span><span class="p">,</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a>                                   <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a>                                   <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a>    <span class="n">fig</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels</span><span class="p">)</span>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a>    <span class="c1"># Save Figure</span>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a>                    <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a>    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.cci_plot">
        <code>cci_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.cci_plot.clustermap_cci">
<code class="highlight language-python"><span class="n">clustermap_cci</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;CCI score&#39;</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a clustermap (heatmap + dendrograms from a hierarchical
clustering) based on CCI scores of cell-cell pairs.</p>
<h6 id="cell2cell.plotting.cci_plot.clustermap_cci--parameters">Parameters</h6>
<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all a distance matrix after running the
    the method compute_pairwise_cci_scores. Alternatively, this object
    can be a numpy-array or a pandas DataFrame. Also, a
    SingleCellInteractions or a BulkInteractions object after running
    the method compute_pairwise_cci_scores.</p>
<p>method : str, default='ward'
    Clustering method for computing a linkage as in
    scipy.cluster.hierarchy.linkage</p>
<p>optimal_leaf : boolean, default=True
    Whether sorting the leaf of the dendrograms to have a minimal distance
    between successive leaves. For more information, see
    scipy.cluster.hierarchy.optimal_leaf_ordering</p>
<p>metadata : pandas.Dataframe, default=None
    Metadata associated with the cells, cell types or samples in the
    matrix containing CCI scores. If None, cells will not be colored
    by major groups.</p>
<p>sample_col : str, default='#SampleID'
    Column in the metadata for the cells, cell types or samples
    in the matrix containing CCI scores.</p>
<p>group_col : str, default='Groups'
    Column in the metadata containing the major groups of cells, cell types
    or samples in the matrix with CCI scores.</p>
<p>meta_cmap : str, default='gist_rainbow'
    Name of the color palette for coloring the major groups of cells.</p>
<p>colors : dict, default=None
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. If colors is specified, meta_cmap will be
    ignored.</p>
<p>excluded_cells : list, default=None
    List containing cell names that are present in the interaction_space
    object but that will be excluded from this plot.</p>
<p>title : str, default=''
    Title of the clustermap.</p>
<p>cbar_title : str, default='CCI score'
    Title for the colorbar, depending on the score employed.</p>
<p>cbar_fontsize : int, default=18
    Font size for the colorbar title as well as labels for axes X and Y.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<p>**kwargs : dict
    Dictionary containing arguments for the seaborn.clustermap function.</p>
<h6 id="cell2cell.plotting.cci_plot.clustermap_cci--returns">Returns</h6>
<p>hier : seaborn.matrix.ClusterGrid
    A seaborn ClusterGrid instance.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/cci_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">clustermap_cci</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                   <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                   <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;CCI score&#39;</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Generates a clustermap (heatmap + dendrograms from a hierarchical</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    clustering) based on CCI scores of cell-cell pairs.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Interaction space that contains all a distance matrix after running the</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        the method compute_pairwise_cci_scores. Alternatively, this object</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        can be a numpy-array or a pandas DataFrame. Also, a</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        SingleCellInteractions or a BulkInteractions object after running</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        the method compute_pairwise_cci_scores.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    method : str, default=&#39;ward&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Clustering method for computing a linkage as in</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        scipy.cluster.hierarchy.linkage</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    optimal_leaf : boolean, default=True</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Whether sorting the leaf of the dendrograms to have a minimal distance</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        between successive leaves. For more information, see</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        scipy.cluster.hierarchy.optimal_leaf_ordering</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    metadata : pandas.Dataframe, default=None</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Metadata associated with the cells, cell types or samples in the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        matrix containing CCI scores. If None, cells will not be colored</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        by major groups.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    sample_col : str, default=&#39;#SampleID&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Column in the metadata for the cells, cell types or samples</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        in the matrix containing CCI scores.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    group_col : str, default=&#39;Groups&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Column in the metadata containing the major groups of cells, cell types</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        or samples in the matrix with CCI scores.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    meta_cmap : str, default=&#39;gist_rainbow&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Name of the color palette for coloring the major groups of cells.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    colors : dict, default=None</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        of major groups of cells. If colors is specified, meta_cmap will be</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    excluded_cells : list, default=None</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        List containing cell names that are present in the interaction_space</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        object but that will be excluded from this plot.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    title : str, default=&#39;&#39;</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Title of the clustermap.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    cbar_title : str, default=&#39;CCI score&#39;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Title for the colorbar, depending on the score employed.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    cbar_fontsize : int, default=18</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Font size for the colorbar title as well as labels for axes X and Y.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Dictionary containing arguments for the seaborn.clustermap function.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    -------</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    hier : seaborn.matrix.ClusterGrid</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        A seaborn ClusterGrid instance.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as an InteractionSpace class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">distance_matrix</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="n">space_type</span> <span class="o">=</span> <span class="s1">&#39;class&#39;</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">elif</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a distance matrix&#39;</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="n">space_type</span> <span class="o">=</span> <span class="s1">&#39;matrix&#39;</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_space&#39;</span><span class="p">):</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a Interactions class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_interactions() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                             <span class="s1">&#39; object to generate a distance matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>            <span class="n">interaction_space</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>            <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">distance_matrix</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>            <span class="n">space_type</span> <span class="o">=</span> <span class="s1">&#39;class&#39;</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_interactions() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>                         <span class="s1">&#39; object to generate a distance matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="c1"># Drop excluded cells</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">),</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                                 <span class="o">~</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>    <span class="c1"># Check symmetry to get linkage</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="n">symmetric</span> <span class="o">=</span> <span class="n">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="n">symmetric</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="s1">&#39;The distance matrix does not have the same elements in rows and columns&#39;</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="c1"># Obtain info for generating plot</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">linkage</span> <span class="o">=</span> <span class="n">_get_distance_matrix_linkages</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>                                            <span class="n">kwargs</span><span class="o">=</span><span class="n">kwargs</span><span class="p">,</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>                                            <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>                                            <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_leaf</span><span class="p">,</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>                                            <span class="n">symmetric</span><span class="o">=</span><span class="n">symmetric</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>                                            <span class="p">)</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="n">kwargs_</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="c1"># PLOT CCI MATRIX</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">if</span> <span class="n">space_type</span> <span class="o">==</span> <span class="s1">&#39;class&#39;</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">distance_matrix</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">),</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>                    <span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)]</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>    <span class="c1"># Colors</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="n">col_colors</span> <span class="o">=</span> <span class="n">map_colors_to_metadata</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>                                            <span class="n">ref_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>                                            <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">,</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>                                            <span class="n">sample_col</span><span class="o">=</span><span class="n">sample_col</span><span class="p">,</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>                                            <span class="n">group_col</span><span class="o">=</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>                                            <span class="n">cmap</span><span class="o">=</span><span class="n">meta_cmap</span><span class="p">)</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">symmetric</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>            <span class="n">row_colors</span> <span class="o">=</span> <span class="n">col_colors</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>            <span class="n">row_colors</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>        <span class="n">col_colors</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="n">row_colors</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>    <span class="c1"># Plot hierarchical clustering (triangular)</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>    <span class="n">hier</span> <span class="o">=</span> <span class="n">_plot_triangular_clustermap</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                                       <span class="n">symmetric</span><span class="o">=</span><span class="n">symmetric</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                                       <span class="n">linkage</span><span class="o">=</span><span class="n">linkage</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                                       <span class="n">col_colors</span><span class="o">=</span><span class="n">col_colors</span><span class="p">,</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                                       <span class="n">row_colors</span><span class="o">=</span><span class="n">row_colors</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                                       <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                                       <span class="n">cbar_title</span><span class="o">=</span><span class="n">cbar_title</span><span class="p">,</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                                       <span class="n">cbar_fontsize</span><span class="o">=</span><span class="n">cbar_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>                                       <span class="o">**</span><span class="n">kwargs_</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>    <span class="k">if</span> <span class="o">~</span><span class="n">symmetric</span><span class="p">:</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>        <span class="n">hier</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Receiver cells&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">cbar_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>        <span class="n">hier</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Sender cells&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">cbar_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>    <span class="k">return</span> <span class="n">hier</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.circular_plot">
        <code>circular_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.circos_plot">
<code class="highlight language-python"><span class="n">circos_plot</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">sender_cells</span><span class="p">,</span> <span class="n">receiver_cells</span><span class="p">,</span> <span class="n">ligands</span><span class="p">,</span> <span class="n">receptors</span><span class="p">,</span> <span class="n">excluded_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;Set2&#39;</span><span class="p">,</span> <span class="n">cells_cmap</span><span class="o">=</span><span class="s1">&#39;Pastel1&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ligand_label_color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">receptor_label_color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates the circos plot in the exact order that sender and
receiver cells are provided. Similarly, ligands and receptors are
sorted by the order they are input.</p>
<h6 id="cell2cell.plotting.circular_plot.circos_plot--parameters">Parameters</h6>
<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all a distance matrix after running the
    the method compute_pairwise_communication_scores. Alternatively, this
    object can a SingleCellInteractions or a BulkInteractions object after
    running the method compute_pairwise_communication_scores.</p>
<p>sender_cells : list
    List of cells to be included as senders.</p>
<p>receiver_cells : list
    List of cells to be included as receivers.</p>
<p>ligands : list
    List of genes/proteins to be included as ligands produced by the
    sender cells.</p>
<p>receptors : list
    List of genes/proteins to be included as receptors produced by the
    receiver cells.</p>
<p>excluded_score : float, default=0
    Rows that have a communication score equal or lower to this will
    be dropped from the network.</p>
<p>metadata : pandas.Dataframe, default=None
    Metadata associated with the cells, cell types or samples in the
    matrix containing CCC scores. If None, cells will be color only by
    individual cells.</p>
<p>sample_col : str, default='#SampleID'
    Column in the metadata for the cells, cell types or samples
    in the matrix containing CCC scores.</p>
<p>group_col : str, default='Groups'
    Column in the metadata containing the major groups of cells, cell types
    or samples in the matrix with CCC scores.</p>
<p>meta_cmap : str, default='Set2'
    Name of the matplotlib color palette for coloring the major groups
    of cells.</p>
<p>cells_cmap : str, default='Pastel1'
    Name of the color palette for coloring individual cells.</p>
<p>colors : dict, default=None
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. If colors is specified, meta_cmap will be
    ignored.</p>
<p>ax : matplotlib.axes.Axes, default=None
    Axes instance for a plot.</p>
<p>figsize : tuple, default=(10, 10)
    Size of the figure (width*height), each in inches.</p>
<p>fontsize : int, default=14
    Font size for ligand and receptor labels.</p>
<p>legend : boolean, default=True
    Whether including legends for cell and cell group colors as well
    as ligand/receptor colors.</p>
<p>ligand_label_color : str, default='dimgray'
    Name of the matplotlib color palette for coloring the labels of
    ligands.</p>
<p>receptor_label_color : str, default='dimgray'
    Name of the matplotlib color palette for coloring the labels of
    receptors.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.circular_plot.circos_plot--returns">Returns</h6>
<p>ax : matplotlib.axes.Axes
    Axes instance containing a circos plot.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">circos_plot</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">sender_cells</span><span class="p">,</span> <span class="n">receiver_cells</span><span class="p">,</span> <span class="n">ligands</span><span class="p">,</span> <span class="n">receptors</span><span class="p">,</span> <span class="n">excluded_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;Set2&#39;</span><span class="p">,</span> <span class="n">cells_cmap</span><span class="o">=</span><span class="s1">&#39;Pastel1&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ligand_label_color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">receptor_label_color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="sd">&#39;&#39;&#39;Generates the circos plot in the exact order that sender and</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    receiver cells are provided. Similarly, ligands and receptors are</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    sorted by the order they are input.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Interaction space that contains all a distance matrix after running the</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        the method compute_pairwise_communication_scores. Alternatively, this</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        object can a SingleCellInteractions or a BulkInteractions object after</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        running the method compute_pairwise_communication_scores.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    sender_cells : list</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        List of cells to be included as senders.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    receiver_cells : list</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        List of cells to be included as receivers.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    ligands : list</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        List of genes/proteins to be included as ligands produced by the</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        sender cells.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    receptors : list</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        List of genes/proteins to be included as receptors produced by the</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        receiver cells.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    excluded_score : float, default=0</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Rows that have a communication score equal or lower to this will</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        be dropped from the network.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    metadata : pandas.Dataframe, default=None</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Metadata associated with the cells, cell types or samples in the</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        matrix containing CCC scores. If None, cells will be color only by</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        individual cells.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    sample_col : str, default=&#39;#SampleID&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Column in the metadata for the cells, cell types or samples</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        in the matrix containing CCC scores.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    group_col : str, default=&#39;Groups&#39;</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Column in the metadata containing the major groups of cells, cell types</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        or samples in the matrix with CCC scores.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    meta_cmap : str, default=&#39;Set2&#39;</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Name of the matplotlib color palette for coloring the major groups</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        of cells.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    cells_cmap : str, default=&#39;Pastel1&#39;</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        Name of the color palette for coloring individual cells.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    colors : dict, default=None</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        of major groups of cells. If colors is specified, meta_cmap will be</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    ax : matplotlib.axes.Axes, default=None</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        Axes instance for a plot.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    figsize : tuple, default=(10, 10)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        Font size for ligand and receptor labels.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    legend : boolean, default=True</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        Whether including legends for cell and cell group colors as well</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        as ligand/receptor colors.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    ligand_label_color : str, default=&#39;dimgray&#39;</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        Name of the matplotlib color palette for coloring the labels of</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        ligands.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    receptor_label_color : str, default=&#39;dimgray&#39;</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Name of the matplotlib color palette for coloring the labels of</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        receptors.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    -------</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">    ax : matplotlib.axes.Axes</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        Axes instance containing a circos plot.</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_elements&#39;</span><span class="p">):</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_communication_scores() before generating circos plots.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>            <span class="n">readable_ccc</span> <span class="o">=</span> <span class="n">get_readable_ccc_matrix</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">])</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>    <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_space&#39;</span><span class="p">):</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="k">if</span> <span class="s1">&#39;communication_matrix&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Run the method compute_pairwise_communication_scores() before generating circos plots.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>            <span class="n">readable_ccc</span> <span class="o">=</span> <span class="n">get_readable_ccc_matrix</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">])</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not a valid interaction_space&#39;</span><span class="p">)</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="c1"># Figure setups</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>        <span class="n">R</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>        <span class="n">center</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">([</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">],</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="n">R</span> <span class="o">*</span> <span class="mf">1.05</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="p">(</span><span class="n">R</span> <span class="o">*</span> <span class="mf">1.05</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="o">-</span><span class="n">R</span> <span class="o">*</span> <span class="mf">1.05</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="p">(</span><span class="n">R</span> <span class="o">*</span> <span class="mf">1.05</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>        <span class="n">xlim</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">()</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">ylim</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="n">x_range</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">xlim</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">xlim</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="n">y_range</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">ylim</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">ylim</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>        <span class="n">R</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">([</span><span class="n">x_range</span><span class="o">/</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">y_range</span><span class="o">/</span><span class="mf">2.0</span><span class="p">])</span> <span class="o">/</span> <span class="mf">1.05</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>        <span class="n">center</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">xlim</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">ylim</span><span class="p">))</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>    <span class="n">current_ax</span> <span class="o">=</span> <span class="n">ax</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="c1"># Elements to build network</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>    <span class="c1"># TODO: Add option to select sort_by: None (as input), cells, proteins or metadata</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>    <span class="n">sorted_nodes</span> <span class="o">=</span> <span class="n">sort_nodes</span><span class="p">(</span><span class="n">sender_cells</span><span class="o">=</span><span class="n">sender_cells</span><span class="p">,</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>                              <span class="n">receiver_cells</span><span class="o">=</span><span class="n">receiver_cells</span><span class="p">,</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>                              <span class="n">ligands</span><span class="o">=</span><span class="n">ligands</span><span class="p">,</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>                              <span class="n">receptors</span><span class="o">=</span><span class="n">receptors</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>                              <span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>    <span class="c1"># Build network</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>    <span class="n">G</span> <span class="o">=</span> <span class="n">_build_network</span><span class="p">(</span><span class="n">sender_cells</span><span class="o">=</span><span class="n">sender_cells</span><span class="p">,</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>                       <span class="n">receiver_cells</span><span class="o">=</span><span class="n">receiver_cells</span><span class="p">,</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>                       <span class="n">ligands</span><span class="o">=</span><span class="n">ligands</span><span class="p">,</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                       <span class="n">receptors</span><span class="o">=</span><span class="n">receptors</span><span class="p">,</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>                       <span class="n">sorted_nodes</span><span class="o">=</span><span class="n">sorted_nodes</span><span class="p">,</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>                       <span class="n">readable_ccc</span><span class="o">=</span><span class="n">readable_ccc</span><span class="p">,</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>                       <span class="n">excluded_score</span><span class="o">=</span><span class="n">excluded_score</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>                       <span class="p">)</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>    <span class="c1"># Get coordinates</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>    <span class="n">nodes_dict</span> <span class="o">=</span> <span class="n">get_arc_angles</span><span class="p">(</span><span class="n">G</span><span class="o">=</span><span class="n">G</span><span class="p">,</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                                <span class="n">sorting_feature</span><span class="o">=</span><span class="s1">&#39;sorting&#39;</span><span class="p">)</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>    <span class="n">edges_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">nodes_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="n">edges_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_cartesian</span><span class="p">(</span><span class="n">theta</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">v</span><span class="p">),</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                                      <span class="n">radius</span><span class="o">=</span><span class="mf">0.95</span><span class="o">*</span><span class="n">R</span><span class="o">/</span><span class="mf">2.</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                                      <span class="n">center</span><span class="o">=</span><span class="n">center</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                                      <span class="n">angle</span><span class="o">=</span><span class="s1">&#39;degrees&#39;</span><span class="p">)</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>    <span class="n">small_R</span> <span class="o">=</span> <span class="n">determine_small_radius</span><span class="p">(</span><span class="n">edges_dict</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>    <span class="c1"># Colors</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="n">cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sender_cells</span><span class="o">+</span><span class="n">receiver_cells</span><span class="p">))</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="n">meta</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="n">meta</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[[</span><span class="n">group_col</span><span class="p">]]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;NA&#39;</span><span class="p">)</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>        <span class="n">labels</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>        <span class="k">if</span> <span class="n">colors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>            <span class="n">colors</span> <span class="o">=</span> <span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">meta_cmap</span><span class="p">)</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>        <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">colors</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">meta</span><span class="p">[</span><span class="n">group_col</span><span class="p">]]</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="n">meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">cells</span><span class="p">)</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>    <span class="c1"># Colors for cells, not major groups</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>    <span class="n">colors</span> <span class="o">=</span> <span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cells_cmap</span><span class="p">)</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>    <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Cells-Color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">colors</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>    <span class="c1"># signal_colors = {&#39;ligand&#39; : &#39;brown&#39;, &#39;receptor&#39; : &#39;black&#39;}</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>    <span class="n">signal_colors</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ligand&#39;</span> <span class="p">:</span> <span class="n">ligand_label_color</span><span class="p">,</span> <span class="s1">&#39;receptor&#39;</span> <span class="p">:</span> <span class="n">receptor_label_color</span><span class="p">}</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>    <span class="c1"># Draw edges</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>    <span class="c1"># TODO: Automatically determine lw given the size of the arcs (each ligand or receptor)</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>    <span class="n">lw</span> <span class="o">=</span> <span class="mi">10</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>    <span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>        <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">([</span><span class="n">edges_dict</span><span class="p">[</span><span class="n">l</span><span class="p">],</span> <span class="n">center</span><span class="p">,</span> <span class="n">edges_dict</span><span class="p">[</span><span class="n">r</span><span class="p">]],</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>                    <span class="p">[</span><span class="n">Path</span><span class="o">.</span><span class="n">MOVETO</span><span class="p">,</span> <span class="n">Path</span><span class="o">.</span><span class="n">CURVE3</span><span class="p">,</span> <span class="n">Path</span><span class="o">.</span><span class="n">CURVE3</span><span class="p">])</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="n">patch</span> <span class="o">=</span> <span class="n">patches</span><span class="o">.</span><span class="n">FancyArrowPatch</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>                                        <span class="n">arrowstyle</span><span class="o">=</span><span class="s2">&quot;-&gt;,head_length=</span><span class="si">{}</span><span class="s2">,head_width=</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">lw</span><span class="o">/</span><span class="mf">3.0</span><span class="p">,</span><span class="n">lw</span><span class="o">/</span><span class="mf">3.0</span><span class="p">),</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>                                        <span class="n">lw</span><span class="o">=</span><span class="n">lw</span><span class="o">/</span><span class="mf">2.</span><span class="p">,</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>                                        <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="c1">#meta.at[l.split(&#39;^&#39;)[0], &#39;Cells-Color&#39;],</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>                                        <span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                                        <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>                                        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.15</span><span class="p">)</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patch</span><span class="p">)</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>    <span class="c1"># Draw nodes</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>    <span class="c1"># TODO: Automatically determine lw given the size of the figure</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>    <span class="n">cell_legend</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>        <span class="n">meta_legend</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>        <span class="n">meta_legend</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">nodes_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>        <span class="n">diff</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="o">*</span> <span class="nb">abs</span><span class="p">(</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1"># Distance to substract and avoid nodes touching each others</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>        <span class="n">cell</span> <span class="o">=</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;^&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>        <span class="n">cell_color</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">cell</span><span class="p">,</span> <span class="s1">&#39;Cells-Color&#39;</span><span class="p">]</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>        <span class="n">cell_legend</span><span class="p">[</span><span class="n">cell</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_color</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">Arc</span><span class="p">(</span><span class="n">center</span><span class="p">,</span> <span class="n">R</span><span class="p">,</span> <span class="n">R</span><span class="p">,</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>                         <span class="n">theta1</span><span class="o">=</span><span class="n">diff</span> <span class="o">+</span> <span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>                         <span class="n">theta2</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">diff</span><span class="p">,</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>                         <span class="n">edgecolor</span><span class="o">=</span><span class="n">cell_color</span><span class="p">,</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>                         <span class="n">lw</span><span class="o">=</span><span class="n">lw</span><span class="p">))</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>        <span class="n">coeff</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>        <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a>            <span class="n">coeff</span> <span class="o">=</span> <span class="mf">1.1</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a>            <span class="n">meta_color</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">cell</span><span class="p">,</span> <span class="s1">&#39;Color&#39;</span><span class="p">]</span>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">Arc</span><span class="p">(</span><span class="n">center</span><span class="p">,</span> <span class="n">coeff</span> <span class="o">*</span> <span class="n">R</span><span class="p">,</span> <span class="n">coeff</span> <span class="o">*</span> <span class="n">R</span><span class="p">,</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a>                             <span class="n">theta1</span><span class="o">=</span><span class="n">diff</span> <span class="o">+</span> <span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a>                             <span class="n">theta2</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">diff</span><span class="p">,</span>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a>                             <span class="n">edgecolor</span><span class="o">=</span><span class="n">meta_color</span><span class="p">,</span>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a>                             <span class="n">lw</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a>            <span class="n">meta_legend</span><span class="p">[</span><span class="n">meta</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">cell</span><span class="p">,</span> <span class="n">group_col</span><span class="p">]]</span> <span class="o">=</span> <span class="n">meta_color</span>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a>        <span class="n">label_coord</span> <span class="o">=</span> <span class="n">get_cartesian</span><span class="p">(</span><span class="n">theta</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">v</span><span class="p">),</span>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a>                                    <span class="n">radius</span><span class="o">=</span><span class="n">coeff</span> <span class="o">*</span> <span class="n">R</span><span class="o">/</span><span class="mf">2.</span> <span class="o">+</span> <span class="nb">min</span><span class="p">([</span><span class="n">small_R</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span> <span class="n">coeff</span> <span class="o">*</span> <span class="n">R</span><span class="o">*</span><span class="mf">0.05</span><span class="p">]),</span>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a>                                    <span class="n">center</span><span class="o">=</span><span class="n">center</span><span class="p">,</span>
<a id="__codelineno-0-221" name="__codelineno-0-221" href="#__codelineno-0-221"></a>                                    <span class="n">angle</span><span class="o">=</span><span class="s1">&#39;degrees&#39;</span><span class="p">)</span>
<a id="__codelineno-0-222" name="__codelineno-0-222" href="#__codelineno-0-222"></a>
<a id="__codelineno-0-223" name="__codelineno-0-223" href="#__codelineno-0-223"></a>        <span class="c1"># Add Labels</span>
<a id="__codelineno-0-224" name="__codelineno-0-224" href="#__codelineno-0-224"></a>        <span class="n">v2</span> <span class="o">=</span> <span class="n">label_coord</span>
<a id="__codelineno-0-225" name="__codelineno-0-225" href="#__codelineno-0-225"></a>        <span class="n">x</span> <span class="o">=</span> <span class="n">v2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-226" name="__codelineno-0-226" href="#__codelineno-0-226"></a>        <span class="n">y</span> <span class="o">=</span> <span class="n">v2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-227" name="__codelineno-0-227" href="#__codelineno-0-227"></a>        <span class="n">rotation</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
<a id="__codelineno-0-228" name="__codelineno-0-228" href="#__codelineno-0-228"></a>
<a id="__codelineno-0-229" name="__codelineno-0-229" href="#__codelineno-0-229"></a>        <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-230" name="__codelineno-0-230" href="#__codelineno-0-230"></a>            <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span>
<a id="__codelineno-0-231" name="__codelineno-0-231" href="#__codelineno-0-231"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-232" name="__codelineno-0-232" href="#__codelineno-0-232"></a>            <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;right&#39;</span>
<a id="__codelineno-0-233" name="__codelineno-0-233" href="#__codelineno-0-233"></a>        <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span>
<a id="__codelineno-0-234" name="__codelineno-0-234" href="#__codelineno-0-234"></a>
<a id="__codelineno-0-235" name="__codelineno-0-235" href="#__codelineno-0-235"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">rotation</span> <span class="o">&lt;=</span> <span class="mi">90</span><span class="p">):</span>
<a id="__codelineno-0-236" name="__codelineno-0-236" href="#__codelineno-0-236"></a>            <span class="n">rotation</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">rotation</span><span class="p">)</span>
<a id="__codelineno-0-237" name="__codelineno-0-237" href="#__codelineno-0-237"></a>        <span class="k">elif</span> <span class="p">(</span><span class="n">rotation</span> <span class="o">&lt;=</span> <span class="mi">180</span><span class="p">):</span>
<a id="__codelineno-0-238" name="__codelineno-0-238" href="#__codelineno-0-238"></a>            <span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="nb">abs</span><span class="p">(</span><span class="n">rotation</span> <span class="o">-</span> <span class="mi">180</span><span class="p">)</span>
<a id="__codelineno-0-239" name="__codelineno-0-239" href="#__codelineno-0-239"></a>        <span class="k">elif</span> <span class="p">(</span><span class="n">rotation</span> <span class="o">&lt;=</span> <span class="mi">270</span><span class="p">):</span>
<a id="__codelineno-0-240" name="__codelineno-0-240" href="#__codelineno-0-240"></a>            <span class="n">rotation</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">rotation</span> <span class="o">-</span> <span class="mi">180</span><span class="p">)</span>
<a id="__codelineno-0-241" name="__codelineno-0-241" href="#__codelineno-0-241"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-242" name="__codelineno-0-242" href="#__codelineno-0-242"></a>            <span class="n">rotation</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">rotation</span><span class="p">)</span>
<a id="__codelineno-0-243" name="__codelineno-0-243" href="#__codelineno-0-243"></a>
<a id="__codelineno-0-244" name="__codelineno-0-244" href="#__codelineno-0-244"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span>
<a id="__codelineno-0-245" name="__codelineno-0-245" href="#__codelineno-0-245"></a>                <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;^&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
<a id="__codelineno-0-246" name="__codelineno-0-246" href="#__codelineno-0-246"></a>                <span class="n">rotation</span><span class="o">=</span><span class="n">rotation</span><span class="p">,</span>
<a id="__codelineno-0-247" name="__codelineno-0-247" href="#__codelineno-0-247"></a>                <span class="n">rotation_mode</span><span class="o">=</span><span class="s2">&quot;anchor&quot;</span><span class="p">,</span>
<a id="__codelineno-0-248" name="__codelineno-0-248" href="#__codelineno-0-248"></a>                <span class="n">horizontalalignment</span><span class="o">=</span><span class="n">ha</span><span class="p">,</span>
<a id="__codelineno-0-249" name="__codelineno-0-249" href="#__codelineno-0-249"></a>                <span class="n">verticalalignment</span><span class="o">=</span><span class="n">va</span><span class="p">,</span>
<a id="__codelineno-0-250" name="__codelineno-0-250" href="#__codelineno-0-250"></a>                <span class="n">color</span><span class="o">=</span><span class="n">signal_colors</span><span class="p">[</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;signal&#39;</span><span class="p">]],</span>
<a id="__codelineno-0-251" name="__codelineno-0-251" href="#__codelineno-0-251"></a>                <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span>
<a id="__codelineno-0-252" name="__codelineno-0-252" href="#__codelineno-0-252"></a>                <span class="p">)</span>
<a id="__codelineno-0-253" name="__codelineno-0-253" href="#__codelineno-0-253"></a>
<a id="__codelineno-0-254" name="__codelineno-0-254" href="#__codelineno-0-254"></a>    <span class="c1"># Draw legend</span>
<a id="__codelineno-0-255" name="__codelineno-0-255" href="#__codelineno-0-255"></a>    <span class="k">if</span> <span class="n">legend</span><span class="p">:</span>
<a id="__codelineno-0-256" name="__codelineno-0-256" href="#__codelineno-0-256"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
<a id="__codelineno-0-257" name="__codelineno-0-257" href="#__codelineno-0-257"></a>        <span class="n">lgd</span> <span class="o">=</span> <span class="n">generate_circos_legend</span><span class="p">(</span><span class="n">cell_legend</span><span class="o">=</span><span class="n">cell_legend</span><span class="p">,</span>
<a id="__codelineno-0-258" name="__codelineno-0-258" href="#__codelineno-0-258"></a>                                     <span class="n">meta_legend</span><span class="o">=</span><span class="n">meta_legend</span><span class="p">,</span>
<a id="__codelineno-0-259" name="__codelineno-0-259" href="#__codelineno-0-259"></a>                                     <span class="n">signal_legend</span><span class="o">=</span><span class="n">signal_colors</span><span class="p">,</span>
<a id="__codelineno-0-260" name="__codelineno-0-260" href="#__codelineno-0-260"></a>                                     <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-261" name="__codelineno-0-261" href="#__codelineno-0-261"></a>                                     <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-262" name="__codelineno-0-262" href="#__codelineno-0-262"></a>                                     <span class="p">)</span>
<a id="__codelineno-0-263" name="__codelineno-0-263" href="#__codelineno-0-263"></a>
<a id="__codelineno-0-264" name="__codelineno-0-264" href="#__codelineno-0-264"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-265" name="__codelineno-0-265" href="#__codelineno-0-265"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-266" name="__codelineno-0-266" href="#__codelineno-0-266"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-267" name="__codelineno-0-267" href="#__codelineno-0-267"></a>    <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.determine_small_radius">
<code class="highlight language-python"><span class="n">determine_small_radius</span><span class="p">(</span><span class="n">coordinate_dict</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the radius of a circle whose diameter is the distance
between the center of two nodes.</p>
<h6 id="cell2cell.plotting.circular_plot.determine_small_radius--parameters">Parameters</h6>
<p>coordinate_dict : dict
    A dictionary containing the coordinates to plot each node.</p>
<h6 id="cell2cell.plotting.circular_plot.determine_small_radius--returns">Returns</h6>
<p>radius : float
    The half of the distance between the center of two nodes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">determine_small_radius</span><span class="p">(</span><span class="n">coordinate_dict</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the radius of a circle whose diameter is the distance</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    between the center of two nodes.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    coordinate_dict : dict</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dictionary containing the coordinates to plot each node.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    -------</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    radius : float</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        The half of the distance between the center of two nodes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>    <span class="c1"># Cartesian coordinates</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">coordinate_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="n">circle1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">coordinate_dict</span><span class="p">[</span><span class="n">keys</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">circle2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">coordinate_dict</span><span class="p">[</span><span class="n">keys</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">diff</span> <span class="o">=</span> <span class="n">circle1</span> <span class="o">-</span> <span class="n">circle2</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">diff</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="n">diff</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">radius</span> <span class="o">=</span> <span class="n">distance</span> <span class="o">/</span> <span class="mf">2.0</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">radius</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.generate_circos_legend">
<code class="highlight language-python"><span class="n">generate_circos_legend</span><span class="p">(</span><span class="n">cell_legend</span><span class="p">,</span> <span class="n">signal_legend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">meta_legend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Adds legends to circos plot.</p>
<h6 id="cell2cell.plotting.circular_plot.generate_circos_legend--parameters">Parameters</h6>
<p>cell_legend : dict
    Dictionary containing the colors for the cells.</p>
<p>signal_legend : dict, default=None
    Dictionary containing the colors for the LR pairs in a given pair
    of cells. Corresponds to the colors of the links.</p>
<p>meta_legend : dict, default=None
    Dictionary containing the colors for the cells given their major
    groups.</p>
<p>fontsize : int, default=14
    Size of the labels in the legend.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_circos_legend</span><span class="p">(</span><span class="n">cell_legend</span><span class="p">,</span> <span class="n">signal_legend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">meta_legend</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Adds legends to circos plot.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    cell_legend : dict</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Dictionary containing the colors for the cells.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    signal_legend : dict, default=None</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Dictionary containing the colors for the LR pairs in a given pair</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        of cells. Corresponds to the colors of the links.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    meta_legend : dict, default=None</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Dictionary containing the colors for the cells given their major</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        groups.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Size of the labels in the legend.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">legend_fontsize</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fontsize</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="c1"># Cell legend</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">lgd</span> <span class="o">=</span> <span class="n">generate_legend</span><span class="p">(</span><span class="n">color_dict</span><span class="o">=</span><span class="n">cell_legend</span><span class="p">,</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>                          <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>                          <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.01</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>                          <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>                          <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>                          <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>                          <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Cells&#39;</span><span class="p">,</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>                          <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>                          <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>                          <span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">signal_legend</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="c1"># Signal legend</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="c1"># generate_legend(color_dict=signal_legend,</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="c1">#                 loc=&#39;upper center&#39;,</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="c1">#                 bbox_to_anchor=(0.5, -0.01),</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="c1">#                 ncol=2,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="c1">#                 fancybox=True,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="c1">#                 shadow=True,</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="c1">#                 title=&#39;Signals&#39;,</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="c1">#                 fontsize=legend_fontsize</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="c1">#                 )</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="k">pass</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="k">if</span> <span class="n">meta_legend</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>        <span class="c1">#ax.add_artist(lgd)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">([</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">],</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">ax2</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="c1"># Meta legend</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="n">lgd3</span> <span class="o">=</span> <span class="n">generate_legend</span><span class="p">(</span><span class="n">color_dict</span><span class="o">=</span><span class="n">meta_legend</span><span class="p">,</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>                               <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center right&#39;</span><span class="p">,</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>                               <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>                               <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>                               <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>                               <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                               <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                               <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                               <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                               <span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="k">return</span> <span class="n">lgd3</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="k">return</span> <span class="n">lgd</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.get_arc_angles">
<code class="highlight language-python"><span class="n">get_arc_angles</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">sorting_feature</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains the angles of polar coordinates to plot nodes as arcs of
 a circumference.</p>
<h6 id="cell2cell.plotting.circular_plot.get_arc_angles--parameters">Parameters</h6>
<p>G : networkx.Graph or networkx.DiGraph
    A networkx graph.</p>
<p>sorting_feature : str, default=None
    A node attribute present in the dictionary associated with
    each node. The values associated with this attributed will
    be used for sorting the nodes.</p>
<h6 id="cell2cell.plotting.circular_plot.get_arc_angles--returns">Returns</h6>
<p>angles : dict
    A dictionary containing the angles for positioning the
    nodes in polar coordinates. Keys are the node names and
    values are tuples with angles for the start and end of
    the arc that represents a node.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_arc_angles</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">sorting_feature</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Obtains the angles of polar coordinates to plot nodes as arcs of</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">     a circumference.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    G : networkx.Graph or networkx.DiGraph</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A networkx graph.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    sorting_feature : str, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A node attribute present in the dictionary associated with</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        each node. The values associated with this attributed will</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        be used for sorting the nodes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    -------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    angles : dict</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        A dictionary containing the angles for positioning the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        nodes in polar coordinates. Keys are the node names and</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        values are tuples with angles for the start and end of</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        the arc that represents a node.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">elements</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()))</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">n_elements</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">elements</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="n">sorting_feature</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">sorting_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="n">sorting_feature</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">elements</span><span class="p">]</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">elements</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">sorting_list</span><span class="p">,</span> <span class="n">elements</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">pair</span><span class="p">:</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">angles</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">elements</span><span class="p">):</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="n">theta1</span> <span class="o">=</span> <span class="p">(</span><span class="mi">360</span> <span class="o">/</span> <span class="n">n_elements</span><span class="p">)</span> <span class="o">*</span> <span class="n">i</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="n">theta2</span> <span class="o">=</span> <span class="p">(</span><span class="mi">360</span> <span class="o">/</span> <span class="n">n_elements</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">angles</span><span class="p">[</span><span class="n">node</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">theta1</span><span class="p">,</span> <span class="n">theta2</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">return</span> <span class="n">angles</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.get_cartesian">
<code class="highlight language-python"><span class="n">get_cartesian</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">angle</span><span class="o">=</span><span class="s1">&#39;radians&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Performs a polar to cartesian coordinates conversion.</p>
<h6 id="cell2cell.plotting.circular_plot.get_cartesian--parameters">Parameters</h6>
<p>theta : float or ndarray
    An angle for a polar coordinate.</p>
<p>radius : float
    The radius in a polar coordinate.</p>
<p>center : tuple, default=(0,0)
    The center of the circle in the cartesian coordinates.</p>
<p>angle : str, default='radians'
    Type of angle that theta is. Options are:
     - 'degrees' : from 0 to 360
     - 'radians' : from 0 to 2*numpy.pi</p>
<h6 id="cell2cell.plotting.circular_plot.get_cartesian--returns">Returns</h6>
<p>(x, y) : tuple
    Cartesian coordinates for X and Y axis respective.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_cartesian</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">angle</span><span class="o">=</span><span class="s1">&#39;radians&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Performs a polar to cartesian coordinates conversion.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    theta : float or ndarray</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        An angle for a polar coordinate.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    radius : float</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        The radius in a polar coordinate.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    center : tuple, default=(0,0)</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        The center of the circle in the cartesian coordinates.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    angle : str, default=&#39;radians&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Type of angle that theta is. Options are:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">         - &#39;degrees&#39; : from 0 to 360</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">         - &#39;radians&#39; : from 0 to 2*numpy.pi</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    (x, y) : tuple</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Cartesian coordinates for X and Y axis respective.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">angle</span> <span class="o">==</span> <span class="s1">&#39;degrees&#39;</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">theta_</span> <span class="o">=</span> <span class="mf">2.0</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="o">*</span><span class="n">theta</span><span class="o">/</span><span class="mf">360.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">elif</span> <span class="n">angle</span> <span class="o">==</span> <span class="s1">&#39;radians&#39;</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">theta_</span> <span class="o">=</span> <span class="n">theta</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not a valid angle. Use randians or degrees.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">x</span> <span class="o">=</span> <span class="n">radius</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">theta_</span><span class="p">)</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">y</span> <span class="o">=</span> <span class="n">radius</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">theta_</span><span class="p">)</span> <span class="o">+</span> <span class="n">center</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.get_node_colors">
<code class="highlight language-python"><span class="n">get_node_colors</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">coloring_feature</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates colors for each node in a network given one of their
properties.</p>
<h6 id="cell2cell.plotting.circular_plot.get_node_colors--parameters">Parameters</h6>
<p>G : networkx.Graph
    A graph containing a list of nodes</p>
<p>coloring_feature : str, default=None
    A node attribute present in the dictionary associated with
    each node. The values associated with this attributed will
    be used for coloring the nodes.</p>
<p>cmap : str, default='viridis'
    Name of a matplotlib color palette for coloring the nodes.</p>
<h6 id="cell2cell.plotting.circular_plot.get_node_colors--returns">Returns</h6>
<p>node_colors : dict
    A dictionary wherein each key is a node and values are
    tuples containing colors in the RGBA format.</p>
<p>feature_colores : dict
    A dictionary wherein each key is a value for the attribute
    of nodes in the coloring_feature property and values are
    tuples containing colors in the RGBA format.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_node_colors</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">coloring_feature</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates colors for each node in a network given one of their</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    properties.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    G : networkx.Graph</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A graph containing a list of nodes</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    coloring_feature : str, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A node attribute present in the dictionary associated with</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        each node. The values associated with this attributed will</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        be used for coloring the nodes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    cmap : str, default=&#39;viridis&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Name of a matplotlib color palette for coloring the nodes.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    node_colors : dict</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        A dictionary wherein each key is a node and values are</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        tuples containing colors in the RGBA format.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    feature_colores : dict</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        A dictionary wherein each key is a value for the attribute</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        of nodes in the coloring_feature property and values are</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        tuples containing colors in the RGBA format.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="n">coloring_feature</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Node feature not specified!&#39;</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="c1"># Get what features we have in the network G</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">features</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">features</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="n">coloring_feature</span><span class="p">])</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="c1"># Generate colors for each feature</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">NUM_COLORS</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">cm</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">feature_colors</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">feature_colors</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">cm</span><span class="p">(</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">i</span> <span class="o">/</span> <span class="n">NUM_COLORS</span><span class="p">)</span>  <span class="c1"># color will now be an RGBA tuple</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="c1"># Map feature colors into nodes</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">node_colors</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">feature</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">n</span><span class="p">][</span><span class="n">coloring_feature</span><span class="p">]</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">node_colors</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature_colors</span><span class="p">[</span><span class="n">feature</span><span class="p">]</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">return</span> <span class="n">node_colors</span><span class="p">,</span> <span class="n">feature_colors</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.get_readable_ccc_matrix">
<code class="highlight language-python"><span class="n">get_readable_ccc_matrix</span><span class="p">(</span><span class="n">ccc_matrix</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Transforms a CCC matrix from an InteractionSpace instance
into a readable dataframe.</p>
<h6 id="cell2cell.plotting.circular_plot.get_readable_ccc_matrix--parameters">Parameters</h6>
<p>ccc_matrix : pandas.DataFrame
    A dataframe containing the communication scores for a given combination
    between a pair of sender-receiver cells and a ligand-receptor pair.
    Columns are pairs of cells and rows LR pairs.</p>
<h6 id="cell2cell.plotting.circular_plot.get_readable_ccc_matrix--returns">Returns</h6>
<p>readable_ccc : pandas.DataFrame
    A dataframe containing flat information in each row about communication
    scores for a given pair of cells and a specific LR pair. A row contains
    the sender and receiver cells as well as the ligand and the receptor
    participating in an interaction and their respective communication
    score. Columns are: ['sender', 'receiver', 'ligand', 'receptor',
    'communication_score']</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_readable_ccc_matrix</span><span class="p">(</span><span class="n">ccc_matrix</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Transforms a CCC matrix from an InteractionSpace instance</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    into a readable dataframe.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ccc_matrix : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe containing the communication scores for a given combination</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        between a pair of sender-receiver cells and a ligand-receptor pair.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Columns are pairs of cells and rows LR pairs.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    readable_ccc : pandas.DataFrame</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        A dataframe containing flat information in each row about communication</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        scores for a given pair of cells and a specific LR pair. A row contains</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        the sender and receiver cells as well as the ligand and the receptor</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        participating in an interaction and their respective communication</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        score. Columns are: [&#39;sender&#39;, &#39;receiver&#39;, &#39;ligand&#39;, &#39;receptor&#39;,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        &#39;communication_score&#39;]</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">readable_ccc</span> <span class="o">=</span> <span class="n">ccc_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">readable_ccc</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;LR&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">readable_ccc</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">readable_ccc</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;LR&#39;</span><span class="p">])</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;sender&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;variable&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;receiver&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;variable&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;ligand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;LR&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;receptor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="p">[</span><span class="s1">&#39;LR&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">readable_ccc</span> <span class="o">=</span> <span class="n">readable_ccc</span><span class="p">[[</span><span class="s1">&#39;sender&#39;</span><span class="p">,</span> <span class="s1">&#39;receiver&#39;</span><span class="p">,</span> <span class="s1">&#39;ligand&#39;</span><span class="p">,</span> <span class="s1">&#39;receptor&#39;</span><span class="p">,</span> <span class="s1">&#39;value&#39;</span><span class="p">]]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">readable_ccc</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sender&#39;</span><span class="p">,</span> <span class="s1">&#39;receiver&#39;</span><span class="p">,</span> <span class="s1">&#39;ligand&#39;</span><span class="p">,</span> <span class="s1">&#39;receptor&#39;</span><span class="p">,</span> <span class="s1">&#39;communication_score&#39;</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">readable_ccc</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.circular_plot.sort_nodes">
<code class="highlight language-python"><span class="n">sort_nodes</span><span class="p">(</span><span class="n">sender_cells</span><span class="p">,</span> <span class="n">receiver_cells</span><span class="p">,</span> <span class="n">ligands</span><span class="p">,</span> <span class="n">receptors</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Sorts cells by senders first and alphabetically and creates pairs of
senders-ligands. If senders and receivers share cells, it creates pairs
of senders-receptors for those shared cells. Then sorts receivers cells
and creates pairs of receivers-receptors, for those cells that are not
shared with senders.</p>
<h6 id="cell2cell.plotting.circular_plot.sort_nodes--parameters">Parameters</h6>
<p>sender_cells : list
    List of sender cells to sort.</p>
<p>receiver_cells : list
    List of receiver cells to sort.</p>
<p>ligands : list
    List of ligands to sort.</p>
<p>receptors : list
    List of receptors to sort.</p>
<h6 id="cell2cell.plotting.circular_plot.sort_nodes--returns">Returns</h6>
<p>sorted_nodes : dict
    A dictionary where keys are the nodes of cells-proteins and values
    are the position they obtained (a ranking from 0 to N, where N is
    the total number of nodes).</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/circular_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">sort_nodes</span><span class="p">(</span><span class="n">sender_cells</span><span class="p">,</span> <span class="n">receiver_cells</span><span class="p">,</span> <span class="n">ligands</span><span class="p">,</span> <span class="n">receptors</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Sorts cells by senders first and alphabetically and creates pairs of</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    senders-ligands. If senders and receivers share cells, it creates pairs</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    of senders-receptors for those shared cells. Then sorts receivers cells</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    and creates pairs of receivers-receptors, for those cells that are not</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    shared with senders.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    sender_cells : list</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List of sender cells to sort.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    receiver_cells : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        List of receiver cells to sort.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ligands : list</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List of ligands to sort.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    receptors : list</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        List of receptors to sort.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    sorted_nodes : dict</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        A dictionary where keys are the nodes of cells-proteins and values</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        are the position they obtained (a ranking from 0 to N, where N is</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        the total number of nodes).</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">sorted_nodes</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">both</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sender_cells</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">receiver_cells</span><span class="p">)</span> <span class="c1"># Intersection</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">sender_cells</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ligands</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>            <span class="n">sorted_nodes</span><span class="p">[(</span><span class="n">c</span> <span class="o">+</span> <span class="s1">&#39;^&#39;</span> <span class="o">+</span> <span class="n">p</span><span class="p">)]</span> <span class="o">=</span> <span class="n">count</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">both</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">receptors</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>                <span class="n">sorted_nodes</span><span class="p">[(</span><span class="n">c</span> <span class="o">+</span> <span class="s1">&#39;^&#39;</span> <span class="o">+</span> <span class="n">p</span><span class="p">)]</span> <span class="o">=</span> <span class="n">count</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">receiver_cells</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">both</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">receptors</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                <span class="n">sorted_nodes</span><span class="p">[(</span><span class="n">c</span> <span class="o">+</span> <span class="s1">&#39;^&#39;</span> <span class="o">+</span> <span class="n">p</span><span class="p">)]</span> <span class="o">=</span> <span class="n">count</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="k">return</span> <span class="n">sorted_nodes</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.factor_plot">
        <code>factor_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.factor_plot.ccc_networks_plot">
<code class="highlight language-python"><span class="n">ccc_networks_plot</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">included_factors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">,</span> <span class="n">ccc_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">network_layout</span><span class="o">=</span><span class="s1">&#39;spring&#39;</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="s1">&#39;magenta&#39;</span><span class="p">,</span> <span class="n">edge_width</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">edge_arrow_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">edge_alpha</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="s1">&#39;#210070&#39;</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">node_alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">node_label_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">node_label_alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">node_label_offset</span><span class="o">=</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">),</span> <span class="n">factor_title_size</span><span class="o">=</span><span class="mi">36</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots factor-specific cell-cell communication networks
resulting from decomposition with Tensor-cell2cell.</p>
<h6 id="cell2cell.plotting.factor_plot.ccc_networks_plot--parameters">Parameters</h6>
<p>factors : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor.</p>
<p>included_factors : list, default=None
    Factors to be included. Factor names must be the same as the key values in
    the factors dictionary.</p>
<p>sender_label : str
    Label for the dimension of sender cells. It is one key of the factors dict.</p>
<p>receiver_label : str
    Label for the dimension of receiver cells. It is one key of the factors dict.</p>
<p>ccc_threshold : float, default=None
    Threshold to consider only edges with a higher weight than this value.</p>
<p>panel_size : tuple, default=(8, 8)
    Size of one subplot or network (width*height), each in inches.</p>
<p>nrows : int, default=2
    Number of rows in the set of subplots.</p>
<p>network_layout : str, default='spring'
    Visualization layout of the networks. It uses algorithms implemented
    in NetworkX, including:
        -'spring' : Fruchterman-Reingold force-directed algorithm.
        -'circular' : Position nodes on a circle.</p>
<p>edge_color : str, default='magenta'
    Color of the edges in the network.</p>
<p>edge_width : int, default=25
    Thickness of the edges in the network.</p>
<p>edge_arrow_size : int, default=20
    Size of the arrow of an edge pointing towards the receiver cells.</p>
<p>edge_alpha : float, default=0.25
    Transparency of the edges. Values must be between 0 and 1. Higher
    values indicates less transparency.</p>
<p>node_color : str, default="#210070"
    Color of the nodes in the network.</p>
<p>node_size : int, default=1000
    Size of the nodes in the network.</p>
<p>node_alpha : float, default=0.9
    Transparency of the nodes. Values must be between 0 and 1. Higher
    values indicates less transparency.</p>
<p>node_label_size : int, default=20
    Size of the labels for the node names.</p>
<p>node_label_alpha : int, default=0.7
    Transparency of the node labeks. Values must be between 0 and 1.
    Higher values indicates less transparency.</p>
<p>node_label_offset : tuple, default=(0.1, -0.2)
    Offset values to move the node labels away from the center of the nodes.</p>
<p>factor_title_size : int, default=36
    Size of the subplot titles. Each network has a title like 'Factor 1',
    'Factor 2', ... ,'Factor R'.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.factor_plot.ccc_networks_plot--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    A matplotlib figure.</p>
<p>axes : matplotlib.axes.Axes or array of Axes
    Matplotlib axes representing the subplots containing the networks.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/factor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">ccc_networks_plot</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">included_factors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sender_label</span><span class="o">=</span><span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="n">receiver_label</span><span class="o">=</span><span class="s1">&#39;Receiver Cells&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                      <span class="n">ccc_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">network_layout</span><span class="o">=</span><span class="s1">&#39;spring&#39;</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="s1">&#39;magenta&#39;</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                      <span class="n">edge_width</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">edge_arrow_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">edge_alpha</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="s2">&quot;#210070&quot;</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                      <span class="n">node_alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">node_label_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">node_label_alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">node_label_offset</span><span class="o">=</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">),</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>                      <span class="n">factor_title_size</span><span class="o">=</span><span class="mi">36</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    <span class="sd">&#39;&#39;&#39;Plots factor-specific cell-cell communication networks</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    resulting from decomposition with Tensor-cell2cell.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    factors : dict</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        dimension/order of the tensor.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    included_factors : list, default=None</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Factors to be included. Factor names must be the same as the key values in</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        the factors dictionary.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    sender_label : str</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Label for the dimension of sender cells. It is one key of the factors dict.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    receiver_label : str</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Label for the dimension of receiver cells. It is one key of the factors dict.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    ccc_threshold : float, default=None</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Threshold to consider only edges with a higher weight than this value.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    panel_size : tuple, default=(8, 8)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Size of one subplot or network (width*height), each in inches.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    nrows : int, default=2</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Number of rows in the set of subplots.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    network_layout : str, default=&#39;spring&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Visualization layout of the networks. It uses algorithms implemented</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        in NetworkX, including:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">            -&#39;spring&#39; : Fruchterman-Reingold force-directed algorithm.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">            -&#39;circular&#39; : Position nodes on a circle.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    edge_color : str, default=&#39;magenta&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Color of the edges in the network.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    edge_width : int, default=25</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Thickness of the edges in the network.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    edge_arrow_size : int, default=20</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Size of the arrow of an edge pointing towards the receiver cells.</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    edge_alpha : float, default=0.25</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        Transparency of the edges. Values must be between 0 and 1. Higher</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        values indicates less transparency.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    node_color : str, default=&quot;#210070&quot;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Color of the nodes in the network.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    node_size : int, default=1000</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Size of the nodes in the network.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    node_alpha : float, default=0.9</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Transparency of the nodes. Values must be between 0 and 1. Higher</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        values indicates less transparency.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    node_label_size : int, default=20</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Size of the labels for the node names.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    node_label_alpha : int, default=0.7</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        Transparency of the node labeks. Values must be between 0 and 1.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Higher values indicates less transparency.</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">    node_label_offset : tuple, default=(0.1, -0.2)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        Offset values to move the node labels away from the center of the nodes.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    factor_title_size : int, default=36</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        Size of the subplot titles. Each network has a title like &#39;Factor 1&#39;,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        &#39;Factor 2&#39;, ... ,&#39;Factor R&#39;.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">    -------</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        A matplotlib figure.</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    axes : matplotlib.axes.Axes or array of Axes</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        Matplotlib axes representing the subplots containing the networks.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">networks</span> <span class="o">=</span> <span class="n">get_factor_specific_ccc_networks</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="n">factors</span><span class="p">,</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>                                                <span class="n">sender_label</span><span class="o">=</span><span class="n">sender_label</span><span class="p">,</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>                                                <span class="n">receiver_label</span><span class="o">=</span><span class="n">receiver_label</span><span class="p">)</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">if</span> <span class="n">included_factors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="n">factor_labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Factor </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">networks</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="n">factor_labels</span> <span class="o">=</span> <span class="n">included_factors</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="n">nrows</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">),</span> <span class="n">nrows</span><span class="p">])</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>    <span class="n">ncols</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">)</span> <span class="o">/</span> <span class="n">nrows</span><span class="p">))</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">panel_size</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">panel_size</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">nrows</span><span class="p">))</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="n">axs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">axes</span><span class="p">])</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>        <span class="n">axs</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">factor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">):</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>        <span class="k">if</span> <span class="n">ccc_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>            <span class="c1"># Considers edges with weight above ccc_threshold</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>            <span class="n">df</span> <span class="o">=</span> <span class="n">networks</span><span class="p">[</span><span class="n">factor</span><span class="p">]</span><span class="o">.</span><span class="n">gt</span><span class="p">(</span><span class="n">ccc_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">networks</span><span class="p">[</span><span class="n">factor</span><span class="p">])</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>            <span class="n">df</span> <span class="o">=</span> <span class="n">networks</span><span class="p">[</span><span class="n">factor</span><span class="p">]</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>        <span class="c1"># Networkx Directed Network</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">convert_matrix</span><span class="o">.</span><span class="n">from_pandas_adjacency</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">create_using</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">())</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="c1"># Layout for visualization - Node positions</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="k">if</span> <span class="n">network_layout</span> <span class="o">==</span> <span class="s1">&#39;spring&#39;</span><span class="p">:</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>            <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">,</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>                                   <span class="n">k</span><span class="o">=</span><span class="mf">1.</span><span class="p">,</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                                   <span class="n">seed</span><span class="o">=</span><span class="mi">888</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>                                   <span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="k">elif</span> <span class="n">network_layout</span> <span class="o">==</span> <span class="s1">&#39;circular&#39;</span><span class="p">:</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>            <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">circular_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;network_layout should be either &#39;spring&#39; or &#39;circular&#39;&quot;</span><span class="p">)</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="c1"># Weights for edge thickness</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>        <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span><span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">[</span><span class="n">e</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">()])</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="c1"># Visualize network</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>                               <span class="n">pos</span><span class="p">,</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                               <span class="n">alpha</span><span class="o">=</span><span class="n">edge_alpha</span><span class="p">,</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>                               <span class="n">arrowsize</span><span class="o">=</span><span class="n">edge_arrow_size</span><span class="p">,</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>                               <span class="n">width</span><span class="o">=</span><span class="n">weights</span> <span class="o">*</span> <span class="n">edge_width</span><span class="p">,</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>                               <span class="n">edge_color</span><span class="o">=</span><span class="n">edge_color</span><span class="p">,</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>                               <span class="n">connectionstyle</span><span class="o">=</span><span class="s2">&quot;arc3,rad=-0.3&quot;</span><span class="p">,</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>                               <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>                               <span class="p">)</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>        <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                               <span class="n">pos</span><span class="p">,</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>                               <span class="n">node_color</span><span class="o">=</span><span class="n">node_color</span><span class="p">,</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                               <span class="n">node_size</span><span class="o">=</span><span class="n">node_size</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                               <span class="n">alpha</span><span class="o">=</span><span class="n">node_alpha</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                               <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                               <span class="p">)</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>        <span class="n">label_options</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ec&quot;</span><span class="p">:</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="s2">&quot;fc&quot;</span><span class="p">:</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="n">node_label_alpha</span><span class="p">}</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>        <span class="n">_</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_labels</span><span class="p">(</span><span class="n">G</span><span class="p">,</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                                    <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">node_label_offset</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">pos</span><span class="o">.</span><span class="n">items</span><span class="p">()},</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>                                    <span class="n">font_size</span><span class="o">=</span><span class="n">node_label_size</span><span class="p">,</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>                                    <span class="n">bbox</span><span class="o">=</span><span class="n">label_options</span><span class="p">,</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>                                    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>                                    <span class="p">)</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_frame_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="n">xlim</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">()</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>        <span class="n">ylim</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>        <span class="n">coeff</span> <span class="o">=</span> <span class="mf">1.4</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="n">xlim</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">coeff</span><span class="p">,</span> <span class="n">xlim</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">coeff</span><span class="p">))</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="n">ylim</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">coeff</span><span class="p">,</span> <span class="n">ylim</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">coeff</span><span class="p">))</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">factor</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">factor_title_size</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">)</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>    <span class="c1"># Remove extra subplots</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">axs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.factor_plot.context_boxplot">
<code class="highlight language-python"><span class="n">context_boxplot</span><span class="p">(</span><span class="n">context_loadings</span><span class="p">,</span> <span class="n">metadict</span><span class="p">,</span> <span class="n">included_factors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">statistical_test</span><span class="o">=</span><span class="s1">&#39;Mann-Whitney&#39;</span><span class="p">,</span> <span class="n">pval_correction</span><span class="o">=</span><span class="s1">&#39;benjamini-hochberg&#39;</span><span class="p">,</span> <span class="n">text_format</span><span class="o">=</span><span class="s1">&#39;star&#39;</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab10&#39;</span><span class="p">,</span> <span class="n">title_size</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">axis_label_size</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">group_label_rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Context Loadings&#39;</span><span class="p">,</span> <span class="n">dot_color</span><span class="o">=</span><span class="s1">&#39;lightsalmon&#39;</span><span class="p">,</span> <span class="n">dot_edge_color</span><span class="o">=</span><span class="s1">&#39;brown&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots a boxplot to compare the loadings of context groups in each
of the factors resulting from a tensor decomposition.</p>
<h6 id="cell2cell.plotting.factor_plot.context_boxplot--parameters">Parameters</h6>
<p>context_loadings : pandas.DataFrame
    Dataframe containing the loadings of each of the contexts
    from a tensor decomposition. Rows are contexts and columns
    are the factors obtained.</p>
<p>metadict : dict
    A dictionary containing the groups where each of the contexts
    belong to. Keys corresponds to the indexes in <code>context_loadings</code>
    and values are the respective groups. For example:
    metadict={'Context 1' : 'Group 1', 'Context 2' : 'Group 1',
              'Context 3' : 'Group 2', 'Context 4' : 'Group 2'}</p>
<p>included_factors : list, default=None
    Factors to be included. Factor names must be the same as column elements
    in the context_loadings.</p>
<p>group_order : list, default=None
    Order of the groups to plot the boxplots. Considering the
    example of the metadict, it could be:
    group_order=['Group 1', 'Group 2'] or
    group_order=['Group 2', 'Group 1']
    If None, the order that groups are found in <code>metadict</code>
    will be considered.</p>
<p>statistical_test : str, default='Mann-Whitney'
    The statistical test to compare context groups within each factor.
    Options include:
    't-test_ind', 't-test_welch', 't-test_paired', 'Mann-Whitney',
    'Mann-Whitney-gt', 'Mann-Whitney-ls', 'Levene', 'Wilcoxon', 'Kruskal'.</p>
<p>pval_correction : str, default='benjamini-hochberg'
    Multiple test correction method to reduce false positives.
    Options include:
    'bonferroni', 'bonf', 'Bonferroni', 'holm-bonferroni', 'HB',
    'Holm-Bonferroni', 'holm', 'benjamini-hochberg', 'BH', 'fdr_bh',
    'Benjamini-Hochberg', 'fdr_by', 'Benjamini-Yekutieli', 'BY', None</p>
<p>text_format : str, default='star'
    Format to display the results of the statistical test.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;star&#39;, to display P- values &lt; 1e-4 as &quot;****&quot;; &lt; 1e-3 as &quot;***&quot;;
          &lt; 1e-2 as &quot;**&quot;; &lt; 0.05 as &quot;*&quot;, and &lt; 1 as &quot;ns&quot;.
- &#39;simple&#39;, to display P-values &lt; 1e-5 as &quot;1e-5&quot;; &lt; 1e-4 as &quot;1e-4&quot;;
          &lt; 1e-3 as &quot;0.001&quot;; &lt; 1e-2 as &quot;0.01&quot;; and &lt; 5e-2 as &quot;0.05&quot;.
</code></pre></div>

<p>nrows : int, default=1
    Number of rows to generate the subplots.</p>
<p>figsize : tuple, default=(12, 6)
    Size of the figure (width*height), each in inches.</p>
<p>cmap : str, default='tab10'
    Name of the color palette for coloring the major groups of contexts.</p>
<p>title_size : int, default=14
    Font size of the title in each of the factor boxplots.</p>
<p>axis_label_size : int, default=12
    Font size of the labels for X and Y axes.</p>
<p>group_label_rotation : int, default=45
    Angle of rotation for the tick labels in the X axis.</p>
<p>ylabel : str, default='Context Loadings'
    Label for the Y axis.</p>
<p>dot_color : str, default='lightsalmon'
    A matplotlib color for the dots representing individual contexts
    in the boxplot. For more info see:
    https://matplotlib.org/stable/gallery/color/named_colors.html</p>
<p>dot_edge_color : str, default='brown'
    A matplotlib color for the edge of the dots in the boxplot.
    For more info see:
    https://matplotlib.org/stable/gallery/color/named_colors.html</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<p>verbose : boolean, default=None
    Whether printing out the result of the pairwise statistical tests
    in each of the factors</p>
<h6 id="cell2cell.plotting.factor_plot.context_boxplot--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    A matplotlib figure.</p>
<p>axes : matplotlib.axes.Axes or array of Axes
       Matplotlib axes representing the subplots containing the boxplots.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/factor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">context_boxplot</span><span class="p">(</span><span class="n">context_loadings</span><span class="p">,</span> <span class="n">metadict</span><span class="p">,</span> <span class="n">included_factors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">statistical_test</span><span class="o">=</span><span class="s1">&#39;Mann-Whitney&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                    <span class="n">pval_correction</span><span class="o">=</span><span class="s1">&#39;benjamini-hochberg&#39;</span><span class="p">,</span> <span class="n">text_format</span><span class="o">=</span><span class="s1">&#39;star&#39;</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab10&#39;</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                    <span class="n">title_size</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">axis_label_size</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">group_label_rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Context Loadings&#39;</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                    <span class="n">dot_color</span><span class="o">=</span><span class="s1">&#39;lightsalmon&#39;</span><span class="p">,</span> <span class="n">dot_edge_color</span><span class="o">=</span><span class="s1">&#39;brown&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="sd">&#39;&#39;&#39;Plots a boxplot to compare the loadings of context groups in each</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    of the factors resulting from a tensor decomposition.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    context_loadings : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Dataframe containing the loadings of each of the contexts</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        from a tensor decomposition. Rows are contexts and columns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        are the factors obtained.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    metadict : dict</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        A dictionary containing the groups where each of the contexts</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        belong to. Keys corresponds to the indexes in `context_loadings`</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        and values are the respective groups. For example:</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        metadict={&#39;Context 1&#39; : &#39;Group 1&#39;, &#39;Context 2&#39; : &#39;Group 1&#39;,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                  &#39;Context 3&#39; : &#39;Group 2&#39;, &#39;Context 4&#39; : &#39;Group 2&#39;}</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    included_factors : list, default=None</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Factors to be included. Factor names must be the same as column elements</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        in the context_loadings.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    group_order : list, default=None</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Order of the groups to plot the boxplots. Considering the</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        example of the metadict, it could be:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        group_order=[&#39;Group 1&#39;, &#39;Group 2&#39;] or</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        group_order=[&#39;Group 2&#39;, &#39;Group 1&#39;]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        If None, the order that groups are found in `metadict`</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        will be considered.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    statistical_test : str, default=&#39;Mann-Whitney&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        The statistical test to compare context groups within each factor.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Options include:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        &#39;t-test_ind&#39;, &#39;t-test_welch&#39;, &#39;t-test_paired&#39;, &#39;Mann-Whitney&#39;,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        &#39;Mann-Whitney-gt&#39;, &#39;Mann-Whitney-ls&#39;, &#39;Levene&#39;, &#39;Wilcoxon&#39;, &#39;Kruskal&#39;.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    pval_correction : str, default=&#39;benjamini-hochberg&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Multiple test correction method to reduce false positives.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Options include:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        &#39;bonferroni&#39;, &#39;bonf&#39;, &#39;Bonferroni&#39;, &#39;holm-bonferroni&#39;, &#39;HB&#39;,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        &#39;Holm-Bonferroni&#39;, &#39;holm&#39;, &#39;benjamini-hochberg&#39;, &#39;BH&#39;, &#39;fdr_bh&#39;,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        &#39;Benjamini-Hochberg&#39;, &#39;fdr_by&#39;, &#39;Benjamini-Yekutieli&#39;, &#39;BY&#39;, None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    text_format : str, default=&#39;star&#39;</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        Format to display the results of the statistical test.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        - &#39;star&#39;, to display P- values &lt; 1e-4 as &quot;****&quot;; &lt; 1e-3 as &quot;***&quot;;</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">                  &lt; 1e-2 as &quot;**&quot;; &lt; 0.05 as &quot;*&quot;, and &lt; 1 as &quot;ns&quot;.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        - &#39;simple&#39;, to display P-values &lt; 1e-5 as &quot;1e-5&quot;; &lt; 1e-4 as &quot;1e-4&quot;;</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">                  &lt; 1e-3 as &quot;0.001&quot;; &lt; 1e-2 as &quot;0.01&quot;; and &lt; 5e-2 as &quot;0.05&quot;.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    nrows : int, default=1</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Number of rows to generate the subplots.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    figsize : tuple, default=(12, 6)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">    cmap : str, default=&#39;tab10&#39;</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        Name of the color palette for coloring the major groups of contexts.</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    title_size : int, default=14</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        Font size of the title in each of the factor boxplots.</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    axis_label_size : int, default=12</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        Font size of the labels for X and Y axes.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">    group_label_rotation : int, default=45</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        Angle of rotation for the tick labels in the X axis.</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">    ylabel : str, default=&#39;Context Loadings&#39;</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        Label for the Y axis.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    dot_color : str, default=&#39;lightsalmon&#39;</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        A matplotlib color for the dots representing individual contexts</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        in the boxplot. For more info see:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">        https://matplotlib.org/stable/gallery/color/named_colors.html</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">    dot_edge_color : str, default=&#39;brown&#39;</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        A matplotlib color for the edge of the dots in the boxplot.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        For more info see:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">        https://matplotlib.org/stable/gallery/color/named_colors.html</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">    verbose : boolean, default=None</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        Whether printing out the result of the pairwise statistical tests</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        in each of the factors</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">    -------</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">        A matplotlib figure.</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">    axes : matplotlib.axes.Axes or array of Axes</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">           Matplotlib axes representing the subplots containing the boxplots.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="k">if</span> <span class="n">group_order</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">group_order</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">metadict</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">metadict</span><span class="o">.</span><span class="n">values</span><span class="p">())),</span> <span class="s2">&quot;All groups in `metadict` must be contained in `group_order`&quot;</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>        <span class="n">group_order</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">metadict</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">context_loadings</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="k">if</span> <span class="n">included_factors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>        <span class="n">factor_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>        <span class="n">factor_labels</span> <span class="o">=</span> <span class="n">included_factors</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="n">rank</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">metadict</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="n">nrows</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">rank</span><span class="p">,</span> <span class="n">nrows</span><span class="p">])</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>    <span class="n">ncols</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">rank</span><span class="o">/</span><span class="n">nrows</span><span class="p">))</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="n">ncols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="k">if</span> <span class="n">rank</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="n">axs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">axes</span><span class="p">])</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="n">axs</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">factor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">factor_labels</span><span class="p">):</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="n">factor</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="n">order</span> <span class="o">=</span> <span class="n">group_order</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="c1"># Plot the boxes</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                         <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>                         <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>                         <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>                         <span class="n">whis</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>                         <span class="n">width</span><span class="o">=</span><span class="mf">.6</span><span class="p">,</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>                         <span class="n">palette</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>                         <span class="n">boxprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.5</span><span class="p">),</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>                         <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                         <span class="p">)</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="c1"># Plot the dots</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>        <span class="n">sns</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                      <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                      <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                      <span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                      <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                      <span class="n">color</span><span class="o">=</span><span class="n">dot_color</span><span class="p">,</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>                      <span class="n">edgecolor</span><span class="o">=</span><span class="n">dot_edge_color</span><span class="p">,</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>                      <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>                      <span class="n">jitter</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>                      <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>                      <span class="p">)</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="k">if</span> <span class="n">statistical_test</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>            <span class="c1"># Add annotations about statistical test</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>            <span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">combinations</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">combinations</span><span class="p">(</span><span class="n">order</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>            <span class="n">annotator</span> <span class="o">=</span> <span class="n">Annotator</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>                                  <span class="n">pairs</span><span class="o">=</span><span class="n">pairs</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>                                  <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>                                  <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>                                  <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>                                  <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>            <span class="n">annotator</span><span class="o">.</span><span class="n">configure</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">statistical_test</span><span class="p">,</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>                                <span class="n">text_format</span><span class="o">=</span><span class="n">text_format</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>                                <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;inside&#39;</span><span class="p">,</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                                <span class="n">comparisons_correction</span><span class="o">=</span><span class="n">pval_correction</span><span class="p">,</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>                                <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                                <span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>            <span class="n">annotator</span><span class="o">.</span><span class="n">apply_and_annotate</span><span class="p">()</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">factor</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">title_size</span><span class="p">)</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_label_size</span><span class="p">)</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(((</span><span class="n">i</span><span class="p">)</span> <span class="o">%</span> <span class="n">ncols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_label_size</span><span class="p">)</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_label_size</span><span class="p">)</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>                           <span class="n">rotation</span><span class="o">=</span><span class="n">group_label_rotation</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>                           <span class="n">rotation_mode</span><span class="o">=</span><span class="s1">&#39;anchor&#39;</span><span class="p">,</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                           <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>                           <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>    <span class="c1"># Remove extra subplots</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">axs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>    <span class="k">if</span> <span class="n">axes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>        <span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">axes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">align_ylabels</span><span class="p">(</span><span class="n">axes</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">rect</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.99</span><span class="p">])</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.factor_plot.loading_clustermap">
<code class="highlight language-python"><span class="n">loading_clustermap</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">loading_threshold</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">use_zscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">heatmap_lw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">tick_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cbar_label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots a clustermap of the tensor-factorization loadings from one tensor dimension or
   the joint loadings from multiple tensor dimensions.</p>
<p>Parameters</p>
<hr />
<p>loadings : pandas.DataFrame
       Loadings for a given tensor dimension after running the tensor
       decomposition. Rows are the elements in one dimension or joint
       pairs/n-tuples in multiple dimensions. It is recommended that
       the loadings resulting from the decomposition should be
       l2-normalized prior to their use, by considering all dimensions
       together. For example, take the factors dictionary found in any
       InteractionTensor or any BaseTensor derived class, and execute
       cell2cell.tensor.normalize(factors).</p>
<p>loading_threshold : float
       Threshold to filter out elements in the loadings dataframe.
       This plot considers elements with loadings greater than this
       threshold in at least one of the factors.</p>
<p>use_zscore : boolean
       Whether converting loadings to z-scores across factors.</p>
<p>metric : str, default='euclidean'
       The distance metric to use. The distance function can be 'braycurtis',
       'canberra', 'chebyshev', 'cityblock', 'correlation', 'cosine', 'dice',
       'euclidean', 'hamming', 'jaccard', 'jensenshannon', 'kulsinski',
       'mahalanobis', 'matching', 'minkowski', 'rogerstanimoto', 'russellrao',
       'seuclidean', 'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule'.</p>
<p>method : str, 'ward' by default
       Method to compute the linkage.
       It could be:</p>
<div class="codehilite"><pre><span></span><code><span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;single&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;average&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;weighted&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;centroid&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;median&#39;</span><span class="w"></span>
<span class="w">   </span><span class="o">-</span><span class="w"> </span><span class="s1">&#39;ward&#39;</span><span class="w"></span>
<span class="w">   </span><span class="k">For</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">details</span><span class="p">,</span><span class="w"> </span><span class="k">go</span><span class="w"> </span><span class="k">to</span><span class="err">:</span><span class="w"></span>
<span class="w">   </span><span class="nl">https</span><span class="p">:</span><span class="o">//</span><span class="n">docs</span><span class="p">.</span><span class="n">scipy</span><span class="p">.</span><span class="n">org</span><span class="o">/</span><span class="n">doc</span><span class="o">/</span><span class="n">scipy</span><span class="o">-</span><span class="mf">0.19.0</span><span class="o">/</span><span class="n">reference</span><span class="o">/</span><span class="n">generated</span><span class="o">/</span><span class="n">scipy</span><span class="p">.</span><span class="n">cluster</span><span class="p">.</span><span class="n">hierarchy</span><span class="p">.</span><span class="n">linkage</span><span class="p">.</span><span class="n">html</span><span class="w"></span>
</code></pre></div>

<p>optimal_leaf : boolean, default=True
       Whether sorting the leaf of the dendrograms to have a minimal distance
       between successive leaves. For more information, see
       scipy.cluster.hierarchy.optimal_leaf_ordering</p>
<p>figsize : tuple, default=(16, 9)
       Size of the figure (width*height), each in inches.</p>
<p>heatmap_lw : float, default=0.2
       Width of the lines that will divide each cell.</p>
<p>cbar_fontsize : int, default=12
       Font size for the colorbar title.</p>
<p>tick_fontsize : int, default=10
       Font size for ticks in the x and y axes.</p>
<p>cmap : str, default=None
       Name of the color palette for coloring the heatmap. If None,
       cmap='Blues' would be used when use_zscore=False; and cmap='vlag' when use_zscore=True.</p>
<p>cbar_label : str, default=None
       Label for the color bar. If None, default labels will be 'Z-scores 
across factors'
       or 'Loadings', depending on <code>use_zcore</code> is True or False, respectively.</p>
<p>filename : str, default=None
       Path to save the figure of the elbow analysis. If None, the figure is not
       saved.</p>
<p>**kwargs : dict
       Dictionary containing arguments for the seaborn.clustermap function.</p>
<p>Returns</p>
<hr />
<p>cm : seaborn.matrix.ClusterGrid
       A seaborn ClusterGrid instance.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/factor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">loading_clustermap</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">loading_threshold</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">use_zscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                       <span class="n">optimal_leaf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">heatmap_lw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">cbar_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">tick_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                       <span class="n">cbar_label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Plots a clustermap of the tensor-factorization loadings from one tensor dimension or</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    the joint loadings from multiple tensor dimensions.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    loadings : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Loadings for a given tensor dimension after running the tensor</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        decomposition. Rows are the elements in one dimension or joint</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        pairs/n-tuples in multiple dimensions. It is recommended that</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        the loadings resulting from the decomposition should be</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        l2-normalized prior to their use, by considering all dimensions</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        together. For example, take the factors dictionary found in any</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        InteractionTensor or any BaseTensor derived class, and execute</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        cell2cell.tensor.normalize(factors).</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    loading_threshold : float</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Threshold to filter out elements in the loadings dataframe.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        This plot considers elements with loadings greater than this</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        threshold in at least one of the factors.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    use_zscore : boolean</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Whether converting loadings to z-scores across factors.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    metric : str, default=&#39;euclidean&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        The distance metric to use. The distance function can be &#39;braycurtis&#39;,</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        &#39;canberra&#39;, &#39;chebyshev&#39;, &#39;cityblock&#39;, &#39;correlation&#39;, &#39;cosine&#39;, &#39;dice&#39;,</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        &#39;euclidean&#39;, &#39;hamming&#39;, &#39;jaccard&#39;, &#39;jensenshannon&#39;, &#39;kulsinski&#39;,</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        &#39;mahalanobis&#39;, &#39;matching&#39;, &#39;minkowski&#39;, &#39;rogerstanimoto&#39;, &#39;russellrao&#39;,</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        &#39;seuclidean&#39;, &#39;sokalmichener&#39;, &#39;sokalsneath&#39;, &#39;sqeuclidean&#39;, &#39;yule&#39;.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    method : str, &#39;ward&#39; by default</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Method to compute the linkage.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        It could be:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        - &#39;single&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        - &#39;complete&#39;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        - &#39;average&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        - &#39;weighted&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        - &#39;centroid&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        - &#39;median&#39;</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        - &#39;ward&#39;</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        For more details, go to:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.cluster.hierarchy.linkage.html</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    optimal_leaf : boolean, default=True</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Whether sorting the leaf of the dendrograms to have a minimal distance</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        between successive leaves. For more information, see</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        scipy.cluster.hierarchy.optimal_leaf_ordering</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    figsize : tuple, default=(16, 9)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    heatmap_lw : float, default=0.2</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Width of the lines that will divide each cell.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    cbar_fontsize : int, default=12</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Font size for the colorbar title.</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">    tick_fontsize : int, default=10</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        Font size for ticks in the x and y axes.</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    cmap : str, default=None</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        Name of the color palette for coloring the heatmap. If None,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        cmap=&#39;Blues&#39; would be used when use_zscore=False; and cmap=&#39;vlag&#39; when use_zscore=True.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    cbar_label : str, default=None</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        Label for the color bar. If None, default labels will be &#39;Z-scores \n across factors&#39;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        or &#39;Loadings&#39;, depending on `use_zcore` is True or False, respectively.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Dictionary containing arguments for the seaborn.clustermap function.</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">    -------</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">    cm : seaborn.matrix.ClusterGrid</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        A seaborn ClusterGrid instance.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">loadings</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="o">.</span><span class="n">T</span> <span class="o">&gt;</span> <span class="n">loading_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()]</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>    <span class="k">if</span> <span class="n">use_zscore</span><span class="p">:</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">zscore</span><span class="p">)</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>        <span class="k">if</span> <span class="n">cmap</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>            <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;vlag&#39;</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="n">val</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">max</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()),</span> <span class="nb">abs</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">())]))</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>        <span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">val</span><span class="p">,</span> <span class="n">val</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>        <span class="k">if</span> <span class="n">cbar_label</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>            <span class="n">cbar_label</span> <span class="o">=</span> <span class="s1">&#39;Z-scores </span><span class="se">\n</span><span class="s1"> across factors&#39;</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="k">if</span> <span class="n">cmap</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>            <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mf">0.</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="k">if</span> <span class="n">cbar_label</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>            <span class="n">cbar_label</span> <span class="o">=</span> <span class="s1">&#39;Loadings&#39;</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="c1"># Clustering</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="n">dm_rows</span> <span class="o">=</span> <span class="n">compute_distance</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="n">row_linkage</span> <span class="o">=</span> <span class="n">compute_linkage</span><span class="p">(</span><span class="n">dm_rows</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_leaf</span><span class="p">)</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">dm_cols</span> <span class="o">=</span> <span class="n">compute_distance</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">)</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="n">col_linkage</span> <span class="o">=</span> <span class="n">compute_linkage</span><span class="p">(</span><span class="n">dm_cols</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">optimal_ordering</span><span class="o">=</span><span class="n">optimal_leaf</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="c1"># Clustermap</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="n">cm</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>                        <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>                        <span class="n">col_linkage</span><span class="o">=</span><span class="n">col_linkage</span><span class="p">,</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>                        <span class="n">row_linkage</span><span class="o">=</span><span class="n">row_linkage</span><span class="p">,</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>                        <span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>                        <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">,</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>                        <span class="n">xticklabels</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>                        <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>                        <span class="n">linewidths</span><span class="o">=</span><span class="n">heatmap_lw</span><span class="p">,</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>                        <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>                        <span class="p">)</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>    <span class="c1"># Color bar label</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="n">cbar</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">collections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">colorbar</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>    <span class="n">cbar</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">cbar_label</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">cbar_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>    <span class="n">cbar</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_label_position</span><span class="p">(</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="c1"># Tick labels</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">tick_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">get_majorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">tick_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>    <span class="c1"># Resize clustermap and dendrograms</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>    <span class="n">hm</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>    <span class="n">w_mult</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>    <span class="n">h_mult</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_position</span><span class="p">([</span><span class="n">hm</span><span class="o">.</span><span class="n">x0</span><span class="p">,</span> <span class="n">hm</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="n">hm</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">w_mult</span><span class="p">,</span> <span class="n">hm</span><span class="o">.</span><span class="n">height</span><span class="p">])</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>    <span class="n">row</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>    <span class="n">row_d_mult</span> <span class="o">=</span> <span class="mf">0.33</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">set_position</span><span class="p">(</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">x0</span> <span class="o">+</span> <span class="n">row</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">row_d_mult</span><span class="p">),</span> <span class="n">row</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">row_d_mult</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">height</span> <span class="o">*</span> <span class="n">h_mult</span><span class="p">])</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>    <span class="n">col</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_position</span><span class="p">([</span><span class="n">col</span><span class="o">.</span><span class="n">x0</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">w_mult</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">height</span> <span class="o">*</span> <span class="mf">0.5</span><span class="p">])</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_xlabel</span><span class="p">(),</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span> <span class="o">*</span> <span class="n">tick_fontsize</span><span class="p">))</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>    <span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_ylabel</span><span class="p">(),</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span> <span class="o">*</span> <span class="n">tick_fontsize</span><span class="p">))</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>    <span class="k">return</span> <span class="n">cm</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.pcoa_plot">
        <code>pcoa_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.pcoa_plot.pcoa_3dplot">
<code class="highlight language-python"><span class="n">pcoa_3dplot</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">pcoa_method</span><span class="o">=</span><span class="s1">&#39;eigh&#39;</span><span class="p">,</span> <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">axis_fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">legend_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">view_angles</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">135</span><span class="p">),</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Projects the cells into an Euclidean space (PCoA) given their distances
based on their CCI scores. Then, plots each cell by their first three
coordinates in a 3D scatter plot.</p>
<h6 id="cell2cell.plotting.pcoa_plot.pcoa_3dplot--parameters">Parameters</h6>
<p>interaction_space : cell2cell.core.interaction_space.InteractionSpace
    Interaction space that contains all a distance matrix after running the
    the method compute_pairwise_cci_scores. Alternatively, this object
    can be a numpy-array or a pandas DataFrame. Also, a
    SingleCellInteractions or a BulkInteractions object after running
    the method compute_pairwise_cci_scores.</p>
<p>metadata : pandas.Dataframe, default=None
    Metadata associated with the cells, cell types or samples in the
    matrix containing CCC scores. If None, cells will not be colored
    by major groups.</p>
<p>sample_col : str, default='#SampleID'
    Column in the metadata for the cells, cell types or samples
    in the matrix containing CCI scores.</p>
<p>group_col : str, default='Groups'
    Column in the metadata containing the major groups of cells, cell types
    or samples in the matrix with CCI scores.</p>
<p>pcoa_method : str, default='eigh'
    Eigendecomposition method to use in performing PCoA.
    By default, uses SciPy's <code>eigh</code>, which computes exact
    eigenvectors and eigenvalues for all dimensions. The alternate
    method, <code>fsvd</code>, uses faster heuristic eigendecomposition but loses
    accuracy. The magnitude of accuracy lost is dependent on dataset.</p>
<p>meta_cmap : str, default='gist_rainbow'
    Name of the color palette for coloring the major groups of cells.</p>
<p>colors : dict, default=None
    Dictionary containing tuples in the RGBA format for indicating colors
    of major groups of cells. If colors is specified, meta_cmap will be
    ignored.</p>
<p>excluded_cells : list, default=None
    List containing cell names that are present in the interaction_space
    object but that will be excluded from this plot.</p>
<p>title : str, default=''
    Title of the PCoA 3D plot.</p>
<p>axis_fontsize : int, default=14
    Size of the font for the labels of each axis (X, Y and Z).</p>
<p>legend_fontsize : int, default=12
    Size of the font for labels in the legend.</p>
<p>figsize : tuple, default=(6, 5)
    Size of the figure (width*height), each in inches.</p>
<p>view_angles : tuple, default=(30, 135)
    Rotation angles of the plot. Set the elevation and
    azimuth of the axes.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.pcoa_plot.pcoa_3dplot--returns">Returns</h6>
<p>results : dict
    Dictionary that contains:</p>
<div class="codehilite"><pre><span></span><code>- &#39;fig&#39; : matplotlib.figure.Figure, containing the whole figure
- &#39;axes&#39; : matplotlib.axes.Axes, containing the axes of the 3D plot
- &#39;ordination&#39; : Ordination or projection obtained from the PCoA
- &#39;distance_matrix&#39; : Distance matrix used to perform the PCoA (usually in
    interaction_space.distance_matrix
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/pcoa_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pcoa_3dplot</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;#SampleID&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Groups&#39;</span><span class="p">,</span> <span class="n">pcoa_method</span><span class="o">=</span><span class="s1">&#39;eigh&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                <span class="n">meta_cmap</span><span class="o">=</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">axis_fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">legend_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">view_angles</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">135</span><span class="p">),</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Projects the cells into an Euclidean space (PCoA) given their distances</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    based on their CCI scores. Then, plots each cell by their first three</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    coordinates in a 3D scatter plot.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    interaction_space : cell2cell.core.interaction_space.InteractionSpace</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Interaction space that contains all a distance matrix after running the</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        the method compute_pairwise_cci_scores. Alternatively, this object</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        can be a numpy-array or a pandas DataFrame. Also, a</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        SingleCellInteractions or a BulkInteractions object after running</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        the method compute_pairwise_cci_scores.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    metadata : pandas.Dataframe, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Metadata associated with the cells, cell types or samples in the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        matrix containing CCC scores. If None, cells will not be colored</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        by major groups.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    sample_col : str, default=&#39;#SampleID&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Column in the metadata for the cells, cell types or samples</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        in the matrix containing CCI scores.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    group_col : str, default=&#39;Groups&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Column in the metadata containing the major groups of cells, cell types</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        or samples in the matrix with CCI scores.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    pcoa_method : str, default=&#39;eigh&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Eigendecomposition method to use in performing PCoA.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        By default, uses SciPy&#39;s `eigh`, which computes exact</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        eigenvectors and eigenvalues for all dimensions. The alternate</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        method, `fsvd`, uses faster heuristic eigendecomposition but loses</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        accuracy. The magnitude of accuracy lost is dependent on dataset.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    meta_cmap : str, default=&#39;gist_rainbow&#39;</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Name of the color palette for coloring the major groups of cells.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    colors : dict, default=None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Dictionary containing tuples in the RGBA format for indicating colors</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        of major groups of cells. If colors is specified, meta_cmap will be</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        ignored.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    excluded_cells : list, default=None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        List containing cell names that are present in the interaction_space</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        object but that will be excluded from this plot.</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    title : str, default=&#39;&#39;</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        Title of the PCoA 3D plot.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    axis_fontsize : int, default=14</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        Size of the font for the labels of each axis (X, Y and Z).</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    legend_fontsize : int, default=12</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        Size of the font for labels in the legend.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">    figsize : tuple, default=(6, 5)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    view_angles : tuple, default=(30, 135)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        Rotation angles of the plot. Set the elevation and</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        azimuth of the axes.</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">    -------</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">    results : dict</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        Dictionary that contains:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        - &#39;fig&#39; : matplotlib.figure.Figure, containing the whole figure</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        - &#39;axes&#39; : matplotlib.axes.Axes, containing the axes of the 3D plot</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">        - &#39;ordination&#39; : Ordination or projection obtained from the PCoA</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        - &#39;distance_matrix&#39; : Distance matrix used to perform the PCoA (usually in</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">            interaction_space.distance_matrix</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as an InteractionSpace class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">distance_matrix</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="k">elif</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">)</span> <span class="ow">is</span> <span class="n">pd</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">frame</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a distance matrix&#39;</span><span class="p">)</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;interaction_space&#39;</span><span class="p">):</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Interaction space detected as a Interactions class&#39;</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span><span class="p">):</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_interactions() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>                             <span class="s1">&#39; object to generate a distance matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>            <span class="n">distance_matrix</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">distance_matrix</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;First run the method compute_pairwise_interactions() in your interaction&#39;</span> <span class="o">+</span> \
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                         <span class="s1">&#39; object to generate a distance matrix.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="c1"># Drop excluded cells</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">),</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                                 <span class="o">~</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)]</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">distance_matrix</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="c1"># PCoA</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">ordination</span> <span class="o">=</span> <span class="n">pcoa</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">pcoa_method</span><span class="p">)</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="n">ordination</span> <span class="o">=</span> <span class="n">_check_ordination</span><span class="p">(</span><span class="n">ordination</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="c1"># Biplot</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="c1">#ax = Axes3D(fig) # Not displayed in newer versions</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="n">metadata</span><span class="p">[</span><span class="n">sample_col</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="n">metadata</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>    <span class="n">meta_</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="n">meta_</span> <span class="o">=</span> <span class="n">meta_</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">meta_</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)]</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="n">labels</span> <span class="o">=</span> <span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>    <span class="k">if</span> <span class="n">colors</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="n">colors</span> <span class="o">=</span> <span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">meta_cmap</span><span class="p">)</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">elem</span> <span class="ow">in</span> <span class="n">colors</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">elem</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>    <span class="c1"># Plot each data point with respective color</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())):</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">meta_</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">meta_</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="k">if</span> <span class="n">colors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC1&#39;</span><span class="p">],</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                       <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">],</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>                       <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC3&#39;</span><span class="p">],</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>                       <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">cell_type</span><span class="p">],</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>                       <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>                       <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>                       <span class="n">label</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC1&#39;</span><span class="p">],</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                       <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">],</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>                       <span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="s1">&#39;PC3&#39;</span><span class="p">],</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                       <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                       <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                       <span class="n">label</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>    <span class="c1"># Plot texts</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;PC1 (</span><span class="si">{}</span><span class="s1">%)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;proportion_explained&#39;</span><span class="p">][</span><span class="s1">&#39;PC1&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;PC2 (</span><span class="si">{}</span><span class="s1">%)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;proportion_explained&#39;</span><span class="p">][</span><span class="s1">&#39;PC2&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s1">&#39;PC3 (</span><span class="si">{}</span><span class="s1">%)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">ordination</span><span class="p">[</span><span class="s1">&#39;proportion_explained&#39;</span><span class="p">][</span><span class="s1">&#39;PC3&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">axis_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([])</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([])</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_zticklabels</span><span class="p">([])</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">view_init</span><span class="p">(</span><span class="n">view_angles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">view_angles</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.35</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>               <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">)</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>    <span class="c1">#distskbio = skbio.DistanceMatrix(df, ids=df.index) # Not using skbio for now</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>    <span class="c1"># Save plot</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;fig&#39;</span> <span class="p">:</span> <span class="n">fig</span><span class="p">,</span> <span class="s1">&#39;axes&#39;</span> <span class="p">:</span> <span class="n">ax</span><span class="p">,</span> <span class="s1">&#39;ordination&#39;</span> <span class="p">:</span> <span class="n">ordination</span><span class="p">,</span> <span class="s1">&#39;distance_matrix&#39;</span> <span class="p">:</span> <span class="n">df</span><span class="p">}</span> <span class="c1"># df used to be distskbio</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>    <span class="k">return</span> <span class="n">results</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.pval_plot">
        <code>pval_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.pval_plot.dot_plot">
<code class="highlight language-python"><span class="n">dot_plot</span><span class="p">(</span><span class="n">sc_interactions</span><span class="p">,</span> <span class="n">evaluation</span><span class="o">=</span><span class="s1">&#39;communication&#39;</span><span class="p">,</span> <span class="n">significance</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">senders</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">receivers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">tick_size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;PuOr&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a dot plot for the CCI or communication scores given their
P-values. Size of the dots are given by the -log10(P-value) and colors
by the value of the CCI or communication score.</p>
<h6 id="cell2cell.plotting.pval_plot.dot_plot--parameters">Parameters</h6>
<p>sc_interactions : cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions
    Interaction class with all necessary methods to run the cell2cell
    pipeline on a single-cell RNA-seq dataset. The method
    permute_cell_labels() must be run before generating this plot.</p>
<p>evaluation : str, default='communication'
    P-values of CCI or communication scores used for this plot.
    - 'interactions' : For CCI scores
    - 'communication' : For communication scores</p>
<p>significance : float, default=0.05
    The significance threshold to be plotted. LR pairs or cell-cell
    pairs with at least one P-value below this threshold will be
    considered.</p>
<p>senders : list, default=None
    Optional filter to plot specific sender cells.</p>
<p>receivers : list, default=None
    Optional filter to plot specific receiver cells.</p>
<p>figsize : tuple, default=(16, 9)
    Size of the figure (width*height), each in inches.</p>
<p>tick_size : int, default=8
    Specifies the size of ticklabels as well as the maximum size
    of the dots.</p>
<p>cmap : str, default='PuOr'
    A matplotlib color palette name.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.pval_plot.dot_plot--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/pval_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">dot_plot</span><span class="p">(</span><span class="n">sc_interactions</span><span class="p">,</span> <span class="n">evaluation</span><span class="o">=</span><span class="s1">&#39;communication&#39;</span><span class="p">,</span> <span class="n">significance</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">senders</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">receivers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>             <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">tick_size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;PuOr&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Generates a dot plot for the CCI or communication scores given their</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    P-values. Size of the dots are given by the -log10(P-value) and colors</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    by the value of the CCI or communication score.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    sc_interactions : cell2cell.analysis.cell2cell_pipelines.SingleCellInteractions</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Interaction class with all necessary methods to run the cell2cell</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        pipeline on a single-cell RNA-seq dataset. The method</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        permute_cell_labels() must be run before generating this plot.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    evaluation : str, default=&#39;communication&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        P-values of CCI or communication scores used for this plot.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        - &#39;interactions&#39; : For CCI scores</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;communication&#39; : For communication scores</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    significance : float, default=0.05</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        The significance threshold to be plotted. LR pairs or cell-cell</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        pairs with at least one P-value below this threshold will be</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        considered.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    senders : list, default=None</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Optional filter to plot specific sender cells.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    receivers : list, default=None</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Optional filter to plot specific receiver cells.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    figsize : tuple, default=(16, 9)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    tick_size : int, default=8</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Specifies the size of ticklabels as well as the maximum size</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        of the dots.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    cmap : str, default=&#39;PuOr&#39;</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        A matplotlib color palette name.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    -------</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">hasattr</span><span class="p">(</span><span class="n">sc_interactions</span><span class="p">,</span> <span class="s1">&#39;ccc_permutation_pvalues&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                <span class="s1">&#39;Run the method permute_cell_labels() with evaluation=&quot;communication&quot; before plotting communication P-values.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>            <span class="n">pvals</span> <span class="o">=</span> <span class="n">sc_interactions</span><span class="o">.</span><span class="n">ccc_permutation_pvalues</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>            <span class="n">scores</span> <span class="o">=</span> <span class="n">sc_interactions</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;communication_matrix&#39;</span><span class="p">]</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">hasattr</span><span class="p">(</span><span class="n">sc_interactions</span><span class="p">,</span> <span class="s1">&#39;cci_permutation_pvalues&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>                <span class="s1">&#39;Run the method permute_cell_labels() with evaluation=&quot;interactions&quot; before plotting interaction P-values.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>            <span class="n">pvals</span> <span class="o">=</span> <span class="n">sc_interactions</span><span class="o">.</span><span class="n">cci_permutation_pvalues</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>            <span class="n">scores</span> <span class="o">=</span> <span class="n">sc_interactions</span><span class="o">.</span><span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;evaluation has to be either &quot;communication&quot; or &quot;interactions&quot;&#39;</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pvals</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="n">score_df</span> <span class="o">=</span> <span class="n">scores</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="c1"># Filter cells</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;communication&#39;</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">senders</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">receivers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>            <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="o">+</span> <span class="s1">&#39;;&#39;</span> <span class="o">+</span> <span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">receivers</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">senders</span><span class="p">]</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="k">elif</span> <span class="n">senders</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>            <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">senders</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="p">(</span><span class="n">s</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])]</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="k">elif</span> <span class="n">receivers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>            <span class="n">new_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">receivers</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="p">(</span><span class="n">r</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])]</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>            <span class="n">new_cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_cols</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;Sender-Receiver Pairs&#39;</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="n">ylabel</span> <span class="o">=</span> <span class="s1">&#39;Ligand-Receptor Pairs&#39;</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;Communication Score&#39;</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="k">elif</span> <span class="n">evaluation</span> <span class="o">==</span> <span class="s1">&#39;interactions&#39;</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="k">if</span> <span class="n">senders</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>            <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">senders</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="k">if</span> <span class="n">receivers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>            <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">receivers</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>        <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;Receiver Cells&#39;</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>        <span class="n">ylabel</span> <span class="o">=</span> <span class="s1">&#39;Sender Cells&#39;</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>        <span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;CCI Score&#39;</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="n">pval_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">))</span> \
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="n">score_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">score_df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="n">score_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39; --&gt; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">))</span> \
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">score_df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="n">fig</span> <span class="o">=</span> <span class="n">generate_dot_plot</span><span class="p">(</span><span class="n">pval_df</span><span class="o">=</span><span class="n">pval_df</span><span class="p">,</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>                            <span class="n">score_df</span><span class="o">=</span><span class="n">score_df</span><span class="p">,</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                            <span class="n">xlabel</span><span class="o">=</span><span class="n">xlabel</span><span class="p">,</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>                            <span class="n">ylabel</span><span class="o">=</span><span class="n">ylabel</span><span class="p">,</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>                            <span class="n">cbar_title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>                            <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>                            <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>                            <span class="n">significance</span><span class="o">=</span><span class="n">significance</span><span class="p">,</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>                            <span class="n">label_size</span><span class="o">=</span><span class="mi">24</span><span class="p">,</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>                            <span class="n">title_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>                            <span class="n">tick_size</span><span class="o">=</span><span class="n">tick_size</span><span class="p">,</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>                            <span class="n">filename</span><span class="o">=</span><span class="n">filename</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>                            <span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.pval_plot.generate_dot_plot">
<code class="highlight language-python"><span class="n">generate_dot_plot</span><span class="p">(</span><span class="n">pval_df</span><span class="p">,</span> <span class="n">score_df</span><span class="p">,</span> <span class="n">significance</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;Score&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;PuOr&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">label_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">title_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tick_size</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a dot plot for given P-values and respective scores.</p>
<h6 id="cell2cell.plotting.pval_plot.generate_dot_plot--parameters">Parameters</h6>
<p>pval_df : pandas.DataFrame
    A dataframe containing the P-values, with multiple elements
    in both rows and columns</p>
<p>score_df : pandas.DataFrame
    A dataframe containing the scores that were tested. Rows and
    columns must be the same as in <code>pval_df</code>.</p>
<p>significance : float, default=0.05
    The significance threshold to be plotted. LR pairs or cell-cell
    pairs with at least one P-value below this threshold will be
    considered.</p>
<p>xlabel : str, default=''
    Name or label of the X axis.</p>
<p>ylabel : str, default=''
    Name or label of the Y axis.</p>
<p>cbar_title : str, default='Score'
    A title for the colorbar associated with the scores in
    <code>score_df</code>. It is usually the name of the score.</p>
<p>cmap : str, default='PuOr'
    A matplotlib color palette name.</p>
<p>figsize : tuple, default=(16, 9)
    Size of the figure (width*height), each in inches.</p>
<p>label_size : int, default=20
    Specifies the size of the labels of both X and Y axes.</p>
<p>title_size : int, default=20
    Specifies the size of the title of the colorbar and P-val sizes.</p>
<p>tick_size : int, default=14
    Specifies the size of ticklabels as well as the maximum size
    of the dots.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.pval_plot.generate_dot_plot--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/pval_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_dot_plot</span><span class="p">(</span><span class="n">pval_df</span><span class="p">,</span> <span class="n">score_df</span><span class="p">,</span> <span class="n">significance</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;Score&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;PuOr&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                      <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">label_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">title_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tick_size</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Generates a dot plot for given P-values and respective scores.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    pval_df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe containing the P-values, with multiple elements</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        in both rows and columns</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    score_df : pandas.DataFrame</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A dataframe containing the scores that were tested. Rows and</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        columns must be the same as in `pval_df`.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    significance : float, default=0.05</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        The significance threshold to be plotted. LR pairs or cell-cell</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        pairs with at least one P-value below this threshold will be</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        considered.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    xlabel : str, default=&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Name or label of the X axis.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    ylabel : str, default=&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Name or label of the Y axis.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    cbar_title : str, default=&#39;Score&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        A title for the colorbar associated with the scores in</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        `score_df`. It is usually the name of the score.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    cmap : str, default=&#39;PuOr&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        A matplotlib color palette name.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    figsize : tuple, default=(16, 9)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    label_size : int, default=20</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Specifies the size of the labels of both X and Y axes.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    title_size : int, default=20</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        Specifies the size of the title of the colorbar and P-val sizes.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    tick_size : int, default=14</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Specifies the size of ticklabels as well as the maximum size</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        of the dots.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    -------</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="c1"># Preprocessing</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pval_df</span><span class="o">.</span><span class="n">lt</span><span class="p">(</span><span class="n">significance</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="c1"># Drop all zeros</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)]</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="n">pval_df</span> <span class="o">=</span> <span class="n">pval_df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">-</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mf">1e-9</span><span class="p">))</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="c1"># Set dot sizes and color range</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="n">max_abs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">score_df</span><span class="p">))),</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">score_df</span><span class="p">)))])</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="n">norm</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=-</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">max_abs</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">max_abs</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">max_size</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="c1"># Colormap</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="c1"># Dot plot</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax2</span><span class="p">,</span> <span class="n">ax</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;height_ratios&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">]})</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>            <span class="n">color</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">cmap</span><span class="p">(</span><span class="n">norm</span><span class="p">(</span><span class="n">score_df</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">idx</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">item</span><span class="p">())))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>            <span class="n">v</span> <span class="o">=</span> <span class="n">pval_df</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">idx</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>            <span class="n">size</span> <span class="o">=</span> <span class="p">((</span><span class="n">max_size</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="n">v</span><span class="p">,</span> <span class="mi">3</span><span class="p">]))</span> <span class="o">*</span> <span class="n">tick_size</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="c1"># Change tick labels</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="n">xlabels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="n">ylabels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">ticks</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">,</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>                       <span class="n">fontsize</span><span class="o">=</span><span class="n">tick_size</span><span class="p">,</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>                       <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                       <span class="n">rotation_mode</span><span class="o">=</span><span class="s1">&#39;anchor&#39;</span><span class="p">,</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>                       <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>                       <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">ticks</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">pval_df</span><span class="o">.</span><span class="n">index</span><span class="p">)))</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabels</span><span class="p">,</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>                       <span class="n">fontsize</span><span class="o">=</span><span class="n">tick_size</span><span class="p">,</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>                       <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                       <span class="p">)</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">invert_yaxis</span><span class="p">()</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                    <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                    <span class="n">bottom</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>                    <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                    <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>                    <span class="n">left</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>                    <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>                    <span class="n">labelbottom</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">xlabel</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">label_size</span><span class="p">)</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">label_size</span><span class="p">)</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="c1"># Colorbar</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="c1"># create an axes on the top side of ax. The width of cax will be 3%</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="c1"># of ax and the padding between cax and ax will be fixed at 0.21 inch.</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="n">divider</span> <span class="o">=</span> <span class="n">make_axes_locatable</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="n">cax</span> <span class="o">=</span> <span class="n">divider</span><span class="o">.</span><span class="n">append_axes</span><span class="p">(</span><span class="s2">&quot;top&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="s2">&quot;3%&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mf">0.21</span><span class="p">)</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">),</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>                        <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">,</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>                        <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;horizontal&#39;</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>                        <span class="p">)</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="n">cbar</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">labelsize</span><span class="o">=</span><span class="n">tick_size</span><span class="p">)</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="n">cax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span>  <span class="c1"># changes apply to the x-axis</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>                    <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>  <span class="c1"># both major and minor ticks are affected</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                    <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># ticks along the bottom edge are off</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>                    <span class="n">top</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># ticks along the top edge are off</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>                    <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># labels along the bottom edge are off</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>                    <span class="n">labeltop</span><span class="o">=</span><span class="kc">True</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>                    <span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>    <span class="n">cax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">cbar_title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">title_size</span><span class="p">)</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">pval_df</span><span class="p">)),</span> <span class="o">-</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="mf">0.99</span><span class="p">)]),</span> <span class="o">-</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">significance</span> <span class="o">+</span> <span class="mf">1e-9</span><span class="p">),</span> <span class="mf">3.0</span><span class="p">]):</span> <span class="c1"># old min np.min(np.min(pval_df))</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="n">ax2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="p">(</span><span class="n">max_size</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="o">*</span> <span class="n">tick_size</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>        <span class="n">ax2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>        <span class="k">if</span> <span class="n">v</span> <span class="o">==</span> <span class="mf">3.0</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>            <span class="n">extra</span> <span class="o">=</span> <span class="s1">&#39;&gt;=&#39;</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="k">elif</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>            <span class="n">extra</span> <span class="o">=</span> <span class="s1">&#39;Threshold: &#39;</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>            <span class="n">extra</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="n">ax2</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">extra</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">v</span><span class="p">),</span> <span class="mi">4</span><span class="p">)),</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">tick_size</span><span class="p">,</span> <span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>    <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>    <span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>    <span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;-log10(P-value) sizes&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">title_size</span><span class="p">)</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.tensor_plot">
        <code>tensor_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.generate_plot_df">
<code class="highlight language-python"><span class="n">generate_plot_df</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a melt dataframe with loadings for each element in all dimensions
across factors</p>
<h6 id="cell2cell.plotting.tensor_plot.generate_plot_df--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor</p>
<h6 id="cell2cell.plotting.tensor_plot.generate_plot_df--returns">Returns</h6>
<p>plot_df : pandas.DataFrame
    A dataframe containing loadings for every element of all dimensions across
    factors from the decomposition. Rows are loadings individual elements of each
    dimension in a given factor, while columns are the following list
    ['Factor', 'Variable', 'Value', 'Order']</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_plot_df</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Generates a melt dataframe with loadings for each element in all dimensions</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    across factors</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell2cell.tensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    plot_df : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A dataframe containing loadings for every element of all dimensions across</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        factors from the decomposition. Rows are loadings individual elements of each</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        dimension in a given factor, while columns are the following list</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        [&#39;Factor&#39;, &#39;Variable&#39;, &#39;Value&#39;, &#39;Order&#39;]</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">tensor_dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">if</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="k">if</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>            <span class="n">factor_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Context&#39;</span><span class="p">,</span> <span class="s1">&#39;LRs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver&#39;</span><span class="p">]</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>            <span class="n">factor_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Context-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">tensor_dim</span> <span class="o">-</span> <span class="mi">3</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;LRs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver&#39;</span><span class="p">]</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>            <span class="n">factor_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;LRs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver&#39;</span><span class="p">]</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Too few dimensions in the tensor&#39;</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">tensor_dim</span><span class="p">,</span> <span class="s2">&quot;The length of order_labels must match the number of orders/dimensions in the tensor&quot;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="n">factor_labels</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">plot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">order_factors</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">factors</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="n">sns_df</span> <span class="o">=</span> <span class="n">order_factors</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="n">sns_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Factors&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">melt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">sns_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span> <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Factors&#39;</span><span class="p">],</span> <span class="n">value_vars</span><span class="o">=</span><span class="n">sns_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">melt_df</span> <span class="o">=</span> <span class="n">melt_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Order</span><span class="o">=</span><span class="n">factor_labels</span><span class="p">[</span><span class="n">lab</span><span class="p">])</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">plot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_df</span><span class="p">,</span> <span class="n">melt_df</span><span class="p">])</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">plot_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Factor&#39;</span><span class="p">,</span> <span class="s1">&#39;Variable&#39;</span><span class="p">,</span> <span class="s1">&#39;Value&#39;</span><span class="p">,</span> <span class="s1">&#39;Order&#39;</span><span class="p">]</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">return</span> <span class="n">plot_df</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.plot_elbow">
<code class="highlight language-python"><span class="n">plot_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">,</span> <span class="n">elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Normalized Error&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots the errors of an elbow analysis with just one run of a tensor factorization
for each rank.</p>
<h6 id="cell2cell.plotting.tensor_plot.plot_elbow--parameters">Parameters</h6>
<p>loss : list
    List of  tuples with (x, y) coordinates for the elbow analysis. X values are
    the different ranks and Y values are the errors of each decomposition.</p>
<p>elbow : int, default=None
    X coordinate to color the error as red. Usually used to represent the detected
    elbow.</p>
<p>figsize : tuple, default=(4, 2.25)
    Figure size, width by height</p>
<p>ylabel : str, default='Normalized Error'
    Label for the y-axis</p>
<p>fontsize : int, default=14
    Fontsize for axis labels.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.tensor_plot.plot_elbow--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">plot_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">,</span> <span class="n">elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Normalized Error&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Plots the errors of an elbow analysis with just one run of a tensor factorization</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    for each rank.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    loss : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of  tuples with (x, y) coordinates for the elbow analysis. X values are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the different ranks and Y values are the errors of each decomposition.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    elbow : int, default=None</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        X coordinate to color the error as red. Usually used to represent the detected</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        elbow.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    figsize : tuple, default=(4, 2.25)</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Figure size, width by height</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    ylabel : str, default=&#39;Normalized Error&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Label for the y-axis</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Fontsize for axis labels.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">*</span><span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">loss</span><span class="p">))</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Rank&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span> <span class="o">*</span> <span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">if</span> <span class="n">elbow</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">*</span><span class="n">loss</span><span class="p">[</span><span class="n">elbow</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;ro&#39;</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.plot_multiple_run_elbow">
<code class="highlight language-python"><span class="n">plot_multiple_run_elbow</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ci</span><span class="o">=</span><span class="s1">&#39;95%&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Normalized Error&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots the errors of an elbow analysis with multiple runs of a tensor
factorization for each rank.</p>
<h6 id="cell2cell.plotting.tensor_plot.plot_multiple_run_elbow--parameters">Parameters</h6>
<p>all_loss : ndarray
    Array containing the errors associated with multiple runs for a given rank.
    This array is of shape (runs, upper_rank).</p>
<p>elbow : int, default=None
    X coordinate to color the error as red. Usually used to represent the detected
    elbow.</p>
<p _95_="'95%'" _std_="'std',">ci : str, default='std'
    Confidence interval for representing the multiple runs in each rank.</p>
<p>figsize : tuple, default=(4, 2.25)
    Figure size, width by height</p>
<p>ylabel : str, default='Normalized Error'
    Label for the y-axis</p>
<p>fontsize : int, default=14
    Fontsize for axis labels.</p>
<p>smooth : boolean, default=False
    Whether smoothing the curve with a Savitzky-Golay filter.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.tensor_plot.plot_multiple_run_elbow--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">plot_multiple_run_elbow</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ci</span><span class="o">=</span><span class="s1">&#39;95%&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Normalized Error&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                            <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Plots the errors of an elbow analysis with multiple runs of a tensor</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    factorization for each rank.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    all_loss : ndarray</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Array containing the errors associated with multiple runs for a given rank.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        This array is of shape (runs, upper_rank).</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    elbow : int, default=None</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        X coordinate to color the error as red. Usually used to represent the detected</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        elbow.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ci : str, default=&#39;std&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Confidence interval for representing the multiple runs in each rank.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        {&#39;std&#39;, &#39;95%&#39;}</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    figsize : tuple, default=(4, 2.25)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Figure size, width by height</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    ylabel : str, default=&#39;Normalized Error&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Label for the y-axis</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Fontsize for axis labels.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    smooth : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether smoothing the curve with a Savitzky-Golay filter.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    -------</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">x</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">all_loss</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">std</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanstd</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">mean</span> <span class="o">=</span> <span class="n">smooth_curve</span><span class="p">(</span><span class="n">mean</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="c1"># Plot Mean</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="s1">&#39;ob&#39;</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="c1"># Plot CI</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">if</span> <span class="n">ci</span> <span class="o">==</span> <span class="s1">&#39;95%&#39;</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">coeff</span> <span class="o">=</span> <span class="mf">1.96</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">elif</span> <span class="n">ci</span> <span class="o">==</span> <span class="s1">&#39;std&#39;</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="n">coeff</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Specify a correct ci. Either &#39;95%&#39; or &#39;std&#39;&quot;</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mean</span><span class="o">-</span><span class="n">coeff</span><span class="o">*</span><span class="n">std</span><span class="p">,</span> <span class="n">mean</span><span class="o">+</span><span class="n">coeff</span><span class="o">*</span><span class="n">std</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;steelblue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>                     <span class="n">label</span><span class="o">=</span><span class="s1">&#39;$\pm$ 1 std&#39;</span><span class="p">)</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Rank&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span> <span class="o">*</span> <span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="k">if</span> <span class="n">elbow</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>        <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">elbow</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="n">mean</span><span class="p">[</span><span class="n">elbow</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;ro&#39;</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.reorder_dimension_elements">
<code class="highlight language-python"><span class="n">reorder_dimension_elements</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Reorders elements in the dataframes including factor loadings.</p>
<h6 id="cell2cell.plotting.tensor_plot.reorder_dimension_elements--parameters">Parameters</h6>
<p>factors : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor.</p>
<p>reorder_elements : dict, default=None
    Dictionary for reordering elements in each of the tensor dimension.
    Keys of this dictionary could be any or all of the keys in
    interaction_tensor.factors. Values are list with the names or labels of the
    elements in a tensor dimension. For example, for the context dimension,
    all elements included in interaction_tensor.factors['Context'].index must
    be present.</p>
<p>metadata : list, default=None
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element.</p>
<h6 id="cell2cell.plotting.tensor_plot.reorder_dimension_elements--returns">Returns</h6>
<p>reordered_factors : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor. This dictionary includes the new orders.</p>
<p>new_metadata : list, default=None
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element. In this case, elements are sorted according to reorder_elements.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">reorder_dimension_elements</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Reorders elements in the dataframes including factor loadings.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    factors : dict</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        dimension/order of the tensor.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    reorder_elements : dict, default=None</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Dictionary for reordering elements in each of the tensor dimension.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Keys of this dictionary could be any or all of the keys in</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        interaction_tensor.factors. Values are list with the names or labels of the</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        elements in a tensor dimension. For example, for the context dimension,</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        all elements included in interaction_tensor.factors[&#39;Context&#39;].index must</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        be present.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    metadata : list, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        element.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    -------</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    reordered_factors : dict</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        dimension/order of the tensor. This dictionary includes the new orders.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    new_metadata : list, default=None</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        element. In this case, elements are sorted according to reorder_elements.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">k</span> <span class="ow">in</span> <span class="n">factors</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">reorder_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s2">&quot;Keys in &#39;reorder_elements&#39; must be only keys in &#39;factors&#39;&quot;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">factors</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">reorder_elements</span><span class="p">[</span><span class="n">key</span><span class="p">])))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">reorder_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s2">&quot;All elements of each dimension included should be present&quot;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">reordered_factors</span> <span class="o">=</span> <span class="n">factors</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">new_metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">reordered_factors</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">reorder_elements</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">reorder_elements</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>            <span class="n">reordered_factors</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)]</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="k">if</span> <span class="n">new_metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>                <span class="n">meta</span> <span class="o">=</span> <span class="n">new_metadata</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Element&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Element&#39;</span><span class="p">],</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">categories</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">reordered_factors</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                <span class="n">new_metadata</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>            <span class="n">reordered_factors</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">return</span> <span class="n">reordered_factors</span><span class="p">,</span> <span class="n">new_metadata</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.tensor_factors_plot">
<code class="highlight language-python"><span class="n">tensor_factors_plot</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">plot_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots the loadings for each element in each dimension of the tensor, generate by
a tensor factorization.</p>
<h6 id="cell2cell.plotting.tensor_plot.tensor_factors_plot--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor.</p>
<p>order_labels : list, default=None
    List with the labels of each dimension to use in the plot. If none, the
    default names given when factorizing the tensor will be used.</p>
<p>reorder_elements : dict, default=None
    Dictionary for reordering elements in each of the tensor dimension.
    Keys of this dictionary could be any or all of the keys in
    interaction_tensor.factors. Values are list with the names or labels of the
    elements in a tensor dimension. For example, for the context dimension,
    all elements included in interaction_tensor.factors['Context'].index must
    be present.</p>
<p>metadata : list, default=None
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element.</p>
<p>sample_col : str, default='Element'
    Name of the column containing the element names in the metadata.</p>
<p>group_col : str, default='Category'
    Name of the column containing the metadata or grouping information for each
    element in the metadata.</p>
<p>meta_cmaps : list, default=None
    A list of colormaps used for coloring elements in each dimension. The length
    of this list is equal to the number of dimensions of the tensor. If None, all
    dimensions will be colores with the colormap 'gist_rainbow'.</p>
<p>fontsize : int, default=20
    Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</p>
<p>plot_legend : boolean, default=True
    Whether plotting the legends for the coloring of each element in their
    respective dimensions.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is
    not saved.</p>
<h6 id="cell2cell.plotting.tensor_plot.tensor_factors_plot--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>
<p>axes : matplotlib.axes.Axes or array of Axes
    List of Axes for each subplot in the figure.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">tensor_factors_plot</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                        <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">plot_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                        <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Plots the loadings for each element in each dimension of the tensor, generate by</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    a tensor factorization.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell2cell.tensor.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    order_labels : list, default=None</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        List with the labels of each dimension to use in the plot. If none, the</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        default names given when factorizing the tensor will be used.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    reorder_elements : dict, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Dictionary for reordering elements in each of the tensor dimension.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Keys of this dictionary could be any or all of the keys in</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        interaction_tensor.factors. Values are list with the names or labels of the</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        elements in a tensor dimension. For example, for the context dimension,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        all elements included in interaction_tensor.factors[&#39;Context&#39;].index must</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        be present.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    metadata : list, default=None</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        element.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    sample_col : str, default=&#39;Element&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Name of the column containing the element names in the metadata.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    group_col : str, default=&#39;Category&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Name of the column containing the metadata or grouping information for each</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        element in the metadata.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    meta_cmaps : list, default=None</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        A list of colormaps used for coloring elements in each dimension. The length</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        of this list is equal to the number of dimensions of the tensor. If None, all</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        dimensions will be colores with the colormap &#39;gist_rainbow&#39;.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    fontsize : int, default=20</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    plot_legend : boolean, default=True</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        Whether plotting the legends for the coloring of each element in their</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        respective dimensions.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        not saved.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    -------</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    axes : matplotlib.axes.Axes or array of Axes</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        List of Axes for each subplot in the figure.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="c1"># Prepare inputs for matplotlib</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="k">assert</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">factors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;First run the method &#39;compute_tensor_factorization&#39; in your InteractionTensor&quot;</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">factors</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">order_labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="k">assert</span> <span class="n">dim</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">order_labels</span><span class="p">),</span> <span class="s2">&quot;The lenght of factor_labels must match the order of the tensor (order </span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dim</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>        <span class="n">order_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">factors</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">rank</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">rank</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">tensor_factors_plot_from_loadings</span><span class="p">(</span><span class="n">factors</span><span class="o">=</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">factors</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                  <span class="n">rank</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                  <span class="n">order_labels</span><span class="o">=</span><span class="n">order_labels</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                                  <span class="n">reorder_elements</span><span class="o">=</span><span class="n">reorder_elements</span><span class="p">,</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                                                  <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                                                  <span class="n">sample_col</span><span class="o">=</span><span class="n">sample_col</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                  <span class="n">group_col</span><span class="o">=</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                                                  <span class="n">meta_cmaps</span><span class="o">=</span><span class="n">meta_cmaps</span><span class="p">,</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>                                                  <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                                                  <span class="n">plot_legend</span><span class="o">=</span><span class="n">plot_legend</span><span class="p">,</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                                                  <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.tensor_plot.tensor_factors_plot_from_loadings">
<code class="highlight language-python"><span class="n">tensor_factors_plot_from_loadings</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">plot_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots the loadings for each element in each dimension of the tensor, generate by
a tensor factorization.</p>
<h6 id="cell2cell.plotting.tensor_plot.tensor_factors_plot_from_loadings--parameters">Parameters</h6>
<p>factors : collections.OrderedDict
    An ordered dictionary wherein keys are the names of each
    tensor dimension, and values are the loadings in a pandas.DataFrame.
    In this dataframe, rows are the elements of the respective dimension
    and columns are the factors from the tensor factorization. Values
    are the corresponding loadings.</p>
<p>rank : int, default=None
    Number of factors generated from the decomposition</p>
<p>order_labels : list, default=None
    List with the labels of each dimension to use in the plot. If none, the
    default names given when factorizing the tensor will be used.</p>
<p>reorder_elements : dict, default=None
    Dictionary for reordering elements in each of the tensor dimension.
    Keys of this dictionary could be any or all of the keys in
    interaction_tensor.factors. Values are list with the names or labels of the
    elements in a tensor dimension. For example, for the context dimension,
    all elements included in interaction_tensor.factors['Context'].index must
    be present.</p>
<p>metadata : list, default=None
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element.</p>
<p>sample_col : str, default='Element'
    Name of the column containing the element names in the metadata.</p>
<p>group_col : str, default='Category'
    Name of the column containing the metadata or grouping information for each
    element in the metadata.</p>
<p>meta_cmaps : list, default=None
    A list of colormaps used for coloring elements in each dimension. The length
    of this list is equal to the number of dimensions of the tensor. If None, all
    dimensions will be colores with the colormap 'gist_rainbow'.</p>
<p>fontsize : int, default=20
    Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</p>
<p>plot_legend : boolean, default=True
    Whether plotting the legends for the coloring of each element in their
    respective dimensions.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is
    not saved.</p>
<h6 id="cell2cell.plotting.tensor_plot.tensor_factors_plot_from_loadings--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>
<p>axes : matplotlib.axes.Axes or array of Axes
    List of Axes for each subplot in the figure.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/tensor_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">tensor_factors_plot_from_loadings</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reorder_elements</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                      <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">,</span> <span class="n">group_col</span><span class="o">=</span><span class="s1">&#39;Category&#39;</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">plot_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                                      <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Plots the loadings for each element in each dimension of the tensor, generate by</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    a tensor factorization.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    factors : collections.OrderedDict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        An ordered dictionary wherein keys are the names of each</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        tensor dimension, and values are the loadings in a pandas.DataFrame.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        In this dataframe, rows are the elements of the respective dimension</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        and columns are the factors from the tensor factorization. Values</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        are the corresponding loadings.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    rank : int, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Number of factors generated from the decomposition</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    order_labels : list, default=None</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        List with the labels of each dimension to use in the plot. If none, the</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        default names given when factorizing the tensor will be used.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    reorder_elements : dict, default=None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Dictionary for reordering elements in each of the tensor dimension.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Keys of this dictionary could be any or all of the keys in</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        interaction_tensor.factors. Values are list with the names or labels of the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        elements in a tensor dimension. For example, for the context dimension,</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        all elements included in interaction_tensor.factors[&#39;Context&#39;].index must</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        be present.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    metadata : list, default=None</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        element.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    sample_col : str, default=&#39;Element&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Name of the column containing the element names in the metadata.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    group_col : str, default=&#39;Category&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Name of the column containing the metadata or grouping information for each</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        element in the metadata.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    meta_cmaps : list, default=None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        A list of colormaps used for coloring elements in each dimension. The length</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        of this list is equal to the number of dimensions of the tensor. If None, all</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        dimensions will be colores with the colormap &#39;gist_rainbow&#39;.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    fontsize : int, default=20</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Font size of the tick labels. Axis labels will be 1.2 times the fontsize.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    plot_legend : boolean, default=True</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Whether plotting the legends for the coloring of each element in their</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        respective dimensions.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        not saved.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">    -------</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    axes : matplotlib.axes.Axes or array of Axes</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        List of Axes for each subplot in the figure.</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="c1"># Prepare inputs for matplotlib</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">rank</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="k">assert</span> <span class="nb">list</span><span class="p">(</span><span class="n">factors</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">rank</span><span class="p">,</span> <span class="s2">&quot;Rank must match the number of columns in dataframes in `factors`&quot;</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">rank</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">factors</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="n">dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">factors</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">if</span> <span class="n">order_labels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="k">assert</span> <span class="n">dim</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">order_labels</span><span class="p">),</span> <span class="s2">&quot;The length of factor_labels must match the order of the tensor (order </span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dim</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="n">order_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">factors</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="n">meta_og</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="k">if</span> <span class="n">reorder_elements</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">factors</span><span class="p">,</span> <span class="n">metadata</span> <span class="o">=</span> <span class="n">reorder_dimension_elements</span><span class="p">(</span><span class="n">factors</span><span class="o">=</span><span class="n">factors</span><span class="p">,</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>                                                       <span class="n">reorder_elements</span><span class="o">=</span><span class="n">reorder_elements</span><span class="p">,</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                                                       <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="k">if</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>        <span class="n">metadata</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">dim</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="n">meta_colors</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">dim</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>        <span class="n">element_colors</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">dim</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="k">if</span> <span class="n">meta_cmaps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>            <span class="n">meta_cmaps</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gist_rainbow&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">meta_cmaps</span><span class="p">),</span> <span class="s2">&quot;Provide a cmap for each order&quot;</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">factors</span><span class="p">),</span> <span class="s2">&quot;Provide a metadata for each order. If there is no metadata for any, replace with None&quot;</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="n">meta_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">get_colors_from_labels</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="n">group_col</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span> <span class="k">if</span> <span class="p">((</span><span class="n">m</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">cmap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">cmap</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">meta_og</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="p">)]</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="n">element_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">map_colors_to_metadata</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">m</span><span class="p">,</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                                                 <span class="n">colors</span><span class="o">=</span><span class="n">mc</span><span class="p">,</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                                                 <span class="n">sample_col</span><span class="o">=</span><span class="n">sample_col</span><span class="p">,</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>                                                 <span class="n">group_col</span><span class="o">=</span><span class="n">group_col</span><span class="p">,</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                                                 <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span> <span class="k">if</span> <span class="p">((</span><span class="n">m</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">cmap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">))</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">cmap</span><span class="p">,</span> <span class="n">mc</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">meta_cmaps</span><span class="p">,</span> <span class="n">meta_colors</span><span class="p">)]</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="c1"># Make the plot</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>                             <span class="n">ncols</span><span class="o">=</span><span class="n">dim</span><span class="p">,</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>                             <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">rank</span> <span class="o">*</span> <span class="mf">1.2</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)),</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>                             <span class="n">sharex</span><span class="o">=</span><span class="s1">&#39;col&#39;</span><span class="p">,</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>                             <span class="c1">#sharey=&#39;col&#39;</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>                             <span class="p">)</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">rank</span><span class="p">,</span> <span class="n">dim</span><span class="p">))</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="c1"># Factor by factor</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">if</span> <span class="n">rank</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="c1"># Iterates horizontally (dimension by dimension)</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="p">(</span><span class="n">order_factors</span><span class="p">,</span> <span class="n">axs</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">factors</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">axes</span><span class="o">.</span><span class="n">T</span><span class="p">)):</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">order_factors</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>                <span class="n">order_factors</span> <span class="o">=</span> <span class="n">order_factors</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>            <span class="c1"># Iterates vertically (factor by factor)</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">df_row</span><span class="p">,</span> <span class="n">ax</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">order_factors</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">iterrows</span><span class="p">(),</span> <span class="n">axs</span><span class="p">)):</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>                <span class="n">factor_name</span> <span class="o">=</span> <span class="n">df_row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                <span class="n">factor</span> <span class="o">=</span> <span class="n">df_row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>                <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">top</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>                <span class="k">if</span> <span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>                    <span class="n">plot_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">element_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">order_factors</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">factor</span><span class="p">)),</span> <span class="n">factor</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">plot_colors</span><span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">factor</span><span class="p">)),</span> <span class="n">factor</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>                <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">factor_name</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>                <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">axs</span><span class="p">):</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>                    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">visible</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>            <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">order_labels</span><span class="p">[</span><span class="n">ind</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">),</span> <span class="n">labelpad</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">order_factors</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">factors</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">order_factors</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>                <span class="n">order_factors</span> <span class="o">=</span> <span class="n">order_factors</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>            <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">order_labels</span><span class="p">[</span><span class="n">ind</span><span class="p">],</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">),</span> <span class="n">labelpad</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">df_row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">order_factors</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()):</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                <span class="n">factor_name</span> <span class="o">=</span> <span class="n">df_row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                <span class="n">factor</span> <span class="o">=</span> <span class="n">df_row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">top</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                <span class="k">if</span> <span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                    <span class="n">plot_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">element_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">order_factors</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">factor</span><span class="p">)),</span> <span class="n">factor</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">plot_colors</span><span class="p">)</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">factor</span><span class="p">)),</span> <span class="n">factor</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>                <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">factor_name</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1.2</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="n">fig</span><span class="o">.</span><span class="n">align_ylabels</span><span class="p">(</span><span class="n">axes</span><span class="p">[:,</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>    <span class="c1"># Include legends of coloring the elements in each dimension.</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>    <span class="k">if</span> <span class="n">plot_legend</span><span class="p">:</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>        <span class="c1"># Set current axis:</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="c1"># Legends</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="n">renderer</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">get_renderer</span><span class="p">()</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">bbox_cords</span> <span class="o">=</span>  <span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">)</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>        <span class="n">N</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">order_labels</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>        <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">order</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">order_labels</span><span class="p">):</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>            <span class="k">if</span> <span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                <span class="n">lgd</span> <span class="o">=</span> <span class="n">generate_legend</span><span class="p">(</span><span class="n">color_dict</span><span class="o">=</span><span class="n">meta_colors</span><span class="p">[</span><span class="n">ind</span><span class="p">],</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>                                      <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="n">bbox_cords</span><span class="p">,</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                                      <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>                                      <span class="n">title</span><span class="o">=</span><span class="n">order_labels</span><span class="p">[</span><span class="n">ind</span><span class="p">],</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>                                      <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>                                      <span class="n">sorted_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>                                      <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>                                      <span class="p">)</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>                <span class="n">cords</span> <span class="o">=</span> <span class="n">lgd</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span><span class="o">.</span><span class="n">transformed</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">transAxes</span><span class="o">.</span><span class="n">inverted</span><span class="p">())</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>                <span class="n">xrange</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">cords</span><span class="o">.</span><span class="n">p0</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">cords</span><span class="o">.</span><span class="n">p1</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>                <span class="n">bbox_cords</span> <span class="o">=</span> <span class="p">(</span><span class="n">bbox_cords</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">xrange</span> <span class="o">+</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">bbox_cords</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>                <span class="k">if</span> <span class="n">ind</span> <span class="o">!=</span> <span class="n">N</span><span class="p">:</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                    <span class="n">ax</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">lgd</span><span class="p">)</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                    <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.plotting.umap_plot">
        <code>umap_plot</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.plotting.umap_plot.umap_biplot">
<code class="highlight language-python"><span class="n">umap_biplot</span><span class="p">(</span><span class="n">umap_df</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab10&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Plots a UMAP biplot for the UMAP embeddings.</p>
<h6 id="cell2cell.plotting.umap_plot.umap_biplot--parameters">Parameters</h6>
<p>umap_df : pandas.DataFrame
    Dataframe containing the UMAP embeddings for the axis analyzed.
    It must contain columns 'umap1 and 'umap2'. If a hue column is
    provided in the parameter 'hue', that column must be provided
    in this dataframe.</p>
<p>figsize : tuple, default=(8, 8)
    Size of the figure (width*height), each in inches.</p>
<p>ax : matplotlib.axes.Axes, default=None
    The matplotlib axes containing a plot.</p>
<p>show_axes : boolean, default=True
    Whether showing lines, ticks and ticklabels of both axes.</p>
<p>show_legend : boolean, default=True
    Whether including the legend when a hue is provided.</p>
<p>hue : vector or key in 'umap_df'
    Grouping variable that will produce points with different colors.
    Can be either categorical or numeric, although color mapping will
    behave differently in latter case.</p>
<p>cmap : str, default='tab10'
    Name of the color palette for coloring elements with UMAP embeddings.</p>
<p>fontsize : int, default=20
    Fontsize of the axis labels (UMAP1 and UMAP2).</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<h6 id="cell2cell.plotting.umap_plot.umap_biplot--returns">Returns</h6>
<div class="codehilite"><pre><span></span><code><span class="n">fig</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">matplotlib</span><span class="o">.</span><span class="na">figure</span><span class="o">.</span><span class="na">Figure</span><span class="w"></span>
<span class="w">    </span><span class="n">A</span><span class="w"> </span><span class="n">matplotlib</span><span class="w"> </span><span class="n">Figure</span><span class="w"> </span><span class="n">instance</span><span class="o">.</span><span class="w"></span>

<span class="n">ax</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">matplotlib</span><span class="o">.</span><span class="na">axes</span><span class="o">.</span><span class="na">Axes</span><span class="w"></span>
<span class="w">    </span><span class="n">The</span><span class="w"> </span><span class="n">matplotlib</span><span class="w"> </span><span class="n">axes</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">plot</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/plotting/umap_plot.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">umap_biplot</span><span class="p">(</span><span class="n">umap_df</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span> <span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab10&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Plots a UMAP biplot for the UMAP embeddings.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    umap_df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Dataframe containing the UMAP embeddings for the axis analyzed.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        It must contain columns &#39;umap1 and &#39;umap2&#39;. If a hue column is</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        provided in the parameter &#39;hue&#39;, that column must be provided</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        in this dataframe.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    figsize : tuple, default=(8, 8)</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Size of the figure (width*height), each in inches.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    ax : matplotlib.axes.Axes, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The matplotlib axes containing a plot.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    show_axes : boolean, default=True</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Whether showing lines, ticks and ticklabels of both axes.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    show_legend : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether including the legend when a hue is provided.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    hue : vector or key in &#39;umap_df&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Grouping variable that will produce points with different colors.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Can be either categorical or numeric, although color mapping will</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        behave differently in latter case.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    cmap : str, default=&#39;tab10&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Name of the color palette for coloring elements with UMAP embeddings.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    fontsize : int, default=20</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Fontsize of the axis labels (UMAP1 and UMAP2).</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    -------</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">            A matplotlib Figure instance.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        ax : matplotlib.axes.Axes</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">            The matplotlib axes containing the plot.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;umap1&#39;</span><span class="p">,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                         <span class="n">y</span><span class="o">=</span><span class="s1">&#39;umap2&#39;</span><span class="p">,</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>                         <span class="n">data</span><span class="o">=</span><span class="n">umap_df</span><span class="p">,</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>                         <span class="n">hue</span><span class="o">=</span><span class="n">hue</span><span class="p">,</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>                         <span class="n">palette</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>                         <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>                         <span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">if</span> <span class="n">show_axes</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>                    <span class="n">offset</span><span class="o">=</span><span class="mi">15</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                    <span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                       <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                       <span class="n">colors</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                       <span class="n">width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                       <span class="n">length</span><span class="o">=</span><span class="mi">5</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                       <span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>            <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="k">for</span> <span class="n">tick</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">():</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_fontproperties</span><span class="p">(</span><span class="s1">&#39;arial&#39;</span><span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_weight</span><span class="p">(</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.7</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="k">for</span> <span class="n">tick</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">():</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_fontproperties</span><span class="p">(</span><span class="s1">&#39;arial&#39;</span><span class="p">)</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_weight</span><span class="p">(</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>        <span class="n">tick</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.7</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;UMAP 2&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">show_legend</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">hue</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>        <span class="c1"># Put the legend out of the figure</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>        <span class="n">legend</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mf">0.</span><span class="p">)</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>        <span class="n">legend</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">hue</span><span class="p">)</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="n">legend</span><span class="o">.</span><span class="n">get_title</span><span class="p">()</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.7</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>        <span class="k">for</span> <span class="n">text</span> <span class="ow">in</span> <span class="n">legend</span><span class="o">.</span><span class="n">get_texts</span><span class="p">():</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>            <span class="n">text</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.7</span><span class="o">*</span><span class="n">fontsize</span><span class="p">))</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.preprocessing">
        <code>preprocessing</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.cutoffs">
        <code>cutoffs</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.cutoffs.get_constant_cutoff">
<code class="highlight language-python"><span class="n">get_constant_cutoff</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">constant_cutoff</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a cutoff/threshold dataframe for all genes
in rnaseq_data assigning a constant value as the cutoff.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_constant_cutoff--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>constant_cutoff : float, default=10
    Cutoff or threshold assigned to each gene.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_constant_cutoff--returns">Returns</h6>
<p>cutoffs : pandas.DataFrame
    A dataframe containing the value corresponding to cutoff or threshold
    assigned to each gene. Rows are genes and the column corresponds to
    'value'. All values are the same and corresponds to the
    constant_cutoff.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/cutoffs.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_constant_cutoff</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">constant_cutoff</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Generates a cutoff/threshold dataframe for all genes</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in rnaseq_data assigning a constant value as the cutoff.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    constant_cutoff : float, default=10</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Cutoff or threshold assigned to each gene.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A dataframe containing the value corresponding to cutoff or threshold</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        assigned to each gene. Rows are genes and the column corresponds to</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        &#39;value&#39;. All values are the same and corresponds to the</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        constant_cutoff.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">cutoffs</span><span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">constant_cutoff</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">return</span> <span class="n">cutoffs</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.cutoffs.get_cutoffs">
<code class="highlight language-python"><span class="n">get_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">parameters</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This function creates cutoff/threshold values for genes
in rnaseq_data and the respective cells/tissues/samples
by a given method or parameter.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_cutoffs--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>parameters : dict
    This dictionary must contain a 'parameter' key and a 'type' key.
    The first one is the respective parameter to compute the threshold
    or cutoff values. The type corresponds to the approach to
    compute the values according to the parameter employed.
    Options of 'type' that can be used:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"></span>
<span class="w">                       </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"></span>
<span class="w">                       </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"></span>
<span class="w">                       </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"></span>
<span class="w">                        </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"></span>
<span class="w">                        </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">                        </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"></span>
<span class="w">                        </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"></span>
<span class="w">           </span><span class="n">the</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"></span>
<span class="w">           </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"></span>
<span class="w">                       </span><span class="n">cutoff</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"></span>
<span class="w">                       </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"></span>
<span class="w">                       </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">                       </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"></span>
<span class="w">                        </span><span class="n">cutoff</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"></span>
<span class="w">                        </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">                     </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"></span>
<span class="w">                     </span><span class="n">the</span><span class="w"> </span><span class="s1">&#39;parameter&#39;</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_cutoffs--returns">Returns</h6>
<p>cutoffs : pandas.DataFrame
    Dataframe wherein rows are genes in rnaseq_data. Depending on the type in
    the parameters dictionary, it may have only one column ('value') or the
    same columns that rnaseq_data has, generating specfic cutoffs for each
    cell/tissue/sample.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/cutoffs.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">parameters</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    This function creates cutoff/threshold values for genes</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in rnaseq_data and the respective cells/tissues/samples</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    by a given method or parameter.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    parameters : dict</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        This dictionary must contain a &#39;parameter&#39; key and a &#39;type&#39; key.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        The first one is the respective parameter to compute the threshold</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        or cutoff values. The type corresponds to the approach to</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        compute the values according to the parameter employed.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Options of &#39;type&#39; that can be used:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;local_percentile&#39; : computes the value of a given percentile,</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">                               for each gene independently. In this case,</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                               the parameter corresponds to the percentile</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                               to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;global_percentile&#39; : computes the value of a given percentile</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                from all genes and samples simultaneously.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                                In this case, the parameter corresponds to</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                                the percentile to compute, as a float value</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                                between 0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                   the path of that file. It must contain the same genes as</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                   index and same samples as columns.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                               cutoff for each gene in each sample. This allows</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">                               to use specific cutoffs for each sample. The</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">                               columns here must be the same as the ones in the</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">                               rnaseq_data.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">                                cutoff for each gene in only one column. These</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">                                cutoffs will be applied to all samples.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">                             expression is greater than the value input in</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">                             the &#39;parameter&#39;.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    -------</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Dataframe wherein rows are genes in rnaseq_data. Depending on the type in</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        the parameters dictionary, it may have only one column (&#39;value&#39;) or the</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        same columns that rnaseq_data has, generating specfic cutoffs for each</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        cell/tissue/sample.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">parameter</span> <span class="o">=</span> <span class="n">parameters</span><span class="p">[</span><span class="s1">&#39;parameter&#39;</span><span class="p">]</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="nb">type</span> <span class="o">=</span> <span class="n">parameters</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Calculating cutoffs for gene abundances&quot;</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;local_percentile&#39;</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">get_local_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">parameter</span><span class="p">)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;global_percentile&#39;</span><span class="p">:</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">get_global_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">parameter</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;constant_value&#39;</span><span class="p">:</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">get_constant_cutoff</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">parameter</span><span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;file&#39;</span><span class="p">:</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">read_data</span><span class="o">.</span><span class="n">load_cutoffs</span><span class="p">(</span><span class="n">parameter</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                         <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">cutoffs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;multi_col_matrix&#39;</span><span class="p">:</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">parameter</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">cutoffs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">cutoffs</span><span class="p">[</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;single_col_matrix&#39;</span><span class="p">:</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">parameter</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">cutoffs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="nb">type</span> <span class="o">+</span> <span class="s1">&#39; is not a valid cutoff&#39;</span><span class="p">)</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>    <span class="k">return</span> <span class="n">cutoffs</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.cutoffs.get_global_percentile_cutoffs">
<code class="highlight language-python"><span class="n">get_global_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">percentile</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains a global value associated with a given percentile across
cells/tissues/samples and genes in a rnaseq_data.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_global_percentile_cutoffs--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>percentile : float, default=0.75
    This is the percentile to be computed.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_global_percentile_cutoffs--returns">Returns</h6>
<p>cutoffs : pandas.DataFrame
    A dataframe containing the value corresponding to the percentile
    across the dataset. Rows are genes and the column corresponds to
    'value'. All values here are the same global percentile.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/cutoffs.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_global_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">percentile</span><span class="o">=</span><span class="mf">0.75</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Obtains a global value associated with a given percentile across</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cells/tissues/samples and genes in a rnaseq_data.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    percentile : float, default=0.75</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        This is the percentile to be computed.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A dataframe containing the value corresponding to the percentile</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        across the dataset. Rows are genes and the column corresponds to</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        &#39;value&#39;. All values here are the same global percentile.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">])</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">cutoffs</span><span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">percentile</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">cutoffs</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.cutoffs.get_local_percentile_cutoffs">
<code class="highlight language-python"><span class="n">get_local_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">percentile</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains a local value associated with a given percentile across
cells/tissues/samples for each gene in a rnaseq_data.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_local_percentile_cutoffs--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>percentile : float, default=0.75
    This is the percentile to be computed.</p>
<h6 id="cell2cell.preprocessing.cutoffs.get_local_percentile_cutoffs--returns">Returns</h6>
<p>cutoffs : pandas.DataFrame
    A dataframe containing the value corresponding to the percentile
    across the genes. Rows are genes and the column corresponds to
    'value'.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/cutoffs.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_local_percentile_cutoffs</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">percentile</span><span class="o">=</span><span class="mf">0.75</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Obtains a local value associated with a given percentile across</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cells/tissues/samples for each gene in a rnaseq_data.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    percentile : float, default=0.75</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        This is the percentile to be computed.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A dataframe containing the value corresponding to the percentile</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        across the genes. Rows are genes and the column corresponds to</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        &#39;value&#39;.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">cutoffs</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">percentile</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">cutoffs</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.find_elements">
        <code>find_elements</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.find_elements.find_duplicates">
<code class="highlight language-python"><span class="n">find_duplicates</span><span class="p">(</span><span class="n">element_list</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Function based on: https://stackoverflow.com/a/5419576/12032899
Finds duplicate items and list their index location.</p>
<h6 id="cell2cell.preprocessing.find_elements.find_duplicates--parameters">Parameters</h6>
<p>element_list : list
    List of elements</p>
<h6 id="cell2cell.preprocessing.find_elements.find_duplicates--returns">Returns</h6>
<p>duplicate_dict : dict
    Dictionary with duplicate items. Keys are the items, and values
    are lists with the respective indexes where they are.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/find_elements.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">find_duplicates</span><span class="p">(</span><span class="n">element_list</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Function based on: https://stackoverflow.com/a/5419576/12032899</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Finds duplicate items and list their index location.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    element_list : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of elements</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    -------</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    duplicate_dict : dict</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Dictionary with duplicate items. Keys are the items, and values</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        are lists with the respective indexes where they are.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="n">tally</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">element_list</span><span class="p">):</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">tally</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">duplicate_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span> <span class="p">:</span> <span class="n">locs</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span><span class="n">locs</span> <span class="ow">in</span> <span class="n">tally</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>                            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">locs</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">}</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">duplicate_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.find_elements.get_element_abundances">
<code class="highlight language-python"><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the fraction of occurrence of each element
in a list of lists.</p>
<h6 id="cell2cell.preprocessing.find_elements.get_element_abundances--parameters">Parameters</h6>
<p>element_lists : list
    List of lists of elements. Elements will be
    counted only once in each of the lists.</p>
<h6 id="cell2cell.preprocessing.find_elements.get_element_abundances--returns">Returns</h6>
<p>abundance_dict : dict
    Dictionary containing the number of times that an
    element was present, divided by the total number of
    lists in <code>element_lists</code>.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/find_elements.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the fraction of occurrence of each element</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    in a list of lists.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    element_lists : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of lists of elements. Elements will be</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        counted only once in each of the lists.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    abundance_dict : dict</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Dictionary containing the number of times that an</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        element was present, divided by the total number of</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        lists in `element_lists`.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">abundance_dict</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">element_lists</span><span class="p">)))</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">element_lists</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">abundance_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span> <span class="p">:</span> <span class="n">v</span><span class="o">/</span><span class="n">total</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">abundance_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">return</span> <span class="n">abundance_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.find_elements.get_elements_over_fraction">
<code class="highlight language-python"><span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="p">,</span> <span class="n">fraction</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains a list of elements with the
fraction of occurrence at least the threshold.</p>
<h6 id="cell2cell.preprocessing.find_elements.get_elements_over_fraction--parameters">Parameters</h6>
<p>abundance_dict : dict
    Dictionary containing the number of times that an
    element was present, divided by the total number of
    possible occurrences.</p>
<p>fraction : float
    Threshold to filter the elements. Elements with at least
    this threshold will be included.</p>
<h6 id="cell2cell.preprocessing.find_elements.get_elements_over_fraction--returns">Returns</h6>
<p>elements : list
    List of elements that met the fraction criteria.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/find_elements.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="p">,</span> <span class="n">fraction</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Obtains a list of elements with the</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    fraction of occurrence at least the threshold.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    abundance_dict : dict</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Dictionary containing the number of times that an</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        element was present, divided by the total number of</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        possible occurrences.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    fraction : float</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Threshold to filter the elements. Elements with at least</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        this threshold will be included.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    elements : list</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        List of elements that met the fraction criteria.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">elements</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">abundance_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;=</span> <span class="n">fraction</span><span class="p">]</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">elements</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.gene_ontology">
        <code>gene_ontology</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.gene_ontology.find_all_children_of_go_term">
<code class="highlight language-python"><span class="n">find_all_children_of_go_term</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">go_term_name</span><span class="p">,</span> <span class="n">output_list</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Finds all children GO terms (below in hierarchy) of
a given GO term.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.find_all_children_of_go_term--parameters">Parameters</h6>
<p>go_terms : networkx.Graph
    NetworkX Graph containing GO terms datasets from .obo file.
    It could be loaded using
    cell2cell.io.read_data.load_go_terms(filename).</p>
<p>go_term_name : str
    Specific GO term to find their children. For example:
    'GO:0007155'.</p>
<p>output_list : list
    List used to perform a Depth First Search and find the
    children in a recursive way. Here the children will be
    automatically written.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/gene_ontology.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">find_all_children_of_go_term</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">go_term_name</span><span class="p">,</span> <span class="n">output_list</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Finds all children GO terms (below in hierarchy) of</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a given GO term.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    go_terms : networkx.Graph</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        NetworkX Graph containing GO terms datasets from .obo file.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        It could be loaded using</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell2cell.io.read_data.load_go_terms(filename).</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    go_term_name : str</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Specific GO term to find their children. For example:</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        &#39;GO:0007155&#39;.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    output_list : list</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        List used to perform a Depth First Search and find the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        children in a recursive way. Here the children will be</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        automatically written.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">networkx</span><span class="o">.</span><span class="n">ancestors</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">go_term_name</span><span class="p">):</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="k">if</span> <span class="n">child</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">output_list</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Retrieving children for &#39;</span> <span class="o">+</span> <span class="n">go_term_name</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>            <span class="n">output_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">find_all_children_of_go_term</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">child</span><span class="p">,</span> <span class="n">output_list</span><span class="p">,</span> <span class="n">verbose</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.gene_ontology.find_go_terms_from_keyword">
<code class="highlight language-python"><span class="n">find_go_terms_from_keyword</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">keyword</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Uses a keyword to find related GO terms.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.find_go_terms_from_keyword--parameters">Parameters</h6>
<p>go_terms : networkx.Graph
    NetworkX Graph containing GO terms datasets from .obo file.
    It could be loaded using
    cell2cell.io.read_data.load_go_terms(filename).</p>
<p>keyword : str
    Keyword to be included in the names of retrieved GO terms.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.find_go_terms_from_keyword--returns">Returns</h6>
<p>go_filter : list
    List containing all GO terms related to a keyword.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/gene_ontology.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">find_go_terms_from_keyword</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">keyword</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Uses a keyword to find related GO terms.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    go_terms : networkx.Graph</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        NetworkX Graph containing GO terms datasets from .obo file.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        It could be loaded using</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell2cell.io.read_data.load_go_terms(filename).</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    keyword : str</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Keyword to be included in the names of retrieved GO terms.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    go_filter : list</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        List containing all GO terms related to a keyword.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">go_filter</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">for</span> <span class="n">go</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">go_terms</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="k">if</span> <span class="n">keyword</span> <span class="ow">in</span> <span class="n">node</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>            <span class="n">go_filter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">go</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>                <span class="nb">print</span><span class="p">(</span><span class="n">go</span><span class="p">,</span> <span class="n">node</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">])</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">return</span> <span class="n">go_filter</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_hierarchy">
<code class="highlight language-python"><span class="n">get_genes_from_go_hierarchy</span><span class="p">(</span><span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_terms</span><span class="p">,</span> <span class="n">go_filter</span><span class="p">,</span> <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Obtains genes associated with specific GO terms and their
children GO terms (below in the hierarchy).</p>
<h6 id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_hierarchy--parameters">Parameters</h6>
<p>go_annotations : pandas.DataFrame
    Dataframe containing information about GO term annotations of each
    gene for a given organism according to the ga file. Can be loading
    with the function cell2cell.io.read_data.load_go_annotations().</p>
<p>go_terms : networkx.Graph
    NetworkX Graph containing GO terms datasets from .obo file.
    It could be loaded using
    cell2cell.io.read_data.load_go_terms(filename).</p>
<p>go_filter : list
    List containing one or more GO-terms to find associated genes.</p>
<p>go_header : str, default='GO'
    Column name wherein GO terms are located in the dataframe.</p>
<p>gene_header : str, default='Gene'
    Column name wherein genes are located in the dataframe.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_hierarchy--returns">Returns</h6>
<p>genes : list
    List of genes that are associated with GO-terms contained in
    go_filter, and related to the children GO terms of those terms.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/gene_ontology.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_genes_from_go_hierarchy</span><span class="p">(</span><span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_terms</span><span class="p">,</span> <span class="n">go_filter</span><span class="p">,</span> <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Obtains genes associated with specific GO terms and their</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    children GO terms (below in the hierarchy).</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    go_annotations : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Dataframe containing information about GO term annotations of each</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        gene for a given organism according to the ga file. Can be loading</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        with the function cell2cell.io.read_data.load_go_annotations().</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    go_terms : networkx.Graph</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        NetworkX Graph containing GO terms datasets from .obo file.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        It could be loaded using</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        cell2cell.io.read_data.load_go_terms(filename).</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    go_filter : list</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        List containing one or more GO-terms to find associated genes.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    go_header : str, default=&#39;GO&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Column name wherein GO terms are located in the dataframe.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    gene_header : str, default=&#39;Gene&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Column name wherein genes are located in the dataframe.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    -------</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    genes : list</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        List of genes that are associated with GO-terms contained in</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        go_filter, and related to the children GO terms of those terms.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">go_hierarchy</span> <span class="o">=</span> <span class="n">go_filter</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="nb">iter</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">go_hierarchy</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">iter</span><span class="p">):</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">find_all_children_of_go_term</span><span class="p">(</span><span class="n">go_terms</span><span class="p">,</span> <span class="n">go_hierarchy</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">go_hierarchy</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">go_hierarchy</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">go_hierarchy</span><span class="p">))</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">genes</span> <span class="o">=</span> <span class="n">get_genes_from_go_terms</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">=</span><span class="n">go_annotations</span><span class="p">,</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>                                    <span class="n">go_filter</span><span class="o">=</span><span class="n">go_hierarchy</span><span class="p">,</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                                    <span class="n">go_header</span><span class="o">=</span><span class="n">go_header</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                    <span class="n">gene_header</span><span class="o">=</span><span class="n">gene_header</span><span class="p">,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                    <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">return</span> <span class="n">genes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_terms">
<code class="highlight language-python"><span class="n">get_genes_from_go_terms</span><span class="p">(</span><span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_filter</span><span class="p">,</span> <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Finds genes associated with specific GO-terms.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_terms--parameters">Parameters</h6>
<p>go_annotations : pandas.DataFrame
    Dataframe containing information about GO term annotations of each
    gene for a given organism according to the ga file. Can be loading
    with the function cell2cell.io.read_data.load_go_annotations().</p>
<p>go_filter : list
    List containing one or more GO-terms to find associated genes.</p>
<p>go_header : str, default='GO'
    Column name wherein GO terms are located in the dataframe.</p>
<p>gene_header : str, default='Gene'
    Column name wherein genes are located in the dataframe.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.gene_ontology.get_genes_from_go_terms--returns">Returns</h6>
<p>genes : list
    List of genes that are associated with GO-terms contained in
    go_filter.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/gene_ontology.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_genes_from_go_terms</span><span class="p">(</span><span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_filter</span><span class="p">,</span> <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Finds genes associated with specific GO-terms.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    go_annotations : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Dataframe containing information about GO term annotations of each</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        gene for a given organism according to the ga file. Can be loading</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        with the function cell2cell.io.read_data.load_go_annotations().</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    go_filter : list</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List containing one or more GO-terms to find associated genes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    go_header : str, default=&#39;GO&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Column name wherein GO terms are located in the dataframe.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    gene_header : str, default=&#39;Gene&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Column name wherein genes are located in the dataframe.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    genes : list</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        List of genes that are associated with GO-terms contained in</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        go_filter.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Filtering genes by using GO terms&#39;</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">go_annotations</span><span class="p">[</span><span class="n">go_header</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">go_filter</span><span class="p">)][</span><span class="n">gene_header</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">genes</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data">
        <code>integrate_data</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data.get_modified_rnaseq">
<code class="highlight language-python"><span class="n">get_modified_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cutoffs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Preprocess gene expression into values used by a communication
scoring function (either continuous or binary).</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_modified_rnaseq--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>cutoffs : pandas.DataFrame
    A dataframe containing the value corresponding to cutoff or threshold
    assigned to each gene. Rows are genes and columns could be either
    'value' for a single threshold for all cell-types/tissues/samples
    or the names of cell-types/tissues/samples for thresholding in a
    specific way.
    They could be obtained through the function
    cell2cell.preprocessing.cutoffs.get_cutoffs()</p>
<p>communication_score : str, default='expression_thresholding'
    Type of communication score used to detect active ligand-receptor
    pairs between each pair of cell. See
    cell2cell.core.communication_scores for more details.
    It can be:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_thresholding&#39;
- &#39;expression_product&#39;
- &#39;expression_mean&#39;
- &#39;expression_gmean&#39;
</code></pre></div>

<h6 id="cell2cell.preprocessing.integrate_data.get_modified_rnaseq--returns">Returns</h6>
<p>modified_rnaseq : pandas.DataFrame
    Preprocessed gene expression given a communication scoring
    function to use. Rows are genes and columns are
    cell-types/tissues/samples.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/integrate_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_modified_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cutoffs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_thresholding&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Preprocess gene expression into values used by a communication</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    scoring function (either continuous or binary).</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        A dataframe containing the value corresponding to cutoff or threshold</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        assigned to each gene. Rows are genes and columns could be either</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        &#39;value&#39; for a single threshold for all cell-types/tissues/samples</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        or the names of cell-types/tissues/samples for thresholding in a</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        specific way.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        They could be obtained through the function</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        cell2cell.preprocessing.cutoffs.get_cutoffs()</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    communication_score : str, default=&#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Type of communication score used to detect active ligand-receptor</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        pairs between each pair of cell. See</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        cell2cell.core.communication_scores for more details.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        It can be:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - &#39;expression_product&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;expression_gmean&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    -------</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    modified_rnaseq : pandas.DataFrame</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Preprocessed gene expression given a communication scoring</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        function to use. Rows are genes and columns are</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">if</span> <span class="n">communication_score</span> <span class="o">==</span> <span class="s1">&#39;expression_thresholding&#39;</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="n">modified_rnaseq</span> <span class="o">=</span> <span class="n">get_thresholded_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cutoffs</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">elif</span> <span class="n">communication_score</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="s1">&#39;expression_gmean&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">modified_rnaseq</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="c1"># As other score types are implemented, other elif condition will be included here.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Score type </span><span class="si">{}</span><span class="s2"> to compute pairwise cell-interactions is not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">communication_score</span><span class="p">))</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">return</span> <span class="n">modified_rnaseq</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_go_terms">
<code class="highlight language-python"><span class="n">get_ppi_dict_from_go_terms</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_terms</span><span class="p">,</span> <span class="n">contact_go_terms</span><span class="p">,</span> <span class="n">mediator_go_terms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_children</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a complete list of protein-protein interactions into
sublists containing proteins involved in different kinds of
intercellular interactions, by provided lists of GO terms.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_go_terms--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>go_annotations : pandas.DataFrame
    Dataframe containing information about GO term annotations of each
    gene for a given organism according to the ga file. Can be loading
    with the function cell2cell.io.read_data.load_go_annotations().</p>
<p>go_terms : networkx.Graph
    NetworkX Graph containing GO terms datasets from .obo file.
    It could be loaded using
    cell2cell.io.read_data.load_go_terms(filename).</p>
<p>contact_go_terms : list
    GO terms for selecting proteins participating in cell contact
    interactions (e.g. surface proteins, receptors).</p>
<p>mediator_go_terms : list, default=None
    GO terms for selecting proteins participating in mediated or
    secreted signaling (e.g. extracellular proteins, ligands).
    If None, only interactions involved in cell contacts
    will be returned.</p>
<p>use_children : boolean, default=True
    Whether considering children GO terms (below in hierarchy) to the
    ones passed as inputs (contact_go_terms and mediator_go_terms).</p>
<p>go_header : str, default='GO'
    Column name wherein GO terms are located in the dataframe.</p>
<p>gene_header : str, default='Gene'
    Column name wherein genes are located in the dataframe.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_go_terms--returns">Returns</h6>
<p>ppi_dict : dict
    Dictionary containing lists of PPIs involving proteins that
    participate in diffferent kinds of intercellular interactions.
    Options are under the keys:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="n">contact</span><span class="w"></span>
<span class="w">        </span><span class="n">interactions</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">mediated</span><span class="w"> </span><span class="ow">or</span><span class="w"></span>
<span class="w">        </span><span class="n">secreted</span><span class="w"> </span><span class="n">signaling</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">interactions</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;combined&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="n">PPIs</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">combinations</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interactions</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">        </span><span class="n">ligands</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">,</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">etc</span><span class="p">)</span><span class="o">.</span><span class="w"></span>
<span class="n">If</span><span class="w"> </span><span class="n">mediator_go_terms</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">None</span><span class="p">,</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">dictionary</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">contain</span><span class="w"></span>
<span class="n">PPIs</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/integrate_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_ppi_dict_from_go_terms</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">go_annotations</span><span class="p">,</span> <span class="n">go_terms</span><span class="p">,</span> <span class="n">contact_go_terms</span><span class="p">,</span> <span class="n">mediator_go_terms</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_children</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                               <span class="n">go_header</span><span class="o">=</span><span class="s1">&#39;GO&#39;</span><span class="p">,</span> <span class="n">gene_header</span><span class="o">=</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a complete list of protein-protein interactions into</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    sublists containing proteins involved in different kinds of</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    intercellular interactions, by provided lists of GO terms.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    go_annotations : pandas.DataFrame</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Dataframe containing information about GO term annotations of each</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        gene for a given organism according to the ga file. Can be loading</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        with the function cell2cell.io.read_data.load_go_annotations().</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    go_terms : networkx.Graph</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        NetworkX Graph containing GO terms datasets from .obo file.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        It could be loaded using</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        cell2cell.io.read_data.load_go_terms(filename).</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    contact_go_terms : list</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        GO terms for selecting proteins participating in cell contact</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        interactions (e.g. surface proteins, receptors).</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    mediator_go_terms : list, default=None</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        GO terms for selecting proteins participating in mediated or</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        secreted signaling (e.g. extracellular proteins, ligands).</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        If None, only interactions involved in cell contacts</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        will be returned.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    use_children : boolean, default=True</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Whether considering children GO terms (below in hierarchy) to the</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        ones passed as inputs (contact_go_terms and mediator_go_terms).</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    go_header : str, default=&#39;GO&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        Column name wherein GO terms are located in the dataframe.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    gene_header : str, default=&#39;Gene&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Column name wherein genes are located in the dataframe.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    -------</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    ppi_dict : dict</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        Dictionary containing lists of PPIs involving proteins that</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        participate in diffferent kinds of intercellular interactions.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        Options are under the keys:</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        - &#39;contacts&#39; : Contains proteins participating in cell contact</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">                interactions (e.g. surface proteins, receptors)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        - &#39;mediated&#39; : Contains proteins participating in mediated or</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">                secreted signaling (e.g. ligand-receptor interactions)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        - &#39;combined&#39; : Contains both &#39;contacts&#39; and &#39;mediated&#39; PPIs.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        - &#39;complete&#39; : Contains all combinations of interactions between</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">                ligands, receptors, surface proteins, etc).</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        If mediator_go_terms input is None, this dictionary will contain</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        PPIs only for &#39;contacts&#39;.</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">use_children</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="n">contact_proteins</span> <span class="o">=</span> <span class="n">gene_ontology</span><span class="o">.</span><span class="n">get_genes_from_go_hierarchy</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">=</span><span class="n">go_annotations</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                                                                     <span class="n">go_terms</span><span class="o">=</span><span class="n">go_terms</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                                                     <span class="n">go_filter</span><span class="o">=</span><span class="n">contact_go_terms</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                                     <span class="n">go_header</span><span class="o">=</span><span class="n">go_header</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                                     <span class="n">gene_header</span><span class="o">=</span><span class="n">gene_header</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                                                     <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="n">mediator_proteins</span> <span class="o">=</span> <span class="n">gene_ontology</span><span class="o">.</span><span class="n">get_genes_from_go_hierarchy</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">=</span><span class="n">go_annotations</span><span class="p">,</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                                                      <span class="n">go_terms</span><span class="o">=</span><span class="n">go_terms</span><span class="p">,</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                                                                      <span class="n">go_filter</span><span class="o">=</span><span class="n">mediator_go_terms</span><span class="p">,</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>                                                                      <span class="n">go_header</span><span class="o">=</span><span class="n">go_header</span><span class="p">,</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                                                                      <span class="n">gene_header</span><span class="o">=</span><span class="n">gene_header</span><span class="p">,</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                                                                      <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">contact_proteins</span> <span class="o">=</span> <span class="n">gene_ontology</span><span class="o">.</span><span class="n">get_genes_from_go_terms</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">=</span><span class="n">go_annotations</span><span class="p">,</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>                                                                 <span class="n">go_filter</span><span class="o">=</span><span class="n">contact_go_terms</span><span class="p">,</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                                                                 <span class="n">go_header</span><span class="o">=</span><span class="n">go_header</span><span class="p">,</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>                                                                 <span class="n">gene_header</span><span class="o">=</span><span class="n">gene_header</span><span class="p">,</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>                                                                 <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="n">mediator_proteins</span> <span class="o">=</span> <span class="n">gene_ontology</span><span class="o">.</span><span class="n">get_genes_from_go_terms</span><span class="p">(</span><span class="n">go_annotations</span><span class="o">=</span><span class="n">go_annotations</span><span class="p">,</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>                                                                  <span class="n">go_filter</span><span class="o">=</span><span class="n">mediator_go_terms</span><span class="p">,</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>                                                                  <span class="n">go_header</span><span class="o">=</span><span class="n">go_header</span><span class="p">,</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>                                                                  <span class="n">gene_header</span><span class="o">=</span><span class="n">gene_header</span><span class="p">,</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                                                                  <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="c1"># Avoid same genes in list</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="c1">#contact_proteins = list(set(contact_proteins) - set(mediator_proteins))</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="n">ppi_dict</span> <span class="o">=</span> <span class="n">get_ppi_dict_from_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                                          <span class="n">contact_proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>                                          <span class="n">mediator_proteins</span><span class="o">=</span><span class="n">mediator_proteins</span><span class="p">,</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>                                          <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="k">return</span> <span class="n">ppi_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_proteins">
<code class="highlight language-python"><span class="n">get_ppi_dict_from_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">bidirectional</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a complete list of protein-protein interactions into
sublists containing proteins involved in different kinds of
intercellular interactions, by provided lists of proteins.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_proteins--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>contact_proteins : list
    Protein names of proteins participating in cell contact
    interactions (e.g. surface proteins, receptors).</p>
<p>mediator_proteins : list, default=None
    Protein names of proteins participating in mediated or
    secreted signaling (e.g. extracellular proteins, ligands).
    If None, only interactions involved in cell contacts
    will be returned.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>bidirectional : boolean, default=True
    Whether duplicating PPIs in both direction of interactions.
    That is, if the list considers ProtA-ProtB interaction,
    the interaction ProtB-ProtA will be also included.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_ppi_dict_from_proteins--returns">Returns</h6>
<p>ppi_dict : dict
    Dictionary containing lists of PPIs involving proteins that
    participate in diffferent kinds of intercellular interactions.
    Options are under the keys:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="n">contact</span><span class="w"></span>
<span class="w">        </span><span class="n">interactions</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">mediated</span><span class="w"> </span><span class="ow">or</span><span class="w"></span>
<span class="w">        </span><span class="n">secreted</span><span class="w"> </span><span class="n">signaling</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">interactions</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;combined&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="n">PPIs</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">combinations</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interactions</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">        </span><span class="n">ligands</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">,</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">etc</span><span class="p">)</span><span class="o">.</span><span class="w"></span>
<span class="n">If</span><span class="w"> </span><span class="n">mediator_proteins</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">None</span><span class="p">,</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">dictionary</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">contain</span><span class="w"></span>
<span class="n">PPIs</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/integrate_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_ppi_dict_from_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                               <span class="n">bidirectional</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a complete list of protein-protein interactions into</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    sublists containing proteins involved in different kinds of</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    intercellular interactions, by provided lists of proteins.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    contact_proteins : list</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Protein names of proteins participating in cell contact</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        interactions (e.g. surface proteins, receptors).</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    mediator_proteins : list, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Protein names of proteins participating in mediated or</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        secreted signaling (e.g. extracellular proteins, ligands).</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        If None, only interactions involved in cell contacts</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        will be returned.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    bidirectional : boolean, default=True</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Whether duplicating PPIs in both direction of interactions.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        That is, if the list considers ProtA-ProtB interaction,</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        the interaction ProtB-ProtA will be also included.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    -------</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    ppi_dict : dict</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Dictionary containing lists of PPIs involving proteins that</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        participate in diffferent kinds of intercellular interactions.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Options are under the keys:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        - &#39;contacts&#39; : Contains proteins participating in cell contact</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">                interactions (e.g. surface proteins, receptors)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        - &#39;mediated&#39; : Contains proteins participating in mediated or</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">                secreted signaling (e.g. ligand-receptor interactions)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        - &#39;combined&#39; : Contains both &#39;contacts&#39; and &#39;mediated&#39; PPIs.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        - &#39;complete&#39; : Contains all combinations of interactions between</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">                ligands, receptors, surface proteins, etc).</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        If mediator_proteins input is None, this dictionary will contain</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        PPIs only for &#39;contacts&#39;.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="n">ppi_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">ppi_dict</span><span class="p">[</span><span class="s1">&#39;contacts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">filter_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>                                                  <span class="n">contact_proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>                                                  <span class="n">mediator_proteins</span><span class="o">=</span><span class="n">mediator_proteins</span><span class="p">,</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>                                                  <span class="n">interaction_type</span><span class="o">=</span><span class="s1">&#39;contacts&#39;</span><span class="p">,</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>                                                  <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                                                  <span class="n">bidirectional</span><span class="o">=</span><span class="n">bidirectional</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                                                  <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="k">if</span> <span class="n">mediator_proteins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="k">for</span> <span class="n">interaction_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;mediated&#39;</span><span class="p">,</span> <span class="s1">&#39;combined&#39;</span><span class="p">,</span> <span class="s1">&#39;complete&#39;</span><span class="p">]:</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="n">ppi_dict</span><span class="p">[</span><span class="n">interaction_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi</span><span class="o">.</span><span class="n">filter_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                                                          <span class="n">contact_proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                                          <span class="n">mediator_proteins</span><span class="o">=</span><span class="n">mediator_proteins</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                                          <span class="n">interaction_type</span><span class="o">=</span><span class="n">interaction_type</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                                                          <span class="n">bidirectional</span><span class="o">=</span><span class="n">bidirectional</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                                          <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">return</span> <span class="n">ppi_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data.get_thresholded_rnaseq">
<code class="highlight language-python"><span class="n">get_thresholded_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cutoffs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Binzarizes a RNA-seq dataset given cutoff or threshold
values.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_thresholded_rnaseq--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>cutoffs : pandas.DataFrame
    A dataframe containing the value corresponding to cutoff or threshold
    assigned to each gene. Rows are genes and columns could be either
    'value' for a single threshold for all cell-types/tissues/samples
    or the names of cell-types/tissues/samples for thresholding in a
    specific way.
    They could be obtained through the function
    cell2cell.preprocessing.cutoffs.get_cutoffs()</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_thresholded_rnaseq--returns">Returns</h6>
<p>binary_rnaseq_data : pandas.DataFrame
    Preprocessed gene expression into binary values given cutoffs
    or thresholds either general or specific for all cell-types/
    tissues/samples.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/integrate_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_thresholded_rnaseq</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">cutoffs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Binzarizes a RNA-seq dataset given cutoff or threshold</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    values.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    cutoffs : pandas.DataFrame</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A dataframe containing the value corresponding to cutoff or threshold</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        assigned to each gene. Rows are genes and columns could be either</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        &#39;value&#39; for a single threshold for all cell-types/tissues/samples</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        or the names of cell-types/tissues/samples for thresholding in a</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        specific way.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        They could be obtained through the function</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        cell2cell.preprocessing.cutoffs.get_cutoffs()</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    binary_rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Preprocessed gene expression into binary values given cutoffs</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        or thresholds either general or specific for all cell-types/</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        tissues/samples.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">binary_rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cutoffs</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s1">&#39;value&#39;</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">binary_rnaseq_data</span> <span class="o">=</span> <span class="n">binary_rnaseq_data</span><span class="o">.</span><span class="n">gt</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cutoffs</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">elif</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>  <span class="c1"># Check there is a column for each cell type</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="n">binary_rnaseq_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">binary_rnaseq_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">gt</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cutoffs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="c1"># ge</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s2">&quot;The cutoff data provided does not have a &#39;value&#39; column or does not match the columns in rnaseq_data.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">binary_rnaseq_data</span> <span class="o">=</span> <span class="n">binary_rnaseq_data</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">return</span> <span class="n">binary_rnaseq_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.integrate_data.get_weighted_ppi">
<code class="highlight language-python"><span class="n">get_weighted_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">modified_rnaseq_data</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Assigns preprocessed gene expression values to
proteins in a list of protein-protein interactions.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_weighted_ppi--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>modified_rnaseq_data : pandas.DataFrame
    Preprocessed gene expression given a communication scoring
    function to use. Rows are genes and columns are
    cell-types/tissues/samples.</p>
<p>column : str, default='value'
    Column name to consider the gene expression values.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.integrate_data.get_weighted_ppi--returns">Returns</h6>
<p>weighted_ppi : pandas.DataFrame
    List of protein-protein interactions that contains gene expression
    values instead of the names of interacting proteins. Gene expression
    values are preprocessed given a communication scoring function to
    use.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/integrate_data.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_weighted_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">modified_rnaseq_data</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Assigns preprocessed gene expression values to</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    proteins in a list of protein-protein interactions.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    modified_rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Preprocessed gene expression given a communication scoring</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        function to use. Rows are genes and columns are</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        cell-types/tissues/samples.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    column : str, default=&#39;value&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Column name to consider the gene expression values.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    -------</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    weighted_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        List of protein-protein interactions that contains gene expression</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        values instead of the names of interacting proteins. Gene expression</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        values are preprocessed given a communication scoring function to</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        use.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">prot_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">prot_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">weighted_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">weighted_ppi</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">weighted_ppi</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">modified_rnaseq_data</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">column</span><span class="p">])</span> <span class="c1"># Replaced .loc by .at</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">weighted_ppi</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">weighted_ppi</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">modified_rnaseq_data</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">column</span><span class="p">])</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">weighted_ppi</span> <span class="o">=</span> <span class="n">weighted_ppi</span><span class="p">[[</span><span class="n">prot_a</span><span class="p">,</span> <span class="n">prot_b</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">return</span> <span class="n">weighted_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes">
        <code>manipulate_dataframes</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.check_presence_in_dataframe">
<code class="highlight language-python"><span class="n">check_presence_in_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">elements</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Searches for elements in a dataframe and returns those
that are present in the dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.check_presence_in_dataframe--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe</p>
<p>elements : list
    List of elements to find in the dataframe. They
    must be a data type contained in the dataframe.</p>
<p>columns : list, default=None
    Names of columns to consider in the search. If
    None, all columns are used.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.check_presence_in_dataframe--returns">Returns</h6>
<p>found_elements : list
    List of elements in the input list that were found
    in the dataframe.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">check_presence_in_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">elements</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Searches for elements in a dataframe and returns those</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    that are present in the dataframe.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        A dataframe</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    elements : list</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        List of elements to find in the dataframe. They</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        must be a data type contained in the dataframe.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    columns : list, default=None</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Names of columns to consider in the search. If</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        None, all columns are used.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    found_elements : list</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        List of elements in the input list that were found</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        in the dataframe.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">df_elements</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">df_elements</span> <span class="o">=</span> <span class="n">df_elements</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df_elements</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">elements</span><span class="p">)]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">found_elements</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df_elements</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">return</span> <span class="n">found_elements</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.check_symmetry">
<code class="highlight language-python"><span class="n">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Checks whether a dataframe is symmetric.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.check_symmetry--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.check_symmetry--returns">Returns</h6>
<p>symmetric : boolean
    Whether a dataframe is symmetric.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Checks whether a dataframe is symmetric.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    -------</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    symmetric : boolean</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Whether a dataframe is symmetric.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>    <span class="n">shape</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">if</span> <span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="n">symmetric</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span> <span class="o">==</span> <span class="n">df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="n">symmetric</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">return</span> <span class="n">symmetric</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.convert_to_distance_matrix">
<code class="highlight language-python"><span class="n">convert_to_distance_matrix</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Converts a symmetric dataframe into a distance dataframe.
That is, diagonal elements are all zero.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.convert_to_distance_matrix--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.convert_to_distance_matrix--returns">Returns</h6>
<p>df_ : pandas.DataFrame
    A copy of df, but with all diagonal elements with a
    value of zero.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">convert_to_distance_matrix</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Converts a symmetric dataframe into a distance dataframe.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    That is, diagonal elements are all zero.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    df_ : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A copy of df, but with all diagonal elements with a</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        value of zero.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="k">if</span> <span class="n">check_symmetry</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">df_</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">trace</span><span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">values</span><span class="p">,)</span> <span class="o">!=</span> <span class="mf">0.0</span><span class="p">:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>            <span class="k">raise</span> <span class="ne">Warning</span><span class="p">(</span><span class="s2">&quot;Diagonal elements are not zero. Automatically replaced by zeros&quot;</span><span class="p">)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;The DataFrame is not symmetric&#39;</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">return</span> <span class="n">df_</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.shuffle_cols_in_df">
<code class="highlight language-python"><span class="n">shuffle_cols_in_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly shuffles specific columns in a dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_cols_in_df--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<p>columns : list
    Names of columns to shuffle.</p>
<p>shuffling_number : int, default=1
    Number of shuffles per column.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_cols_in_df--returns">Returns</h6>
<p>df_ : pandas.DataFrame
    A shuffled dataframe.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">shuffle_cols_in_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly shuffles specific columns in a dataframe.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    columns : list</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Names of columns to shuffle.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    shuffling_number : int, default=1</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Number of shuffles per column.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    df_ : pandas.DataFrame</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        A shuffled dataframe.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">columns</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">columns</span><span class="p">]</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">shuffling_number</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>            <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>                <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>            <span class="n">df_</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">permutation</span><span class="p">(</span><span class="n">df_</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">df_</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.shuffle_dataframe">
<code class="highlight language-python"><span class="n">shuffle_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly shuffles a whole dataframe across a given axis.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_dataframe--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<p>shuffling_number : int, default=1
    Number of shuffles per column.</p>
<p>axis : int, default=0
    An axis of the dataframe (0 across rows, 1 across columns).
    Across rows means that shuffles each column independently,
    and across columns shuffles each row independently.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_dataframe--returns">Returns</h6>
<p>df_ : pandas.DataFrame
    A shuffled dataframe.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">shuffle_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly shuffles a whole dataframe across a given axis.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    shuffling_number : int, default=1</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Number of shuffles per column.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        An axis of the dataframe (0 across rows, 1 across columns).</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Across rows means that shuffles each column independently,</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        and across columns shuffles each row independently.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    -------</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    df_ : pandas.DataFrame</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        A shuffled dataframe.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">axis</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="ow">not</span> <span class="n">axis</span><span class="p">)</span>  <span class="c1"># pandas.DataFrame is always 2D</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">to_shuffle</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">rollaxis</span><span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">axis</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">shuffling_number</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">view</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">to_shuffle</span><span class="p">):</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>            <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>                <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">view</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">rollaxis</span><span class="p">(</span><span class="n">to_shuffle</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">df_</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df_</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">return</span> <span class="n">df_</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.shuffle_rows_in_df">
<code class="highlight language-python"><span class="n">shuffle_rows_in_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">rows</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly shuffles specific rows in a dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_rows_in_df--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<p>rows : list
    Names of rows (or indexes) to shuffle.</p>
<p>shuffling_number : int, default=1
    Number of shuffles per row.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.shuffle_rows_in_df--returns">Returns</h6>
<p>df_.T : pandas.DataFrame
    A shuffled dataframe.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">shuffle_rows_in_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">rows</span><span class="p">,</span> <span class="n">shuffling_number</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly shuffles specific rows in a dataframe.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    rows : list</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Names of rows (or indexes) to shuffle.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    shuffling_number : int, default=1</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Number of shuffles per row.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    df_.T : pandas.DataFrame</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        A shuffled dataframe.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">rows</span> <span class="o">=</span> <span class="p">[</span><span class="n">rows</span><span class="p">]</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">shuffling_number</span><span class="p">):</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>            <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>                <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>            <span class="n">df_</span><span class="p">[</span><span class="n">row</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">permutation</span><span class="p">(</span><span class="n">df_</span><span class="p">[</span><span class="n">row</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">return</span> <span class="n">df_</span><span class="o">.</span><span class="n">T</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.manipulate_dataframes.subsample_dataframe">
<code class="highlight language-python"><span class="n">subsample_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_samples</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly subsamples rows of a dataframe.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.subsample_dataframe--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe.</p>
<p>n_samples : int
    Number of samples, rows in this case. If
    n_samples is larger than the number of rows,
    the entire dataframe will be returned, but
    shuffled.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<h6 id="cell2cell.preprocessing.manipulate_dataframes.subsample_dataframe--returns">Returns</h6>
<p>subsampled_df : pandas.DataFrame
    A subsampled and shuffled dataframe.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/manipulate_dataframes.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">subsample_dataframe</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_samples</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly subsamples rows of a dataframe.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A dataframe.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    n_samples : int</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Number of samples, rows in this case. If</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        n_samples is larger than the number of rows,</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        the entire dataframe will be returned, but</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        shuffled.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    subsampled_df : pandas.DataFrame</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        A subsampled and shuffled dataframe.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">items</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">n_samples</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">items</span><span class="p">):</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">n_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">items</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">random_state</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">random_state</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">items</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">subsampled_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">items</span><span class="p">[:</span><span class="n">n_samples</span><span class="p">],:]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">return</span> <span class="n">subsampled_df</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.ppi">
        <code>ppi</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.bidirectional_ppi_for_cci">
<code class="highlight language-python"><span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Makes a list of protein-protein interactions to be bidirectional.
That is, repeating a PPI like ProtA-ProtB but in the other direction
(ProtB-ProtA) if not present.</p>
<h6 id="cell2cell.preprocessing.ppi.bidirectional_ppi_for_cci--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.bidirectional_ppi_for_cci--returns">Returns</h6>
<p>bi_ppi_data : pandas.DataFrame
    Bidirectional ppi_data. Contains duplicated PPIs in both directions.
    That is, it contains both ProtA-ProtB and ProtB-ProtA interactions.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Makes a list of protein-protein interactions to be bidirectional.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    That is, repeating a PPI like ProtA-ProtB but in the other direction</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    (ProtB-ProtA) if not present.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    bi_ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Bidirectional ppi_data. Contains duplicated PPIs in both directions.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        That is, it contains both ProtA-ProtB and ProtB-ProtA interactions.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Making bidirectional PPI for CCI.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">if</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="k">return</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">ppi_A</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">col1</span> <span class="o">=</span> <span class="n">ppi_A</span><span class="p">[[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]]]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">col2</span> <span class="o">=</span> <span class="n">ppi_A</span><span class="p">[[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]]]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">ppi_B</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">ppi_B</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">col2</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">ppi_B</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">col1</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Removing duplicates in bidirectional PPI network.&quot;</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">bi_ppi_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">ppi_A</span><span class="p">,</span> <span class="n">ppi_B</span><span class="p">],</span> <span class="n">join</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">bi_ppi_data</span> <span class="o">=</span> <span class="n">bi_ppi_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">bi_ppi_data</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">return</span> <span class="n">bi_ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.filter_complex_ppi_by_proteins">
<code class="highlight language-python"><span class="n">filter_complex_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="s1">&#39;&amp;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions that for sure contains
protein complexes to contain only interacting proteins or subunites
in a list of specific protein or gene names.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_complex_ppi_by_proteins--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>proteins : list
    A list of protein names to filter PPIs.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the protein names in the list of proteins and
    the names in the ppi_data to match their names and integrate their
    Useful when there are inconsistencies in the names that comes from a
    expression matrix and from ligand-receptor annotations.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_complex_ppi_by_proteins--returns">Returns</h6>
<p>integrated_ppi : pandas.DataFrame
    A filtered list of PPIs, containing protein complexes in some cases,
    by a given list of proteins or gene names.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">filter_complex_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="s1">&#39;&amp;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                   <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a list of protein-protein interactions that for sure contains</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    protein complexes to contain only interacting proteins or subunites</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    in a list of specific protein or gene names.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    proteins : list</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        A list of protein names to filter PPIs.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Whether making uppercase the protein names in the list of proteins and</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        the names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Useful when there are inconsistencies in the names that comes from a</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        expression matrix and from ligand-receptor annotations.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    -------</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    integrated_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        A filtered list of PPIs, containing protein complexes in some cases,</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        by a given list of proteins or gene names.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">col_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">col_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">if</span> <span class="n">upper_letter_comparison</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">prots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">proteins</span><span class="p">])</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">prots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">proteins</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="n">col_a_genes</span><span class="p">,</span> <span class="n">complex_a</span><span class="p">,</span> <span class="n">col_b_genes</span><span class="p">,</span> <span class="n">complex_b</span><span class="p">,</span> <span class="n">complexes</span> <span class="o">=</span> <span class="n">get_genes_from_complexes</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">integrated_ppi</span><span class="p">,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                                                                                         <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>                                                                                         <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>                                                                                         <span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="n">shared_a_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">col_a_genes</span> <span class="o">&amp;</span> <span class="n">prots</span><span class="p">)</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">shared_b_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">col_b_genes</span> <span class="o">&amp;</span> <span class="n">prots</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="n">shared_a_complexes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">complex_a</span> <span class="o">&amp;</span> <span class="n">prots</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="n">shared_b_complexes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">complex_b</span> <span class="o">&amp;</span> <span class="n">prots</span><span class="p">)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>    <span class="n">integrated_a_complexes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="n">integrated_b_complexes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">complexes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">p</span> <span class="ow">in</span> <span class="n">shared_a_complexes</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">v</span><span class="p">):</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="n">integrated_a_complexes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="k">elif</span> <span class="nb">all</span><span class="p">(</span><span class="n">p</span> <span class="ow">in</span> <span class="n">shared_b_complexes</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">v</span><span class="p">):</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>            <span class="n">integrated_b_complexes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">integrated_a</span> <span class="o">=</span> <span class="n">shared_a_genes</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">integrated_a_complexes</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">integrated_b</span> <span class="o">=</span> <span class="n">shared_b_genes</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">integrated_b_complexes</span><span class="p">)</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="nb">filter</span> <span class="o">=</span> <span class="p">(</span><span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">integrated_a</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">integrated_b</span><span class="p">))</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">filter</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">return</span> <span class="n">integrated_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.filter_ppi_by_proteins">
<code class="highlight language-python"><span class="n">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions to contain
only interacting proteins in a list of specific protein or gene names.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_ppi_by_proteins--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>proteins : list
    A list of protein names to filter PPIs.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the protein names in the list of proteins and
    the names in the ppi_data to match their names and integrate their
    Useful when there are inconsistencies in the names that comes from a
    expression matrix and from ligand-receptor annotations.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_ppi_by_proteins--returns">Returns</h6>
<p>integrated_ppi : pandas.DataFrame
    A filtered list of PPIs by a given list of proteins or gene names.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Filters a list of protein-protein interactions to contain</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    only interacting proteins in a list of specific protein or gene names.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    proteins : list</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        A list of protein names to filter PPIs.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Whether making uppercase the protein names in the list of proteins and</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        the names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Useful when there are inconsistencies in the names that comes from a</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        expression matrix and from ligand-receptor annotations.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    -------</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    integrated_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        A filtered list of PPIs by a given list of proteins or gene names.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">col_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">col_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">if</span> <span class="n">upper_letter_comparison</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">prots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">proteins</span><span class="p">])</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">prots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">proteins</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">filter_complex_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">integrated_ppi</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                                                        <span class="n">proteins</span><span class="o">=</span><span class="n">prots</span><span class="p">,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                                                        <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>                                                        <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="c1"># Because it was ran above</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>                                                        <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>                                                        <span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="p">[(</span><span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">prots</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">integrated_ppi</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">prots</span><span class="p">))]</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">integrated_ppi</span> <span class="o">=</span> <span class="n">integrated_ppi</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">return</span> <span class="n">integrated_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.filter_ppi_network">
<code class="highlight language-python"><span class="n">filter_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reference_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bidirectional</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_type</span><span class="o">=</span><span class="s1">&#39;contacts&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions to contain interacting
proteins involved in different kinds of cell-cell interactions/communication.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_ppi_network--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>contact_proteins : list
    Protein names of proteins participating in cell contact
    interactions (e.g. surface proteins, receptors).</p>
<p>mediator_proteins : list, default=None
    Protein names of proteins participating in mediated or
    secreted signaling (e.g. extracellular proteins, ligands).
    If None, only interactions involved in cell contacts
    will be returned.</p>
<p>reference_list : list, default=None
    Reference list of protein names. Filtered PPIs from contact_proteins
    and mediator proteins will be keep only if those proteins are also
    present in this list when is not None.</p>
<p>bidirectional : boolean, default=True
    Whether duplicating PPIs in both direction of interactions.
    That is, if the list considers ProtA-ProtB interaction,
    the interaction ProtB-ProtA will be also included.</p>
<p>interaction_type : str, default='contacts'
    Type of intercellular interactions/communication where the proteins
    have to be involved in.
    Available types are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="n">contact</span><span class="w"></span>
<span class="w">        </span><span class="n">interactions</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">mediated</span><span class="w"> </span><span class="ow">or</span><span class="w"></span>
<span class="w">        </span><span class="n">secreted</span><span class="w"> </span><span class="n">signaling</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">interactions</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;combined&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="n">PPIs</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">combinations</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interactions</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">        </span><span class="n">ligands</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">,</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">etc</span><span class="p">)</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.filter_ppi_network--returns">Returns</h6>
<p>new_ppi_data : pandas.DataFrame
    A filtered list of PPIs by a given list of proteins or gene names
    depending on the type of intercellular communication.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">filter_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reference_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bidirectional</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                       <span class="n">interaction_type</span><span class="o">=</span><span class="s1">&#39;contacts&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a list of protein-protein interactions to contain interacting</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    proteins involved in different kinds of cell-cell interactions/communication.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    contact_proteins : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Protein names of proteins participating in cell contact</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        interactions (e.g. surface proteins, receptors).</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    mediator_proteins : list, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Protein names of proteins participating in mediated or</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        secreted signaling (e.g. extracellular proteins, ligands).</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        If None, only interactions involved in cell contacts</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        will be returned.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    reference_list : list, default=None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Reference list of protein names. Filtered PPIs from contact_proteins</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        and mediator proteins will be keep only if those proteins are also</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        present in this list when is not None.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    bidirectional : boolean, default=True</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Whether duplicating PPIs in both direction of interactions.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        That is, if the list considers ProtA-ProtB interaction,</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        the interaction ProtB-ProtA will be also included.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    interaction_type : str, default=&#39;contacts&#39;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Type of intercellular interactions/communication where the proteins</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        have to be involved in.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Available types are:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        - &#39;contacts&#39; : Contains proteins participating in cell contact</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">                interactions (e.g. surface proteins, receptors)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        - &#39;mediated&#39; : Contains proteins participating in mediated or</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">                secreted signaling (e.g. ligand-receptor interactions)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        - &#39;combined&#39; : Contains both &#39;contacts&#39; and &#39;mediated&#39; PPIs.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        - &#39;complete&#39; : Contains all combinations of interactions between</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">                ligands, receptors, surface proteins, etc).</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    -------</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">    new_ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        A filtered list of PPIs by a given list of proteins or gene names</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        depending on the type of intercellular communication.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">get_filtered_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                                            <span class="n">contact_proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                                            <span class="n">mediator_proteins</span><span class="o">=</span><span class="n">mediator_proteins</span><span class="p">,</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                                            <span class="n">reference_list</span><span class="o">=</span><span class="n">reference_list</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                                            <span class="n">interaction_type</span><span class="o">=</span><span class="n">interaction_type</span><span class="p">,</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                                            <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                                            <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">if</span> <span class="n">bidirectional</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">bidirectional_ppi_for_cci</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">new_ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>                                                 <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>                                                 <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="k">return</span> <span class="n">new_ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.get_all_to_all_ppi">
<code class="highlight language-python"><span class="n">get_all_to_all_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions to
contain only proteins in a given list in both
columns of interacting partners.</p>
<h6 id="cell2cell.preprocessing.ppi.get_all_to_all_ppi--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>proteins : list
    A list of protein names to filter PPIs.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.ppi.get_all_to_all_ppi--returns">Returns</h6>
<p>new_ppi_data : pandas.DataFrame
    A filtered list of PPIs by a given list of proteins or gene names.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_all_to_all_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Filters a list of protein-protein interactions to</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    contain only proteins in a given list in both</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    columns of interacting partners.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    proteins : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A list of protein names to filter PPIs.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    -------</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    new_ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        A filtered list of PPIs by a given list of proteins or gene names.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">header_interactorA</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">header_interactorB</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorA</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorB</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins</span><span class="p">)]</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">new_ppi_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">return</span> <span class="n">new_ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.get_filtered_ppi_network">
<code class="highlight language-python"><span class="n">get_filtered_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reference_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_type</span><span class="o">=</span><span class="s1">&#39;contacts&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions to contain interacting
proteins involved in different kinds of cell-cell interactions/communication.</p>
<h6 id="cell2cell.preprocessing.ppi.get_filtered_ppi_network--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>contact_proteins : list
    Protein names of proteins participating in cell contact
    interactions (e.g. surface proteins, receptors).</p>
<p>mediator_proteins : list, default=None
    Protein names of proteins participating in mediated or
    secreted signaling (e.g. extracellular proteins, ligands).
    If None, only interactions involved in cell contacts
    will be returned.</p>
<p>reference_list : list, default=None
    Reference list of protein names. Filtered PPIs from contact_proteins
    and mediator proteins will be keep only if those proteins are also
    present in this list when is not None.</p>
<p>interaction_type : str, default='contacts'
    Type of intercellular interactions/communication where the proteins
    have to be involved in.
    Available types are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="n">contact</span><span class="w"></span>
<span class="w">        </span><span class="n">interactions</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">proteins</span><span class="w"> </span><span class="n">participating</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">mediated</span><span class="w"> </span><span class="ow">or</span><span class="w"></span>
<span class="w">        </span><span class="n">secreted</span><span class="w"> </span><span class="n">signaling</span><span class="w"> </span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">interactions</span><span class="p">)</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;combined&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="s1">&#39;contacts&#39;</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="s1">&#39;mediated&#39;</span><span class="w"> </span><span class="n">PPIs</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;complete&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Contains</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">combinations</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interactions</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">        </span><span class="n">ligands</span><span class="p">,</span><span class="w"> </span><span class="n">receptors</span><span class="p">,</span><span class="w"> </span><span class="n">surface</span><span class="w"> </span><span class="n">proteins</span><span class="p">,</span><span class="w"> </span><span class="n">etc</span><span class="p">)</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.get_filtered_ppi_network--returns">Returns</h6>
<p>new_ppi_data : pandas.DataFrame
    A filtered list of PPIs by a given list of proteins or gene names
    depending on the type of intercellular communication.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_filtered_ppi_network</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">contact_proteins</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">reference_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                             <span class="n">interaction_type</span><span class="o">=</span><span class="s1">&#39;contacts&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a list of protein-protein interactions to contain interacting</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    proteins involved in different kinds of cell-cell interactions/communication.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    contact_proteins : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Protein names of proteins participating in cell contact</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        interactions (e.g. surface proteins, receptors).</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    mediator_proteins : list, default=None</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Protein names of proteins participating in mediated or</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        secreted signaling (e.g. extracellular proteins, ligands).</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        If None, only interactions involved in cell contacts</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        will be returned.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    reference_list : list, default=None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Reference list of protein names. Filtered PPIs from contact_proteins</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        and mediator proteins will be keep only if those proteins are also</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        present in this list when is not None.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    interaction_type : str, default=&#39;contacts&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        Type of intercellular interactions/communication where the proteins</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        have to be involved in.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Available types are:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;contacts&#39; : Contains proteins participating in cell contact</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                interactions (e.g. surface proteins, receptors)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        - &#39;mediated&#39; : Contains proteins participating in mediated or</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">                secreted signaling (e.g. ligand-receptor interactions)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        - &#39;combined&#39; : Contains both &#39;contacts&#39; and &#39;mediated&#39; PPIs.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        - &#39;complete&#39; : Contains all combinations of interactions between</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">                ligands, receptors, surface proteins, etc).</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">    -------</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">    new_ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        A filtered list of PPIs by a given list of proteins or gene names</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        depending on the type of intercellular communication.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">mediator_proteins</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">interaction_type</span> <span class="o">!=</span> <span class="s1">&#39;contacts&#39;</span><span class="p">):</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mediator_proteins cannot be None when interaction_type is not contacts&quot;</span><span class="p">)</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="c1"># Consider only genes that are in the reference_list:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="k">if</span> <span class="n">reference_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="n">contact_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reference_list</span> <span class="p">,</span> <span class="n">contact_proteins</span><span class="p">))</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">if</span> <span class="n">mediator_proteins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="n">mediator_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reference_list</span><span class="p">,</span> <span class="n">mediator_proteins</span><span class="p">))</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Filtering PPI interactions by using a list of genes for </span><span class="si">{}</span><span class="s1"> interactions&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">interaction_type</span><span class="p">))</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">if</span> <span class="n">interaction_type</span> <span class="o">==</span> <span class="s1">&#39;contacts&#39;</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">get_all_to_all_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                          <span class="n">proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">elif</span> <span class="n">interaction_type</span> <span class="o">==</span> <span class="s1">&#39;complete&#39;</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="n">total_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">contact_proteins</span> <span class="o">+</span> <span class="n">mediator_proteins</span><span class="p">))</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">get_all_to_all_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                          <span class="n">proteins</span><span class="o">=</span><span class="n">total_proteins</span><span class="p">,</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="c1"># All the following interactions incorporate contacts-mediator interactions</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="n">mediated</span> <span class="o">=</span> <span class="n">get_one_group_to_other_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>                                              <span class="n">proteins_a</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>                                              <span class="n">proteins_b</span><span class="o">=</span><span class="n">mediator_proteins</span><span class="p">,</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                                              <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="k">if</span> <span class="n">interaction_type</span> <span class="o">==</span> <span class="s1">&#39;mediated&#39;</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>            <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">mediated</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>        <span class="k">elif</span> <span class="n">interaction_type</span> <span class="o">==</span> <span class="s1">&#39;combined&#39;</span><span class="p">:</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>            <span class="n">contacts</span> <span class="o">=</span> <span class="n">get_all_to_all_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>                                          <span class="n">proteins</span><span class="o">=</span><span class="n">contact_proteins</span><span class="p">,</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>                                          <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>            <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">contacts</span><span class="p">,</span> <span class="n">mediated</span><span class="p">],</span> <span class="n">ignore_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>            <span class="k">raise</span> <span class="ne">NameError</span><span class="p">(</span><span class="s1">&#39;Not valid interaction type to filter the PPI network&#39;</span><span class="p">)</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="n">new_ppi_data</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="k">return</span> <span class="n">new_ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.get_genes_from_complexes">
<code class="highlight language-python"><span class="n">get_genes_from_complexes</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="s1">&#39;&amp;&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Gets protein/gene names for individual proteins (subunits when in complex)
in a list of PPIs. If protein is a complex, for example ProtA&amp;ProtB, it will
return ProtA and ProtB separately.</p>
<h6 id="cell2cell.preprocessing.ppi.get_genes_from_complexes--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.ppi.get_genes_from_complexes--returns">Returns</h6>
<p>col_a_genes : list
    List of protein/gene names for proteins and subunits in the first column
    of interacting partners.</p>
<p>complex_a : list
    List of list of subunits of each complex that were present in the first
    column of interacting partners and that were returned as subunits in the
    previous list.</p>
<p>col_b_genes : list
    List of protein/gene names for proteins and subunits in the second column
    of interacting partners.</p>
<p>complex_b : list
    List of list of subunits of each complex that were present in the second
    column of interacting partners and that were returned as subunits in the
    previous list.</p>
<p>complexes : dict
    Dictionary where keys are the complex names in the list of PPIs, while
    values are list of subunits for the respective complex names.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_genes_from_complexes</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="s1">&#39;&amp;&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Gets protein/gene names for individual proteins (subunits when in complex)</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in a list of PPIs. If protein is a complex, for example ProtA&amp;ProtB, it will</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    return ProtA and ProtB separately.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    col_a_genes : list</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        List of protein/gene names for proteins and subunits in the first column</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        of interacting partners.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    complex_a : list</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        List of list of subunits of each complex that were present in the first</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        column of interacting partners and that were returned as subunits in the</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        previous list.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    col_b_genes : list</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        List of protein/gene names for proteins and subunits in the second column</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        of interacting partners.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    complex_b : list</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        List of list of subunits of each complex that were present in the second</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        column of interacting partners and that were returned as subunits in the</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        previous list.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    complexes : dict</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Dictionary where keys are the complex names in the list of PPIs, while</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        values are list of subunits for the respective complex names.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">col_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">col_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">col_a_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="n">col_b_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="n">complexes</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="n">complex_a</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">complex_b</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">prot_a</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">col_a</span><span class="p">]</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="n">prot_b</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">col_b</span><span class="p">]</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">in</span> <span class="n">prot_a</span><span class="p">:</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>            <span class="n">comp</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">l</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">prot_a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">complex_sep</span><span class="p">)])</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="n">complexes</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">comp</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>            <span class="n">complex_a</span> <span class="o">=</span> <span class="n">complex_a</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">comp</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>            <span class="n">col_a_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prot_a</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">in</span> <span class="n">prot_b</span><span class="p">:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>            <span class="n">comp</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">prot_b</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">complex_sep</span><span class="p">)])</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>            <span class="n">complexes</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">comp</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>            <span class="n">complex_b</span> <span class="o">=</span> <span class="n">complex_b</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">comp</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>            <span class="n">col_b_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prot_b</span><span class="p">)</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">return</span> <span class="n">col_a_genes</span><span class="p">,</span> <span class="n">complex_a</span><span class="p">,</span> <span class="n">col_b_genes</span><span class="p">,</span> <span class="n">complex_b</span><span class="p">,</span> <span class="n">complexes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.get_one_group_to_other_ppi">
<code class="highlight language-python"><span class="n">get_one_group_to_other_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins_a</span><span class="p">,</span> <span class="n">proteins_b</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a list of protein-protein interactions to
contain specific proteins in the first column of
interacting partners an other specific proteins in
the second column.</p>
<h6 id="cell2cell.preprocessing.ppi.get_one_group_to_other_ppi--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>proteins_a : list
    A list of protein names to filter the first column of interacting
    proteins in a list of PPIs.</p>
<p>proteins_b : list
    A list of protein names to filter the second column of interacting
    proteins in a list of PPIs.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<h6 id="cell2cell.preprocessing.ppi.get_one_group_to_other_ppi--returns">Returns</h6>
<p>new_ppi_data : pandas.DataFrame
    A filtered list of PPIs by a given lists of proteins or gene names.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_one_group_to_other_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">proteins_a</span><span class="p">,</span> <span class="n">proteins_b</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Filters a list of protein-protein interactions to</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    contain specific proteins in the first column of</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    interacting partners an other specific proteins in</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    the second column.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    proteins_a : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A list of protein names to filter the first column of interacting</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        proteins in a list of PPIs.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    proteins_b : list</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        A list of protein names to filter the second column of interacting</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        proteins in a list of PPIs.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    -------</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    new_ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        A filtered list of PPIs by a given lists of proteins or gene names.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">header_interactorA</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">header_interactorB</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">direction1</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorA</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins_a</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorB</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins_b</span><span class="p">)]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">direction2</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorA</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins_b</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">ppi_data</span><span class="p">[</span><span class="n">header_interactorB</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">proteins_a</span><span class="p">)]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">direction2</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">header_interactorB</span><span class="p">,</span> <span class="n">header_interactorA</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">]</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">direction2</span> <span class="o">=</span> <span class="n">direction2</span><span class="p">[[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">]]</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">new_ppi_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">direction1</span><span class="p">,</span> <span class="n">direction2</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">return</span> <span class="n">new_ppi_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.preprocess_ppi_data">
<code class="highlight language-python"><span class="n">preprocess_ppi_data</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rnaseq_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strna</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Preprocess a list of protein-protein interactions by
removed bidirectionality and keeping the minimum number of columns.</p>
<h6 id="cell2cell.preprocessing.ppi.preprocess_ppi_data--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>sort_values : str, default=None
    Column name to sort PPIs in an ascending manner. If None,
    sorting is not done.</p>
<p>rnaseq_genes : list, default=None
    List of protein or gene names to filter the PPIs.</p>
<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>dropna : boolean, default=False
    Whether dropping incomplete PPIs (with NaN values).</p>
<p>strna : str, default=''
    String to replace empty or NaN values with.</p>
<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the gene names in the expression matrices and the
    protein names in the ppi_data to match their names and integrate their
    respective expression level. Useful when there are inconsistencies in the
    names between the expression matrix and the ligand-receptor annotations.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.preprocess_ppi_data--returns">Returns</h6>
<p>simplified_ppi : pandas.DataFrame
    A simplified list of protein-protein interactions. It does not contains
    duplicated interactions in both directions (if ProtA-ProtB and
    ProtB-ProtA interactions are present, only the one that appears first
    is kept) either extra columns beyond interacting ones. It contains only
    three columns: 'A', 'B', 'score', wherein 'A' and 'B' are the interacting
    partners in the PPI and 'score' represents a weight of the interaction
    for computing cell-cell interactions/communication.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">preprocess_ppi_data</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">sort_values</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rnaseq_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>             <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strna</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Preprocess a list of protein-protein interactions by</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    removed bidirectionality and keeping the minimum number of columns.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    sort_values : str, default=None</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Column name to sort PPIs in an ascending manner. If None,</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        sorting is not done.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    rnaseq_genes : list, default=None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        List of protein or gene names to filter the PPIs.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    dropna : boolean, default=False</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Whether dropping incomplete PPIs (with NaN values).</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    strna : str, default=&#39;&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        String to replace empty or NaN values with.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Whether making uppercase the gene names in the expression matrices and the</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        protein names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        respective expression level. Useful when there are inconsistencies in the</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        names between the expression matrix and the ligand-receptor annotations.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    -------</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    simplified_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        A simplified list of protein-protein interactions. It does not contains</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        duplicated interactions in both directions (if ProtA-ProtB and</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        ProtB-ProtA interactions are present, only the one that appears first</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        is kept) either extra columns beyond interacting ones. It contains only</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        three columns: &#39;A&#39;, &#39;B&#39;, &#39;score&#39;, wherein &#39;A&#39; and &#39;B&#39; are the interacting</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        partners in the PPI and &#39;score&#39; represents a weight of the interaction</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        for computing cell-cell interactions/communication.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">if</span> <span class="n">sort_values</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">sort_values</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="k">if</span> <span class="n">dropna</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi_data</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">,:]</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">if</span> <span class="n">strna</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">strna</span><span class="p">,</span> <span class="nb">str</span><span class="p">)),</span> <span class="s2">&quot;strna has to be an string.&quot;</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">strna</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="n">unidirectional_ppi</span> <span class="o">=</span> <span class="n">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">simplified_ppi</span> <span class="o">=</span> <span class="n">simplify_ppi</span><span class="p">(</span><span class="n">unidirectional_ppi</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="k">if</span> <span class="n">rnaseq_genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="n">simplified_ppi</span> <span class="o">=</span> <span class="n">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">simplified_ppi</span><span class="p">,</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                                                    <span class="n">proteins</span><span class="o">=</span><span class="n">rnaseq_genes</span><span class="p">,</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                                                    <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span><span class="p">,</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                                                    <span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>            <span class="n">simplified_ppi</span> <span class="o">=</span> <span class="n">filter_complex_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">simplified_ppi</span><span class="p">,</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>                                                            <span class="n">proteins</span><span class="o">=</span><span class="n">rnaseq_genes</span><span class="p">,</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>                                                            <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>                                                            <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                                                            <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                                                            <span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="n">simplified_ppi</span> <span class="o">=</span> <span class="n">simplified_ppi</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="k">return</span> <span class="n">simplified_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.remove_ppi_bidirectionality">
<code class="highlight language-python"><span class="n">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Removes duplicate interactions. For example, when ProtA-ProtB and
ProtB-ProtA interactions are present in the dataset, only one of
them will be kept.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.remove_ppi_bidirectionality--returns">Returns</h6>
<p>unidirectional_ppi : pandas.DataFrame
    List of protein-protein interactions without duplicated interactions
    in both directions (if ProtA-ProtB and ProtB-ProtA interactions are
    present, only the one that appears first is kept).</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">remove_ppi_bidirectionality</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Removes duplicate interactions. For example, when ProtA-ProtB and</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    ProtB-ProtA interactions are present in the dataset, only one of</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    them will be kept.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    -------</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    unidirectional_ppi : pandas.DataFrame</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        List of protein-protein interactions without duplicated interactions</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        in both directions (if ProtA-ProtB and ProtB-ProtA interactions are</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        present, only the one that appears first is kept).</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Removing bidirectionality of PPI network&#39;</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">header_interactorA</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">header_interactorB</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">IA</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="p">[[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">]]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">IB</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="p">[[</span><span class="n">header_interactorB</span><span class="p">,</span> <span class="n">header_interactorA</span><span class="p">]]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">IB</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">]</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">repeated_interactions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">IA</span><span class="p">,</span> <span class="n">IB</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">])</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">repeated</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">repeated_interactions</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">df</span> <span class="o">=</span>  <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">combinations</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">repeated</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">])</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="n">header_interactorB</span><span class="p">,</span> <span class="n">header_interactorA</span><span class="p">]]</span>   <span class="c1"># To keep lexicographically sorted interactions</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">]</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">unidirectional_ppi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">indicator</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;_merge==&quot;left_only&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;_merge&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">unidirectional_ppi</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">return</span> <span class="n">unidirectional_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.ppi.simplify_ppi">
<code class="highlight language-python"><span class="n">simplify_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Reduces a dataframe of protein-protein interactions into
a simpler version with only three columns (A, B and score).</p>
<h6 id="cell2cell.preprocessing.ppi.simplify_ppi--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>interaction_columns : tuple
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>score : str, default=None
    Column name where weights for the PPIs are specified. If
    None, a default score of one is automatically assigned
    to each PPI.</p>
<p>verbose : boolean, default=True
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.preprocessing.ppi.simplify_ppi--returns">Returns</h6>
<div class="codehilite"><pre><span></span><code><span class="nv">A</span> <span class="nv">simplified</span> <span class="nv">dataframe</span> <span class="nv">of</span> <span class="nv">protein</span><span class="o">-</span><span class="nv">protein</span> <span class="nv">interactions</span> <span class="nv">with</span> <span class="nv">only</span>
<span class="nv">three</span> <span class="nv">columns</span>: <span class="s1">&#39;</span><span class="s">A</span><span class="s1">&#39;</span>, <span class="s1">&#39;</span><span class="s">B</span><span class="s1">&#39;</span>, <span class="s1">&#39;</span><span class="s">score</span><span class="s1">&#39;</span>, <span class="nv">wherein</span> <span class="s1">&#39;</span><span class="s">A</span><span class="s1">&#39;</span> <span class="nv">and</span> <span class="s1">&#39;</span><span class="s">B</span><span class="s1">&#39;</span> <span class="nv">are</span> <span class="nv">the</span>
<span class="nv">interacting</span> <span class="nv">partners</span> <span class="nv">in</span> <span class="nv">the</span> <span class="nv">PPI</span> <span class="nv">and</span> <span class="s1">&#39;</span><span class="s">score</span><span class="s1">&#39;</span> <span class="nv">represents</span> <span class="nv">a</span> <span class="nv">weight</span>
<span class="nv">of</span> <span class="nv">the</span> <span class="nv">interaction</span> <span class="k">for</span> <span class="nv">computing</span> <span class="nv">cell</span><span class="o">-</span><span class="nv">cell</span> <span class="nv">interactions</span><span class="o">/</span><span class="nv">communication</span>.
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/ppi.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">simplify_ppi</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Reduces a dataframe of protein-protein interactions into</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a simpler version with only three columns (A, B and score).</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    interaction_columns : tuple</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    score : str, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Column name where weights for the PPIs are specified. If</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        None, a default score of one is automatically assigned</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        to each PPI.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    verbose : boolean, default=True</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    -------</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        A simplified dataframe of protein-protein interactions with only</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        three columns: &#39;A&#39;, &#39;B&#39;, &#39;score&#39;, wherein &#39;A&#39; and &#39;B&#39; are the</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        interacting partners in the PPI and &#39;score&#39; represents a weight</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        of the interaction for computing cell-cell interactions/communication.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Simplying PPI network&#39;</span><span class="p">)</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">header_interactorA</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">header_interactorB</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="k">if</span> <span class="n">score</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="n">simple_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="p">[[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">]]</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">simple_ppi</span> <span class="o">=</span> <span class="n">simple_ppi</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">score</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">simple_ppi</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="p">[[</span><span class="n">header_interactorA</span><span class="p">,</span> <span class="n">header_interactorB</span><span class="p">,</span> <span class="n">score</span><span class="p">]]</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">simple_ppi</span><span class="p">[</span><span class="n">score</span><span class="p">]</span> <span class="o">=</span> <span class="n">simple_ppi</span><span class="p">[</span><span class="n">score</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">(</span><span class="n">simple_ppi</span><span class="p">[</span><span class="n">score</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">]</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">simple_ppi</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">cols</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">return</span> <span class="n">simple_ppi</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq">
        <code>rnaseq</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.add_complexes_to_expression">
<code class="highlight language-python"><span class="n">add_complexes_to_expression</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">complexes</span><span class="p">,</span> <span class="n">agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Adds multimeric complexes into the gene expression matrix.
Their gene expressions are the minimum expression value
among the respective subunits composing them.</p>
<h6 id="cell2cell.preprocessing.rnaseq.add_complexes_to_expression--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>complexes : dict
    Dictionary where keys are the complex names in the list of PPIs, while
    values are list of subunits for the respective complex names.</p>
<p>agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<h6 id="cell2cell.preprocessing.rnaseq.add_complexes_to_expression--returns">Returns</h6>
<p>tmp_rna : pandas.DataFrame
    Gene expression data for RNA-seq experiment containing multimeric
    complex names. Their gene expressions are the minimum expression value
    among the respective subunits composing them. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">add_complexes_to_expression</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">complexes</span><span class="p">,</span> <span class="n">agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Adds multimeric complexes into the gene expression matrix.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Their gene expressions are the minimum expression value</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    among the respective subunits composing them.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    complexes : dict</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Dictionary where keys are the complex names in the list of PPIs, while</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        values are list of subunits for the respective complex names.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    -------</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    tmp_rna : pandas.DataFrame</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Gene expression data for RNA-seq experiment containing multimeric</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        complex names. Their gene expressions are the minimum expression value</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        among the respective subunits composing them. Columns are</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">tmp_rna</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">complexes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">g</span> <span class="ow">in</span> <span class="n">tmp_rna</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">v</span><span class="p">):</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="n">df</span> <span class="o">=</span> <span class="n">tmp_rna</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">v</span><span class="p">,</span> <span class="p">:]</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>            <span class="k">if</span> <span class="n">agg_method</span> <span class="o">==</span> <span class="s1">&#39;min&#39;</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>                <span class="n">tmp_rna</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>            <span class="k">elif</span> <span class="n">agg_method</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>                <span class="n">tmp_rna</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>            <span class="k">elif</span> <span class="n">agg_method</span> <span class="o">==</span> <span class="s1">&#39;gmean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>                <span class="n">tmp_rna</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">))))</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> is not a valid agg_method&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">agg_method</span><span class="p">))</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="n">tmp_rna</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">tmp_rna</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">return</span> <span class="n">tmp_rna</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.aggregate_single_cells">
<code class="highlight language-python"><span class="n">aggregate_single_cells</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcodes&#39;</span><span class="p">,</span> <span class="n">celltype_col</span><span class="o">=</span><span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span> <span class="n">transposed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Aggregates gene expression of single cells into cell types for each gene.</p>
<h6 id="cell2cell.preprocessing.rnaseq.aggregate_single_cells--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a single-cell RNA-seq experiment. Columns are
    single cells and rows are genes. If columns are genes and rows are
    single cells, specify transposed=True.</p>
<p>metadata : pandas.Dataframe
    Metadata containing the cell types for each single cells in the
    RNA-seq dataset.</p>
<p>barcode_col : str, default='barcodes'
    Column-name for the single cells in the metadata.</p>
<p>celltype_col : str, default='cell_types'
    Column-name in the metadata for the grouping single cells into cell types
    by the selected aggregation method.</p>
<p>method : str, default='average
    Specifies the method to use to aggregate gene expression of single
    cells into their respective cell types. Used to perform the CCI
    analysis since it is on the cell types rather than single cells.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">nn_cell_fraction</span><span class="s1">&#39;</span> : <span class="nv">Among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span>, <span class="nv">it</span>
    <span class="nv">calculates</span> <span class="nv">the</span> <span class="nv">fraction</span> <span class="nv">of</span> <span class="nv">single</span> <span class="nv">cells</span> <span class="nv">with</span> <span class="nv">non</span><span class="o">-</span><span class="nv">zero</span> <span class="nv">count</span> <span class="nv">values</span>
    <span class="nv">of</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">average</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">gene</span> <span class="nv">expression</span> <span class="nv">among</span> <span class="nv">the</span> <span class="nv">single</span> <span class="nv">cells</span>
    <span class="nv">composing</span> <span class="nv">a</span> <span class="nv">cell</span> <span class="nv">type</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">gene</span>.
</code></pre></div>

<p>transposed : boolean, default=True
    Whether the rnaseq_data is organized with columns as
    genes and rows as single cells.</p>
<h6 id="cell2cell.preprocessing.rnaseq.aggregate_single_cells--returns">Returns</h6>
<p>agg_df : pandas.DataFrame
    Dataframe containing the gene expression values that were aggregated
    by cell types. Columns are cell types and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">aggregate_single_cells</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcodes&#39;</span><span class="p">,</span> <span class="n">celltype_col</span><span class="o">=</span><span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                           <span class="n">transposed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Aggregates gene expression of single cells into cell types for each gene.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for a single-cell RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        single cells and rows are genes. If columns are genes and rows are</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        single cells, specify transposed=True.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    metadata : pandas.Dataframe</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Metadata containing the cell types for each single cells in the</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        RNA-seq dataset.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    barcode_col : str, default=&#39;barcodes&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Column-name for the single cells in the metadata.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    celltype_col : str, default=&#39;cell_types&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Column-name in the metadata for the grouping single cells into cell types</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        by the selected aggregation method.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    method : str, default=&#39;average</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Specifies the method to use to aggregate gene expression of single</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        cells into their respective cell types. Used to perform the CCI</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        analysis since it is on the cell types rather than single cells.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;nn_cell_fraction&#39; : Among the single cells composing a cell type, it</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">            calculates the fraction of single cells with non-zero count values</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">            of a given gene.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &#39;average&#39; : Computes the average gene expression among the single cells</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">            composing a cell type for a given gene.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    transposed : boolean, default=True</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Whether the rnaseq_data is organized with columns as</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        genes and rows as single cells.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    -------</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    agg_df : pandas.DataFrame</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Dataframe containing the gene expression values that were aggregated</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        by cell types. Columns are cell types and rows are genes.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">assert</span> <span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Please provide metadata containing the barcodes and cell-type annotation.&quot;</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">assert</span> <span class="n">method</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;average&#39;</span><span class="p">,</span> <span class="s1">&#39;nn_cell_fraction&#39;</span><span class="p">],</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> is not a valid option for method&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">meta</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">meta</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[[</span><span class="n">barcode_col</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">]]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">barcode_col</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">mapper</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="n">celltype_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">if</span> <span class="n">transposed</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">rnaseq_data</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">df</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">mapper</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;celltype&#39;</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="n">agg_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="k">for</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">ct_df</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;celltype&#39;</span><span class="p">):</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="n">ct_df</span> <span class="o">=</span> <span class="n">ct_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;average&#39;</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>            <span class="n">agg</span> <span class="o">=</span> <span class="n">ct_df</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;nn_cell_fraction&#39;</span><span class="p">:</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="n">agg</span> <span class="o">=</span> <span class="p">((</span><span class="n">ct_df</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">ct_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>        <span class="n">agg_df</span><span class="p">[</span><span class="n">celltype</span><span class="p">]</span> <span class="o">=</span> <span class="n">agg</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="k">return</span> <span class="n">agg_df</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.divide_expression_by_max">
<code class="highlight language-python"><span class="n">divide_expression_by_max</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Normalizes each gene value given the max value across an axis.</p>
<h6 id="cell2cell.preprocessing.rnaseq.divide_expression_by_max--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>axis : int, default=0
    Axis to perform the max-value normalization. Options
    are {0 for normalizing across rows (column-wise) or 1 for
    normalizing across columns (row-wise)}.</p>
<h6 id="cell2cell.preprocessing.rnaseq.divide_expression_by_max--returns">Returns</h6>
<p>new_data : pandas.DataFrame
    A gene expression data for RNA-seq experiment with
    normalized values. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">divide_expression_by_max</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Normalizes each gene value given the max value across an axis.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Axis to perform the max-value normalization. Options</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        are {0 for normalizing across rows (column-wise) or 1 for</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        normalizing across columns (row-wise)}.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    new_data : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A gene expression data for RNA-seq experiment with</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        normalized values. Columns are</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">new_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="ow">not</span> <span class="n">axis</span><span class="p">))</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">new_data</span> <span class="o">=</span> <span class="n">new_data</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">new_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.divide_expression_by_mean">
<code class="highlight language-python"><span class="n">divide_expression_by_mean</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Normalizes each gene value given the mean value across an axis.</p>
<h6 id="cell2cell.preprocessing.rnaseq.divide_expression_by_mean--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>axis : int, default=0
    Axis to perform the mean-value normalization. Options
    are {0 for normalizing across rows (column-wise) or 1 for
    normalizing across columns (row-wise)}.</p>
<h6 id="cell2cell.preprocessing.rnaseq.divide_expression_by_mean--returns">Returns</h6>
<p>new_data : pandas.DataFrame
    A gene expression data for RNA-seq experiment with
    normalized values. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">divide_expression_by_mean</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Normalizes each gene value given the mean value across an axis.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Axis to perform the mean-value normalization. Options</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        are {0 for normalizing across rows (column-wise) or 1 for</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        normalizing across columns (row-wise)}.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    new_data : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A gene expression data for RNA-seq experiment with</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        normalized values. Columns are</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">new_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="ow">not</span> <span class="n">axis</span><span class="p">))</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">new_data</span> <span class="o">=</span> <span class="n">new_data</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">new_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.drop_empty_genes">
<code class="highlight language-python"><span class="n">drop_empty_genes</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Drops genes that are all zeroes and/or without
expression values for all cell-types/tissues/samples.</p>
<h6 id="cell2cell.preprocessing.rnaseq.drop_empty_genes--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<h6 id="cell2cell.preprocessing.rnaseq.drop_empty_genes--returns">Returns</h6>
<p>data : pandas.DataFrame
    A gene expression data for RNA-seq experiment without
    empty genes. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">drop_empty_genes</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Drops genes that are all zeroes and/or without</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    expression values for all cell-types/tissues/samples.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    data : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A gene expression data for RNA-seq experiment without</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        empty genes. Columns are</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># Put zeros to all missing values</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>  <span class="c1"># Drop rows will 0 among all cell/tissues</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.log10_transformation">
<code class="highlight language-python"><span class="n">log10_transformation</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">addition</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Log-transforms gene expression values in a
gene expression matrix for a RNA-seq experiment.</p>
<h6 id="cell2cell.preprocessing.rnaseq.log10_transformation--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<h6 id="cell2cell.preprocessing.rnaseq.log10_transformation--returns">Returns</h6>
<p>data : pandas.DataFrame
    A gene expression data for RNA-seq experiment with
    log-transformed values. Values are log10(expression + addition).
    Columns are cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">log10_transformation</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">addition</span> <span class="o">=</span> <span class="mf">1e-6</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Log-transforms gene expression values in a</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    gene expression matrix for a RNA-seq experiment.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    -------</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    data : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A gene expression data for RNA-seq experiment with</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        log-transformed values. Values are log10(expression + addition).</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Columns are cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="c1">### Apply this only after applying &quot;drop_empty_genes&quot; function</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">addition</span><span class="p">))</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">replace</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.rnaseq.scale_expression_by_sum">
<code class="highlight language-python"><span class="n">scale_expression_by_sum</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_value</span><span class="o">=</span><span class="mf">1000000.0</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Normalizes all samples to sum up the same scale factor.</p>
<h6 id="cell2cell.preprocessing.rnaseq.scale_expression_by_sum--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for RNA-seq experiment. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>axis : int, default=0
    Axis to perform the global-scaling normalization. Options
    are {0 for normalizing across rows (column-wise) or 1 for
    normalizing across columns (row-wise)}.</p>
<p>sum_value : float, default=1e6
    Scaling factor. Normalized axis will sum up this value.</p>
<h6 id="cell2cell.preprocessing.rnaseq.scale_expression_by_sum--returns">Returns</h6>
<p>scaled_data : pandas.DataFrame
    A gene expression data for RNA-seq experiment with
    scaled values. All rows or columns, depending on the specified
    axis sum up to the same value. Columns are
    cell-types/tissues/samples and rows are genes.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/rnaseq.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">scale_expression_by_sum</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_value</span><span class="o">=</span><span class="mf">1e6</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Normalizes all samples to sum up the same scale factor.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for RNA-seq experiment. Columns are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Axis to perform the global-scaling normalization. Options</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        are {0 for normalizing across rows (column-wise) or 1 for</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        normalizing across columns (row-wise)}.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    sum_value : float, default=1e6</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Scaling factor. Normalized axis will sum up this value.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    -------</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    scaled_data : pandas.DataFrame</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        A gene expression data for RNA-seq experiment with</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        scaled values. All rows or columns, depending on the specified</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        axis sum up to the same value. Columns are</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">sum_value</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">))</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">scaled_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">return</span> <span class="n">scaled_data</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.preprocessing.signal">
        <code>signal</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.preprocessing.signal.smooth_curve">
<code class="highlight language-python"><span class="n">smooth_curve</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">window_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">polyorder</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Apply a Savitzky-Golay filter to an array to smooth the curve.</p>
<h6 id="cell2cell.preprocessing.signal.smooth_curve--parameters">Parameters</h6>
<p>values : array-like
    An array or list of values.</p>
<p>window_length : int, default=None
    Size of the window of values to use too smooth the curve.</p>
<p>polyorder : int, default=3
    The order of the polynomial used to fit the samples.</p>
<p>**kwargs : dict
    Extra arguments for the scipy.signal.savgol_filter function.</p>
<h6 id="cell2cell.preprocessing.signal.smooth_curve--returns">Returns</h6>
<p>smooth_values : array-like
    An array or list of values representing the smooth curvee.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/preprocessing/signal.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">smooth_curve</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">window_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">polyorder</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Apply a Savitzky-Golay filter to an array to smooth the curve.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    values : array-like</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        An array or list of values.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    window_length : int, default=None</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Size of the window of values to use too smooth the curve.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    polyorder : int, default=3</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        The order of the polynomial used to fit the samples.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Extra arguments for the scipy.signal.savgol_filter function.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    smooth_values : array-like</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        An array or list of values representing the smooth curvee.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">if</span> <span class="n">window_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="n">window_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span> <span class="o">/</span> <span class="nb">min</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="p">]))</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">if</span> <span class="n">window_length</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">window_length</span> <span class="o">+=</span> <span class="mi">1</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">assert</span><span class="p">(</span><span class="n">polyorder</span> <span class="o">&lt;</span> <span class="n">window_length</span><span class="p">),</span> <span class="s2">&quot;polyorder must be less than window_length.&quot;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">smooth_values</span> <span class="o">=</span> <span class="n">savgol_filter</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">window_length</span><span class="p">,</span> <span class="n">polyorder</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">return</span> <span class="n">smooth_values</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.spatial">
        <code>spatial</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.spatial.distances">
        <code>distances</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.distances.celltype_pair_distance">
<code class="highlight language-python"><span class="n">celltype_pair_distance</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates the distance between two sets of data points (single cell coordinates)
represented by df1 and df2. It supports two distance metrics: Euclidean and Manhattan
distances. The method parameter allows you to specify how the distances between the
two sets are aggregated.</p>
<h6 id="cell2cell.spatial.distances.celltype_pair_distance--parameters">Parameters</h6>
<p>df1 : pandas.DataFrame
    The first set of single cell coordinates.</p>
<p>df1 : pandas.DataFrame
    The second set of single cell coordinates.</p>
<p>method : str, default='min'
    The aggregation method for the calculated distances. It can be one of 'min',
    'max', or 'mean'.</p>
<p>distance : str, default='euclidean'
    The distance metric to use. It can be 'euclidean' or 'manhattan'.</p>
<h6 id="cell2cell.spatial.distances.celltype_pair_distance--returns">Returns</h6>
<p>agg_dist : numpy.float
    The aggregated distance between the two sets of data points based on the specified
    method and distance metric.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/distances.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">celltype_pair_distance</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Calculates the distance between two sets of data points (single cell coordinates)</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    represented by df1 and df2. It supports two distance metrics: Euclidean and Manhattan</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    distances. The method parameter allows you to specify how the distances between the</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    two sets are aggregated.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    df1 : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        The first set of single cell coordinates.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    df1 : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        The second set of single cell coordinates.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The aggregation method for the calculated distances. It can be one of &#39;min&#39;,</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        &#39;max&#39;, or &#39;mean&#39;.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    distance : str, default=&#39;euclidean&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        The distance metric to use. It can be &#39;euclidean&#39; or &#39;manhattan&#39;.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    -------</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    agg_dist : numpy.float</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        The aggregated distance between the two sets of data points based on the specified</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        method and distance metric.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">if</span> <span class="n">distance</span> <span class="o">==</span> <span class="s1">&#39;euclidean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">distances</span> <span class="o">=</span> <span class="n">euclidean_distances</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">elif</span> <span class="n">distance</span> <span class="o">==</span> <span class="s1">&#39;manhattan&#39;</span><span class="p">:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="n">distances</span> <span class="o">=</span> <span class="n">manhattan_distances</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> distance is not implemented.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">distance</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()))</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;min&#39;</span><span class="p">:</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="n">agg_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">agg_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">agg_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;Method </span><span class="si">{}</span><span class="s1"> is not implemented.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">))</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">return</span> <span class="n">agg_dist</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.distances.pairwise_celltype_distances">
<code class="highlight language-python"><span class="n">pairwise_celltype_distances</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group_col</span><span class="p">,</span> <span class="n">coord_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">pairs</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates pairwise distances between groups of single cells. It computes an
aggregate distance between all possible combinations of groups.</p>
<h6 id="cell2cell.spatial.distances.pairwise_celltype_distances--parameters">Parameters</h6>
<p>df : pandas.DataFrame
    A dataframe where each row is a single cell, and there are columns containing
    spatial coordinates and cell group.</p>
<p>group_col : str
    The name of the column that defines the groups for which distances are calculated.</p>
<p>coord_cols : list, default=None
    The list of column names that represent the coordinates of the single cells.</p>
<p>pairs : list
    A list of specific group pairs for which distances should be calculated.
    If not provided, all possible combinations of group pairs will be considered.</p>
<h6 id="cell2cell.spatial.distances.pairwise_celltype_distances--returns">Returns</h6>
<p>distances : pandas.DataFrame
    The pairwise distances between groups based on the specified group column.
    In this dataframe rows and columns are the cell groups used to compute distances.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/distances.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pairwise_celltype_distances</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group_col</span><span class="p">,</span> <span class="n">coord_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">],</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                <span class="n">method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">pairs</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Calculates pairwise distances between groups of single cells. It computes an</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    aggregate distance between all possible combinations of groups.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    df : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        A dataframe where each row is a single cell, and there are columns containing</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        spatial coordinates and cell group.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    group_col : str</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        The name of the column that defines the groups for which distances are calculated.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    coord_cols : list, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The list of column names that represent the coordinates of the single cells.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    pairs : list</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        A list of specific group pairs for which distances should be calculated.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        If not provided, all possible combinations of group pairs will be considered.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    -------</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    distances : pandas.DataFrame</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        The pairwise distances between groups based on the specified group column.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        In this dataframe rows and columns are the cell groups used to compute distances.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="c1"># TODO: Adapt code below to receive AnnData or MuData objects</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="c1"># df_ = pd.DataFrame(adata.obsm[&#39;spatial&#39;], index=adata.obs_names, columns=[&#39;X&#39;, &#39;Y&#39;])</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="c1"># df = adata.obs[[group_col]]</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">df_</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">coord_cols</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">groups</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="n">distances</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">))),</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>                             <span class="n">index</span><span class="o">=</span><span class="n">groups</span><span class="p">,</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>                             <span class="n">columns</span><span class="o">=</span><span class="n">groups</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="n">pairs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">dist</span> <span class="o">=</span> <span class="n">celltype_pair_distance</span><span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">df_</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">group_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                                      <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                      <span class="n">distance</span><span class="o">=</span><span class="n">distance</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                      <span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">dist</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">dist</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="k">return</span> <span class="n">distances</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.spatial.filtering">
        <code>filtering</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.filtering.dist_filter_liana">
<code class="highlight language-python"><span class="n">dist_filter_liana</span><span class="p">(</span><span class="n">liana_outputs</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">source_col</span><span class="o">=</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">keep_dist</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters a dataframe with outputs from LIANA based on a distance threshold
defined applied to another dataframe containing distances between cell groups.</p>
<h6 id="cell2cell.spatial.filtering.dist_filter_liana--parameters">Parameters</h6>
<p>liana_outputs : pandas.DataFrame
    Dataframe containing the results from LIANA, where rows are pairs of
    ligand-receptor interactions by pair of source-target cell groups.</p>
<p>distances : pandas.DataFrame
    Square dataframe containing distances between pairs of cell groups.</p>
<p>max_dist : float
    The distance threshold used to filter the pairs from the liana_outputs dataframe.</p>
<p>min_dist : float, default=0
    The minimum distance between cell pairs to consider them in the interaction tensor.</p>
<p>source_col : str, default='source'
    Column name in both dataframes that represents the source cell groups.</p>
<p>target_col : str, default='target'
     Column name in both dataframes that represents the target cell groups.</p>
<p>keep_dist : bool, default=False
    To determine whether to keep the 'distance' column in the filtered output.
    If set to True, the 'distance' column will be retained; otherwise, it will be dropped
    and the LIANA dataframe will contain the original columns.</p>
<h6 id="cell2cell.spatial.filtering.dist_filter_liana--returns">Returns</h6>
<p>filtered_liana_outputs : pandas.DataFrame
    It containing pairs from the liana_outputs dataframe that meet the distance
    threshold criteria.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/filtering.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">dist_filter_liana</span><span class="p">(</span><span class="n">liana_outputs</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">source_col</span><span class="o">=</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s1">&#39;target&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                      <span class="n">keep_dist</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Filters a dataframe with outputs from LIANA based on a distance threshold</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    defined applied to another dataframe containing distances between cell groups.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    liana_outputs : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Dataframe containing the results from LIANA, where rows are pairs of</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        ligand-receptor interactions by pair of source-target cell groups.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    distances : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Square dataframe containing distances between pairs of cell groups.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    max_dist : float</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The distance threshold used to filter the pairs from the liana_outputs dataframe.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    min_dist : float, default=0</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        The minimum distance between cell pairs to consider them in the interaction tensor.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    source_col : str, default=&#39;source&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Column name in both dataframes that represents the source cell groups.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    target_col : str, default=&#39;target&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">         Column name in both dataframes that represents the target cell groups.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    keep_dist : bool, default=False</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        To determine whether to keep the &#39;distance&#39; column in the filtered output.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        If set to True, the &#39;distance&#39; column will be retained; otherwise, it will be dropped</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        and the LIANA dataframe will contain the original columns.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    -------</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    filtered_liana_outputs : pandas.DataFrame</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        It containing pairs from the liana_outputs dataframe that meet the distance</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        threshold criteria.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="c1"># Convert distances to a long-form dataframe</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">distances</span> <span class="o">=</span> <span class="n">distances</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">distances</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">source_col</span><span class="p">,</span> <span class="n">target_col</span><span class="p">,</span> <span class="s1">&#39;distance&#39;</span><span class="p">]</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Merge the long-form distances DataFrame with pairs_df</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">liana_outputs</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">distances</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="n">source_col</span><span class="p">,</span> <span class="n">target_col</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="c1"># Filter based on the distance threshold</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="n">filtered_liana_outputs</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[(</span><span class="n">min_dist</span> <span class="o">&lt;=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_dist</span><span class="p">)]</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">if</span> <span class="n">keep_dist</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">filtered_liana_outputs</span> <span class="o">=</span> <span class="n">filtered_liana_outputs</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">return</span> <span class="n">filtered_liana_outputs</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.filtering.dist_filter_tensor">
<code class="highlight language-python"><span class="n">dist_filter_tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">source_axis</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">target_axis</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Filters an Interaction Tensor based on intercellular distances between cell types.</p>
<h6 id="cell2cell.spatial.filtering.dist_filter_tensor--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor</p>
<p>distances : pandas.DataFrame
    Square dataframe containing distances between pairs of cell groups. It must contain
    all cell groups that act as sender and receiver cells in the tensor.</p>
<p>max_dist : float
    The maximum distance between cell pairs to consider them in the interaction tensor.</p>
<p>min_dist : float, default=0
    The minimum distance between cell pairs to consider them in the interaction tensor.</p>
<p>source_axis : int, default=2
    The index indicating the axis in the tensor corresponding to sender cells.</p>
<p>target_axis : int, default=3
    The index indicating the axis in the tensor corresponding to receiver cells.</p>
<h6 id="cell2cell.spatial.filtering.dist_filter_tensor--returns">Returns</h6>
<p>new_interaction_tensor : cell2cell.tensor.BaseTensor
    A tensor with communication scores made zero for cell type pairs with intercellular
    distance over the distance threshold.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/filtering.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">dist_filter_tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">source_axis</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">target_axis</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Filters an Interaction Tensor based on intercellular distances between cell types.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell2cell.tensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    distances : pandas.DataFrame</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Square dataframe containing distances between pairs of cell groups. It must contain</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        all cell groups that act as sender and receiver cells in the tensor.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    max_dist : float</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        The maximum distance between cell pairs to consider them in the interaction tensor.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    min_dist : float, default=0</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        The minimum distance between cell pairs to consider them in the interaction tensor.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    source_axis : int, default=2</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        The index indicating the axis in the tensor corresponding to sender cells.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    target_axis : int, default=3</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        The index indicating the axis in the tensor corresponding to receiver cells.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    -------</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    new_interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        A tensor with communication scores made zero for cell type pairs with intercellular</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        distance over the distance threshold.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="c1"># Evaluate whether we provide distances for all cell types in the tensor</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">cell</span> <span class="ow">in</span> <span class="n">distances</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>                <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">source_axis</span><span class="p">]]),</span> <span class="s2">&quot;Distances not provided for all sender cells&quot;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">cell</span> <span class="ow">in</span> <span class="n">distances</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>                <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">target_axis</span><span class="p">]]),</span> <span class="s2">&quot;Distances not provided for all receiver cells&quot;</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">source_cell_groups</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">source_axis</span><span class="p">]</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">target_cell_groups</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">target_axis</span><span class="p">]</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="c1"># Use only cell types in the tensor</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">dist_df</span> <span class="o">=</span> <span class="n">distances</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">source_cell_groups</span><span class="p">,</span> <span class="n">target_cell_groups</span><span class="p">]</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="c1"># Filter cell types by intercellular distances</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">dist</span> <span class="o">=</span> <span class="p">((</span><span class="n">min_dist</span> <span class="o">&lt;=</span> <span class="n">dist_df</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">dist_df</span> <span class="o">&lt;=</span> <span class="n">max_dist</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="c1"># Mapping what re-arrange should be done to keep the original tensor shape</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">tensor_shape</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">original_order</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tensor_shape</span><span class="p">)))</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">new_order</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="c1"># Generate template tensor with cells to keep</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="n">template_tensor</span> <span class="o">=</span> <span class="n">dist</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tensor_shape</span><span class="p">):</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">!=</span> <span class="n">source_axis</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">i</span> <span class="o">!=</span> <span class="n">target_axis</span><span class="p">):</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>            <span class="n">template_tensor</span> <span class="o">=</span> <span class="p">[</span><span class="n">template_tensor</span><span class="p">]</span> <span class="o">*</span> <span class="n">size</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>            <span class="n">new_order</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="n">template_tensor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">template_tensor</span><span class="p">)</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="n">new_order</span> <span class="o">+=</span> <span class="p">[</span><span class="n">source_axis</span><span class="p">,</span> <span class="n">target_axis</span><span class="p">]</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="n">changes_needed</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_order</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">original_order</span><span class="p">]</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="c1"># Re-arrange axes by the order</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="n">template_tensor</span> <span class="o">=</span> <span class="n">template_tensor</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">changes_needed</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="c1"># Create tensorly object</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="n">template_tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">template_tensor</span><span class="p">,</span> <span class="o">**</span><span class="n">tl</span><span class="o">.</span><span class="n">context</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">))</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="k">assert</span> <span class="n">template_tensor</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="s2">&quot;Filtering of cells was not properly done. Revise code of this function (template tensor)&quot;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="c1"># tensor = tl.zeros_like(interaction_tensor.tensor, **tl.context(tensor))</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="n">new_interaction_tensor</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="n">new_interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">new_interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">*</span> <span class="n">template_tensor</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="c1"># Make masked cells by distance to be real zeros</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="n">new_interaction_tensor</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="p">(</span><span class="n">new_interaction_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="n">new_interaction_tensor</span><span class="o">.</span><span class="n">loc_nans</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">return</span> <span class="n">new_interaction_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.spatial.neighborhoods">
        <code>neighborhoods</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.neighborhoods.add_sliding_window_info_to_adata">
<code class="highlight language-python"><span class="n">add_sliding_window_info_to_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">window_mapping</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Adds window information to the AnnData object's .obs DataFrame. Each window is represented
as a column, and cells/spots belonging to a window are marked with a 1.0, while others are marked
with a 0.0. It modifies the <code>adata</code> object in place.</p>
<h6 id="cell2cell.spatial.neighborhoods.add_sliding_window_info_to_adata--parameters">Parameters</h6>
<p>adata : AnnData
    The AnnData object to which the window information will be added.</p>
<p>window_mapping : dict
    A dictionary mapping each window to a set of cell/spot indeces or barcodes.
    This is the output from the <code>create_moving_windows</code> function.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/neighborhoods.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">add_sliding_window_info_to_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">window_mapping</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Adds window information to the AnnData object&#39;s .obs DataFrame. Each window is represented</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    as a column, and cells/spots belonging to a window are marked with a 1.0, while others are marked</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    with a 0.0. It modifies the `adata` object in place.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    adata : AnnData</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        The AnnData object to which the window information will be added.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    window_mapping : dict</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        A dictionary mapping each window to a set of cell/spot indeces or barcodes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        This is the output from the `create_moving_windows` function.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="c1"># Initialize all window columns to 0.0</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">for</span> <span class="n">window</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">window_mapping</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">window</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="c1"># Mark cells that belong to each window</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">for</span> <span class="n">window</span><span class="p">,</span> <span class="n">barcode_indeces</span> <span class="ow">in</span> <span class="n">window_mapping</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">barcode_indeces</span><span class="p">,</span> <span class="n">window</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.neighborhoods.calculate_window_size">
<code class="highlight language-python"><span class="n">calculate_window_size</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">num_windows</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Calculates the window size required to fit a specified number of windows
across the width of the coordinate space in spatial transcriptomics data.</p>
<h6 id="cell2cell.spatial.neighborhoods.calculate_window_size--parameters">Parameters</h6>
<p>adata : AnnData
    The AnnData object containing spatial transcriptomics data. The spatial coordinates
    must be stored in <code>adata.obsm['spatial']</code>.</p>
<p>num_windows : int
    The desired number of windows to fit across the width of the coordinate space.</p>
<h6 id="cell2cell.spatial.neighborhoods.calculate_window_size--returns">Returns</h6>
<p>window_size : float
    The calculated size of each window to fit the specified number of windows
    across the width of the coordinate space.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/neighborhoods.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">calculate_window_size</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">num_windows</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Calculates the window size required to fit a specified number of windows</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    across the width of the coordinate space in spatial transcriptomics data.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    adata : AnnData</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        The AnnData object containing spatial transcriptomics data. The spatial coordinates</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        must be stored in `adata.obsm[&#39;spatial&#39;]`.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    num_windows : int</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        The desired number of windows to fit across the width of the coordinate space.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    -------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    window_size : float</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        The calculated size of each window to fit the specified number of windows</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        across the width of the coordinate space.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="c1"># Extract X coordinates</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">x_coords</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="c1"># Determine the range of X coordinates</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">x_coords</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">x_coords</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="c1"># Calculate the window size</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">window_size</span> <span class="o">=</span> <span class="p">(</span><span class="n">x_max</span> <span class="o">-</span> <span class="n">x_min</span><span class="p">)</span> <span class="o">/</span> <span class="n">num_windows</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">return</span> <span class="n">window_size</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.neighborhoods.create_sliding_windows">
<code class="highlight language-python"><span class="n">create_sliding_windows</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">stride</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Maps windows to the cells they contain based on spatial transcriptomics data.
Returns a dictionary where keys are window identifiers and values are sets of cell indices.</p>
<h6 id="cell2cell.spatial.neighborhoods.create_sliding_windows--parameters">Parameters</h6>
<p>adata : AnnData
    The AnnData object containing spatial transcriptomics data. The spatial coordinates
    must be stored in <code>adata.obsm['spatial']</code>.</p>
<p>window_size : float
    The size of each square window along each dimension.</p>
<p>stride : float
    The stride with which the window moves along each dimension.</p>
<h6 id="cell2cell.spatial.neighborhoods.create_sliding_windows--returns">Returns</h6>
<p>window_mapping : dict
    A dictionary mapping each window to a set of cell indices that fall within that window.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/neighborhoods.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">create_sliding_windows</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">stride</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Maps windows to the cells they contain based on spatial transcriptomics data.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Returns a dictionary where keys are window identifiers and values are sets of cell indices.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    adata : AnnData</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        The AnnData object containing spatial transcriptomics data. The spatial coordinates</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        must be stored in `adata.obsm[&#39;spatial&#39;]`.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    window_size : float</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        The size of each square window along each dimension.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    stride : float</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        The stride with which the window moves along each dimension.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    -------</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    window_mapping : dict</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        A dictionary mapping each window to a set of cell indices that fall within that window.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="c1"># Get the spatial coordinates</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="n">coords</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">])</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="c1"># Define the range of the sliding windows</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="n">x_min</span><span class="p">,</span> <span class="n">y_min</span> <span class="o">=</span> <span class="n">coords</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">x_max</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">=</span> <span class="n">coords</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">x_windows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">+</span> <span class="n">stride</span><span class="p">,</span> <span class="n">stride</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">y_windows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">+</span> <span class="n">stride</span><span class="p">,</span> <span class="n">stride</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="c1"># Function to find all windows a point belongs to</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">def</span> <span class="nf">find_windows</span><span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">window_edges</span><span class="p">):</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">return</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">edge</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">window_edges</span><span class="p">)</span> <span class="k">if</span> <span class="n">edge</span> <span class="o">&lt;=</span> <span class="n">coord</span> <span class="o">&lt;</span> <span class="n">edge</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="c1"># Initialize the window mapping</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">window_mapping</span> <span class="o">=</span> <span class="p">{}</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="c1"># Assign cells to all overlapping windows</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">for</span> <span class="n">cell_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">],</span> <span class="n">coords</span><span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">])):</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="n">cell_windows</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;window_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">wx</span><span class="p">,</span> <span class="n">wy</span><span class="p">)</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                        <span class="k">for</span> <span class="n">wx</span> <span class="ow">in</span> <span class="n">find_windows</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">x_windows</span><span class="p">)</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                        <span class="k">for</span> <span class="n">wy</span> <span class="ow">in</span> <span class="n">find_windows</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">y_windows</span><span class="p">)]</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="k">for</span> <span class="n">win</span> <span class="ow">in</span> <span class="n">cell_windows</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>            <span class="k">if</span> <span class="n">win</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">window_mapping</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>                <span class="n">window_mapping</span><span class="p">[</span><span class="n">win</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>            <span class="n">window_mapping</span><span class="p">[</span><span class="n">win</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">coords</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">cell_idx</span><span class="p">])</span>  <span class="c1"># This stores the cell/spot barcodes</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="c1"># For memory efficiency, it could be `window_mapping[win].add(cell_idx)` instead</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">return</span> <span class="n">window_mapping</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.spatial.neighborhoods.create_spatial_grid">
<code class="highlight language-python"><span class="n">create_spatial_grid</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Segments spatial transcriptomics data into a square grid based on spatial coordinates
and annotates each cell or spot with its corresponding grid position.</p>
<h6 id="cell2cell.spatial.neighborhoods.create_spatial_grid--parameters">Parameters</h6>
<p>adata : AnnData
    The AnnData object containing spatial transcriptomics data. The spatial coordinates
    must be stored in <code>adata.obsm['spatial']</code>. This object is either modified in place
    or a copy is returned based on the <code>copy</code> parameter.</p>
<p>num_bins : int
    The number of bins (squares) along each dimension of the grid. The grid is square,
    so this number applies to both the horizontal and vertical divisions.</p>
<p>copy : bool, default=False
    If True, the function operates on and returns a copy of the input AnnData object.
    If False, the function modifies the input AnnData object in place.</p>
<h6 id="cell2cell.spatial.neighborhoods.create_spatial_grid--returns">Returns</h6>
<p>adata_ : AnnData or None
    If <code>copy=True</code>, a new AnnData object with added grid annotations is returned.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/spatial/neighborhoods.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">create_spatial_grid</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Segments spatial transcriptomics data into a square grid based on spatial coordinates</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    and annotates each cell or spot with its corresponding grid position.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    adata : AnnData</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        The AnnData object containing spatial transcriptomics data. The spatial coordinates</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        must be stored in `adata.obsm[&#39;spatial&#39;]`. This object is either modified in place</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        or a copy is returned based on the `copy` parameter.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    num_bins : int</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        The number of bins (squares) along each dimension of the grid. The grid is square,</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        so this number applies to both the horizontal and vertical divisions.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    copy : bool, default=False</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        If True, the function operates on and returns a copy of the input AnnData object.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        If False, the function modifies the input AnnData object in place.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    -------</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    adata_ : AnnData or None</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        If `copy=True`, a new AnnData object with added grid annotations is returned.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">if</span> <span class="n">copy</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">adata_</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="n">adata_</span> <span class="o">=</span> <span class="n">adata</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="c1"># Get the spatial coordinates</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">coords</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">])</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="c1"># Define the bins for each dimension</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">x_min</span><span class="p">,</span> <span class="n">y_min</span> <span class="o">=</span> <span class="n">coords</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">x_max</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">=</span> <span class="n">coords</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">x_bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">,</span> <span class="n">num_bins</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">y_bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">,</span> <span class="n">num_bins</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="c1"># Digitize the coordinates into bins</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_x&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">digitize</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">],</span> <span class="n">x_bins</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_y&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">digitize</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">],</span> <span class="n">y_bins</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="c1"># Adjust indices to start from 0 and end at num_bins - 1</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_x&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_x&#39;</span><span class="p">],</span> <span class="mi">0</span><span class="p">,</span> <span class="n">num_bins</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_y&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_y&#39;</span><span class="p">],</span> <span class="mi">0</span><span class="p">,</span> <span class="n">num_bins</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="c1"># Combine grid indices to form a grid cell identifier</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_cell&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_x&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;grid_y&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">if</span> <span class="n">copy</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="k">return</span> <span class="n">adata_</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.stats">
        <code>stats</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.stats.enrichment">
        <code>enrichment</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.enrichment.fisher_representation">
<code class="highlight language-python"><span class="n">fisher_representation</span><span class="p">(</span><span class="n">sample_size</span><span class="p">,</span> <span class="n">class_in_sample</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Performs an analysis of enrichment/depletion based on observation
in a sample. It computes a p-value given a fisher exact test.</p>
<h6 id="cell2cell.stats.enrichment.fisher_representation--parameters">Parameters</h6>
<p>sample_size : int
    Size of the sample obtained or number of elements
    obtained from the analysis.</p>
<p>class_in_sample : int
    Number of elements of a given class that are
    contained in the sample. This is the class to be tested.</p>
<p>population_size : int
    Size of the sampling space. That is, the total number
    of possible elements to be chosen when sampling.</p>
<p>class_in_population : int
    Number of elements of a given class that are contained
    in the population. This is the class to be tested.</p>
<h6 id="cell2cell.stats.enrichment.fisher_representation--returns">Returns</h6>
<p>results : dict
    A dictionary containing the odd ratios and p-values for
    depletion and enrichment analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/enrichment.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">fisher_representation</span><span class="p">(</span><span class="n">sample_size</span><span class="p">,</span> <span class="n">class_in_sample</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Performs an analysis of enrichment/depletion based on observation</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in a sample. It computes a p-value given a fisher exact test.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    sample_size : int</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Size of the sample obtained or number of elements</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        obtained from the analysis.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    class_in_sample : int</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Number of elements of a given class that are</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        contained in the sample. This is the class to be tested.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    population_size : int</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Size of the sampling space. That is, the total number</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        of possible elements to be chosen when sampling.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    class_in_population : int</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Number of elements of a given class that are contained</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        in the population. This is the class to be tested.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    results : dict</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        A dictionary containing the odd ratios and p-values for</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        depletion and enrichment analysis.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="c1"># Computing the number of elements that are not in the same class</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">nonclass_in_sample</span> <span class="o">=</span> <span class="n">sample_size</span> <span class="o">-</span> <span class="n">class_in_sample</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">nonclass_in_population</span> <span class="o">=</span> <span class="n">population_size</span> <span class="o">-</span> <span class="n">class_in_population</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># Remaining elements in population after sampling</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">rem_class</span> <span class="o">=</span> <span class="n">class_in_population</span> <span class="o">-</span> <span class="n">class_in_sample</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">rem_nonclass</span> <span class="o">=</span> <span class="n">nonclass_in_population</span> <span class="o">-</span> <span class="n">nonclass_in_sample</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="c1"># Depletion Analysis</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">depletion_odds</span><span class="p">,</span> <span class="n">depletion_fisher_p_val</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">fisher_exact</span><span class="p">([[</span><span class="n">class_in_sample</span><span class="p">,</span> <span class="n">rem_class</span><span class="p">],</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>                                                              <span class="p">[</span><span class="n">nonclass_in_sample</span><span class="p">,</span> <span class="n">rem_nonclass</span><span class="p">]],</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>                                                             <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;less&#39;</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Enrichment Analysis</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">enrichment_odds</span><span class="p">,</span> <span class="n">enrichment_fisher_p_val</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">fisher_exact</span><span class="p">([[</span><span class="n">class_in_sample</span><span class="p">,</span> <span class="n">rem_class</span><span class="p">],</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                                                <span class="p">[</span><span class="n">nonclass_in_sample</span><span class="p">,</span> <span class="n">rem_nonclass</span><span class="p">]],</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                                                               <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;greater&#39;</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="n">p_vals</span> <span class="o">=</span> <span class="p">(</span><span class="n">depletion_fisher_p_val</span><span class="p">,</span> <span class="n">enrichment_fisher_p_val</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">odds</span> <span class="o">=</span> <span class="p">(</span><span class="n">depletion_odds</span><span class="p">,</span> <span class="n">enrichment_odds</span><span class="p">)</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;pval&#39;</span> <span class="p">:</span> <span class="n">p_vals</span><span class="p">,</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>               <span class="s1">&#39;odds&#39;</span> <span class="p">:</span> <span class="n">odds</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>              <span class="p">}</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="k">return</span> <span class="n">results</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.enrichment.hypergeom_representation">
<code class="highlight language-python"><span class="n">hypergeom_representation</span><span class="p">(</span><span class="n">sample_size</span><span class="p">,</span> <span class="n">class_in_sample</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Performs an analysis of enrichment/depletion based on observation
in a sample. It computes a p-value given a hypergeometric
distribution.</p>
<h6 id="cell2cell.stats.enrichment.hypergeom_representation--parameters">Parameters</h6>
<p>sample_size : int
    Size of the sample obtained or number of elements
    obtained from the analysis.</p>
<p>class_in_sample : int
    Number of elements of a given class that are
    contained in the sample. This is the class to be tested.</p>
<p>population_size : int
    Size of the sampling space. That is, the total number
    of possible elements to be chosen when sampling.</p>
<p>class_in_population : int
    Number of elements of a given class that are contained
    in the population. This is the class to be tested.</p>
<h6 id="cell2cell.stats.enrichment.hypergeom_representation--returns">Returns</h6>
<p>p_vals : tuple
    A tuple containing the p-values for depletion and
    enrichment analysis, respectively.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/enrichment.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">hypergeom_representation</span><span class="p">(</span><span class="n">sample_size</span><span class="p">,</span> <span class="n">class_in_sample</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Performs an analysis of enrichment/depletion based on observation</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    in a sample. It computes a p-value given a hypergeometric</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    distribution.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    sample_size : int</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Size of the sample obtained or number of elements</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        obtained from the analysis.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    class_in_sample : int</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Number of elements of a given class that are</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        contained in the sample. This is the class to be tested.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    population_size : int</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Size of the sampling space. That is, the total number</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        of possible elements to be chosen when sampling.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    class_in_population : int</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Number of elements of a given class that are contained</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        in the population. This is the class to be tested.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    -------</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    p_vals : tuple</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        A tuple containing the p-values for depletion and</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        enrichment analysis, respectively.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="c1"># Computing the number of elements that are not in the same class</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">nonclass_in_sample</span> <span class="o">=</span> <span class="n">sample_size</span> <span class="o">-</span> <span class="n">class_in_sample</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">nonclass_in_population</span> <span class="o">=</span> <span class="n">population_size</span> <span class="o">-</span> <span class="n">class_in_population</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="c1"># Remaining elements in population after sampling</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">rem_class</span> <span class="o">=</span> <span class="n">class_in_population</span> <span class="o">-</span> <span class="n">class_in_sample</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">rem_nonclass</span> <span class="o">=</span> <span class="n">nonclass_in_population</span> <span class="o">-</span> <span class="n">nonclass_in_sample</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="c1"># Depletion Analysis</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">depletion_hyp_p_val</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">hypergeom</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">class_in_sample</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">,</span> <span class="n">sample_size</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="c1"># Enrichment Analysis</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">enrichment_hyp_p_val</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">st</span><span class="o">.</span><span class="n">hypergeom</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">class_in_sample</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">population_size</span><span class="p">,</span> <span class="n">class_in_population</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                                  <span class="n">sample_size</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">p_vals</span> <span class="o">=</span> <span class="p">(</span><span class="n">depletion_hyp_p_val</span><span class="p">,</span> <span class="n">enrichment_hyp_p_val</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">return</span> <span class="n">p_vals</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.stats.gini">
        <code>gini</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.gini.gini_coefficient">
<code class="highlight language-python"><span class="n">gini_coefficient</span><span class="p">(</span><span class="n">distribution</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the Gini coefficient of an array of values.
Code borrowed from:
https://stackoverflow.com/questions/39512260/calculating-gini-coefficient-in-python-numpy</p>
<h6 id="cell2cell.stats.gini.gini_coefficient--parameters">Parameters</h6>
<p>distribution : array-like
    An array of values representing the distribution
    to be evaluated.</p>
<h6 id="cell2cell.stats.gini.gini_coefficient--returns">Returns</h6>
<p>gini : float
    Gini coefficient for the evaluated distribution.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/gini.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">gini_coefficient</span><span class="p">(</span><span class="n">distribution</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;Computes the Gini coefficient of an array of values.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Code borrowed from:</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    https://stackoverflow.com/questions/39512260/calculating-gini-coefficient-in-python-numpy</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    distribution : array-like</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        An array of values representing the distribution</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        to be evaluated.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    gini : float</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Gini coefficient for the evaluated distribution.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="n">diffsum</span> <span class="o">=</span> <span class="mi">0</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">xi</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">distribution</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="n">diffsum</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">xi</span> <span class="o">-</span> <span class="n">distribution</span><span class="p">[</span><span class="n">i</span><span class="p">:]))</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">gini</span> <span class="o">=</span> <span class="n">diffsum</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">distribution</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">distribution</span><span class="p">))</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">return</span> <span class="n">gini</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.stats.multitest">
        <code>multitest</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.multitest.compute_fdrcorrection_asymmetric_matrix">
<code class="highlight language-python"><span class="n">compute_fdrcorrection_asymmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes and FDR correction or Benjamini-Hochberg procedure
on a asymmetric matrix of p-values. Here, the correction
is performed for every value in X.</p>
<h6 id="cell2cell.stats.multitest.compute_fdrcorrection_asymmetric_matrix--parameters">Parameters</h6>
<p>X : pandas.DataFrame
    An asymmetric dataframe of P-values.</p>
<p>alpha : float, default=0.1
    Error rate of the FDR correction. Must be 0 &lt; alpha &lt; 1.</p>
<h6 id="cell2cell.stats.multitest.compute_fdrcorrection_asymmetric_matrix--returns">Returns</h6>
<p>adj_X : pandas.DataFrame
    An asymmetric dataframe with adjusted P-values of X.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/multitest.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_fdrcorrection_asymmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes and FDR correction or Benjamini-Hochberg procedure</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    on a asymmetric matrix of p-values. Here, the correction</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    is performed for every value in X.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    X : pandas.DataFrame</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        An asymmetric dataframe of P-values.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    alpha : float, default=0.1</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Error rate of the FDR correction. Must be 0 &lt; alpha &lt; 1.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    -------</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    adj_X : pandas.DataFrame</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        An asymmetric dataframe with adjusted P-values of X.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">pandas</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">a</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="n">pandas</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="n">a</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">index</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">columns</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">columns</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="c1"># Original data</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="c1"># New data</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">rej</span><span class="p">,</span> <span class="n">adj_pvals</span> <span class="o">=</span> <span class="n">fdrcorrection</span><span class="p">(</span><span class="n">pvals</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="c1"># Reorder_data</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="c1">#adj_X = adj_pvals.reshape(-1, a.shape[1])</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">adj_X</span> <span class="o">=</span> <span class="n">adj_pvals</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="c1"># Allows using tensors</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">if</span> <span class="n">pandas</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>        <span class="n">adj_X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adj_X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">return</span> <span class="n">adj_X</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.multitest.compute_fdrcorrection_symmetric_matrix">
<code class="highlight language-python"><span class="n">compute_fdrcorrection_symmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes and FDR correction or Benjamini-Hochberg procedure
on a symmetric matrix of p-values. Here, only the diagonal
and values on the upper triangle are considered to avoid
repetition with the lower triangle.</p>
<h6 id="cell2cell.stats.multitest.compute_fdrcorrection_symmetric_matrix--parameters">Parameters</h6>
<p>X : pandas.DataFrame
    A symmetric dataframe of P-values.</p>
<p>alpha : float, default=0.1
    Error rate of the FDR correction. Must be 0 &lt; alpha &lt; 1.</p>
<h6 id="cell2cell.stats.multitest.compute_fdrcorrection_symmetric_matrix--returns">Returns</h6>
<p>adj_X : pandas.DataFrame
    A symmetric dataframe with adjusted P-values of X.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/multitest.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_fdrcorrection_symmetric_matrix</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes and FDR correction or Benjamini-Hochberg procedure</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    on a symmetric matrix of p-values. Here, only the diagonal</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    and values on the upper triangle are considered to avoid</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    repetition with the lower triangle.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    X : pandas.DataFrame</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A symmetric dataframe of P-values.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    alpha : float, default=0.1</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Error rate of the FDR correction. Must be 0 &lt; alpha &lt; 1.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    adj_X : pandas.DataFrame</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A symmetric dataframe with adjusted P-values of X.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="n">pandas</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">a</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="n">pandas</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">a</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">index</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">columns</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">columns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="c1"># Original data</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="n">upper_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">triu_indices_from</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="n">upper_idx</span><span class="p">]</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># New data</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">adj_X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">rej</span><span class="p">,</span> <span class="n">adj_pvals</span> <span class="o">=</span> <span class="n">fdrcorrection</span><span class="p">(</span><span class="n">pvals</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="c1"># Reorder_data</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">adj_X</span><span class="p">[</span><span class="n">upper_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">adj_pvals</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">adj_X</span> <span class="o">=</span> <span class="n">adj_X</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">triu</span><span class="p">(</span><span class="n">adj_X</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">if</span> <span class="n">pandas</span><span class="p">:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="n">adj_X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adj_X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">return</span> <span class="n">adj_X</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.stats.permutation">
        <code>permutation</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.permutation.compute_pvalue_from_dist">
<code class="highlight language-python"><span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="p">,</span> <span class="n">dist</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;upper&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the probability of observing a value in a given distribution.</p>
<h6 id="cell2cell.stats.permutation.compute_pvalue_from_dist--parameters">Parameters</h6>
<p>obs_value : float
    An observed value used to get a p-value from a distribution.</p>
<p>dist : array-like
    A simulated oe empirical distribution of values used to compare
    the observed value and get a p-value.</p>
<p>consider_size : boolean, default=False
    Whether considering the size of the distribution for limiting
    small probabilities to be as minimal as the reciprocal of the size.</p>
<p>comparison : str, default='upper'
    Type of hypothesis testing:</p>
<div class="codehilite"><pre><span></span><code>- &#39;lower&#39; : Lower-tailed, whether the value is smaller than most
    of the values in the distribution.
- &#39;upper&#39; : Upper-tailed, whether the value is greater than most
    of the values in the distribution.
- &#39;different&#39; : Two-tailed, whether the value is different than
    most of the values in the distribution.
</code></pre></div>

<h6 id="cell2cell.stats.permutation.compute_pvalue_from_dist--returns">Returns</h6>
<p>pval : float
    P-value obtained from comparing the observed value and values in the
    distribution.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/permutation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="p">,</span> <span class="n">dist</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;upper&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes the probability of observing a value in a given distribution.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    obs_value : float</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        An observed value used to get a p-value from a distribution.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    dist : array-like</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A simulated oe empirical distribution of values used to compare</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        the observed value and get a p-value.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    consider_size : boolean, default=False</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Whether considering the size of the distribution for limiting</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        small probabilities to be as minimal as the reciprocal of the size.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    comparison : str, default=&#39;upper&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Type of hypothesis testing:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;lower&#39; : Lower-tailed, whether the value is smaller than most</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">            of the values in the distribution.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;upper&#39; : Upper-tailed, whether the value is greater than most</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">            of the values in the distribution.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;different&#39; : Two-tailed, whether the value is different than</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">            most of the values in the distribution.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    -------</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    pval : float</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        P-value obtained from comparing the observed value and values in the</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        distribution.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># Omit nan values</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="n">dist_</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">dist</span> <span class="k">if</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="c1"># All values in dist are NaNs or obs_value is NaN</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dist_</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">obs_value</span><span class="p">):</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">return</span> <span class="mf">1.0</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="c1"># No NaN values</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">if</span> <span class="n">comparison</span> <span class="o">==</span> <span class="s1">&#39;lower&#39;</span><span class="p">:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="n">pval</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">percentileofscore</span><span class="p">(</span><span class="n">dist_</span><span class="p">,</span> <span class="n">obs_value</span><span class="p">)</span> <span class="o">/</span> <span class="mf">100.0</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">elif</span> <span class="n">comparison</span> <span class="o">==</span> <span class="s1">&#39;upper&#39;</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="n">pval</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">percentileofscore</span><span class="p">(</span><span class="n">dist_</span><span class="p">,</span> <span class="n">obs_value</span><span class="p">)</span> <span class="o">/</span> <span class="mf">100.0</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="k">elif</span> <span class="n">comparison</span> <span class="o">==</span> <span class="s1">&#39;different&#39;</span><span class="p">:</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">percentile</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">percentileofscore</span><span class="p">(</span><span class="n">dist_</span><span class="p">,</span> <span class="n">obs_value</span><span class="p">)</span> <span class="o">/</span> <span class="mf">100.0</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="k">if</span> <span class="n">percentile</span> <span class="o">&lt;=</span> <span class="mf">0.5</span><span class="p">:</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>            <span class="n">pval</span> <span class="o">=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">percentile</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>            <span class="n">pval</span> <span class="o">=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="n">percentile</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s1">&#39;Comparison </span><span class="si">{}</span><span class="s1"> is not implemented&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">comparison</span><span class="p">))</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">if</span> <span class="p">(</span><span class="n">consider_size</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pval</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">):</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">pval</span> <span class="o">=</span> <span class="mf">1.</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dist_</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">elif</span> <span class="n">pval</span> <span class="o">&lt;</span> <span class="mf">0.</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">pval</span> <span class="o">=</span> <span class="mf">1.</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dist_</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="k">return</span> <span class="n">pval</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.permutation.pvalue_from_dist">
<code class="highlight language-python"><span class="n">pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="p">,</span> <span class="n">dist</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;upper&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes a p-value for an observed value given a simulated or
empirical distribution. It plots the distribution and prints
the p-value.</p>
<h6 id="cell2cell.stats.permutation.pvalue_from_dist--parameters">Parameters</h6>
<p>obs_value : float
    An observed value used to get a p-value from a distribution.</p>
<p>dist : array-like
    A simulated oe empirical distribution of values used to compare
    the observed value and get a p-value.</p>
<p>label : str, default=''
    Label used for the histogram plot. Useful for identifying it
    across multiple plots.</p>
<p>consider_size : boolean, default=False
    Whether considering the size of the distribution for limiting
    small probabilities to be as minimal as the reciprocal of the size.</p>
<p>comparison : str, default='upper'
    Type of hypothesis testing:</p>
<div class="codehilite"><pre><span></span><code>- &#39;lower&#39; : Lower-tailed, whether the value is smaller than most
    of the values in the distribution.
- &#39;upper&#39; : Upper-tailed, whether the value is greater than most
    of the values in the distribution.
- &#39;different&#39; : Two-tailed, whether the value is different than
    most of the values in the distribution.
</code></pre></div>

<h6 id="cell2cell.stats.permutation.pvalue_from_dist--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure that shows the histogram for dist.</p>
<p>pval : float
    P-value obtained from comparing the observed value and values in the
    distribution.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/permutation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="p">,</span> <span class="n">dist</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;upper&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes a p-value for an observed value given a simulated or</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    empirical distribution. It plots the distribution and prints</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    the p-value.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    obs_value : float</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        An observed value used to get a p-value from a distribution.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    dist : array-like</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        A simulated oe empirical distribution of values used to compare</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        the observed value and get a p-value.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    label : str, default=&#39;&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Label used for the histogram plot. Useful for identifying it</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        across multiple plots.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    consider_size : boolean, default=False</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Whether considering the size of the distribution for limiting</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        small probabilities to be as minimal as the reciprocal of the size.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    comparison : str, default=&#39;upper&#39;</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        Type of hypothesis testing:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;lower&#39; : Lower-tailed, whether the value is smaller than most</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">            of the values in the distribution.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;upper&#39; : Upper-tailed, whether the value is greater than most</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">            of the values in the distribution.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        - &#39;different&#39; : Two-tailed, whether the value is different than</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">            most of the values in the distribution.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    -------</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Figure that shows the histogram for dist.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    pval : float</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        P-value obtained from comparing the observed value and values in the</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        distribution.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">pval</span> <span class="o">=</span> <span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">obs_value</span><span class="o">=</span><span class="n">obs_value</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                    <span class="n">dist</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                    <span class="n">consider_size</span><span class="o">=</span><span class="n">consider_size</span><span class="p">,</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>                                    <span class="n">comparison</span><span class="o">=</span><span class="n">comparison</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                                    <span class="p">)</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;P-value is: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pval</span><span class="p">))</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="k">with</span> <span class="n">sns</span><span class="o">.</span><span class="n">axes_style</span><span class="p">(</span><span class="s2">&quot;darkgrid&quot;</span><span class="p">):</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>        <span class="k">if</span> <span class="n">label</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>            <span class="k">if</span> <span class="n">pval</span> <span class="o">&gt;</span> <span class="mf">1.</span> <span class="o">-</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">):</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="s1">&#39;p-val: &gt;</span><span class="si">{:g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{:.1g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mf">1.</span> <span class="o">-</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">))))</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>            <span class="k">elif</span> <span class="n">pval</span> <span class="o">&lt;</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">):</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="s1">&#39;p-val: &lt;</span><span class="si">{:g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{:.1g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">))))</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="s1">&#39;p-val: </span><span class="si">{0:.2E}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pval</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>            <span class="k">if</span> <span class="n">pval</span> <span class="o">&gt;</span> <span class="mf">1.</span> <span class="o">-</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">):</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="n">label</span> <span class="o">+</span> <span class="s1">&#39; - p-val: &gt;</span><span class="si">{:g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{:.1g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mf">1.</span> <span class="o">-</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">))))</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>            <span class="k">elif</span> <span class="n">pval</span> <span class="o">&lt;</span> <span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">):</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="n">label</span> <span class="o">+</span> <span class="s1">&#39; - p-val: &lt;</span><span class="si">{:g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{:.1g}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mf">1.</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="p">))))</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                <span class="n">label_</span> <span class="o">=</span> <span class="n">label</span> <span class="o">+</span> <span class="s1">&#39; - p-val: </span><span class="si">{0:.2E}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pval</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>        <span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">dist</span><span class="p">,</span> <span class="n">hist</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm_hist</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">rug</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">label_</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">obs_value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">get_lines</span><span class="p">()[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_c</span><span class="p">(),</span> <span class="n">ls</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>        <span class="n">fig</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;major&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="n">lgd</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.01</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                     <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">pval</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.permutation.random_switching_ppi_labels">
<code class="highlight language-python"><span class="n">random_switching_ppi_labels</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">permuted_column</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly permutes the labels of interacting proteins in
a list of protein-protein interactions.</p>
<h6 id="cell2cell.stats.permutation.random_switching_ppi_labels--parameters">Parameters</h6>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>genes : list, default=None
    List of genes, with names matching proteins in the PPIs, to exclusively
    consider in the analysis.</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a
    dataframe of protein-protein interactions. If the list is for
    ligand-receptor pairs, the first column is for the ligands and the second
    for the receptors.</p>
<p>permuted_column : str, default='both'
    Column among the interacting_columns to permute.
    Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;first&#39; : To permute labels considering only proteins in the first
    column.
- &#39;second&#39; : To permute labels considering only proteins in the second
    column.
- &#39; both&#39; : To permute labels considering all the proteins in the list.
</code></pre></div>

<h6 id="cell2cell.stats.permutation.random_switching_ppi_labels--returns">Returns</h6>
<p>ppi_data_ : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) with randomly
    permuted labels of proteins.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/permutation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">random_switching_ppi_labels</span><span class="p">(</span><span class="n">ppi_data</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">permuted_column</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly permutes the labels of interacting proteins in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    a list of protein-protein interactions.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    genes : list, default=None</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of genes, with names matching proteins in the PPIs, to exclusively</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        consider in the analysis.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Contains the names of the columns where to find the partners in a</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        dataframe of protein-protein interactions. If the list is for</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        ligand-receptor pairs, the first column is for the ligands and the second</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        for the receptors.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    permuted_column : str, default=&#39;both&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Column among the interacting_columns to permute.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;first&#39; : To permute labels considering only proteins in the first</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">            column.</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        - &#39;second&#39; : To permute labels considering only proteins in the second</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">            column.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39; both&#39; : To permute labels considering all the proteins in the list.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    -------</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    ppi_data_ : pandas.DataFrame</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) with randomly</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        permuted labels of proteins.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">ppi_data_</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">prot_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="n">prot_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">if</span> <span class="n">permuted_column</span> <span class="o">==</span> <span class="s1">&#39;both&#39;</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">ppi_data_</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="n">mapper</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">shuffle</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)))</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mapper</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mapper</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="k">elif</span> <span class="n">permuted_column</span> <span class="o">==</span> <span class="s1">&#39;first&#39;</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>        <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="n">mapper</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">shuffle</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)))</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_a</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mapper</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>    <span class="k">elif</span> <span class="n">permuted_column</span> <span class="o">==</span> <span class="s1">&#39;second&#39;</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="n">mapper</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">shuffle</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)))</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span> <span class="o">=</span> <span class="n">ppi_data_</span><span class="p">[</span><span class="n">prot_b</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mapper</span><span class="p">[</span><span class="n">x</span><span class="p">])</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="k">else</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not valid option&#39;</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="k">return</span> <span class="n">ppi_data_</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.stats.permutation.run_label_permutation">
<code class="highlight language-python"><span class="n">run_label_permutation</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">analysis_setup</span><span class="p">,</span> <span class="n">cutoff_setup</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">permuted_label</span><span class="o">=</span><span class="s1">&#39;gene_labels&#39;</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Permutes a label before computing cell-cell interaction scores.</p>
<h6 id="cell2cell.stats.permutation.run_label_permutation--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression data for a bulk RNA-seq experiment or a single-cell
    experiment after aggregation into cell types. Columns are
    cell-types/tissues/samples and rows are genes.</p>
<p>ppi_data : pandas.DataFrame
    List of protein-protein interactions (or ligand-receptor pairs) used
    for inferring the cell-cell interactions and communication.</p>
<p>genes : list
    List of genes in rnaseq_data to exclusively consider in the analysis.</p>
<p>analysis_setup : dict
    Contains main setup for running the cell-cell interactions and communication
    analyses. Three main setups are needed (passed as keys):</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">communication_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">communication</span><span class="w"> </span><span class="n">score</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">detect</span><span class="w"></span>
<span class="w">    </span><span class="n">active</span><span class="w"> </span><span class="n">ligand</span><span class="o">-</span><span class="n">receptor</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">pair</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">cell</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_thresholding</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_product</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">expression_mean</span><span class="p">&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_score</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">scoring</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">communication</span><span class="w"></span>
<span class="w">    </span><span class="n">scores</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">It</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">bray_curtis</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">jaccard</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">count</span><span class="p">&#39;</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">cci_type</span><span class="p">&#39;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">type</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">interaction</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="n">two</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"> </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"></span>
<span class="w">    </span><span class="n">undirected</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">both</span><span class="w"> </span><span class="n">cells</span><span class="p">.</span><span class="w"></span>
<span class="w">    </span><span class="n">If</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="n">directed</span><span class="p">,</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptors</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">other</span><span class="w"></span>
<span class="w">     </span><span class="n">are</span><span class="w"> </span><span class="n">considered</span><span class="w"> </span><span class="n">separately</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="n">respect</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">ligands</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">second</span><span class="w"></span>
<span class="w">     </span><span class="n">cell</span><span class="w"> </span><span class="k">and</span><span class="w"> </span><span class="n">receptor</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">first</span><span class="w"> </span><span class="n">one</span><span class="p">.</span><span class="w"></span>
<span class="w">     </span><span class="n">So</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">can</span><span class="w"> </span><span class="nl">be:</span><span class="w"></span>

<span class="w">     </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">undirected</span><span class="p">&#39;</span><span class="w"></span>
<span class="w">     </span><span class="o">-</span><span class="w"> </span><span class="p">&#39;</span><span class="n">directed</span><span class="w"></span>
</code></pre></div>

<p>cutoff_setup : dict
    Contains two keys: 'type' and 'parameter'. The first key represent the
    way to use a cutoff or threshold, while parameter is the value used
    to binarize the expression values.
    The key 'type' can be:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;local_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="p">,</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"></span>
<span class="w">    </span><span class="n">gene</span><span class="w"> </span><span class="n">independently</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"> </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"> </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;global_percentile&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">computes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">given</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">all</span><span class="w"></span>
<span class="w">    </span><span class="n">genes</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">samples</span><span class="w"> </span><span class="n">simultaneously</span><span class="o">.</span><span class="w"> </span><span class="n">In</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="p">,</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="w"></span>
<span class="w">    </span><span class="n">corresponds</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">percentile</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">compute</span><span class="p">,</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nb nb-Type">float</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">between</span><span class="w"></span>
<span class="w">    </span><span class="mi">0</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="mf">1.</span><span class="w"> </span><span class="n">All</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">cutoff</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;file&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="nb">load</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"> </span><span class="n">table</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">Parameter</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">this</span><span class="w"> </span><span class="n">case</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">that</span><span class="w"> </span><span class="n">file</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">contain</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">same</span><span class="w"></span>
<span class="w">    </span><span class="n">samples</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">columns</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;multi_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">This</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">use</span><span class="w"> </span><span class="n">specific</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="k">for</span><span class="w"></span>
<span class="w">    </span><span class="n">each</span><span class="w"> </span><span class="n">sample</span><span class="o">.</span><span class="w"> </span><span class="n">The</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="n">here</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">same</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">ones</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">    </span><span class="n">rnaseq_data</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;single_col_matrix&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">dataframe</span><span class="w"> </span><span class="n">must</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">provided</span><span class="p">,</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">cutoff</span><span class="w"></span>
<span class="w">    </span><span class="k">for</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">only</span><span class="w"> </span><span class="n">one</span><span class="w"> </span><span class="n">column</span><span class="o">.</span><span class="w"> </span><span class="n">These</span><span class="w"> </span><span class="n">cutoffs</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">applied</span><span class="w"> </span><span class="n">to</span><span class="w"></span>
<span class="w">    </span><span class="n">all</span><span class="w"> </span><span class="n">samples</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constant_value&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">binarizes</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">expression</span><span class="o">.</span><span class="w"> </span><span class="n">Evaluates</span><span class="w"> </span><span class="n">whether</span><span class="w"></span>
<span class="w">    </span><span class="n">expression</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">greater</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="n">input</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">parameter</span><span class="o">.</span><span class="w"></span>
</code></pre></div>

<p>permutations : int, default=100
        Number of permutations where in each of them a random
        shuffle of labels is performed, followed of computing
        CCI scores to create a null distribution.</p>
<p>permuted_label : str, default='gene_labels'
    Label to be permuted.
    Types are:</p>
<div class="codehilite"><pre><span></span><code>    - &#39;genes&#39; : Permutes cell-labels in a gene-specific way.
    - &#39;gene_labels&#39; : Permutes the labels of genes in the RNA-seq dataset.
    - &#39;cell_labels&#39; : Permutes the labels of cell-types/tissues/samples
        in the RNA-seq dataset.
</code></pre></div>

<p>excluded_cells : list, default=None
    List of cells to exclude from the analysis.</p>
<p>consider_size : boolean, default=True
    Whether considering the size of the distribution for limiting
    small probabilities to be as minimal as the reciprocal of the size.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.stats.permutation.run_label_permutation--returns">Returns</h6>
<p>cci_pvals : pandas.DataFrame
    Matrix where rows and columns are cell-types/tissues/samples and
    each value is a P-value for the corresponding CCI score.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/stats/permutation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">run_label_permutation</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">analysis_setup</span><span class="p">,</span> <span class="n">cutoff_setup</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                          <span class="n">permuted_label</span><span class="o">=</span><span class="s1">&#39;gene_labels&#39;</span><span class="p">,</span> <span class="n">excluded_cells</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">consider_size</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Permutes a label before computing cell-cell interaction scores.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression data for a bulk RNA-seq experiment or a single-cell</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        experiment after aggregation into cell types. Columns are</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        cell-types/tissues/samples and rows are genes.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        List of protein-protein interactions (or ligand-receptor pairs) used</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        for inferring the cell-cell interactions and communication.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    genes : list</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List of genes in rnaseq_data to exclusively consider in the analysis.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    analysis_setup : dict</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Contains main setup for running the cell-cell interactions and communication</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        analyses. Three main setups are needed (passed as keys):</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;communication_score&#39; : is the type of communication score used to detect</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">            active ligand-receptor pairs between each pair of cell.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">            - &#39;expression_thresholding&#39;</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">            - &#39;expression_product&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">            - &#39;expression_mean&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;cci_score&#39; : is the scoring function to aggregate the communication</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">            scores.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">            It can be:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">            - &#39;bray_curtis&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">            - &#39;jaccard&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">            - &#39;count&#39;</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        - &#39;cci_type&#39; : is the type of interaction between two cells. If it is</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">            undirected, all ligands and receptors are considered from both cells.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">            If it is directed, ligands from one cell and receptors from the other</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">             are considered separately with respect to ligands from the second</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">             cell and receptor from the first one.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">             So, it can be:</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">             - &#39;undirected&#39;</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">             - &#39;directed</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">    cutoff_setup : dict</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        Contains two keys: &#39;type&#39; and &#39;parameter&#39;. The first key represent the</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        way to use a cutoff or threshold, while parameter is the value used</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        to binarize the expression values.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        The key &#39;type&#39; can be:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        - &#39;local_percentile&#39; : computes the value of a given percentile, for each</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">            gene independently. In this case, the parameter corresponds to the</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">            percentile to compute, as a float value between 0 and 1.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        - &#39;global_percentile&#39; : computes the value of a given percentile from all</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">            genes and samples simultaneously. In this case, the parameter</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">            corresponds to the percentile to compute, as a float value between</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">            0 and 1. All genes have the same cutoff.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        - &#39;file&#39; : load a cutoff table from a file. Parameter in this case is the</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">            path of that file. It must contain the same genes as index and same</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">            samples as columns.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        - &#39;multi_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">            for each gene in each sample. This allows to use specific cutoffs for</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">            each sample. The columns here must be the same as the ones in the</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">            rnaseq_data.</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        - &#39;single_col_matrix&#39; : a dataframe must be provided, containing a cutoff</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">            for each gene in only one column. These cutoffs will be applied to</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">            all samples.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        - &#39;constant_value&#39; : binarizes the expression. Evaluates whether</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">            expression is greater than the value input in the parameter.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    permutations : int, default=100</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">            Number of permutations where in each of them a random</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">            shuffle of labels is performed, followed of computing</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">            CCI scores to create a null distribution.</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">    permuted_label : str, default=&#39;gene_labels&#39;</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        Label to be permuted.</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">        Types are:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">            - &#39;genes&#39; : Permutes cell-labels in a gene-specific way.</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">            - &#39;gene_labels&#39; : Permutes the labels of genes in the RNA-seq dataset.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">            - &#39;cell_labels&#39; : Permutes the labels of cell-types/tissues/samples</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">                in the RNA-seq dataset.</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">    excluded_cells : list, default=None</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        List of cells to exclude from the analysis.</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">    consider_size : boolean, default=True</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">        Whether considering the size of the distribution for limiting</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        small probabilities to be as minimal as the reciprocal of the size.</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">    -------</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">    cci_pvals : pandas.DataFrame</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">        Matrix where rows and columns are cell-types/tissues/samples and</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">        each value is a P-value for the corresponding CCI score.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>    <span class="c1"># Placeholders</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>    <span class="n">scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">diag</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="c1"># Info to use</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="k">if</span> <span class="n">excluded_cells</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>        <span class="n">included_cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">excluded_cells</span><span class="p">)))</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>        <span class="n">included_cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="n">rnaseq_data_</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">,</span> <span class="n">included_cells</span><span class="p">]</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>    <span class="c1"># Permutations</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">permutations</span><span class="p">)):</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="k">if</span> <span class="n">permuted_label</span> <span class="o">==</span> <span class="s1">&#39;genes&#39;</span><span class="p">:</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>            <span class="c1"># Shuffle genes</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>            <span class="n">shuffled_rnaseq_data</span> <span class="o">=</span> <span class="n">shuffle_rows_in_df</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">rnaseq_data_</span><span class="p">,</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                                                      <span class="n">rows</span><span class="o">=</span><span class="n">genes</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="k">elif</span> <span class="n">permuted_label</span> <span class="o">==</span> <span class="s1">&#39;cell_labels&#39;</span><span class="p">:</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>            <span class="c1"># Shuffle cell labels</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>            <span class="n">shuffled_rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data_</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>            <span class="n">shuffled_cells</span> <span class="o">=</span> <span class="n">shuffle</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">shuffled_rnaseq_data</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>            <span class="n">shuffled_rnaseq_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">shuffled_cells</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="k">elif</span> <span class="n">permuted_label</span> <span class="o">==</span> <span class="s1">&#39;gene_labels&#39;</span><span class="p">:</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>            <span class="c1"># Shuffle gene labels</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>            <span class="n">shuffled_rnaseq_data</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>            <span class="n">shuffled_genes</span> <span class="o">=</span> <span class="n">shuffle</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">shuffled_rnaseq_data</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>            <span class="n">shuffled_rnaseq_data</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">shuffled_genes</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Not a valid shuffle_type.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="n">interaction_space</span> <span class="o">=</span> <span class="n">ispace</span><span class="o">.</span><span class="n">InteractionSpace</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">shuffled_rnaseq_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">,</span> <span class="n">included_cells</span><span class="p">],</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>                                                    <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                                                    <span class="n">gene_cutoffs</span><span class="o">=</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>                                                    <span class="n">communication_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;communication_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                                                    <span class="n">cci_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                                                    <span class="n">cci_type</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">],</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                                                    <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>        <span class="c1"># Keep scores</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>        <span class="n">cci</span> <span class="o">=</span> <span class="n">interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">included_cells</span><span class="p">,</span> <span class="n">included_cells</span><span class="p">]</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>        <span class="n">cci_diag</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">cci</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">cci</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>        <span class="n">iter_scores</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">cci</span><span class="p">)</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>        <span class="n">iter_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">iter_scores</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">iter_scores</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="n">iter_diag</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">cci_diag</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cci_diag</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">T</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>        <span class="k">if</span> <span class="n">scores</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">0</span><span class="p">,):</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>            <span class="n">scores</span> <span class="o">=</span> <span class="n">iter_scores</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>            <span class="n">scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">scores</span><span class="p">,</span> <span class="n">iter_scores</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="k">if</span> <span class="n">diag</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">0</span><span class="p">,):</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>            <span class="n">diag</span> <span class="o">=</span> <span class="n">iter_diag</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>            <span class="n">diag</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">diag</span><span class="p">,</span> <span class="n">iter_diag</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>    <span class="c1"># Base CCI scores</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="n">base_interaction_space</span> <span class="o">=</span> <span class="n">ispace</span><span class="o">.</span><span class="n">InteractionSpace</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">rnaseq_data_</span><span class="p">[</span><span class="n">included_cells</span><span class="p">],</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>                                                     <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>                                                     <span class="n">gene_cutoffs</span><span class="o">=</span><span class="n">cutoff_setup</span><span class="p">,</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>                                                     <span class="n">communication_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;communication_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>                                                     <span class="n">cci_score</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_score&#39;</span><span class="p">],</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>                                                     <span class="n">cci_type</span><span class="o">=</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">],</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>                                                     <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>    <span class="c1"># Keep scores</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>    <span class="n">base_cci</span> <span class="o">=</span> <span class="n">base_interaction_space</span><span class="o">.</span><span class="n">interaction_elements</span><span class="p">[</span><span class="s1">&#39;cci_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">included_cells</span><span class="p">,</span> <span class="n">included_cells</span><span class="p">]</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>    <span class="n">base_cci_diag</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">base_cci</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>    <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">base_cci</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>    <span class="n">base_scores</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">base_cci</span><span class="p">)</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>    <span class="c1"># P-values</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>    <span class="n">pvals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>    <span class="n">diag_pvals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">diag</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">scores</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>        <span class="n">pvals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">base_scores</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>                                            <span class="n">scores</span><span class="p">[:,</span> <span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>                                            <span class="n">consider_size</span><span class="o">=</span><span class="n">consider_size</span><span class="p">,</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>                                            <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;different&#39;</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>                                            <span class="p">)</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">diag</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>        <span class="n">diag_pvals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">compute_pvalue_from_dist</span><span class="p">(</span><span class="n">base_cci_diag</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>                                                 <span class="n">diag</span><span class="p">[:,</span> <span class="n">i</span><span class="p">],</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>                                                 <span class="n">consider_size</span><span class="o">=</span><span class="n">consider_size</span><span class="p">,</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                                                 <span class="n">comparison</span><span class="o">=</span><span class="s1">&#39;different&#39;</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>                                                 <span class="p">)</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>    <span class="c1"># DataFrame</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>    <span class="n">cci_pvals</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">spatial</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">pvals</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">pvals</span><span class="p">),)))</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">diag_pvals</span><span class="o">.</span><span class="n">flatten</span><span class="p">()):</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>        <span class="n">cci_pvals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a>    <span class="n">cci_pvals</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cci_pvals</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">base_cci</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">base_cci</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a>    <span class="k">return</span> <span class="n">cci_pvals</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.tensor">
        <code>tensor</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.external_scores">
        <code>external_scores</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.external_scores.dataframes_to_tensor">
<code class="highlight language-python"><span class="n">dataframes_to_tensor</span><span class="p">(</span><span class="n">context_df_dict</span><span class="p">,</span> <span class="n">sender_col</span><span class="p">,</span> <span class="n">receiver_col</span><span class="p">,</span> <span class="n">ligand_col</span><span class="p">,</span> <span class="n">receptor_col</span><span class="p">,</span> <span class="n">score_col</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">lr_fill</span><span class="o">=</span><span class="n">nan</span><span class="p">,</span> <span class="n">cell_fill</span><span class="o">=</span><span class="n">nan</span><span class="p">,</span> <span class="n">lr_sep</span><span class="o">=</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="n">dup_aggregation</span><span class="o">=</span><span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="n">context_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sort_elements</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates an InteractionTensor from a dictionary
containing dataframes for all contexts.</p>
<h6 id="cell2cell.tensor.external_scores.dataframes_to_tensor--parameters">Parameters</h6>
<p>context_df_dict : dict
    Dictionary containing a dataframe for each context. The dataframe
    must contain columns containing sender cells, receiver cells,
    ligands, receptors, and communication scores, separately.
    Keys are context names and values are dataframes.</p>
<p>sender_col : str
    Name of the column containing the sender cells in all context
    dataframes.</p>
<p>receiver_col : str
    Name of the column containing the receiver cells in all context
    dataframes.</p>
<p>ligand_col : str
    Name of the column containing the ligands in all context
    dataframes.</p>
<p>receptor_col : str
    Name of the column containing the receptors in all context
    dataframes.</p>
<p>score_col : str
    Name of the column containing the communication scores in all context
    dataframes.</p>
<p>how : str, default='inner'
    Approach to consider cell types and genes present across multiple contexts.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">inner</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">LR</span> <span class="nv">pairs</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
            <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">LR</span> <span class="nv">pairs</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span>
            <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_lrs</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">LR</span> <span class="nv">pairs</span> <span class="nv">that</span> <span class="nv">are</span>
                <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_cells</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">LR</span> <span class="nv">pairs</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
</code></pre></div>

<p>outer_fraction : float, default=0.0
    Threshold to filter the elements when <code>how</code> includes any outer option.
    Elements with a fraction abundance across contexts (in <code>context_df_dict</code>)
    at least this threshold will be included. When this value is 0, considers
    all elements across the samples. When this value is 1, it acts as using
    <code>how='inner'</code>.</p>
<p>lr_fill : float, default=numpy.nan
    Value to fill communication scores when a ligand-receptor pair is not
    present across all contexts.</p>
<p>cell_fill : float, default=numpy.nan
    Value to fill communication scores when a cell is not
    present across all ligand-receptor pairs or all contexts.</p>
<p>lr_sep : str, default='^'
    Separation character to join ligands and receptors into a LR pair name.</p>
<p>dup_aggregation : str, default='max'
    Approach to aggregate communication score if there are multiple instances
    of an LR pair for a specific sender-receiver pair in one of the dataframes.</p>
<div class="codehilite"><pre><span></span><code>- &#39;max&#39; : Maximum of the multiple instances
- &#39;min&#39; : Minimum of the multiple instances
- &#39;mean&#39; : Average of the multiple instances
- &#39;median&#39; : Median of the multiple instances
</code></pre></div>

<p>context_order : list, default=None
    List used to sort the contexts when building the tensor. Elements must
    be all elements in context_df_dict.keys().</p>
<p>order_labels : list, default=None
    List containing the labels for each order or dimension of the tensor. For
    example: ['Contexts', 'Ligand-Receptor Pairs', 'Sender Cells', 'Receiver Cells']</p>
<p>sort_elements : boolean, default=True
    Whether alphabetically sorting elements in the InteractionTensor.
    The Context Dimension is not sorted if a 'context_order' list is provided.</p>
<p None="None" _cpu_="'cpu'," _cuda:0_="'cuda:0',">device : str, default=None
        Device to use when backend is pytorch. Options are:</p>
<h6 id="cell2cell.tensor.external_scores.dataframes_to_tensor--returns">Returns</h6>
<p>interaction_tensor : cell2cell.tensor.PreBuiltTensor
    A communication tensor generated for the Tensor-cell2cell pipeline.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/external_scores.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">dataframes_to_tensor</span><span class="p">(</span><span class="n">context_df_dict</span><span class="p">,</span> <span class="n">sender_col</span><span class="p">,</span> <span class="n">receiver_col</span><span class="p">,</span> <span class="n">ligand_col</span><span class="p">,</span> <span class="n">receptor_col</span><span class="p">,</span> <span class="n">score_col</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                         <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">lr_fill</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">cell_fill</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">lr_sep</span><span class="o">=</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="n">dup_aggregation</span><span class="o">=</span><span class="s1">&#39;max&#39;</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                         <span class="n">context_order</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sort_elements</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Generates an InteractionTensor from a dictionary</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    containing dataframes for all contexts.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    context_df_dict : dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Dictionary containing a dataframe for each context. The dataframe</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        must contain columns containing sender cells, receiver cells,</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        ligands, receptors, and communication scores, separately.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Keys are context names and values are dataframes.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    sender_col : str</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Name of the column containing the sender cells in all context</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        dataframes.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    receiver_col : str</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Name of the column containing the receiver cells in all context</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        dataframes.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    ligand_col : str</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Name of the column containing the ligands in all context</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        dataframes.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    receptor_col : str</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Name of the column containing the receptors in all context</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        dataframes.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    score_col : str</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Name of the column containing the communication scores in all context</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        dataframes.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    how : str, default=&#39;inner&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Approach to consider cell types and genes present across multiple contexts.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        - &#39;inner&#39; : Considers only cell types and LR pairs that are present in all</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">                    contexts (intersection).</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        - &#39;outer&#39; : Considers all cell types and LR pairs that are present</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">                    across contexts (union).</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        - &#39;outer_lrs&#39; : Considers only cell types that are present in all</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">                        contexts (intersection), while all LR pairs that are</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">                        present across contexts (union).</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        - &#39;outer_cells&#39; : Considers only LR pairs that are present in all</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">                          contexts (intersection), while all cell types that are</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">    outer_fraction : float, default=0.0</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        Threshold to filter the elements when `how` includes any outer option.</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        Elements with a fraction abundance across contexts (in `context_df_dict`)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        at least this threshold will be included. When this value is 0, considers</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        all elements across the samples. When this value is 1, it acts as using</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        `how=&#39;inner&#39;`.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    lr_fill : float, default=numpy.nan</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Value to fill communication scores when a ligand-receptor pair is not</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        present across all contexts.</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">    cell_fill : float, default=numpy.nan</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        Value to fill communication scores when a cell is not</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        present across all ligand-receptor pairs or all contexts.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    lr_sep : str, default=&#39;^&#39;</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        Separation character to join ligands and receptors into a LR pair name.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    dup_aggregation : str, default=&#39;max&#39;</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Approach to aggregate communication score if there are multiple instances</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        of an LR pair for a specific sender-receiver pair in one of the dataframes.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        - &#39;max&#39; : Maximum of the multiple instances</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        - &#39;min&#39; : Minimum of the multiple instances</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">        - &#39;mean&#39; : Average of the multiple instances</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        - &#39;median&#39; : Median of the multiple instances</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">    context_order : list, default=None</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        List used to sort the contexts when building the tensor. Elements must</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        be all elements in context_df_dict.keys().</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    order_labels : list, default=None</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        List containing the labels for each order or dimension of the tensor. For</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        example: [&#39;Contexts&#39;, &#39;Ligand-Receptor Pairs&#39;, &#39;Sender Cells&#39;, &#39;Receiver Cells&#39;]</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">    sort_elements : boolean, default=True</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">        Whether alphabetically sorting elements in the InteractionTensor.</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">        The Context Dimension is not sorted if a &#39;context_order&#39; list is provided.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">            Device to use when backend is pytorch. Options are:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">             {&#39;cpu&#39;, &#39;cuda:0&#39;, None}</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">    -------</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">    interaction_tensor : cell2cell.tensor.PreBuiltTensor</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        A communication tensor generated for the Tensor-cell2cell pipeline.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>    <span class="c1"># Make sure that all contexts contain needed info</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>    <span class="k">if</span> <span class="n">context_order</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>        <span class="n">sort_context</span> <span class="o">=</span> <span class="n">sort_elements</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>        <span class="n">context_order</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">context_df_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>        <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">c</span> <span class="ow">in</span> <span class="n">context_df_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">context_order</span><span class="p">]),</span> <span class="s2">&quot;The list &#39;context_order&#39; must contain all context names contained in the keys of &#39;context_dict&#39;&quot;</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">context_order</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">context_df_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s2">&quot;Each context name must be contained only once in the list &#39;context_order&#39;&quot;</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>        <span class="n">sort_context</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">sender_col</span><span class="p">,</span> <span class="n">receiver_col</span><span class="p">,</span> <span class="n">ligand_col</span><span class="p">,</span> <span class="n">receptor_col</span><span class="p">,</span> <span class="n">score_col</span><span class="p">]</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">c</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cols</span> <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">context_df_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()]),</span> <span class="s2">&quot;All input columns must be contained in all dataframes included in &#39;context_dict&#39;&quot;</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>    <span class="c1"># Copy context dict to make modifications</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="n">cont_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span> <span class="p">:</span> <span class="n">v</span><span class="o">.</span><span class="n">copy</span><span class="p">()[</span><span class="n">cols</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">context_df_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="c1"># Labels for each dimension</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="k">if</span> <span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>        <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts&#39;</span><span class="p">,</span> <span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="c1"># Find all existing LR pairs, sender and receiver cells across contexts</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="n">lr_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>    <span class="n">sender_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>    <span class="n">receiver_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">cont_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;LRs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">ligand_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">lr_sep</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="n">receptor_col</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>        <span class="c1"># This is to consider only LR pairs in each context that are present in all cell pairs. Disabled for now.</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="c1"># if how in [&#39;inner&#39;, &#39;outer_cells&#39;]:</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>        <span class="c1">#     ccc_df = df.pivot(index=&#39;LRs&#39;, columns=[sender_col, receiver_col], values=score_col)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="c1">#     ccc_df = ccc_df.dropna(how=&#39;any&#39;)</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>        <span class="c1">#     lr_dict[k].update(list(ccc_df.index))</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="c1"># else:</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="n">lr_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;LRs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>        <span class="n">sender_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">sender_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="n">receiver_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">receiver_col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>    <span class="c1"># Subset LR pairs, sender and receiver cells given parameter &#39;how&#39;</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>    <span class="n">df_lrs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">lr_dict</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">context_order</span><span class="p">]</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>    <span class="n">df_senders</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">sender_dict</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">context_order</span><span class="p">]</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>    <span class="n">df_receivers</span>  <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">receiver_dict</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">context_order</span><span class="p">]</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;inner&#39;</span><span class="p">:</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="n">lr_pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_lrs</span><span class="p">)))</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>        <span class="n">sender_cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_senders</span><span class="p">)))</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="n">receiver_cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_receivers</span><span class="p">)))</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer&#39;</span><span class="p">:</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>        <span class="n">lr_pairs</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_lrs</span><span class="p">),</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>                                              <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="n">sender_cells</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_senders</span><span class="p">),</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                                              <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="n">receiver_cells</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_receivers</span><span class="p">),</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                                              <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer_lrs&#39;</span><span class="p">:</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>        <span class="n">lr_pairs</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_lrs</span><span class="p">),</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                                              <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="n">sender_cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_senders</span><span class="p">)))</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="n">receiver_cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_receivers</span><span class="p">)))</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer_cells&#39;</span><span class="p">:</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>        <span class="n">lr_pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_lrs</span><span class="p">)))</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>        <span class="n">sender_cells</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_senders</span><span class="p">),</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>                                                  <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="n">receiver_cells</span> <span class="o">=</span> <span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_receivers</span><span class="p">),</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>                                                    <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">)</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Not a valid input for parameter &#39;how&#39;&quot;</span><span class="p">)</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>    <span class="k">if</span> <span class="n">sort_elements</span><span class="p">:</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="k">if</span> <span class="n">sort_context</span><span class="p">:</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>            <span class="n">context_order</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">context_order</span><span class="p">)</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="n">lr_pairs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">lr_pairs</span><span class="p">)</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">sender_cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">sender_cells</span><span class="p">)</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>        <span class="n">receiver_cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">receiver_cells</span><span class="p">)</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="c1"># Build temporal tensor to pass to PreBuiltTensor</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>    <span class="n">tmp_tensor</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">context_order</span><span class="p">):</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>        <span class="n">v</span> <span class="o">=</span> <span class="n">cont_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>        <span class="c1"># 3D tensor for the context</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>        <span class="n">tmp_3d_tensor</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>        <span class="k">for</span> <span class="n">lr</span> <span class="ow">in</span> <span class="n">lr_pairs</span><span class="p">:</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>            <span class="n">df</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;LRs&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">lr</span><span class="p">]</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>            <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># TODO: Check behavior when df is empty</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>                <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">lr_fill</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">sender_cells</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">receiver_cells</span><span class="p">)</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>                <span class="k">if</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>                    <span class="k">assert</span> <span class="n">dup_aggregation</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;median&#39;</span><span class="p">],</span> <span class="s2">&quot;Please use a valid option for `dup_aggregation`.&quot;</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>                    <span class="n">df</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">cols</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]),</span> <span class="n">dup_aggregation</span><span class="p">)()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">sender_col</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">receiver_col</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">score_col</span><span class="p">)</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">sender_cells</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">cell_fill</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">receiver_cells</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">cell_fill</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>            <span class="n">tmp_3d_tensor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>        <span class="n">tmp_tensor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_3d_tensor</span><span class="p">)</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>    <span class="c1"># Create InteractionTensor using PreBuiltTensor</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>    <span class="n">tensor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">tmp_tensor</span><span class="p">)</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>    <span class="k">if</span> <span class="n">how</span> <span class="o">!=</span> <span class="s1">&#39;inner&#39;</span><span class="p">:</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">tensor</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>        <span class="n">loc_nans</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">tensor</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>        <span class="n">loc_nans</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>    <span class="n">interaction_tensor</span> <span class="o">=</span> <span class="n">PreBuiltTensor</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                                        <span class="n">order_names</span><span class="o">=</span><span class="p">[</span><span class="n">context_order</span><span class="p">,</span> <span class="n">lr_pairs</span><span class="p">,</span> <span class="n">sender_cells</span><span class="p">,</span> <span class="n">receiver_cells</span><span class="p">],</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>                                        <span class="n">order_labels</span><span class="o">=</span><span class="n">order_labels</span><span class="p">,</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>                                        <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>                                        <span class="n">loc_nans</span><span class="o">=</span><span class="n">loc_nans</span><span class="p">,</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>                                        <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>    <span class="k">return</span> <span class="n">interaction_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.factor_manipulation">
        <code>factor_manipulation</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.factor_manipulation.normalize_factors">
<code class="highlight language-python"><span class="n">normalize_factors</span><span class="p">(</span><span class="n">factors</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>L2-normalizes the factors considering all tensor dimensions
from a tensor decomposition result</p>
<h6 id="cell2cell.tensor.factor_manipulation.normalize_factors--parameters">Parameters</h6>
<p>factors : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor. This is the result from a tensor decomposition,
    it can be found as the attribute <code>factors</code> in any tensor class derived from the
    class BaseTensor (e.g. BaseTensor.factors).</p>
<h6 id="cell2cell.tensor.factor_manipulation.normalize_factors--returns">Returns</h6>
<p>norm_factors : dict
    The normalized factors.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/factor_manipulation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">normalize_factors</span><span class="p">(</span><span class="n">factors</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    L2-normalizes the factors considering all tensor dimensions</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    from a tensor decomposition result</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    factors : dict</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        dimension/order of the tensor. This is the result from a tensor decomposition,</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        it can be found as the attribute `factors` in any tensor class derived from the</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        class BaseTensor (e.g. BaseTensor.factors).</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    -------</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    norm_factors : dict</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The normalized factors.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">norm_factors</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">factors</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>        <span class="n">norm_factors</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="k">return</span> <span class="n">norm_factors</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.factor_manipulation.shuffle_factors">
<code class="highlight language-python"><span class="n">shuffle_factors</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Randomly shuffles the values of the factors in the tensor decomposition.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/factor_manipulation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">shuffle_factors</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Randomly shuffles the values of the factors in the tensor decomposition.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="k">raise</span> <span class="ne">NotImplementedError</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.factorization">
        <code>factorization</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.factorization.normalized_error">
<code class="highlight language-python"><span class="n">normalized_error</span><span class="p">(</span><span class="n">reference_tensor</span><span class="p">,</span> <span class="n">reconstructed_tensor</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes a normalized error between two tensors</p>
<h6 id="cell2cell.tensor.factorization.normalized_error--parameters">Parameters</h6>
<p>reference_tensor : ndarray list
    A tensor that could be a list of lists, a multidimensional numpy array or
    a tensorly.tensor. This tensor is the input of a tensor decomposition and
    used as reference in the normalized error for a new tensor reconstructed
    from the factors of the tensor decomposition.</p>
<p>reconstructed_tensor : ndarray list
    A tensor that could be a list of lists, a multidimensional numpy array or
    a tensorly.tensor. This tensor is an approximation of the reference_tensor
    by using the resulting factors of a tensor decomposition to compute it.</p>
<h6 id="cell2cell.tensor.factorization.normalized_error--returns">Returns</h6>
<p>norm_error : float
    The normalized error between a reference tensor and a reconstructed tensor.
    The error is normalized by dividing by the Frobinius norm of the reference
    tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/factorization.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">normalized_error</span><span class="p">(</span><span class="n">reference_tensor</span><span class="p">,</span> <span class="n">reconstructed_tensor</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes a normalized error between two tensors</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    reference_tensor : ndarray list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        A tensor that could be a list of lists, a multidimensional numpy array or</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        a tensorly.tensor. This tensor is the input of a tensor decomposition and</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        used as reference in the normalized error for a new tensor reconstructed</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        from the factors of the tensor decomposition.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    reconstructed_tensor : ndarray list</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        A tensor that could be a list of lists, a multidimensional numpy array or</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        a tensorly.tensor. This tensor is an approximation of the reference_tensor</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        by using the resulting factors of a tensor decomposition to compute it.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    norm_error : float</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        The normalized error between a reference tensor and a reconstructed tensor.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        The error is normalized by dividing by the Frobinius norm of the reference</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        tensor.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="n">norm_error</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">reference_tensor</span> <span class="o">-</span> <span class="n">reconstructed_tensor</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">reference_tensor</span><span class="p">)</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">return</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">norm_error</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.metrics">
        <code>metrics</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.metrics.correlation_index">
<code class="highlight language-python"><span class="n">correlation_index</span><span class="p">(</span><span class="n">factors_1</span><span class="p">,</span> <span class="n">factors_2</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">5e-16</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;stacked&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>CorrIndex implementation to assess tensor decomposition outputs.
From [1] Sobhani et al 2022 (https://doi.org/10.1016/j.sigpro.2022.108457).
Metric is scaling and column-permutation invariant, wherein each column is a factor.</p>
<h6 id="cell2cell.tensor.metrics.correlation_index--parameters">Parameters</h6>
<p>factors_1 : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor. This is the result from a tensor decomposition,
    it can be found as the attribute <code>factors</code> in any tensor class derived from the
    class BaseTensor (e.g. BaseTensor.factors).</p>
<p>factors_2 : dict
    Similar to factors_1 but coming from another tensor decomposition of a tensor
    with equal shape.</p>
<p>tol : float, default=5e-16
    Precision threshold below which to call the CorrIndex score 0.</p>
<p>method : str, default='stacked'
    Method to obtain the CorrIndex by comparing the A matrices from two decompositions.
    Possible options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">stacked</span><span class="s1">&#39;</span> : <span class="nv">The</span> <span class="nv">original</span> <span class="nv">method</span> <span class="nv">implemented</span> <span class="nv">in</span> [<span class="mi">1</span>]. <span class="nv">Here</span> <span class="nv">all</span> <span class="nv">A</span> <span class="nv">matrices</span> <span class="nv">from</span> <span class="nv">the</span> <span class="nv">same</span> <span class="nv">decomposition</span> <span class="nv">are</span>
              <span class="nv">vertically</span> <span class="nv">concatenated</span>, <span class="nv">building</span> <span class="nv">a</span> <span class="nv">big</span> <span class="nv">A</span> <span class="nv">matrix</span> <span class="k">for</span> <span class="nv">each</span> <span class="nv">decomposition</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span> : <span class="nv">This</span> <span class="nv">computes</span> <span class="nv">the</span> <span class="nv">CorrIndex</span> <span class="k">for</span> <span class="nv">each</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">A</span> <span class="nv">matrices</span> <span class="ss">(</span><span class="nv">i</span>.<span class="nv">e</span>. <span class="nv">between</span> <span class="nv">A_1</span> <span class="nv">in</span> <span class="nv">factors_1</span> <span class="nv">and</span>
                <span class="nv">factors_2</span>, <span class="nv">between</span> <span class="nv">A_2</span> <span class="nv">in</span> <span class="nv">factors_1</span> <span class="nv">and</span> <span class="nv">factors_2</span>, <span class="nv">and</span> <span class="nv">so</span> <span class="nv">on</span><span class="ss">)</span>. <span class="k">Then</span> <span class="nv">the</span> <span class="nv">max</span> <span class="nv">score</span> <span class="nv">is</span>
                <span class="nv">selected</span> <span class="ss">(</span><span class="nv">the</span> <span class="nv">most</span> <span class="nv">conservative</span> <span class="nv">approach</span><span class="ss">)</span>. <span class="nv">In</span> <span class="nv">other</span> <span class="nv">words</span>, <span class="nv">it</span> <span class="nv">selects</span> <span class="nv">the</span> <span class="nv">max</span> <span class="nv">score</span> <span class="nv">among</span> <span class="nv">the</span>
                <span class="nv">CorrIndexes</span> <span class="nv">computed</span> <span class="nv">dimension</span><span class="o">-</span><span class="nv">wise</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">min_score</span><span class="s1">&#39;</span> : <span class="nv">Similar</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span>, <span class="nv">but</span> <span class="nv">the</span> <span class="nv">min</span> <span class="nv">score</span> <span class="nv">is</span> <span class="nv">selected</span> <span class="ss">(</span><span class="nv">the</span> <span class="nv">least</span> <span class="nv">conservative</span> <span class="nv">approach</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">avg_score</span><span class="s1">&#39;</span> : <span class="nv">Similar</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span>, <span class="nv">but</span> <span class="nv">the</span> <span class="nv">avg</span> <span class="nv">score</span> <span class="nv">is</span> <span class="nv">selected</span>.
</code></pre></div>

<h6 id="cell2cell.tensor.metrics.correlation_index--returns">Returns</h6>
<p>score : float
     CorrIndex metric [0,1]; lower score indicates higher similarity between matrices</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/metrics.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">correlation_index</span><span class="p">(</span><span class="n">factors_1</span><span class="p">,</span> <span class="n">factors_2</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">5e-16</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;stacked&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    CorrIndex implementation to assess tensor decomposition outputs.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    From [1] Sobhani et al 2022 (https://doi.org/10.1016/j.sigpro.2022.108457).</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Metric is scaling and column-permutation invariant, wherein each column is a factor.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    factors_1 : dict</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        dimension/order of the tensor. This is the result from a tensor decomposition,</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        it can be found as the attribute `factors` in any tensor class derived from the</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        class BaseTensor (e.g. BaseTensor.factors).</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    factors_2 : dict</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        Similar to factors_1 but coming from another tensor decomposition of a tensor</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        with equal shape.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    tol : float, default=5e-16</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Precision threshold below which to call the CorrIndex score 0.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    method : str, default=&#39;stacked&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Method to obtain the CorrIndex by comparing the A matrices from two decompositions.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Possible options are:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        - &#39;stacked&#39; : The original method implemented in [1]. Here all A matrices from the same decomposition are</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                      vertically concatenated, building a big A matrix for each decomposition.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - &#39;max_score&#39; : This computes the CorrIndex for each pair of A matrices (i.e. between A_1 in factors_1 and</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                        factors_2, between A_2 in factors_1 and factors_2, and so on). Then the max score is</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                        selected (the most conservative approach). In other words, it selects the max score among the</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                        CorrIndexes computed dimension-wise.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &#39;min_score&#39; : Similar to &#39;max_score&#39;, but the min score is selected (the least conservative approach).</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        - &#39;avg_score&#39; : Similar to &#39;max_score&#39;, but the avg score is selected.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    -------</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    score : float</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">         CorrIndex metric [0,1]; lower score indicates higher similarity between matrices</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">factors_1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">factors_1</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">factors_2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">factors_2</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># check input factors shape</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="k">for</span> <span class="n">factors</span> <span class="ow">in</span> <span class="p">[</span><span class="n">factors_1</span><span class="p">,</span> <span class="n">factors_2</span><span class="p">]:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">({</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">A</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="k">for</span> <span class="n">A</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">})</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Factors should be a list of loading matrices of the same rank&#39;</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="c1"># check method</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="n">options</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;stacked&#39;</span><span class="p">,</span> <span class="s1">&#39;max_score&#39;</span><span class="p">,</span> <span class="s1">&#39;min_score&#39;</span><span class="p">,</span> <span class="s1">&#39;avg_score&#39;</span><span class="p">]</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">if</span> <span class="n">method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">options</span><span class="p">:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The `method` must be either option among </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">options</span><span class="p">))</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;stacked&#39;</span><span class="p">:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="c1"># vertically stack loading matrices -- shape sum(tensor.shape)xR)</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="n">X_1</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">factors_1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)]</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">X_2</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">factors_2</span><span class="p">,</span> <span class="mi">0</span><span class="p">)]</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">X_1</span> <span class="o">=</span> <span class="n">factors_1</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="n">X_2</span> <span class="o">=</span> <span class="n">factors_2</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">for</span> <span class="n">x1</span><span class="p">,</span> <span class="n">x2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">X_1</span><span class="p">,</span> <span class="n">X_2</span><span class="p">):</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">x1</span><span class="p">)</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">x2</span><span class="p">):</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Factor matrices should be of the same shapes&#39;</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>    <span class="c1"># normalize columns to L2 norm - even if ran decomposition with normalize_factors=True</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>    <span class="n">col_norm_1</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x1</span> <span class="ow">in</span> <span class="n">X_1</span><span class="p">]</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>    <span class="n">col_norm_2</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">x2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x2</span> <span class="ow">in</span> <span class="n">X_2</span><span class="p">]</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>    <span class="k">for</span> <span class="n">cn1</span><span class="p">,</span> <span class="n">cn2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">col_norm_1</span><span class="p">,</span> <span class="n">col_norm_2</span><span class="p">):</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">cn1</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">cn2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Column norms must be non-zero&#39;</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">X_1</span> <span class="o">=</span> <span class="p">[</span><span class="n">x1</span> <span class="o">/</span> <span class="n">cn1</span> <span class="k">for</span> <span class="n">x1</span><span class="p">,</span> <span class="n">cn1</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">X_1</span><span class="p">,</span> <span class="n">col_norm_1</span><span class="p">)]</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">X_2</span> <span class="o">=</span> <span class="p">[</span><span class="n">x2</span> <span class="o">/</span> <span class="n">cn2</span> <span class="k">for</span> <span class="n">x2</span><span class="p">,</span> <span class="n">cn2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">X_2</span><span class="p">,</span> <span class="n">col_norm_2</span><span class="p">)]</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="n">corr_idxs</span> <span class="o">=</span> <span class="p">[</span><span class="n">_compute_correlation_index</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">x2</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">)</span> <span class="k">for</span> <span class="n">x1</span><span class="p">,</span> <span class="n">x2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">X_1</span><span class="p">,</span> <span class="n">X_2</span><span class="p">)]</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;stacked&#39;</span><span class="p">:</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="n">score</span> <span class="o">=</span> <span class="n">corr_idxs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;max_score&#39;</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>        <span class="n">score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">corr_idxs</span><span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;min_score&#39;</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="n">score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">corr_idxs</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;avg_score&#39;</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="n">score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">corr_idxs</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="n">score</span> <span class="o">=</span> <span class="mf">1.0</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="k">return</span> <span class="n">score</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.metrics.pairwise_correlation_index">
<code class="highlight language-python"><span class="n">pairwise_correlation_index</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">5e-16</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;stacked&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the CorrIndex between all pairs of factors</p>
<h6 id="cell2cell.tensor.metrics.pairwise_correlation_index--parameters">Parameters</h6>
<p>factors : list
    List with multiple Ordered dictionaries, each containing a dataframe with
    the factor loadings for each dimension/order of the tensor. This is the
    result from a tensor decomposition, it can be found as the attribute
    <code>factors</code> in any tensor class derived from the class BaseTensor
    (e.g. BaseTensor.factors).</p>
<p>tol : float, default=5e-16
    Precision threshold below which to call the CorrIndex score 0.</p>
<p>method : str, default='stacked'
    Method to obtain the CorrIndex by comparing the A matrices from two decompositions.
    Possible options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">stacked</span><span class="s1">&#39;</span> : <span class="nv">The</span> <span class="nv">original</span> <span class="nv">method</span> <span class="nv">implemented</span> <span class="nv">in</span> [<span class="mi">1</span>]. <span class="nv">Here</span> <span class="nv">all</span> <span class="nv">A</span> <span class="nv">matrices</span> <span class="nv">from</span> <span class="nv">the</span> <span class="nv">same</span> <span class="nv">decomposition</span> <span class="nv">are</span>
              <span class="nv">vertically</span> <span class="nv">concatenated</span>, <span class="nv">building</span> <span class="nv">a</span> <span class="nv">big</span> <span class="nv">A</span> <span class="nv">matrix</span> <span class="k">for</span> <span class="nv">each</span> <span class="nv">decomposition</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span> : <span class="nv">This</span> <span class="nv">computes</span> <span class="nv">the</span> <span class="nv">CorrIndex</span> <span class="k">for</span> <span class="nv">each</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">A</span> <span class="nv">matrices</span> <span class="ss">(</span><span class="nv">i</span>.<span class="nv">e</span>. <span class="nv">between</span> <span class="nv">A_1</span> <span class="nv">in</span> <span class="nv">factors_1</span> <span class="nv">and</span>
                <span class="nv">factors_2</span>, <span class="nv">between</span> <span class="nv">A_2</span> <span class="nv">in</span> <span class="nv">factors_1</span> <span class="nv">and</span> <span class="nv">factors_2</span>, <span class="nv">and</span> <span class="nv">so</span> <span class="nv">on</span><span class="ss">)</span>. <span class="k">Then</span> <span class="nv">the</span> <span class="nv">max</span> <span class="nv">score</span> <span class="nv">is</span>
                <span class="nv">selected</span> <span class="ss">(</span><span class="nv">the</span> <span class="nv">most</span> <span class="nv">conservative</span> <span class="nv">approach</span><span class="ss">)</span>. <span class="nv">In</span> <span class="nv">other</span> <span class="nv">words</span>, <span class="nv">it</span> <span class="nv">selects</span> <span class="nv">the</span> <span class="nv">max</span> <span class="nv">score</span> <span class="nv">among</span> <span class="nv">the</span>
                <span class="nv">CorrIndexes</span> <span class="nv">computed</span> <span class="nv">dimension</span><span class="o">-</span><span class="nv">wise</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">min_score</span><span class="s1">&#39;</span> : <span class="nv">Similar</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span>, <span class="nv">but</span> <span class="nv">the</span> <span class="nv">min</span> <span class="nv">score</span> <span class="nv">is</span> <span class="nv">selected</span> <span class="ss">(</span><span class="nv">the</span> <span class="nv">least</span> <span class="nv">conservative</span> <span class="nv">approach</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">avg_score</span><span class="s1">&#39;</span> : <span class="nv">Similar</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">max_score</span><span class="s1">&#39;</span>, <span class="nv">but</span> <span class="nv">the</span> <span class="nv">avg</span> <span class="nv">score</span> <span class="nv">is</span> <span class="nv">selected</span>.
</code></pre></div>

<h6 id="cell2cell.tensor.metrics.pairwise_correlation_index--returns">Returns</h6>
<p>scores : pd.DataFrame
     Dataframe with CorrIndex metric for each pair of decompositions.
     This metric bounds are [0,1]; lower score indicates higher similarity between matrices</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/metrics.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">pairwise_correlation_index</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">5e-16</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;stacked&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes the CorrIndex between all pairs of factors</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    factors : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List with multiple Ordered dictionaries, each containing a dataframe with</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        the factor loadings for each dimension/order of the tensor. This is the</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        result from a tensor decomposition, it can be found as the attribute</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        `factors` in any tensor class derived from the class BaseTensor</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        (e.g. BaseTensor.factors).</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    tol : float, default=5e-16</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Precision threshold below which to call the CorrIndex score 0.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    method : str, default=&#39;stacked&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Method to obtain the CorrIndex by comparing the A matrices from two decompositions.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Possible options are:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;stacked&#39; : The original method implemented in [1]. Here all A matrices from the same decomposition are</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">                      vertically concatenated, building a big A matrix for each decomposition.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;max_score&#39; : This computes the CorrIndex for each pair of A matrices (i.e. between A_1 in factors_1 and</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                        factors_2, between A_2 in factors_1 and factors_2, and so on). Then the max score is</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">                        selected (the most conservative approach). In other words, it selects the max score among the</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                        CorrIndexes computed dimension-wise.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;min_score&#39; : Similar to &#39;max_score&#39;, but the min score is selected (the least conservative approach).</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - &#39;avg_score&#39; : Similar to &#39;max_score&#39;, but the avg score is selected.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    -------</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    scores : pd.DataFrame</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">         Dataframe with CorrIndex metric for each pair of decompositions.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">         This metric bounds are [0,1]; lower score indicates higher similarity between matrices</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">factors</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>    <span class="n">idxs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">combinations</span><span class="p">(</span><span class="n">idxs</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">scores</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">N</span><span class="p">,</span> <span class="n">N</span><span class="p">)),</span><span class="n">index</span><span class="o">=</span><span class="n">idxs</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">idxs</span><span class="p">)</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">for</span> <span class="n">p1</span><span class="p">,</span> <span class="n">p2</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="n">corrindex</span> <span class="o">=</span> <span class="n">correlation_index</span><span class="p">(</span><span class="n">factors_1</span><span class="o">=</span><span class="n">factors</span><span class="p">[</span><span class="n">p1</span><span class="p">],</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>                                      <span class="n">factors_2</span><span class="o">=</span><span class="n">factors</span><span class="p">[</span><span class="n">p2</span><span class="p">],</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>                                      <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>                                      <span class="n">method</span><span class="o">=</span><span class="n">method</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>                                      <span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="n">scores</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">]</span> <span class="o">=</span> <span class="n">corrindex</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">scores</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">p2</span><span class="p">,</span> <span class="n">p1</span><span class="p">]</span> <span class="o">=</span> <span class="n">corrindex</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">return</span> <span class="n">scores</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.subset">
        <code>subset</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.subset.find_element_indexes">
<code class="highlight language-python"><span class="n">find_element_indexes</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">elements</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">original_order</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Finds the location/indexes of a list of elements in one of the
axis of an InteractionTensor.</p>
<h6 id="cell2cell.tensor.subset.find_element_indexes--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor</p>
<p>elements : list
    A list of names for the elements to find in one of the axis.</p>
<p>axis : int, default=0
    An axis of the interaction_tensor, representing one of
    its dimensions.</p>
<p>remove_duplicates : boolean, default=True
    Whether removing duplicated names in <code>elements</code>.</p>
<p>keep : str, default='first'
    Determines which duplicates (if any) to keep.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="nv">first</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">first</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">last</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">last</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">False</span> : <span class="nv">Drop</span> <span class="nv">all</span> <span class="nv">duplicates</span>.
</code></pre></div>

<p>original_order : boolean, default=False
    Whether keeping the original order of the elements in
    interaction_tensor.order_names[axis] or keeping the
    new order as indicated in <code>elements</code>.</p>
<h6 id="cell2cell.tensor.subset.find_element_indexes--returns">Returns</h6>
<p>indexes : list
    List of indexes for the elements that where found in the
    axis indicated of the interaction_tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/subset.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">find_element_indexes</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">elements</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">original_order</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Finds the location/indexes of a list of elements in one of the</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    axis of an InteractionTensor.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell2cell.tensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    elements : list</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A list of names for the elements to find in one of the axis.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    axis : int, default=0</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        An axis of the interaction_tensor, representing one of</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        its dimensions.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    remove_duplicates : boolean, default=True</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Whether removing duplicated names in `elements`.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    keep : str, default=&#39;first&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Determines which duplicates (if any) to keep.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - first : Drop duplicates except for the first occurrence.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        - last : Drop duplicates except for the last occurrence.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - False : Drop all duplicates.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    original_order : boolean, default=False</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        Whether keeping the original order of the elements in</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        interaction_tensor.order_names[axis] or keeping the</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        new order as indicated in `elements`.</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    -------</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    indexes : list</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        List of indexes for the elements that where found in the</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        axis indicated of the interaction_tensor.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="k">assert</span> <span class="n">axis</span> <span class="o">&lt;</span> <span class="nb">len</span> \
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="s2">&quot;List index out of range. &#39;axis&#39; must be one of the axis in the tensor.&quot;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="k">assert</span> <span class="n">axis</span> <span class="o">&lt;</span> <span class="nb">len</span> \
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">),</span> <span class="s2">&quot;List index out of range. interaction_tensor.order_names must have element names for each axis of the tensor.&quot;</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">elements</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">elements</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">elements</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">if</span> <span class="n">original_order</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="c1"># Avoids error for considering elements not in the tensor</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="n">elements</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">elements</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">axis</span><span class="p">]))</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">elements</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">elements</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">axis</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="c1"># Find duplicates if we are removing them</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="n">to_exclude</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">if</span> <span class="n">remove_duplicates</span><span class="p">:</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">dup_dict</span> <span class="o">=</span> <span class="n">find_duplicates</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">axis</span><span class="p">])</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dup_dict</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Only if we have duplicate items</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>            <span class="k">if</span> <span class="n">keep</span> <span class="o">==</span> <span class="s1">&#39;first&#39;</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">dup_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>                    <span class="n">to_exclude</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>            <span class="k">elif</span> <span class="n">keep</span> <span class="o">==</span> <span class="s1">&#39;last&#39;</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">dup_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                    <span class="n">to_exclude</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">v</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>            <span class="k">elif</span> <span class="ow">not</span> <span class="n">keep</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">dup_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                    <span class="n">to_exclude</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Not a valid option was selected for the parameter `keep`&quot;</span><span class="p">)</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="c1"># Find indexes in the tensor</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">indexes</span> <span class="o">=</span> <span class="nb">sum</span> \
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>        <span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">axis</span><span class="p">])</span> <span class="o">==</span> <span class="n">element</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">elements</span><span class="p">],</span> <span class="p">[])</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="c1"># Exclude duplicates if any to exclude</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="n">indexes</span> <span class="o">=</span> <span class="p">[</span><span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">indexes</span> <span class="k">if</span> <span class="n">idx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">]</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>    <span class="k">return</span> <span class="n">indexes</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.subset.subset_metadata">
<code class="highlight language-python"><span class="n">subset_metadata</span><span class="p">(</span><span class="n">tensor_metadata</span><span class="p">,</span> <span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Subsets the metadata of an InteractionTensor to contain only
elements in a reference InteractionTensor (interaction_tensor).</p>
<h6 id="cell2cell.tensor.subset.subset_metadata--parameters">Parameters</h6>
<p>tensor_metadata : list
    List of pandas dataframes with metadata information for elements of each
    dimension in the tensor. A column called as the variable <code>sample_col</code> contains
    the name of each element in the tensor while another column called as the
    variable <code>group_col</code> contains the metadata or grouping information of each
    element.</p>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor. This tensor is used as reference to subset the metadata.
    The subset metadata will contain only elements that are present in this
    tensor, so if metadata was originally built for another tensor, the elements
    that are exclusive for that original tensor will be excluded.</p>
<p>sample_col : str, default='Element'
    Name of the column containing the element names in the metadata.</p>
<h6 id="cell2cell.tensor.subset.subset_metadata--returns">Returns</h6>
<p>subset_metadata : list
    List of pandas dataframes with metadata information for elements contained
    in <code>interaction_tensor.order_names</code>. It is a subset of <code>tensor_metadata</code>.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/subset.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">subset_metadata</span><span class="p">(</span><span class="n">tensor_metadata</span><span class="p">,</span> <span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s1">&#39;Element&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Subsets the metadata of an InteractionTensor to contain only</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    elements in a reference InteractionTensor (interaction_tensor).</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    tensor_metadata : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of pandas dataframes with metadata information for elements of each</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        dimension in the tensor. A column called as the variable `sample_col` contains</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        the name of each element in the tensor while another column called as the</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        variable `group_col` contains the metadata or grouping information of each</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        element.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        cell2cell.tensor. This tensor is used as reference to subset the metadata.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        The subset metadata will contain only elements that are present in this</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        tensor, so if metadata was originally built for another tensor, the elements</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        that are exclusive for that original tensor will be excluded.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    sample_col : str, default=&#39;Element&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Name of the column containing the element names in the metadata.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    -------</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    subset_metadata : list</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        List of pandas dataframes with metadata information for elements contained</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        in `interaction_tensor.order_names`. It is a subset of `tensor_metadata`.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="n">subset_metadata</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">meta</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tensor_metadata</span><span class="p">):</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">if</span> <span class="n">meta</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>            <span class="n">tmp_meta</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sample_col</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>            <span class="n">tmp_meta</span> <span class="o">=</span> <span class="n">tmp_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="p">:]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>            <span class="n">tmp_meta</span> <span class="o">=</span> <span class="n">tmp_meta</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="n">subset_metadata</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_meta</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>            <span class="n">subset_metadata</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="k">return</span> <span class="n">subset_metadata</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.subset.subset_tensor">
<code class="highlight language-python"><span class="n">subset_tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">subset_dict</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">original_order</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Subsets an InteractionTensor to contain only specific elements in
respective dimensions.</p>
<h6 id="cell2cell.tensor.subset.subset_tensor--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor generated with any of the tensor class in
    cell2cell.tensor</p>
<p>subset_dict : dict
    Dictionary to subset the tensor. It must contain the axes or
    dimensions that will be subset as the keys of the dictionary
    and the values corresponds to lists of element names for the
    respective axes or dimensions. Those axes that are not present
    in this dictionary will not be subset.
    E.g. {0 : ['Context 1', 'Context2'], 1: ['LR 10', 'LR 100']}</p>
<p>remove_duplicates : boolean, default=True
    Whether removing duplicated names in <code>elements</code>.</p>
<p>keep : str, default='first'
    Determines which duplicates (if any) to keep.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="nv">first</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">first</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">last</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">last</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">False</span> : <span class="nv">Drop</span> <span class="nv">all</span> <span class="nv">duplicates</span>.
</code></pre></div>

<p>original_order : boolean, default=False
    Whether keeping the original order of the elements in
    interaction_tensor.order_names or keeping the
    new order as indicated in the lists in the <code>subset_dict</code>.</p>
<h6 id="cell2cell.tensor.subset.subset_tensor--returns">Returns</h6>
<p>subset_tensor : cell2cell.tensor.BaseTensor
    A copy of interaction_tensor that was subset to contain
    only the elements specified for the respective axis in the
    <code>subset_dict</code>. Corresponds to a communication tensor
    generated with any of the tensor class in cell2cell.tensor</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/subset.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">subset_tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">subset_dict</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">original_order</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Subsets an InteractionTensor to contain only specific elements in</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    respective dimensions.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A communication tensor generated with any of the tensor class in</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        cell2cell.tensor</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    subset_dict : dict</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Dictionary to subset the tensor. It must contain the axes or</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        dimensions that will be subset as the keys of the dictionary</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        and the values corresponds to lists of element names for the</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        respective axes or dimensions. Those axes that are not present</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        in this dictionary will not be subset.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        E.g. {0 : [&#39;Context 1&#39;, &#39;Context2&#39;], 1: [&#39;LR 10&#39;, &#39;LR 100&#39;]}</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    remove_duplicates : boolean, default=True</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Whether removing duplicated names in `elements`.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    keep : str, default=&#39;first&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Determines which duplicates (if any) to keep.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        - first : Drop duplicates except for the first occurrence.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - last : Drop duplicates except for the last occurrence.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - False : Drop all duplicates.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    original_order : boolean, default=False</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Whether keeping the original order of the elements in</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        interaction_tensor.order_names or keeping the</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        new order as indicated in the lists in the `subset_dict`.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    -------</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    subset_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        A copy of interaction_tensor that was subset to contain</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        only the elements specified for the respective axis in the</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        `subset_dict`. Corresponds to a communication tensor</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        generated with any of the tensor class in cell2cell.tensor</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Perform a deep copy of the original tensor and reset previous factorization</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">subset_tensor</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">subset_tensor</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>    <span class="n">subset_tensor</span><span class="o">.</span><span class="n">tl_object</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="n">subset_tensor</span><span class="o">.</span><span class="n">factors</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="c1"># Initialize tensor into a numpy object for performing subset</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="n">context</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">context</span><span class="p">(</span><span class="n">subset_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>    <span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">subset_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>    <span class="n">mask</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="k">if</span> <span class="n">subset_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">subset_tensor</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>    <span class="c1"># Search for indexes</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="n">axis_idxs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">subset_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>        <span class="k">if</span> <span class="n">k</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>                <span class="n">idx</span> <span class="o">=</span> <span class="n">find_element_indexes</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">=</span><span class="n">subset_tensor</span><span class="p">,</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>                                           <span class="n">elements</span><span class="o">=</span><span class="n">v</span><span class="p">,</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>                                           <span class="n">axis</span><span class="o">=</span><span class="n">k</span><span class="p">,</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>                                           <span class="n">remove_duplicates</span><span class="o">=</span><span class="n">remove_duplicates</span><span class="p">,</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>                                           <span class="n">keep</span><span class="o">=</span><span class="n">keep</span><span class="p">,</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>                                           <span class="n">original_order</span><span class="o">=</span><span class="n">original_order</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>                                           <span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No elements found for axis </span><span class="si">{}</span><span class="s2">. It will return an empty tensor.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>                <span class="n">axis_idxs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">idx</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Axis </span><span class="si">{}</span><span class="s2"> is out of index, not considering elements in this axis.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="c1"># Subset tensor</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">axis_idxs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>        <span class="k">if</span> <span class="n">tensor</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,):</span>  <span class="c1"># Avoids error when returned empty tensor</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>            <span class="n">tensor</span> <span class="o">=</span> <span class="n">tensor</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">indices</span><span class="o">=</span><span class="n">v</span><span class="p">,</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>                                 <span class="n">axis</span><span class="o">=</span><span class="n">k</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                                 <span class="p">)</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>            <span class="n">subset_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">subset_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">v</span><span class="p">]</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>            <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>                <span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">indices</span><span class="o">=</span><span class="n">v</span><span class="p">,</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>                                 <span class="n">axis</span><span class="o">=</span><span class="n">k</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>                                 <span class="p">)</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>    <span class="c1"># Restore tensor and mask properties</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>    <span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor</span><span class="p">,</span> <span class="o">**</span><span class="n">context</span><span class="p">)</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="o">**</span><span class="n">context</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>    <span class="n">subset_tensor</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tensor</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="n">subset_tensor</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="k">return</span> <span class="n">subset_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.tensor">
        <code>tensor</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor">
        <code>
BaseTensor        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Empty base tensor class that contains the main functions for the Tensor
Factorization of a Communication Tensor</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor--attributes">Attributes</h6>
<p>communication_score : str
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                           of a ligand from a sender cell and the
                           expression of a receptor on a receiver cell.
</code></pre></div>

<p>how : str
    Approach to consider cell types and genes present across multiple contexts.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">inner</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
            <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span>
            <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_genes</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_cells</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
</code></pre></div>

<p>outer_fraction : float
    Threshold to filter the elements when <code>how</code> includes any outer option.
    Elements with a fraction abundance across samples (in <code>rnaseq_matrices</code>)
    at least this threshold will be included. When this value is 0, considers
    all elements across the samples. When this value is 1, it acts as using
    <code>how='inner'</code>.</p>
<p>tensor : tensorly.tensor
    Tensor object created with the library tensorly.</p>
<p>genes : list
    List of strings detailing the genes used through all contexts. Obtained
    depending on the attribute 'how'.</p>
<p>cells : list
    List of strings detailing the cells used through all contexts. Obtained
    depending on the attribute 'how'.</p>
<p>order_names : list
    List of lists containing the string names of each element in each of the
    dimensions or orders in the tensor. For a 4D-Communication tensor, the first
    list should contain the names of the contexts, the second the names of the
    ligand-receptor interactions, the third the names of the sender cells and the
    fourth the names of the receiver cells.</p>
<p>order_labels : list
    List of labels for dimensions or orders in the tensor.</p>
<p>tl_object : ndarray list
    A tensorly object containing a list of initialized factors of the tensor
    decomposition where element <code>i</code> is of shape (tensor.shape[i], rank).</p>
<p>norm_tl_object : ndarray list
    A tensorly object containing a list of initialized factors of the tensor
    decomposition where element <code>i</code> is of shape (tensor.shape[i], rank). This
    results from normalizing the factor loadings of the tl_object.</p>
<p>factors : dict
    Ordered dictionary containing a dataframe with the factor loadings for each
    dimension/order of the tensor.</p>
<p>rank : int
    Rank of the Tensor Factorization (number of factors to deconvolve the original
    tensor).</p>
<p>mask : ndarray list
    Helps avoiding missing values during a tensor factorization. A mask should be
    a boolean array of the same shape as the original tensor and should be 0
    where the values are missing and 1 everywhere else.</p>
<p>explained_variance : float
    Explained variance score for a tnesor factorization.</p>
<p>explained_variance_ratio_ : ndarray list
    Percentage of variance explained by each of the factors. Only present when
    "normalize_loadings" is True. Otherwise, it is None.</p>
<p>loc_nans : ndarray list
    An array of shape equal to <code>tensor</code> with ones where NaN values were assigned
    when building the tensor. Other values are zeros. It stores the
    location of the NaN values.</p>
<p>loc_zeros : ndarray list
    An array of shape equal to <code>tensor</code> with ones where zeros that are not in
    <code>loc_nans</code> are located. Other values are assigned a zero. It tracks the
    real zero values rather than NaN values that were converted to zero.</p>
<p>elbow_metric : str
    Stores the metric used to perform the elbow analysis (y-axis).</p>
<div class="codehilite"><pre><span></span><code>    - &#39;error&#39; : Normalized error to compute the elbow.
    - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).
</code></pre></div>

<p>elbow_metric_mean : ndarray list
    Metric computed from the elbow analysis for each of the different rank
    evaluated. This list contains (X,Y) pairs where X values are the
    different ranks and Y values are the mean value of the metric employed.
    This mean is computed from multiple runs, or the values for just one run.
    Metric could be the normalized error of the decomposition or the similarity
    between multiple runs with different initialization, based on the
    CorrIndex.</p>
<p>elbow_metric_raw : ndarray list
    Similar to <code>elbow_metric_mean</code>, but instead of containing (X, Y) pairs,
    it is an array of shape runs by ranks that were used for the analysis.
    It contains all the metrics for each run in each of the evaluated ranks.</p>
<p>shape : tuple
    Shape of the tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">BaseTensor</span><span class="p">():</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Empty base tensor class that contains the main functions for the Tensor</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Factorization of a Communication Tensor</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Attributes</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    communication_score : str</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">                                   of a ligand from a sender cell and the</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                                   expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    how : str</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Approach to consider cell types and genes present across multiple contexts.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;inner&#39; : Considers only cell types and genes that are present in all</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                    contexts (intersection).</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;outer&#39; : Considers all cell types and genes that are present</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                    across contexts (union).</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;outer_genes&#39; : Considers only cell types that are present in all</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                          contexts (intersection), while all genes that are</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &#39;outer_cells&#39; : Considers only genes that are present in all</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">                          contexts (intersection), while all cell types that are</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    outer_fraction : float</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Threshold to filter the elements when `how` includes any outer option.</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Elements with a fraction abundance across samples (in `rnaseq_matrices`)</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        at least this threshold will be included. When this value is 0, considers</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        all elements across the samples. When this value is 1, it acts as using</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        `how=&#39;inner&#39;`.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    tensor : tensorly.tensor</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Tensor object created with the library tensorly.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    genes : list</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        List of strings detailing the genes used through all contexts. Obtained</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        depending on the attribute &#39;how&#39;.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">    cells : list</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        List of strings detailing the cells used through all contexts. Obtained</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        depending on the attribute &#39;how&#39;.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    order_names : list</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        List of lists containing the string names of each element in each of the</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        dimensions or orders in the tensor. For a 4D-Communication tensor, the first</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        list should contain the names of the contexts, the second the names of the</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        ligand-receptor interactions, the third the names of the sender cells and the</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        fourth the names of the receiver cells.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    order_labels : list</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        List of labels for dimensions or orders in the tensor.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    tl_object : ndarray list</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        A tensorly object containing a list of initialized factors of the tensor</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        decomposition where element `i` is of shape (tensor.shape[i], rank).</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    norm_tl_object : ndarray list</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        A tensorly object containing a list of initialized factors of the tensor</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        decomposition where element `i` is of shape (tensor.shape[i], rank). This</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        results from normalizing the factor loadings of the tl_object.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">    factors : dict</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        Ordered dictionary containing a dataframe with the factor loadings for each</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        dimension/order of the tensor.</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">    rank : int</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        Rank of the Tensor Factorization (number of factors to deconvolve the original</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        tensor).</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">    mask : ndarray list</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        Helps avoiding missing values during a tensor factorization. A mask should be</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        a boolean array of the same shape as the original tensor and should be 0</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        where the values are missing and 1 everywhere else.</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    explained_variance : float</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        Explained variance score for a tnesor factorization.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">    explained_variance_ratio_ : ndarray list</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">        Percentage of variance explained by each of the factors. Only present when</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">        &quot;normalize_loadings&quot; is True. Otherwise, it is None.</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">    loc_nans : ndarray list</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">        An array of shape equal to `tensor` with ones where NaN values were assigned</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">        when building the tensor. Other values are zeros. It stores the</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        location of the NaN values.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">    loc_zeros : ndarray list</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">        An array of shape equal to `tensor` with ones where zeros that are not in</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">        `loc_nans` are located. Other values are assigned a zero. It tracks the</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">        real zero values rather than NaN values that were converted to zero.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">    elbow_metric : str</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">        Stores the metric used to perform the elbow analysis (y-axis).</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">            - &#39;error&#39; : Normalized error to compute the elbow.</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">            - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">    elbow_metric_mean : ndarray list</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a><span class="sd">        Metric computed from the elbow analysis for each of the different rank</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a><span class="sd">        evaluated. This list contains (X,Y) pairs where X values are the</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a><span class="sd">        different ranks and Y values are the mean value of the metric employed.</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a><span class="sd">        This mean is computed from multiple runs, or the values for just one run.</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a><span class="sd">        Metric could be the normalized error of the decomposition or the similarity</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a><span class="sd">        between multiple runs with different initialization, based on the</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a><span class="sd">        CorrIndex.</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a><span class="sd">    elbow_metric_raw : ndarray list</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a><span class="sd">        Similar to `elbow_metric_mean`, but instead of containing (X, Y) pairs,</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a><span class="sd">        it is an array of shape runs by ranks that were used for the analysis.</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a><span class="sd">        It contains all the metrics for each run in each of the evaluated ranks.</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a><span class="sd">    shape : tuple</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a><span class="sd">        Shape of the tensor.</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="c1"># Save variables for this class</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">how</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">outer_fraction</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cells</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">order_names</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">norm_tl_object</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">factors</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_mean</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_raw</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>        <span class="sd">&#39;&#39;&#39;Performs a deep copy of this object.&#39;&#39;&#39;</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="kn">import</span> <span class="nn">copy</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="k">return</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="nd">@property</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>    <span class="k">def</span> <span class="nf">shape</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="sd">&#39;&#39;&#39;Returns the shape of the tensor&#39;&#39;&#39;</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="s1">&#39;shape&#39;</span><span class="p">):</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>            <span class="k">return</span> <span class="p">()</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>    <span class="k">def</span> <span class="nf">write_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>        <span class="sd">&#39;&#39;&#39;Exports this object into a pickle file.</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a><span class="sd">        filename : str</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a><span class="sd">            Complete path to the file wherein the variable will be</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a><span class="sd">            stored. For example:</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a><span class="sd">            /home/user/variable.pkl</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>        <span class="kn">from</span> <span class="nn">cell2cell.io.save_data</span> <span class="kn">import</span> <span class="n">export_variable_with_pickle</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>        <span class="n">export_variable_with_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>    <span class="k">def</span> <span class="nf">to_device</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">device</span><span class="p">):</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>        <span class="sd">&#39;&#39;&#39;Changes the device where the tensor</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a><span class="sd">        is analyzed.</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a><span class="sd">        device : str</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a><span class="sd">            Device name to use for the decomposition.</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a><span class="sd">            Options could be &#39;cpu&#39;, &#39;cuda&#39;, &#39;gpu&#39;, depending on</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a><span class="sd">            the backend used with tensorly.</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>        <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>        <span class="k">except</span><span class="p">:</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Device is either not available or the backend used with tensorly does not support this device.</span><span class="se">\</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a><span class="s1">                   Try changing it with tensorly.set_backend(&quot;&lt;backend_name&gt;&quot;) before.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>    <span class="k">def</span> <span class="nf">compute_tensor_factorization</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;svd&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>                                     <span class="n">runs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">normalize_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">var_ordered_factors</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">10e-7</span><span class="p">,</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>                                     <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>        <span class="sd">&#39;&#39;&#39;Performs a Tensor Factorization.</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a><span class="sd">        There are no returns, instead the attributes factors and rank</span>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a><span class="sd">         of the Tensor class are updated.</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-206" name="__codelineno-0-206" href="#__codelineno-0-206"></a><span class="sd">        rank : int</span>
<a id="__codelineno-0-207" name="__codelineno-0-207" href="#__codelineno-0-207"></a><span class="sd">            Rank of the Tensor Factorization (number of factors to deconvolve the original</span>
<a id="__codelineno-0-208" name="__codelineno-0-208" href="#__codelineno-0-208"></a><span class="sd">            tensor).</span>
<a id="__codelineno-0-209" name="__codelineno-0-209" href="#__codelineno-0-209"></a>
<a id="__codelineno-0-210" name="__codelineno-0-210" href="#__codelineno-0-210"></a><span class="sd">        tf_type : str, default=&#39;non_negative_cp&#39;</span>
<a id="__codelineno-0-211" name="__codelineno-0-211" href="#__codelineno-0-211"></a><span class="sd">            Type of Tensor Factorization.</span>
<a id="__codelineno-0-212" name="__codelineno-0-212" href="#__codelineno-0-212"></a>
<a id="__codelineno-0-213" name="__codelineno-0-213" href="#__codelineno-0-213"></a><span class="sd">            - &#39;non_negative_cp&#39; : Non-negative PARAFAC through the traditional ALS.</span>
<a id="__codelineno-0-214" name="__codelineno-0-214" href="#__codelineno-0-214"></a><span class="sd">            - &#39;non_negative_cp_hals&#39; : Non-negative PARAFAC through the Hierarchical ALS.</span>
<a id="__codelineno-0-215" name="__codelineno-0-215" href="#__codelineno-0-215"></a><span class="sd">                                       It reaches an optimal solution faster than the</span>
<a id="__codelineno-0-216" name="__codelineno-0-216" href="#__codelineno-0-216"></a><span class="sd">                                       traditional ALS, but it does not allow a mask.</span>
<a id="__codelineno-0-217" name="__codelineno-0-217" href="#__codelineno-0-217"></a><span class="sd">            - &#39;parafac&#39; : PARAFAC through the traditional ALS. It allows negative loadings.</span>
<a id="__codelineno-0-218" name="__codelineno-0-218" href="#__codelineno-0-218"></a><span class="sd">            - &#39;constrained_parafac&#39; : PARAFAC through the traditional ALS. It allows</span>
<a id="__codelineno-0-219" name="__codelineno-0-219" href="#__codelineno-0-219"></a><span class="sd">                                      negative loadings. Also, it incorporates L1 and L2</span>
<a id="__codelineno-0-220" name="__codelineno-0-220" href="#__codelineno-0-220"></a><span class="sd">                                      regularization, includes a &#39;non_negative&#39; option, and</span>
<a id="__codelineno-0-221" name="__codelineno-0-221" href="#__codelineno-0-221"></a><span class="sd">                                      allows constraining the sparsity of the decomposition.</span>
<a id="__codelineno-0-222" name="__codelineno-0-222" href="#__codelineno-0-222"></a><span class="sd">                                      For more information, see</span>
<a id="__codelineno-0-223" name="__codelineno-0-223" href="#__codelineno-0-223"></a><span class="sd">                                      http://tensorly.org/stable/modules/generated/tensorly.decomposition.constrained_parafac.html#tensorly.decomposition.constrained_parafac</span>
<a id="__codelineno-0-224" name="__codelineno-0-224" href="#__codelineno-0-224"></a>
<a id="__codelineno-0-225" name="__codelineno-0-225" href="#__codelineno-0-225"></a>
<a id="__codelineno-0-226" name="__codelineno-0-226" href="#__codelineno-0-226"></a><span class="sd">        init : str, default=&#39;svd&#39;</span>
<a id="__codelineno-0-227" name="__codelineno-0-227" href="#__codelineno-0-227"></a><span class="sd">            Initialization method for computing the Tensor Factorization.</span>
<a id="__codelineno-0-228" name="__codelineno-0-228" href="#__codelineno-0-228"></a><span class="sd">            {svd, random}</span>
<a id="__codelineno-0-229" name="__codelineno-0-229" href="#__codelineno-0-229"></a>
<a id="__codelineno-0-230" name="__codelineno-0-230" href="#__codelineno-0-230"></a><span class="sd">        svd : str, default=&#39;numpy_svd&#39;</span>
<a id="__codelineno-0-231" name="__codelineno-0-231" href="#__codelineno-0-231"></a><span class="sd">            Function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS</span>
<a id="__codelineno-0-232" name="__codelineno-0-232" href="#__codelineno-0-232"></a>
<a id="__codelineno-0-233" name="__codelineno-0-233" href="#__codelineno-0-233"></a><span class="sd">        random_state : int, default=None</span>
<a id="__codelineno-0-234" name="__codelineno-0-234" href="#__codelineno-0-234"></a><span class="sd">            Seed for randomization.</span>
<a id="__codelineno-0-235" name="__codelineno-0-235" href="#__codelineno-0-235"></a>
<a id="__codelineno-0-236" name="__codelineno-0-236" href="#__codelineno-0-236"></a><span class="sd">        runs : int, default=1</span>
<a id="__codelineno-0-237" name="__codelineno-0-237" href="#__codelineno-0-237"></a><span class="sd">            Number of models to choose among and find the lowest error.</span>
<a id="__codelineno-0-238" name="__codelineno-0-238" href="#__codelineno-0-238"></a><span class="sd">            This helps to avoid local minima when using runs &gt; 1.</span>
<a id="__codelineno-0-239" name="__codelineno-0-239" href="#__codelineno-0-239"></a>
<a id="__codelineno-0-240" name="__codelineno-0-240" href="#__codelineno-0-240"></a><span class="sd">        normalize_loadings : boolean, default=True</span>
<a id="__codelineno-0-241" name="__codelineno-0-241" href="#__codelineno-0-241"></a><span class="sd">            Whether normalizing the loadings in each factor to unit</span>
<a id="__codelineno-0-242" name="__codelineno-0-242" href="#__codelineno-0-242"></a><span class="sd">            Euclidean length.</span>
<a id="__codelineno-0-243" name="__codelineno-0-243" href="#__codelineno-0-243"></a>
<a id="__codelineno-0-244" name="__codelineno-0-244" href="#__codelineno-0-244"></a><span class="sd">        var_ordered_factors : boolean, default=True</span>
<a id="__codelineno-0-245" name="__codelineno-0-245" href="#__codelineno-0-245"></a><span class="sd">            Whether ordering factors by the variance they explain. The order is from</span>
<a id="__codelineno-0-246" name="__codelineno-0-246" href="#__codelineno-0-246"></a><span class="sd">            highest to lowest variance. `normalize_loadings` must be True. Otherwise,</span>
<a id="__codelineno-0-247" name="__codelineno-0-247" href="#__codelineno-0-247"></a><span class="sd">            this parameter is ignored.</span>
<a id="__codelineno-0-248" name="__codelineno-0-248" href="#__codelineno-0-248"></a>
<a id="__codelineno-0-249" name="__codelineno-0-249" href="#__codelineno-0-249"></a><span class="sd">        tol : float, default=10e-7</span>
<a id="__codelineno-0-250" name="__codelineno-0-250" href="#__codelineno-0-250"></a><span class="sd">            Tolerance for the decomposition algorithm to stop when the variation in</span>
<a id="__codelineno-0-251" name="__codelineno-0-251" href="#__codelineno-0-251"></a><span class="sd">            the reconstruction error is less than the tolerance. Lower `tol` helps</span>
<a id="__codelineno-0-252" name="__codelineno-0-252" href="#__codelineno-0-252"></a><span class="sd">            to improve the solution obtained from the decomposition, but it takes</span>
<a id="__codelineno-0-253" name="__codelineno-0-253" href="#__codelineno-0-253"></a><span class="sd">            longer to run.</span>
<a id="__codelineno-0-254" name="__codelineno-0-254" href="#__codelineno-0-254"></a>
<a id="__codelineno-0-255" name="__codelineno-0-255" href="#__codelineno-0-255"></a><span class="sd">        n_iter_max : int, default=100</span>
<a id="__codelineno-0-256" name="__codelineno-0-256" href="#__codelineno-0-256"></a><span class="sd">            Maximum number of iteration to reach an optimal solution with the</span>
<a id="__codelineno-0-257" name="__codelineno-0-257" href="#__codelineno-0-257"></a><span class="sd">            decomposition algorithm. Higher `n_iter_max`helps to improve the solution</span>
<a id="__codelineno-0-258" name="__codelineno-0-258" href="#__codelineno-0-258"></a><span class="sd">            obtained from the decomposition, but it takes longer to run.</span>
<a id="__codelineno-0-259" name="__codelineno-0-259" href="#__codelineno-0-259"></a>
<a id="__codelineno-0-260" name="__codelineno-0-260" href="#__codelineno-0-260"></a><span class="sd">        verbose : boolean, default=False</span>
<a id="__codelineno-0-261" name="__codelineno-0-261" href="#__codelineno-0-261"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-262" name="__codelineno-0-262" href="#__codelineno-0-262"></a>
<a id="__codelineno-0-263" name="__codelineno-0-263" href="#__codelineno-0-263"></a><span class="sd">        **kwargs : dict</span>
<a id="__codelineno-0-264" name="__codelineno-0-264" href="#__codelineno-0-264"></a><span class="sd">            Extra arguments for the tensor factorization according to inputs in tensorly.</span>
<a id="__codelineno-0-265" name="__codelineno-0-265" href="#__codelineno-0-265"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-266" name="__codelineno-0-266" href="#__codelineno-0-266"></a>        <span class="n">tensor_dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-267" name="__codelineno-0-267" href="#__codelineno-0-267"></a>        <span class="n">best_err</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span>
<a id="__codelineno-0-268" name="__codelineno-0-268" href="#__codelineno-0-268"></a>        <span class="n">tf</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-269" name="__codelineno-0-269" href="#__codelineno-0-269"></a>
<a id="__codelineno-0-270" name="__codelineno-0-270" href="#__codelineno-0-270"></a>        <span class="k">if</span> <span class="n">kwargs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-271" name="__codelineno-0-271" href="#__codelineno-0-271"></a>            <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;return_errors&#39;</span> <span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<a id="__codelineno-0-272" name="__codelineno-0-272" href="#__codelineno-0-272"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-273" name="__codelineno-0-273" href="#__codelineno-0-273"></a>            <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;return_errors&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-274" name="__codelineno-0-274" href="#__codelineno-0-274"></a>
<a id="__codelineno-0-275" name="__codelineno-0-275" href="#__codelineno-0-275"></a>        <span class="k">for</span> <span class="n">run</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">runs</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="n">runs</span><span class="o">==</span><span class="mi">1</span><span class="p">)):</span>
<a id="__codelineno-0-276" name="__codelineno-0-276" href="#__codelineno-0-276"></a>            <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-277" name="__codelineno-0-277" href="#__codelineno-0-277"></a>                <span class="n">random_state_</span> <span class="o">=</span> <span class="n">random_state</span> <span class="o">+</span> <span class="n">run</span>
<a id="__codelineno-0-278" name="__codelineno-0-278" href="#__codelineno-0-278"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-279" name="__codelineno-0-279" href="#__codelineno-0-279"></a>                <span class="n">random_state_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-280" name="__codelineno-0-280" href="#__codelineno-0-280"></a>            <span class="n">local_tf</span><span class="p">,</span> <span class="n">errors</span> <span class="o">=</span> <span class="n">_compute_tensor_factorization</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-281" name="__codelineno-0-281" href="#__codelineno-0-281"></a>                                                             <span class="n">rank</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-282" name="__codelineno-0-282" href="#__codelineno-0-282"></a>                                                             <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-283" name="__codelineno-0-283" href="#__codelineno-0-283"></a>                                                             <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-284" name="__codelineno-0-284" href="#__codelineno-0-284"></a>                                                             <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-285" name="__codelineno-0-285" href="#__codelineno-0-285"></a>                                                             <span class="n">random_state</span><span class="o">=</span><span class="n">random_state_</span><span class="p">,</span>
<a id="__codelineno-0-286" name="__codelineno-0-286" href="#__codelineno-0-286"></a>                                                             <span class="n">mask</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-287" name="__codelineno-0-287" href="#__codelineno-0-287"></a>                                                             <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-288" name="__codelineno-0-288" href="#__codelineno-0-288"></a>                                                             <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-289" name="__codelineno-0-289" href="#__codelineno-0-289"></a>                                                             <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-290" name="__codelineno-0-290" href="#__codelineno-0-290"></a>                                                             <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-291" name="__codelineno-0-291" href="#__codelineno-0-291"></a>            <span class="c1"># This helps to obtain proper error when the mask is not None.</span>
<a id="__codelineno-0-292" name="__codelineno-0-292" href="#__codelineno-0-292"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-293" name="__codelineno-0-293" href="#__codelineno-0-293"></a>                <span class="n">err</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">errors</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-294" name="__codelineno-0-294" href="#__codelineno-0-294"></a>                <span class="k">if</span> <span class="n">best_err</span> <span class="o">&gt;</span> <span class="n">err</span><span class="p">:</span>
<a id="__codelineno-0-295" name="__codelineno-0-295" href="#__codelineno-0-295"></a>                    <span class="n">best_err</span> <span class="o">=</span> <span class="n">err</span>
<a id="__codelineno-0-296" name="__codelineno-0-296" href="#__codelineno-0-296"></a>                    <span class="n">tf</span> <span class="o">=</span> <span class="n">local_tf</span>
<a id="__codelineno-0-297" name="__codelineno-0-297" href="#__codelineno-0-297"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-298" name="__codelineno-0-298" href="#__codelineno-0-298"></a>                <span class="n">err</span> <span class="o">=</span> <span class="n">_compute_norm_error</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="n">local_tf</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-299" name="__codelineno-0-299" href="#__codelineno-0-299"></a>                <span class="k">if</span> <span class="n">best_err</span> <span class="o">&gt;</span> <span class="n">err</span><span class="p">:</span>
<a id="__codelineno-0-300" name="__codelineno-0-300" href="#__codelineno-0-300"></a>                    <span class="n">best_err</span> <span class="o">=</span> <span class="n">err</span>
<a id="__codelineno-0-301" name="__codelineno-0-301" href="#__codelineno-0-301"></a>                    <span class="n">tf</span> <span class="o">=</span> <span class="n">local_tf</span>
<a id="__codelineno-0-302" name="__codelineno-0-302" href="#__codelineno-0-302"></a>
<a id="__codelineno-0-303" name="__codelineno-0-303" href="#__codelineno-0-303"></a>        <span class="k">if</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-304" name="__codelineno-0-304" href="#__codelineno-0-304"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Best model has a normalized error of: </span><span class="si">{0:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_err</span><span class="p">))</span>
<a id="__codelineno-0-305" name="__codelineno-0-305" href="#__codelineno-0-305"></a>
<a id="__codelineno-0-306" name="__codelineno-0-306" href="#__codelineno-0-306"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span> <span class="o">=</span> <span class="n">tf</span>
<a id="__codelineno-0-307" name="__codelineno-0-307" href="#__codelineno-0-307"></a>        <span class="k">if</span> <span class="n">normalize_loadings</span><span class="p">:</span>
<a id="__codelineno-0-308" name="__codelineno-0-308" href="#__codelineno-0-308"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">norm_tl_object</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">cp_tensor</span><span class="o">.</span><span class="n">cp_normalize</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span><span class="p">)</span>
<a id="__codelineno-0-309" name="__codelineno-0-309" href="#__codelineno-0-309"></a>
<a id="__codelineno-0-310" name="__codelineno-0-310" href="#__codelineno-0-310"></a>        <span class="n">factor_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Factor </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">rank</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-311" name="__codelineno-0-311" href="#__codelineno-0-311"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-312" name="__codelineno-0-312" href="#__codelineno-0-312"></a>            <span class="k">if</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-313" name="__codelineno-0-313" href="#__codelineno-0-313"></a>                <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts&#39;</span><span class="p">,</span> <span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-314" name="__codelineno-0-314" href="#__codelineno-0-314"></a>            <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-315" name="__codelineno-0-315" href="#__codelineno-0-315"></a>                <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">tensor_dim</span><span class="o">-</span><span class="mi">3</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-316" name="__codelineno-0-316" href="#__codelineno-0-316"></a>            <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
<a id="__codelineno-0-317" name="__codelineno-0-317" href="#__codelineno-0-317"></a>                <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-318" name="__codelineno-0-318" href="#__codelineno-0-318"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-319" name="__codelineno-0-319" href="#__codelineno-0-319"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Too few dimensions in the tensor&#39;</span><span class="p">)</span>
<a id="__codelineno-0-320" name="__codelineno-0-320" href="#__codelineno-0-320"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-321" name="__codelineno-0-321" href="#__codelineno-0-321"></a>            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">tensor_dim</span><span class="p">,</span> <span class="s2">&quot;The length of order_labels must match the number of orders/dimensions in the tensor&quot;</span>
<a id="__codelineno-0-322" name="__codelineno-0-322" href="#__codelineno-0-322"></a>            <span class="n">order_labels</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span>
<a id="__codelineno-0-323" name="__codelineno-0-323" href="#__codelineno-0-323"></a>
<a id="__codelineno-0-324" name="__codelineno-0-324" href="#__codelineno-0-324"></a>        <span class="k">if</span> <span class="n">normalize_loadings</span><span class="p">:</span>
<a id="__codelineno-0-325" name="__codelineno-0-325" href="#__codelineno-0-325"></a>            <span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">norm_tl_object</span>
<a id="__codelineno-0-326" name="__codelineno-0-326" href="#__codelineno-0-326"></a>            <span class="n">weights</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-327" name="__codelineno-0-327" href="#__codelineno-0-327"></a>            <span class="k">if</span> <span class="n">var_ordered_factors</span><span class="p">:</span>
<a id="__codelineno-0-328" name="__codelineno-0-328" href="#__codelineno-0-328"></a>                <span class="n">w_order</span> <span class="o">=</span> <span class="n">weights</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-329" name="__codelineno-0-329" href="#__codelineno-0-329"></a>                <span class="n">factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">)[:,</span> <span class="n">w_order</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">]</span>
<a id="__codelineno-0-330" name="__codelineno-0-330" href="#__codelineno-0-330"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">w_order</span><span class="p">]</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-331" name="__codelineno-0-331" href="#__codelineno-0-331"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-332" name="__codelineno-0-332" href="#__codelineno-0-332"></a>                <span class="n">factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">]</span>
<a id="__codelineno-0-333" name="__codelineno-0-333" href="#__codelineno-0-333"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">weights</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-334" name="__codelineno-0-334" href="#__codelineno-0-334"></a>
<a id="__codelineno-0-335" name="__codelineno-0-335" href="#__codelineno-0-335"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-336" name="__codelineno-0-336" href="#__codelineno-0-336"></a>            <span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span>
<a id="__codelineno-0-337" name="__codelineno-0-337" href="#__codelineno-0-337"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-338" name="__codelineno-0-338" href="#__codelineno-0-338"></a>
<a id="__codelineno-0-339" name="__codelineno-0-339" href="#__codelineno-0-339"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance</span><span class="p">()</span>
<a id="__codelineno-0-340" name="__codelineno-0-340" href="#__codelineno-0-340"></a>
<a id="__codelineno-0-341" name="__codelineno-0-341" href="#__codelineno-0-341"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">factors</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">order_labels</span><span class="p">,</span>
<a id="__codelineno-0-342" name="__codelineno-0-342" href="#__codelineno-0-342"></a>                                       <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">factor_names</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_names</span><span class="p">)]))</span>
<a id="__codelineno-0-343" name="__codelineno-0-343" href="#__codelineno-0-343"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="n">rank</span>
<a id="__codelineno-0-344" name="__codelineno-0-344" href="#__codelineno-0-344"></a>
<a id="__codelineno-0-345" name="__codelineno-0-345" href="#__codelineno-0-345"></a>    <span class="k">def</span> <span class="nf">elbow_rank_selection</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">upper_rank</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">runs</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;random&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span>
<a id="__codelineno-0-346" name="__codelineno-0-346" href="#__codelineno-0-346"></a>                             <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;error&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">10e-7</span><span class="p">,</span> <span class="n">automatic_elbow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-347" name="__codelineno-0-347" href="#__codelineno-0-347"></a>                             <span class="n">manual_elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ci</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span>
<a id="__codelineno-0-348" name="__codelineno-0-348" href="#__codelineno-0-348"></a>                             <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_fig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-349" name="__codelineno-0-349" href="#__codelineno-0-349"></a>        <span class="sd">&#39;&#39;&#39;Elbow analysis on the error achieved by the Tensor Factorization for</span>
<a id="__codelineno-0-350" name="__codelineno-0-350" href="#__codelineno-0-350"></a><span class="sd">        selecting the number of factors to use. A plot is made with the results.</span>
<a id="__codelineno-0-351" name="__codelineno-0-351" href="#__codelineno-0-351"></a>
<a id="__codelineno-0-352" name="__codelineno-0-352" href="#__codelineno-0-352"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-353" name="__codelineno-0-353" href="#__codelineno-0-353"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-354" name="__codelineno-0-354" href="#__codelineno-0-354"></a><span class="sd">        upper_rank : int, default=50</span>
<a id="__codelineno-0-355" name="__codelineno-0-355" href="#__codelineno-0-355"></a><span class="sd">            Upper bound of ranks to explore with the elbow analysis.</span>
<a id="__codelineno-0-356" name="__codelineno-0-356" href="#__codelineno-0-356"></a>
<a id="__codelineno-0-357" name="__codelineno-0-357" href="#__codelineno-0-357"></a><span class="sd">        runs : int, default=20</span>
<a id="__codelineno-0-358" name="__codelineno-0-358" href="#__codelineno-0-358"></a><span class="sd">            Number of tensor factorization performed for a given rank. Each</span>
<a id="__codelineno-0-359" name="__codelineno-0-359" href="#__codelineno-0-359"></a><span class="sd">            factorization varies in the seed of initialization.</span>
<a id="__codelineno-0-360" name="__codelineno-0-360" href="#__codelineno-0-360"></a>
<a id="__codelineno-0-361" name="__codelineno-0-361" href="#__codelineno-0-361"></a><span class="sd">        tf_type : str, default=&#39;non_negative_cp&#39;</span>
<a id="__codelineno-0-362" name="__codelineno-0-362" href="#__codelineno-0-362"></a><span class="sd">            Type of Tensor Factorization.</span>
<a id="__codelineno-0-363" name="__codelineno-0-363" href="#__codelineno-0-363"></a>
<a id="__codelineno-0-364" name="__codelineno-0-364" href="#__codelineno-0-364"></a><span class="sd">            - &#39;non_negative_cp&#39; : Non-negative PARAFAC through the traditional ALS.</span>
<a id="__codelineno-0-365" name="__codelineno-0-365" href="#__codelineno-0-365"></a><span class="sd">            - &#39;non_negative_cp_hals&#39; : Non-negative PARAFAC through the Hierarchical ALS.</span>
<a id="__codelineno-0-366" name="__codelineno-0-366" href="#__codelineno-0-366"></a><span class="sd">                                       It reaches an optimal solution faster than the</span>
<a id="__codelineno-0-367" name="__codelineno-0-367" href="#__codelineno-0-367"></a><span class="sd">                                       traditional ALS, but it does not allow a mask.</span>
<a id="__codelineno-0-368" name="__codelineno-0-368" href="#__codelineno-0-368"></a><span class="sd">            - &#39;parafac&#39; : PARAFAC through the traditional ALS. It allows negative loadings.</span>
<a id="__codelineno-0-369" name="__codelineno-0-369" href="#__codelineno-0-369"></a><span class="sd">            - &#39;constrained_parafac&#39; : PARAFAC through the traditional ALS. It allows</span>
<a id="__codelineno-0-370" name="__codelineno-0-370" href="#__codelineno-0-370"></a><span class="sd">                                      negative loadings. Also, it incorporates L1 and L2</span>
<a id="__codelineno-0-371" name="__codelineno-0-371" href="#__codelineno-0-371"></a><span class="sd">                                      regularization, includes a &#39;non_negative&#39; option, and</span>
<a id="__codelineno-0-372" name="__codelineno-0-372" href="#__codelineno-0-372"></a><span class="sd">                                      allows constraining the sparsity of the decomposition.</span>
<a id="__codelineno-0-373" name="__codelineno-0-373" href="#__codelineno-0-373"></a><span class="sd">                                      For more information, see</span>
<a id="__codelineno-0-374" name="__codelineno-0-374" href="#__codelineno-0-374"></a><span class="sd">                                      http://tensorly.org/stable/modules/generated/tensorly.decomposition.constrained_parafac.html#tensorly.decomposition.constrained_parafac</span>
<a id="__codelineno-0-375" name="__codelineno-0-375" href="#__codelineno-0-375"></a>
<a id="__codelineno-0-376" name="__codelineno-0-376" href="#__codelineno-0-376"></a><span class="sd">        init : str, default=&#39;svd&#39;</span>
<a id="__codelineno-0-377" name="__codelineno-0-377" href="#__codelineno-0-377"></a><span class="sd">            Initialization method for computing the Tensor Factorization.</span>
<a id="__codelineno-0-378" name="__codelineno-0-378" href="#__codelineno-0-378"></a><span class="sd">            {svd, random}</span>
<a id="__codelineno-0-379" name="__codelineno-0-379" href="#__codelineno-0-379"></a>
<a id="__codelineno-0-380" name="__codelineno-0-380" href="#__codelineno-0-380"></a><span class="sd">        svd : str, default=&#39;numpy_svd&#39;</span>
<a id="__codelineno-0-381" name="__codelineno-0-381" href="#__codelineno-0-381"></a><span class="sd">            Function to compute the SVD, acceptable values in tensorly.SVD_FUNS</span>
<a id="__codelineno-0-382" name="__codelineno-0-382" href="#__codelineno-0-382"></a>
<a id="__codelineno-0-383" name="__codelineno-0-383" href="#__codelineno-0-383"></a><span class="sd">        metric : str, default=&#39;error&#39;</span>
<a id="__codelineno-0-384" name="__codelineno-0-384" href="#__codelineno-0-384"></a><span class="sd">            Metric to perform the elbow analysis (y-axis)</span>
<a id="__codelineno-0-385" name="__codelineno-0-385" href="#__codelineno-0-385"></a>
<a id="__codelineno-0-386" name="__codelineno-0-386" href="#__codelineno-0-386"></a><span class="sd">            - &#39;error&#39; : Normalized error to compute the elbow.</span>
<a id="__codelineno-0-387" name="__codelineno-0-387" href="#__codelineno-0-387"></a><span class="sd">            - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).</span>
<a id="__codelineno-0-388" name="__codelineno-0-388" href="#__codelineno-0-388"></a>
<a id="__codelineno-0-389" name="__codelineno-0-389" href="#__codelineno-0-389"></a><span class="sd">        random_state : int, default=None</span>
<a id="__codelineno-0-390" name="__codelineno-0-390" href="#__codelineno-0-390"></a><span class="sd">            Seed for randomization.</span>
<a id="__codelineno-0-391" name="__codelineno-0-391" href="#__codelineno-0-391"></a>
<a id="__codelineno-0-392" name="__codelineno-0-392" href="#__codelineno-0-392"></a><span class="sd">        tol : float, default=10e-7</span>
<a id="__codelineno-0-393" name="__codelineno-0-393" href="#__codelineno-0-393"></a><span class="sd">            Tolerance for the decomposition algorithm to stop when the variation in</span>
<a id="__codelineno-0-394" name="__codelineno-0-394" href="#__codelineno-0-394"></a><span class="sd">            the reconstruction error is less than the tolerance. Lower `tol` helps</span>
<a id="__codelineno-0-395" name="__codelineno-0-395" href="#__codelineno-0-395"></a><span class="sd">            to improve the solution obtained from the decomposition, but it takes</span>
<a id="__codelineno-0-396" name="__codelineno-0-396" href="#__codelineno-0-396"></a><span class="sd">            longer to run.</span>
<a id="__codelineno-0-397" name="__codelineno-0-397" href="#__codelineno-0-397"></a>
<a id="__codelineno-0-398" name="__codelineno-0-398" href="#__codelineno-0-398"></a><span class="sd">        n_iter_max : int, default=100</span>
<a id="__codelineno-0-399" name="__codelineno-0-399" href="#__codelineno-0-399"></a><span class="sd">            Maximum number of iteration to reach an optimal solution with the</span>
<a id="__codelineno-0-400" name="__codelineno-0-400" href="#__codelineno-0-400"></a><span class="sd">            decomposition algorithm. Higher `n_iter_max`helps to improve the solution</span>
<a id="__codelineno-0-401" name="__codelineno-0-401" href="#__codelineno-0-401"></a><span class="sd">            obtained from the decomposition, but it takes longer to run.</span>
<a id="__codelineno-0-402" name="__codelineno-0-402" href="#__codelineno-0-402"></a>
<a id="__codelineno-0-403" name="__codelineno-0-403" href="#__codelineno-0-403"></a><span class="sd">        automatic_elbow : boolean, default=True</span>
<a id="__codelineno-0-404" name="__codelineno-0-404" href="#__codelineno-0-404"></a><span class="sd">            Whether using an automatic strategy to find the elbow. If True, the method</span>
<a id="__codelineno-0-405" name="__codelineno-0-405" href="#__codelineno-0-405"></a><span class="sd">            implemented by the package kneed is used.</span>
<a id="__codelineno-0-406" name="__codelineno-0-406" href="#__codelineno-0-406"></a>
<a id="__codelineno-0-407" name="__codelineno-0-407" href="#__codelineno-0-407"></a><span class="sd">        manual_elbow : int, default=None</span>
<a id="__codelineno-0-408" name="__codelineno-0-408" href="#__codelineno-0-408"></a><span class="sd">            Rank or number of factors to highlight in the curve of error achieved by</span>
<a id="__codelineno-0-409" name="__codelineno-0-409" href="#__codelineno-0-409"></a><span class="sd">            the Tensor Factorization. This input is considered only when</span>
<a id="__codelineno-0-410" name="__codelineno-0-410" href="#__codelineno-0-410"></a><span class="sd">            `automatic_elbow=True`</span>
<a id="__codelineno-0-411" name="__codelineno-0-411" href="#__codelineno-0-411"></a>
<a id="__codelineno-0-412" name="__codelineno-0-412" href="#__codelineno-0-412"></a><span class="sd">        smooth : boolean, default=False</span>
<a id="__codelineno-0-413" name="__codelineno-0-413" href="#__codelineno-0-413"></a><span class="sd">            Whether smoothing the curve with a Savitzky-Golay filter.</span>
<a id="__codelineno-0-414" name="__codelineno-0-414" href="#__codelineno-0-414"></a>
<a id="__codelineno-0-415" name="__codelineno-0-415" href="#__codelineno-0-415"></a><span class="sd">        mask : ndarray list, default=None</span>
<a id="__codelineno-0-416" name="__codelineno-0-416" href="#__codelineno-0-416"></a><span class="sd">            Helps avoiding missing values during a tensor factorization. A mask should be</span>
<a id="__codelineno-0-417" name="__codelineno-0-417" href="#__codelineno-0-417"></a><span class="sd">            a boolean array of the same shape as the original tensor and should be 0</span>
<a id="__codelineno-0-418" name="__codelineno-0-418" href="#__codelineno-0-418"></a><span class="sd">            where the values are missing and 1 everywhere else.</span>
<a id="__codelineno-0-419" name="__codelineno-0-419" href="#__codelineno-0-419"></a>
<a id="__codelineno-0-420" name="__codelineno-0-420" href="#__codelineno-0-420"></a><span class="sd">        ci : str, default=&#39;std&#39;</span>
<a id="__codelineno-0-421" name="__codelineno-0-421" href="#__codelineno-0-421"></a><span class="sd">            Confidence interval for representing the multiple runs in each rank.</span>
<a id="__codelineno-0-422" name="__codelineno-0-422" href="#__codelineno-0-422"></a><span class="sd">            {&#39;std&#39;, &#39;95%&#39;}</span>
<a id="__codelineno-0-423" name="__codelineno-0-423" href="#__codelineno-0-423"></a>
<a id="__codelineno-0-424" name="__codelineno-0-424" href="#__codelineno-0-424"></a><span class="sd">        figsize : tuple, default=(4, 2.25)</span>
<a id="__codelineno-0-425" name="__codelineno-0-425" href="#__codelineno-0-425"></a><span class="sd">            Figure size, width by height</span>
<a id="__codelineno-0-426" name="__codelineno-0-426" href="#__codelineno-0-426"></a>
<a id="__codelineno-0-427" name="__codelineno-0-427" href="#__codelineno-0-427"></a><span class="sd">        fontsize : int, default=14</span>
<a id="__codelineno-0-428" name="__codelineno-0-428" href="#__codelineno-0-428"></a><span class="sd">            Fontsize for axis labels.</span>
<a id="__codelineno-0-429" name="__codelineno-0-429" href="#__codelineno-0-429"></a>
<a id="__codelineno-0-430" name="__codelineno-0-430" href="#__codelineno-0-430"></a><span class="sd">        filename : str, default=None</span>
<a id="__codelineno-0-431" name="__codelineno-0-431" href="#__codelineno-0-431"></a><span class="sd">            Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-432" name="__codelineno-0-432" href="#__codelineno-0-432"></a><span class="sd">            saved.</span>
<a id="__codelineno-0-433" name="__codelineno-0-433" href="#__codelineno-0-433"></a>
<a id="__codelineno-0-434" name="__codelineno-0-434" href="#__codelineno-0-434"></a><span class="sd">        output_fig : boolean, default=True</span>
<a id="__codelineno-0-435" name="__codelineno-0-435" href="#__codelineno-0-435"></a><span class="sd">            Whether generating the figure with matplotlib.</span>
<a id="__codelineno-0-436" name="__codelineno-0-436" href="#__codelineno-0-436"></a>
<a id="__codelineno-0-437" name="__codelineno-0-437" href="#__codelineno-0-437"></a><span class="sd">        verbose : boolean, default=False</span>
<a id="__codelineno-0-438" name="__codelineno-0-438" href="#__codelineno-0-438"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-439" name="__codelineno-0-439" href="#__codelineno-0-439"></a>
<a id="__codelineno-0-440" name="__codelineno-0-440" href="#__codelineno-0-440"></a><span class="sd">        **kwargs : dict</span>
<a id="__codelineno-0-441" name="__codelineno-0-441" href="#__codelineno-0-441"></a><span class="sd">            Extra arguments for the tensor factorization according to inputs in</span>
<a id="__codelineno-0-442" name="__codelineno-0-442" href="#__codelineno-0-442"></a><span class="sd">            tensorly.</span>
<a id="__codelineno-0-443" name="__codelineno-0-443" href="#__codelineno-0-443"></a>
<a id="__codelineno-0-444" name="__codelineno-0-444" href="#__codelineno-0-444"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-445" name="__codelineno-0-445" href="#__codelineno-0-445"></a><span class="sd">        -------</span>
<a id="__codelineno-0-446" name="__codelineno-0-446" href="#__codelineno-0-446"></a><span class="sd">        fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-447" name="__codelineno-0-447" href="#__codelineno-0-447"></a><span class="sd">            Figure object made with matplotlib</span>
<a id="__codelineno-0-448" name="__codelineno-0-448" href="#__codelineno-0-448"></a>
<a id="__codelineno-0-449" name="__codelineno-0-449" href="#__codelineno-0-449"></a><span class="sd">        loss : list</span>
<a id="__codelineno-0-450" name="__codelineno-0-450" href="#__codelineno-0-450"></a><span class="sd">            List of normalized errors for each rank. Here the errors are te average</span>
<a id="__codelineno-0-451" name="__codelineno-0-451" href="#__codelineno-0-451"></a><span class="sd">            across distinct runs for each rank.</span>
<a id="__codelineno-0-452" name="__codelineno-0-452" href="#__codelineno-0-452"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-453" name="__codelineno-0-453" href="#__codelineno-0-453"></a>        <span class="k">assert</span> <span class="n">metric</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;similarity&#39;</span><span class="p">,</span> <span class="s1">&#39;error&#39;</span><span class="p">],</span> <span class="s2">&quot;`metric` must be either &#39;similarity&#39; or &#39;error&#39;&quot;</span>
<a id="__codelineno-0-454" name="__codelineno-0-454" href="#__codelineno-0-454"></a>        <span class="n">ylabel</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;similarity&#39;</span> <span class="p">:</span> <span class="s1">&#39;Similarity</span><span class="se">\n</span><span class="s1">(1-CorrIndex)&#39;</span><span class="p">,</span> <span class="s1">&#39;error&#39;</span> <span class="p">:</span> <span class="s1">&#39;Normalized Error&#39;</span><span class="p">}</span>
<a id="__codelineno-0-455" name="__codelineno-0-455" href="#__codelineno-0-455"></a>
<a id="__codelineno-0-456" name="__codelineno-0-456" href="#__codelineno-0-456"></a>        <span class="c1"># Run analysis</span>
<a id="__codelineno-0-457" name="__codelineno-0-457" href="#__codelineno-0-457"></a>        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-458" name="__codelineno-0-458" href="#__codelineno-0-458"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running Elbow Analysis&#39;</span><span class="p">)</span>
<a id="__codelineno-0-459" name="__codelineno-0-459" href="#__codelineno-0-459"></a>
<a id="__codelineno-0-460" name="__codelineno-0-460" href="#__codelineno-0-460"></a>        <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-461" name="__codelineno-0-461" href="#__codelineno-0-461"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-462" name="__codelineno-0-462" href="#__codelineno-0-462"></a>                <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span>
<a id="__codelineno-0-463" name="__codelineno-0-463" href="#__codelineno-0-463"></a>
<a id="__codelineno-0-464" name="__codelineno-0-464" href="#__codelineno-0-464"></a>        <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;similarity&#39;</span><span class="p">:</span>
<a id="__codelineno-0-465" name="__codelineno-0-465" href="#__codelineno-0-465"></a>            <span class="k">assert</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;`runs` must be greater than 1 when `metric` = &#39;similarity&#39;&quot;</span>
<a id="__codelineno-0-466" name="__codelineno-0-466" href="#__codelineno-0-466"></a>        <span class="k">if</span> <span class="n">runs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-467" name="__codelineno-0-467" href="#__codelineno-0-467"></a>            <span class="n">loss</span> <span class="o">=</span> <span class="n">_run_elbow_analysis</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-468" name="__codelineno-0-468" href="#__codelineno-0-468"></a>                                       <span class="n">upper_rank</span><span class="o">=</span><span class="n">upper_rank</span><span class="p">,</span>
<a id="__codelineno-0-469" name="__codelineno-0-469" href="#__codelineno-0-469"></a>                                       <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-470" name="__codelineno-0-470" href="#__codelineno-0-470"></a>                                       <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-471" name="__codelineno-0-471" href="#__codelineno-0-471"></a>                                       <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-472" name="__codelineno-0-472" href="#__codelineno-0-472"></a>                                       <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-473" name="__codelineno-0-473" href="#__codelineno-0-473"></a>                                       <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-474" name="__codelineno-0-474" href="#__codelineno-0-474"></a>                                       <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-475" name="__codelineno-0-475" href="#__codelineno-0-475"></a>                                       <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-476" name="__codelineno-0-476" href="#__codelineno-0-476"></a>                                       <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-477" name="__codelineno-0-477" href="#__codelineno-0-477"></a>                                       <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-478" name="__codelineno-0-478" href="#__codelineno-0-478"></a>                                       <span class="p">)</span>
<a id="__codelineno-0-479" name="__codelineno-0-479" href="#__codelineno-0-479"></a>            <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">())</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]</span>
<a id="__codelineno-0-480" name="__codelineno-0-480" href="#__codelineno-0-480"></a>            <span class="n">all_loss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]])</span>
<a id="__codelineno-0-481" name="__codelineno-0-481" href="#__codelineno-0-481"></a>            <span class="k">if</span> <span class="n">automatic_elbow</span><span class="p">:</span>
<a id="__codelineno-0-482" name="__codelineno-0-482" href="#__codelineno-0-482"></a>                <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
<a id="__codelineno-0-483" name="__codelineno-0-483" href="#__codelineno-0-483"></a>                    <span class="n">loss_</span> <span class="o">=</span> <span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]</span>
<a id="__codelineno-0-484" name="__codelineno-0-484" href="#__codelineno-0-484"></a>                    <span class="n">loss</span> <span class="o">=</span> <span class="n">smooth_curve</span><span class="p">(</span><span class="n">loss_</span><span class="p">)</span>
<a id="__codelineno-0-485" name="__codelineno-0-485" href="#__codelineno-0-485"></a>                    <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">loss</span><span class="p">)]</span>
<a id="__codelineno-0-486" name="__codelineno-0-486" href="#__codelineno-0-486"></a>                <span class="n">rank</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">_compute_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">))</span>
<a id="__codelineno-0-487" name="__codelineno-0-487" href="#__codelineno-0-487"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-488" name="__codelineno-0-488" href="#__codelineno-0-488"></a>                <span class="n">rank</span> <span class="o">=</span> <span class="n">manual_elbow</span>
<a id="__codelineno-0-489" name="__codelineno-0-489" href="#__codelineno-0-489"></a>            <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
<a id="__codelineno-0-490" name="__codelineno-0-490" href="#__codelineno-0-490"></a>                <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_elbow</span><span class="p">(</span><span class="n">loss</span><span class="o">=</span><span class="n">loss</span><span class="p">,</span>
<a id="__codelineno-0-491" name="__codelineno-0-491" href="#__codelineno-0-491"></a>                                 <span class="n">elbow</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-492" name="__codelineno-0-492" href="#__codelineno-0-492"></a>                                 <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-493" name="__codelineno-0-493" href="#__codelineno-0-493"></a>                                 <span class="n">ylabel</span><span class="o">=</span><span class="n">ylabel</span><span class="p">[</span><span class="n">metric</span><span class="p">],</span>
<a id="__codelineno-0-494" name="__codelineno-0-494" href="#__codelineno-0-494"></a>                                 <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-495" name="__codelineno-0-495" href="#__codelineno-0-495"></a>                                 <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-496" name="__codelineno-0-496" href="#__codelineno-0-496"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-497" name="__codelineno-0-497" href="#__codelineno-0-497"></a>                <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-498" name="__codelineno-0-498" href="#__codelineno-0-498"></a>        <span class="k">elif</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-499" name="__codelineno-0-499" href="#__codelineno-0-499"></a>            <span class="n">all_loss</span> <span class="o">=</span> <span class="n">_multiple_runs_elbow_analysis</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-500" name="__codelineno-0-500" href="#__codelineno-0-500"></a>                                                     <span class="n">upper_rank</span><span class="o">=</span><span class="n">upper_rank</span><span class="p">,</span>
<a id="__codelineno-0-501" name="__codelineno-0-501" href="#__codelineno-0-501"></a>                                                     <span class="n">runs</span><span class="o">=</span><span class="n">runs</span><span class="p">,</span>
<a id="__codelineno-0-502" name="__codelineno-0-502" href="#__codelineno-0-502"></a>                                                     <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-503" name="__codelineno-0-503" href="#__codelineno-0-503"></a>                                                     <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-504" name="__codelineno-0-504" href="#__codelineno-0-504"></a>                                                     <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-505" name="__codelineno-0-505" href="#__codelineno-0-505"></a>                                                     <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
<a id="__codelineno-0-506" name="__codelineno-0-506" href="#__codelineno-0-506"></a>                                                     <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-507" name="__codelineno-0-507" href="#__codelineno-0-507"></a>                                                     <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-508" name="__codelineno-0-508" href="#__codelineno-0-508"></a>                                                     <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-509" name="__codelineno-0-509" href="#__codelineno-0-509"></a>                                                     <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-510" name="__codelineno-0-510" href="#__codelineno-0-510"></a>                                                     <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-511" name="__codelineno-0-511" href="#__codelineno-0-511"></a>                                                     <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-512" name="__codelineno-0-512" href="#__codelineno-0-512"></a>                                                     <span class="p">)</span>
<a id="__codelineno-0-513" name="__codelineno-0-513" href="#__codelineno-0-513"></a>
<a id="__codelineno-0-514" name="__codelineno-0-514" href="#__codelineno-0-514"></a>            <span class="c1"># Same outputs as runs = 1</span>
<a id="__codelineno-0-515" name="__codelineno-0-515" href="#__codelineno-0-515"></a>            <span class="n">loss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-516" name="__codelineno-0-516" href="#__codelineno-0-516"></a>            <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
<a id="__codelineno-0-517" name="__codelineno-0-517" href="#__codelineno-0-517"></a>                <span class="n">loss</span> <span class="o">=</span> <span class="n">smooth_curve</span><span class="p">(</span><span class="n">loss</span><span class="p">)</span>
<a id="__codelineno-0-518" name="__codelineno-0-518" href="#__codelineno-0-518"></a>            <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">loss</span><span class="p">)]</span>
<a id="__codelineno-0-519" name="__codelineno-0-519" href="#__codelineno-0-519"></a>
<a id="__codelineno-0-520" name="__codelineno-0-520" href="#__codelineno-0-520"></a>            <span class="k">if</span> <span class="n">automatic_elbow</span><span class="p">:</span>
<a id="__codelineno-0-521" name="__codelineno-0-521" href="#__codelineno-0-521"></a>                <span class="n">rank</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">_compute_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">))</span>
<a id="__codelineno-0-522" name="__codelineno-0-522" href="#__codelineno-0-522"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-523" name="__codelineno-0-523" href="#__codelineno-0-523"></a>                <span class="n">rank</span> <span class="o">=</span> <span class="n">manual_elbow</span>
<a id="__codelineno-0-524" name="__codelineno-0-524" href="#__codelineno-0-524"></a>
<a id="__codelineno-0-525" name="__codelineno-0-525" href="#__codelineno-0-525"></a>            <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
<a id="__codelineno-0-526" name="__codelineno-0-526" href="#__codelineno-0-526"></a>                <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_multiple_run_elbow</span><span class="p">(</span><span class="n">all_loss</span><span class="o">=</span><span class="n">all_loss</span><span class="p">,</span>
<a id="__codelineno-0-527" name="__codelineno-0-527" href="#__codelineno-0-527"></a>                                              <span class="n">ci</span><span class="o">=</span><span class="n">ci</span><span class="p">,</span>
<a id="__codelineno-0-528" name="__codelineno-0-528" href="#__codelineno-0-528"></a>                                              <span class="n">elbow</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-529" name="__codelineno-0-529" href="#__codelineno-0-529"></a>                                              <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-530" name="__codelineno-0-530" href="#__codelineno-0-530"></a>                                              <span class="n">ylabel</span><span class="o">=</span><span class="n">ylabel</span><span class="p">[</span><span class="n">metric</span><span class="p">],</span>
<a id="__codelineno-0-531" name="__codelineno-0-531" href="#__codelineno-0-531"></a>                                              <span class="n">smooth</span><span class="o">=</span><span class="n">smooth</span><span class="p">,</span>
<a id="__codelineno-0-532" name="__codelineno-0-532" href="#__codelineno-0-532"></a>                                              <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-533" name="__codelineno-0-533" href="#__codelineno-0-533"></a>                                              <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-534" name="__codelineno-0-534" href="#__codelineno-0-534"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-535" name="__codelineno-0-535" href="#__codelineno-0-535"></a>                <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-536" name="__codelineno-0-536" href="#__codelineno-0-536"></a>
<a id="__codelineno-0-537" name="__codelineno-0-537" href="#__codelineno-0-537"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-538" name="__codelineno-0-538" href="#__codelineno-0-538"></a>            <span class="k">assert</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Input runs must be an integer greater than 0&quot;</span>
<a id="__codelineno-0-539" name="__codelineno-0-539" href="#__codelineno-0-539"></a>
<a id="__codelineno-0-540" name="__codelineno-0-540" href="#__codelineno-0-540"></a>        <span class="c1"># Store results</span>
<a id="__codelineno-0-541" name="__codelineno-0-541" href="#__codelineno-0-541"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="n">rank</span>
<a id="__codelineno-0-542" name="__codelineno-0-542" href="#__codelineno-0-542"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric</span> <span class="o">=</span> <span class="n">metric</span>
<a id="__codelineno-0-543" name="__codelineno-0-543" href="#__codelineno-0-543"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_mean</span> <span class="o">=</span> <span class="n">loss</span>
<a id="__codelineno-0-544" name="__codelineno-0-544" href="#__codelineno-0-544"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_raw</span> <span class="o">=</span> <span class="n">all_loss</span>
<a id="__codelineno-0-545" name="__codelineno-0-545" href="#__codelineno-0-545"></a>
<a id="__codelineno-0-546" name="__codelineno-0-546" href="#__codelineno-0-546"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-547" name="__codelineno-0-547" href="#__codelineno-0-547"></a>            <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="nb">int</span><span class="p">)),</span> <span class="s1">&#39;rank must be an integer.&#39;</span>
<a id="__codelineno-0-548" name="__codelineno-0-548" href="#__codelineno-0-548"></a>            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;The rank at the elbow is: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rank</span><span class="p">))</span>
<a id="__codelineno-0-549" name="__codelineno-0-549" href="#__codelineno-0-549"></a>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">loss</span>
<a id="__codelineno-0-550" name="__codelineno-0-550" href="#__codelineno-0-550"></a>
<a id="__codelineno-0-551" name="__codelineno-0-551" href="#__codelineno-0-551"></a>    <span class="k">def</span> <span class="nf">get_top_factor_elements</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">order_name</span><span class="p">,</span> <span class="n">factor_name</span><span class="p">,</span> <span class="n">top_number</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<a id="__codelineno-0-552" name="__codelineno-0-552" href="#__codelineno-0-552"></a>        <span class="sd">&#39;&#39;&#39;Obtains the top-elements with higher loadings for a given factor</span>
<a id="__codelineno-0-553" name="__codelineno-0-553" href="#__codelineno-0-553"></a>
<a id="__codelineno-0-554" name="__codelineno-0-554" href="#__codelineno-0-554"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-555" name="__codelineno-0-555" href="#__codelineno-0-555"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-556" name="__codelineno-0-556" href="#__codelineno-0-556"></a><span class="sd">        order_name : str</span>
<a id="__codelineno-0-557" name="__codelineno-0-557" href="#__codelineno-0-557"></a><span class="sd">            Name of the dimension/order in the tensor according to the keys of the</span>
<a id="__codelineno-0-558" name="__codelineno-0-558" href="#__codelineno-0-558"></a><span class="sd">            dictionary in BaseTensor.factors. The attribute factors is built once the</span>
<a id="__codelineno-0-559" name="__codelineno-0-559" href="#__codelineno-0-559"></a><span class="sd">            tensor factorization is run.</span>
<a id="__codelineno-0-560" name="__codelineno-0-560" href="#__codelineno-0-560"></a>
<a id="__codelineno-0-561" name="__codelineno-0-561" href="#__codelineno-0-561"></a><span class="sd">        factor_name : str</span>
<a id="__codelineno-0-562" name="__codelineno-0-562" href="#__codelineno-0-562"></a><span class="sd">            Name of one of the factors. E.g., &#39;Factor 1&#39;</span>
<a id="__codelineno-0-563" name="__codelineno-0-563" href="#__codelineno-0-563"></a>
<a id="__codelineno-0-564" name="__codelineno-0-564" href="#__codelineno-0-564"></a><span class="sd">        top_number : int, default=10</span>
<a id="__codelineno-0-565" name="__codelineno-0-565" href="#__codelineno-0-565"></a><span class="sd">            Number of top-elements to return</span>
<a id="__codelineno-0-566" name="__codelineno-0-566" href="#__codelineno-0-566"></a>
<a id="__codelineno-0-567" name="__codelineno-0-567" href="#__codelineno-0-567"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-568" name="__codelineno-0-568" href="#__codelineno-0-568"></a><span class="sd">        -------</span>
<a id="__codelineno-0-569" name="__codelineno-0-569" href="#__codelineno-0-569"></a><span class="sd">        top_elements : pandas.DataFrame</span>
<a id="__codelineno-0-570" name="__codelineno-0-570" href="#__codelineno-0-570"></a><span class="sd">            A dataframe with the loadings of the top-elements for the given factor.</span>
<a id="__codelineno-0-571" name="__codelineno-0-571" href="#__codelineno-0-571"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-572" name="__codelineno-0-572" href="#__codelineno-0-572"></a>        <span class="n">top_elements</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="n">order_name</span><span class="p">][</span><span class="n">factor_name</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_number</span><span class="p">)</span>
<a id="__codelineno-0-573" name="__codelineno-0-573" href="#__codelineno-0-573"></a>        <span class="k">return</span> <span class="n">top_elements</span>
<a id="__codelineno-0-574" name="__codelineno-0-574" href="#__codelineno-0-574"></a>
<a id="__codelineno-0-575" name="__codelineno-0-575" href="#__codelineno-0-575"></a>    <span class="k">def</span> <span class="nf">export_factor_loadings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-576" name="__codelineno-0-576" href="#__codelineno-0-576"></a>        <span class="sd">&#39;&#39;&#39;Exports the factor loadings of the tensor into an Excel file</span>
<a id="__codelineno-0-577" name="__codelineno-0-577" href="#__codelineno-0-577"></a>
<a id="__codelineno-0-578" name="__codelineno-0-578" href="#__codelineno-0-578"></a><span class="sd">        Parameters</span>
<a id="__codelineno-0-579" name="__codelineno-0-579" href="#__codelineno-0-579"></a><span class="sd">        ----------</span>
<a id="__codelineno-0-580" name="__codelineno-0-580" href="#__codelineno-0-580"></a><span class="sd">        filename : str</span>
<a id="__codelineno-0-581" name="__codelineno-0-581" href="#__codelineno-0-581"></a><span class="sd">            Full path and filename to store the file. E.g., &#39;/home/user/Loadings.xlsx&#39;</span>
<a id="__codelineno-0-582" name="__codelineno-0-582" href="#__codelineno-0-582"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-583" name="__codelineno-0-583" href="#__codelineno-0-583"></a>        <span class="n">writer</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-584" name="__codelineno-0-584" href="#__codelineno-0-584"></a>        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">factors</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-585" name="__codelineno-0-585" href="#__codelineno-0-585"></a>            <span class="n">v</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="n">k</span><span class="p">)</span>
<a id="__codelineno-0-586" name="__codelineno-0-586" href="#__codelineno-0-586"></a>        <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<a id="__codelineno-0-587" name="__codelineno-0-587" href="#__codelineno-0-587"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Loadings of the tensor factorization were successfully saved into </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
<a id="__codelineno-0-588" name="__codelineno-0-588" href="#__codelineno-0-588"></a>
<a id="__codelineno-0-589" name="__codelineno-0-589" href="#__codelineno-0-589"></a>    <span class="k">def</span> <span class="nf">excluded_value_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-590" name="__codelineno-0-590" href="#__codelineno-0-590"></a>        <span class="sd">&#39;&#39;&#39;Returns the fraction of excluded values in the tensor,</span>
<a id="__codelineno-0-591" name="__codelineno-0-591" href="#__codelineno-0-591"></a><span class="sd">        given the values that are masked in tensor.mask</span>
<a id="__codelineno-0-592" name="__codelineno-0-592" href="#__codelineno-0-592"></a>
<a id="__codelineno-0-593" name="__codelineno-0-593" href="#__codelineno-0-593"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-594" name="__codelineno-0-594" href="#__codelineno-0-594"></a><span class="sd">        -------</span>
<a id="__codelineno-0-595" name="__codelineno-0-595" href="#__codelineno-0-595"></a><span class="sd">        excluded_fraction : float</span>
<a id="__codelineno-0-596" name="__codelineno-0-596" href="#__codelineno-0-596"></a><span class="sd">            Fraction of missing/excluded values in the tensor.</span>
<a id="__codelineno-0-597" name="__codelineno-0-597" href="#__codelineno-0-597"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-598" name="__codelineno-0-598" href="#__codelineno-0-598"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-599" name="__codelineno-0-599" href="#__codelineno-0-599"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have masked values&quot;</span><span class="p">)</span>
<a id="__codelineno-0-600" name="__codelineno-0-600" href="#__codelineno-0-600"></a>            <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-601" name="__codelineno-0-601" href="#__codelineno-0-601"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-602" name="__codelineno-0-602" href="#__codelineno-0-602"></a>            <span class="n">fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-603" name="__codelineno-0-603" href="#__codelineno-0-603"></a>            <span class="n">excluded_fraction</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-604" name="__codelineno-0-604" href="#__codelineno-0-604"></a>            <span class="k">return</span> <span class="n">excluded_fraction</span>
<a id="__codelineno-0-605" name="__codelineno-0-605" href="#__codelineno-0-605"></a>
<a id="__codelineno-0-606" name="__codelineno-0-606" href="#__codelineno-0-606"></a>    <span class="k">def</span> <span class="nf">sparsity_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-607" name="__codelineno-0-607" href="#__codelineno-0-607"></a>        <span class="sd">&#39;&#39;&#39;Returns the fraction of values that are zeros in the tensor,</span>
<a id="__codelineno-0-608" name="__codelineno-0-608" href="#__codelineno-0-608"></a><span class="sd">        given the values that are in tensor.loc_zeros</span>
<a id="__codelineno-0-609" name="__codelineno-0-609" href="#__codelineno-0-609"></a>
<a id="__codelineno-0-610" name="__codelineno-0-610" href="#__codelineno-0-610"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-611" name="__codelineno-0-611" href="#__codelineno-0-611"></a><span class="sd">        -------</span>
<a id="__codelineno-0-612" name="__codelineno-0-612" href="#__codelineno-0-612"></a><span class="sd">        sparsity_fraction : float</span>
<a id="__codelineno-0-613" name="__codelineno-0-613" href="#__codelineno-0-613"></a><span class="sd">            Fraction of values that are real zeros.</span>
<a id="__codelineno-0-614" name="__codelineno-0-614" href="#__codelineno-0-614"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-615" name="__codelineno-0-615" href="#__codelineno-0-615"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-616" name="__codelineno-0-616" href="#__codelineno-0-616"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have zeros&quot;</span><span class="p">)</span>
<a id="__codelineno-0-617" name="__codelineno-0-617" href="#__codelineno-0-617"></a>            <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-618" name="__codelineno-0-618" href="#__codelineno-0-618"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-619" name="__codelineno-0-619" href="#__codelineno-0-619"></a>            <span class="n">sparsity_fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-620" name="__codelineno-0-620" href="#__codelineno-0-620"></a>        <span class="n">sparsity_fraction</span> <span class="o">=</span> <span class="n">sparsity_fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-621" name="__codelineno-0-621" href="#__codelineno-0-621"></a>        <span class="k">return</span> <span class="n">sparsity_fraction</span>
<a id="__codelineno-0-622" name="__codelineno-0-622" href="#__codelineno-0-622"></a>
<a id="__codelineno-0-623" name="__codelineno-0-623" href="#__codelineno-0-623"></a>    <span class="k">def</span> <span class="nf">missing_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-624" name="__codelineno-0-624" href="#__codelineno-0-624"></a>        <span class="sd">&#39;&#39;&#39;Returns the fraction of values that are missing (NaNs) in the tensor,</span>
<a id="__codelineno-0-625" name="__codelineno-0-625" href="#__codelineno-0-625"></a><span class="sd">        given the values that are in tensor.loc_nans</span>
<a id="__codelineno-0-626" name="__codelineno-0-626" href="#__codelineno-0-626"></a>
<a id="__codelineno-0-627" name="__codelineno-0-627" href="#__codelineno-0-627"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-628" name="__codelineno-0-628" href="#__codelineno-0-628"></a><span class="sd">        -------</span>
<a id="__codelineno-0-629" name="__codelineno-0-629" href="#__codelineno-0-629"></a><span class="sd">        missing_fraction : float</span>
<a id="__codelineno-0-630" name="__codelineno-0-630" href="#__codelineno-0-630"></a><span class="sd">            Fraction of values that are real zeros.</span>
<a id="__codelineno-0-631" name="__codelineno-0-631" href="#__codelineno-0-631"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-632" name="__codelineno-0-632" href="#__codelineno-0-632"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-633" name="__codelineno-0-633" href="#__codelineno-0-633"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have missing values&quot;</span><span class="p">)</span>
<a id="__codelineno-0-634" name="__codelineno-0-634" href="#__codelineno-0-634"></a>            <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-635" name="__codelineno-0-635" href="#__codelineno-0-635"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-636" name="__codelineno-0-636" href="#__codelineno-0-636"></a>            <span class="n">missing_fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-637" name="__codelineno-0-637" href="#__codelineno-0-637"></a>        <span class="n">missing_fraction</span> <span class="o">=</span> <span class="n">missing_fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-638" name="__codelineno-0-638" href="#__codelineno-0-638"></a>        <span class="k">return</span> <span class="n">missing_fraction</span>
<a id="__codelineno-0-639" name="__codelineno-0-639" href="#__codelineno-0-639"></a>
<a id="__codelineno-0-640" name="__codelineno-0-640" href="#__codelineno-0-640"></a>    <span class="k">def</span> <span class="nf">explained_variance</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-641" name="__codelineno-0-641" href="#__codelineno-0-641"></a>        <span class="sd">&#39;&#39;&#39;Computes the explained variance score for a tensor decomposition. Inspired on the</span>
<a id="__codelineno-0-642" name="__codelineno-0-642" href="#__codelineno-0-642"></a><span class="sd">        function in sklearn.metrics.explained_variance_score.</span>
<a id="__codelineno-0-643" name="__codelineno-0-643" href="#__codelineno-0-643"></a>
<a id="__codelineno-0-644" name="__codelineno-0-644" href="#__codelineno-0-644"></a><span class="sd">        Returns</span>
<a id="__codelineno-0-645" name="__codelineno-0-645" href="#__codelineno-0-645"></a><span class="sd">        -------</span>
<a id="__codelineno-0-646" name="__codelineno-0-646" href="#__codelineno-0-646"></a><span class="sd">        explained_variance : float</span>
<a id="__codelineno-0-647" name="__codelineno-0-647" href="#__codelineno-0-647"></a><span class="sd">            Explained variance score for a tnesor factorization.</span>
<a id="__codelineno-0-648" name="__codelineno-0-648" href="#__codelineno-0-648"></a><span class="sd">        &#39;&#39;&#39;</span>
<a id="__codelineno-0-649" name="__codelineno-0-649" href="#__codelineno-0-649"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must run compute_tensor_factorization before using this method.&quot;</span>
<a id="__codelineno-0-650" name="__codelineno-0-650" href="#__codelineno-0-650"></a>        <span class="n">tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span>
<a id="__codelineno-0-651" name="__codelineno-0-651" href="#__codelineno-0-651"></a>        <span class="n">rec_tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span><span class="o">.</span><span class="n">to_tensor</span><span class="p">()</span>
<a id="__codelineno-0-652" name="__codelineno-0-652" href="#__codelineno-0-652"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span>
<a id="__codelineno-0-653" name="__codelineno-0-653" href="#__codelineno-0-653"></a>
<a id="__codelineno-0-654" name="__codelineno-0-654" href="#__codelineno-0-654"></a>        <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-655" name="__codelineno-0-655" href="#__codelineno-0-655"></a>            <span class="n">tensor</span> <span class="o">=</span> <span class="n">tensor</span> <span class="o">*</span> <span class="n">mask</span>
<a id="__codelineno-0-656" name="__codelineno-0-656" href="#__codelineno-0-656"></a>            <span class="n">rec_tensor</span> <span class="o">=</span> <span class="n">tensor</span> <span class="o">*</span> <span class="n">mask</span>
<a id="__codelineno-0-657" name="__codelineno-0-657" href="#__codelineno-0-657"></a>
<a id="__codelineno-0-658" name="__codelineno-0-658" href="#__codelineno-0-658"></a>        <span class="n">y_diff_avg</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">rec_tensor</span><span class="p">)</span>
<a id="__codelineno-0-659" name="__codelineno-0-659" href="#__codelineno-0-659"></a>        <span class="n">numerator</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">rec_tensor</span> <span class="o">-</span> <span class="n">y_diff_avg</span><span class="p">)</span>
<a id="__codelineno-0-660" name="__codelineno-0-660" href="#__codelineno-0-660"></a>
<a id="__codelineno-0-661" name="__codelineno-0-661" href="#__codelineno-0-661"></a>        <span class="n">tensor_avg</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-662" name="__codelineno-0-662" href="#__codelineno-0-662"></a>        <span class="n">denominator</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">tensor_avg</span><span class="p">)</span>
<a id="__codelineno-0-663" name="__codelineno-0-663" href="#__codelineno-0-663"></a>
<a id="__codelineno-0-664" name="__codelineno-0-664" href="#__codelineno-0-664"></a>        <span class="k">if</span> <span class="n">denominator</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">:</span>
<a id="__codelineno-0-665" name="__codelineno-0-665" href="#__codelineno-0-665"></a>            <span class="n">explained_variance</span> <span class="o">=</span> <span class="mf">0.0</span>
<a id="__codelineno-0-666" name="__codelineno-0-666" href="#__codelineno-0-666"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-667" name="__codelineno-0-667" href="#__codelineno-0-667"></a>            <span class="n">explained_variance</span> <span class="o">=</span>  <span class="mf">1.</span> <span class="o">-</span> <span class="p">(</span><span class="n">numerator</span> <span class="o">/</span> <span class="n">denominator</span><span class="p">)</span>
<a id="__codelineno-0-668" name="__codelineno-0-668" href="#__codelineno-0-668"></a>            <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">explained_variance</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-669" name="__codelineno-0-669" href="#__codelineno-0-669"></a>        <span class="k">return</span> <span class="n">explained_variance</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">






  <div class="doc doc-object doc-attribute">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.shape">
<code class="highlight language-python"><span class="n">shape</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-property"><code>property</code></small>
      <small class="doc doc-property doc-property-readonly"><code>readonly</code></small>
  </span>

</h5>

    <div class="doc doc-contents ">

      <p>Returns the shape of the tensor</p>
    </div>

  </div>







  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.compute_tensor_factorization">
<code class="highlight language-python"><span class="n">compute_tensor_factorization</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;svd&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">runs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">normalize_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">var_ordered_factors</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Performs a Tensor Factorization.
There are no returns, instead the attributes factors and rank
 of the Tensor class are updated.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.compute_tensor_factorization--parameters">Parameters</h6>
<p>rank : int
    Rank of the Tensor Factorization (number of factors to deconvolve the original
    tensor).</p>
<p>tf_type : str, default='non_negative_cp'
    Type of Tensor Factorization.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;non_negative_cp&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Non</span><span class="o">-</span><span class="n">negative</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;non_negative_cp_hals&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Non</span><span class="o">-</span><span class="n">negative</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">Hierarchical</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"></span>
<span class="w">                           </span><span class="n">It</span><span class="w"> </span><span class="n">reaches</span><span class="w"> </span><span class="n">an</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">solution</span><span class="w"> </span><span class="n">faster</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">                           </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="p">,</span><span class="w"> </span><span class="n">but</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">does</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">allow</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">mask</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;parafac&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">negative</span><span class="w"> </span><span class="n">loadings</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constrained_parafac&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">allows</span><span class="w"></span>
<span class="w">                          </span><span class="n">negative</span><span class="w"> </span><span class="n">loadings</span><span class="o">.</span><span class="w"> </span><span class="n">Also</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">incorporates</span><span class="w"> </span><span class="n">L1</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">L2</span><span class="w"></span>
<span class="w">                          </span><span class="n">regularization</span><span class="p">,</span><span class="w"> </span><span class="n">includes</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="s1">&#39;non_negative&#39;</span><span class="w"> </span><span class="n">option</span><span class="p">,</span><span class="w"> </span><span class="ow">and</span><span class="w"></span>
<span class="w">                          </span><span class="n">allows</span><span class="w"> </span><span class="n">constraining</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">sparsity</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">decomposition</span><span class="o">.</span><span class="w"></span>
<span class="w">                          </span><span class="n">For</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">information</span><span class="p">,</span><span class="w"> </span><span class="n">see</span><span class="w"></span>
<span class="w">                          </span><span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">tensorly</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">stable</span><span class="o">/</span><span class="n">modules</span><span class="o">/</span><span class="n">generated</span><span class="o">/</span><span class="n">tensorly</span><span class="o">.</span><span class="n">decomposition</span><span class="o">.</span><span class="n">constrained_parafac</span><span class="o">.</span><span class="n">html</span><span class="c1">#tensorly.decomposition.constrained_parafac</span><span class="w"></span>
</code></pre></div>

<p _random_="random" _svd_="svd,">init : str, default='svd'
    Initialization method for computing the Tensor Factorization.</p>
<p>svd : str, default='numpy_svd'
    Function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS</p>
<p>random_state : int, default=None
    Seed for randomization.</p>
<p>runs : int, default=1
    Number of models to choose among and find the lowest error.
    This helps to avoid local minima when using runs &gt; 1.</p>
<p>normalize_loadings : boolean, default=True
    Whether normalizing the loadings in each factor to unit
    Euclidean length.</p>
<p>var_ordered_factors : boolean, default=True
    Whether ordering factors by the variance they explain. The order is from
    highest to lowest variance. <code>normalize_loadings</code> must be True. Otherwise,
    this parameter is ignored.</p>
<p>tol : float, default=10e-7
    Tolerance for the decomposition algorithm to stop when the variation in
    the reconstruction error is less than the tolerance. Lower <code>tol</code> helps
    to improve the solution obtained from the decomposition, but it takes
    longer to run.</p>
<p>n_iter_max : int, default=100
    Maximum number of iteration to reach an optimal solution with the
    decomposition algorithm. Higher <code>n_iter_max</code>helps to improve the solution
    obtained from the decomposition, but it takes longer to run.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for the tensor factorization according to inputs in tensorly.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">compute_tensor_factorization</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;svd&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                 <span class="n">runs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">normalize_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">var_ordered_factors</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">10e-7</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                                 <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Performs a Tensor Factorization.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    There are no returns, instead the attributes factors and rank</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">     of the Tensor class are updated.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    rank : int</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Rank of the Tensor Factorization (number of factors to deconvolve the original</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        tensor).</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    tf_type : str, default=&#39;non_negative_cp&#39;</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Type of Tensor Factorization.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        - &#39;non_negative_cp&#39; : Non-negative PARAFAC through the traditional ALS.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        - &#39;non_negative_cp_hals&#39; : Non-negative PARAFAC through the Hierarchical ALS.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">                                   It reaches an optimal solution faster than the</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">                                   traditional ALS, but it does not allow a mask.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;parafac&#39; : PARAFAC through the traditional ALS. It allows negative loadings.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        - &#39;constrained_parafac&#39; : PARAFAC through the traditional ALS. It allows</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                                  negative loadings. Also, it incorporates L1 and L2</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                                  regularization, includes a &#39;non_negative&#39; option, and</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">                                  allows constraining the sparsity of the decomposition.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                  For more information, see</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                                  http://tensorly.org/stable/modules/generated/tensorly.decomposition.constrained_parafac.html#tensorly.decomposition.constrained_parafac</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    init : str, default=&#39;svd&#39;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Initialization method for computing the Tensor Factorization.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        {svd, random}</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    svd : str, default=&#39;numpy_svd&#39;</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    runs : int, default=1</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Number of models to choose among and find the lowest error.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        This helps to avoid local minima when using runs &gt; 1.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">    normalize_loadings : boolean, default=True</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Whether normalizing the loadings in each factor to unit</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Euclidean length.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    var_ordered_factors : boolean, default=True</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Whether ordering factors by the variance they explain. The order is from</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        highest to lowest variance. `normalize_loadings` must be True. Otherwise,</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        this parameter is ignored.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">    tol : float, default=10e-7</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        Tolerance for the decomposition algorithm to stop when the variation in</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        the reconstruction error is less than the tolerance. Lower `tol` helps</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        to improve the solution obtained from the decomposition, but it takes</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        longer to run.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    n_iter_max : int, default=100</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Maximum number of iteration to reach an optimal solution with the</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        decomposition algorithm. Higher `n_iter_max`helps to improve the solution</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        obtained from the decomposition, but it takes longer to run.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Extra arguments for the tensor factorization according to inputs in tensorly.</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="n">tensor_dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">best_err</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">tf</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="k">if</span> <span class="n">kwargs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;return_errors&#39;</span> <span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;return_errors&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>    <span class="k">for</span> <span class="n">run</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">runs</span><span class="p">),</span> <span class="n">disable</span><span class="o">=</span><span class="p">(</span><span class="n">runs</span><span class="o">==</span><span class="mi">1</span><span class="p">)):</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="k">if</span> <span class="n">random_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>            <span class="n">random_state_</span> <span class="o">=</span> <span class="n">random_state</span> <span class="o">+</span> <span class="n">run</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>            <span class="n">random_state_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="n">local_tf</span><span class="p">,</span> <span class="n">errors</span> <span class="o">=</span> <span class="n">_compute_tensor_factorization</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>                                                         <span class="n">rank</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>                                                         <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                                                         <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>                                                         <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>                                                         <span class="n">random_state</span><span class="o">=</span><span class="n">random_state_</span><span class="p">,</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>                                                         <span class="n">mask</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>                                                         <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>                                                         <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>                                                         <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>                                                         <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>        <span class="c1"># This helps to obtain proper error when the mask is not None.</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>            <span class="n">err</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">errors</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>            <span class="k">if</span> <span class="n">best_err</span> <span class="o">&gt;</span> <span class="n">err</span><span class="p">:</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                <span class="n">best_err</span> <span class="o">=</span> <span class="n">err</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                <span class="n">tf</span> <span class="o">=</span> <span class="n">local_tf</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>            <span class="n">err</span> <span class="o">=</span> <span class="n">_compute_norm_error</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="n">local_tf</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a>            <span class="k">if</span> <span class="n">best_err</span> <span class="o">&gt;</span> <span class="n">err</span><span class="p">:</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>                <span class="n">best_err</span> <span class="o">=</span> <span class="n">err</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>                <span class="n">tf</span> <span class="o">=</span> <span class="n">local_tf</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>    <span class="k">if</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Best model has a normalized error of: </span><span class="si">{0:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_err</span><span class="p">))</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span> <span class="o">=</span> <span class="n">tf</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="k">if</span> <span class="n">normalize_loadings</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">norm_tl_object</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">cp_tensor</span><span class="o">.</span><span class="n">cp_normalize</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span><span class="p">)</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="n">factor_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Factor </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">rank</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="k">if</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>            <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts&#39;</span><span class="p">,</span> <span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>            <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">tensor_dim</span><span class="o">-</span><span class="mi">3</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>            <span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Too few dimensions in the tensor&#39;</span><span class="p">)</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">tensor_dim</span><span class="p">,</span> <span class="s2">&quot;The length of order_labels must match the number of orders/dimensions in the tensor&quot;</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>        <span class="n">order_labels</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>    <span class="k">if</span> <span class="n">normalize_loadings</span><span class="p">:</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>        <span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">norm_tl_object</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="n">weights</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="k">if</span> <span class="n">var_ordered_factors</span><span class="p">:</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>            <span class="n">w_order</span> <span class="o">=</span> <span class="n">weights</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>            <span class="n">factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">)[:,</span> <span class="n">w_order</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">]</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[</span><span class="n">w_order</span><span class="p">]</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>            <span class="n">factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">]</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">weights</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance</span><span class="p">()</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">factors</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">order_labels</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                                   <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">f</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">factor_names</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">factors</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">order_names</span><span class="p">)]))</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="n">rank</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.copy">
<code class="highlight language-python"><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Performs a deep copy of this object.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Performs a deep copy of this object.&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="kn">import</span> <span class="nn">copy</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="k">return</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.elbow_rank_selection">
<code class="highlight language-python"><span class="n">elbow_rank_selection</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">upper_rank</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">runs</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;random&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;error&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">,</span> <span class="n">automatic_elbow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">manual_elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ci</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_fig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Elbow analysis on the error achieved by the Tensor Factorization for
selecting the number of factors to use. A plot is made with the results.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.elbow_rank_selection--parameters">Parameters</h6>
<p>upper_rank : int, default=50
    Upper bound of ranks to explore with the elbow analysis.</p>
<p>runs : int, default=20
    Number of tensor factorization performed for a given rank. Each
    factorization varies in the seed of initialization.</p>
<p>tf_type : str, default='non_negative_cp'
    Type of Tensor Factorization.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span><span class="s1">&#39;non_negative_cp&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Non</span><span class="o">-</span><span class="n">negative</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;non_negative_cp_hals&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">Non</span><span class="o">-</span><span class="n">negative</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">Hierarchical</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"></span>
<span class="w">                           </span><span class="n">It</span><span class="w"> </span><span class="n">reaches</span><span class="w"> </span><span class="n">an</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">solution</span><span class="w"> </span><span class="n">faster</span><span class="w"> </span><span class="n">than</span><span class="w"> </span><span class="n">the</span><span class="w"></span>
<span class="w">                           </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="p">,</span><span class="w"> </span><span class="n">but</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">does</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">allow</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">mask</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;parafac&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">allows</span><span class="w"> </span><span class="n">negative</span><span class="w"> </span><span class="n">loadings</span><span class="o">.</span><span class="w"></span>
<span class="o">-</span><span class="w"> </span><span class="s1">&#39;constrained_parafac&#39;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="n">PARAFAC</span><span class="w"> </span><span class="n">through</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">traditional</span><span class="w"> </span><span class="n">ALS</span><span class="o">.</span><span class="w"> </span><span class="n">It</span><span class="w"> </span><span class="n">allows</span><span class="w"></span>
<span class="w">                          </span><span class="n">negative</span><span class="w"> </span><span class="n">loadings</span><span class="o">.</span><span class="w"> </span><span class="n">Also</span><span class="p">,</span><span class="w"> </span><span class="n">it</span><span class="w"> </span><span class="n">incorporates</span><span class="w"> </span><span class="n">L1</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">L2</span><span class="w"></span>
<span class="w">                          </span><span class="n">regularization</span><span class="p">,</span><span class="w"> </span><span class="n">includes</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="s1">&#39;non_negative&#39;</span><span class="w"> </span><span class="n">option</span><span class="p">,</span><span class="w"> </span><span class="ow">and</span><span class="w"></span>
<span class="w">                          </span><span class="n">allows</span><span class="w"> </span><span class="n">constraining</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">sparsity</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">decomposition</span><span class="o">.</span><span class="w"></span>
<span class="w">                          </span><span class="n">For</span><span class="w"> </span><span class="n">more</span><span class="w"> </span><span class="n">information</span><span class="p">,</span><span class="w"> </span><span class="n">see</span><span class="w"></span>
<span class="w">                          </span><span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">tensorly</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">stable</span><span class="o">/</span><span class="n">modules</span><span class="o">/</span><span class="n">generated</span><span class="o">/</span><span class="n">tensorly</span><span class="o">.</span><span class="n">decomposition</span><span class="o">.</span><span class="n">constrained_parafac</span><span class="o">.</span><span class="n">html</span><span class="c1">#tensorly.decomposition.constrained_parafac</span><span class="w"></span>
</code></pre></div>

<p _random_="random" _svd_="svd,">init : str, default='svd'
    Initialization method for computing the Tensor Factorization.</p>
<p>svd : str, default='numpy_svd'
    Function to compute the SVD, acceptable values in tensorly.SVD_FUNS</p>
<p>metric : str, default='error'
    Metric to perform the elbow analysis (y-axis)</p>
<div class="codehilite"><pre><span></span><code>- &#39;error&#39; : Normalized error to compute the elbow.
- &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).
</code></pre></div>

<p>random_state : int, default=None
    Seed for randomization.</p>
<p>tol : float, default=10e-7
    Tolerance for the decomposition algorithm to stop when the variation in
    the reconstruction error is less than the tolerance. Lower <code>tol</code> helps
    to improve the solution obtained from the decomposition, but it takes
    longer to run.</p>
<p>n_iter_max : int, default=100
    Maximum number of iteration to reach an optimal solution with the
    decomposition algorithm. Higher <code>n_iter_max</code>helps to improve the solution
    obtained from the decomposition, but it takes longer to run.</p>
<p>automatic_elbow : boolean, default=True
    Whether using an automatic strategy to find the elbow. If True, the method
    implemented by the package kneed is used.</p>
<p>manual_elbow : int, default=None
    Rank or number of factors to highlight in the curve of error achieved by
    the Tensor Factorization. This input is considered only when
    <code>automatic_elbow=True</code></p>
<p>smooth : boolean, default=False
    Whether smoothing the curve with a Savitzky-Golay filter.</p>
<p>mask : ndarray list, default=None
    Helps avoiding missing values during a tensor factorization. A mask should be
    a boolean array of the same shape as the original tensor and should be 0
    where the values are missing and 1 everywhere else.</p>
<p _95_="'95%'" _std_="'std',">ci : str, default='std'
    Confidence interval for representing the multiple runs in each rank.</p>
<p>figsize : tuple, default=(4, 2.25)
    Figure size, width by height</p>
<p>fontsize : int, default=14
    Fontsize for axis labels.</p>
<p>filename : str, default=None
    Path to save the figure of the elbow analysis. If None, the figure is not
    saved.</p>
<p>output_fig : boolean, default=True
    Whether generating the figure with matplotlib.</p>
<p>verbose : boolean, default=False
    Whether printing or not steps of the analysis.</p>
<p>**kwargs : dict
    Extra arguments for the tensor factorization according to inputs in
    tensorly.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.elbow_rank_selection--returns">Returns</h6>
<p>fig : matplotlib.figure.Figure
    Figure object made with matplotlib</p>
<p>loss : list
    List of normalized errors for each rank. Here the errors are te average
    across distinct runs for each rank.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">elbow_rank_selection</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">upper_rank</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">runs</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tf_type</span><span class="o">=</span><span class="s1">&#39;non_negative_cp&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="s1">&#39;random&#39;</span><span class="p">,</span> <span class="n">svd</span><span class="o">=</span><span class="s1">&#39;numpy_svd&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                         <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;error&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_iter_max</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">10e-7</span><span class="p">,</span> <span class="n">automatic_elbow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                         <span class="n">manual_elbow</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ci</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mf">2.25</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>                         <span class="n">filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_fig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="sd">&#39;&#39;&#39;Elbow analysis on the error achieved by the Tensor Factorization for</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    selecting the number of factors to use. A plot is made with the results.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    upper_rank : int, default=50</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        Upper bound of ranks to explore with the elbow analysis.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    runs : int, default=20</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Number of tensor factorization performed for a given rank. Each</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        factorization varies in the seed of initialization.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    tf_type : str, default=&#39;non_negative_cp&#39;</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Type of Tensor Factorization.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        - &#39;non_negative_cp&#39; : Non-negative PARAFAC through the traditional ALS.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;non_negative_cp_hals&#39; : Non-negative PARAFAC through the Hierarchical ALS.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">                                   It reaches an optimal solution faster than the</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                                   traditional ALS, but it does not allow a mask.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        - &#39;parafac&#39; : PARAFAC through the traditional ALS. It allows negative loadings.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;constrained_parafac&#39; : PARAFAC through the traditional ALS. It allows</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                  negative loadings. Also, it incorporates L1 and L2</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                                  regularization, includes a &#39;non_negative&#39; option, and</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                                  allows constraining the sparsity of the decomposition.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                                  For more information, see</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                                  http://tensorly.org/stable/modules/generated/tensorly.decomposition.constrained_parafac.html#tensorly.decomposition.constrained_parafac</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    init : str, default=&#39;svd&#39;</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        Initialization method for computing the Tensor Factorization.</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        {svd, random}</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    svd : str, default=&#39;numpy_svd&#39;</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Function to compute the SVD, acceptable values in tensorly.SVD_FUNS</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    metric : str, default=&#39;error&#39;</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        Metric to perform the elbow analysis (y-axis)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        - &#39;error&#39; : Normalized error to compute the elbow.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        - &#39;similarity&#39; : Similarity based on CorrIndex (1-CorrIndex).</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">    random_state : int, default=None</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        Seed for randomization.</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">    tol : float, default=10e-7</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Tolerance for the decomposition algorithm to stop when the variation in</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        the reconstruction error is less than the tolerance. Lower `tol` helps</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        to improve the solution obtained from the decomposition, but it takes</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        longer to run.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">    n_iter_max : int, default=100</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">        Maximum number of iteration to reach an optimal solution with the</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">        decomposition algorithm. Higher `n_iter_max`helps to improve the solution</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        obtained from the decomposition, but it takes longer to run.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">    automatic_elbow : boolean, default=True</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        Whether using an automatic strategy to find the elbow. If True, the method</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        implemented by the package kneed is used.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    manual_elbow : int, default=None</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        Rank or number of factors to highlight in the curve of error achieved by</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        the Tensor Factorization. This input is considered only when</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        `automatic_elbow=True`</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    smooth : boolean, default=False</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        Whether smoothing the curve with a Savitzky-Golay filter.</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">    mask : ndarray list, default=None</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        Helps avoiding missing values during a tensor factorization. A mask should be</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">        a boolean array of the same shape as the original tensor and should be 0</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        where the values are missing and 1 everywhere else.</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">    ci : str, default=&#39;std&#39;</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        Confidence interval for representing the multiple runs in each rank.</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        {&#39;std&#39;, &#39;95%&#39;}</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    figsize : tuple, default=(4, 2.25)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        Figure size, width by height</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">    fontsize : int, default=14</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">        Fontsize for axis labels.</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        Path to save the figure of the elbow analysis. If None, the figure is not</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a><span class="sd">        saved.</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">    output_fig : boolean, default=True</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">        Whether generating the figure with matplotlib.</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">        Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">    **kwargs : dict</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">        Extra arguments for the tensor factorization according to inputs in</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">        tensorly.</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">    -------</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">    fig : matplotlib.figure.Figure</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">        Figure object made with matplotlib</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">    loss : list</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">        List of normalized errors for each rank. Here the errors are te average</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">        across distinct runs for each rank.</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>    <span class="k">assert</span> <span class="n">metric</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;similarity&#39;</span><span class="p">,</span> <span class="s1">&#39;error&#39;</span><span class="p">],</span> <span class="s2">&quot;`metric` must be either &#39;similarity&#39; or &#39;error&#39;&quot;</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>    <span class="n">ylabel</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;similarity&#39;</span> <span class="p">:</span> <span class="s1">&#39;Similarity</span><span class="se">\n</span><span class="s1">(1-CorrIndex)&#39;</span><span class="p">,</span> <span class="s1">&#39;error&#39;</span> <span class="p">:</span> <span class="s1">&#39;Normalized Error&#39;</span><span class="p">}</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="c1"># Run analysis</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Running Elbow Analysis&#39;</span><span class="p">)</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>    <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;similarity&#39;</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>        <span class="k">assert</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;`runs` must be greater than 1 when `metric` = &#39;similarity&#39;&quot;</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="k">if</span> <span class="n">runs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="n">loss</span> <span class="o">=</span> <span class="n">_run_elbow_analysis</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                                   <span class="n">upper_rank</span><span class="o">=</span><span class="n">upper_rank</span><span class="p">,</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>                                   <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>                                   <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>                                   <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>                                   <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>                                   <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>                                   <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>                                   <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>                                   <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>                                   <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>                                   <span class="p">)</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>        <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">())</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="n">all_loss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]])</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>        <span class="k">if</span> <span class="n">automatic_elbow</span><span class="p">:</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>            <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>                <span class="n">loss_</span> <span class="o">=</span> <span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">loss</span><span class="p">]</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>                <span class="n">loss</span> <span class="o">=</span> <span class="n">smooth_curve</span><span class="p">(</span><span class="n">loss_</span><span class="p">)</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>                <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">loss</span><span class="p">)]</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>            <span class="n">rank</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">_compute_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">))</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>            <span class="n">rank</span> <span class="o">=</span> <span class="n">manual_elbow</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>        <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>            <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_elbow</span><span class="p">(</span><span class="n">loss</span><span class="o">=</span><span class="n">loss</span><span class="p">,</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                             <span class="n">elbow</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                             <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                             <span class="n">ylabel</span><span class="o">=</span><span class="n">ylabel</span><span class="p">[</span><span class="n">metric</span><span class="p">],</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>                             <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>                             <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>            <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="k">elif</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>        <span class="n">all_loss</span> <span class="o">=</span> <span class="n">_multiple_runs_elbow_analysis</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>                                                 <span class="n">upper_rank</span><span class="o">=</span><span class="n">upper_rank</span><span class="p">,</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>                                                 <span class="n">runs</span><span class="o">=</span><span class="n">runs</span><span class="p">,</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>                                                 <span class="n">tf_type</span><span class="o">=</span><span class="n">tf_type</span><span class="p">,</span>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>                                                 <span class="n">init</span><span class="o">=</span><span class="n">init</span><span class="p">,</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>                                                 <span class="n">svd</span><span class="o">=</span><span class="n">svd</span><span class="p">,</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>                                                 <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>                                                 <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">,</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>                                                 <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>                                                 <span class="n">n_iter_max</span><span class="o">=</span><span class="n">n_iter_max</span><span class="p">,</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>                                                 <span class="n">tol</span><span class="o">=</span><span class="n">tol</span><span class="p">,</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>                                                 <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>                                                 <span class="o">**</span><span class="n">kwargs</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>                                                 <span class="p">)</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>        <span class="c1"># Same outputs as runs = 1</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>        <span class="n">loss</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">all_loss</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>        <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>            <span class="n">loss</span> <span class="o">=</span> <span class="n">smooth_curve</span><span class="p">(</span><span class="n">loss</span><span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>        <span class="n">loss</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">loss</span><span class="p">)]</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>        <span class="k">if</span> <span class="n">automatic_elbow</span><span class="p">:</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>            <span class="n">rank</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">_compute_elbow</span><span class="p">(</span><span class="n">loss</span><span class="p">))</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>            <span class="n">rank</span> <span class="o">=</span> <span class="n">manual_elbow</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>        <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>            <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_multiple_run_elbow</span><span class="p">(</span><span class="n">all_loss</span><span class="o">=</span><span class="n">all_loss</span><span class="p">,</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                                          <span class="n">ci</span><span class="o">=</span><span class="n">ci</span><span class="p">,</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>                                          <span class="n">elbow</span><span class="o">=</span><span class="n">rank</span><span class="p">,</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>                                          <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>                                          <span class="n">ylabel</span><span class="o">=</span><span class="n">ylabel</span><span class="p">[</span><span class="n">metric</span><span class="p">],</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                                          <span class="n">smooth</span><span class="o">=</span><span class="n">smooth</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>                                          <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                                          <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>            <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-194" name="__codelineno-0-194" href="#__codelineno-0-194"></a>        <span class="k">assert</span> <span class="n">runs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Input runs must be an integer greater than 0&quot;</span>
<a id="__codelineno-0-195" name="__codelineno-0-195" href="#__codelineno-0-195"></a>
<a id="__codelineno-0-196" name="__codelineno-0-196" href="#__codelineno-0-196"></a>    <span class="c1"># Store results</span>
<a id="__codelineno-0-197" name="__codelineno-0-197" href="#__codelineno-0-197"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="n">rank</span>
<a id="__codelineno-0-198" name="__codelineno-0-198" href="#__codelineno-0-198"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric</span> <span class="o">=</span> <span class="n">metric</span>
<a id="__codelineno-0-199" name="__codelineno-0-199" href="#__codelineno-0-199"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_mean</span> <span class="o">=</span> <span class="n">loss</span>
<a id="__codelineno-0-200" name="__codelineno-0-200" href="#__codelineno-0-200"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">elbow_metric_raw</span> <span class="o">=</span> <span class="n">all_loss</span>
<a id="__codelineno-0-201" name="__codelineno-0-201" href="#__codelineno-0-201"></a>
<a id="__codelineno-0-202" name="__codelineno-0-202" href="#__codelineno-0-202"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">rank</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-203" name="__codelineno-0-203" href="#__codelineno-0-203"></a>        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="nb">int</span><span class="p">)),</span> <span class="s1">&#39;rank must be an integer.&#39;</span>
<a id="__codelineno-0-204" name="__codelineno-0-204" href="#__codelineno-0-204"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;The rank at the elbow is: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rank</span><span class="p">))</span>
<a id="__codelineno-0-205" name="__codelineno-0-205" href="#__codelineno-0-205"></a>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">loss</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.excluded_value_fraction">
<code class="highlight language-python"><span class="n">excluded_value_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Returns the fraction of excluded values in the tensor,
given the values that are masked in tensor.mask</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.excluded_value_fraction--returns">Returns</h6>
<p>excluded_fraction : float
    Fraction of missing/excluded values in the tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">excluded_value_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Returns the fraction of excluded values in the tensor,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    given the values that are masked in tensor.mask</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    -------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    excluded_fraction : float</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Fraction of missing/excluded values in the tensor.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have masked values&quot;</span><span class="p">)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>        <span class="n">fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>        <span class="n">excluded_fraction</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>        <span class="k">return</span> <span class="n">excluded_fraction</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.explained_variance">
<code class="highlight language-python"><span class="n">explained_variance</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Computes the explained variance score for a tensor decomposition. Inspired on the
function in sklearn.metrics.explained_variance_score.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.explained_variance--returns">Returns</h6>
<p>explained_variance : float
    Explained variance score for a tnesor factorization.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">explained_variance</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes the explained variance score for a tensor decomposition. Inspired on the</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    function in sklearn.metrics.explained_variance_score.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    -------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    explained_variance : float</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Explained variance score for a tnesor factorization.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must run compute_tensor_factorization before using this method.&quot;</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>    <span class="n">tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>    <span class="n">rec_tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tl_object</span><span class="o">.</span><span class="n">to_tensor</span><span class="p">()</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>        <span class="n">tensor</span> <span class="o">=</span> <span class="n">tensor</span> <span class="o">*</span> <span class="n">mask</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="n">rec_tensor</span> <span class="o">=</span> <span class="n">tensor</span> <span class="o">*</span> <span class="n">mask</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>    <span class="n">y_diff_avg</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">rec_tensor</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>    <span class="n">numerator</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">rec_tensor</span> <span class="o">-</span> <span class="n">y_diff_avg</span><span class="p">)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">tensor_avg</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="n">denominator</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">tensor</span> <span class="o">-</span> <span class="n">tensor_avg</span><span class="p">)</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="k">if</span> <span class="n">denominator</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">:</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="n">explained_variance</span> <span class="o">=</span> <span class="mf">0.0</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">explained_variance</span> <span class="o">=</span>  <span class="mf">1.</span> <span class="o">-</span> <span class="p">(</span><span class="n">numerator</span> <span class="o">/</span> <span class="n">denominator</span><span class="p">)</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">explained_variance</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">return</span> <span class="n">explained_variance</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.export_factor_loadings">
<code class="highlight language-python"><span class="n">export_factor_loadings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Exports the factor loadings of the tensor into an Excel file</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.export_factor_loadings--parameters">Parameters</h6>
<p>filename : str
    Full path and filename to store the file. E.g., '/home/user/Loadings.xlsx'</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">export_factor_loadings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Exports the factor loadings of the tensor into an Excel file</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Full path and filename to store the file. E.g., &#39;/home/user/Loadings.xlsx&#39;</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>    <span class="n">writer</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">factors</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="n">v</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="n">k</span><span class="p">)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>    <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Loadings of the tensor factorization were successfully saved into </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.get_top_factor_elements">
<code class="highlight language-python"><span class="n">get_top_factor_elements</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">order_name</span><span class="p">,</span> <span class="n">factor_name</span><span class="p">,</span> <span class="n">top_number</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Obtains the top-elements with higher loadings for a given factor</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.get_top_factor_elements--parameters">Parameters</h6>
<p>order_name : str
    Name of the dimension/order in the tensor according to the keys of the
    dictionary in BaseTensor.factors. The attribute factors is built once the
    tensor factorization is run.</p>
<p>factor_name : str
    Name of one of the factors. E.g., 'Factor 1'</p>
<p>top_number : int, default=10
    Number of top-elements to return</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.get_top_factor_elements--returns">Returns</h6>
<p>top_elements : pandas.DataFrame
    A dataframe with the loadings of the top-elements for the given factor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">get_top_factor_elements</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">order_name</span><span class="p">,</span> <span class="n">factor_name</span><span class="p">,</span> <span class="n">top_number</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Obtains the top-elements with higher loadings for a given factor</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    order_name : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Name of the dimension/order in the tensor according to the keys of the</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        dictionary in BaseTensor.factors. The attribute factors is built once the</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        tensor factorization is run.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    factor_name : str</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Name of one of the factors. E.g., &#39;Factor 1&#39;</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    top_number : int, default=10</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Number of top-elements to return</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    -------</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    top_elements : pandas.DataFrame</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        A dataframe with the loadings of the top-elements for the given factor.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">top_elements</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="n">order_name</span><span class="p">][</span><span class="n">factor_name</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_number</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">return</span> <span class="n">top_elements</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.missing_fraction">
<code class="highlight language-python"><span class="n">missing_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Returns the fraction of values that are missing (NaNs) in the tensor,
given the values that are in tensor.loc_nans</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.missing_fraction--returns">Returns</h6>
<p>missing_fraction : float
    Fraction of values that are real zeros.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">missing_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Returns the fraction of values that are missing (NaNs) in the tensor,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    given the values that are in tensor.loc_nans</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    -------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    missing_fraction : float</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Fraction of values that are real zeros.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have missing values&quot;</span><span class="p">)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>        <span class="n">missing_fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>    <span class="n">missing_fraction</span> <span class="o">=</span> <span class="n">missing_fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">return</span> <span class="n">missing_fraction</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.sparsity_fraction">
<code class="highlight language-python"><span class="n">sparsity_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Returns the fraction of values that are zeros in the tensor,
given the values that are in tensor.loc_zeros</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.sparsity_fraction--returns">Returns</h6>
<p>sparsity_fraction : float
    Fraction of values that are real zeros.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">sparsity_fraction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Returns the fraction of values that are zeros in the tensor,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    given the values that are in tensor.loc_zeros</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    -------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    sparsity_fraction : float</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Fraction of values that are real zeros.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The interaction tensor does not have zeros&quot;</span><span class="p">)</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>        <span class="k">return</span> <span class="mf">0.0</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>        <span class="n">sparsity_fraction</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span> <span class="o">/</span> <span class="n">tl</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>    <span class="n">sparsity_fraction</span> <span class="o">=</span> <span class="n">sparsity_fraction</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">return</span> <span class="n">sparsity_fraction</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.to_device">
<code class="highlight language-python"><span class="n">to_device</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Changes the device where the tensor
is analyzed.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.to_device--parameters">Parameters</h6>
<p>device : str
    Device name to use for the decomposition.
    Options could be 'cpu', 'cuda', 'gpu', depending on
    the backend used with tensorly.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">to_device</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">device</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Changes the device where the tensor</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    is analyzed.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    device : str</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Device name to use for the decomposition.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        Options could be &#39;cpu&#39;, &#39;cuda&#39;, &#39;gpu&#39;, depending on</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        the backend used with tensorly.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>    <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>    <span class="k">except</span><span class="p">:</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Device is either not available or the backend used with tensorly does not support this device.</span><span class="se">\</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="s1">               Try changing it with tensorly.set_backend(&quot;&lt;backend_name&gt;&quot;) before.&#39;</span><span class="p">)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h5 class="doc doc-heading" id="cell2cell.tensor.tensor.BaseTensor.write_file">
<code class="highlight language-python"><span class="n">write_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span></code>


</h5>

    <div class="doc doc-contents ">

      <p>Exports this object into a pickle file.</p>
<h6 id="cell2cell.tensor.tensor.BaseTensor.write_file--parameters">Parameters</h6>
<p>filename : str
    Complete path to the file wherein the variable will be
    stored. For example:
    /home/user/variable.pkl</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">write_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Exports this object into a pickle file.</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    filename : str</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Complete path to the file wherein the variable will be</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        stored. For example:</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        /home/user/variable.pkl</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>    <span class="kn">from</span> <span class="nn">cell2cell.io.save_data</span> <span class="kn">import</span> <span class="n">export_variable_with_pickle</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>    <span class="n">export_variable_with_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">filename</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.InteractionTensor">
        <code>
InteractionTensor            (<a class="autorefs autorefs-internal" title="cell2cell.tensor.tensor.BaseTensor" href="#cell2cell.tensor.tensor.BaseTensor">BaseTensor</a>)
        </code>



</h4>

    <div class="doc doc-contents ">

      <p>4D-Communication Tensor built from gene expression matrices for different contexts
 and a list of ligand-receptor pairs</p>
<h6 id="cell2cell.tensor.tensor.InteractionTensor--parameters">Parameters</h6>
<p>rnaseq_matrices : list
    A list with dataframes of gene expression wherein the rows are the genes and
    columns the cell types, tissues or samples.</p>
<p>ppi_data : pandas.DataFrame
    A dataframe containing protein-protein interactions (rows). It has to
    contain at least two columns, one for the first protein partner in the
    interaction as well as the second protein partner.</p>
<p>order_labels : list, default=None
    List containing the labels for each order or dimension of the tensor. For
    example: ['Contexts', 'Ligand-Receptor Pairs', 'Sender Cells', 'Receiver Cells']</p>
<p>context_names : list, default=None
    A list of strings containing the names of the corresponding contexts to each
    rnaseq_matrix. The length of this list must match the length of the list
    rnaseq_matrices.</p>
<p>how : str, default='inner'
    Approach to consider cell types and genes present across multiple contexts.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">inner</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
            <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span>
            <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_genes</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_cells</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
</code></pre></div>

<p>outer_fraction : float, default=0.0
    Threshold to filter the elements when <code>how</code> includes any outer option.
    Elements with a fraction abundance across samples (in <code>rnaseq_matrices</code>)
    at least this threshold will be included. When this value is 0, considers
    all elements across the samples. When this value is 1, it acts as using
    <code>how='inner'</code>.</p>
<p>communication_score : str, default='expression_mean'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                       of a ligand from a sender cell and the
                       expression of a receptor on a receiver cell.
</code></pre></div>

<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>complex_agg_method : str, default='min'
    Method to aggregate the expression value of multiple genes in a
    complex.</p>
<div class="codehilite"><pre><span></span><code>- &#39;min&#39; : Minimum expression value among all genes.
- &#39;mean&#39; : Average expression value among all genes.
- &#39;gmean&#39; : Geometric mean expression value among all genes.
</code></pre></div>

<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the gene names in the expression matrices and the
    protein names in the ppi_data to match their names and integrate their
    respective expression level. Useful when there are inconsistencies in the
    names between the expression matrix and the ligand-receptor annotations.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a dataframe of
    protein-protein interactions. If the list is for ligand-receptor pairs, the
    first column is for the ligands and the second for the receptors.</p>
<p>group_ppi_by : str, default=None
    Column name in the list of PPIs used for grouping individual PPIs into major
    groups such as signaling pathways.</p>
<p>group_ppi_method : str, default='gmean'
    Method for aggregating multiple PPIs into major groups.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">mean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">communication</span> <span class="nv">score</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
           <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">gmean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">geometric</span> <span class="nv">mean</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span>
            <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">sum</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">sum</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
          <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
</code></pre></div>

<p None="None" _cpu_="'cpu'," _cuda:0_="'cuda:0',">device : str, default=None
    Device to use when backend allows using multiple devices. Options are:</p>
<p>verbose : boolean, default=False
        Whether printing or not steps of the analysis.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">InteractionTensor</span><span class="p">(</span><span class="n">BaseTensor</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;4D-Communication Tensor built from gene expression matrices for different contexts</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">     and a list of ligand-receptor pairs</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_matrices : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A list with dataframes of gene expression wherein the rows are the genes and</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        columns the cell types, tissues or samples.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A dataframe containing protein-protein interactions (rows). It has to</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        contain at least two columns, one for the first protein partner in the</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        interaction as well as the second protein partner.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    order_labels : list, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List containing the labels for each order or dimension of the tensor. For</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        example: [&#39;Contexts&#39;, &#39;Ligand-Receptor Pairs&#39;, &#39;Sender Cells&#39;, &#39;Receiver Cells&#39;]</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    context_names : list, default=None</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        A list of strings containing the names of the corresponding contexts to each</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        rnaseq_matrix. The length of this list must match the length of the list</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        rnaseq_matrices.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    how : str, default=&#39;inner&#39;</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        Approach to consider cell types and genes present across multiple contexts.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - &#39;inner&#39; : Considers only cell types and genes that are present in all</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                    contexts (intersection).</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;outer&#39; : Considers all cell types and genes that are present</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                    across contexts (union).</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        - &#39;outer_genes&#39; : Considers only cell types that are present in all</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">                          contexts (intersection), while all genes that are</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        - &#39;outer_cells&#39; : Considers only genes that are present in all</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">                          contexts (intersection), while all cell types that are</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    outer_fraction : float, default=0.0</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        Threshold to filter the elements when `how` includes any outer option.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        Elements with a fraction abundance across samples (in `rnaseq_matrices`)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        at least this threshold will be included. When this value is 0, considers</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        all elements across the samples. When this value is 1, it acts as using</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        `how=&#39;inner&#39;`.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">    communication_score : str, default=&#39;expression_mean&#39;</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">                               of a ligand from a sender cell and the</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">                               expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">    complex_agg_method : str, default=&#39;min&#39;</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">        Method to aggregate the expression value of multiple genes in a</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        complex.</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        - &#39;min&#39; : Minimum expression value among all genes.</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a><span class="sd">        - &#39;mean&#39; : Average expression value among all genes.</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">        - &#39;gmean&#39; : Geometric mean expression value among all genes.</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a><span class="sd">        Whether making uppercase the gene names in the expression matrices and the</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">        protein names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        respective expression level. Useful when there are inconsistencies in the</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a><span class="sd">        names between the expression matrix and the ligand-receptor annotations.</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        Contains the names of the columns where to find the partners in a dataframe of</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">        protein-protein interactions. If the list is for ligand-receptor pairs, the</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        first column is for the ligands and the second for the receptors.</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a><span class="sd">    group_ppi_by : str, default=None</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">        Column name in the list of PPIs used for grouping individual PPIs into major</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">        groups such as signaling pathways.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">    group_ppi_method : str, default=&#39;gmean&#39;</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">        Method for aggregating multiple PPIs into major groups.</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        - &#39;mean&#39; : Computes the average communication score among all PPIs of the</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">                   group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a><span class="sd">        - &#39;gmean&#39; : Computes the geometric mean of the communication scores among all</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">                    PPIs of the group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        - &#39;sum&#39; : Computes the sum of the communication scores among all PPIs of the</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a><span class="sd">                  group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a><span class="sd">        Device to use when backend allows using multiple devices. Options are:</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">         {&#39;cpu&#39;, &#39;cuda:0&#39;, None}</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rnaseq_matrices</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">context_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a>                 <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_mean&#39;</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">complex_agg_method</span><span class="o">=</span><span class="s1">&#39;min&#39;</span><span class="p">,</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a>                 <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">group_ppi_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a>                 <span class="n">group_ppi_method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a>        <span class="c1"># Asserts</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>        <span class="k">if</span> <span class="n">group_ppi_by</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>            <span class="k">assert</span> <span class="n">group_ppi_by</span> <span class="ow">in</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;Using </span><span class="si">{}</span><span class="s2"> for grouping PPIs is not possible. Not present among columns in ppi_data&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">group_ppi_by</span><span class="p">)</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="c1"># Init BaseTensor</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="n">BaseTensor</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="c1"># Generate expression values for protein complexes in PPI data</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>        <span class="k">if</span> <span class="n">complex_sep</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Getting expression values for protein complexes&#39;</span><span class="p">)</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>            <span class="n">col_a_genes</span><span class="p">,</span> <span class="n">complex_a</span><span class="p">,</span> <span class="n">col_b_genes</span><span class="p">,</span> <span class="n">complex_b</span><span class="p">,</span> <span class="n">complexes</span> <span class="o">=</span> <span class="n">get_genes_from_complexes</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>                                                                                                 <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                                                                                                 <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>                                                                                                 <span class="p">)</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>            <span class="n">mod_rnaseq_matrices</span> <span class="o">=</span> <span class="p">[</span><span class="n">add_complexes_to_expression</span><span class="p">(</span><span class="n">rnaseq</span><span class="p">,</span> <span class="n">complexes</span><span class="p">,</span> <span class="n">agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span><span class="p">)</span> <span class="k">for</span> <span class="n">rnaseq</span> <span class="ow">in</span> <span class="n">rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>            <span class="n">mod_rnaseq_matrices</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>        <span class="c1"># Uppercase for Gene names</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="k">if</span> <span class="n">upper_letter_comparison</span><span class="p">:</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>            <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">mod_rnaseq_matrices</span><span class="p">:</span>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>                <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">idx</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>        <span class="c1"># Deduplicate gene names</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>        <span class="n">mod_rnaseq_matrices</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)]</span> <span class="k">for</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">mod_rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>        <span class="c1"># Get context CCC tensor</span>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>        <span class="n">tensor</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">cells</span><span class="p">,</span> <span class="n">ppi_names</span><span class="p">,</span> <span class="n">mask</span> <span class="o">=</span> <span class="n">build_context_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_matrices</span><span class="o">=</span><span class="n">mod_rnaseq_matrices</span><span class="p">,</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>                                                                         <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>                                                                         <span class="n">how</span><span class="o">=</span><span class="n">how</span><span class="p">,</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>                                                                         <span class="n">outer_fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">,</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>                                                                         <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>                                                                         <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>                                                                         <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                                                                         <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                                                                         <span class="n">group_ppi_by</span><span class="o">=</span><span class="n">group_ppi_by</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                                                                         <span class="n">group_ppi_method</span><span class="o">=</span><span class="n">group_ppi_method</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                                                                         <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">)</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>        <span class="c1"># Distinguish NaNs from real zeros</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="p">(</span><span class="n">tensor</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>        <span class="c1"># Generate names for the elements in each dimension (order) in the tensor</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="k">if</span> <span class="n">context_names</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>            <span class="n">context_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;C-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mod_rnaseq_matrices</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>            <span class="c1"># for PPIS use ppis, and sender &amp; receiver cells, use cells</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>        <span class="c1"># Save variables for this class</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">communication_score</span> <span class="o">=</span> <span class="n">communication_score</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">how</span> <span class="o">=</span> <span class="n">how</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">outer_fraction</span> <span class="o">=</span> <span class="n">outer_fraction</span>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>        <span class="k">if</span> <span class="n">device</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>            <span class="k">if</span> <span class="n">tl</span><span class="o">.</span><span class="n">get_backend</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pytorch&#39;</span><span class="p">,</span> <span class="s1">&#39;tensorflow&#39;</span><span class="p">]:</span> <span class="c1"># Potential TODO: Include other backends that support different devices</span>
<a id="__codelineno-0-175" name="__codelineno-0-175" href="#__codelineno-0-175"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-176" name="__codelineno-0-176" href="#__codelineno-0-176"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-177" name="__codelineno-0-177" href="#__codelineno-0-177"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-178" name="__codelineno-0-178" href="#__codelineno-0-178"></a>                <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-179" name="__codelineno-0-179" href="#__codelineno-0-179"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<a id="__codelineno-0-180" name="__codelineno-0-180" href="#__codelineno-0-180"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-181" name="__codelineno-0-181" href="#__codelineno-0-181"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-182" name="__codelineno-0-182" href="#__codelineno-0-182"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-183" name="__codelineno-0-183" href="#__codelineno-0-183"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-184" name="__codelineno-0-184" href="#__codelineno-0-184"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span>
<a id="__codelineno-0-185" name="__codelineno-0-185" href="#__codelineno-0-185"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span>
<a id="__codelineno-0-186" name="__codelineno-0-186" href="#__codelineno-0-186"></a>                <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-187" name="__codelineno-0-187" href="#__codelineno-0-187"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-188" name="__codelineno-0-188" href="#__codelineno-0-188"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-189" name="__codelineno-0-189" href="#__codelineno-0-189"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-190" name="__codelineno-0-190" href="#__codelineno-0-190"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span>
<a id="__codelineno-0-191" name="__codelineno-0-191" href="#__codelineno-0-191"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cells</span> <span class="o">=</span> <span class="n">cells</span>
<a id="__codelineno-0-192" name="__codelineno-0-192" href="#__codelineno-0-192"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="o">=</span> <span class="n">order_labels</span>
<a id="__codelineno-0-193" name="__codelineno-0-193" href="#__codelineno-0-193"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">order_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">context_names</span><span class="p">,</span> <span class="n">ppi_names</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">cells</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">cells</span><span class="p">]</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">












  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.PreBuiltTensor">
        <code>
PreBuiltTensor            (<a class="autorefs autorefs-internal" title="cell2cell.tensor.tensor.BaseTensor" href="#cell2cell.tensor.tensor.BaseTensor">BaseTensor</a>)
        </code>



</h4>

    <div class="doc doc-contents ">

      <p>Initializes a cell2cell.tensor.BaseTensor with a prebuilt communication tensor</p>
<h6 id="cell2cell.tensor.tensor.PreBuiltTensor--parameters">Parameters</h6>
<p>tensor : ndarray list
    Prebuilt tensor. Could be a list of lists, a numpy array or a tensorly.tensor.</p>
<p>order_names : list
    List of lists containing the string names of each element in each of the
    dimensions or orders in the tensor. For a 4D-Communication tensor, the first
    list should contain the names of the contexts, the second the names of the
    ligand-receptor interactions, the third the names of the sender cells and the
    fourth the names of the receiver cells.</p>
<p>order_labels : list, default=None
    List containing the labels for each order or dimension of the tensor. For
    example: ['Contexts', 'Ligand-Receptor Pairs', 'Sender Cells', 'Receiver Cells']</p>
<p>mask : ndarray list, default=None
    Helps avoiding missing values during a tensor factorization. A mask should be
    a boolean array of the same shape as the original tensor and should be 0
    where the values are missing and 1 everywhere else.</p>
<p>loc_nans : ndarray list, default=None
    An array of shape equal to <code>tensor</code> with ones where NaN values were assigned
    when building the tensor. Other values are zeros. It stores the
    location of the NaN values.</p>
<p None="None" _cpu_="'cpu'," _cuda:0_="'cuda:0',">device : str, default=None
    Device to use when backend allows using multiple devices. Options are:</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">class</span> <span class="nc">PreBuiltTensor</span><span class="p">(</span><span class="n">BaseTensor</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Initializes a cell2cell.tensor.BaseTensor with a prebuilt communication tensor</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    tensor : ndarray list</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">        Prebuilt tensor. Could be a list of lists, a numpy array or a tensorly.tensor.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    order_names : list</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of lists containing the string names of each element in each of the</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        dimensions or orders in the tensor. For a 4D-Communication tensor, the first</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        list should contain the names of the contexts, the second the names of the</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        ligand-receptor interactions, the third the names of the sender cells and the</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        fourth the names of the receiver cells.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    order_labels : list, default=None</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List containing the labels for each order or dimension of the tensor. For</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        example: [&#39;Contexts&#39;, &#39;Ligand-Receptor Pairs&#39;, &#39;Sender Cells&#39;, &#39;Receiver Cells&#39;]</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    mask : ndarray list, default=None</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Helps avoiding missing values during a tensor factorization. A mask should be</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        a boolean array of the same shape as the original tensor and should be 0</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        where the values are missing and 1 everywhere else.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    loc_nans : ndarray list, default=None</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        An array of shape equal to `tensor` with ones where NaN values were assigned</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        when building the tensor. Other values are zeros. It stores the</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        location of the NaN values.</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        Device to use when backend allows using multiple devices. Options are:</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">         {&#39;cpu&#39;, &#39;cuda:0&#39;, None}</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tensor</span><span class="p">,</span> <span class="n">order_names</span><span class="p">,</span> <span class="n">order_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">loc_nans</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="c1"># Init BaseTensor</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">BaseTensor</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>        <span class="c1"># Initialize tensor</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="n">context</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">context</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>        <span class="k">except</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>            <span class="n">context</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;dtype&#39;</span><span class="p">:</span> <span class="n">tensor</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span> <span class="s1">&#39;device&#39;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>            <span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>        <span class="c1"># Location of NaNs and zeros</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">tmp_nans</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">tensor</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="c1"># Find extra NaNs that were not considered</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="k">if</span> <span class="n">loc_nans</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>            <span class="k">assert</span> <span class="n">loc_nans</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="s2">&quot;`loc_nans` and `tensor` must be of the same shape&quot;</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">loc_nans</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">+</span> <span class="n">tmp_nans</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>        <span class="c1"># Store tensor</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>        <span class="n">tensor_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>        <span class="k">if</span> <span class="n">device</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>            <span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">device</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>        <span class="k">if</span> <span class="s1">&#39;device&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">context</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor_</span><span class="p">)</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">tensor_</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_nans</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loc_zeros</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>            <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">mask</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">order_names</span> <span class="o">=</span> <span class="n">order_names</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="k">if</span> <span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Dimension-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">))]</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span> <span class="o">=</span> <span class="n">order_labels</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">order_labels</span><span class="p">),</span> <span class="s2">&quot;The length of order_labels must match the number of orders/dimensions in the tensor&quot;</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">












  </div>

    </div>

  </div>




  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.aggregate_ccc_tensor">
<code class="highlight language-python"><span class="n">aggregate_ccc_tensor</span><span class="p">(</span><span class="n">ccc_tensor</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">group_ppi_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_ppi_method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Aggregates communication scores of multiple PPIs into major groups
(e.g., pathways) in a communication tensor</p>
<h6 id="cell2cell.tensor.tensor.aggregate_ccc_tensor--parameters">Parameters</h6>
<p>ccc_tensor : ndarray list
    List of directed cell-cell communication matrices, one for each ligand-
    receptor pair in ppi_data. These matrices contain the communication score for
    pairs of cells for the corresponding PPI. This tensor represent a
    3D-communication tensor for the context.</p>
<p>ppi_data : pandas.DataFrame
    A dataframe containing protein-protein interactions (rows). It has to
    contain at least two columns, one for the first protein partner in the
    interaction as well as the second protein partner.</p>
<p>group_ppi_by : str, default=None
    Column name in the list of PPIs used for grouping individual PPIs into major
    groups such as signaling pathways.</p>
<p>group_ppi_method : str, default='gmean'
    Method for aggregating multiple PPIs into major groups.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">mean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">communication</span> <span class="nv">score</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
           <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">gmean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">geometric</span> <span class="nv">mean</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span>
            <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">sum</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">sum</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
          <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
</code></pre></div>

<h6 id="cell2cell.tensor.tensor.aggregate_ccc_tensor--returns">Returns</h6>
<p>aggregated_tensor : ndarray list
    List of directed cell-cell communication matrices, one for each major group of
    ligand-receptor pair in ppi_data. These matrices contain the communication
    score for pairs of cells for the corresponding PPI group. This tensor
    represent a 3D-communication tensor for the context, but for major groups
    instead of individual PPIs.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">aggregate_ccc_tensor</span><span class="p">(</span><span class="n">ccc_tensor</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">group_ppi_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_ppi_method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Aggregates communication scores of multiple PPIs into major groups</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    (e.g., pathways) in a communication tensor</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ccc_tensor : ndarray list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of directed cell-cell communication matrices, one for each ligand-</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        receptor pair in ppi_data. These matrices contain the communication score for</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        pairs of cells for the corresponding PPI. This tensor represent a</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        3D-communication tensor for the context.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        A dataframe containing protein-protein interactions (rows). It has to</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        contain at least two columns, one for the first protein partner in the</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        interaction as well as the second protein partner.</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    group_ppi_by : str, default=None</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Column name in the list of PPIs used for grouping individual PPIs into major</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        groups such as signaling pathways.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">    group_ppi_method : str, default=&#39;gmean&#39;</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        Method for aggregating multiple PPIs into major groups.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;mean&#39; : Computes the average communication score among all PPIs of the</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                   group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;gmean&#39; : Computes the geometric mean of the communication scores among all</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                    PPIs of the group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        - &#39;sum&#39; : Computes the sum of the communication scores among all PPIs of the</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                  group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    -------</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    aggregated_tensor : ndarray list</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        List of directed cell-cell communication matrices, one for each major group of</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        ligand-receptor pair in ppi_data. These matrices contain the communication</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        score for pairs of cells for the corresponding PPI group. This tensor</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        represent a 3D-communication tensor for the context, but for major groups</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        instead of individual PPIs.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="n">tensor_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ccc_tensor</span><span class="p">)</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="n">aggregated_tensor</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">group_ppi_by</span><span class="p">):</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="n">lr_idx</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>        <span class="n">ccc_matrices</span> <span class="o">=</span> <span class="n">tensor_</span><span class="p">[</span><span class="n">lr_idx</span><span class="p">]</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">aggregated_tensor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aggregate_ccc_matrices</span><span class="p">(</span><span class="n">ccc_matrices</span><span class="o">=</span><span class="n">ccc_matrices</span><span class="p">,</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>                                                        <span class="n">method</span><span class="o">=</span><span class="n">group_ppi_method</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="k">return</span> <span class="n">aggregated_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.build_context_ccc_tensor">
<code class="highlight language-python"><span class="n">build_context_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_matrices</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">group_ppi_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_ppi_method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Builds a 4D-Communication tensor.
Takes the gene expression matrices and the list of PPIs to compute
the communication scores between the interacting cells for each PPI.
This is done for each context.</p>
<h6 id="cell2cell.tensor.tensor.build_context_ccc_tensor--parameters">Parameters</h6>
<p>rnaseq_matrices : list
    A list with dataframes of gene expression wherein the rows are the genes and
    columns the cell types, tissues or samples.</p>
<p>ppi_data : pandas.DataFrame
    A dataframe containing protein-protein interactions (rows). It has to
    contain at least two columns, one for the first protein partner in the
    interaction as well as the second protein partner.</p>
<p>how : str, default='inner'
    Approach to consider cell types and genes present across multiple contexts.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">inner</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
            <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span>
            <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_genes</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_cells</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
</code></pre></div>

<p>outer_fraction : float, default=0.0
    Threshold to filter the elements when <code>how</code> includes any outer option.
    Elements with a fraction abundance across samples (in <code>rnaseq_matrices</code>)
    at least this threshold will be included. When this value is 0, considers
    all elements across the samples. When this value is 1, it acts as using
    <code>how='inner'</code>.</p>
<p>communication_score : str, default='expression_mean'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                       of a ligand from a sender cell and the
                       expression of a receptor on a receiver cell.
</code></pre></div>

<p>complex_sep : str, default=None
    Symbol that separates the protein subunits in a multimeric complex.
    For example, '&amp;' is the complex_sep for a list of ligand-receptor pairs
    where a protein partner could be "CD74&amp;CD44".</p>
<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the gene names in the expression matrices and the
    protein names in the ppi_data to match their names and integrate their
    respective expression level. Useful when there are inconsistencies in the
    names between the expression matrix and the ligand-receptor annotations.</p>
<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a dataframe of
    protein-protein interactions. If the list is for ligand-receptor pairs, the
    first column is for the ligands and the second for the receptors.</p>
<p>group_ppi_by : str, default=None
    Column name in the list of PPIs used for grouping individual PPIs into major
    groups such as signaling pathways.</p>
<p>group_ppi_method : str, default='gmean'
    Method for aggregating multiple PPIs into major groups.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">mean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">average</span> <span class="nv">communication</span> <span class="nv">score</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
           <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">gmean</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">geometric</span> <span class="nv">mean</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span>
            <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">sum</span><span class="s1">&#39;</span> : <span class="nv">Computes</span> <span class="nv">the</span> <span class="nv">sum</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">communication</span> <span class="nv">scores</span> <span class="nv">among</span> <span class="nv">all</span> <span class="nv">PPIs</span> <span class="nv">of</span> <span class="nv">the</span>
          <span class="nv">group</span> <span class="k">for</span> <span class="nv">a</span> <span class="nv">given</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">cells</span><span class="o">/</span><span class="nv">tissues</span><span class="o">/</span><span class="nv">samples</span>
</code></pre></div>

<p>verbose : boolean, default=False
        Whether printing or not steps of the analysis.</p>
<h6 id="cell2cell.tensor.tensor.build_context_ccc_tensor--returns">Returns</h6>
<p>tensors : list
    List of 3D-Communication tensors for each context. This list corresponds to
    the 4D-Communication tensor.</p>
<p>genes : list
    List of genes included in the tensor.</p>
<p>cells : list
    List of cells included in the tensor.</p>
<div class="admonition ppi_names">
<p class="admonition-title">list</p>
<p>List of names for each of the PPIs included in the tensor. Used as labels for the
elements in the cognate tensor dimension (in the attribute order_names of the
InteractionTensor)</p>
</div>
<div class="admonition mask_tensor">
<p class="admonition-title">numpy.array</p>
<p>Mask used to exclude values in the tensor. When using how='outer' it masks
missing values (e.g., cell types that are not present in a given context),
while using how='inner' makes the mask_tensor to be None.</p>
</div>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">build_context_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_matrices</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                             <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">complex_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>                             <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">),</span> <span class="n">group_ppi_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">group_ppi_method</span><span class="o">=</span><span class="s1">&#39;gmean&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>    <span class="sd">&#39;&#39;&#39;Builds a 4D-Communication tensor.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Takes the gene expression matrices and the list of PPIs to compute</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    the communication scores between the interacting cells for each PPI.</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    This is done for each context.</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    rnaseq_matrices : list</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A list with dataframes of gene expression wherein the rows are the genes and</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        columns the cell types, tissues or samples.</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">        A dataframe containing protein-protein interactions (rows). It has to</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        contain at least two columns, one for the first protein partner in the</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        interaction as well as the second protein partner.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    how : str, default=&#39;inner&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Approach to consider cell types and genes present across multiple contexts.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - &#39;inner&#39; : Considers only cell types and genes that are present in all</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">                    contexts (intersection).</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - &#39;outer&#39; : Considers all cell types and genes that are present</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                    across contexts (union).</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;outer_genes&#39; : Considers only cell types that are present in all</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                          contexts (intersection), while all genes that are</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        - &#39;outer_cells&#39; : Considers only genes that are present in all</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">                          contexts (intersection), while all cell types that are</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">    outer_fraction : float, default=0.0</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">        Threshold to filter the elements when `how` includes any outer option.</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Elements with a fraction abundance across samples (in `rnaseq_matrices`)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        at least this threshold will be included. When this value is 0, considers</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        all elements across the samples. When this value is 1, it acts as using</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        `how=&#39;inner&#39;`.</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">    communication_score : str, default=&#39;expression_mean&#39;</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">                               of a ligand from a sender cell and the</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">                               expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a><span class="sd">    complex_sep : str, default=None</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">        Symbol that separates the protein subunits in a multimeric complex.</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        For example, &#39;&amp;&#39; is the complex_sep for a list of ligand-receptor pairs</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        where a protein partner could be &quot;CD74&amp;CD44&quot;.</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a><span class="sd">        Whether making uppercase the gene names in the expression matrices and the</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">        protein names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">        respective expression level. Useful when there are inconsistencies in the</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">        names between the expression matrix and the ligand-receptor annotations.</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">        Contains the names of the columns where to find the partners in a dataframe of</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a><span class="sd">        protein-protein interactions. If the list is for ligand-receptor pairs, the</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a><span class="sd">        first column is for the ligands and the second for the receptors.</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a><span class="sd">    group_ppi_by : str, default=None</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a><span class="sd">        Column name in the list of PPIs used for grouping individual PPIs into major</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a><span class="sd">        groups such as signaling pathways.</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a><span class="sd">    group_ppi_method : str, default=&#39;gmean&#39;</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a><span class="sd">        Method for aggregating multiple PPIs into major groups.</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a><span class="sd">        - &#39;mean&#39; : Computes the average communication score among all PPIs of the</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a><span class="sd">                   group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a><span class="sd">        - &#39;gmean&#39; : Computes the geometric mean of the communication scores among all</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a><span class="sd">                    PPIs of the group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a><span class="sd">        - &#39;sum&#39; : Computes the sum of the communication scores among all PPIs of the</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a><span class="sd">                  group for a given pair of cells/tissues/samples</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a><span class="sd">    verbose : boolean, default=False</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a><span class="sd">            Whether printing or not steps of the analysis.</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a><span class="sd">    -------</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a><span class="sd">    tensors : list</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a><span class="sd">        List of 3D-Communication tensors for each context. This list corresponds to</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a><span class="sd">        the 4D-Communication tensor.</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a><span class="sd">    genes : list</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a><span class="sd">        List of genes included in the tensor.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a><span class="sd">    cells : list</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a><span class="sd">        List of cells included in the tensor.</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a><span class="sd">    ppi_names: list</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a><span class="sd">        List of names for each of the PPIs included in the tensor. Used as labels for the</span>
<a id="__codelineno-0-103" name="__codelineno-0-103" href="#__codelineno-0-103"></a><span class="sd">        elements in the cognate tensor dimension (in the attribute order_names of the</span>
<a id="__codelineno-0-104" name="__codelineno-0-104" href="#__codelineno-0-104"></a><span class="sd">        InteractionTensor)</span>
<a id="__codelineno-0-105" name="__codelineno-0-105" href="#__codelineno-0-105"></a>
<a id="__codelineno-0-106" name="__codelineno-0-106" href="#__codelineno-0-106"></a><span class="sd">    mask_tensor: numpy.array</span>
<a id="__codelineno-0-107" name="__codelineno-0-107" href="#__codelineno-0-107"></a><span class="sd">        Mask used to exclude values in the tensor. When using how=&#39;outer&#39; it masks</span>
<a id="__codelineno-0-108" name="__codelineno-0-108" href="#__codelineno-0-108"></a><span class="sd">        missing values (e.g., cell types that are not present in a given context),</span>
<a id="__codelineno-0-109" name="__codelineno-0-109" href="#__codelineno-0-109"></a><span class="sd">        while using how=&#39;inner&#39; makes the mask_tensor to be None.</span>
<a id="__codelineno-0-110" name="__codelineno-0-110" href="#__codelineno-0-110"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-111" name="__codelineno-0-111" href="#__codelineno-0-111"></a>    <span class="n">df_idxs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">rnaseq</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">rnaseq</span> <span class="ow">in</span> <span class="n">rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-112" name="__codelineno-0-112" href="#__codelineno-0-112"></a>    <span class="n">df_cols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">rnaseq</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="k">for</span> <span class="n">rnaseq</span> <span class="ow">in</span> <span class="n">rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-113" name="__codelineno-0-113" href="#__codelineno-0-113"></a>
<a id="__codelineno-0-114" name="__codelineno-0-114" href="#__codelineno-0-114"></a>    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;inner&#39;</span><span class="p">:</span>
<a id="__codelineno-0-115" name="__codelineno-0-115" href="#__codelineno-0-115"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_idxs</span><span class="p">))</span>
<a id="__codelineno-0-116" name="__codelineno-0-116" href="#__codelineno-0-116"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_cols</span><span class="p">))</span>
<a id="__codelineno-0-117" name="__codelineno-0-117" href="#__codelineno-0-117"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer&#39;</span><span class="p">:</span>
<a id="__codelineno-0-118" name="__codelineno-0-118" href="#__codelineno-0-118"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_idxs</span><span class="p">),</span>
<a id="__codelineno-0-119" name="__codelineno-0-119" href="#__codelineno-0-119"></a>                                               <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">))</span>
<a id="__codelineno-0-120" name="__codelineno-0-120" href="#__codelineno-0-120"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_cols</span><span class="p">),</span>
<a id="__codelineno-0-121" name="__codelineno-0-121" href="#__codelineno-0-121"></a>                                               <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">))</span>
<a id="__codelineno-0-122" name="__codelineno-0-122" href="#__codelineno-0-122"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer_genes&#39;</span><span class="p">:</span>
<a id="__codelineno-0-123" name="__codelineno-0-123" href="#__codelineno-0-123"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_idxs</span><span class="p">),</span>
<a id="__codelineno-0-124" name="__codelineno-0-124" href="#__codelineno-0-124"></a>                                               <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">))</span>
<a id="__codelineno-0-125" name="__codelineno-0-125" href="#__codelineno-0-125"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_cols</span><span class="p">))</span>
<a id="__codelineno-0-126" name="__codelineno-0-126" href="#__codelineno-0-126"></a>    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s1">&#39;outer_cells&#39;</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127" href="#__codelineno-0-127"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="nb">map</span><span class="p">(</span><span class="nb">set</span><span class="p">,</span> <span class="n">df_idxs</span><span class="p">))</span>
<a id="__codelineno-0-128" name="__codelineno-0-128" href="#__codelineno-0-128"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_elements_over_fraction</span><span class="p">(</span><span class="n">abundance_dict</span><span class="o">=</span><span class="n">get_element_abundances</span><span class="p">(</span><span class="n">element_lists</span><span class="o">=</span><span class="n">df_cols</span><span class="p">),</span>
<a id="__codelineno-0-129" name="__codelineno-0-129" href="#__codelineno-0-129"></a>                                               <span class="n">fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">))</span>
<a id="__codelineno-0-130" name="__codelineno-0-130" href="#__codelineno-0-130"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-131" name="__codelineno-0-131" href="#__codelineno-0-131"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Provide a valid way to build the tensor; &quot;how&quot; must be &quot;inner&quot;, &quot;outer&quot;, &quot;outer_genes&quot; or &quot;outer_cells&quot;&#39;</span><span class="p">)</span>
<a id="__codelineno-0-132" name="__codelineno-0-132" href="#__codelineno-0-132"></a>
<a id="__codelineno-0-133" name="__codelineno-0-133" href="#__codelineno-0-133"></a>    <span class="c1"># Preserve order or sort new set (either inner or outer)</span>
<a id="__codelineno-0-134" name="__codelineno-0-134" href="#__codelineno-0-134"></a>    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_idxs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="n">genes</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135" href="#__codelineno-0-135"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="n">df_idxs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-136" name="__codelineno-0-136" href="#__codelineno-0-136"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137" href="#__codelineno-0-137"></a>        <span class="n">genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>
<a id="__codelineno-0-138" name="__codelineno-0-138" href="#__codelineno-0-138"></a>
<a id="__codelineno-0-139" name="__codelineno-0-139" href="#__codelineno-0-139"></a>    <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_cols</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="n">cells</span><span class="p">:</span>
<a id="__codelineno-0-140" name="__codelineno-0-140" href="#__codelineno-0-140"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="n">df_cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-141" name="__codelineno-0-141" href="#__codelineno-0-141"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-142" name="__codelineno-0-142" href="#__codelineno-0-142"></a>        <span class="n">cells</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cells</span><span class="p">))</span>
<a id="__codelineno-0-143" name="__codelineno-0-143" href="#__codelineno-0-143"></a>
<a id="__codelineno-0-144" name="__codelineno-0-144" href="#__codelineno-0-144"></a>    <span class="c1"># Filter PPI data for</span>
<a id="__codelineno-0-145" name="__codelineno-0-145" href="#__codelineno-0-145"></a>    <span class="n">ppi_data_</span> <span class="o">=</span> <span class="n">filter_ppi_by_proteins</span><span class="p">(</span><span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146" href="#__codelineno-0-146"></a>                                       <span class="n">proteins</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
<a id="__codelineno-0-147" name="__codelineno-0-147" href="#__codelineno-0-147"></a>                                       <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-148" name="__codelineno-0-148" href="#__codelineno-0-148"></a>                                       <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span><span class="p">,</span>
<a id="__codelineno-0-149" name="__codelineno-0-149" href="#__codelineno-0-149"></a>                                       <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span>
<a id="__codelineno-0-150" name="__codelineno-0-150" href="#__codelineno-0-150"></a>
<a id="__codelineno-0-151" name="__codelineno-0-151" href="#__codelineno-0-151"></a>    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
<a id="__codelineno-0-152" name="__codelineno-0-152" href="#__codelineno-0-152"></a>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Building tensor for the provided context&#39;</span><span class="p">)</span>
<a id="__codelineno-0-153" name="__codelineno-0-153" href="#__codelineno-0-153"></a>
<a id="__codelineno-0-154" name="__codelineno-0-154" href="#__codelineno-0-154"></a>    <span class="n">tensors</span> <span class="o">=</span> <span class="p">[</span><span class="n">generate_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="o">=</span><span class="n">rnaseq</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">),</span>
<a id="__codelineno-0-155" name="__codelineno-0-155" href="#__codelineno-0-155"></a>                                   <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data_</span><span class="p">,</span>
<a id="__codelineno-0-156" name="__codelineno-0-156" href="#__codelineno-0-156"></a>                                   <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-157" name="__codelineno-0-157" href="#__codelineno-0-157"></a>                                   <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">)</span> <span class="k">for</span> <span class="n">rnaseq</span> <span class="ow">in</span> <span class="n">rnaseq_matrices</span><span class="p">]</span>
<a id="__codelineno-0-158" name="__codelineno-0-158" href="#__codelineno-0-158"></a>
<a id="__codelineno-0-159" name="__codelineno-0-159" href="#__codelineno-0-159"></a>    <span class="k">if</span> <span class="n">group_ppi_by</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-160" name="__codelineno-0-160" href="#__codelineno-0-160"></a>        <span class="n">ppi_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">group</span> <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">ppi_data_</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">group_ppi_by</span><span class="p">)]</span>
<a id="__codelineno-0-161" name="__codelineno-0-161" href="#__codelineno-0-161"></a>        <span class="n">tensors</span> <span class="o">=</span> <span class="p">[</span><span class="n">aggregate_ccc_tensor</span><span class="p">(</span><span class="n">ccc_tensor</span><span class="o">=</span><span class="n">t</span><span class="p">,</span>
<a id="__codelineno-0-162" name="__codelineno-0-162" href="#__codelineno-0-162"></a>                                        <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data_</span><span class="p">,</span>
<a id="__codelineno-0-163" name="__codelineno-0-163" href="#__codelineno-0-163"></a>                                        <span class="n">group_ppi_by</span><span class="o">=</span><span class="n">group_ppi_by</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164" href="#__codelineno-0-164"></a>                                        <span class="n">group_ppi_method</span><span class="o">=</span><span class="n">group_ppi_method</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tensors</span><span class="p">]</span>
<a id="__codelineno-0-165" name="__codelineno-0-165" href="#__codelineno-0-165"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-166" name="__codelineno-0-166" href="#__codelineno-0-166"></a>        <span class="n">ppi_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span> <span class="s1">&#39;^&#39;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ppi_data_</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
<a id="__codelineno-0-167" name="__codelineno-0-167" href="#__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168" href="#__codelineno-0-168"></a>    <span class="c1"># Generate mask:</span>
<a id="__codelineno-0-169" name="__codelineno-0-169" href="#__codelineno-0-169"></a>    <span class="k">if</span> <span class="n">how</span> <span class="o">!=</span> <span class="s1">&#39;inner&#39;</span><span class="p">:</span>
<a id="__codelineno-0-170" name="__codelineno-0-170" href="#__codelineno-0-170"></a>        <span class="n">mask_tensor</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">tensors</span><span class="p">)))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<a id="__codelineno-0-171" name="__codelineno-0-171" href="#__codelineno-0-171"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-172" name="__codelineno-0-172" href="#__codelineno-0-172"></a>        <span class="n">mask_tensor</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-173" name="__codelineno-0-173" href="#__codelineno-0-173"></a>    <span class="n">tensors</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">tensors</span><span class="p">)</span>
<a id="__codelineno-0-174" name="__codelineno-0-174" href="#__codelineno-0-174"></a>    <span class="k">return</span> <span class="n">tensors</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">cells</span><span class="p">,</span> <span class="n">ppi_names</span><span class="p">,</span> <span class="n">mask_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.generate_ccc_tensor">
<code class="highlight language-python"><span class="n">generate_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">))</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes a 3D-Communication tensor for a given context based on the gene
expression matrix and the list of PPIS</p>
<h6 id="cell2cell.tensor.tensor.generate_ccc_tensor--parameters">Parameters</h6>
<p>rnaseq_data : pandas.DataFrame
    Gene expression matrix for a given context, sample or condition. Rows are
    genes and columns are cell types/tissues/samples.</p>
<p>ppi_data : pandas.DataFrame
    A dataframe containing protein-protein interactions (rows). It has to
    contain at least two columns, one for the first protein partner in the
    interaction as well as the second protein partner.</p>
<p>communication_score : str, default='expression_mean'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                       of a ligand from a sender cell and the
                       expression of a receptor on a receiver cell.
</code></pre></div>

<p>interaction_columns : tuple, default=('A', 'B')
    Contains the names of the columns where to find the partners in a dataframe of
    protein-protein interactions. If the list is for ligand-receptor pairs, the
    first column is for the ligands and the second for the receptors.</p>
<h6 id="cell2cell.tensor.tensor.generate_ccc_tensor--returns">Returns</h6>
<p>ccc_tensor : ndarray list
    List of directed cell-cell communication matrices, one for each ligand-
    receptor pair in ppi_data. These matrices contain the communication score for
    pairs of cells for the corresponding PPI. This tensor represent a
    3D-communication tensor for the context.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_ccc_tensor</span><span class="p">(</span><span class="n">rnaseq_data</span><span class="p">,</span> <span class="n">ppi_data</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">interaction_columns</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">)):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Computes a 3D-Communication tensor for a given context based on the gene</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    expression matrix and the list of PPIS</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    rnaseq_data : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        Gene expression matrix for a given context, sample or condition. Rows are</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        genes and columns are cell types/tissues/samples.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    ppi_data : pandas.DataFrame</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        A dataframe containing protein-protein interactions (rows). It has to</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        contain at least two columns, one for the first protein partner in the</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        interaction as well as the second protein partner.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    communication_score : str, default=&#39;expression_mean&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">                               of a ligand from a sender cell and the</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                               expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">    interaction_columns : tuple, default=(&#39;A&#39;, &#39;B&#39;)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">        Contains the names of the columns where to find the partners in a dataframe of</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">        protein-protein interactions. If the list is for ligand-receptor pairs, the</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        first column is for the ligands and the second for the receptors.</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    -------</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    ccc_tensor : ndarray list</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        List of directed cell-cell communication matrices, one for each ligand-</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        receptor pair in ppi_data. These matrices contain the communication score for</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        pairs of cells for the corresponding PPI. This tensor represent a</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">        3D-communication tensor for the context.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">ppi_a</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="n">ppi_b</span> <span class="o">=</span> <span class="n">interaction_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="n">ccc_tensor</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">ppi</span> <span class="ow">in</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>        <span class="n">v</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi</span><span class="p">[</span><span class="n">ppi_a</span><span class="p">],</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="n">w</span> <span class="o">=</span> <span class="n">rnaseq_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ppi</span><span class="p">[</span><span class="n">ppi_b</span><span class="p">],</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="n">ccc_tensor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">compute_ccc_matrix</span><span class="p">(</span><span class="n">prot_a_exp</span><span class="o">=</span><span class="n">v</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>                                             <span class="n">prot_b_exp</span><span class="o">=</span><span class="n">w</span><span class="p">,</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                                             <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="k">return</span> <span class="n">ccc_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.generate_tensor_metadata">
<code class="highlight language-python"><span class="n">generate_tensor_metadata</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">metadata_dicts</span><span class="p">,</span> <span class="n">fill_with_order_elements</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Uses a list of of dicts (or None when a dict is missing) to generate a list of
metadata for each order in the tensor.</p>
<h6 id="cell2cell.tensor.tensor.generate_tensor_metadata--parameters">Parameters</h6>
<p>interaction_tensor : cell2cell.tensor.BaseTensor
    A communication tensor.</p>
<p>metadata_dicts : list
    A list of dictionaries. Each dictionary represents an order of the tensor. In
    an interaction tensor these orders should be contexts, LR pairs, sender cells
    and receiver cells. The keys are the elements in each order (they are
    contained in interaction_tensor.order_names) and the values are the categories
    that each elements will be assigned as metadata.</p>
<p>fill_with_order_elements : boolean, default=True
    Whether using each element of a dimension as its own metadata when a None is
    passed instead of a dictionary for the respective order/dimension. If True,
    each element in that order will be use itself, that dimension will not contain
    metadata.</p>
<h6 id="cell2cell.tensor.tensor.generate_tensor_metadata--returns">Returns</h6>
<p>metadata : list
    A list of pandas.DataFrames that will be used as an input of the
    cell2cell.plot.tensor_factors_plot.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_tensor_metadata</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="p">,</span> <span class="n">metadata_dicts</span><span class="p">,</span> <span class="n">fill_with_order_elements</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;Uses a list of of dicts (or None when a dict is missing) to generate a list of</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    metadata for each order in the tensor.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensor : cell2cell.tensor.BaseTensor</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        A communication tensor.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    metadata_dicts : list</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        A list of dictionaries. Each dictionary represents an order of the tensor. In</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        an interaction tensor these orders should be contexts, LR pairs, sender cells</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        and receiver cells. The keys are the elements in each order (they are</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        contained in interaction_tensor.order_names) and the values are the categories</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        that each elements will be assigned as metadata.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    fill_with_order_elements : boolean, default=True</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        Whether using each element of a dimension as its own metadata when a None is</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        passed instead of a dictionary for the respective order/dimension. If True,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        each element in that order will be use itself, that dimension will not contain</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        metadata.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">    -------</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">    metadata : list</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        A list of pandas.DataFrames that will be used as an input of the</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">        cell2cell.plot.tensor_factors_plot.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>    <span class="n">tensor_dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>    <span class="k">assert</span> <span class="p">(</span><span class="n">tensor_dim</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">metadata_dicts</span><span class="p">)),</span> <span class="s2">&quot;metadata_dicts should be of the same size as the number of orders/dimensions in the tensor&quot;</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="k">if</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>        <span class="k">if</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>            <span class="n">default_cats</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts&#39;</span><span class="p">,</span> <span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>            <span class="n">default_cats</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Contexts-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">tensor_dim</span> <span class="o">-</span> <span class="mi">3</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>        <span class="k">elif</span> <span class="n">tensor_dim</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>            <span class="n">default_cats</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Ligand-Receptor Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;Sender Cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Receiver Cells&#39;</span><span class="p">]</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Too few dimensions in the tensor&#39;</span><span class="p">)</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">tensor_dim</span><span class="p">,</span> <span class="s2">&quot;The length of order_labels must match the number of orders/dimensions in the tensor&quot;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>        <span class="n">default_cats</span> <span class="o">=</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_labels</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">if</span> <span class="n">fill_with_order_elements</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">metadata</span> <span class="o">=</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">names</span><span class="p">)</span> <span class="k">for</span> <span class="n">names</span> <span class="ow">in</span> <span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">]</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">metadata</span> <span class="o">=</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">names</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">meta</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">names</span><span class="p">,</span> <span class="n">meta</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">,</span> <span class="n">metadata_dicts</span><span class="p">)]</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">meta</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metadata</span><span class="p">):</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>        <span class="k">if</span> <span class="n">meta</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>            <span class="k">if</span> <span class="n">metadata_dicts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>                <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">metadata_dicts</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>            <span class="k">else</span><span class="p">:</span> <span class="c1"># dict is None and fill with order elements TRUE</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">75</span><span class="p">:</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>                    <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">index</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>                <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>                    <span class="n">meta</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">*</span> <span class="p">[</span><span class="n">default_cats</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>            <span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Element&#39;</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>            <span class="n">meta</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="k">return</span> <span class="n">metadata</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor.interactions_to_tensor">
<code class="highlight language-python"><span class="n">interactions_to_tensor</span><span class="p">(</span><span class="n">interactions</span><span class="p">,</span> <span class="n">experiment</span><span class="o">=</span><span class="s1">&#39;single_cell&#39;</span><span class="p">,</span> <span class="n">context_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Takes a list of Interaction pipelines (see classes in
cell2cell.analysis.pipelines) and generates a communication
tensor.</p>
<h6 id="cell2cell.tensor.tensor.interactions_to_tensor--parameters">Parameters</h6>
<p>interactions : list
    List of Interaction pipelines. The Interactions has to be all either
    BulkInteractions or SingleCellInteractions.</p>
<p>experiment : str, default='single_cell'
    Type of Interaction pipelines in the list. Either 'single_cell' or 'bulk'.</p>
<p>context_names : list
    List of context names or labels for each of the Interaction pipelines. This
    list matches the length of interactions and the labels have to follows the
    same order.</p>
<p>how : str, default='inner'
    Approach to consider cell types and genes present across multiple contexts.</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="s1">&#39;</span><span class="s">inner</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
            <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">and</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span>
            <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_genes</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
<span class="o">-</span> <span class="s1">&#39;</span><span class="s">outer_cells</span><span class="s1">&#39;</span> : <span class="nv">Considers</span> <span class="nv">only</span> <span class="nv">genes</span> <span class="nv">that</span> <span class="nv">are</span> <span class="nv">present</span> <span class="nv">in</span> <span class="nv">all</span>
                  <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">intersection</span><span class="ss">)</span>, <span class="k">while</span> <span class="nv">all</span> <span class="nv">cell</span> <span class="nv">types</span> <span class="nv">that</span> <span class="nv">are</span>
                  <span class="nv">present</span> <span class="nv">across</span> <span class="nv">contexts</span> <span class="ss">(</span><span class="nv">union</span><span class="ss">)</span>.
</code></pre></div>

<p>outer_fraction : float, default=0.0
    Threshold to filter the elements when <code>how</code> includes any outer option.
    Elements with a fraction abundance across samples at least this
    threshold will be included. When this value is 0, considers
    all elements across the samples. When this value is 1, it acts as using
    <code>how='inner'</code>.</p>
<p>communication_score : str, default='expression_mean'
    Type of communication score to infer the potential use of a given ligand-
    receptor pair by a pair of cells/tissues/samples.
    Available communication_scores are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;expression_mean&#39; : Computes the average between the expression of a ligand
                      from a sender cell and the expression of a receptor on a
                      receiver cell.
- &#39;expression_product&#39; : Computes the product between the expression of a
                        ligand from a sender cell and the expression of a
                        receptor on a receiver cell.
- &#39;expression_gmean&#39; : Computes the geometric mean between the expression
                       of a ligand from a sender cell and the
                       expression of a receptor on a receiver cell.
</code></pre></div>

<p>upper_letter_comparison : boolean, default=True
    Whether making uppercase the gene names in the expression matrices and the
    protein names in the ppi_data to match their names and integrate their
    respective expression level. Useful when there are inconsistencies in the
    names between the expression matrix and the ligand-receptor annotations.</p>
<h6 id="cell2cell.tensor.tensor.interactions_to_tensor--returns">Returns</h6>
<p>tensor : cell2cell.tensor.InteractionTensor
    A 4D-communication tensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">interactions_to_tensor</span><span class="p">(</span><span class="n">interactions</span><span class="p">,</span> <span class="n">experiment</span><span class="o">=</span><span class="s1">&#39;single_cell&#39;</span><span class="p">,</span> <span class="n">context_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">outer_fraction</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                           <span class="n">communication_score</span><span class="o">=</span><span class="s1">&#39;expression_product&#39;</span><span class="p">,</span> <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Takes a list of Interaction pipelines (see classes in</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    cell2cell.analysis.pipelines) and generates a communication</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    tensor.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    interactions : list</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        List of Interaction pipelines. The Interactions has to be all either</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        BulkInteractions or SingleCellInteractions.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    experiment : str, default=&#39;single_cell&#39;</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Type of Interaction pipelines in the list. Either &#39;single_cell&#39; or &#39;bulk&#39;.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    context_names : list</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        List of context names or labels for each of the Interaction pipelines. This</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">        list matches the length of interactions and the labels have to follows the</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        same order.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">    how : str, default=&#39;inner&#39;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="sd">        Approach to consider cell types and genes present across multiple contexts.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        - &#39;inner&#39; : Considers only cell types and genes that are present in all</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">                    contexts (intersection).</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">        - &#39;outer&#39; : Considers all cell types and genes that are present</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">                    across contexts (union).</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        - &#39;outer_genes&#39; : Considers only cell types that are present in all</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">                          contexts (intersection), while all genes that are</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        - &#39;outer_cells&#39; : Considers only genes that are present in all</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a><span class="sd">                          contexts (intersection), while all cell types that are</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">                          present across contexts (union).</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">    outer_fraction : float, default=0.0</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a><span class="sd">        Threshold to filter the elements when `how` includes any outer option.</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">        Elements with a fraction abundance across samples at least this</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">        threshold will be included. When this value is 0, considers</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">        all elements across the samples. When this value is 1, it acts as using</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        `how=&#39;inner&#39;`.</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    communication_score : str, default=&#39;expression_mean&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a><span class="sd">        Type of communication score to infer the potential use of a given ligand-</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a><span class="sd">        receptor pair by a pair of cells/tissues/samples.</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a><span class="sd">        Available communication_scores are:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a><span class="sd">        - &#39;expression_mean&#39; : Computes the average between the expression of a ligand</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a><span class="sd">                              from a sender cell and the expression of a receptor on a</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a><span class="sd">                              receiver cell.</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a><span class="sd">        - &#39;expression_product&#39; : Computes the product between the expression of a</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a><span class="sd">                                ligand from a sender cell and the expression of a</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a><span class="sd">                                receptor on a receiver cell.</span>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a><span class="sd">        - &#39;expression_gmean&#39; : Computes the geometric mean between the expression</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a><span class="sd">                               of a ligand from a sender cell and the</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a><span class="sd">                               expression of a receptor on a receiver cell.</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a><span class="sd">    upper_letter_comparison : boolean, default=True</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a><span class="sd">        Whether making uppercase the gene names in the expression matrices and the</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a><span class="sd">        protein names in the ppi_data to match their names and integrate their</span>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a><span class="sd">        respective expression level. Useful when there are inconsistencies in the</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a><span class="sd">        names between the expression matrix and the ligand-receptor annotations.</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a><span class="sd">    -------</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a><span class="sd">    tensor : cell2cell.tensor.InteractionTensor</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a><span class="sd">        A 4D-communication tensor.</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>    <span class="n">ppis</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="n">rnaseq_matrices</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="n">complex_sep</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">complex_sep</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>    <span class="n">complex_agg_method</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">complex_agg_method</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>    <span class="n">interaction_columns</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">interaction_columns</span>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="k">for</span> <span class="n">Int_</span> <span class="ow">in</span> <span class="n">interactions</span><span class="p">:</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>        <span class="k">if</span> <span class="n">Int_</span><span class="o">.</span><span class="n">analysis_setup</span><span class="p">[</span><span class="s1">&#39;cci_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;undirected&#39;</span><span class="p">:</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>            <span class="n">ppis</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Int_</span><span class="o">.</span><span class="n">ref_ppi</span><span class="p">)</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>            <span class="n">ppis</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Int_</span><span class="o">.</span><span class="n">ppi_data</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>        <span class="k">if</span> <span class="n">experiment</span> <span class="o">==</span> <span class="s1">&#39;single_cell&#39;</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>            <span class="n">rnaseq_matrices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Int_</span><span class="o">.</span><span class="n">aggregated_expression</span><span class="p">)</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>        <span class="k">elif</span> <span class="n">experiment</span> <span class="o">==</span> <span class="s1">&#39;bulk&#39;</span><span class="p">:</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>            <span class="n">rnaseq_matrices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Int_</span><span class="o">.</span><span class="n">rnaseq_data</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;experiment must be &#39;single_cell&#39; or &#39;bulk&#39;&quot;</span><span class="p">)</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>    <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">ppis</span><span class="p">)</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>    <span class="n">ppi_data</span> <span class="o">=</span> <span class="n">ppi_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>    <span class="n">tensor</span> <span class="o">=</span> <span class="n">InteractionTensor</span><span class="p">(</span><span class="n">rnaseq_matrices</span><span class="o">=</span><span class="n">rnaseq_matrices</span><span class="p">,</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>                               <span class="n">ppi_data</span><span class="o">=</span><span class="n">ppi_data</span><span class="p">,</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>                               <span class="n">context_names</span><span class="o">=</span><span class="n">context_names</span><span class="p">,</span>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>                               <span class="n">how</span><span class="o">=</span><span class="n">how</span><span class="p">,</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>                               <span class="n">outer_fraction</span><span class="o">=</span><span class="n">outer_fraction</span><span class="p">,</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>                               <span class="n">complex_sep</span><span class="o">=</span><span class="n">complex_sep</span><span class="p">,</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>                               <span class="n">complex_agg_method</span><span class="o">=</span><span class="n">complex_agg_method</span><span class="p">,</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>                               <span class="n">interaction_columns</span><span class="o">=</span><span class="n">interaction_columns</span><span class="p">,</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>                               <span class="n">communication_score</span><span class="o">=</span><span class="n">communication_score</span><span class="p">,</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                               <span class="n">upper_letter_comparison</span><span class="o">=</span><span class="n">upper_letter_comparison</span><span class="p">,</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>                               <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span>
<a id="__codelineno-0-101" name="__codelineno-0-101" href="#__codelineno-0-101"></a>                               <span class="p">)</span>
<a id="__codelineno-0-102" name="__codelineno-0-102" href="#__codelineno-0-102"></a>    <span class="k">return</span> <span class="n">tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.tensor.tensor_manipulation">
        <code>tensor_manipulation</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.tensor.tensor_manipulation.concatenate_interaction_tensors">
<code class="highlight language-python"><span class="n">concatenate_interaction_tensors</span><span class="p">(</span><span class="n">interaction_tensors</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">order_labels</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Concatenates interaction tensors in a given tensor dimension or axis.</p>
<h6 id="cell2cell.tensor.tensor_manipulation.concatenate_interaction_tensors--parameters">Parameters</h6>
<p>interaction_tensors : list
    List of any tensor class in cell2cell.tensor.</p>
<p>axis : int
    The axis along which the arrays will be joined. If axis is None, arrays are flattened before use.</p>
<p>order_labels : list
    List of labels for dimensions or orders in the tensor.</p>
<p>remove_duplicates : boolean, default=False
    Whether removing duplicated names in the concatenated axis.</p>
<p>keep : str, default='first'
    Determines which duplicates (if any) to keep.
    Options are:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span> <span class="nv">first</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">first</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">last</span> : <span class="nv">Drop</span> <span class="nv">duplicates</span> <span class="nv">except</span> <span class="k">for</span> <span class="nv">the</span> <span class="nv">last</span> <span class="nv">occurrence</span>.
<span class="o">-</span> <span class="nv">False</span> : <span class="nv">Drop</span> <span class="nv">all</span> <span class="nv">duplicates</span>.
</code></pre></div>

<p>mask : ndarray list
    Helps avoiding missing values during a tensor factorization. A mask should be
    a boolean array of the same shape as the original tensor and should be 0
    where the values are missing and 1 everywhere else. This must be of equal shape
    as the concatenated tensor.</p>
<p None="None" _cpu_="'cpu'," _cuda_="'cuda',">device : str, default=None
    Device to use when backend is pytorch. Options are:</p>
<h6 id="cell2cell.tensor.tensor_manipulation.concatenate_interaction_tensors--returns">Returns</h6>
<p>concatenated_tensor : cell2cell.tensor.PreBuiltTensor
    Final tensor after concatenation. It is a PreBuiltTensor that works
    any interaction tensor based on the class BaseTensor.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/tensor/tensor_manipulation.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">concatenate_interaction_tensors</span><span class="p">(</span><span class="n">interaction_tensors</span><span class="p">,</span> <span class="n">axis</span><span class="p">,</span> <span class="n">order_labels</span><span class="p">,</span> <span class="n">remove_duplicates</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">,</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>                                    <span class="n">mask</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    <span class="sd">&#39;&#39;&#39;Concatenates interaction tensors in a given tensor dimension or axis.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    interaction_tensors : list</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        List of any tensor class in cell2cell.tensor.</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">    axis : int</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        The axis along which the arrays will be joined. If axis is None, arrays are flattened before use.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    order_labels : list</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        List of labels for dimensions or orders in the tensor.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    remove_duplicates : boolean, default=False</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Whether removing duplicated names in the concatenated axis.</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">    keep : str, default=&#39;first&#39;</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        Determines which duplicates (if any) to keep.</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        Options are:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">        - first : Drop duplicates except for the first occurrence.</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        - last : Drop duplicates except for the last occurrence.</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        - False : Drop all duplicates.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a><span class="sd">    mask : ndarray list</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a><span class="sd">        Helps avoiding missing values during a tensor factorization. A mask should be</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a><span class="sd">        a boolean array of the same shape as the original tensor and should be 0</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a><span class="sd">        where the values are missing and 1 everywhere else. This must be of equal shape</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a><span class="sd">        as the concatenated tensor.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a><span class="sd">    device : str, default=None</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a><span class="sd">        Device to use when backend is pytorch. Options are:</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a><span class="sd">         {&#39;cpu&#39;, &#39;cuda&#39;, None}</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a><span class="sd">    -------</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a><span class="sd">    concatenated_tensor : cell2cell.tensor.PreBuiltTensor</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a><span class="sd">        Final tensor after concatenation. It is a PreBuiltTensor that works</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a><span class="sd">        any interaction tensor based on the class BaseTensor.</span>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="c1"># Assert if all other dimensions contains the same elements:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>    <span class="n">shape</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">shape</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">1</span><span class="p">:]),</span> <span class="s2">&quot;Tensors must have same number of dimensions&quot;</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">shape</span><span class="p">):</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">axis</span><span class="p">:</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>            <span class="n">elements</span> <span class="o">=</span> <span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>            <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
<a id="__codelineno-0-51" name="__codelineno-0-51" href="#__codelineno-0-51"></a>                <span class="k">assert</span> <span class="n">elements</span> <span class="o">==</span> <span class="n">tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;Tensors must have the same elements in the other axes.&quot;</span>
<a id="__codelineno-0-52" name="__codelineno-0-52" href="#__codelineno-0-52"></a>
<a id="__codelineno-0-53" name="__codelineno-0-53" href="#__codelineno-0-53"></a>    <span class="c1"># Initialize tensors into a numpy object for performing subset</span>
<a id="__codelineno-0-54" name="__codelineno-0-54" href="#__codelineno-0-54"></a>    <span class="c1"># Use the same context as first tensor for everything</span>
<a id="__codelineno-0-55" name="__codelineno-0-55" href="#__codelineno-0-55"></a>    <span class="k">try</span><span class="p">:</span>
<a id="__codelineno-0-56" name="__codelineno-0-56" href="#__codelineno-0-56"></a>        <span class="n">context</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">context</span><span class="p">(</span><span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tensor</span><span class="p">)</span>
<a id="__codelineno-0-57" name="__codelineno-0-57" href="#__codelineno-0-57"></a>    <span class="k">except</span><span class="p">:</span>
<a id="__codelineno-0-58" name="__codelineno-0-58" href="#__codelineno-0-58"></a>        <span class="n">context</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;dtype&#39;</span><span class="p">:</span> <span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span> <span class="s1">&#39;device&#39;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<a id="__codelineno-0-59" name="__codelineno-0-59" href="#__codelineno-0-59"></a>
<a id="__codelineno-0-60" name="__codelineno-0-60" href="#__codelineno-0-60"></a>    <span class="c1"># Concatenate tensors</span>
<a id="__codelineno-0-61" name="__codelineno-0-61" href="#__codelineno-0-61"></a>    <span class="n">concat_tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">tensor</span><span class="o">.</span><span class="n">tensor</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;cpu&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
<a id="__codelineno-0-62" name="__codelineno-0-62" href="#__codelineno-0-62"></a>    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63" href="#__codelineno-0-63"></a>        <span class="k">assert</span> <span class="n">mask</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">concat_tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="s2">&quot;Mask must have the same shape of the concatenated tensor. Here: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">concat_tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64" href="#__codelineno-0-64"></a>    <span class="k">else</span><span class="p">:</span> <span class="c1"># Generate a new mask from all previous masks if all are not None</span>
<a id="__codelineno-0-65" name="__codelineno-0-65" href="#__codelineno-0-65"></a>        <span class="k">if</span> <span class="nb">all</span><span class="p">([</span><span class="n">tensor</span><span class="o">.</span><span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">]):</span>
<a id="__codelineno-0-66" name="__codelineno-0-66" href="#__codelineno-0-66"></a>            <span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">tensor</span><span class="o">.</span><span class="n">mask</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;cpu&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
<a id="__codelineno-0-67" name="__codelineno-0-67" href="#__codelineno-0-67"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68" href="#__codelineno-0-68"></a>            <span class="n">mask</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-69" name="__codelineno-0-69" href="#__codelineno-0-69"></a>
<a id="__codelineno-0-70" name="__codelineno-0-70" href="#__codelineno-0-70"></a>    <span class="n">concat_tensor</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">concat_tensor</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-71" name="__codelineno-0-71" href="#__codelineno-0-71"></a>    <span class="k">if</span> <span class="n">mask</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-72" name="__codelineno-0-72" href="#__codelineno-0-72"></a>        <span class="n">mask</span> <span class="o">=</span> <span class="n">tl</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">context</span><span class="p">[</span><span class="s1">&#39;device&#39;</span><span class="p">])</span>
<a id="__codelineno-0-73" name="__codelineno-0-73" href="#__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74" href="#__codelineno-0-74"></a>    <span class="c1"># Concatenate names of elements for the given axis but keep the others as in one tensor</span>
<a id="__codelineno-0-75" name="__codelineno-0-75" href="#__codelineno-0-75"></a>    <span class="n">order_names</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-76" name="__codelineno-0-76" href="#__codelineno-0-76"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">shape</span><span class="p">):</span>
<a id="__codelineno-0-77" name="__codelineno-0-77" href="#__codelineno-0-77"></a>        <span class="n">tmp_names</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-78" name="__codelineno-0-78" href="#__codelineno-0-78"></a>        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">axis</span><span class="p">:</span>
<a id="__codelineno-0-79" name="__codelineno-0-79" href="#__codelineno-0-79"></a>            <span class="k">for</span> <span class="n">tensor</span> <span class="ow">in</span> <span class="n">interaction_tensors</span><span class="p">:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80" href="#__codelineno-0-80"></a>                <span class="n">tmp_names</span> <span class="o">+=</span> <span class="n">tensor</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-81" name="__codelineno-0-81" href="#__codelineno-0-81"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-82" name="__codelineno-0-82" href="#__codelineno-0-82"></a>            <span class="n">tmp_names</span> <span class="o">=</span> <span class="n">interaction_tensors</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">order_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-83" name="__codelineno-0-83" href="#__codelineno-0-83"></a>        <span class="n">order_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_names</span><span class="p">)</span>
<a id="__codelineno-0-84" name="__codelineno-0-84" href="#__codelineno-0-84"></a>
<a id="__codelineno-0-85" name="__codelineno-0-85" href="#__codelineno-0-85"></a>    <span class="c1"># Generate final object</span>
<a id="__codelineno-0-86" name="__codelineno-0-86" href="#__codelineno-0-86"></a>    <span class="n">concatenated_tensor</span> <span class="o">=</span> <span class="n">PreBuiltTensor</span><span class="p">(</span><span class="n">tensor</span><span class="o">=</span><span class="n">concat_tensor</span><span class="p">,</span>
<a id="__codelineno-0-87" name="__codelineno-0-87" href="#__codelineno-0-87"></a>                                         <span class="n">order_names</span><span class="o">=</span><span class="n">order_names</span><span class="p">,</span>
<a id="__codelineno-0-88" name="__codelineno-0-88" href="#__codelineno-0-88"></a>                                         <span class="n">order_labels</span><span class="o">=</span><span class="n">order_labels</span><span class="p">,</span>
<a id="__codelineno-0-89" name="__codelineno-0-89" href="#__codelineno-0-89"></a>                                         <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">,</span>  <span class="c1"># Change if you want to omit values in the decomposition</span>
<a id="__codelineno-0-90" name="__codelineno-0-90" href="#__codelineno-0-90"></a>                                         <span class="n">device</span><span class="o">=</span><span class="n">device</span>
<a id="__codelineno-0-91" name="__codelineno-0-91" href="#__codelineno-0-91"></a>                                        <span class="p">)</span>
<a id="__codelineno-0-92" name="__codelineno-0-92" href="#__codelineno-0-92"></a>
<a id="__codelineno-0-93" name="__codelineno-0-93" href="#__codelineno-0-93"></a>    <span class="c1"># Remove duplicates</span>
<a id="__codelineno-0-94" name="__codelineno-0-94" href="#__codelineno-0-94"></a>    <span class="k">if</span> <span class="n">remove_duplicates</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95" href="#__codelineno-0-95"></a>        <span class="n">concatenated_tensor</span> <span class="o">=</span> <span class="n">subset_tensor</span><span class="p">(</span><span class="n">interaction_tensor</span><span class="o">=</span><span class="n">concatenated_tensor</span><span class="p">,</span>
<a id="__codelineno-0-96" name="__codelineno-0-96" href="#__codelineno-0-96"></a>                                            <span class="n">subset_dict</span><span class="o">=</span><span class="p">{</span><span class="n">axis</span><span class="p">:</span> <span class="n">order_names</span><span class="p">[</span><span class="n">axis</span><span class="p">]},</span>
<a id="__codelineno-0-97" name="__codelineno-0-97" href="#__codelineno-0-97"></a>                                            <span class="n">remove_duplicates</span><span class="o">=</span><span class="n">remove_duplicates</span><span class="p">,</span>
<a id="__codelineno-0-98" name="__codelineno-0-98" href="#__codelineno-0-98"></a>                                            <span class="n">keep</span><span class="o">=</span><span class="n">keep</span><span class="p">,</span>
<a id="__codelineno-0-99" name="__codelineno-0-99" href="#__codelineno-0-99"></a>                                            <span class="n">original_order</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-100" name="__codelineno-0-100" href="#__codelineno-0-100"></a>    <span class="k">return</span> <span class="n">concatenated_tensor</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="cell2cell.utils">
        <code>utils</code>


  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.utils.networks">
        <code>networks</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.utils.networks.export_network_to_cytoscape">
<code class="highlight language-python"><span class="n">export_network_to_cytoscape</span><span class="p">(</span><span class="n">network</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Exports a network into a spreadsheet that is readable
by the software Gephi.</p>
<h6 id="cell2cell.utils.networks.export_network_to_cytoscape--parameters">Parameters</h6>
<p>network : networkx.Graph, networkx.DiGraph or a pandas.DataFrame
    A networkx Graph or Directed Graph, or an adjacency matrix,
    where in rows and columns are nodes and values represents a
    weight for the respective edge.</p>
<p>filename : str, default=None
    Path to save the network into a Cytoscape-readable format
    (JSON file in this case). E.g. '/home/user/network.json'</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/utils/networks.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">export_network_to_cytoscape</span><span class="p">(</span><span class="n">network</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Exports a network into a spreadsheet that is readable</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    by the software Gephi.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    network : networkx.Graph, networkx.DiGraph or a pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        A networkx Graph or Directed Graph, or an adjacency matrix,</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        where in rows and columns are nodes and values represents a</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        weight for the respective edge.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Path to save the network into a Cytoscape-readable format</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        (JSON file in this case). E.g. &#39;/home/user/network.json&#39;</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="c1"># This allows to pass a network directly or an adjacency matrix</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">network</span><span class="p">)</span> <span class="o">!=</span> <span class="n">nx</span><span class="o">.</span><span class="n">classes</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="n">network</span> <span class="o">=</span> <span class="n">generate_network_from_adjacency</span><span class="p">(</span><span class="n">network</span><span class="p">,</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>                                                  <span class="n">package</span><span class="o">=</span><span class="s1">&#39;networkx&#39;</span><span class="p">)</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>    <span class="n">data</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">readwrite</span><span class="o">.</span><span class="n">json_graph</span><span class="o">.</span><span class="n">cytoscape</span><span class="o">.</span><span class="n">cytoscape_data</span><span class="p">(</span><span class="n">network</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="c1"># Export</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>    <span class="kn">import</span> <span class="nn">json</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="n">json_str</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json_str</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.utils.networks.export_network_to_gephi">
<code class="highlight language-python"><span class="n">export_network_to_gephi</span><span class="p">(</span><span class="n">network</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;excel&#39;</span><span class="p">,</span> <span class="n">network_type</span><span class="o">=</span><span class="s1">&#39;Undirected&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Exports a network into a spreadsheet that is readable
by the software Gephi.</p>
<h6 id="cell2cell.utils.networks.export_network_to_gephi--parameters">Parameters</h6>
<p>network : networkx.Graph, networkx.DiGraph or a pandas.DataFrame
    A networkx Graph or Directed Graph, or an adjacency matrix,
    where in rows and columns are nodes and values represents a
    weight for the respective edge.</p>
<p>filename : str, default=None
    Path to save the network into a Gephi-readable format.</p>
<p>format : str, default='excel'
    Format to export the spreadsheet. Options are:</p>
<div class="codehilite"><pre><span></span><code>- &#39;excel&#39; : An excel file, either .xls or .xlsx
- &#39;csv&#39; : Comma separated value format
- &#39;tsv&#39; : Tab separated value format
</code></pre></div>

<p>network_type : str, default='Undirected'
    Type of edges in the network. They could be either
    'Undirected' or 'Directed'.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/utils/networks.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">export_network_to_gephi</span><span class="p">(</span><span class="n">network</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;excel&#39;</span><span class="p">,</span> <span class="n">network_type</span><span class="o">=</span><span class="s1">&#39;Undirected&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Exports a network into a spreadsheet that is readable</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    by the software Gephi.</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    network : networkx.Graph, networkx.DiGraph or a pandas.DataFrame</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        A networkx Graph or Directed Graph, or an adjacency matrix,</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        where in rows and columns are nodes and values represents a</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">        weight for the respective edge.</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    filename : str, default=None</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        Path to save the network into a Gephi-readable format.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    format : str, default=&#39;excel&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">        Format to export the spreadsheet. Options are:</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        - &#39;excel&#39; : An excel file, either .xls or .xlsx</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">        - &#39;csv&#39; : Comma separated value format</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="sd">        - &#39;tsv&#39; : Tab separated value format</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="sd">    network_type : str, default=&#39;Undirected&#39;</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a><span class="sd">        Type of edges in the network. They could be either</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a><span class="sd">        &#39;Undirected&#39; or &#39;Directed&#39;.</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>    <span class="c1"># This allows to pass a network directly or an adjacency matrix</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">network</span><span class="p">)</span> <span class="o">!=</span> <span class="n">nx</span><span class="o">.</span><span class="n">classes</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="n">network</span> <span class="o">=</span> <span class="n">generate_network_from_adjacency</span><span class="p">(</span><span class="n">network</span><span class="p">,</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>                                                  <span class="n">package</span><span class="o">=</span><span class="s1">&#39;networkx&#39;</span><span class="p">)</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>    <span class="n">gephi_df</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">to_pandas_edgelist</span><span class="p">(</span><span class="n">network</span><span class="p">)</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="n">gephi_df</span> <span class="o">=</span> <span class="n">gephi_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Type</span><span class="o">=</span><span class="n">network_type</span><span class="p">)</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>    <span class="c1"># When weight is not in the network</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;weight&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gephi_df</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
<a id="__codelineno-0-36" name="__codelineno-0-36" href="#__codelineno-0-36"></a>        <span class="n">gephi_df</span> <span class="o">=</span> <span class="n">gephi_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-37" name="__codelineno-0-37" href="#__codelineno-0-37"></a>
<a id="__codelineno-0-38" name="__codelineno-0-38" href="#__codelineno-0-38"></a>    <span class="c1"># Transform column names</span>
<a id="__codelineno-0-39" name="__codelineno-0-39" href="#__codelineno-0-39"></a>    <span class="n">gephi_df</span> <span class="o">=</span> <span class="n">gephi_df</span><span class="p">[[</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span> <span class="s1">&#39;weight&#39;</span><span class="p">]]</span>
<a id="__codelineno-0-40" name="__codelineno-0-40" href="#__codelineno-0-40"></a>    <span class="n">gephi_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">gephi_df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<a id="__codelineno-0-41" name="__codelineno-0-41" href="#__codelineno-0-41"></a>
<a id="__codelineno-0-42" name="__codelineno-0-42" href="#__codelineno-0-42"></a>    <span class="c1"># Save with different formats</span>
<a id="__codelineno-0-43" name="__codelineno-0-43" href="#__codelineno-0-43"></a>    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;excel&#39;</span><span class="p">:</span>
<a id="__codelineno-0-44" name="__codelineno-0-44" href="#__codelineno-0-44"></a>        <span class="n">gephi_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Edges&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-45" name="__codelineno-0-45" href="#__codelineno-0-45"></a>    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;csv&#39;</span><span class="p">:</span>
<a id="__codelineno-0-46" name="__codelineno-0-46" href="#__codelineno-0-46"></a>        <span class="n">gephi_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-47" name="__codelineno-0-47" href="#__codelineno-0-47"></a>    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;tsv&#39;</span><span class="p">:</span>
<a id="__codelineno-0-48" name="__codelineno-0-48" href="#__codelineno-0-48"></a>        <span class="n">gephi_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-0-49" name="__codelineno-0-49" href="#__codelineno-0-49"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-50" name="__codelineno-0-50" href="#__codelineno-0-50"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Format not supported.&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.utils.networks.generate_network_from_adjacency">
<code class="highlight language-python"><span class="n">generate_network_from_adjacency</span><span class="p">(</span><span class="n">adjacency_matrix</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s1">&#39;networkx&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Generates a network or graph object from an adjacency matrix.</p>
<h6 id="cell2cell.utils.networks.generate_network_from_adjacency--parameters">Parameters</h6>
<p>adjacency_matrix : pandas.DataFrame
    An adjacency matrix, where in rows and columns are nodes
    and values represents a weight for the respective edge.</p>
<p>package : str, default='networkx'
    Package or python library to built the network.
    Implemented optios are {'networkx'}. Soon will be
    available for 'igraph'.</p>
<h6 id="cell2cell.utils.networks.generate_network_from_adjacency--returns">Returns</h6>
<p>network : graph-like
    A graph object built with a python-library for networks.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/utils/networks.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">generate_network_from_adjacency</span><span class="p">(</span><span class="n">adjacency_matrix</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s1">&#39;networkx&#39;</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Generates a network or graph object from an adjacency matrix.</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    adjacency_matrix : pandas.DataFrame</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">        An adjacency matrix, where in rows and columns are nodes</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">        and values represents a weight for the respective edge.</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="sd">    package : str, default=&#39;networkx&#39;</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">        Package or python library to built the network.</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">        Implemented optios are {&#39;networkx&#39;}. Soon will be</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">        available for &#39;igraph&#39;.</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="sd">    -------</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="sd">    network : graph-like</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="sd">        A graph object built with a python-library for networks.</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">if</span> <span class="n">package</span> <span class="o">==</span> <span class="s1">&#39;networkx&#39;</span><span class="p">:</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>        <span class="n">network</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_pandas_adjacency</span><span class="p">(</span><span class="n">adjacency_matrix</span><span class="p">)</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>    <span class="k">elif</span> <span class="n">package</span> <span class="o">==</span> <span class="s1">&#39;igraph&#39;</span><span class="p">:</span>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>        <span class="c1"># A = adjacency_matrix.values</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="c1"># network = igraph.Graph.Weighted_Adjacency((A &gt; 0).tolist(), mode=igraph.ADJ_UNDIRECTED)</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>        <span class="c1">#</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="c1"># # Add edge weights and node labels.</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>        <span class="c1"># network.es[&#39;weight&#39;] = A[A.nonzero()]</span>
<a id="__codelineno-0-29" name="__codelineno-0-29" href="#__codelineno-0-29"></a>        <span class="c1"># network.vs[&#39;label&#39;] = list(adjacency_matrix.columns)</span>
<a id="__codelineno-0-30" name="__codelineno-0-30" href="#__codelineno-0-30"></a>        <span class="c1">#</span>
<a id="__codelineno-0-31" name="__codelineno-0-31" href="#__codelineno-0-31"></a>        <span class="c1"># Warning(&quot;iGraph functionalities are not completely implemented yet.&quot;)</span>
<a id="__codelineno-0-32" name="__codelineno-0-32" href="#__codelineno-0-32"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Network using package </span><span class="si">{}</span><span class="s2"> not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">package</span><span class="p">))</span>
<a id="__codelineno-0-33" name="__codelineno-0-33" href="#__codelineno-0-33"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-34" name="__codelineno-0-34" href="#__codelineno-0-34"></a>        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Network using package </span><span class="si">{}</span><span class="s2"> not implemented&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">package</span><span class="p">))</span>
<a id="__codelineno-0-35" name="__codelineno-0-35" href="#__codelineno-0-35"></a>    <span class="k">return</span> <span class="n">network</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h3 class="doc doc-heading" id="cell2cell.utils.parallel_computing">
        <code>parallel_computing</code>



</h3>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.utils.parallel_computing.agents_number">
<code class="highlight language-python"><span class="n">agents_number</span><span class="p">(</span><span class="n">n_jobs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Computes the number of agents/cores/threads that the
computer can really provide given a number of
jobs/threads requested.</p>
<h6 id="cell2cell.utils.parallel_computing.agents_number--parameters">Parameters</h6>
<p>n_jobs : int
    Number of threads for parallelization.</p>
<h6 id="cell2cell.utils.parallel_computing.agents_number--returns">Returns</h6>
<p>agents : int
    Number of threads that the computer can really provide.</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/utils/parallel_computing.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">agents_number</span><span class="p">(</span><span class="n">n_jobs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Computes the number of agents/cores/threads that the</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    computer can really provide given a number of</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="sd">    jobs/threads requested.</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="sd">    Parameters</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="sd">    ----------</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="sd">    n_jobs : int</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="sd">        Number of threads for parallelization.</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="sd">    Returns</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="sd">    -------</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="sd">    agents : int</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="sd">        Number of threads that the computer can really provide.</span>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>    <span class="k">if</span> <span class="n">n_jobs</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>        <span class="n">agents</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">n_jobs</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>        <span class="k">if</span> <span class="n">agents</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>            <span class="n">agents</span> <span class="o">=</span> <span class="mi">1</span>
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>    <span class="k">elif</span> <span class="n">n_jobs</span> <span class="o">&gt;</span> <span class="n">cpu_count</span><span class="p">():</span>
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>        <span class="n">agents</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span>
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>    <span class="k">elif</span> <span class="n">n_jobs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<a id="__codelineno-0-25" name="__codelineno-0-25" href="#__codelineno-0-25"></a>        <span class="n">agents</span> <span class="o">=</span> <span class="mi">1</span>
<a id="__codelineno-0-26" name="__codelineno-0-26" href="#__codelineno-0-26"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27" href="#__codelineno-0-27"></a>        <span class="n">agents</span> <span class="o">=</span> <span class="n">n_jobs</span>
<a id="__codelineno-0-28" name="__codelineno-0-28" href="#__codelineno-0-28"></a>    <span class="k">return</span> <span class="n">agents</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h4 class="doc doc-heading" id="cell2cell.utils.parallel_computing.parallel_spatial_ccis">
<code class="highlight language-python"><span class="n">parallel_spatial_ccis</span><span class="p">(</span><span class="n">inputs</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Parallel computing in cell2cell2.analysis.pipelines.SpatialSingleCellInteractions</p>

        <details class="quote">
          <summary>Source code in <code>cell2cell/utils/parallel_computing.py</code></summary>
          <div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="k">def</span> <span class="nf">parallel_spatial_ccis</span><span class="p">(</span><span class="n">inputs</span><span class="p">):</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    <span class="sd">&#39;&#39;&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="sd">    Parallel computing in cell2cell2.analysis.pipelines.SpatialSingleCellInteractions</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="sd">    &#39;&#39;&#39;</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>    <span class="c1"># TODO: Implement this for enabling spatial analysis and compute interactions in parallel</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>    <span class="c1"># from cell2cell.core import spatial_operation</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>    <span class="c1">#results = spatial_operation()</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>    <span class="c1"># return results</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>    <span class="k">pass</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>




  </div>

    </div>

  </div>




  </div>

    </div>

  </div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href=".." class="btn btn-neutral float-left" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../tutorials/Toy-Example-BulkPipeline/" class="btn btn-neutral float-right" title="Cell-cell communication from bulk dataset">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/earmingol/cell2cell" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href=".." style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../tutorials/Toy-Example-BulkPipeline/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
